<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>MuSiC包根据scRNAseq预测Bulk细胞组成 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="转录组">
<meta name="description" content="
MuSiC(MUlti-Subject SIngle Cell deconvolution)是来自宾夕法尼亚大学Biostatistics, Epidemiology and Informatics系的Mingyao Li课题组于2019年发表于Nature Communication的一个工具R包，可根据单细胞转录组信息推测Bulk RNA-seq细胞组成。而后，该团队又于2022年在Briefing in bioinformatics发表了扩展版本MuSiC2，可以考虑更复杂的场景。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/074music%E5%8C%85%E6%A0%B9%E6%8D%AEscrnaseq%E9%A2%84%E6%B5%8Bbulk%E7%BB%86%E8%83%9E%E7%BB%84%E6%88%90/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/074music%E5%8C%85%E6%A0%B9%E6%8D%AEscrnaseq%E9%A2%84%E6%B5%8Bbulk%E7%BB%86%E8%83%9E%E7%BB%84%E6%88%90/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="MuSiC包根据scRNAseq预测Bulk细胞组成" />
<meta property="og:description" content="
MuSiC(MUlti-Subject SIngle Cell deconvolution)是来自宾夕法尼亚大学Biostatistics, Epidemiology and Informatics系的Mingyao Li课题组于2019年发表于Nature Communication的一个工具R包，可根据单细胞转录组信息推测Bulk RNA-seq细胞组成。而后，该团队又于2022年在Briefing in bioinformatics发表了扩展版本MuSiC2，可以考虑更复杂的场景。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/074music%E5%8C%85%E6%A0%B9%E6%8D%AEscrnaseq%E9%A2%84%E6%B5%8Bbulk%E7%BB%86%E8%83%9E%E7%BB%84%E6%88%90/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-03-26T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-03-26T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="MuSiC包根据scRNAseq预测Bulk细胞组成"/>
<meta name="twitter:description" content="
MuSiC(MUlti-Subject SIngle Cell deconvolution)是来自宾夕法尼亚大学Biostatistics, Epidemiology and Informatics系的Mingyao Li课题组于2019年发表于Nature Communication的一个工具R包，可根据单细胞转录组信息推测Bulk RNA-seq细胞组成。而后，该团队又于2022年在Briefing in bioinformatics发表了扩展版本MuSiC2，可以考虑更复杂的场景。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "MuSiC包根据scRNAseq预测Bulk细胞组成",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/074music%E5%8C%85%E6%A0%B9%E6%8D%AEscrnaseq%E9%A2%84%E6%B5%8Bbulk%E7%BB%86%E8%83%9E%E7%BB%84%E6%88%90/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "MuSiC包根据scRNAseq预测Bulk细胞组成",
  "name": "MuSiC包根据scRNAseq预测Bulk细胞组成",
  "description": " MuSiC(MUlti-Subject SIngle Cell deconvolution)是来自宾夕法尼亚大学Biostatistics, Epidemiology and Informatics系的Mingyao Li课题组于2019年发表于Nature Communication的一个工具R包，可根据单细胞转录组信息推测Bulk RNA-seq细胞组成。而后，该团队又于2022年在Briefing in bioinformatics发表了扩展版本MuSiC2，可以考虑更复杂的场景。\n",
  "keywords": [
    "转录组"
  ],
  "articleBody": " MuSiC(MUlti-Subject SIngle Cell deconvolution)是来自宾夕法尼亚大学Biostatistics, Epidemiology and Informatics系的Mingyao Li课题组于2019年发表于Nature Communication的一个工具R包，可根据单细胞转录组信息推测Bulk RNA-seq细胞组成。而后，该团队又于2022年在Briefing in bioinformatics发表了扩展版本MuSiC2，可以考虑更复杂的场景。\nPaper–MuSic：https://doi.org/10.1038/s41467-018-08023-x Paper–MuSic2：https://doi.org/10.1093/bib/bbac430 Github：https://xuranw.github.io/MuSiC/index.html 1、算法简介 1.1 MuSiC 如下是流程图，由于单细胞数据一般存在多个样本，MuSiC的核心是“marker gene consistency”，即基因的表达一致性体现在同种细胞类型间(cross cell)，也体现在不同样本间(cross subject)。在去卷积注释时针对这些基因会赋予较高的权重。 另一方面考虑到，作者认为Solid tissues often contain closely related cell types，即一般算法难以对这些类似的细胞进行区分。对此，MuSiC分两步预测。首先进行general prediction，然后针对每个cluster分别单独预测具体的细胞类型比例。 1.2 MuSiC2 MuSiC2算法在上述基础考虑到Bulk RNAseq的组间差异带来的影响。一般单细胞数据来自健康人组织，而Bulk RNAseq可能有健康，疾病等多组。 作者认为需要避免健康与疾病组间的差异基因表达对细胞类型去卷积产生的干扰，即选择stable genes对疾病组样本注释(删去组间细胞类型差异表达基因)。 2、R包用法 1 2 3 4 5 6 7 8 # 安装 # BiocManager::install(\"TOAST\") devtools::install_github('xuranw/MuSiC') ## 参考Github教程提前下载示例数据 # Bulk data为ExpressionSet格式 # scRNA-seq data为SingleCellExperiment format格式 # 表达矩阵均为原始read counts，不需要标准化 2.1 MuSiC简单用法 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 # (1) Bulk data GSE50244.bulk.eset = readRDS('./GSE50244bulkeset.rds') dim(GSE50244.bulk.eset) # Features Samples # 32581 89 bulk.mtx = exprs(GSE50244.bulk.eset) # (2) scRNA-seq data EMTAB.sce = readRDS('./EMTABsce_healthy.rds') dim(EMTAB.sce) # [1] 25453 1097 # 主要需要两列注释信息：样本名与细胞类型 table(EMTAB.sce$cellType, EMTAB.sce$sampleID) # 1 2 4 5 6 8 # acinar 4 20 3 80 3 2 # alpha 28 117 92 26 136 44 # beta 12 48 35 32 34 10 # co-expression 3 3 6 5 6 3 # delta 7 21 12 2 7 10 # ductal 4 19 14 67 8 23 # endothelial 1 1 8 0 1 2 # epsilon 0 1 3 1 0 0 # ...... # (3) Estimate cell type proportions music_pred_obj = music_prop(bulk.mtx = bulk.mtx, # bulk exp sc.sce = EMTAB.sce, # scRNAseq obj clusters = 'cellType', # cluster column samples = 'sampleID', # sample column select.ct = c('alpha', 'beta', 'delta', 'gamma', 'acinar', 'ductal'), verbose = F) names(music_pred_obj) # [1] \"Est.prop.weighted\" \"Est.prop.allgene\" \"Weight.gene\" \"r.squared.full\" \"Var.prop\" ## 前两者分别表示MuSiC与NNLS算法的结果：每个样本(行)对于每种细胞类型(列)的比例，并进行样本水平归一化 music_pred_obj$Est.prop.weighted[1:4,] # alpha beta delta gamma acinar ductal # Sub1 0.2079168 0.1746030 0.002203054 0.0015887041 0.09597685 0.5177117 # Sub2 0.6964173 0.1197928 0.003683287 0.0007471486 0.03078049 0.1485789 # Sub3 0.1913485 0.3521117 0.032037770 0.0003117540 0.19143687 0.2327533 # Sub4 0.2339358 0.4673282 0.009425930 0.0226075392 0.05385258 0.2128500 2.2 MuSiC分步预测 即针对细胞类型相似的Bulk组织类型（Solid tissues often contain closely related cell types）\n（1）预处理 1 2 3 4 5 6 7 8 9 10 11 12 13 14 Mouse.bulk.eset = readRDS('./Mousebulkeset.rds') Mouse.bulk.eset # 10 19033 Mousesub.sce = readRDS('./Mousesub_sce.rds') Mousesub.sce # 10000 16273 Mousesub.basis = music_basis(Mousesub.sce, clusters = 'cellType', samples = 'sampleID', select.ct = c('Endo', 'Podo', 'PT', 'LOH', 'DCT', 'CD-PC', 'CD-IC', 'Fib','Macro', 'Neutro','B lymph', 'T lymph', 'NK')) （2）根据单细胞的标签定义相似的细胞组 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 d \u003c- dist(t(log(Mousesub.basis$Disgn.mtx + 1e-6)), method = \"euclidean\") # d \u003c- dist(t(log(Mousesub.basis$M.theta + 1e-8)), method = \"euclidean\") hc1 \u003c- hclust(d, method = \"complete\" ) plot(hc1, cex = 0.6, hang = -1, main = 'Cluster log(Design Matrix)') clusters.type = list(C1 = 'Neutro', C2 = 'Podo', C3 = c('Endo', 'CD-PC', 'LOH', 'CD-IC', 'DCT', 'PT'), C4 = c('Macro', 'Fib', 'B lymph', 'NK', 'T lymph')) # 添加到单细胞数据中 cl.type = as.character(Mousesub.sce$cellType) for(cl in 1:length(clusters.type)){ cl.type[cl.type %in% clusters.type[[cl]]] = names(clusters.type)[cl] } Mousesub.sce$clusterType = factor(cl.type, levels = c(names(clusters.type), 'CD-Trans', 'Novel1', 'Novel2')) （3）鉴定每大组内的组间marker基因(marker genes within each group) 1 2 3 4 5 6 7 8 9 load('./IEmarkers.RData') IEmarkers = list(NULL, NULL, Epith.marker, Immune.marker) names(IEmarkers) = c('C1', 'C2', 'C3', 'C4') str(IEmarkers) # List of 4 # $ C1: NULL # $ C2: NULL # $ C3: chr [1:2614] \"Rp1\" \"Sox17\" \"1700034P13Rik\" \"Prex2\" ... # $ C4: chr [1:8126] \"Sox17\" \"Gm37988\" \"Gm26901\" \"Mybl1\" ... （4）Estimate cell type proportions 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 music_pred_obj = music_prop.cluster(bulk.mtx = exprs(Mouse.bulk.eset), sc.sce = Mousesub.sce, group.markers = IEmarkers, clusters = 'cellType', groups = 'clusterType', samples = 'sampleID', clusters.type = clusters.type) dim(music_pred_obj$Est.prop.weighted.cluster) # 10, 13 music_pred_obj$Est.prop.weighted.cluster[1:4, 1:4] # Neutro Podo Endo CD-PC # control.NA.27 0 0.009505281 0.04474493 0.02680222 # control.NA.30 0 0.021948575 0.04305809 0.02687187 # control.NA.39 0 0.005628454 0.04412269 0.02802505 # control.NAP.3 0 0.018769776 0.04305809 0.02655007 2.3 MuSiC2用法 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 # (1) 50个对照与50个疾病，共100个样本的Bulk data benchmark.eset = readRDS(\"./bulk-eset.rds\") benchmark.eset table(benchmark.eset$group) # healthy t2d # 50 50 # (2) 来自健康样本的scRNA-seq Data seger.sce = readRDS(\"./EMTABsce_healthy.rds\") seger.sce # (3) Estimate cell type proportions music2_pred_obj = music2_prop(bulk.control.mtx = bulk.control.mtx, bulk.case.mtx = bulk.case.mtx, sc.sce = seger.sce, clusters = 'cellType', samples = 'sampleID', select.ct = c('acinar','alpha','beta','delta','ductal','gamma'), n_resample=20, sample_prop=0.5, cutoff_c=0.05,cutoff_r=0.01) dim(music2_pred_obj$Est.prop) # [1] 100 6 # acinar alpha beta delta ductal gamma # 1 0.154031473 0.03854683 0.5124088 0.001545064 0.29346786 0.000000000 # 10 0.559249172 0.13773168 0.2220337 0.005554692 0.07290038 0.002530353 # 12 0.062448310 0.57696213 0.1321646 0.000000000 0.22842496 0.000000000 # 14 0.113631970 0.34670271 0.3154881 0.000000000 0.16988614 0.054291121 # 17 0.078927767 0.31450931 0.2647605 0.000000000 0.34180242 0.000000000 # 2 0.008209709 0.41839211 0.1390657 0.019564262 0.41476824 0.000000000 ",
  "wordCount" : "1799",
  "inLanguage": "en",
  "datePublished": "2023-03-26T00:00:00Z",
  "dateModified": "2023-03-26T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/074music%E5%8C%85%E6%A0%B9%E6%8D%AEscrnaseq%E9%A2%84%E6%B5%8Bbulk%E7%BB%86%E8%83%9E%E7%BB%84%E6%88%90/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      MuSiC包根据scRNAseq预测Bulk细胞组成
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-03-26 00:00:00 &#43;0000 UTC&#39;&gt;2023-03-26&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-03-26&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1799&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e7%ae%97%e6%b3%95%e7%ae%80%e4%bb%8b" aria-label="1、算法简介">1、算法简介</a><ul>
                            
                    <li>
                        <a href="#11-music" aria-label="1.1 MuSiC">1.1 MuSiC</a></li>
                    <li>
                        <a href="#12-music2" aria-label="1.2 MuSiC2">1.2 MuSiC2</a></li></ul>
                    </li>
                    <li>
                        <a href="#2r%e5%8c%85%e7%94%a8%e6%b3%95" aria-label="2、R包用法">2、R包用法</a><ul>
                            <ul>
                            
                    <li>
                        <a href="#21-music%e7%ae%80%e5%8d%95%e7%94%a8%e6%b3%95" aria-label="2.1 MuSiC简单用法">2.1 MuSiC简单用法</a></li>
                    <li>
                        <a href="#22-music%e5%88%86%e6%ad%a5%e9%a2%84%e6%b5%8b" aria-label="2.2 MuSiC分步预测">2.2 MuSiC分步预测</a></li>
                    <li>
                        <a href="#23-music2%e7%94%a8%e6%b3%95" aria-label="2.3 MuSiC2用法">2.3 MuSiC2用法</a>
                    </li>
                </ul>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>MuSiC(MUlti-Subject SIngle Cell deconvolution)是来自宾夕法尼亚大学Biostatistics, Epidemiology and Informatics系的Mingyao Li课题组于2019年发表于Nature Communication的一个工具R包，可根据单细胞转录组信息推测Bulk RNA-seq细胞组成。而后，该团队又于2022年在Briefing in bioinformatics发表了扩展版本MuSiC2，可以考虑更复杂的场景。</p>
<ul>
<li>Paper&ndash;MuSic：https://doi.org/10.1038/s41467-018-08023-x</li>
<li>Paper&ndash;MuSic2：https://doi.org/10.1093/bib/bbac430</li>
<li>Github：https://xuranw.github.io/MuSiC/index.html</li>
</ul></blockquote>
<h2 id="1算法简介">1、算法简介<a hidden class="anchor" aria-hidden="true" href="#1算法简介">#</a></h2>
<h3 id="11-music">1.1 MuSiC<a hidden class="anchor" aria-hidden="true" href="#11-music">#</a></h3>
<ul>
<li>如下是流程图，由于单细胞数据一般存在多个样本，MuSiC的核心是“marker gene consistency”，即基因的表达一致性体现在同种细胞类型间(cross cell)，也体现在不同样本间(cross subject)。在去卷积注释时针对这些基因会赋予较高的权重。</li>
<li>另一方面考虑到，作者认为Solid tissues often contain closely related cell types，即一般算法难以对这些类似的细胞进行区分。对此，MuSiC分两步预测。首先进行general prediction，然后针对每个cluster分别单独预测具体的细胞类型比例。</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230326181456787.png" alt="image-20230326181456787" style="zoom:67%;" />
<h3 id="12-music2">1.2 MuSiC2<a hidden class="anchor" aria-hidden="true" href="#12-music2">#</a></h3>
<ul>
<li>MuSiC2算法在上述基础考虑到Bulk RNAseq的组间差异带来的影响。一般单细胞数据来自健康人组织，而Bulk RNAseq可能有健康，疾病等多组。</li>
<li>作者认为需要避免健康与疾病组间的差异基因表达对细胞类型去卷积产生的干扰，即选择stable genes对疾病组样本注释(删去组间细胞类型差异表达基因)。</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230326182814771.png" alt="image-20230326182814771" style="zoom: 67%;" />
<h2 id="2r包用法">2、R包用法<a hidden class="anchor" aria-hidden="true" href="#2r包用法">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 安装</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># BiocManager::install(&#34;TOAST&#34;)</span>
</span></span><span style="display:flex;"><span>devtools::install_github(<span style="color:#0ff;font-weight:bold">&#39;xuranw/MuSiC&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 参考Github教程提前下载示例数据</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Bulk data为ExpressionSet格式</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># scRNA-seq data为SingleCellExperiment format格式</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 表达矩阵均为原始read counts，不需要标准化</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h4 id="21-music简单用法">2.1 MuSiC简单用法<a hidden class="anchor" aria-hidden="true" href="#21-music简单用法">#</a></h4>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># (1) Bulk data</span>
</span></span><span style="display:flex;"><span>GSE50244.bulk.eset = readRDS(<span style="color:#0ff;font-weight:bold">&#39;./GSE50244bulkeset.rds&#39;</span>)
</span></span><span style="display:flex;"><span>dim(GSE50244.bulk.eset)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Features  Samples </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 32581       89 </span>
</span></span><span style="display:flex;"><span>bulk.mtx = exprs(GSE50244.bulk.eset)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2) scRNA-seq data</span>
</span></span><span style="display:flex;"><span>EMTAB.sce = readRDS(<span style="color:#0ff;font-weight:bold">&#39;./EMTABsce_healthy.rds&#39;</span>)
</span></span><span style="display:flex;"><span>dim(EMTAB.sce)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 25453  1097</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 主要需要两列注释信息：样本名与细胞类型</span>
</span></span><span style="display:flex;"><span>table(EMTAB.sce$cellType, EMTAB.sce$sampleID)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                          1   2   4   5   6   8</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># acinar                   4  20   3  80   3   2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># alpha                   28 117  92  26 136  44</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># beta                    12  48  35  32  34  10</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># co-expression            3   3   6   5   6   3</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># delta                    7  21  12   2   7  10</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ductal                   4  19  14  67   8  23</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># endothelial              1   1   8   0   1   2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># epsilon                  0   1   3   1   0   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ......</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (3) Estimate cell type proportions</span>
</span></span><span style="display:flex;"><span>music_pred_obj = music_prop(bulk.mtx = bulk.mtx, <span style="color:#007f7f"># bulk exp</span>
</span></span><span style="display:flex;"><span>                            sc.sce = EMTAB.sce, <span style="color:#007f7f"># scRNAseq obj</span>
</span></span><span style="display:flex;"><span>                            clusters = <span style="color:#0ff;font-weight:bold">&#39;cellType&#39;</span>,  <span style="color:#007f7f"># cluster column</span>
</span></span><span style="display:flex;"><span>                            samples = <span style="color:#0ff;font-weight:bold">&#39;sampleID&#39;</span>,   <span style="color:#007f7f"># sample  column</span>
</span></span><span style="display:flex;"><span>                            select.ct = c(<span style="color:#0ff;font-weight:bold">&#39;alpha&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;beta&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;delta&#39;</span>, 
</span></span><span style="display:flex;"><span>                                          <span style="color:#0ff;font-weight:bold">&#39;gamma&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;acinar&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;ductal&#39;</span>), 
</span></span><span style="display:flex;"><span>                            verbose = <span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>names(music_pred_obj)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;Est.prop.weighted&#34; &#34;Est.prop.allgene&#34;  &#34;Weight.gene&#34;       &#34;r.squared.full&#34;    &#34;Var.prop&#34;      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 前两者分别表示MuSiC与NNLS算法的结果：每个样本(行)对于每种细胞类型(列)的比例，并进行样本水平归一化</span>
</span></span><span style="display:flex;"><span>music_pred_obj$Est.prop.weighted[1:<span style="color:#ff0;font-weight:bold">4</span>,]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#          alpha      beta       delta        gamma     acinar    ductal</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Sub1 0.2079168 0.1746030 0.002203054 0.0015887041 0.09597685 0.5177117</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Sub2 0.6964173 0.1197928 0.003683287 0.0007471486 0.03078049 0.1485789</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Sub3 0.1913485 0.3521117 0.032037770 0.0003117540 0.19143687 0.2327533</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Sub4 0.2339358 0.4673282 0.009425930 0.0226075392 0.05385258 0.2128500</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h4 id="22-music分步预测">2.2 MuSiC分步预测<a hidden class="anchor" aria-hidden="true" href="#22-music分步预测">#</a></h4>
<p>即针对细胞类型相似的Bulk组织类型（Solid tissues often contain closely related cell types）</p>
<ul>
<li>（1）预处理</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>Mouse.bulk.eset = readRDS(<span style="color:#0ff;font-weight:bold">&#39;./Mousebulkeset.rds&#39;</span>)
</span></span><span style="display:flex;"><span>Mouse.bulk.eset
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 10 19033</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>Mousesub.sce = readRDS(<span style="color:#0ff;font-weight:bold">&#39;./Mousesub_sce.rds&#39;</span>)
</span></span><span style="display:flex;"><span>Mousesub.sce
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 10000 16273</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>Mousesub.basis = music_basis(Mousesub.sce, 
</span></span><span style="display:flex;"><span>                             clusters = <span style="color:#0ff;font-weight:bold">&#39;cellType&#39;</span>, 
</span></span><span style="display:flex;"><span>                             samples = <span style="color:#0ff;font-weight:bold">&#39;sampleID&#39;</span>, 
</span></span><span style="display:flex;"><span>                             select.ct = c(<span style="color:#0ff;font-weight:bold">&#39;Endo&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;Podo&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;PT&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;LOH&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;DCT&#39;</span>, 
</span></span><span style="display:flex;"><span>                                           <span style="color:#0ff;font-weight:bold">&#39;CD-PC&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;CD-IC&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;Fib&#39;</span>,<span style="color:#0ff;font-weight:bold">&#39;Macro&#39;</span>, 
</span></span><span style="display:flex;"><span>                                           <span style="color:#0ff;font-weight:bold">&#39;Neutro&#39;</span>,<span style="color:#0ff;font-weight:bold">&#39;B lymph&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;T lymph&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;NK&#39;</span>))
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（2）根据单细胞的标签定义相似的细胞组</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>d &lt;- dist(t(log(Mousesub.basis$Disgn.mtx + <span style="color:#ff0;font-weight:bold">1e-6</span>)), method = <span style="color:#0ff;font-weight:bold">&#34;euclidean&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># d &lt;- dist(t(log(Mousesub.basis$M.theta + 1e-8)), method = &#34;euclidean&#34;)</span>
</span></span><span style="display:flex;"><span>hc1 &lt;- hclust(d, method = <span style="color:#0ff;font-weight:bold">&#34;complete&#34;</span> )
</span></span><span style="display:flex;"><span>plot(hc1, cex = <span style="color:#ff0;font-weight:bold">0.6</span>, hang = <span style="color:#ff0;font-weight:bold">-1</span>, main = <span style="color:#0ff;font-weight:bold">&#39;Cluster log(Design Matrix)&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>clusters.type = list(C1 = <span style="color:#0ff;font-weight:bold">&#39;Neutro&#39;</span>, 
</span></span><span style="display:flex;"><span>                     C2 = <span style="color:#0ff;font-weight:bold">&#39;Podo&#39;</span>, 
</span></span><span style="display:flex;"><span>                     C3 = c(<span style="color:#0ff;font-weight:bold">&#39;Endo&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;CD-PC&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;LOH&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;CD-IC&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;DCT&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;PT&#39;</span>), 
</span></span><span style="display:flex;"><span>                     C4 = c(<span style="color:#0ff;font-weight:bold">&#39;Macro&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;Fib&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;B lymph&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;NK&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;T lymph&#39;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 添加到单细胞数据中</span>
</span></span><span style="display:flex;"><span>cl.type = as.character(Mousesub.sce$cellType)
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span>(cl <span style="color:#fff;font-weight:bold">in</span> <span style="color:#ff0;font-weight:bold">1</span>:length(clusters.type)){
</span></span><span style="display:flex;"><span>  cl.type[cl.type %in% clusters.type[[cl]]] = names(clusters.type)[cl]
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>Mousesub.sce$clusterType = factor(cl.type, 
</span></span><span style="display:flex;"><span>                                  levels = c(names(clusters.type), <span style="color:#0ff;font-weight:bold">&#39;CD-Trans&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;Novel1&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;Novel2&#39;</span>))
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230326185057707.png" alt="image-20230326185057707" style="zoom:67%;" />
<ul>
<li>（3）鉴定每大组内的组间marker基因(marker genes within each group)</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>load(<span style="color:#0ff;font-weight:bold">&#39;./IEmarkers.RData&#39;</span>)
</span></span><span style="display:flex;"><span>IEmarkers = list(<span style="color:#fff;font-weight:bold">NULL</span>, <span style="color:#fff;font-weight:bold">NULL</span>, Epith.marker, Immune.marker)
</span></span><span style="display:flex;"><span>names(IEmarkers) = c(<span style="color:#0ff;font-weight:bold">&#39;C1&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;C2&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;C3&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;C4&#39;</span>)
</span></span><span style="display:flex;"><span>str(IEmarkers)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># List of 4</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  $ C1: NULL</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  $ C2: NULL</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  $ C3: chr [1:2614] &#34;Rp1&#34; &#34;Sox17&#34; &#34;1700034P13Rik&#34; &#34;Prex2&#34; ...</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  $ C4: chr [1:8126] &#34;Sox17&#34; &#34;Gm37988&#34; &#34;Gm26901&#34; &#34;Mybl1&#34; ...</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（4）Estimate cell type proportions</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>music_pred_obj = music_prop.cluster(bulk.mtx = exprs(Mouse.bulk.eset), 
</span></span><span style="display:flex;"><span>                                    sc.sce = Mousesub.sce, 
</span></span><span style="display:flex;"><span>                                    group.markers = IEmarkers, 
</span></span><span style="display:flex;"><span>                                    clusters = <span style="color:#0ff;font-weight:bold">&#39;cellType&#39;</span>, 
</span></span><span style="display:flex;"><span>                                    groups = <span style="color:#0ff;font-weight:bold">&#39;clusterType&#39;</span>, 
</span></span><span style="display:flex;"><span>                                    samples = <span style="color:#0ff;font-weight:bold">&#39;sampleID&#39;</span>, 
</span></span><span style="display:flex;"><span>                                    clusters.type = clusters.type)
</span></span><span style="display:flex;"><span>dim(music_pred_obj$Est.prop.weighted.cluster)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 10, 13</span>
</span></span><span style="display:flex;"><span>music_pred_obj$Est.prop.weighted.cluster[1:<span style="color:#ff0;font-weight:bold">4</span>, <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#               Neutro        Podo       Endo      CD-PC</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># control.NA.27      0 0.009505281 0.04474493 0.02680222</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># control.NA.30      0 0.021948575 0.04305809 0.02687187</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># control.NA.39      0 0.005628454 0.04412269 0.02802505</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># control.NAP.3      0 0.018769776 0.04305809 0.02655007</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h4 id="23-music2用法">2.3 MuSiC2用法<a hidden class="anchor" aria-hidden="true" href="#23-music2用法">#</a></h4>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># (1) 50个对照与50个疾病，共100个样本的Bulk data</span>
</span></span><span style="display:flex;"><span>benchmark.eset = readRDS(<span style="color:#0ff;font-weight:bold">&#34;./bulk-eset.rds&#34;</span>)
</span></span><span style="display:flex;"><span>benchmark.eset
</span></span><span style="display:flex;"><span>table(benchmark.eset$group)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># healthy     t2d </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#      50      50</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2) 来自健康样本的scRNA-seq Data</span>
</span></span><span style="display:flex;"><span>seger.sce = readRDS(<span style="color:#0ff;font-weight:bold">&#34;./EMTABsce_healthy.rds&#34;</span>)
</span></span><span style="display:flex;"><span>seger.sce
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (3) Estimate cell type proportions</span>
</span></span><span style="display:flex;"><span>music2_pred_obj = music2_prop(bulk.control.mtx = bulk.control.mtx, 
</span></span><span style="display:flex;"><span>                                           bulk.case.mtx = bulk.case.mtx, 
</span></span><span style="display:flex;"><span>                                           sc.sce = seger.sce, 
</span></span><span style="display:flex;"><span>                                           clusters = <span style="color:#0ff;font-weight:bold">&#39;cellType&#39;</span>, 
</span></span><span style="display:flex;"><span>                                           samples = <span style="color:#0ff;font-weight:bold">&#39;sampleID&#39;</span>, 
</span></span><span style="display:flex;"><span>                                           select.ct = c(<span style="color:#0ff;font-weight:bold">&#39;acinar&#39;</span>,<span style="color:#0ff;font-weight:bold">&#39;alpha&#39;</span>,<span style="color:#0ff;font-weight:bold">&#39;beta&#39;</span>,<span style="color:#0ff;font-weight:bold">&#39;delta&#39;</span>,<span style="color:#0ff;font-weight:bold">&#39;ductal&#39;</span>,<span style="color:#0ff;font-weight:bold">&#39;gamma&#39;</span>), 
</span></span><span style="display:flex;"><span>                                           n_resample=<span style="color:#ff0;font-weight:bold">20</span>, sample_prop=<span style="color:#ff0;font-weight:bold">0.5</span>,
</span></span><span style="display:flex;"><span>                                           cutoff_c=<span style="color:#ff0;font-weight:bold">0.05</span>,cutoff_r=<span style="color:#ff0;font-weight:bold">0.01</span>)
</span></span><span style="display:flex;"><span>dim(music2_pred_obj$Est.prop)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 100   6</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#         acinar      alpha      beta       delta     ductal       gamma</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1  0.154031473 0.03854683 0.5124088 0.001545064 0.29346786 0.000000000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 10 0.559249172 0.13773168 0.2220337 0.005554692 0.07290038 0.002530353</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 12 0.062448310 0.57696213 0.1321646 0.000000000 0.22842496 0.000000000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 14 0.113631970 0.34670271 0.3154881 0.000000000 0.16988614 0.054291121</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 17 0.078927767 0.31450931 0.2647605 0.000000000 0.34180242 0.000000000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2  0.008209709 0.41839211 0.1390657 0.019564262 0.41476824 0.000000000</span>
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">转录组</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/059survival%E5%8C%85%E7%94%9F%E5%AD%98%E5%88%86%E6%9E%90%E5%8F%8Aglmnet%E5%8C%85lasso%E5%9B%9E%E5%BD%92/">
    <span class="title">« Prev Page</span>
    <br>
    <span>survival包生存分析及glmnet包lasso回归</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/060%E4%BB%8Ernaseq%E7%9A%84fastq.gz%E6%8F%90%E5%8F%96%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5/">
    <span class="title">Next Page »</span>
    <br>
    <span>从RNAseq的fastq.gz提取表达矩阵</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
