<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--infercnv拷贝数变异鉴定 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
inferCNV包可根据肿瘤组织相关的单细胞表达数据，推测肿瘤细胞的拷贝数变异情况，从而完成恶性细胞的鉴定。


1
2
3


BiocManager::install(&#34;infercnv&#34;)
library(infercnv)
# 暂时在window安装出现点问题，目前在linux使用conda安装、学习。



">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/018%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--infercnv%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82%E9%89%B4%E5%AE%9A/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/018%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--infercnv%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82%E9%89%B4%E5%AE%9A/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--infercnv拷贝数变异鉴定" />
<meta property="og:description" content="
inferCNV包可根据肿瘤组织相关的单细胞表达数据，推测肿瘤细胞的拷贝数变异情况，从而完成恶性细胞的鉴定。


1
2
3


BiocManager::install(&#34;infercnv&#34;)
library(infercnv)
# 暂时在window安装出现点问题，目前在linux使用conda安装、学习。



" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/018%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--infercnv%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82%E9%89%B4%E5%AE%9A/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-01-20T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-01-20T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--infercnv拷贝数变异鉴定"/>
<meta name="twitter:description" content="
inferCNV包可根据肿瘤组织相关的单细胞表达数据，推测肿瘤细胞的拷贝数变异情况，从而完成恶性细胞的鉴定。


1
2
3


BiocManager::install(&#34;infercnv&#34;)
library(infercnv)
# 暂时在window安装出现点问题，目前在linux使用conda安装、学习。



"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--infercnv拷贝数变异鉴定",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/018%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--infercnv%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82%E9%89%B4%E5%AE%9A/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--infercnv拷贝数变异鉴定",
  "name": "单细胞分析工具--infercnv拷贝数变异鉴定",
  "description": " inferCNV包可根据肿瘤组织相关的单细胞表达数据，推测肿瘤细胞的拷贝数变异情况，从而完成恶性细胞的鉴定。\n1 2 3 BiocManager::install(\u0026#34;infercnv\u0026#34;) library(infercnv) # 暂时在window安装出现点问题，目前在linux使用conda安装、学习。 ",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " inferCNV包可根据肿瘤组织相关的单细胞表达数据，推测肿瘤细胞的拷贝数变异情况，从而完成恶性细胞的鉴定。\n1 2 3 BiocManager::install(\"infercnv\") library(infercnv) # 暂时在window安装出现点问题，目前在linux使用conda安装、学习。 1、背景原理 拷贝数变异是指指染色体上大于1 kb的DNA片段的扩增(amplification)或者减少(deletion)，对基因的表达有很大的影响（扩增/降低）。而肿瘤恶性细胞通常伴随着拷贝数变异，通过影响相关基因的表达促进肿瘤发生。 在肿瘤单细胞数据分析过程中，肿瘤细胞类型的注释可通过tumor related marker gene的表达情况(是否高表达)做出判断。而inferCNV可以从拷贝数变异的角度进一步验证肿瘤细胞类型的注释。 inferCNV的算法是在完成肿瘤微环境的细胞类型注释的基础之上，以“Normal”细胞的基因表达情况做对照，计算“tumor”-annotated 细胞中的某些染色体区域的基因表达是否发生明显的增多或减少，从而推测出细胞的拷贝数变异图谱（并可以进一步聚类），从而验证之前的注释结果。 inferCNV从计算步骤来说分为两大步：第一步根据Normal细胞对比，计算得到tumor-like细胞的CNV图谱（preliminary infercnv object）；然后第二步是可选项，包括降噪处理和HMM预测，可分别得到两种结果。 2、输入数据 infercnv分析主要需要三类数据，以R包自带的示例文件为例：\n（1）单细胞表达矩阵\n已完成预处理步骤(过滤低质量细胞等)的单细胞raw count表达矩阵，matrix矩阵格式即可。\n1 2 3 4 5 6 7 8 9 10 11 12 mat_dir = system.file(\"extdata\", \"oligodendroglioma_expression_downsampled.counts.matrix.gz\", package = \"infercnv\") raw_counts = data.table::fread(mat_dir, data.table = F) rownames(raw_counts) = raw_counts[,1] raw_counts = raw_counts[,-1] raw_counts[1:4,1:4] # MGH54_P16_F12 MGH54_P12_C10 MGH54_P11_C11 MGH54_P15_D06 # A2M 0 0.000 0.000 0.000 # A4GALT 0 0.000 0.000 0.000 # AAAS 0 37.008 30.935 21.011 # AACS 0 0.000 0.000 0.000 dim(raw_counts) # [1] 10338 184 （2）细胞类型注释文件 包含两列：第一列是细胞ID(对应表达矩阵的列名)，第二列是细胞类型注释结果； 至少需要包含两种细胞类型：已知正常的细胞类型（免疫细胞、内皮细胞..）、可能为肿瘤细胞的细胞类型（肿瘤细胞、上皮细胞、成纤维细胞…）； 此外由于肿瘤患者的异质性，不同病人来源的肿瘤细胞的拷贝数变异情况可能差别很大，因此可以在第二列的肿瘤细胞类型进行病源的注释，例如tumor_P1，tumor_P2表示分别来自病人P1、P2的肿瘤细胞。 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 anno_file_dir = system.file(\"extdata\", \"oligodendroglioma_annotations_downsampled.txt\", package = \"infercnv\") anno_file = data.table::fread(anno_file_dir, data.table = F, header = F) head(anno_file) dim(anno_file) anno_file$V2 = stringr::str_split(anno_file$V2,\"_\",simplify = T)[,1] head(anno_file) # V1 V2 # 1 MGH54_P2_C12 Microglia/Macrophage # 2 MGH36_P6_F03 Microglia/Macrophage # 3 MGH53_P4_H08 Microglia/Macrophage # 4 MGH53_P2_E09 Microglia/Macrophage # 5 MGH36_P5_E12 Microglia/Macrophage # 6 MGH54_P2_H07 Microglia/Macrophage write.table(anno_file, quote = F, row.names = F, col.names = F, sep = \"\\t\", file = \"anno_file.txt\") （3）基因坐标文件 包含四列，分别为：基因名(对应表达矩阵的行名)、染色体信息、起始位点、终止位点。 https://data.broadinstitute.org/Trinity/CTAT/cnv/ 提供有人类的基因坐标信息，其中genecode_19对应hg19/GRCh37，genecode_v21对应hg38/GRCh38 1 2 3 4 5 6 7 8 9 10 11 12 gene_order_file_dir = system.file(\"extdata\", \"gencode_downsampled.EXAMPLE_ONLY_DONT_REUSE.txt\", package = \"infercnv\") gene_order_file = data.table::fread(gene_order_file_dir, data.table = F, header = F) head(gene_order_file) # V1 V2 V3 V4 # 1 WASH7P chr1 14363 29806 # 2 LINC00115 chr1 761586 762902 # 3 NOC2L chr1 879584 894689 # 4 MIR200A chr1 1103243 1103332 # 5 SDF4 chr1 1152288 1167411 # 6 UBE2J2 chr1 1189289 1209265 write.table(gene_order_file, quote = F, row.names = F, col.names = F, sep = \"\\t\", file = \"gene_order_file.txt\") 3、示例分析 3.1 构建对象 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 ## 模板参数 infercnv_obj = CreateInfercnvObject( #原始count矩阵 raw_counts_matrix=raw_counts_matrix, #细胞类型注释信息 annotations_file=annotations_file, #对应annotations_file里，认为是normal细胞的细胞类型 ref_group_names=c(\"celltype1\",\"celltype2\") #基因坐标信息 gene_order_file=gene_order_file, #指定上述两个文件的分隔符 delim=\"\\t\") ## 示例分析 infercnv_obj = CreateInfercnvObject(raw_counts_matrix=raw_counts, annotations_file=\"anno_file.txt\", delim=\"\\t\", gene_order_file=\"gene_order_file.txt\", ref_group_names=c(\"Microglia/Macrophage\", \"Oligodendrocytes (non-malignant)\")) 3.2 变异分析 默认计算得到preliminary infercnv object，可分别设置参数交代是否进行进一步降噪(de-noising)或者CNV的HMM预测。\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 ## 模板参数 infercnv_obj = infercnv::run( #上一步构建的infercnv对象 infercnv_obj, #筛选基因的阈值：基因在所有细胞的平均表达量 # use 1 for smart-seq, 0.1 for 10x-genomics cutoff=1, #存储输出结果的文件夹名（每一步的中间文件都会保存） out_dir=\"output_dir\", #是否将肿瘤细胞按照病源(病人之间异质性)分群计算CNV图谱 cluster_by_groups=T, #是否降噪处理 denoise=T, #是否利用HMM算法预测CNV状态 HMM=T, #使用的线程数 num_threads = 8) ## 示例分析 infercnv_obj = infercnv::run(infercnv_obj, cutoff=1, out_dir=\"infer_out\", cluster_by_groups=TRUE, denoise=T, HMM=F) 3.3 结果解读 如下图为infercnv的分析结果可视化呈现，可分为3部分：上半部分热图、下半部分热图以及左上角的图例\n首先关于左上角的图例：（0，0.5，1，1.5，2）分别表示相对于Normal细胞的染色体区域基因表达量的倍数，红色表示该区域基因量相对增多，蓝色表示该区域基因量相对减少。柱子的长度表示对应区域的多少； 上半部分的热图：表示指定为Normal细胞的表达分布情况，正常情况下应该都是白色，没有明显集中的CNV区域； 下半部分的热图：相对于上半部分的Normal cell，计算的得到的每个tumor-like细胞的CNV图谱；然后根据所有细胞的相似性进行树状图聚类。 ",
  "wordCount" : "1907",
  "inLanguage": "en",
  "datePublished": "2023-01-20T00:00:00Z",
  "dateModified": "2023-01-20T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/018%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--infercnv%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82%E9%89%B4%E5%AE%9A/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--infercnv拷贝数变异鉴定
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-20 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-20&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-20&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1907&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e8%83%8c%e6%99%af%e5%8e%9f%e7%90%86" aria-label="1、背景原理">1、背景原理</a></li>
                    <li>
                        <a href="#2%e8%be%93%e5%85%a5%e6%95%b0%e6%8d%ae" aria-label="2、输入数据">2、输入数据</a></li>
                    <li>
                        <a href="#3%e7%a4%ba%e4%be%8b%e5%88%86%e6%9e%90" aria-label="3、示例分析">3、示例分析</a><ul>
                            
                    <li>
                        <a href="#31-%e6%9e%84%e5%bb%ba%e5%af%b9%e8%b1%a1" aria-label="3.1 构建对象">3.1 构建对象</a></li>
                    <li>
                        <a href="#32-%e5%8f%98%e5%bc%82%e5%88%86%e6%9e%90" aria-label="3.2 变异分析">3.2 变异分析</a></li>
                    <li>
                        <a href="#33-%e7%bb%93%e6%9e%9c%e8%a7%a3%e8%af%bb" aria-label="3.3 结果解读">3.3 结果解读</a>
                    </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>inferCNV包可根据肿瘤组织相关的单细胞表达数据，推测肿瘤细胞的拷贝数变异情况，从而完成恶性细胞的鉴定。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>BiocManager::install(<span style="color:#0ff;font-weight:bold">&#34;infercnv&#34;</span>)
</span></span><span style="display:flex;"><span>library(infercnv)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 暂时在window安装出现点问题，目前在linux使用conda安装、学习。</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/infercnv_logo.png" alt="InferCNV"  />
</p></blockquote>
<h2 id="1背景原理">1、背景原理<a hidden class="anchor" aria-hidden="true" href="#1背景原理">#</a></h2>
<ul>
<li>拷贝数变异是指指染色体上大于1 kb的DNA片段的扩增(amplification)或者减少(deletion)，对基因的表达有很大的影响（扩增/降低）。而肿瘤恶性细胞通常伴随着拷贝数变异，通过影响相关基因的表达促进肿瘤发生。</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-9c44a6b872bce4a3.png" alt="Copy Number alterations. WT cell, since diploid organisms, carry two&hellip; |  Download Scientific Diagram"  />
</p>
<ul>
<li>在肿瘤单细胞数据分析过程中，肿瘤细胞类型的注释可通过tumor related marker gene的表达情况(是否高表达)做出判断。而<strong>inferCNV可以从拷贝数变异的角度进一步验证肿瘤细胞类型的注释</strong>。</li>
<li>inferCNV的算法是在完成肿瘤微环境的细胞类型注释的基础之上，<strong>以“Normal”细胞的基因表达情况做对照，计算“tumor”-annotated 细胞中的某些染色体区域的基因表达是否发生明显的增多或减少</strong>，从而推测出细胞的拷贝数变异图谱（并可以进一步聚类），从而验证之前的注释结果。</li>
<li>inferCNV从计算步骤来说分为两大步：第一步根据Normal细胞对比，计算得到tumor-like细胞的CNV图谱（preliminary infercnv object）；然后第二步是可选项，包括降噪处理和HMM预测，可分别得到两种结果。</li>
</ul>
<img src="https://github.com/broadinstitute/inferCNV/wiki/images/InferCNV_procedure.png" alt="img" style="zoom: 67%;" />
<h2 id="2输入数据">2、输入数据<a hidden class="anchor" aria-hidden="true" href="#2输入数据">#</a></h2>
<p>infercnv分析主要需要三类数据，以R包自带的示例文件为例：</p>
<ul>
<li>
<p>（1）单细胞表达矩阵</p>
<p>已完成预处理步骤(过滤低质量细胞等)的单细胞raw count表达矩阵，matrix矩阵格式即可。</p>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>mat_dir = system.file(<span style="color:#0ff;font-weight:bold">&#34;extdata&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;oligodendroglioma_expression_downsampled.counts.matrix.gz&#34;</span>, package = <span style="color:#0ff;font-weight:bold">&#34;infercnv&#34;</span>)
</span></span><span style="display:flex;"><span>raw_counts = data.table::fread(mat_dir, data.table = <span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>rownames(raw_counts) = raw_counts[,<span style="color:#ff0;font-weight:bold">1</span>]
</span></span><span style="display:flex;"><span>raw_counts = raw_counts[,<span style="color:#ff0;font-weight:bold">-1</span>]
</span></span><span style="display:flex;"><span>raw_counts[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        MGH54_P16_F12 MGH54_P12_C10 MGH54_P11_C11 MGH54_P15_D06</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># A2M                0         0.000         0.000         0.000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># A4GALT             0         0.000         0.000         0.000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAAS               0        37.008        30.935        21.011</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AACS               0         0.000         0.000         0.000</span>
</span></span><span style="display:flex;"><span>dim(raw_counts)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 10338   184</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（2）细胞类型注释文件
<ul>
<li>包含两列：第一列是细胞ID(对应表达矩阵的列名)，第二列是细胞类型注释结果；</li>
<li>至少需要包含两种细胞类型：已知正常的细胞类型（免疫细胞、内皮细胞..）、可能为肿瘤细胞的细胞类型（肿瘤细胞、上皮细胞、成纤维细胞&hellip;）；</li>
<li>此外由于肿瘤患者的异质性，不同病人来源的肿瘤细胞的拷贝数变异情况可能差别很大，因此可以在第二列的肿瘤细胞类型进行病源的注释，例如<code>tumor_P1</code>，<code>tumor_P2</code>表示分别来自病人P1、P2的肿瘤细胞。</li>
</ul>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>anno_file_dir = system.file(<span style="color:#0ff;font-weight:bold">&#34;extdata&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;oligodendroglioma_annotations_downsampled.txt&#34;</span>, package = <span style="color:#0ff;font-weight:bold">&#34;infercnv&#34;</span>)
</span></span><span style="display:flex;"><span>anno_file = data.table::fread(anno_file_dir, data.table = <span style="color:#fff;font-weight:bold">F</span>, header = <span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>head(anno_file)
</span></span><span style="display:flex;"><span>dim(anno_file)
</span></span><span style="display:flex;"><span>anno_file$V2 = stringr::str_split(anno_file$V2,<span style="color:#0ff;font-weight:bold">&#34;_&#34;</span>,simplify = <span style="color:#fff;font-weight:bold">T</span>)[,<span style="color:#ff0;font-weight:bold">1</span>]
</span></span><span style="display:flex;"><span>head(anno_file)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#             V1                   V2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 MGH54_P2_C12 Microglia/Macrophage</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 MGH36_P6_F03 Microglia/Macrophage</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 MGH53_P4_H08 Microglia/Macrophage</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 MGH53_P2_E09 Microglia/Macrophage</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5 MGH36_P5_E12 Microglia/Macrophage</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6 MGH54_P2_H07 Microglia/Macrophage</span>
</span></span><span style="display:flex;"><span>write.table(anno_file, quote = <span style="color:#fff;font-weight:bold">F</span>, row.names = <span style="color:#fff;font-weight:bold">F</span>, col.names = <span style="color:#fff;font-weight:bold">F</span>,
</span></span><span style="display:flex;"><span>            sep = <span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, file = <span style="color:#0ff;font-weight:bold">&#34;anno_file.txt&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（3）基因坐标文件
<ul>
<li>包含四列，分别为：基因名(对应表达矩阵的行名)、染色体信息、起始位点、终止位点。</li>
<li><a href="https://data.broadinstitute.org/Trinity/CTAT/cnv/">https://data.broadinstitute.org/Trinity/CTAT/cnv/</a> 提供有人类的基因坐标信息，其中genecode_19对应hg19/GRCh37，genecode_v21对应hg38/GRCh38</li>
</ul>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>gene_order_file_dir = system.file(<span style="color:#0ff;font-weight:bold">&#34;extdata&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;gencode_downsampled.EXAMPLE_ONLY_DONT_REUSE.txt&#34;</span>, package = <span style="color:#0ff;font-weight:bold">&#34;infercnv&#34;</span>)
</span></span><span style="display:flex;"><span>gene_order_file = data.table::fread(gene_order_file_dir, data.table = <span style="color:#fff;font-weight:bold">F</span>, header = <span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>head(gene_order_file)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#          V1   V2      V3      V4</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1    WASH7P chr1   14363   29806</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 LINC00115 chr1  761586  762902</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3     NOC2L chr1  879584  894689</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4   MIR200A chr1 1103243 1103332</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5      SDF4 chr1 1152288 1167411</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6    UBE2J2 chr1 1189289 1209265</span>
</span></span><span style="display:flex;"><span>write.table(gene_order_file, quote = <span style="color:#fff;font-weight:bold">F</span>, row.names = <span style="color:#fff;font-weight:bold">F</span>, col.names = <span style="color:#fff;font-weight:bold">F</span>,
</span></span><span style="display:flex;"><span>            sep = <span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, file = <span style="color:#0ff;font-weight:bold">&#34;gene_order_file.txt&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="3示例分析">3、示例分析<a hidden class="anchor" aria-hidden="true" href="#3示例分析">#</a></h2>
<h3 id="31-构建对象">3.1 构建对象<a hidden class="anchor" aria-hidden="true" href="#31-构建对象">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## 模板参数</span>
</span></span><span style="display:flex;"><span>infercnv_obj = CreateInfercnvObject(
</span></span><span style="display:flex;"><span>						<span style="color:#007f7f">#原始count矩阵</span>
</span></span><span style="display:flex;"><span>						raw_counts_matrix=raw_counts_matrix,
</span></span><span style="display:flex;"><span>    					<span style="color:#007f7f">#细胞类型注释信息</span>
</span></span><span style="display:flex;"><span>						annotations_file=annotations_file,
</span></span><span style="display:flex;"><span>    					<span style="color:#007f7f">#对应annotations_file里，认为是normal细胞的细胞类型</span>
</span></span><span style="display:flex;"><span>						ref_group_names=c(<span style="color:#0ff;font-weight:bold">&#34;celltype1&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;celltype2&#34;</span>)
</span></span><span style="display:flex;"><span>    					<span style="color:#007f7f">#基因坐标信息</span>
</span></span><span style="display:flex;"><span>						gene_order_file=gene_order_file,
</span></span><span style="display:flex;"><span>    					<span style="color:#007f7f">#指定上述两个文件的分隔符</span>
</span></span><span style="display:flex;"><span>						delim=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 示例分析</span>
</span></span><span style="display:flex;"><span>infercnv_obj = CreateInfercnvObject(raw_counts_matrix=raw_counts,
</span></span><span style="display:flex;"><span>                                    annotations_file=<span style="color:#0ff;font-weight:bold">&#34;anno_file.txt&#34;</span>,
</span></span><span style="display:flex;"><span>                                    delim=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>,
</span></span><span style="display:flex;"><span>                                    gene_order_file=<span style="color:#0ff;font-weight:bold">&#34;gene_order_file.txt&#34;</span>,
</span></span><span style="display:flex;"><span>                                    ref_group_names=c(<span style="color:#0ff;font-weight:bold">&#34;Microglia/Macrophage&#34;</span>,
</span></span><span style="display:flex;"><span>                                                      <span style="color:#0ff;font-weight:bold">&#34;Oligodendrocytes (non-malignant)&#34;</span>)) 
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="32-变异分析">3.2 变异分析<a hidden class="anchor" aria-hidden="true" href="#32-变异分析">#</a></h3>
<p>默认计算得到<code>preliminary infercnv object</code>，可分别设置参数交代是否进行进一步降噪(de-noising)或者CNV的HMM预测。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## 模板参数</span>
</span></span><span style="display:flex;"><span>infercnv_obj = infercnv::run(
</span></span><span style="display:flex;"><span>						<span style="color:#007f7f">#上一步构建的infercnv对象</span>
</span></span><span style="display:flex;"><span>						infercnv_obj,
</span></span><span style="display:flex;"><span>    					<span style="color:#007f7f">#筛选基因的阈值：基因在所有细胞的平均表达量</span>
</span></span><span style="display:flex;"><span>    					<span style="color:#007f7f"># use 1 for smart-seq, 0.1 for 10x-genomics</span>
</span></span><span style="display:flex;"><span>    					cutoff=<span style="color:#ff0;font-weight:bold">1</span>,  
</span></span><span style="display:flex;"><span>    					<span style="color:#007f7f">#存储输出结果的文件夹名（每一步的中间文件都会保存）</span>
</span></span><span style="display:flex;"><span>                        out_dir=<span style="color:#0ff;font-weight:bold">&#34;output_dir&#34;</span>, 
</span></span><span style="display:flex;"><span>    					<span style="color:#007f7f">#是否将肿瘤细胞按照病源(病人之间异质性)分群计算CNV图谱</span>
</span></span><span style="display:flex;"><span>    					cluster_by_groups=<span style="color:#fff;font-weight:bold">T</span>, 
</span></span><span style="display:flex;"><span>    					<span style="color:#007f7f">#是否降噪处理</span>
</span></span><span style="display:flex;"><span>    					denoise=<span style="color:#fff;font-weight:bold">T</span>,
</span></span><span style="display:flex;"><span>    					<span style="color:#007f7f">#是否利用HMM算法预测CNV状态</span>
</span></span><span style="display:flex;"><span>    					HMM=<span style="color:#fff;font-weight:bold">T</span>,
</span></span><span style="display:flex;"><span>						<span style="color:#007f7f">#使用的线程数</span>
</span></span><span style="display:flex;"><span>						num_threads = <span style="color:#ff0;font-weight:bold">8</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 示例分析</span>
</span></span><span style="display:flex;"><span>infercnv_obj = infercnv::run(infercnv_obj,
</span></span><span style="display:flex;"><span>                             cutoff=<span style="color:#ff0;font-weight:bold">1</span>, 
</span></span><span style="display:flex;"><span>                             out_dir=<span style="color:#0ff;font-weight:bold">&#34;infer_out&#34;</span>,
</span></span><span style="display:flex;"><span>                             cluster_by_groups=<span style="color:#fff;font-weight:bold">TRUE</span>,
</span></span><span style="display:flex;"><span>                             denoise=<span style="color:#fff;font-weight:bold">T</span>,
</span></span><span style="display:flex;"><span>                             HMM=<span style="color:#fff;font-weight:bold">F</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="33-结果解读">3.3 结果解读<a hidden class="anchor" aria-hidden="true" href="#33-结果解读">#</a></h3>
<p>如下图为infercnv的分析结果可视化呈现，可分为3部分：上半部分热图、下半部分热图以及左上角的图例</p>
<ul>
<li>首先关于左上角的图例：（0，0.5，1，1.5，2）分别表示相对于Normal细胞的染色体区域基因表达量的倍数，红色表示该区域基因量相对增多，蓝色表示该区域基因量相对减少。柱子的长度表示对应区域的多少；</li>
<li>上半部分的热图：表示指定为Normal细胞的表达分布情况，正常情况下应该都是白色，没有明显集中的CNV区域；</li>
<li>下半部分的热图：相对于上半部分的Normal cell，计算的得到的每个tumor-like细胞的CNV图谱；然后根据所有细胞的相似性进行树状图聚类。</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230120220503723.png" alt="image-20230120220503723" style="zoom: 67%;" />

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/022%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--singlecellexperiment%E5%AF%B9%E8%B1%A1%E6%A0%BC%E5%BC%8F/">
    <span class="title">« Prev Page</span>
    <br>
    <span>单细胞分析工具--SingleCellExperiment对象格式</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/023%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scanpy%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/">
    <span class="title">Next Page »</span>
    <br>
    <span>单细胞分析工具--scanpy基础用法(anndata,loom格式)</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
