<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--SingleCellExperiment对象格式 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
SingleCellExperiment是通过SingleCellExperiment包创建的单细胞数据分析对象，已有几十个单细胞R包支持。
其衍生自SummarizedExperiment，之前在GEO数据挖掘学习时，了解过相关知识，主要是assay与pData两个函数的使用。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/022%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--singlecellexperiment%E5%AF%B9%E8%B1%A1%E6%A0%BC%E5%BC%8F/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/022%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--singlecellexperiment%E5%AF%B9%E8%B1%A1%E6%A0%BC%E5%BC%8F/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--SingleCellExperiment对象格式" />
<meta property="og:description" content="
SingleCellExperiment是通过SingleCellExperiment包创建的单细胞数据分析对象，已有几十个单细胞R包支持。
其衍生自SummarizedExperiment，之前在GEO数据挖掘学习时，了解过相关知识，主要是assay与pData两个函数的使用。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/022%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--singlecellexperiment%E5%AF%B9%E8%B1%A1%E6%A0%BC%E5%BC%8F/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-01-22T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-01-22T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--SingleCellExperiment对象格式"/>
<meta name="twitter:description" content="
SingleCellExperiment是通过SingleCellExperiment包创建的单细胞数据分析对象，已有几十个单细胞R包支持。
其衍生自SummarizedExperiment，之前在GEO数据挖掘学习时，了解过相关知识，主要是assay与pData两个函数的使用。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--SingleCellExperiment对象格式",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/022%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--singlecellexperiment%E5%AF%B9%E8%B1%A1%E6%A0%BC%E5%BC%8F/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--SingleCellExperiment对象格式",
  "name": "单细胞分析工具--SingleCellExperiment对象格式",
  "description": " SingleCellExperiment是通过SingleCellExperiment包创建的单细胞数据分析对象，已有几十个单细胞R包支持。\n其衍生自SummarizedExperiment，之前在GEO数据挖掘学习时，了解过相关知识，主要是assay与pData两个函数的使用。\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " SingleCellExperiment是通过SingleCellExperiment包创建的单细胞数据分析对象，已有几十个单细胞R包支持。\n其衍生自SummarizedExperiment，之前在GEO数据挖掘学习时，了解过相关知识，主要是assay与pData两个函数的使用。\n参考文章：https://osca.bioconductor.org/data-infrastructure.html 0、创建sce 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 library(tidyverse) library(SingleCellExperiment) ## 表达矩阵 # https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE81861\u0026format=file\u0026file=GSE81861%5FCRC%5FNM%5Fepithelial%5Fcells%5FCOUNT%2Ecsv%2Egz nm_count \u003c- read.csv(\"GSE81861_CRC_NM_epithelial_cells_COUNT.csv\") %\u003e% tibble::column_to_rownames(\"X\") %\u003e% as.matrix() rownames(nm_count) \u003c- sapply(strsplit(rownames(nm_count),\"_\"),\"[\",2) colnames(nm_count) \u003c- sapply(strsplit(colnames(nm_count),\"_\"),\"[\",1) dim(nm_count) # [1] 57241 160 nm_count[1:4,1:4] # RHC4104 RHC6087 RHL2880 RHC5949 # TSPAN6 0 428 66 141 # TNMD 0 0 0 0 # DPM1 0 179 0 1 # SCYL3 465 0 0 0 ## metadata group.dat \u003c- data.frame(group=rep(\"nomal\",ncol(nm_count))) ## 创建sce sce \u003c- SingleCellExperiment(assays = list(counts = nm_count), colData = group.dat) sce 1、assay结构 assays 中最基本的是列为sample，行为feature（一般是gene）的表达矩阵。如上代码，创建了名为counts的基础表达矩阵； 后续可以根据需要，增添标准化、归一化等行列名不变的矩阵，相当于在count矩阵基础上“新盖的几层楼”。 1.1 查看矩阵 1 2 3 4 5 6 #根据矩阵名（楼层名）选择查看 assay(sce, \"counts\")[1:4,1:4] #或者如下，但我觉得初学者使用第一种方法更能适应对象的结构 #counts(sce)[1:4,1:4] assays(sce) #查看所有的层楼名，目前只有一个 1.2 新添矩阵 一般创建对象时都是引入counts矩阵。根据此基础矩阵可新建相关标准化等矩阵。\n手动计算添加 1 2 3 4 5 6 7 8 counts \u003c- assay(sce, \"counts\") libsizes \u003c- colSums(counts) size.factors \u003c- libsizes/mean(libsizes) assay(sce, \"logcounts\") \u003c- log2(t(t(counts)/size.factors) + 1) assay(sce, \"logcounts\")[1:4,1:4] assays(sce) # List of length 2 # names(2): counts logcounts 利用相关R包可自动添加 1 2 3 4 5 #注意下面自动添加的log标准化矩阵与上面的logcounts一致，可不运行 sce \u003c- scater::logNormCounts(sce) sce assays(sce) #注意是assays，不是assay assay(sce, \"logcounts\")[1:4,1:4] 关于assays的命名，我们可以随意命名。为了能够命名有意义并且配合R包之间的接口，官方给了如下的命名建议。 counts: Raw count data, e.g., number of reads or transcripts for a particular gene. normcounts: Normalized values on the same scale as the original counts. For example, counts divided by cell-specific size factors that are centred at unity. logcounts: Log-transformed counts or count-like values. In most cases, this will be defined as log-transformed normcounts, e.g., using log base 2 and a pseudo-count of 1. cpm: Counts-per-million. This is the read count for each gene in each cell, divided by the library size of each cell in millions. tpm: Transcripts-per-million. This is the number of transcripts for each gene in each cell, divided by the total number of transcripts in that cell (in millions).\n2、colData结构 本质上为行名为sample的data.frame，即为细胞添加注释信息 如上，创建对象时，引入了sample的group分组信息 1 2 colData(sce) %\u003e% head() table(sce$group) 2.1 新增colData $符可用于查看colData，也能用于新增。（这一点等同于dataframe） 1 2 sce$anything \u003c- runif(ncol(sce)) colnames(colData(sce)) scater包自动添加 1 2 sce \u003c- scater::addPerCellQC(sce) colData(sce)[, 1:5] 2.2 筛选sample 1 2 sce[, sce$detected \u003e 3000] sce #原来是160个sample 3、rowData结构 参考colData，本质上为行名为feature的data.frame，即为基因添加注释信息。\n可以在创建对象时添加(基因的不同ID)，或者利用scater包添加注释。 1 sce \u003c- scater::addPerFeatureQC(sce) 手动添加可用cbind函数 1 2 rowData(x) \u003c- cbind(rowData(x), feature) #同理colData也类似，但使用$符也很方便 筛选feature 1 2 sce[rownames(nm_count)[c(1:10)], ] sce[1:10, ] Furthermore, there is a special rowRanges slot to hold genomic coordinates in the form of a GRanges or GRangesList. This stores describes the chromosome, start, and end coordinates of the features (genes, genomic regions) in a manner that is easy to query and manipulate via the GenomicRanges framework.\n4、reducedDims 上面三点基本源于SummarizedExperiment对象，但reducedDims是sce对象独有的一点，一般储存PCA，tSNE，uMAP等细胞降维聚类信息； 本质是行名为sample的一组data.frame 1 2 3 4 5 6 7 8 9 sce \u003c- scater::logNormCounts(sce) sce \u003c- scater::runPCA(sce) reducedDim(sce, \"PCA\") reducedDim(sce, \"PCA\")[1:4,1:4] # PC1 PC2 PC3 PC4 #RHC4104 46.66 -46.44 -9.488 1.79 #RHC6087 -54.38 -6.15 -0.155 13.16 #RHL2880 -7.89 -38.89 -1.319 -2.90 #RHC5949 -53.27 -20.88 -12.089 -7.73 1 2 3 4 5 6 7 8 sce \u003c- scater::runTSNE(sce, perplexity = 0.1) reducedDim(sce, \"TSNE\") #可用于绘图的二维坐标信息 reducedDims(sce) #类比assays u \u003c- uwot::umap(t(logcounts(sce)), n_neighbors = 2) reducedDim(sce, \"UMAP_uwot\") \u003c- u reducedDims(sce) # Now stored in the object. reducedDim(sce, \"UMAP_uwot\") #可用于绘图的二维坐标信息 ",
  "wordCount" : "1462",
  "inLanguage": "en",
  "datePublished": "2023-01-22T00:00:00Z",
  "dateModified": "2023-01-22T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/022%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--singlecellexperiment%E5%AF%B9%E8%B1%A1%E6%A0%BC%E5%BC%8F/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--SingleCellExperiment对象格式
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-22 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-22&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-22&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1462&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#0%e5%88%9b%e5%bb%basce" aria-label="0、创建sce">0、创建sce</a></li>
                    <li>
                        <a href="#1assay%e7%bb%93%e6%9e%84" aria-label="1、assay结构">1、assay结构</a><ul>
                            
                    <li>
                        <a href="#11-%e6%9f%a5%e7%9c%8b%e7%9f%a9%e9%98%b5" aria-label="1.1 查看矩阵">1.1 查看矩阵</a></li>
                    <li>
                        <a href="#12-%e6%96%b0%e6%b7%bb%e7%9f%a9%e9%98%b5" aria-label="1.2 新添矩阵">1.2 新添矩阵</a></li></ul>
                    </li>
                    <li>
                        <a href="#2coldata%e7%bb%93%e6%9e%84" aria-label="2、colData结构">2、colData结构</a><ul>
                            
                    <li>
                        <a href="#21-%e6%96%b0%e5%a2%9ecoldata" aria-label="2.1 新增colData">2.1 新增colData</a></li>
                    <li>
                        <a href="#22-%e7%ad%9b%e9%80%89sample" aria-label="2.2 筛选sample">2.2 筛选sample</a></li></ul>
                    </li>
                    <li>
                        <a href="#3rowdata%e7%bb%93%e6%9e%84" aria-label="3、rowData结构">3、rowData结构</a></li>
                    <li>
                        <a href="#4reduceddims" aria-label="4、reducedDims">4、reducedDims</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p><code>SingleCellExperiment</code>是通过SingleCellExperiment包创建的单细胞数据分析对象，已有几十个单细胞R包支持。</p>
<p>其衍生自<a href="https://bioconductor.org/packages/3.11/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">SummarizedExperiment</a>，之前在GEO数据挖掘学习时，了解过相关知识，主要是<code>assay</code>与<code>pData</code>两个函数的使用。</p>
<ul>
<li>参考文章：https://osca.bioconductor.org/data-infrastructure.html</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-e3eecf59ed4d32eb.png" alt="SingleCellExperiment" style="zoom:67%;" /></blockquote>
<h2 id="0创建sce">0、创建sce<a hidden class="anchor" aria-hidden="true" href="#0创建sce">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>library(SingleCellExperiment)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 表达矩阵</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE81861&amp;format=file&amp;file=GSE81861%5FCRC%5FNM%5Fepithelial%5Fcells%5FCOUNT%2Ecsv%2Egz</span>
</span></span><span style="display:flex;"><span>nm_count &lt;- read.csv(<span style="color:#0ff;font-weight:bold">&#34;GSE81861_CRC_NM_epithelial_cells_COUNT.csv&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  tibble::column_to_rownames(<span style="color:#0ff;font-weight:bold">&#34;X&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  as.matrix()
</span></span><span style="display:flex;"><span>rownames(nm_count) &lt;- sapply(strsplit(rownames(nm_count),<span style="color:#0ff;font-weight:bold">&#34;_&#34;</span>),<span style="color:#0ff;font-weight:bold">&#34;[&#34;</span>,<span style="color:#ff0;font-weight:bold">2</span>)
</span></span><span style="display:flex;"><span>colnames(nm_count) &lt;- sapply(strsplit(colnames(nm_count),<span style="color:#0ff;font-weight:bold">&#34;_&#34;</span>),<span style="color:#0ff;font-weight:bold">&#34;[&#34;</span>,<span style="color:#ff0;font-weight:bold">1</span>)
</span></span><span style="display:flex;"><span>dim(nm_count)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 57241   160</span>
</span></span><span style="display:flex;"><span>nm_count[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        RHC4104 RHC6087 RHL2880 RHC5949</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TSPAN6       0     428      66     141</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TNMD         0       0       0       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DPM1         0     179       0       1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SCYL3      465       0       0       0</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## metadata</span>
</span></span><span style="display:flex;"><span>group.dat &lt;- data.frame(group=rep(<span style="color:#0ff;font-weight:bold">&#34;nomal&#34;</span>,ncol(nm_count)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 创建sce</span>
</span></span><span style="display:flex;"><span>sce &lt;- SingleCellExperiment(assays = list(counts = nm_count),
</span></span><span style="display:flex;"><span>                            colData = group.dat)
</span></span><span style="display:flex;"><span>sce
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-3bb99b9482f4ee5e.png" alt="0"  />
</p>
<h2 id="1assay结构">1、assay结构<a hidden class="anchor" aria-hidden="true" href="#1assay结构">#</a></h2>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-22c0ea289c7d8785.png" alt="1-1"  />
</p>
<ul>
<li><code>assays</code> 中最基本的是列为sample，行为feature（一般是gene）的表达矩阵。如上代码，创建了名为counts的基础表达矩阵；</li>
<li>后续可以根据需要，增添标准化、归一化等行列名不变的矩阵，相当于在count矩阵基础上“新盖的几层楼”。</li>
</ul>
<h3 id="11-查看矩阵">1.1 查看矩阵<a hidden class="anchor" aria-hidden="true" href="#11-查看矩阵">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">#根据矩阵名（楼层名）选择查看</span>
</span></span><span style="display:flex;"><span>assay(sce, <span style="color:#0ff;font-weight:bold">&#34;counts&#34;</span>)[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#或者如下，但我觉得初学者使用第一种方法更能适应对象的结构</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#counts(sce)[1:4,1:4]</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>assays(sce) <span style="color:#007f7f">#查看所有的层楼名，目前只有一个</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="12-新添矩阵">1.2 新添矩阵<a hidden class="anchor" aria-hidden="true" href="#12-新添矩阵">#</a></h3>
<p>一般创建对象时都是引入counts矩阵。根据此基础矩阵可新建相关标准化等矩阵。</p>
<ul>
<li>手动计算添加</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>counts &lt;- assay(sce, <span style="color:#0ff;font-weight:bold">&#34;counts&#34;</span>)
</span></span><span style="display:flex;"><span>libsizes &lt;- colSums(counts)
</span></span><span style="display:flex;"><span>size.factors &lt;- libsizes/mean(libsizes)
</span></span><span style="display:flex;"><span>assay(sce, <span style="color:#0ff;font-weight:bold">&#34;logcounts&#34;</span>) &lt;- log2(t(t(counts)/size.factors) + <span style="color:#ff0;font-weight:bold">1</span>)
</span></span><span style="display:flex;"><span>assay(sce, <span style="color:#0ff;font-weight:bold">&#34;logcounts&#34;</span>)[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span>assays(sce)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># List of length 2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># names(2): counts logcounts</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>利用相关R包可自动添加</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">#注意下面自动添加的log标准化矩阵与上面的logcounts一致，可不运行</span>
</span></span><span style="display:flex;"><span>sce &lt;- scater::logNormCounts(sce)
</span></span><span style="display:flex;"><span>sce
</span></span><span style="display:flex;"><span>assays(sce) <span style="color:#007f7f">#注意是assays，不是assay</span>
</span></span><span style="display:flex;"><span>assay(sce, <span style="color:#0ff;font-weight:bold">&#34;logcounts&#34;</span>)[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>关于assays的命名，我们可以随意命名。为了能够命名有意义并且配合R包之间的接口，官方给了如下的命名建议。
<code>counts</code>: Raw count data, e.g., number of reads or transcripts for a particular gene.
<code>normcounts</code>: Normalized values on the same scale as the original counts. For example, counts divided by cell-specific size factors that are centred at unity.
<code>logcounts</code>: Log-transformed counts or count-like values. In most cases, this will be defined as log-transformed  normcounts, e.g., using log base 2 and a pseudo-count of 1.
<code>cpm</code>: Counts-per-million. This is the read count for each gene in each cell, divided by the library size of each cell in millions.
<code>tpm</code>: Transcripts-per-million. This is the number of transcripts for each gene in each cell, divided by the total number of transcripts in that cell (in millions).</p></blockquote>
<h2 id="2coldata结构">2、colData结构<a hidden class="anchor" aria-hidden="true" href="#2coldata结构">#</a></h2>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-1962099096313355.png" alt="2-1"  />
</p>
<ul>
<li>本质上为行名为sample的data.frame，即为细胞添加注释信息
如上，创建对象时，引入了sample的group分组信息</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>colData(sce) %&gt;% head() 
</span></span><span style="display:flex;"><span>table(sce$group)
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="21-新增coldata">2.1 新增colData<a hidden class="anchor" aria-hidden="true" href="#21-新增coldata">#</a></h3>
<ul>
<li><code>$</code>符可用于查看colData，也能用于新增。（这一点等同于dataframe）</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sce$anything &lt;- runif(ncol(sce))
</span></span><span style="display:flex;"><span>colnames(colData(sce))
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>scater包自动添加</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sce &lt;- scater::addPerCellQC(sce)
</span></span><span style="display:flex;"><span>colData(sce)[, <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">5</span>]
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="22-筛选sample">2.2 筛选sample<a hidden class="anchor" aria-hidden="true" href="#22-筛选sample">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sce[, sce$detected &gt; <span style="color:#ff0;font-weight:bold">3000</span>]
</span></span><span style="display:flex;"><span>sce  <span style="color:#007f7f">#原来是160个sample</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-50aa28e55c2b152c.png" alt="2-2"  />
</p>
<h2 id="3rowdata结构">3、rowData结构<a hidden class="anchor" aria-hidden="true" href="#3rowdata结构">#</a></h2>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-0ff93ccd6cfba93f.png" alt="3" style="zoom: 67%;" />
<p>参考colData，本质上为行名为feature的data.frame，即为基因添加注释信息。</p>
<ul>
<li>可以在创建对象时添加(基因的不同ID)，或者利用scater包添加注释。</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sce &lt;- scater::addPerFeatureQC(sce)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>手动添加可用<code>cbind</code>函数</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>rowData(x) &lt;- cbind(rowData(x), feature)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#同理colData也类似，但使用$符也很方便</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>筛选feature</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sce[rownames(nm_count)[c(<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">10</span>)], ]
</span></span><span style="display:flex;"><span>sce[1:<span style="color:#ff0;font-weight:bold">10</span>, ]
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>Furthermore, there is a special <code>rowRanges</code> slot to hold genomic coordinates in the form of a <code>GRanges</code> or <code>GRangesList</code>. This stores describes the chromosome, start, and end coordinates of the features (genes, genomic regions) in a manner that is easy to query and manipulate via the <a href="https://bioconductor.org/packages/3.11/GenomicRanges">GenomicRanges</a> framework.</p></blockquote>
<h2 id="4reduceddims">4、reducedDims<a hidden class="anchor" aria-hidden="true" href="#4reduceddims">#</a></h2>
<p><img loading="lazy" src="https://upload-images.jianshu.io/upload_images/20354525-2433b75d26d1caf5.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="4"  />
</p>
<ul>
<li>上面三点基本源于SummarizedExperiment对象，但reducedDims是sce对象独有的一点，一般储存PCA，tSNE，uMAP等细胞降维聚类信息；</li>
<li>本质是行名为sample的一组data.frame</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sce &lt;- scater::logNormCounts(sce)
</span></span><span style="display:flex;"><span>sce &lt;- scater::runPCA(sce) 
</span></span><span style="display:flex;"><span>reducedDim(sce, <span style="color:#0ff;font-weight:bold">&#34;PCA&#34;</span>)
</span></span><span style="display:flex;"><span>reducedDim(sce, <span style="color:#0ff;font-weight:bold">&#34;PCA&#34;</span>)[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#           PC1    PC2     PC3   PC4</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#RHC4104  46.66 -46.44  -9.488  1.79</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#RHC6087 -54.38  -6.15  -0.155 13.16</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#RHL2880  -7.89 -38.89  -1.319 -2.90</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#RHC5949 -53.27 -20.88 -12.089 -7.73</span>
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sce &lt;- scater::runTSNE(sce, perplexity = <span style="color:#ff0;font-weight:bold">0.1</span>)
</span></span><span style="display:flex;"><span>reducedDim(sce, <span style="color:#0ff;font-weight:bold">&#34;TSNE&#34;</span>)  <span style="color:#007f7f">#可用于绘图的二维坐标信息</span>
</span></span><span style="display:flex;"><span>reducedDims(sce) <span style="color:#007f7f">#类比assays</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>u &lt;- uwot::umap(t(logcounts(sce)), n_neighbors = <span style="color:#ff0;font-weight:bold">2</span>)
</span></span><span style="display:flex;"><span>reducedDim(sce, <span style="color:#0ff;font-weight:bold">&#34;UMAP_uwot&#34;</span>) &lt;- u
</span></span><span style="display:flex;"><span>reducedDims(sce) <span style="color:#007f7f"># Now stored in the object.</span>
</span></span><span style="display:flex;"><span>reducedDim(sce, <span style="color:#0ff;font-weight:bold">&#34;UMAP_uwot&#34;</span>) <span style="color:#007f7f">#可用于绘图的二维坐标信息</span>
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/021%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--%E5%A4%9A%E6%A0%B7%E6%9C%AC%E6%89%B9%E6%AC%A1%E6%A0%A1%E6%AD%A3/">
    <span class="title">« Prev Page</span>
    <br>
    <span>单细胞分析工具--多样本批次校正</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/018%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--infercnv%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82%E9%89%B4%E5%AE%9A/">
    <span class="title">Next Page »</span>
    <br>
    <span>单细胞分析工具--infercnv拷贝数变异鉴定</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
