<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>表达矩阵TPM,FPKM标准化 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="转录组">
<meta name="description" content="1、计算公式

  
      
          
          sample 1
          sample 2
          &hellip;&hellip;..
          sample k
      
  
  
      
          Gene 1
          10
          12
          
          30
      
      
          Gene 2
          20
          25
          
          60
      
      
          &hellip;&hellip;
          &hellip;
          &hellip;
          &hellip;
          &hellip;
      
      
          Gene n
          0
          0
          &hellip;
          1
      
  

对于(n*k)表达矩阵中k个样本的n个基因的count表达数据。
在任意样本（列）中，基因i的长度为l，count表达值为q；则基因i的FPKM与TPM标准化方式计算如下">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/047%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5tpmfpkm%E6%A0%87%E5%87%86%E5%8C%96/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/047%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5tpmfpkm%E6%A0%87%E5%87%86%E5%8C%96/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="表达矩阵TPM,FPKM标准化" />
<meta property="og:description" content="1、计算公式

  
      
          
          sample 1
          sample 2
          &hellip;&hellip;..
          sample k
      
  
  
      
          Gene 1
          10
          12
          
          30
      
      
          Gene 2
          20
          25
          
          60
      
      
          &hellip;&hellip;
          &hellip;
          &hellip;
          &hellip;
          &hellip;
      
      
          Gene n
          0
          0
          &hellip;
          1
      
  

对于(n*k)表达矩阵中k个样本的n个基因的count表达数据。
在任意样本（列）中，基因i的长度为l，count表达值为q；则基因i的FPKM与TPM标准化方式计算如下" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/047%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5tpmfpkm%E6%A0%87%E5%87%86%E5%8C%96/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-04-30T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2022-04-30T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="表达矩阵TPM,FPKM标准化"/>
<meta name="twitter:description" content="1、计算公式

  
      
          
          sample 1
          sample 2
          &hellip;&hellip;..
          sample k
      
  
  
      
          Gene 1
          10
          12
          
          30
      
      
          Gene 2
          20
          25
          
          60
      
      
          &hellip;&hellip;
          &hellip;
          &hellip;
          &hellip;
          &hellip;
      
      
          Gene n
          0
          0
          &hellip;
          1
      
  

对于(n*k)表达矩阵中k个样本的n个基因的count表达数据。
在任意样本（列）中，基因i的长度为l，count表达值为q；则基因i的FPKM与TPM标准化方式计算如下"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "表达矩阵TPM,FPKM标准化",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/047%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5tpmfpkm%E6%A0%87%E5%87%86%E5%8C%96/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "表达矩阵TPM,FPKM标准化",
  "name": "表达矩阵TPM,FPKM标准化",
  "description": "1、计算公式 sample 1 sample 2 \u0026hellip;\u0026hellip;.. sample k Gene 1 10 12 30 Gene 2 20 25 60 \u0026hellip;\u0026hellip; \u0026hellip; \u0026hellip; \u0026hellip; \u0026hellip; Gene n 0 0 \u0026hellip; 1 对于(n*k)表达矩阵中k个样本的n个基因的count表达数据。\n在任意样本（列）中，基因i的长度为l，count表达值为q；则基因i的FPKM与TPM标准化方式计算如下\n",
  "keywords": [
    "转录组"
  ],
  "articleBody": "1、计算公式 sample 1 sample 2 …….. sample k Gene 1 10 12 30 Gene 2 20 25 60 …… … … … … Gene n 0 0 … 1 对于(n*k)表达矩阵中k个样本的n个基因的count表达数据。\n在任意样本（列）中，基因i的长度为l，count表达值为q；则基因i的FPKM与TPM标准化方式计算如下\n$$ \\begin{align} \u0026 FPKM_i = \\frac{\\frac{q_i}{l_i/10^3}}{\\sum_{j=1}^nq_j/10^6} = \\frac{\\frac{q_i}{l_i}}{{\\sum_{j=1}^nq_j}}*10^9 \\tag{1} \\ \\ \u0026 TPM_i = \\frac{\\frac{q_i}{l_i/10^3}}{\\sum_{j=1}^n(\\frac{q_j}{l_j/10^3})/10^6} = \\frac{\\frac{q_i}{l_i}}{\\sum_{j=1}^n(\\frac{q_j}{l_j})} * 10^6 \\tag{2}\n\\end{align} $$\nFPKM : fragments per kilobase of exon per million mapped fragments. TPM：transcript per million, and the sum of all TPM values is the same in all samples. 2、基因长度 2.1 gtf注释外显子之和 参考https://www.biostars.org/p/456800/，计算基因的所有外显子长度之和作为基因长度。 基因以及对应外显子坐标信息来自基因组注释文件gtf，下载方式参考之前Refgenie笔记。 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 library(GenomicFeatures) #如下选取的是hg19版本的注释基因组文件 txdb \u003c- makeTxDbFromGFF(\"gencode.v32lift37.annotation.gtf\",format=\"gtf\") exons_per_gene \u003c- exonsBy(txdb,by=\"gene\") gene_len_from_exon \u003c- as.data.frame(sum(width(reduce(exons_per_gene)))) colnames(gene_len_from_exon)[1] = \"length\" gene_len_from_exon$ENSEMBL = str_split(rownames(gene_len_from_exon),\"[.]\", simplify = T)[,1] head(gene_len_from_exon) # length ENSEMBL # ENSG00000000003.15_3 4536 ENSG00000000003 # ENSG00000000005.6_3 1476 ENSG00000000005 # ENSG00000000419.12_4 1207 ENSG00000000419 ##接下来将基因ENSEMBL名转为Symbol格式 gene_ids\u003c-AnnotationDbi::select(org.Hs.eg.db, keys=gene_len_from_exon$ENSEMBL, columns=\"SYMBOL\", #目标格式 keytype=\"ENSEMBL\") #目前的格式 gene_ids = gene_ids %\u003e% dplyr::distinct(ENSEMBL, .keep_all = T) %\u003e% na.omit() gene_len = dplyr::left_join(gene_len_from_exon, gene_ids) %\u003e% dplyr::distinct(ENSEMBL, .keep_all = T) %\u003e% dplyr::distinct(SYMBOL, .keep_all = T) dim(gene_len) # [1] 34102 3 head(gene_len) # length ENSEMBL SYMBOL # 1 4536 ENSG00000000003 TSPAN6 # 2 1476 ENSG00000000005 TNMD # 3 1207 ENSG00000000419 DPM1 2.2 TCGAbiolinks包内置的基因长度信息 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 ##(1) hg19 geneInfo_hg19 = TCGAbiolinks::geneInfo %\u003e% na.omit() %\u003e% as.data.frame() head(geneInfo_hg19) ##(1) hg38 geneInfo_hg38 = TCGAbiolinks::geneInfoHT %\u003e% tibble::rownames_to_column(\"ENSEMBL\") library(org.Hs.eg.db) gene_ids = AnnotationDbi::select(org.Hs.eg.db, keys=geneInfo_hg38$ENSEMBL, columns=c(\"SYMBOL\"), #目标格式 keytype=\"ENSEMBL\") #目前的格式 gene_ids = gene_ids %\u003e% distinct(ENSEMBL, .keep_all = T) %\u003e% distinct(SYMBOL, .keep_all = T) %\u003e% na.omit() head(gene_ids) geneInfo_hg38 = dplyr::inner_join(geneInfo_hg38, gene_ids) %\u003e% tibble::column_to_rownames(\"SYMBOL\") head(geneInfo_hg38) 3、定义函数 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 ##(1) counts_to_tpm counts_to_tpm \u003c- function(counts, geneLength) { # Process one column at a time. tpm \u003c- do.call(cbind, lapply(1:ncol(counts), function(i) { # log的减法相当于除法：每一列（单个样本的count）/对应基因的长度 rate = log(counts[,i]) - log(geneLength) #上一步更新后的新的细胞文库大小 denom = log(sum(exp(rate))) #按照细胞文库标准化 exp(rate - denom + log(1e6)) })) colnames(tpm) \u003c- colnames(counts) rownames(tpm) \u003c- rownames(counts) return(tpm) } ##(2) counts_to_fpkm counts_to_fpkm \u003c- function(counts, geneLength) { fpkm \u003c- do.call(cbind, lapply(1:ncol(counts), function(i) { N \u003c- sum(counts[,i]) exp(log(counts[,i]) - log(geneLength) - log(N) + log(1e9)) })) colnames(fpkm) \u003c- colnames(counts) rownames(fpkm) \u003c- rownames(counts) return(fpkm) } ##(3) fpkm_to_Tpm fpkm_to_tpm \u003c- function(fpkm) { exp(log(fpkm) - log(colsum(fpkm)) + log(1e6)) } 4、示例分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 ##(1)基因表达矩阵 exp_count = fread(\"GEO/GSE81861_CRC_NM_all_cells_COUNT.csv.gz\",data.table = F) %\u003e% dplyr::mutate(gene=str_split(V1,\"_\",simplify = T)[,2]) %\u003e% dplyr::distinct(gene, .keep_all = T) %\u003e% tibble::column_to_rownames(\"gene\") %\u003e% dplyr::select(!V1) %\u003e% dplyr::filter(rownames(.) %in% gene_len$SYMBOL) dim(exp_count) # [1] 26258 266 exp_count[1:4,1:4] # RHC3934__Bcell__.7DEA7B RHC3944__Bcell__.7DEA7B RHC3962__Tcell__.C6E879 RHC4003__Bcell__.7DEA7B # TSPAN6 0 0 0 0 # TNMD 0 0 0 0 # DPM1 0 0 0 0 # SCYL3 0 0 0 1 ##(2)基因长度 gene_len_sub = gene_len[match(rownames(exp_count),gene_len$SYMBOL), ] gene_len_sub = data.frame(length=gene_len_sub$length, row.names = gene_len_sub$SYMBOL) head(gene_len_sub) # length # TSPAN6 4536 # TNMD 1476 # DPM1 1207 identical(rownames(exp_count), rownames(gene_len_sub)) #[1] TRUE ##(3) count 2 TPM exp_tpm = counts_to_tpm(exp_count, gene_len_sub) ##(4) count 2 FPKM exp_fpkm = counts_to_fpkm(exp_count, gene_len_sub) ##(5) fpkm 2 tpm exp_tpm2 = as.data.frame(apply(exp_fpkm, 2, fpkm_to_tpm)) ",
  "wordCount" : "934",
  "inLanguage": "en",
  "datePublished": "2022-04-30T00:00:00Z",
  "dateModified": "2022-04-30T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/047%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5tpmfpkm%E6%A0%87%E5%87%86%E5%8C%96/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      表达矩阵TPM,FPKM标准化
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-30 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-30&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-04-30&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;934&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e8%ae%a1%e7%ae%97%e5%85%ac%e5%bc%8f" aria-label="1、计算公式">1、计算公式</a></li>
                    <li>
                        <a href="#2%e5%9f%ba%e5%9b%a0%e9%95%bf%e5%ba%a6" aria-label="2、基因长度">2、基因长度</a><ul>
                            
                    <li>
                        <a href="#21-gtf%e6%b3%a8%e9%87%8a%e5%a4%96%e6%98%be%e5%ad%90%e4%b9%8b%e5%92%8c" aria-label="2.1 gtf注释外显子之和">2.1 gtf注释外显子之和</a></li>
                    <li>
                        <a href="#22-tcgabiolinks%e5%8c%85%e5%86%85%e7%bd%ae%e7%9a%84%e5%9f%ba%e5%9b%a0%e9%95%bf%e5%ba%a6%e4%bf%a1%e6%81%af" aria-label="2.2 TCGAbiolinks包内置的基因长度信息">2.2 TCGAbiolinks包内置的基因长度信息</a></li>
                    <li>
                        <a href="#3%e5%ae%9a%e4%b9%89%e5%87%bd%e6%95%b0" aria-label="3、定义函数">3、定义函数</a></li>
                    <li>
                        <a href="#4%e7%a4%ba%e4%be%8b%e5%88%86%e6%9e%90" aria-label="4、示例分析">4、示例分析</a>
                    </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><h1 id="1计算公式">1、计算公式<a hidden class="anchor" aria-hidden="true" href="#1计算公式">#</a></h1>
<table>
  <thead>
      <tr>
          <th style="text-align: center"></th>
          <th style="text-align: center">sample 1</th>
          <th style="text-align: center">sample 2</th>
          <th style="text-align: center">&hellip;&hellip;..</th>
          <th style="text-align: center">sample k</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td style="text-align: center">Gene 1</td>
          <td style="text-align: center">10</td>
          <td style="text-align: center">12</td>
          <td style="text-align: center"></td>
          <td style="text-align: center">30</td>
      </tr>
      <tr>
          <td style="text-align: center">Gene 2</td>
          <td style="text-align: center">20</td>
          <td style="text-align: center">25</td>
          <td style="text-align: center"></td>
          <td style="text-align: center">60</td>
      </tr>
      <tr>
          <td style="text-align: center">&hellip;&hellip;</td>
          <td style="text-align: center">&hellip;</td>
          <td style="text-align: center">&hellip;</td>
          <td style="text-align: center">&hellip;</td>
          <td style="text-align: center">&hellip;</td>
      </tr>
      <tr>
          <td style="text-align: center">Gene n</td>
          <td style="text-align: center">0</td>
          <td style="text-align: center">0</td>
          <td style="text-align: center">&hellip;</td>
          <td style="text-align: center">1</td>
      </tr>
  </tbody>
</table>
<p>对于(n*k)表达矩阵中k个样本的n个基因的count表达数据。</p>
<p>在任意样本（列）中，基因i的长度为<code>l</code>，count表达值为<code>q</code>；则基因i的FPKM与TPM标准化方式计算如下</p>
<p>$$
\begin{align}
&amp; FPKM_i = \frac{\frac{q_i}{l_i/10^3}}{\sum_{j=1}^nq_j/10^6} = \frac{\frac{q_i}{l_i}}{{\sum_{j=1}^nq_j}}*10^9 \tag{1} \
\
&amp; TPM_i = \frac{\frac{q_i}{l_i/10^3}}{\sum_{j=1}^n(\frac{q_j}{l_j/10^3})/10^6} =
\frac{\frac{q_i}{l_i}}{\sum_{j=1}^n(\frac{q_j}{l_j})} * 10^6  \tag{2}</p>
<p>\end{align}
$$</p>
<blockquote>
<ul>
<li>FPKM : fragments per kilobase of exon per million mapped fragments.</li>
<li>TPM：transcript per million, and the sum of all TPM values is the same in all samples.</li>
</ul></blockquote>
<h1 id="2基因长度">2、基因长度<a hidden class="anchor" aria-hidden="true" href="#2基因长度">#</a></h1>
<h2 id="21-gtf注释外显子之和">2.1 gtf注释外显子之和<a hidden class="anchor" aria-hidden="true" href="#21-gtf注释外显子之和">#</a></h2>
<ul>
<li>参考https://www.biostars.org/p/456800/，计算基因的所有外显子长度之和作为基因长度。</li>
<li>基因以及对应外显子坐标信息来自基因组注释文件gtf，下载方式参考之前Refgenie笔记。</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(GenomicFeatures) 
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#如下选取的是hg19版本的注释基因组文件</span>
</span></span><span style="display:flex;"><span>txdb &lt;- makeTxDbFromGFF(<span style="color:#0ff;font-weight:bold">&#34;gencode.v32lift37.annotation.gtf&#34;</span>,format=<span style="color:#0ff;font-weight:bold">&#34;gtf&#34;</span>)
</span></span><span style="display:flex;"><span>exons_per_gene &lt;- exonsBy(txdb,by=<span style="color:#0ff;font-weight:bold">&#34;gene&#34;</span>)
</span></span><span style="display:flex;"><span>gene_len_from_exon &lt;- as.data.frame(sum(width(reduce(exons_per_gene))))
</span></span><span style="display:flex;"><span>colnames(gene_len_from_exon)[1] = <span style="color:#0ff;font-weight:bold">&#34;length&#34;</span>
</span></span><span style="display:flex;"><span>gene_len_from_exon$ENSEMBL = str_split(rownames(gene_len_from_exon),<span style="color:#0ff;font-weight:bold">&#34;[.]&#34;</span>,
</span></span><span style="display:flex;"><span>                                        simplify = <span style="color:#fff;font-weight:bold">T</span>)[,<span style="color:#ff0;font-weight:bold">1</span>]
</span></span><span style="display:flex;"><span>head(gene_len_from_exon)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                      length         ENSEMBL</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ENSG00000000003.15_3   4536 ENSG00000000003</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ENSG00000000005.6_3    1476 ENSG00000000005</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ENSG00000000419.12_4   1207 ENSG00000000419</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##接下来将基因ENSEMBL名转为Symbol格式</span>
</span></span><span style="display:flex;"><span>gene_ids&lt;-AnnotationDbi::select(org.Hs.eg.db, keys=gene_len_from_exon$ENSEMBL, 
</span></span><span style="display:flex;"><span>                                columns=<span style="color:#0ff;font-weight:bold">&#34;SYMBOL&#34;</span>, <span style="color:#007f7f">#目标格式</span>
</span></span><span style="display:flex;"><span>                                keytype=<span style="color:#0ff;font-weight:bold">&#34;ENSEMBL&#34;</span>) <span style="color:#007f7f">#目前的格式</span>
</span></span><span style="display:flex;"><span>gene_ids = gene_ids %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::distinct(ENSEMBL, .keep_all = <span style="color:#fff;font-weight:bold">T</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  na.omit()
</span></span><span style="display:flex;"><span>gene_len = dplyr::left_join(gene_len_from_exon, gene_ids) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::distinct(ENSEMBL, .keep_all = <span style="color:#fff;font-weight:bold">T</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::distinct(SYMBOL, .keep_all = <span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>dim(gene_len)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 34102     3</span>
</span></span><span style="display:flex;"><span>head(gene_len)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   length         ENSEMBL   SYMBOL</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1   4536 ENSG00000000003   TSPAN6</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2   1476 ENSG00000000005     TNMD</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3   1207 ENSG00000000419     DPM1</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="22-tcgabiolinks包内置的基因长度信息">2.2 TCGAbiolinks包内置的基因长度信息<a hidden class="anchor" aria-hidden="true" href="#22-tcgabiolinks包内置的基因长度信息">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1) hg19</span>
</span></span><span style="display:flex;"><span>geneInfo_hg19 = TCGAbiolinks::geneInfo %&gt;% na.omit() %&gt;% 
</span></span><span style="display:flex;"><span>  as.data.frame()
</span></span><span style="display:flex;"><span>head(geneInfo_hg19)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(1) hg38</span>
</span></span><span style="display:flex;"><span>geneInfo_hg38 = TCGAbiolinks::geneInfoHT %&gt;% 
</span></span><span style="display:flex;"><span>  tibble::rownames_to_column(<span style="color:#0ff;font-weight:bold">&#34;ENSEMBL&#34;</span>)
</span></span><span style="display:flex;"><span>library(org.Hs.eg.db)
</span></span><span style="display:flex;"><span>gene_ids = AnnotationDbi::select(org.Hs.eg.db, keys=geneInfo_hg38$ENSEMBL, 
</span></span><span style="display:flex;"><span>                                columns=c(<span style="color:#0ff;font-weight:bold">&#34;SYMBOL&#34;</span>), <span style="color:#007f7f">#目标格式</span>
</span></span><span style="display:flex;"><span>                                keytype=<span style="color:#0ff;font-weight:bold">&#34;ENSEMBL&#34;</span>) <span style="color:#007f7f">#目前的格式</span>
</span></span><span style="display:flex;"><span>gene_ids = gene_ids %&gt;% 
</span></span><span style="display:flex;"><span>  distinct(ENSEMBL, .keep_all = <span style="color:#fff;font-weight:bold">T</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  distinct(SYMBOL, .keep_all = <span style="color:#fff;font-weight:bold">T</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  na.omit()
</span></span><span style="display:flex;"><span>head(gene_ids)
</span></span><span style="display:flex;"><span>geneInfo_hg38 = dplyr::inner_join(geneInfo_hg38, gene_ids) %&gt;% 
</span></span><span style="display:flex;"><span>  tibble::column_to_rownames(<span style="color:#0ff;font-weight:bold">&#34;SYMBOL&#34;</span>)
</span></span><span style="display:flex;"><span>head(geneInfo_hg38)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="3定义函数">3、定义函数<a hidden class="anchor" aria-hidden="true" href="#3定义函数">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1) counts_to_tpm</span>
</span></span><span style="display:flex;"><span>counts_to_tpm &lt;- <span style="color:#fff;font-weight:bold">function</span>(counts, geneLength) {
</span></span><span style="display:flex;"><span>  <span style="color:#007f7f"># Process one column at a time.</span>
</span></span><span style="display:flex;"><span>  tpm &lt;- do.call(cbind, lapply(<span style="color:#ff0;font-weight:bold">1</span>:ncol(counts), <span style="color:#fff;font-weight:bold">function</span>(i) {
</span></span><span style="display:flex;"><span>    <span style="color:#007f7f"># log的减法相当于除法：每一列（单个样本的count）/对应基因的长度</span>
</span></span><span style="display:flex;"><span>    rate = log(counts[,i]) - log(geneLength) 
</span></span><span style="display:flex;"><span>    <span style="color:#007f7f">#上一步更新后的新的细胞文库大小</span>
</span></span><span style="display:flex;"><span>    denom = log(sum(exp(rate))) 
</span></span><span style="display:flex;"><span>    <span style="color:#007f7f">#按照细胞文库标准化</span>
</span></span><span style="display:flex;"><span>    exp(rate - denom + log(<span style="color:#ff0;font-weight:bold">1e6</span>)) 
</span></span><span style="display:flex;"><span>  }))
</span></span><span style="display:flex;"><span>  colnames(tpm) &lt;- colnames(counts)
</span></span><span style="display:flex;"><span>  rownames(tpm) &lt;- rownames(counts)
</span></span><span style="display:flex;"><span>  <span style="color:#fff;font-weight:bold">return</span>(tpm)
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) counts_to_fpkm</span>
</span></span><span style="display:flex;"><span>counts_to_fpkm &lt;- <span style="color:#fff;font-weight:bold">function</span>(counts, geneLength) {
</span></span><span style="display:flex;"><span>  fpkm &lt;- do.call(cbind, lapply(<span style="color:#ff0;font-weight:bold">1</span>:ncol(counts), <span style="color:#fff;font-weight:bold">function</span>(i) {
</span></span><span style="display:flex;"><span>    N &lt;- sum(counts[,i])
</span></span><span style="display:flex;"><span>    exp(log(counts[,i]) - log(geneLength) - log(N) + log(<span style="color:#ff0;font-weight:bold">1e9</span>))
</span></span><span style="display:flex;"><span>  }))
</span></span><span style="display:flex;"><span>  colnames(fpkm) &lt;- colnames(counts)
</span></span><span style="display:flex;"><span>  rownames(fpkm) &lt;- rownames(counts)
</span></span><span style="display:flex;"><span>  <span style="color:#fff;font-weight:bold">return</span>(fpkm)
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) fpkm_to_Tpm</span>
</span></span><span style="display:flex;"><span>fpkm_to_tpm &lt;- <span style="color:#fff;font-weight:bold">function</span>(fpkm) {
</span></span><span style="display:flex;"><span>  exp(log(fpkm) - log(colsum(fpkm)) + log(<span style="color:#ff0;font-weight:bold">1e6</span>))
</span></span><span style="display:flex;"><span>}
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="4示例分析">4、示例分析<a hidden class="anchor" aria-hidden="true" href="#4示例分析">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1)基因表达矩阵</span>
</span></span><span style="display:flex;"><span>exp_count = fread(<span style="color:#0ff;font-weight:bold">&#34;GEO/GSE81861_CRC_NM_all_cells_COUNT.csv.gz&#34;</span>,data.table = <span style="color:#fff;font-weight:bold">F</span>) %&gt;%
</span></span><span style="display:flex;"><span>  dplyr::mutate(gene=str_split(V1,<span style="color:#0ff;font-weight:bold">&#34;_&#34;</span>,simplify = <span style="color:#fff;font-weight:bold">T</span>)[,<span style="color:#ff0;font-weight:bold">2</span>]) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::distinct(gene, .keep_all = <span style="color:#fff;font-weight:bold">T</span>) %&gt;%
</span></span><span style="display:flex;"><span>  tibble::column_to_rownames(<span style="color:#0ff;font-weight:bold">&#34;gene&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::select(!V1) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::filter(rownames(.) %in% gene_len$SYMBOL)
</span></span><span style="display:flex;"><span>dim(exp_count)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 26258   266</span>
</span></span><span style="display:flex;"><span>exp_count[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        RHC3934__Bcell__.7DEA7B RHC3944__Bcell__.7DEA7B RHC3962__Tcell__.C6E879 RHC4003__Bcell__.7DEA7B</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TSPAN6                       0                       0                       0                       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TNMD                         0                       0                       0                       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DPM1                         0                       0                       0                       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SCYL3                        0                       0                       0                       1</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2)基因长度</span>
</span></span><span style="display:flex;"><span>gene_len_sub = gene_len[match(rownames(exp_count),gene_len$SYMBOL), ]
</span></span><span style="display:flex;"><span>gene_len_sub = data.frame(length=gene_len_sub$length,
</span></span><span style="display:flex;"><span>                          row.names = gene_len_sub$SYMBOL)
</span></span><span style="display:flex;"><span>head(gene_len_sub)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#          length</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TSPAN6     4536</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TNMD       1476</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DPM1       1207</span>
</span></span><span style="display:flex;"><span>identical(rownames(exp_count), rownames(gene_len_sub))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1] TRUE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) count 2 TPM</span>
</span></span><span style="display:flex;"><span>exp_tpm = counts_to_tpm(exp_count, gene_len_sub)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(4) count 2 FPKM</span>
</span></span><span style="display:flex;"><span>exp_fpkm = counts_to_fpkm(exp_count, gene_len_sub)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(5) fpkm 2 tpm</span>
</span></span><span style="display:flex;"><span>exp_tpm2 = as.data.frame(apply(exp_fpkm, <span style="color:#ff0;font-weight:bold">2</span>, fpkm_to_tpm))
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">转录组</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/046tcgabiolinks%E5%8C%85-tcga%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD%E4%B8%8E%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/">
    <span class="title">« Prev Page</span>
    <br>
    <span>TCGAbiolinks包-TCGA数据下载与差异分析</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/048%E4%BB%8Egeo%E4%B8%8B%E8%BD%BD%E8%8A%AF%E7%89%87%E6%88%96rnaseq%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE/">
    <span class="title">Next Page »</span>
    <br>
    <span>从GEO下载芯片或RNAseq测序数据</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
