<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰ | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="">
<meta name="description" content="ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰ - Li&#39;s Bioinfo-Blog">
<meta name="author" content="">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" type="application/rss+xml" href="https://lishensuo.github.io/en/posts/bioinfo/index.xml">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰" />
<meta property="og:description" content="æç”³é”çš„åšå®¢ï¼Œè®°å½•ç”Ÿä¿¡å­¦ä¹ è¿‡ç¨‹ã€‚R/Python/Shell" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰"/>
<meta name="twitter:description" content="æç”³é”çš„åšå®¢ï¼Œè®°å½•ç”Ÿä¿¡å­¦ä¹ è¿‡ç¨‹ã€‚R/Python/Shell"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "åˆ†ç±»",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }
  ]
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="list" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="ä¸»é¡µ">
                    <span>ä¸»é¡µ</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="åˆ†ç±»">
                    <span>åˆ†ç±»</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="æ ‡ç­¾">
                    <span>æ ‡ç­¾</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="å½’æ¡£">
                    <span>å½’æ¡£</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="å…³äº">
                    <span>å…³äº</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="æœç´¢ (Alt &#43; /)" accesskey=/>
                    <span>æœç´¢</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main"> 
<header class="page-header"><div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/">åˆ†ç±»</a></div>
  <h1>ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰</h1>
</header>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>è‚¿ç˜¤äºšå‹åˆ†ç¾¤å·¥å…·ConsensusClusterPlus
			</h2>
		</header>
		<section class="entry-content">
			<p> ConsensusClusterPlusåŒ…æ˜¯è‚¿ç˜¤åˆ†å‹ç ”ç©¶çš„å¸¸ç”¨å·¥å…·ï¼Œå…¶äº2010å¹´å‘è¡¨äºBioinformaticsã€‚
Paperï¼šhttps://academic.oup.com/bioinformatics/article/26/12/1572/281699 Tutorialï¼šhttps://bioconductor.org/packages/release/bioc/vignettes/ConsensusClusterPlus/inst/doc/ConsensusClusterPlus.pdf 1 2 # BiocManager::install(&#34;ConsensusClusterPlus&#34;) library(ConsensusClusterPlus) 1ã€ç¤ºä¾‹è¡¨è¾¾çŸ©é˜µ è¡Œåä¸ºåŸºå› ï¼Œåˆ—åä¸ºæ ·æœ¬çš„åŸºå› è¡¨è¾¾çŸ©é˜µï¼›æ ¹æ®éœ€è¦ï¼Œè¿›è¡Œæ ‡å‡†åŒ–å¤„ç† åŸºå› é›†çš„é€‰æ‹©æ˜¯å…³é”®çš„ä¸€æ­¥ï¼Œå¯æ ¹æ®ç»Ÿè®¡å­¦ï¼ˆé«˜å˜åŸºå› ï¼‰æˆ–è€…ç”Ÿç‰©å­¦ï¼ˆç‰¹å®šåŠŸèƒ½ç›¸å…³åŸºå› ï¼‰è¿›è¡Œé€‰æ‹© å¦‚ä¸‹æ˜¯å‚è€ƒæ•™ç¨‹çš„æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 library(ALL) data(ALL) d=exprs(ALL) dim(d) # [1] 12625 128 mads=apply(d,1,mad) d=d[rev(order(mads))[1:5000],] # ç­›é€‰é«˜æ–¹å·®åŸºå›  dim(d) # [1] 5000 128 ## æ ‡å‡†åŒ–å¤„ç†ï¼šæ¯è¡Œï¼ˆåŸºå› ï¼‰å‡å»æ¯è¡Œçš„ä¸­ä½æ•° d = sweep(d,1, apply(d,1,median,na.rm=T)) d[1:4,1:4] # 01005 01010 03002 04006 # 36638_at 1.556121 0.9521271 -0.05018082 4.780378 # 39318_at 1.191353 2.5013225 -2.38793537 -1.199521 # 38514_at 1.020716 3.2785671 1.55949145 -3.345919 # 266_s_at 1.829260 0.3624327 1.54913247 -1.286294 2ã€æ ·æœ¬äºšå‹é‰´å®š å®é™…å°±æ˜¯ä¸€ä¸ªå‡½æ•°å³å¯ï¼šConsensusClusterPlus() å…¶ä¸­æ¶‰åŠåˆ°è¾ƒå¤šå‚æ•°çš„é€‰æ‹©ï¼Œå…·ä½“å¦‚ä¸‹ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 ## é»˜è®¤å‚æ•° ConsensusClusterPlus(d, maxK=3, reps=10, pItem=0.8, pFeature=1, title=&#34;untitled_consensus_cluster&#34;, clusterAlg=&#34;hc&#34;, distance=&#34;pearson&#34;, plot=NULL, writeTable=FALSE, seed=42) ## å‚æ•°å«ä¹‰ # maxKï¼šè€ƒè™‘çš„æœ€å¤§èšç±»æ•°ï¼Œå»ºè®®å–10~20 # repsï¼šæŠ½æ ·æ¬¡æ•°ï¼Œå»ºè®®1000 # pItemä¸pFeatureï¼šåˆ†åˆ«è¡¨ç¤ºå¯¹æ ·æœ¬ä¸åŸºå› çš„æŠ½æ ·æ¯”ä¾‹ # titleï¼šå›¾ç‰‡æˆ–æ–‡ä»¶çš„ä¿å­˜è·¯å¾„ # clusterAlgï¼šèšç±»æ–¹å¼ c(&#34;hc&#34;,&#34;pam&#34;,&#34;km&#34;) # distance: è·ç¦»è®¡ç®—æ–¹å¼ c(&#34;pearson&#34;,&#34;spearman&#34;,&#34;euclidean&#34;,&#34;binay&#34;,&#34;maximum&#34;,&#34;canberra&#34;,&#34;minkowski&#34;) # plot : æ˜¯å¦ç»˜å›¾ï¼Œä»¥åŠå›¾å½¢ç±»å‹ c(NULL, &#34;pdf&#34;, &#34;png&#34;, &#34;pngBMP&#34;) # writeTable: æ˜¯å¦ä¿å­˜æ–‡ä»¶ # seedï¼šéšæœºç§å­ results = ConsensusClusterPlus(d, maxK=10, reps=1000, pItem=0.8, pFeature=1, title=&#34;./tmp/&#34;, clusterAlg=&#34;hc&#34;, distance=&#34;pearson&#34;, plot=&#34;pdf&#34;, writeTable=FALSE, seed=42) Rå¯¹è±¡ç»“æœï¼šlistæ ¼å¼ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 ## æŸ¥çœ‹èšç±»æ•°ä¸º2çš„ç»“æœä¿¡æ¯ names(results[[3]]) # [1] &#34;consensusMatrix&#34; &#34;consensusTree&#34; &#34;consensusClass&#34; &#34;ml&#34; &#34;clrs&#34; table(results[[3]]$consensusClass) # 1 2 3 # 69 28 31 dim(results[[3]]$consensusMatrix) # [1] 128 128 results[[3]]$consensusMatrix[1:4,1:4] # [,1] [,2] [,3] [,4] # [1,] 1.0000000 0.3408360 0.7996870 0.3515249 # [2,] 0.3408360 1.0000000 0.1224806 1.0000000 # [3,] 0.7996870 0.1224806 1.0000000 0.1190108 # [4,] 0.3515249 1.0000000 0.1190108 1.0000000 ## consensus values 0 (never clustered together) ## consensus values 1 (always clustered together) å›¾å½¢è¾“å‡ºç»“æœï¼Œä¸»è¦ä¸¤ç±»å›¾ ï¼ˆ1ï¼‰ç‰¹å®šèšç±»ç»“æœçš„çƒ­å›¾
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-04-20 00:00:00 &#43;0000 UTC&#39;&gt;2023-04-20&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-04-20&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;905&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to è‚¿ç˜¤äºšå‹åˆ†ç¾¤å·¥å…·ConsensusClusterPlus" href="https://lishensuo.github.io/en/posts/bioinfo/077%E8%82%BF%E7%98%A4%E4%BA%9A%E5%9E%8B%E5%88%86%E7%BE%A4%E5%B7%A5%E5%85%B7consensusclusterplus/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>ä½¿ç”¨IsoformSwitchAnalyzeRåŒ…è¿›è¡Œå¯å˜å‰ªåˆ‡åˆ†æ
			</h2>
		</header>
		<section class="entry-content">
			<p> https://bioconductor.org/packages/release/bioc/html/IsoformSwitchAnalyzeR.html 1ã€èƒŒæ™¯çŸ¥è¯† The usage of Alternative Transcription Start sites (aTSSå¯é€‰æ‹©è½¬å½•èµ·å§‹ä½ç‚¹), Alternative Splicing (ASå¯é€‰æ‹©å‰ªåˆ‡ä½ç‚¹) and alternative Transcription Termination Sites (aTTSå¯é€‰æ‹©ç»ˆæ­¢ä½ç‚¹) are collectively collectively results in the production of different isoforms. Alternative isoforms are widely used as recently demonstrated by The ENCODE Consortium, which found that on average, 6.3 different transcripts are generated per gene; a number which may vary considerably per gene.
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-16 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-16&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-04-16&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;3284&amp;nbsp;|&amp;nbsp;7 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to ä½¿ç”¨IsoformSwitchAnalyzeRåŒ…è¿›è¡Œå¯å˜å‰ªåˆ‡åˆ†æ" href="https://lishensuo.github.io/en/posts/bioinfo/070%E4%BD%BF%E7%94%A8isoformswitchanalyzer%E5%8C%85%E8%BF%9B%E8%A1%8C%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87%E5%88%86%E6%9E%90/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>è¯ºæ¨¡å›¾æˆ–åˆ—çº¿å›¾ç»˜åˆ¶åˆ†æ
			</h2>
		</header>
		<section class="entry-content">
			<p> è‚¿ç˜¤é¢„åç±»æ–‡ç« çš„å¸¸è§„æ­¥éª¤ä¹‹ä¸€æ˜¯ç»˜åˆ¶è¯ºæ¨¡å›¾ï¼Œå¹¶è¿›è¡Œç›¸å…³åˆ†æã€‚ä»¥ä¸‹æ€»ç»“äº†ç›¸å…³åŸºç¡€ç»˜åˆ¶å·¥å…·ã€‚
ç¤ºä¾‹æ•°æ®é›† 1 2 3 4 5 6 library(survival) head(lung) # inst time status age sex ph.ecog ph.karno pat.karno meal.cal wt.loss # 1 3 306 2 74 1 1 90 100 1175 NA # 2 3 455 2 68 1 0 90 90 1225 15 # 3 3 1010 1 56 1 0 90 90 NA 15 1ã€rmsåŒ… å‚è€ƒç”¨æ³•ï¼šhttps://atm.amegroups.com/article/view/14736/15089 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 library(rms) # var.labels = c(age=&#34;Age in Years&#34;, # lac=&#34;lactate&#34;, # sex=&#34;Sex of the participant&#34;, # shock=&#34;shock&#34;, # y=&#34;outcome&#34;, # Y=&#34;ordinal&#34;) # label(data) = lapply(names(var.labels), # function(x) label(data[,x]) = var.labels[x]) mod.cox &lt;- cph(Surv(time,status) ~ ph.ecog&#43;sex&#43;age,lung, surv=TRUE) ddist &lt;- datadist(lung) options(datadist=&#39;ddist&#39;) surv.cox &lt;- Survival(mod.cox) # å®šä¹‰å‡½æ•° med &lt;- Quantile(mod.cox) surv &lt;- Survival(mod.cox) nom.cox &lt;- nomogram(mod.cox, # æ ¹æ®total pointè¿›è¡Œç‰¹å®šå‡½æ•°è®¡ç®— fun=list(function(x) surv.cox(365, x), # ä¸€å¹´ç”Ÿå­˜ç‡ function(x) med(lp=x, q=0.5)), # ä¸­ä½ç”Ÿå­˜æ—¶é—´ funlabel=c(&#34;200-Day Sur. Prob.&#34;, &#34;Median Survival Time&#34;), lp=F, # ä¸å±•ç¤ºLinear Predictor conf.int=c(0.1,0.5) # ä¸¤ä¸ªç½®ä¿¡åŒºé—´ ) plot(nom.cox, col.conf=c(&#39;red&#39;,&#39;green&#39;), # ç½®ä¿¡åŒºé—´çš„é¢œè‰² col.grid = c(&#34;grey30&#34;,&#34;grey&#34;) # ç½‘æ ¼çš„é¢œè‰² ) # f = cph(Surv(time, status) ~ age &#43; sex &#43; ph.karno, data = lung, # x = T, y = T, sur = T) # pred_score = apply(lung, 1, function(x){ # pred=Predict(f, age=x[&#34;age&#34;], sex=x[&#34;sex&#34;], ph.karno=x[&#34;ph.karno&#34;]) # return(pred$yhat) # }) %&gt;% unlist() # summary(pred_score) æ ¡å‡†æ›²çº¿
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-04-28 00:00:00 &#43;0000 UTC&#39;&gt;2023-04-28&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-04-28&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;585&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to è¯ºæ¨¡å›¾æˆ–åˆ—çº¿å›¾ç»˜åˆ¶åˆ†æ" href="https://lishensuo.github.io/en/posts/bioinfo/078%E8%AF%BA%E6%A8%A1%E5%9B%BE%E6%88%96%E5%88%97%E7%BA%BF%E5%9B%BE%E7%BB%98%E5%88%B6%E5%88%86%E6%9E%90/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>UCSCXenaShinyåŒ…è‚¿ç˜¤æ•°æ®åˆ†æå¯è§†åŒ–
			</h2>
		</header>
		<section class="entry-content">
			<p> UCSCXenaShinyæ˜¯åŸºäºé›†æˆäº†å¤šç§è‚¿ç˜¤æ•°æ®åº“çš„UCSCXenaå¹³å°ï¼Œè¿›è¡Œæ•°æ®ä¸‹è½½ã€åˆ†æã€å¯è§†åŒ–çš„Shinyå·¥å…·ï¼ˆä»¥åŠåŒåRåŒ…ï¼‰ï¼Œç”±ä¸Šæµ·ç§‘æŠ€å¤§å­¦ç‹è¯—ç¿”åšå£«ç­‰å…±åŒå¼€å‘ï¼›äº2021å¹´6æœˆå‘è¡¨äºBioinformaticsã€‚ä¸‹é¢ä¸»è¦å­¦ä¹ å…¶RåŒ…çš„ç›¸å…³å‡½æ•°ï¼Œäº†è§£å…¶æ ¸å¿ƒåŠŸèƒ½ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-05-12 00:00:00 &#43;0000 UTC&#39;&gt;2023-05-12&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-05-12&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1745&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to UCSCXenaShinyåŒ…è‚¿ç˜¤æ•°æ®åˆ†æå¯è§†åŒ–" href="https://lishensuo.github.io/en/posts/bioinfo/079ucscxenashiny%E5%8C%85%E8%82%BF%E7%98%A4%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E5%8F%AF%E8%A7%86%E5%8C%96/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>EnhancedVolcanoåŒ…ç»˜åˆ¶ç«å±±å›¾
			</h2>
		</header>
		<section class="entry-content">
			<p> EnhancedVolcanoåŒ…å¯æ ¹æ®å·®å¼‚åˆ†æç»“æœï¼ŒåŸºäºggplot2ç»˜å›¾ç»“æ„ï¼Œæ–¹ä¾¿åœ°ç»˜åˆ¶ç¾è§‚çš„ç«å±±å›¾ï¼Œä¸‹é¢æ ¹æ®è‡ªå·±çš„ç†è§£å°ç»“ä¸‹åŸºæœ¬ç”¨æ³•ã€‚
å®˜æ–¹å…¨é¢çš„æ•™ç¨‹ï¼šhttps://github.com/kevinblighe/EnhancedVolcano ç¤ºä¾‹å·®å¼‚åŸºå› æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 library(airway) library(magrittr) data(&#39;airway&#39;) airway$dex %&lt;&gt;% relevel(&#39;untrt&#39;) ens &lt;- rownames(airway) library(org.Hs.eg.db) symbols &lt;- mapIds(org.Hs.eg.db, keys = ens, column = c(&#39;SYMBOL&#39;), keytype = &#39;ENSEMBL&#39;) symbols &lt;- symbols[!is.na(symbols)] symbols &lt;- symbols[match(rownames(airway), names(symbols))] rownames(airway) &lt;- symbols keep &lt;- !is.na(rownames(airway)) airway &lt;- airway[keep,] library(&#39;DESeq2&#39;) dds &lt;- DESeqDataSet(airway, design = ~ cell &#43; dex) dds &lt;- DESeq(dds, betaPrior=FALSE) res &lt;- results(dds, contrast = c(&#39;dex&#39;,&#39;trt&#39;,&#39;untrt&#39;)) res &lt;- lfcShrink(dds, contrast = c(&#39;dex&#39;,&#39;trt&#39;,&#39;untrt&#39;), res=res, type = &#39;normal&#39;) res &lt;- as.data.frame(res) head(res) # baseMean log2FoldChange lfcSE stat pvalue padj # TSPAN6 710.0931707 -0.37807189 0.09851236 -3.8404448 0.0001228116 0.0009522932 # TNMD 0.0000000 NA NA NA NA NA # DPM1 521.2572396 0.19826365 0.10931684 1.8155169 0.0694445184 0.1910397405 # SCYL3 237.6068046 0.03234467 0.13821470 0.2371917 0.8125081096 0.9118161375 # C1orf112 58.0358739 -0.08835419 0.25056704 -0.3194810 0.7493618190 0.8773885438 # FGR 0.3194343 -0.08459224 0.15186225 -0.3948862 0.6929268648 NA å¦‚ä¸Šï¼Œåªè¦åŒ…å«åŒ…å«åŸºå› åã€å·®å¼‚å€æ•°ã€På€¼ä¸‰éƒ¨åˆ†ä¿¡æ¯çš„å·®å¼‚ç»“æœå°±å¯ä»¥ç”¨äºç»˜åˆ¶ç«å±±å›¾ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-16 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-16&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-04-16&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;980&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to EnhancedVolcanoåŒ…ç»˜åˆ¶ç«å±±å›¾" href="https://lishensuo.github.io/en/posts/bioinfo/080enhancedvolcano%E5%8C%85%E7%BB%98%E5%88%B6%E7%81%AB%E5%B1%B1%E5%9B%BE/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>IOBRåŒ…è‚¿ç˜¤signatureæ‰“åˆ†ä¸å…ç–«æµ¸æ¶¦åˆ†æ
			</h2>
		</header>
		<section class="entry-content">
			<p> IOBRåŒ…é›†signatureæ‰“åˆ†ä¸å…ç–«æµ¸æ¶¦åˆ†æä¸ºä¸€ä½“çš„è‚¿ç˜¤æ•°æ®åˆ†æå·¥å…·ï¼Œç”±å—æ–¹åŒ»ç§‘å¤§å­¦å—æ–¹åŒ»é™¢å»–æ—ºå†›æ•™æˆï¼Œæ›¾ä¸œå¼ºåšå£«ç­‰äººäº2021å¹´7æœˆå‘è¡¨äºFrontiers in Immunologyï¼Œå¼•ç”¨æ•°ä»¥è¶…è¿‡100ä½™æ¬¡ã€‚ç°æ ¹æ®å…¶githubæ•™ç¨‹å­¦ä¹ å…¶éƒ¨åˆ†åŠŸèƒ½ç”¨æ³•ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-05-15 00:00:00 &#43;0000 UTC&#39;&gt;2023-05-15&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-05-15&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1667&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to IOBRåŒ…è‚¿ç˜¤signatureæ‰“åˆ†ä¸å…ç–«æµ¸æ¶¦åˆ†æ" href="https://lishensuo.github.io/en/posts/bioinfo/081iobr%E5%8C%85%E8%82%BF%E7%98%A4signature%E6%89%93%E5%88%86%E4%B8%8E%E5%85%8D%E7%96%AB%E6%B5%B8%E6%B6%A6%E5%88%86%E6%9E%90/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>BayesPrismåŒ…æ ¹æ®scRNAseqé¢„æµ‹Bulkç»†èƒç»„æˆ
			</h2>
		</header>
		<section class="entry-content">
			<p> BayesPrismæ˜¯ç”±ç¾å›½åº·å¥ˆå°”å¤§å­¦Tin Yi Chuç­‰äººå¼€å‘çš„RåŒ…å·¥å…·ï¼Œäº2022å¹´4æœˆå‘è¡¨åœ¨Nature Cancerã€‚ç®€å•æ¥è¯´ï¼Œè¯¥æ–¹æ³•ä½¿ç”¨å•ç»†èƒRNA-seqä½œä¸ºå…ˆéªŒä¿¡æ¯ï¼Œé€šè¿‡ä¼°è®¡æ‰¹é‡æ ·æœ¬ä¸­ç»†èƒç±»å‹æ¯”ä¾‹å’Œç»†èƒç±»å‹ç‰¹å¼‚æ€§åŸºå› è¡¨è¾¾çš„è”åˆåéªŒåˆ†å¸ƒã€‚P(Î¸,Z|X,Ï•)ï¼Œå‚æ•°æ¦‚å¿µå¦‚ä¸‹
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-06-03 00:00:00 &#43;0000 UTC&#39;&gt;2023-06-03&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-06-03&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1352&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to BayesPrismåŒ…æ ¹æ®scRNAseqé¢„æµ‹Bulkç»†èƒç»„æˆ" href="https://lishensuo.github.io/en/posts/bioinfo/082bayesprism%E5%8C%85%E6%A0%B9%E6%8D%AEscrnaseq%E9%A2%84%E6%B5%8Bbulk%E7%BB%86%E8%83%9E%E7%BB%84%E6%88%90/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>RåŒ…NMFéè´ŸçŸ©é˜µåˆ†è§£
			</h2>
		</header>
		<section class="entry-content">
			<p> ï¼ˆ1ï¼‰NMFæ˜¯éè´ŸçŸ©é˜µåˆ†è§£ï¼ˆNon-negative Matrix Factorizationï¼‰çš„ç¼©å†™ã€‚å®ƒæ˜¯å°†ä¸€ä¸ªéè´Ÿæ•°æ®çŸ©é˜µåˆ†è§£ä¸ºä¸¤ä¸ªéè´ŸçŸ©é˜µçš„ä¹˜ç§¯ï¼Œå…¶ä¸­ä¸€ä¸ªçŸ©é˜µè¡¨ç¤ºç‰¹å¾çš„åŸºçŸ©é˜µï¼Œå¦ä¸€ä¸ªçŸ©é˜µè¡¨ç¤ºæ¯ä¸ªæ ·æœ¬åœ¨è¿™äº›ç‰¹å¾ä¸Šçš„ç³»æ•°çŸ©é˜µã€‚è¿™æ ·çš„åˆ†è§£å¯ä»¥å°†åŸå§‹æ•°æ®è¡¨ç¤ºä¸ºä¸€ç»„éè´ŸåŸºå‘é‡çš„åŠ æƒç»„åˆï¼Œä»è€Œå®ç°æ•°æ®çš„é™ç»´å’Œç‰¹å¾æå–ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-06-17 00:00:00 &#43;0000 UTC&#39;&gt;2023-06-17&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-06-17&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1038&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to RåŒ…NMFéè´ŸçŸ©é˜µåˆ†è§£" href="https://lishensuo.github.io/en/posts/bioinfo/083r%E5%8C%85nmf%E9%9D%9E%E8%B4%9F%E7%9F%A9%E9%98%B5%E5%88%86%E8%A7%A3/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>UCSCXenaShiny V2ç®€è¦æ•™ç¨‹
			</h2>
		</header>
		<section class="entry-content">
			<p> Githubä»“åº“ï¼šhttps://github.com/openbiox/UCSCXenaShiny
Online Appï¼šhttps://shiny.zhoulab.ac.cn/UCSCXenaShiny/
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2024-10-05 00:00:00 &#43;0000 UTC&#39;&gt;2024-10-05&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2024-10-05&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1999&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to UCSCXenaShiny V2ç®€è¦æ•™ç¨‹" href="https://lishensuo.github.io/en/posts/bioinfo/084ucscxenashiny-v2%E7%AE%80%E8%A6%81%E6%95%99%E7%A8%8B/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>MAGeCK-VISPRè´¨æ§åˆ†æCRISPRæµ‹åºæ•°æ®
			</h2>
		</header>
		<section class="entry-content">
			<p>å‚è€ƒ
https://bitbucket.org/liulab/mageck-vispr MAGeCK-VISPRï¼šhttps://doi.org/10.1186%2Fs13059-015-0843-6 1. å®‰è£…ç¯å¢ƒ 1 2 3 conda create -n mageck-vispr python=3.7 conda activate mageck-vispr conda install -c bioconda -c conda-forge mageck-vispr ç»è¿‡å®é™…æ“ä½œï¼Œéƒ¨åˆ†æºä»£ç éœ€è¦ä¿®æ”¹
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2025-07-27 00:00:00 &#43;0000 UTC&#39;&gt;2025-07-27&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2025-07-27&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;982&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to MAGeCK-VISPRè´¨æ§åˆ†æCRISPRæµ‹åºæ•°æ®" href="https://lishensuo.github.io/en/posts/bioinfo/085mageck-vispr%E8%B4%A8%E6%8E%A7%E5%88%86%E6%9E%90crispr%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE/"></a>
</article>
<footer class="page-footer">
  <nav class="pagination">
    <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/page/6/">Â« Prev Page</a>
    <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/page/8/">Next Page Â»</a>
  </nav>
</footer>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>æ‚¨æ˜¯æœ¬ç«™ç¬¬ <span id="busuanzi_value_site_uv"></span> ä½è®¿é—®è€…ï¼Œæ€»æµè§ˆé‡ä¸º <span id="busuanzi_value_site_pv"></span> æ¬¡
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
