<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>使用IsoformSwitchAnalyzeR包进行可变剪切分析 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="转录组">
<meta name="description" content="
https://bioconductor.org/packages/release/bioc/html/IsoformSwitchAnalyzeR.html

1、背景知识

The usage of Alternative Transcription Start sites (aTSS可选择转录起始位点), Alternative Splicing (AS可选择剪切位点) and alternative Transcription Termination Sites (aTTS可选择终止位点) are collectively collectively results in the production of different isoforms.




Alternative isoforms are widely used as recently demonstrated by The ENCODE Consortium, which found that on average, 6.3 different transcripts are generated per gene; a number which may vary considerably per gene.">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/070%E4%BD%BF%E7%94%A8isoformswitchanalyzer%E5%8C%85%E8%BF%9B%E8%A1%8C%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87%E5%88%86%E6%9E%90/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/070%E4%BD%BF%E7%94%A8isoformswitchanalyzer%E5%8C%85%E8%BF%9B%E8%A1%8C%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87%E5%88%86%E6%9E%90/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="使用IsoformSwitchAnalyzeR包进行可变剪切分析" />
<meta property="og:description" content="
https://bioconductor.org/packages/release/bioc/html/IsoformSwitchAnalyzeR.html

1、背景知识

The usage of Alternative Transcription Start sites (aTSS可选择转录起始位点), Alternative Splicing (AS可选择剪切位点) and alternative Transcription Termination Sites (aTTS可选择终止位点) are collectively collectively results in the production of different isoforms.




Alternative isoforms are widely used as recently demonstrated by The ENCODE Consortium, which found that on average, 6.3 different transcripts are generated per gene; a number which may vary considerably per gene." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/070%E4%BD%BF%E7%94%A8isoformswitchanalyzer%E5%8C%85%E8%BF%9B%E8%A1%8C%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87%E5%88%86%E6%9E%90/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-04-16T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2022-04-16T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="使用IsoformSwitchAnalyzeR包进行可变剪切分析"/>
<meta name="twitter:description" content="
https://bioconductor.org/packages/release/bioc/html/IsoformSwitchAnalyzeR.html

1、背景知识

The usage of Alternative Transcription Start sites (aTSS可选择转录起始位点), Alternative Splicing (AS可选择剪切位点) and alternative Transcription Termination Sites (aTTS可选择终止位点) are collectively collectively results in the production of different isoforms.




Alternative isoforms are widely used as recently demonstrated by The ENCODE Consortium, which found that on average, 6.3 different transcripts are generated per gene; a number which may vary considerably per gene."/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "使用IsoformSwitchAnalyzeR包进行可变剪切分析",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/070%E4%BD%BF%E7%94%A8isoformswitchanalyzer%E5%8C%85%E8%BF%9B%E8%A1%8C%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87%E5%88%86%E6%9E%90/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "使用IsoformSwitchAnalyzeR包进行可变剪切分析",
  "name": "使用IsoformSwitchAnalyzeR包进行可变剪切分析",
  "description": " https://bioconductor.org/packages/release/bioc/html/IsoformSwitchAnalyzeR.html 1、背景知识 The usage of Alternative Transcription Start sites (aTSS可选择转录起始位点), Alternative Splicing (AS可选择剪切位点) and alternative Transcription Termination Sites (aTTS可选择终止位点) are collectively collectively results in the production of different isoforms. Alternative isoforms are widely used as recently demonstrated by The ENCODE Consortium, which found that on average, 6.3 different transcripts are generated per gene; a number which may vary considerably per gene.\n",
  "keywords": [
    "转录组"
  ],
  "articleBody": " https://bioconductor.org/packages/release/bioc/html/IsoformSwitchAnalyzeR.html 1、背景知识 The usage of Alternative Transcription Start sites (aTSS可选择转录起始位点), Alternative Splicing (AS可选择剪切位点) and alternative Transcription Termination Sites (aTTS可选择终止位点) are collectively collectively results in the production of different isoforms. Alternative isoforms are widely used as recently demonstrated by The ENCODE Consortium, which found that on average, 6.3 different transcripts are generated per gene; a number which may vary considerably per gene.\n目前基因转录本水平的定量表达分析软件有：Kallisto, Salmon, RSEM or StringTie等\nIsoformSwitchAnalyzeR包可基于上述转录本定量结果，分析同一基因在对照/实验条件下(normal/disease)是否会表达不同比例的转录本。\n如上图所示，IsoformSwitchAnalyzeR分析流程主要分为两大部分：(1) identify isoform switch[基础分析]; (2) predict potential functional consequences of the identified isoform switches[进阶分析]\n2、基础分析 1 2 3 library(IsoformSwitchAnalyzeR) library(tidyverse) library(data.table) 2.1 初步创建switchAnalyzeRlist对象 2.1.1 导入转录本定量数据 使用的是IsoformSwitchAnalyzeR的示例数据，为salmon软件的定量结果。\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 #2组，各两个样本 list.files(\"data/salmon\", recursive = T, full.names = T) # [1] \"data/salmon/hESC_0/quant.sf.gz\" # [2] \"data/salmon/hESC_1/quant.sf.gz\" # [3] \"data/salmon/iPS_0/quant.sf.gz\" # [4] \"data/salmon/iPS_1/quant.sf.gz\" fread(\"data/salmon/hESC_0/quant.sf.gz\") %\u003e% head #包含5列，分别为转录本id， 转录本长度， 有效长度，TPM标准化，原始count # Name Length EffectiveLength TPM NumReads # 1: TCONS_00000001 1652 1652 0.000 0.00000 # 2: TCONS_00000002 1488 1488 0.000 0.00000 # 3: TCONS_00000003 1595 1595 0.000 0.00000 # 4: TCONS_00000006 78 78 1377990.000 4.67384 # 5: TCONS_00000007 2750 2750 329.487 589.15400 # 6: TCONS_00000008 4369 4369 358.067 1050.07000 #importIsoformExpression会自动识别salmon的定量结果，储存到list对象里 salmonQuant \u003c- importIsoformExpression( parentDir = \"data/salmon\" ) salmonQuant$abundance %\u003e% head(3) # isoform_id hESC_0 hESC_1 iPS_0 iPS_1 # 1 TCONS_00000001 0 0.000000 0.000000 4.659597 # 2 TCONS_00000002 0 1.564879 5.504247 2.818824 # 3 TCONS_00000003 0 0.000000 0.000000 0.000000 salmonQuant$counts %\u003e% head(3) # isoform_id hESC_0 hESC_1 iPS_0 iPS_1 # 1 TCONS_00000001 0 0.0000000 0.00000 18.13313 # 2 TCONS_00000002 0 0.1116201 21.10248 10.96964 # 3 TCONS_00000003 0 0.0000000 0.00000 0.00000 2.1.2 样本分组信息 1 2 3 4 5 6 7 8 myDesign \u003c- data.frame( sampleID = c(\"hESC_0\", \"hESC_1\", \"iPS_0\", \"iPS_1\"), condition = c(\"hESC\", \"hESC\", \"iPS\", \"iPS\")) # sampleID condition # 1 hESC_0 hESC # 2 hESC_1 hESC # 3 iPS_0 iPS # 4 iPS_1 iPS 2.1.3 其它注释信息 （1）gtf文件: 转录本的外显子组成以及对应基因 1 2 3 4 5 6 7 8 9 10 11 12 13 fread(\"data/example.gtf.gz\") %\u003e% head(2) # V1 V2 V3 V4 V5 V6 V7 V8 # 1: chr1 cufflinks exon 11874 12227 . + . # 2: chr1 cufflinks exon 11874 12227 . + . # V9 # 1: transcript_id \"TCONS_00000001\"; gene_id \"XLOC_000001\"; # 2: transcript_id \"TCONS_00000002\"; gene_id \"XLOC_000001\"; fread(\"data/example.gtf.gz\") %\u003e% dplyr::count(V3) # V3 n # 1: CDS 9612 # 2: exon 10929 （2）fasta文件：转录本的核苷酸序列 1 2 3 4 5 6 7 8 fread(\"data/example_isoform_nt.fasta\") %\u003e% head() # \u003eTCONS_00000001 # 1: CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTATTTGCTGTCTCTTAGCCCAGACTTCCCGTGTCCT # 2: TTCCACCGGGCCTTTGAGAGGTCACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTCCAGCAACTGCTG # 3: GCCTGTGCCAGGGTGCAAGCTGAGCACTGGAGTGGAGTTTTCCTGTGGAGAGGAGCCATGCCTAGAGTGGGATGGGCCAT # 4: TGTTCATCTTCTGGCCCCTGTTGTCTGCATGTAACTTAATACCACAACCAGGCATAGGGGAAAGATTGGAGGAAAGATGA # 5: GTGAGAGCATCAACTTCTCTCACAACCTAGGCCAGTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCA # 6: GAGCTGCAGAAGACGACGGCCGACTTGGATCACACTCTTGTGAGTGTCCCCAGTGTTGCAGAGGCAGGGCCATCAGGCAC 2.1.4 创建switchAnalyzeRlist对象 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 sar1 \u003c- importRdata( isoformCountMatrix = salmonQuant$counts, isoformRepExpression = salmonQuant$abundance, designMatrix = myDesign, isoformExonAnnoation = \"data/example.gtf.gz\", isoformNtFasta = \"data/example_isoform_nt.fasta\" ) sar1 %\u003e% length() # 11 names(sar1) # [1] \"isoformFeatures\" \"exons\" \"conditions\" # [4] \"designMatrix\" \"sourceId\" \"isoformCountMatrix\" # [7] \"isoformRepExpression\" \"runInfo\" \"orfAnalysis\" # [10] \"isoformRepIF\" \"ntSequence\" sar1$isoformFeatures %\u003e% head(2) # iso_ref/gene_ref : 新的唯一编号，不重复 # gene_overall_mean/iso_overall_mean : 基因/转录本在所有样本的平均表达量 # IF_overall(isoform fraction=isoform_exp / gene_exp): 在所有样本中，特定基因表达该转录本的比例 # iso_ref gene_ref isoform_id gene_id condition_1 condition_2 # 280 isoComp_00000001 geneComp_00000001 TCONS_00000316 AADACL3 hESC iPS # 281 isoComp_00000002 geneComp_00000001 TCONS_00000317 AADACL3 hESC iPS # gene_name gene_biotype iso_biotype gene_overall_mean gene_value_1 gene_value_2 # 280 AADACL3 NA NA 25.37693 32.66891 18.08494 # 281 AADACL3 NA NA 25.37693 32.66891 18.08494 # gene_stderr_1 gene_stderr_2 gene_log2_fold_change gene_q_value iso_overall_mean # 280 10.1063 6.337321 -0.8527734 NA 3.704839 # 281 10.1063 6.337321 -0.8527734 NA 21.672089 # iso_value_1 iso_value_2 iso_stderr_1 iso_stderr_2 iso_log2_fold_change iso_q_value # 280 4.846923 2.562755 3.823911 2.562511 -0.9167289 NA # 281 27.821989 15.522189 6.282387 3.774811 -0.8414829 NA # IF_overall IF1 IF2 dIF isoform_switch_q_value gene_switch_q_value PTC # 280 0.114475 0.124 0.10495 -0.01905 NA NA FALSE # 281 0.885525 0.876 0.89505 0.01905 NA NA FALSE sar1$isoformFeatures %\u003e% dplyr::filter(gene_id==\"AADACL3\") %\u003e% dplyr::select(isoform_id, gene_id, dplyr::contains(\"IF\")) # isoform_id gene_id IF_overall IF1 IF2 dIF # 1 TCONS_00000316 AADACL3 0.114475 0.124 0.10495 -0.01905 # 2 TCONS_00000317 AADACL3 0.885525 0.876 0.89505 0.0190 #ORF转录本开放阅读框注释 sar1$orfAnalysis %\u003e% head(2) # orfTransciptStart/orfTransciptEnd 基于转录本核苷酸序列(剪切后)的起始位置 # orfStartGenomic/orfEndGenomic 基于基因组坐标(剪切前)的起始位置 # isoform_id orfTransciptStart orfTransciptEnd orfTransciptLength # 1 TCONS_00000001 1402 1629 228 # 2 TCONS_00000002 317 715 399 # orfStarExon orfEndExon orfStartGenomic orfEndGenomic # 1 3 3 14159 14386 # 2 1 3 12190 13636 # stopDistanceToLastJunction stopIndex PTC orf_origin # 1 -1163 3 FALSE Annotation # 2 -231 3 FALSE Annotation 2.2 过滤 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 sar2 \u003c- preFilter(sar1) #默认过滤标准 # geneExpressionCutoff = 1 : 根据gene_overall_mean列过滤低表达基因 # isoformExpressionCutoff = 0 : 根据iso_overall_mean列过滤低表达基因 # IFcutoff=0.01: 根据IF_overall过滤 # dIFcutoff = 0.1: 根据dIF过滤（绝对值） # removeSingleIsoformGenes = TRUE: 删除只有一个转录本的基因 dim(sar1) # [1] 1060 30 dim(sar2) # [1] 760 30 sar2 # This switchAnalyzeRlist list contains: # 760 isoforms from 208 genes # 1 comparison from 2 conditions (in total 4 samples) # # Feature analyzed: # [1] \"ORFs, ntSequence\" 2.3 isoformSwitch分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 sar3 \u003c- isoformSwitchTestDEXSeq(sar2) sar3$isoformFeatures %\u003e% head(2) # 如下，主要增加了isoform_switch_q_value与gene_switch_q_value两列信息 # iso_ref gene_ref isoform_id gene_id condition_1 # 557 isoComp_00000004 geneComp_00000003 TCONS_00003880 ACAP3 hESC # 558 isoComp_00000005 geneComp_00000003 TCONS_00003881 ACAP3 hESC # condition_2 gene_name gene_biotype iso_biotype gene_overall_mean # 557 iPS ACAP3 NA NA 910.5955 # 558 iPS ACAP3 NA NA 910.5955 # gene_value_1 gene_value_2 gene_stderr_1 gene_stderr_2 # 557 1771.599 49.59168 377.02 15.68552 # 558 1771.599 49.59168 377.02 15.68552 # gene_log2_fold_change gene_q_value iso_overall_mean iso_value_1 # 557 -5.158528 NA 241.0513 478.7646 # 558 -5.158528 NA 186.6521 365.0107 # iso_value_2 iso_stderr_1 iso_stderr_2 iso_log2_fold_change # 557 3.337951 175.44243 3.337951 -7.159924 # 558 8.293521 89.61604 1.997000 -5.458111 # iso_q_value IF_overall IF1 IF2 dIF # 557 NA 0.156075 0.26100 0.05115 -0.20985 # 558 NA 0.188100 0.20455 0.17165 -0.03290 # isoform_switch_q_value gene_switch_q_value PTC # 557 0.2601812 0.02341811 FALSE # 558 0.7079831 0.02341811 FALSE #如下结果gene_switch_q_value即为isoform_switch_q_value的最小值 sar3$isoformFeatures %\u003e% dplyr::filter(gene_id==\"ACAP3\") %\u003e% dplyr::select(isoform_id, gene_id,dIF, dplyr::contains(\"switch\")) # isoform_id gene_id dIF isoform_switch_q_value gene_switch_q_value # 1 TCONS_00003880 ACAP3 -0.20985 0.26018123 0.02341811 # 2 TCONS_00003881 ACAP3 -0.03290 0.70798307 0.02341811 # 3 TCONS_00003882 ACAP3 0.29825 0.02341811 0.02341811 # 4 TCONS_00003883 ACAP3 -0.05555 0.58385549 0.02341811 switchPlot(sar3, gene = 'ACAP3', condition1=\"hESC\", condition2=\"iPS\") 1 2 3 4 5 6 7 8 9 10 11 12 13 #四个子图 switchPlotGeneExp(sar3, gene = 'ACAP3', condition1=\"hESC\", condition2=\"iPS\") switchPlotIsoExp(sar3, gene = 'ACAP3', condition1=\"hESC\", condition2=\"iPS\") switchPlotIsoUsage(sar3, gene = 'ACAP3', condition1=\"hESC\", condition2=\"iPS\") switchPlotTranscript(sar3, gene = 'ACAP3', condition1=\"hESC\", condition2=\"iPS\") 2.4 isoformSwitch结果探索 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 sar3$isoformSwitchAnalysis %\u003e% head() # iso_ref gene_ref isoform_id condition_1 condition_2 dIF pvalue # 1 isoComp_00000922 geneComp_00000294 TCONS_00000007 hESC iPS 0.01615 8.645971e-01 # 2 isoComp_00000923 geneComp_00000294 TCONS_00000008 hESC iPS -0.25885 4.859083e-02 # 3 isoComp_00000924 geneComp_00000294 TCONS_00000009 hESC iPS 0.24275 7.547031e-05 # 4 isoComp_00000929 geneComp_00000298 TCONS_00000017 hESC iPS 0.08785 7.690961e-01 # 5 isoComp_00000930 geneComp_00000298 TCONS_00000018 hESC iPS -0.47220 6.113540e-14 # 6 isoComp_00000931 geneComp_00000298 TCONS_00000019 hESC iPS 0.23965 2.119346e-03 # padj IF1 IF2 # 1 9.665421e-01 0.51665 0.53280 # 2 1.464787e-01 0.48320 0.22435 # 3 7.536969e-04 0.00015 0.24290 # 4 9.355274e-01 0.00000 0.08785 # 5 3.815868e-12 0.47220 0.00000 # 6 1.250306e-02 0.00000 0.23965 2.4.1 Top Switches–gene 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 sar.ge_df=extractTopSwitches( sar3, filterForConsequences = FALSE, n = NA, extractGenes = TRUE, # when FALSE isoforms are returned sortByQvals = TRUE ) dim(sar.ge_df) # [1] 68 7 head(sar.ge_df) # gene_ref gene_id gene_name condition_1 condition_2 gene_switch_q_value Rank # 1 geneComp_00000088 ENO1 ENO1 hESC iPS 2.429932e-90 1 # 4 geneComp_00000324 XLOC_001217 hESC iPS 1.464063e-24 2 # 5 geneComp_00000171 NBL1 NBL1 hESC iPS 1.039215e-21 3 # 7 geneComp_00000053 CASP9 CASP9 hESC iPS 5.232094e-18 4 # 10 geneComp_00000282 UBR4 UBR4 hESC iPS 1.597415e-17 5 # 14 geneComp_00000072 CROCC CROCC hESC iPS 1.752050e-17 6 2.4.2 Top Switches–isoform 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 sar.iso_df=extractTopSwitches( sar3, filterForConsequences = FALSE, n = NA, extractGenes = FALSE, # when FALSE isoforms are returned sortByQvals = TRUE ) dim(sar.iso_df) # [1] 121 12 head(sar.iso_df) # iso_ref gene_ref isoform_id gene_id gene_name condition_1 condition_2 IF1 # 1 isoComp_00000297 geneComp_00000088 TCONS_00004087 ENO1 ENO1 hESC iPS 0.157 # 2 isoComp_00000985 geneComp_00000324 TCONS_00003819 XLOC_001217 hESC iPS 0.084 # 3 isoComp_00000551 geneComp_00000171 TCONS_00000463 NBL1 NBL1 hESC iPS 0.112 # 4 isoComp_00000549 geneComp_00000171 TCONS_00000460 NBL1 NBL1 hESC iPS 0.881 # 5 isoComp_00000152 geneComp_00000053 TCONS_00004180 CASP9 CASP9 hESC iPS 0.474 # 6 isoComp_00000242 geneComp_00000072 TCONS_00000420 CROCC CROCC hESC iPS 0.683 # IF2 dIF isoform_switch_q_value Rank # 1 0.000 -0.157 2.429932e-90 1 # 2 0.434 0.350 1.464063e-24 2 # 3 1.000 0.888 1.039215e-21 3 # 4 0.000 -0.881 1.194327e-19 4 # 5 0.000 -0.474 5.232094e-18 5 # 6 0.173 -0.510 1.752050e-17 6 2.4.3 火山图 1 2 3 4 5 6 7 8 9 10 ggplot(data=sar3$isoformFeatures, aes(x=dIF, y=-log10(isoform_switch_q_value))) + geom_point( aes( color=abs(dIF) \u003e 0.1 \u0026 isoform_switch_q_value \u003c 0.05 ), # default cutoff size=1 ) + geom_hline(yintercept = -log10(0.05), linetype='dashed') + # default cutoff geom_vline(xintercept = c(-0.1, 0.1), linetype='dashed') + # default cutoff scale_color_manual('Signficant\\nIsoform Switch', values = c('black','red')) + labs(x='dIF', y='-Log10 ( Isoform Switch Q Value )') + theme_bw() 3、进阶分析 对于不同组样本的某一基因，存在其中一种isoform相对上调，另一种isoform相对下调的情况。可以对isoform的特点进行分析，从而进一步分析isoformswitch的影响与意义。\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 sar3 # This switchAnalyzeRlist list contains: # 312 isoforms from 68 genes # 1 comparison from 2 conditions (in total 4 samples) # # Switching features: # Comparison Isoforms Switches Genes # 1 hESC vs iPS 121 95 68 # # Feature analyzed: # [1] \"Isoform Switch Identification, ORFs, ntSequence\" names(sar3) # [1] \"isoformFeatures\" \"exons\" \"conditions\" \"designMatrix\" # [5] \"sourceId\" \"isoformCountMatrix\" \"isoformRepExpression\" \"runInfo\" # [9] \"orfAnalysis\" \"isoformRepIF\" \"ntSequence\" 3.1 可变剪切Alternative Splicing 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 sar4 \u003c- analyzeAlternativeSplicing(sar3) sar4$AlternativeSplicingAnalysis %\u003e% head(2) # isoform_id ES ES_genomic_start ES_genomic_end MEE MEE_genomic_start MEE_genomic_end MES # 1 TCONS_00003880 0 0 0 # 2 TCONS_00003881 1 1230098 1230196 0 0 # MES_genomic_start MES_genomic_end IR IR_genomic_start IR_genomic_end A5 A5_genomic_start A5_genomic_end # 1 0 0 # 2 0 0 # A3 A3_genomic_start A3_genomic_end ATSS # 1 0 1 # 2 0 1 # ATSS_genomic_start # 1 1234725;1235211;1235353;1235538;1235889;1237368;1238302;1238542;1239466;1243149;1244822 # 2 1243149;1244822 # ATSS_genomic_end ATTS # 1 1234736;1235285;1235448;1235582;1236072;1237426;1238355;1238661;1241309;1243269;1244989 0 # 2 1243269;1244989 0 # ATTS_genomic_start ATTS_genomic_end # 1 # 2 1 2 extractSplicingSummary(sar4) # 一个significant isoform如果发生了A3事件，如果其dIF\u003e0,则iPS相对hESC used more;反之used less 1 extractSplicingEnrichment(sar4) 1 extractSplicingGenomeWide(sar4) 3.2 isoform Consequences预测 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 sar5 \u003c- extractSequence(sar4, writeToFile=TRUE, pathToOutput = 'data/output') list.files(\"data/output/\") # [1] \"isoformSwitchAnalyzeR_isoform_AA.fasta\" # [2] \"isoformSwitchAnalyzeR_isoform_nt.fasta\" sar5$ntSequence %\u003e% head(3) #DNAStringSet object of length 6: # width seq names # [1] 2750 GGGTCTCCCTCTGTTG...CCCACGCGGACAGAG TCONS_00000007 # [2] 4369 TTACTGTTGATTGTGA...AAAATATCGCCCACG TCONS_00000008 # [3] 4272 TTACTGTTGATTGTGA...AAAATATCGCCCACG TCONS_00000009 sar5$aaSequence %\u003e% head(3) #AAStringSet object of length 3: # width seq names # [1] 389 MLLPPGSLSRPRTFSS...QAQLLPHSGPFRPNS TCONS_00000007 # [2] 389 MLLPPGSLSRPRTFSS...QAQLLPHSGPFRPNS TCONS_00000008 # [3] 389 MLLPPGSLSRPRTFSS...QAQLLPHSGPFRPNS TCONS_00000009 CPAT : The Coding-Potential Assessment Tool which is a tool for predicting whether an isoform is coding or not. (_nt.fasta) CPC2 : The Coding Potential Calculator 2 which is a tool for predicting whether an isoform is coding or not. (_nt.fasta) Pfam : Prediction of protein domains (_AA.fasta) SignalP : Prediction of Signal Peptides(_AA.fasta) IUPred2A: Predicts Intrinsically Disordered Regions (IDR) and Intrinsically Disordered Binding Regions (IDBR)(_AA.fasta) NetSurfP-2 : Prediction of Intrinsically Disordered Regions (IDR) (_AA.fasta) 上传核苷酸/氨基酸序列到相应网站，下载预测的结果，再导入到switchAnalyzeRlist对象中\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 sar6 \u003c- analyzeCPAT( switchAnalyzeRlist = sar5, pathToCPATresultFile = \"data/external/cpat_results.txt\", codingCutoff = 0.725, removeNoncodinORFs = TRUE ) sar6 \u003c- analyzePFAM( switchAnalyzeRlist = sar6, pathToPFAMresultFile = \"data/external/pfam_results.txt\" ) sar6 \u003c- analyzeSignalP( switchAnalyzeRlist = sar6, pathToSignalPresultFile = \"data/external/signalP_results.txt\" ) sar6 \u003c- analyzeIUPred2A( switchAnalyzeRlist = sar6, pathToIUPred2AresultFile = \"data/external/iupred2a_result.txt.gz\" ) sar7 \u003c- analyzeSwitchConsequences(sar6) length(sar7) # [1] 18 names(sar7) # [1] \"isoformFeatures\" \"exons\" # [3] \"conditions\" \"designMatrix\" # [5] \"sourceId\" \"isoformCountMatrix\" # [7] \"isoformRepExpression\" \"runInfo\" # [9] \"orfAnalysis\" \"isoformRepIF\" # [11] \"ntSequence\" \"isoformSwitchAnalysis\" # [13] \"AlternativeSplicingAnalysis\" \"aaSequence\" # [15] \"domainAnalysis\" \"signalPeptideAnalysis\" # [17] \"idrAnalysis\" \"switchConsequence\" sar7$switchConsequence %\u003e% dplyr::filter(isoformsDifferent==\"TRUE\") %\u003e% head() # gene_ref gene_id gene_name condition_1 condition_2 isoformUpregulated # 1 geneComp_00000003 ACAP3 ACAP3 hESC iPS TCONS_00003882 # 2 geneComp_00000004 ACOT7 ACOT7 hESC iPS TCONS_00004035 # 3 geneComp_00000004 ACOT7 ACOT7 hESC iPS TCONS_00004035 # 4 geneComp_00000004 ACOT7 ACOT7 hESC iPS TCONS_00004035 # 5 geneComp_00000004 ACOT7 ACOT7 hESC iPS TCONS_00004035 # 6 geneComp_00000004 ACOT7 ACOT7 hESC iPS TCONS_00004036 # isoformDownregulated iso_ref_up iso_ref_down featureCompared # 1 TCONS_00003880 isoComp_00000006 isoComp_00000004 ORF_seq_similarity # 2 TCONS_00004039 isoComp_00000010 isoComp_00000014 ORF_seq_similarity # 3 TCONS_00004039 isoComp_00000010 isoComp_00000014 NMD_status # 4 TCONS_00004039 isoComp_00000010 isoComp_00000014 domains_identified # 5 TCONS_00004039 isoComp_00000010 isoComp_00000014 IDR_identified # 6 TCONS_00004039 isoComp_00000011 isoComp_00000014 NMD_status # isoformsDifferent switchConsequence # 1 TRUE ORF is longer # 2 TRUE ORF is longer # 3 TRUE NMD insensitive # 4 TRUE Domain gain # 5 TRUE IDR gain # 6 TRUE NMD insensitive extractConsequenceSummary(sar7) 1 extractConsequenceEnrichment(sar7) 1 2 3 switchPlot(sar7, gene = 'ACAP3', condition1=\"hESC\", condition2=\"iPS\") ",
  "wordCount" : "3284",
  "inLanguage": "en",
  "datePublished": "2022-04-16T00:00:00Z",
  "dateModified": "2022-04-16T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/070%E4%BD%BF%E7%94%A8isoformswitchanalyzer%E5%8C%85%E8%BF%9B%E8%A1%8C%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87%E5%88%86%E6%9E%90/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      使用IsoformSwitchAnalyzeR包进行可变剪切分析
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-16 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-16&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-04-16&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;3284&amp;nbsp;|&amp;nbsp;7 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e8%83%8c%e6%99%af%e7%9f%a5%e8%af%86" aria-label="1、背景知识">1、背景知识</a></li>
                    <li>
                        <a href="#2%e5%9f%ba%e7%a1%80%e5%88%86%e6%9e%90" aria-label="2、基础分析">2、基础分析</a><ul>
                            
                    <li>
                        <a href="#21-%e5%88%9d%e6%ad%a5%e5%88%9b%e5%bb%baswitchanalyzerlist%e5%af%b9%e8%b1%a1" aria-label="2.1 初步创建switchAnalyzeRlist对象">2.1 初步创建switchAnalyzeRlist对象</a><ul>
                            
                    <li>
                        <a href="#211--%e5%af%bc%e5%85%a5%e8%bd%ac%e5%bd%95%e6%9c%ac%e5%ae%9a%e9%87%8f%e6%95%b0%e6%8d%ae" aria-label="2.1.1  导入转录本定量数据">2.1.1  导入转录本定量数据</a></li>
                    <li>
                        <a href="#212-%e6%a0%b7%e6%9c%ac%e5%88%86%e7%bb%84%e4%bf%a1%e6%81%af" aria-label="2.1.2 样本分组信息">2.1.2 样本分组信息</a></li>
                    <li>
                        <a href="#213-%e5%85%b6%e5%ae%83%e6%b3%a8%e9%87%8a%e4%bf%a1%e6%81%af" aria-label="2.1.3 其它注释信息">2.1.3 其它注释信息</a><ul>
                            
                    <li>
                        <a href="#1gtf%e6%96%87%e4%bb%b6-%e8%bd%ac%e5%bd%95%e6%9c%ac%e7%9a%84%e5%a4%96%e6%98%be%e5%ad%90%e7%bb%84%e6%88%90%e4%bb%a5%e5%8f%8a%e5%af%b9%e5%ba%94%e5%9f%ba%e5%9b%a0" aria-label="（1）gtf文件: 转录本的外显子组成以及对应基因">（1）gtf文件: 转录本的外显子组成以及对应基因</a></li>
                    <li>
                        <a href="#2fasta%e6%96%87%e4%bb%b6%e8%bd%ac%e5%bd%95%e6%9c%ac%e7%9a%84%e6%a0%b8%e8%8b%b7%e9%85%b8%e5%ba%8f%e5%88%97" aria-label="（2）fasta文件：转录本的核苷酸序列">（2）fasta文件：转录本的核苷酸序列</a></li></ul>
                    </li>
                    <li>
                        <a href="#214-%e5%88%9b%e5%bb%baswitchanalyzerlist%e5%af%b9%e8%b1%a1" aria-label="2.1.4 创建switchAnalyzeRlist对象">2.1.4 创建switchAnalyzeRlist对象</a></li></ul>
                    </li>
                    <li>
                        <a href="#22-%e8%bf%87%e6%bb%a4" aria-label="2.2 过滤">2.2 过滤</a></li>
                    <li>
                        <a href="#23-isoformswitch%e5%88%86%e6%9e%90" aria-label="2.3 isoformSwitch分析">2.3 isoformSwitch分析</a></li>
                    <li>
                        <a href="#24-isoformswitch%e7%bb%93%e6%9e%9c%e6%8e%a2%e7%b4%a2" aria-label="2.4 isoformSwitch结果探索">2.4 isoformSwitch结果探索</a><ul>
                            
                    <li>
                        <a href="#241-top-switchesgene" aria-label="2.4.1 Top Switches&ndash;gene">2.4.1 Top Switches&ndash;gene</a></li>
                    <li>
                        <a href="#242-top-switchesisoform" aria-label="2.4.2 Top Switches&ndash;isoform">2.4.2 Top Switches&ndash;isoform</a></li>
                    <li>
                        <a href="#243-%e7%81%ab%e5%b1%b1%e5%9b%be" aria-label="2.4.3 火山图">2.4.3 火山图</a></li></ul>
                    </li></ul>
                    </li>
                    <li>
                        <a href="#3%e8%bf%9b%e9%98%b6%e5%88%86%e6%9e%90" aria-label="3、进阶分析">3、进阶分析</a><ul>
                            
                    <li>
                        <a href="#31-%e5%8f%af%e5%8f%98%e5%89%aa%e5%88%87alternative-splicing" aria-label="3.1 可变剪切Alternative Splicing">3.1 可变剪切Alternative Splicing</a></li>
                    <li>
                        <a href="#32-isoform-consequences%e9%a2%84%e6%b5%8b" aria-label="3.2 isoform Consequences预测">3.2 isoform Consequences预测</a>
                    </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><ul>
<li><a href="https://bioconductor.org/packages/release/bioc/html/IsoformSwitchAnalyzeR.html">https://bioconductor.org/packages/release/bioc/html/IsoformSwitchAnalyzeR.html</a></li>
</ul>
<h1 id="1背景知识">1、背景知识<a hidden class="anchor" aria-hidden="true" href="#1背景知识">#</a></h1>
<ul>
<li>The usage of Alternative Transcription Start sites (aTSS可选择转录起始位点), Alternative Splicing (AS可选择剪切位点) and alternative Transcription Termination Sites (aTTS可选择终止位点) are collectively collectively results in the production of <strong>different isoforms.</strong></li>
</ul>
<img src="https://www.researchgate.net/publication/327264306/figure/fig1/AS:664597191204868@1535463726583/A-Schematic-representation-of-a-mRNA-transcript-Isoforms-can-be-generated-by.png" alt="A) Schematic representation of a mRNA transcript. Isoforms can be... |  Download Scientific Diagram" style="zoom:50%;" />
<ul>
<li>
<p>Alternative isoforms are widely used as recently demonstrated by The ENCODE Consortium, which found that on average, <strong>6.3 different transcripts are generated per gene</strong>; a number which may vary considerably per gene.</p>
</li>
<li>
<p>目前基因转录本水平的定量表达分析软件有：Kallisto, <strong>Salmon</strong>, RSEM or StringTie等</p>
</li>
<li>
<p>IsoformSwitchAnalyzeR包可基于上述转录本定量结果，分析同一基因在对照/实验条件下(normal/disease)是否会表达不同比例的转录本。</p>
<img src="https://s6.jpg.cm/2022/02/02/LwNUYk.png" alt="LwNUYk.png" style="zoom:50%;" />
</li>
<li>
<p>如上图所示，IsoformSwitchAnalyzeR分析流程主要分为两大部分：(1) identify isoform switch[基础分析]; (2) predict potential functional consequences of the identified isoform switches[进阶分析]</p>
</li>
</ul>
<h1 id="2基础分析">2、基础分析<a hidden class="anchor" aria-hidden="true" href="#2基础分析">#</a></h1>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(IsoformSwitchAnalyzeR)
</span></span><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>library(data.table)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="21-初步创建switchanalyzerlist对象">2.1 初步创建switchAnalyzeRlist对象<a hidden class="anchor" aria-hidden="true" href="#21-初步创建switchanalyzerlist对象">#</a></h2>
<h3 id="211--导入转录本定量数据">2.1.1  导入转录本定量数据<a hidden class="anchor" aria-hidden="true" href="#211--导入转录本定量数据">#</a></h3>
<p>使用的是IsoformSwitchAnalyzeR的示例数据，为salmon软件的定量结果。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">#2组，各两个样本</span>
</span></span><span style="display:flex;"><span>list.files(<span style="color:#0ff;font-weight:bold">&#34;data/salmon&#34;</span>, 
</span></span><span style="display:flex;"><span>           recursive = <span style="color:#fff;font-weight:bold">T</span>, full.names = <span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;data/salmon/hESC_0/quant.sf.gz&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [2] &#34;data/salmon/hESC_1/quant.sf.gz&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [3] &#34;data/salmon/iPS_0/quant.sf.gz&#34; </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [4] &#34;data/salmon/iPS_1/quant.sf.gz&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>fread(<span style="color:#0ff;font-weight:bold">&#34;data/salmon/hESC_0/quant.sf.gz&#34;</span>) %&gt;% head
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#包含5列，分别为转录本id， 转录本长度， 有效长度，TPM标准化，原始count</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#              Name Length EffectiveLength         TPM   NumReads</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1: TCONS_00000001   1652            1652       0.000    0.00000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2: TCONS_00000002   1488            1488       0.000    0.00000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3: TCONS_00000003   1595            1595       0.000    0.00000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4: TCONS_00000006     78              78 1377990.000    4.67384</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5: TCONS_00000007   2750            2750     329.487  589.15400</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6: TCONS_00000008   4369            4369     358.067 1050.07000</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#importIsoformExpression会自动识别salmon的定量结果，储存到list对象里</span>
</span></span><span style="display:flex;"><span>salmonQuant &lt;- importIsoformExpression(
</span></span><span style="display:flex;"><span>  parentDir = <span style="color:#0ff;font-weight:bold">&#34;data/salmon&#34;</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>salmonQuant$abundance %&gt;% head(<span style="color:#ff0;font-weight:bold">3</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#       isoform_id hESC_0   hESC_1    iPS_0    iPS_1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 TCONS_00000001      0 0.000000 0.000000 4.659597</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 TCONS_00000002      0 1.564879 5.504247 2.818824</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 TCONS_00000003      0 0.000000 0.000000 0.000000</span>
</span></span><span style="display:flex;"><span>salmonQuant$counts %&gt;% head(<span style="color:#ff0;font-weight:bold">3</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#       isoform_id hESC_0    hESC_1    iPS_0    iPS_1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 TCONS_00000001      0 0.0000000  0.00000 18.13313</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 TCONS_00000002      0 0.1116201 21.10248 10.96964</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 TCONS_00000003      0 0.0000000  0.00000  0.00000</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="212-样本分组信息">2.1.2 样本分组信息<a hidden class="anchor" aria-hidden="true" href="#212-样本分组信息">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>myDesign &lt;- data.frame(
</span></span><span style="display:flex;"><span>  sampleID = c(<span style="color:#0ff;font-weight:bold">&#34;hESC_0&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;hESC_1&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;iPS_0&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;iPS_1&#34;</span>),
</span></span><span style="display:flex;"><span>  condition = c(<span style="color:#0ff;font-weight:bold">&#34;hESC&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;hESC&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;iPS&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;iPS&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   sampleID condition</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1   hESC_0      hESC</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2   hESC_1      hESC</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3    iPS_0       iPS</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4    iPS_1       iPS</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="213-其它注释信息">2.1.3 其它注释信息<a hidden class="anchor" aria-hidden="true" href="#213-其它注释信息">#</a></h3>
<h4 id="1gtf文件-转录本的外显子组成以及对应基因">（1）gtf文件: 转录本的外显子组成以及对应基因<a hidden class="anchor" aria-hidden="true" href="#1gtf文件-转录本的外显子组成以及对应基因">#</a></h4>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>fread(<span style="color:#0ff;font-weight:bold">&#34;data/example.gtf.gz&#34;</span>) %&gt;% head(<span style="color:#ff0;font-weight:bold">2</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#      V1        V2   V3    V4    V5 V6 V7 V8</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1: chr1 cufflinks exon 11874 12227  .  +  .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2: chr1 cufflinks exon 11874 12227  .  +  .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                                                        V9</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1: transcript_id &#34;TCONS_00000001&#34;; gene_id &#34;XLOC_000001&#34;;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2: transcript_id &#34;TCONS_00000002&#34;; gene_id &#34;XLOC_000001&#34;;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>fread(<span style="color:#0ff;font-weight:bold">&#34;data/example.gtf.gz&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::count(V3)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#      V3     n</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1:  CDS  9612</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2: exon 10929</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h4 id="2fasta文件转录本的核苷酸序列">（2）fasta文件：转录本的核苷酸序列<a hidden class="anchor" aria-hidden="true" href="#2fasta文件转录本的核苷酸序列">#</a></h4>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>fread(<span style="color:#0ff;font-weight:bold">&#34;data/example_isoform_nt.fasta&#34;</span>) %&gt;% head()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                                                                     &gt;TCONS_00000001</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1: CTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTATTTGCTGTCTCTTAGCCCAGACTTCCCGTGTCCT</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2: TTCCACCGGGCCTTTGAGAGGTCACAGGGTCTTGATGCTGTGGTCTTCATCTGCAGGTGTCTGACTTCCAGCAACTGCTG</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3: GCCTGTGCCAGGGTGCAAGCTGAGCACTGGAGTGGAGTTTTCCTGTGGAGAGGAGCCATGCCTAGAGTGGGATGGGCCAT</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4: TGTTCATCTTCTGGCCCCTGTTGTCTGCATGTAACTTAATACCACAACCAGGCATAGGGGAAAGATTGGAGGAAAGATGA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5: GTGAGAGCATCAACTTCTCTCACAACCTAGGCCAGTGTGTGGTGATGCCAGGCATGCCCTTCCCCAGCATCAGGTCTCCA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6: GAGCTGCAGAAGACGACGGCCGACTTGGATCACACTCTTGTGAGTGTCCCCAGTGTTGCAGAGGCAGGGCCATCAGGCAC</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="214-创建switchanalyzerlist对象">2.1.4 创建switchAnalyzeRlist对象<a hidden class="anchor" aria-hidden="true" href="#214-创建switchanalyzerlist对象">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">43
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">44
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">45
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">46
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">47
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">48
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">49
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">50
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">51
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">52
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">53
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">54
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">55
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">56
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">57
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">58
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sar1 &lt;- importRdata(
</span></span><span style="display:flex;"><span>  isoformCountMatrix   = salmonQuant$counts,
</span></span><span style="display:flex;"><span>  isoformRepExpression = salmonQuant$abundance,
</span></span><span style="display:flex;"><span>  designMatrix         = myDesign,
</span></span><span style="display:flex;"><span>  isoformExonAnnoation = <span style="color:#0ff;font-weight:bold">&#34;data/example.gtf.gz&#34;</span>,
</span></span><span style="display:flex;"><span>  isoformNtFasta       = <span style="color:#0ff;font-weight:bold">&#34;data/example_isoform_nt.fasta&#34;</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sar1 %&gt;% length()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 11</span>
</span></span><span style="display:flex;"><span>names(sar1)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;isoformFeatures&#34;      &#34;exons&#34;                &#34;conditions&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [4] &#34;designMatrix&#34;         &#34;sourceId&#34;             &#34;isoformCountMatrix&#34;  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [7] &#34;isoformRepExpression&#34; &#34;runInfo&#34;              &#34;orfAnalysis&#34;         </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [10] &#34;isoformRepIF&#34;         &#34;ntSequence&#34; </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sar1$isoformFeatures %&gt;% head(<span style="color:#ff0;font-weight:bold">2</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># iso_ref/gene_ref : 新的唯一编号，不重复</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gene_overall_mean/iso_overall_mean : 基因/转录本在所有样本的平均表达量</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># IF_overall(isoform fraction=isoform_exp / gene_exp): 在所有样本中，特定基因表达该转录本的比例</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#              iso_ref          gene_ref     isoform_id gene_id condition_1 condition_2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 280 isoComp_00000001 geneComp_00000001 TCONS_00000316 AADACL3        hESC         iPS</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 281 isoComp_00000002 geneComp_00000001 TCONS_00000317 AADACL3        hESC         iPS</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     gene_name gene_biotype iso_biotype gene_overall_mean gene_value_1 gene_value_2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 280   AADACL3           NA          NA          25.37693     32.66891     18.08494</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 281   AADACL3           NA          NA          25.37693     32.66891     18.08494</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     gene_stderr_1 gene_stderr_2 gene_log2_fold_change gene_q_value iso_overall_mean</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 280       10.1063      6.337321            -0.8527734           NA         3.704839</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 281       10.1063      6.337321            -0.8527734           NA        21.672089</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     iso_value_1 iso_value_2 iso_stderr_1 iso_stderr_2 iso_log2_fold_change iso_q_value</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 280    4.846923    2.562755     3.823911     2.562511           -0.9167289          NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 281   27.821989   15.522189     6.282387     3.774811           -0.8414829          NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     IF_overall   IF1     IF2      dIF isoform_switch_q_value gene_switch_q_value   PTC</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 280   0.114475 0.124 0.10495 -0.01905                     NA                  NA FALSE</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 281   0.885525 0.876 0.89505  0.01905                     NA                  NA FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sar1$isoformFeatures %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::filter(gene_id==<span style="color:#0ff;font-weight:bold">&#34;AADACL3&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::select(isoform_id, gene_id, dplyr::contains(<span style="color:#0ff;font-weight:bold">&#34;IF&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#       isoform_id gene_id IF_overall   IF1     IF2      dIF</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 TCONS_00000316 AADACL3   0.114475 0.124 0.10495 -0.01905</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 TCONS_00000317 AADACL3   0.885525 0.876 0.89505  0.0190</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#ORF转录本开放阅读框注释</span>
</span></span><span style="display:flex;"><span>sar1$orfAnalysis %&gt;% head(<span style="color:#ff0;font-weight:bold">2</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># orfTransciptStart/orfTransciptEnd 基于转录本核苷酸序列(剪切后)的起始位置</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># orfStartGenomic/orfEndGenomic 基于基因组坐标(剪切前)的起始位置</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#       isoform_id orfTransciptStart orfTransciptEnd orfTransciptLength</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 TCONS_00000001              1402            1629                228</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 TCONS_00000002               317             715                399</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   orfStarExon orfEndExon orfStartGenomic orfEndGenomic</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1           3          3           14159         14386</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2           1          3           12190         13636</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   stopDistanceToLastJunction stopIndex   PTC orf_origin</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1                      -1163         3 FALSE Annotation</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2                       -231         3 FALSE Annotation</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="22-过滤">2.2 过滤<a hidden class="anchor" aria-hidden="true" href="#22-过滤">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sar2 &lt;- preFilter(sar1)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#默认过滤标准</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># geneExpressionCutoff = 1 : 根据gene_overall_mean列过滤低表达基因</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># isoformExpressionCutoff = 0 : 根据iso_overall_mean列过滤低表达基因</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># IFcutoff=0.01: 根据IF_overall过滤</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># dIFcutoff = 0.1: 根据dIF过滤（绝对值）</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># removeSingleIsoformGenes = TRUE: 删除只有一个转录本的基因</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>dim(sar1)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 1060   30</span>
</span></span><span style="display:flex;"><span>dim(sar2)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 760  30</span>
</span></span><span style="display:flex;"><span>sar2
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># This switchAnalyzeRlist list contains:</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 760 isoforms from 208 genes</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 comparison from 2 conditions (in total 4 samples)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Feature analyzed:</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   [1] &#34;ORFs, ntSequence&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="23-isoformswitch分析">2.3 isoformSwitch分析<a hidden class="anchor" aria-hidden="true" href="#23-isoformswitch分析">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sar3 &lt;- isoformSwitchTestDEXSeq(sar2)
</span></span><span style="display:flex;"><span>sar3$isoformFeatures %&gt;% head(<span style="color:#ff0;font-weight:bold">2</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 如下，主要增加了isoform_switch_q_value与gene_switch_q_value两列信息</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     iso_ref          gene_ref     isoform_id gene_id condition_1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 557 isoComp_00000004 geneComp_00000003 TCONS_00003880   ACAP3        hESC</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 558 isoComp_00000005 geneComp_00000003 TCONS_00003881   ACAP3        hESC</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     condition_2 gene_name gene_biotype iso_biotype gene_overall_mean</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 557         iPS     ACAP3           NA          NA          910.5955</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 558         iPS     ACAP3           NA          NA          910.5955</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     gene_value_1 gene_value_2 gene_stderr_1 gene_stderr_2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 557     1771.599     49.59168        377.02      15.68552</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 558     1771.599     49.59168        377.02      15.68552</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     gene_log2_fold_change gene_q_value iso_overall_mean iso_value_1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 557             -5.158528           NA         241.0513    478.7646</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 558             -5.158528           NA         186.6521    365.0107</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     iso_value_2 iso_stderr_1 iso_stderr_2 iso_log2_fold_change</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 557    3.337951    175.44243     3.337951            -7.159924</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 558    8.293521     89.61604     1.997000            -5.458111</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     iso_q_value IF_overall     IF1     IF2      dIF</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 557          NA   0.156075 0.26100 0.05115 -0.20985</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 558          NA   0.188100 0.20455 0.17165 -0.03290</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     isoform_switch_q_value gene_switch_q_value   PTC</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 557              0.2601812          0.02341811 FALSE</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 558              0.7079831          0.02341811 FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#如下结果gene_switch_q_value即为isoform_switch_q_value的最小值</span>
</span></span><span style="display:flex;"><span>sar3$isoformFeatures %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::filter(gene_id==<span style="color:#0ff;font-weight:bold">&#34;ACAP3&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::select(isoform_id, gene_id,dIF, 
</span></span><span style="display:flex;"><span>                dplyr::contains(<span style="color:#0ff;font-weight:bold">&#34;switch&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#       isoform_id gene_id      dIF isoform_switch_q_value gene_switch_q_value</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 TCONS_00003880   ACAP3 -0.20985             0.26018123          0.02341811</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 TCONS_00003881   ACAP3 -0.03290             0.70798307          0.02341811</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 TCONS_00003882   ACAP3  0.29825             0.02341811          0.02341811</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 TCONS_00003883   ACAP3 -0.05555             0.58385549          0.02341811</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>switchPlot(sar3, gene = <span style="color:#0ff;font-weight:bold">&#39;ACAP3&#39;</span>,
</span></span><span style="display:flex;"><span>           condition1=<span style="color:#0ff;font-weight:bold">&#34;hESC&#34;</span>,
</span></span><span style="display:flex;"><span>           condition2=<span style="color:#0ff;font-weight:bold">&#34;iPS&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://s6.jpg.cm/2022/02/02/LwNiaw.png" alt="LwNiaw.png" style="zoom:50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">#四个子图</span>
</span></span><span style="display:flex;"><span>switchPlotGeneExp(sar3, gene = <span style="color:#0ff;font-weight:bold">&#39;ACAP3&#39;</span>,
</span></span><span style="display:flex;"><span>                  condition1=<span style="color:#0ff;font-weight:bold">&#34;hESC&#34;</span>,
</span></span><span style="display:flex;"><span>                  condition2=<span style="color:#0ff;font-weight:bold">&#34;iPS&#34;</span>)
</span></span><span style="display:flex;"><span>switchPlotIsoExp(sar3, gene = <span style="color:#0ff;font-weight:bold">&#39;ACAP3&#39;</span>,
</span></span><span style="display:flex;"><span>                  condition1=<span style="color:#0ff;font-weight:bold">&#34;hESC&#34;</span>,
</span></span><span style="display:flex;"><span>                  condition2=<span style="color:#0ff;font-weight:bold">&#34;iPS&#34;</span>)
</span></span><span style="display:flex;"><span>switchPlotIsoUsage(sar3, gene = <span style="color:#0ff;font-weight:bold">&#39;ACAP3&#39;</span>,
</span></span><span style="display:flex;"><span>                 condition1=<span style="color:#0ff;font-weight:bold">&#34;hESC&#34;</span>,
</span></span><span style="display:flex;"><span>                 condition2=<span style="color:#0ff;font-weight:bold">&#34;iPS&#34;</span>)
</span></span><span style="display:flex;"><span>switchPlotTranscript(sar3, gene = <span style="color:#0ff;font-weight:bold">&#39;ACAP3&#39;</span>,
</span></span><span style="display:flex;"><span>                   condition1=<span style="color:#0ff;font-weight:bold">&#34;hESC&#34;</span>,
</span></span><span style="display:flex;"><span>                   condition2=<span style="color:#0ff;font-weight:bold">&#34;iPS&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="24-isoformswitch结果探索">2.4 isoformSwitch结果探索<a hidden class="anchor" aria-hidden="true" href="#24-isoformswitch结果探索">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sar3$isoformSwitchAnalysis %&gt;% head()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            iso_ref          gene_ref     isoform_id condition_1 condition_2      dIF       pvalue</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 isoComp_00000922 geneComp_00000294 TCONS_00000007        hESC         iPS  0.01615 8.645971e-01</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 isoComp_00000923 geneComp_00000294 TCONS_00000008        hESC         iPS -0.25885 4.859083e-02</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 isoComp_00000924 geneComp_00000294 TCONS_00000009        hESC         iPS  0.24275 7.547031e-05</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 isoComp_00000929 geneComp_00000298 TCONS_00000017        hESC         iPS  0.08785 7.690961e-01</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5 isoComp_00000930 geneComp_00000298 TCONS_00000018        hESC         iPS -0.47220 6.113540e-14</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6 isoComp_00000931 geneComp_00000298 TCONS_00000019        hESC         iPS  0.23965 2.119346e-03</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#           padj     IF1     IF2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 9.665421e-01 0.51665 0.53280</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 1.464787e-01 0.48320 0.22435</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 7.536969e-04 0.00015 0.24290</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 9.355274e-01 0.00000 0.08785</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5 3.815868e-12 0.47220 0.00000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6 1.250306e-02 0.00000 0.23965</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="241-top-switchesgene">2.4.1 Top Switches&ndash;gene<a hidden class="anchor" aria-hidden="true" href="#241-top-switchesgene">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sar.ge_df=extractTopSwitches( 
</span></span><span style="display:flex;"><span>  sar3, 
</span></span><span style="display:flex;"><span>  filterForConsequences = <span style="color:#fff;font-weight:bold">FALSE</span>, 
</span></span><span style="display:flex;"><span>  n = <span style="color:#fff;font-weight:bold">NA</span>,                
</span></span><span style="display:flex;"><span>  extractGenes = <span style="color:#fff;font-weight:bold">TRUE</span>,    <span style="color:#007f7f"># when FALSE isoforms are returned</span>
</span></span><span style="display:flex;"><span>  sortByQvals = <span style="color:#fff;font-weight:bold">TRUE</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>dim(sar.ge_df)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 68  7</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>head(sar.ge_df)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#             gene_ref     gene_id gene_name condition_1 condition_2 gene_switch_q_value Rank</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1  geneComp_00000088        ENO1      ENO1        hESC         iPS        2.429932e-90    1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4  geneComp_00000324 XLOC_001217      &lt;NA&gt;        hESC         iPS        1.464063e-24    2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5  geneComp_00000171        NBL1      NBL1        hESC         iPS        1.039215e-21    3</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 7  geneComp_00000053       CASP9     CASP9        hESC         iPS        5.232094e-18    4</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 10 geneComp_00000282        UBR4      UBR4        hESC         iPS        1.597415e-17    5</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 14 geneComp_00000072       CROCC     CROCC        hESC         iPS        1.752050e-17    6</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="242-top-switchesisoform">2.4.2 Top Switches&ndash;isoform<a hidden class="anchor" aria-hidden="true" href="#242-top-switchesisoform">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sar.iso_df=extractTopSwitches( 
</span></span><span style="display:flex;"><span>  sar3, 
</span></span><span style="display:flex;"><span>  filterForConsequences = <span style="color:#fff;font-weight:bold">FALSE</span>, 
</span></span><span style="display:flex;"><span>  n = <span style="color:#fff;font-weight:bold">NA</span>,                
</span></span><span style="display:flex;"><span>  extractGenes = <span style="color:#fff;font-weight:bold">FALSE</span>,    <span style="color:#007f7f"># when FALSE isoforms are returned</span>
</span></span><span style="display:flex;"><span>  sortByQvals = <span style="color:#fff;font-weight:bold">TRUE</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>dim(sar.iso_df)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 121  12</span>
</span></span><span style="display:flex;"><span>head(sar.iso_df)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            iso_ref          gene_ref     isoform_id     gene_id gene_name condition_1 condition_2   IF1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 isoComp_00000297 geneComp_00000088 TCONS_00004087        ENO1      ENO1        hESC         iPS 0.157</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 isoComp_00000985 geneComp_00000324 TCONS_00003819 XLOC_001217      &lt;NA&gt;        hESC         iPS 0.084</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 isoComp_00000551 geneComp_00000171 TCONS_00000463        NBL1      NBL1        hESC         iPS 0.112</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 isoComp_00000549 geneComp_00000171 TCONS_00000460        NBL1      NBL1        hESC         iPS 0.881</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5 isoComp_00000152 geneComp_00000053 TCONS_00004180       CASP9     CASP9        hESC         iPS 0.474</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6 isoComp_00000242 geneComp_00000072 TCONS_00000420       CROCC     CROCC        hESC         iPS 0.683</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     IF2    dIF isoform_switch_q_value Rank</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 0.000 -0.157           2.429932e-90    1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 0.434  0.350           1.464063e-24    2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 1.000  0.888           1.039215e-21    3</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 0.000 -0.881           1.194327e-19    4</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5 0.000 -0.474           5.232094e-18    5</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6 0.173 -0.510           1.752050e-17    6</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="243-火山图">2.4.3 火山图<a hidden class="anchor" aria-hidden="true" href="#243-火山图">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>ggplot(data=sar3$isoformFeatures, aes(x=dIF, y=-log10(isoform_switch_q_value))) +
</span></span><span style="display:flex;"><span>  geom_point(
</span></span><span style="display:flex;"><span>    aes( color=abs(dIF) &gt; <span style="color:#ff0;font-weight:bold">0.1</span> &amp; isoform_switch_q_value &lt; <span style="color:#ff0;font-weight:bold">0.05</span> ), <span style="color:#007f7f"># default cutoff</span>
</span></span><span style="display:flex;"><span>    size=<span style="color:#ff0;font-weight:bold">1</span>
</span></span><span style="display:flex;"><span>  ) +
</span></span><span style="display:flex;"><span>  geom_hline(yintercept = -log10(<span style="color:#ff0;font-weight:bold">0.05</span>), linetype=<span style="color:#0ff;font-weight:bold">&#39;dashed&#39;</span>) + <span style="color:#007f7f"># default cutoff</span>
</span></span><span style="display:flex;"><span>  geom_vline(xintercept = c(<span style="color:#ff0;font-weight:bold">-0.1</span>, <span style="color:#ff0;font-weight:bold">0.1</span>), linetype=<span style="color:#0ff;font-weight:bold">&#39;dashed&#39;</span>) + <span style="color:#007f7f"># default cutoff</span>
</span></span><span style="display:flex;"><span>  scale_color_manual(<span style="color:#0ff;font-weight:bold">&#39;Signficant\nIsoform Switch&#39;</span>, values = c(<span style="color:#0ff;font-weight:bold">&#39;black&#39;</span>,<span style="color:#0ff;font-weight:bold">&#39;red&#39;</span>)) +
</span></span><span style="display:flex;"><span>  labs(x=<span style="color:#0ff;font-weight:bold">&#39;dIF&#39;</span>, y=<span style="color:#0ff;font-weight:bold">&#39;-Log10 ( Isoform Switch Q Value )&#39;</span>) +
</span></span><span style="display:flex;"><span>  theme_bw()
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://s6.jpg.cm/2022/02/02/LwNDe8.png" alt="LwNDe8.png" style="zoom:50%;" />
<h1 id="3进阶分析">3、进阶分析<a hidden class="anchor" aria-hidden="true" href="#3进阶分析">#</a></h1>
<p>对于不同组样本的某一基因，存在其中一种isoform相对上调，另一种isoform相对下调的情况。可以对isoform的特点进行分析，从而进一步分析isoformswitch的影响与意义。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sar3
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># This switchAnalyzeRlist list contains:</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 312 isoforms from 68 genes</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 comparison from 2 conditions (in total 4 samples)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Switching features:</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Comparison Isoforms Switches Genes</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 hESC vs iPS      121       95    68</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Feature analyzed:</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;Isoform Switch Identification, ORFs, ntSequence&#34;</span>
</span></span><span style="display:flex;"><span>names(sar3)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;isoformFeatures&#34;       &#34;exons&#34;                 &#34;conditions&#34;            &#34;designMatrix&#34;         </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;sourceId&#34;              &#34;isoformCountMatrix&#34;    &#34;isoformRepExpression&#34;  &#34;runInfo&#34;              </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [9] &#34;orfAnalysis&#34;           &#34;isoformRepIF&#34;          &#34;ntSequence&#34; </span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="31-可变剪切alternative-splicing">3.1 可变剪切Alternative Splicing<a hidden class="anchor" aria-hidden="true" href="#31-可变剪切alternative-splicing">#</a></h2>
<img src="https://nanoporetech.com/sites/default/files/s3/Figure-splice-varation.png" alt="Six alternative splicing categories" style="zoom: 10%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sar4 &lt;- analyzeAlternativeSplicing(sar3)
</span></span><span style="display:flex;"><span>sar4$AlternativeSplicingAnalysis %&gt;% head(<span style="color:#ff0;font-weight:bold">2</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#       isoform_id ES ES_genomic_start ES_genomic_end MEE MEE_genomic_start MEE_genomic_end MES</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 TCONS_00003880  0             &lt;NA&gt;           &lt;NA&gt;   0              &lt;NA&gt;            &lt;NA&gt;   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 TCONS_00003881  1          1230098        1230196   0              &lt;NA&gt;            &lt;NA&gt;   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   MES_genomic_start MES_genomic_end IR IR_genomic_start IR_genomic_end A5 A5_genomic_start A5_genomic_end</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1              &lt;NA&gt;            &lt;NA&gt;  0             &lt;NA&gt;           &lt;NA&gt;  0             &lt;NA&gt;           &lt;NA&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2              &lt;NA&gt;            &lt;NA&gt;  0             &lt;NA&gt;           &lt;NA&gt;  0             &lt;NA&gt;           &lt;NA&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   A3 A3_genomic_start A3_genomic_end ATSS</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1  0             &lt;NA&gt;           &lt;NA&gt;    1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2  0             &lt;NA&gt;           &lt;NA&gt;    1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                                                                        ATSS_genomic_start</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 1234725;1235211;1235353;1235538;1235889;1237368;1238302;1238542;1239466;1243149;1244822</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2                                                                         1243149;1244822</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                                                                          ATSS_genomic_end ATTS</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 1234736;1235285;1235448;1235582;1236072;1237426;1238355;1238661;1241309;1243269;1244989    0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2                                                                         1243269;1244989    0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   ATTS_genomic_start ATTS_genomic_end</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1               &lt;NA&gt;             &lt;NA&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2               &lt;NA&gt;             &lt;NA&gt;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>extractSplicingSummary(sar4)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 一个significant isoform如果发生了A3事件，如果其dIF&gt;0,则iPS相对hESC used more;反之used less</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://s6.jpg.cm/2022/02/02/LwNMri.png" alt="LwNMri.png" style="zoom:50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>extractSplicingEnrichment(sar4)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://s6.jpg.cm/2022/02/02/LwSGWe.png" alt="LwSGWe.png" style="zoom:50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-fallback" data-lang="fallback"><span style="display:flex;"><span>extractSplicingGenomeWide(sar4)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://s6.jpg.cm/2022/02/02/LwSlW4.png" alt="LwSlW4.png" style="zoom: 80%;" />
<h2 id="32-isoform-consequences预测">3.2 isoform Consequences预测<a hidden class="anchor" aria-hidden="true" href="#32-isoform-consequences预测">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sar5 &lt;- extractSequence(sar4,
</span></span><span style="display:flex;"><span>                        writeToFile=<span style="color:#fff;font-weight:bold">TRUE</span>,
</span></span><span style="display:flex;"><span>                        pathToOutput = <span style="color:#0ff;font-weight:bold">&#39;data/output&#39;</span>)
</span></span><span style="display:flex;"><span>list.files(<span style="color:#0ff;font-weight:bold">&#34;data/output/&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;isoformSwitchAnalyzeR_isoform_AA.fasta&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [2] &#34;isoformSwitchAnalyzeR_isoform_nt.fasta&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sar5$ntSequence %&gt;% head(<span style="color:#ff0;font-weight:bold">3</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#DNAStringSet object of length 6:</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     width seq                                names               </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1]  2750 GGGTCTCCCTCTGTTG...CCCACGCGGACAGAG TCONS_00000007</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [2]  4369 TTACTGTTGATTGTGA...AAAATATCGCCCACG TCONS_00000008</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [3]  4272 TTACTGTTGATTGTGA...AAAATATCGCCCACG TCONS_00000009</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sar5$aaSequence %&gt;% head(<span style="color:#ff0;font-weight:bold">3</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#AAStringSet object of length 3:</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     width seq                                names               </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1]   389 MLLPPGSLSRPRTFSS...QAQLLPHSGPFRPNS TCONS_00000007</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [2]   389 MLLPPGSLSRPRTFSS...QAQLLPHSGPFRPNS TCONS_00000008</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [3]   389 MLLPPGSLSRPRTFSS...QAQLLPHSGPFRPNS TCONS_00000009</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>CPAT : The Coding-Potential Assessment Tool which is a tool for predicting whether an isoform is coding or not. (_nt.fasta)</li>
<li>CPC2 : The Coding Potential Calculator 2 which is a tool for predicting whether an isoform is coding or not. (_nt.fasta)</li>
<li>Pfam : Prediction of protein domains (_AA.fasta)</li>
<li>SignalP : Prediction of Signal Peptides(_AA.fasta)</li>
<li>IUPred2A: Predicts Intrinsically Disordered Regions (IDR) and Intrinsically Disordered Binding Regions (IDBR)(_AA.fasta)</li>
<li>NetSurfP-2 : Prediction of Intrinsically Disordered Regions (IDR) (_AA.fasta)</li>
</ul>
<p>上传核苷酸/氨基酸序列到相应网站，下载预测的结果，再导入到switchAnalyzeRlist对象中</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">43
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">44
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">45
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">46
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">47
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">48
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">49
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">50
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">51
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">52
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">53
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">54
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">55
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">56
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">57
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">58
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">59
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">60
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sar6 &lt;- analyzeCPAT(
</span></span><span style="display:flex;"><span>  switchAnalyzeRlist   = sar5,
</span></span><span style="display:flex;"><span>  pathToCPATresultFile = <span style="color:#0ff;font-weight:bold">&#34;data/external/cpat_results.txt&#34;</span>,
</span></span><span style="display:flex;"><span>  codingCutoff         = <span style="color:#ff0;font-weight:bold">0.725</span>,
</span></span><span style="display:flex;"><span>  removeNoncodinORFs   = <span style="color:#fff;font-weight:bold">TRUE</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>sar6 &lt;- analyzePFAM(
</span></span><span style="display:flex;"><span>  switchAnalyzeRlist   = sar6,
</span></span><span style="display:flex;"><span>  pathToPFAMresultFile = <span style="color:#0ff;font-weight:bold">&#34;data/external/pfam_results.txt&#34;</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>sar6 &lt;- analyzeSignalP(
</span></span><span style="display:flex;"><span>  switchAnalyzeRlist       = sar6,
</span></span><span style="display:flex;"><span>  pathToSignalPresultFile  = <span style="color:#0ff;font-weight:bold">&#34;data/external/signalP_results.txt&#34;</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sar6 &lt;- analyzeIUPred2A(
</span></span><span style="display:flex;"><span>  switchAnalyzeRlist        = sar6,
</span></span><span style="display:flex;"><span>  pathToIUPred2AresultFile = <span style="color:#0ff;font-weight:bold">&#34;data/external/iupred2a_result.txt.gz&#34;</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sar7 &lt;- analyzeSwitchConsequences(sar6)
</span></span><span style="display:flex;"><span>length(sar7)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 18</span>
</span></span><span style="display:flex;"><span>names(sar7)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;isoformFeatures&#34;             &#34;exons&#34;                      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [3] &#34;conditions&#34;                  &#34;designMatrix&#34;               </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;sourceId&#34;                    &#34;isoformCountMatrix&#34;         </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [7] &#34;isoformRepExpression&#34;        &#34;runInfo&#34;                    </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [9] &#34;orfAnalysis&#34;                 &#34;isoformRepIF&#34;               </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [11] &#34;ntSequence&#34;                  &#34;isoformSwitchAnalysis&#34;      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [13] &#34;AlternativeSplicingAnalysis&#34; &#34;aaSequence&#34;                 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [15] &#34;domainAnalysis&#34;              &#34;signalPeptideAnalysis&#34;      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [17] &#34;idrAnalysis&#34;        &#34;switchConsequence&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sar7$switchConsequence %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::filter(isoformsDifferent==<span style="color:#0ff;font-weight:bold">&#34;TRUE&#34;</span>) %&gt;% head()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            gene_ref gene_id gene_name condition_1 condition_2 isoformUpregulated</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 geneComp_00000003   ACAP3     ACAP3        hESC         iPS     TCONS_00003882</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 geneComp_00000004   ACOT7     ACOT7        hESC         iPS     TCONS_00004035</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 geneComp_00000004   ACOT7     ACOT7        hESC         iPS     TCONS_00004035</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 geneComp_00000004   ACOT7     ACOT7        hESC         iPS     TCONS_00004035</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5 geneComp_00000004   ACOT7     ACOT7        hESC         iPS     TCONS_00004035</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6 geneComp_00000004   ACOT7     ACOT7        hESC         iPS     TCONS_00004036</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   isoformDownregulated       iso_ref_up     iso_ref_down    featureCompared</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1       TCONS_00003880 isoComp_00000006 isoComp_00000004 ORF_seq_similarity</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2       TCONS_00004039 isoComp_00000010 isoComp_00000014 ORF_seq_similarity</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3       TCONS_00004039 isoComp_00000010 isoComp_00000014         NMD_status</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4       TCONS_00004039 isoComp_00000010 isoComp_00000014 domains_identified</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5       TCONS_00004039 isoComp_00000010 isoComp_00000014     IDR_identified</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6       TCONS_00004039 isoComp_00000011 isoComp_00000014         NMD_status</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   isoformsDifferent switchConsequence</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1              TRUE     ORF is longer</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2              TRUE     ORF is longer</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3              TRUE   NMD insensitive</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4              TRUE       Domain gain</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5              TRUE          IDR gain</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6              TRUE   NMD insensitive</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>extractConsequenceSummary(sar7)
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://s6.jpg.cm/2022/02/02/LwS1sh.png" alt="LwS1sh.png"  />
</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>extractConsequenceEnrichment(sar7)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://s6.jpg.cm/2022/02/02/LwSHhS.png" alt="LwSHhS.png" style="zoom:50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>switchPlot(sar7, gene = <span style="color:#0ff;font-weight:bold">&#39;ACAP3&#39;</span>,
</span></span><span style="display:flex;"><span>           condition1=<span style="color:#0ff;font-weight:bold">&#34;hESC&#34;</span>,
</span></span><span style="display:flex;"><span>           condition2=<span style="color:#0ff;font-weight:bold">&#34;iPS&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://s6.jpg.cm/2022/02/02/LwStnW.png" alt="LwStnW.png" style="zoom: 67%;" />


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">转录组</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/077%E8%82%BF%E7%98%A4%E4%BA%9A%E5%9E%8B%E5%88%86%E7%BE%A4%E5%B7%A5%E5%85%B7consensusclusterplus/">
    <span class="title">« Prev Page</span>
    <br>
    <span>肿瘤亚型分群工具ConsensusClusterPlus</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/078%E8%AF%BA%E6%A8%A1%E5%9B%BE%E6%88%96%E5%88%97%E7%BA%BF%E5%9B%BE%E7%BB%98%E5%88%B6%E5%88%86%E6%9E%90/">
    <span class="title">Next Page »</span>
    <br>
    <span>诺模图或列线图绘制分析</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
