<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--cellranger | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
单细胞转录组测序技术历经10余年的发展，目前以高通量的Droplet技术广为应用。其中以10X Genomics公司开发的实验及分析系统最为流行。如下简单学习测序原理、结果以及上游比对流程。
">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/005%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellranger%E6%AF%94%E5%AF%B9/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/005%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellranger%E6%AF%94%E5%AF%B9/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--cellranger" />
<meta property="og:description" content="
单细胞转录组测序技术历经10余年的发展，目前以高通量的Droplet技术广为应用。其中以10X Genomics公司开发的实验及分析系统最为流行。如下简单学习测序原理、结果以及上游比对流程。
" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/005%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellranger%E6%AF%94%E5%AF%B9/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-05-24T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-01-25T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--cellranger"/>
<meta name="twitter:description" content="
单细胞转录组测序技术历经10余年的发展，目前以高通量的Droplet技术广为应用。其中以10X Genomics公司开发的实验及分析系统最为流行。如下简单学习测序原理、结果以及上游比对流程。
"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--cellranger",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/005%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellranger%E6%AF%94%E5%AF%B9/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--cellranger",
  "name": "单细胞分析工具--cellranger",
  "description": " 单细胞转录组测序技术历经10余年的发展，目前以高通量的Droplet技术广为应用。其中以10X Genomics公司开发的实验及分析系统最为流行。如下简单学习测序原理、结果以及上游比对流程。\r",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " 单细胞转录组测序技术历经10余年的发展，目前以高通量的Droplet技术广为应用。其中以10X Genomics公司开发的实验及分析系统最为流行。如下简单学习测序原理、结果以及上游比对流程。\r1、测序原理 （1）如下图所示，首先每个凝胶微珠(Gel Beads)表面提前接上大量定制的DNA片段。该片段主要由3部分组成：10x Barcode、UMI以及Poly(dT)尾巴。\n10x Barcode：长度为16的核苷酸序列，用于区分不同的微珠(细胞)； UMI：长度为10的核苷酸序列，用于区分不同的cDNA分子； Poly(dT)：用于结合mRNA的Poly(A)尾巴； （2）理想情况下，一个微珠会捕捉到一个细胞，并形成油包水的封闭体系。然后在酶反应作用下，裂解细胞，释放的mRNA分子结合到微珠表面的DNA片段上；最后逆转录形成相应的cDNA分子。\n（3）提取所有水相里的cDNA，进行PCR扩增，之后接入illumina高通量双端测序分析。\nRead 1，即R1，通常表示长度为26的10X barcode与UMI标签序列； Read 2，即R2，通常为长度为98的mRNA测序片段。 参考https://www.biostars.org/p/9529864/#9529942，如果下载的R1、R2文件为150bp长度，也可直接分析。 UMI标签主要用于消除PCR扩增带来的误差，即表示细胞裂解是结合到微珠的mRNA分子数（即cDNA分子数）。\nSeurat V2的meta.data最初两列：nUMI， nGenes\n后续更新的Seurat V3,4的列名分别改为：nCount_RNA，nFeature_RNA\n2、cellranger比对 示例数据：GSE178911共4个样本，每个样本有4个lane的比对结果(SRRxxxxxx id)\n（1）原始测序数据fastq.gz\nhttps://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/fastq-input\n如上测序原理的双端测序，一个样本通常会产生一对fastq.gz结果； 若同一样本在不同的lane分别测序，则会产生若干对的fastq.gz结果。 如下示例，为样本名为MySample的测序结果。\n1 2 3 4 5 6 7 PROJECT_FOLDER |-- MySample_S1_L001_I1_001.fastq.gz |-- MySample_S1_L001_R1_001.fastq.gz |-- MySample_S1_L001_R2_001.fastq.gz |-- MySample_S1_L002_I1_001.fastq.gz |-- MySample_S1_L002_R1_001.fastq.gz |-- MySample_S1_L002_R2_001.fastq.gz I1文件表示sample index，用于区分不同样本的标签，在后序的比对分析中可不提供。\n在分析公共数据时，可使用aspera或者prefetch方式下载原始数据；并修改成上述的规范文件名。\n（2）cellranger比对环境\nhttps://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest?\n在上述网址内下载比对软件cellranger与相应物种的参考数据集，解压即可使用。\n1 2 3 4 5 6 7 8 9 10 cellranger-7.1.0/bin/cellranger --version # cellranger cellranger-7.1.0 tree -L 1 refdata-gex-GRCh38-2020-A/ # refdata-gex-GRCh38-2020-A/ # ├── fasta # ├── genes # ├── pickle # ├── reference.json # └── star （3）cellranger比对命令\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 bin=/path/to/cellranger-7.1.0/bin/cellranger ref=/path/to/cellranger-7.1.0/refer/refdata-gex-GRCh38-2020-A fq_dir=/path/to/PROJECT_FOLDER $bin count \\ --fastqs=${fq_dir} \\ --sample=MySample_S1 \\ --transcriptome=$ref \\ --id=MySample_S1 \\ --localcores=10 \\ --no-bam \\ --nosecondary # --fastqs 交代测速数据路径（文件夹名） # --sample 交代待比对样本名的前缀（因为该文件夹内可能有许多样本的测序数据） # --transcriptome 交代参考基因组文件夹 # --id 交代储存结果的文件夹，如果没有会自动创建 # --localcores 多线程 # --no-bam 不生成bam文件(视情况而定：如果进行RNA速率分析等，需要保留bam文件) # --nosecondary 不进行后续分析 ##输出结果一般在 ${id} 的out文件夹 （4）比对结果输出\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 tree -L 2 MySample_S1 # MySample_S1 # ├── MySample_S1.mri.tgz # ├── _cmdline # ├── _filelist # ├── _finalstate # ├── _invocation # ├── _jobmode # ├── _log # ├── _mrosource # ├── outs # │ ├── cellsorted_possorted_genome_bam.bam # │ ├── filtered_feature_bc_matrix # │ ├── filtered_feature_bc_matrix.h5 # │ ├── metrics_summary.csv # │ ├── molecule_info.h5 # │ ├── possorted_genome_bam.bam # │ ├── possorted_genome_bam.bam.bai # │ ├── raw_feature_bc_matrix # │ ├── raw_feature_bc_matrix.h5 # │ └── web_summary.html # ├── _perf # ├── SC_RNA_COUNTER_CS # │ ├── CELLRANGER_PREFLIGHT # │ ├── CELLRANGER_PREFLIGHT_LOCAL # │ ├── fork0 # │ ├── FULL_COUNT_INPUTS # │ ├── GET_AGGREGATE_BARCODES_OUT # │ ├── SC_MULTI_CORE # │ ├── _STRUCTIFY # │ └── WRITE_GENE_INDEX # ├── _sitecheck # ├── _tags # ├── _timestamp # ├── _uuid # ├── _vdrkill # └── _versions ",
  "wordCount" : "1522",
  "inLanguage": "en",
  "datePublished": "2022-05-24T00:00:00Z",
  "dateModified": "2023-01-25T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/005%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellranger%E6%AF%94%E5%AF%B9/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--cellranger
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2022-05-24 00:00:00 &#43;0000 UTC&#39;&gt;2022-05-24&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-25&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1522&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e6%b5%8b%e5%ba%8f%e5%8e%9f%e7%90%86" aria-label="1、测序原理">1、测序原理</a></li>
                    <li>
                        <a href="#2cellranger%e6%af%94%e5%af%b9" aria-label="2、cellranger比对">2、cellranger比对</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>单细胞转录组测序技术历经10余年的发展，目前以高通量的Droplet技术广为应用。其中以10X Genomics公司开发的实验及分析系统最为流行。如下简单学习测序原理、结果以及上游比对流程。<img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/01-technologies.jpg" alt="Tools and techniques for single-cell RNA sequencing data"  />
</p></blockquote>
<h2 id="1测序原理">1、测序原理<a hidden class="anchor" aria-hidden="true" href="#1测序原理">#</a></h2>
<p>（1）如下图所示，首先每个凝胶微珠(Gel Beads)表面提前接上大量定制的DNA片段。该片段主要由3部分组成：10x Barcode、UMI以及Poly(dT)尾巴。</p>
<ul>
<li>10x Barcode：长度为16的核苷酸序列，用于区分不同的微珠(细胞)；</li>
<li>UMI：长度为10的核苷酸序列，用于区分不同的cDNA分子；</li>
<li>Poly(dT)：用于结合mRNA的Poly(A)尾巴；</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230125180226221.png" alt="image-20230125180226221" style="zoom:67%;" />
<p>（2）理想情况下，一个微珠会捕捉到一个细胞，并形成油包水的封闭体系。然后在酶反应作用下，裂解细胞，释放的mRNA分子结合到微珠表面的DNA片段上；最后逆转录形成相应的cDNA分子。</p>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230125183220166.png" alt="image-20230125183220166" style="zoom: 25%;" />
<p>（3）提取所有水相里的cDNA，进行PCR扩增，之后接入illumina高通量双端测序分析。</p>
<ul>
<li>Read 1，即R1，通常表示长度为26的10X barcode与UMI标签序列；</li>
<li>Read 2，即R2，通常为长度为98的mRNA测序片段。</li>
<li>参考https://www.biostars.org/p/9529864/#9529942，如果下载的R1、R2文件为150bp长度，也可直接分析。</li>
</ul>
<blockquote>
<p>UMI标签主要用于消除PCR扩增带来的误差，即表示细胞裂解是结合到微珠的mRNA分子数（即cDNA分子数）。</p>
<p>Seurat V2的meta.data最初两列：nUMI， nGenes</p>
<p>后续更新的Seurat V3,4的列名分别改为：nCount_RNA，nFeature_RNA</p></blockquote>
<h2 id="2cellranger比对">2、cellranger比对<a hidden class="anchor" aria-hidden="true" href="#2cellranger比对">#</a></h2>
<blockquote>
<p>示例数据：GSE178911共4个样本，每个样本有4个lane的比对结果(SRRxxxxxx id)</p></blockquote>
<p>（1）原始测序数据fastq.gz</p>
<p><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/fastq-input">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/using/fastq-input</a></p>
<ul>
<li>如上测序原理的双端测序，一个样本通常会产生一对fastq.gz结果；</li>
<li>若同一样本在不同的lane分别测序，则会产生若干对的fastq.gz结果。</li>
</ul>
<p>如下示例，为样本名为MySample的测序结果。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-fallback" data-lang="fallback"><span style="display:flex;"><span>PROJECT_FOLDER
</span></span><span style="display:flex;"><span>|-- MySample_S1_L001_I1_001.fastq.gz
</span></span><span style="display:flex;"><span>|-- MySample_S1_L001_R1_001.fastq.gz
</span></span><span style="display:flex;"><span>|-- MySample_S1_L001_R2_001.fastq.gz  
</span></span><span style="display:flex;"><span>|-- MySample_S1_L002_I1_001.fastq.gz
</span></span><span style="display:flex;"><span>|-- MySample_S1_L002_R1_001.fastq.gz
</span></span><span style="display:flex;"><span>|-- MySample_S1_L002_R2_001.fastq.gz
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>I1文件表示sample index，用于区分不同样本的标签，在后序的比对分析中可不提供。</p></blockquote>
<p>在分析公共数据时，可使用aspera或者prefetch方式下载原始数据；并修改成上述的规范文件名。</p>
<p>（2）cellranger比对环境</p>
<p><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest">https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest</a>?</p>
<p>在上述网址内下载比对软件cellranger与相应物种的参考数据集，解压即可使用。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>cellranger-7.1.0/bin/cellranger --version
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cellranger cellranger-7.1.0</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>tree -L <span style="color:#ff0;font-weight:bold">1</span> refdata-gex-GRCh38-2020-A/
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># refdata-gex-GRCh38-2020-A/</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── fasta</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── genes</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── pickle</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── reference.json</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># └── star</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>（3）cellranger比对命令</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>bin=/path/to/cellranger-7.1.0/bin/cellranger 
</span></span><span style="display:flex;"><span>ref=/path/to/cellranger-7.1.0/refer/refdata-gex-GRCh38-2020-A
</span></span><span style="display:flex;"><span>fq_dir=/path/to/PROJECT_FOLDER
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>$bin count <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--fastqs=<span style="color:#0ff;font-weight:bold">${</span>fq_dir<span style="color:#0ff;font-weight:bold">}</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--sample=MySample_S1 <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--transcriptome=$ref <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--id=MySample_S1 <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--localcores=<span style="color:#ff0;font-weight:bold">10</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--no-bam <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--nosecondary 
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># --fastqs 交代测速数据路径（文件夹名）</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># --sample 交代待比对样本名的前缀（因为该文件夹内可能有许多样本的测序数据）</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># --transcriptome 交代参考基因组文件夹</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># --id 交代储存结果的文件夹，如果没有会自动创建</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># --localcores 多线程</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># --no-bam 不生成bam文件(视情况而定：如果进行RNA速率分析等，需要保留bam文件)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># --nosecondary 不进行后续分析</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##输出结果一般在 ${id} 的out文件夹</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>（4）比对结果输出</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>tree -L <span style="color:#ff0;font-weight:bold">2</span> MySample_S1
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MySample_S1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── MySample_S1.mri.tgz</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _cmdline</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _filelist</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _finalstate</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _invocation</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _jobmode</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _log</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _mrosource</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── outs</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── cellsorted_possorted_genome_bam.bam</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── filtered_feature_bc_matrix</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── filtered_feature_bc_matrix.h5</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── metrics_summary.csv</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── molecule_info.h5</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── possorted_genome_bam.bam</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── possorted_genome_bam.bam.bai</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── raw_feature_bc_matrix</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── raw_feature_bc_matrix.h5</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   └── web_summary.html</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _perf</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── SC_RNA_COUNTER_CS</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── CELLRANGER_PREFLIGHT</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── CELLRANGER_PREFLIGHT_LOCAL</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── fork0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── FULL_COUNT_INPUTS</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── GET_AGGREGATE_BARCODES_OUT</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── SC_MULTI_CORE</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── _STRUCTIFY</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   └── WRITE_GENE_INDEX</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _sitecheck</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _tags</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _timestamp</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _uuid</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── _vdrkill</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># └── _versions</span>
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/basic/005%E7%BB%9F%E8%AE%A1%E5%AD%A6%E5%9F%BA%E7%A1%80--%E4%BA%8C%E5%88%86%E7%B1%BB%E6%A8%A1%E5%9E%8B%E8%AF%84%E4%BB%B7%E6%8C%87%E6%A0%87/">
    <span class="title">« Prev Page</span>
    <br>
    <span>统计学基础--二分类模型评价指标</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/program/005python%E5%9F%BA%E7%A1%80-5%E7%B1%BBclass%E4%B8%8E%E5%AE%9E%E4%BE%8Binstance/">
    <span class="title">Next Page »</span>
    <br>
    <span>Python基础-(5)类Class与实例Instance</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
