<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--Seurat5基础用法 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
Seurat V5版本有一段时间了，由于时间原因未来得及了解。现根据其官方文档简单整理其用法，与V4比较类似的地方就不多叙述了。此外，V5的亮点之一还在于单细胞多组学的整合分析，此次就不做记录了。（PS：中秋快乐~）">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/041%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat5%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/041%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat5%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--Seurat5基础用法" />
<meta property="og:description" content="
Seurat V5版本有一段时间了，由于时间原因未来得及了解。现根据其官方文档简单整理其用法，与V4比较类似的地方就不多叙述了。此外，V5的亮点之一还在于单细胞多组学的整合分析，此次就不做记录了。（PS：中秋快乐~）" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/041%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat5%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2024-09-17T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2024-09-17T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--Seurat5基础用法"/>
<meta name="twitter:description" content="
Seurat V5版本有一段时间了，由于时间原因未来得及了解。现根据其官方文档简单整理其用法，与V4比较类似的地方就不多叙述了。此外，V5的亮点之一还在于单细胞多组学的整合分析，此次就不做记录了。（PS：中秋快乐~）"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--Seurat5基础用法",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/041%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat5%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--Seurat5基础用法",
  "name": "单细胞分析工具--Seurat5基础用法",
  "description": " Seurat V5版本有一段时间了，由于时间原因未来得及了解。现根据其官方文档简单整理其用法，与V4比较类似的地方就不多叙述了。此外，V5的亮点之一还在于单细胞多组学的整合分析，此次就不做记录了。（PS：中秋快乐~）\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " Seurat V5版本有一段时间了，由于时间原因未来得及了解。现根据其官方文档简单整理其用法，与V4比较类似的地方就不多叙述了。此外，V5的亮点之一还在于单细胞多组学的整合分析，此次就不做记录了。（PS：中秋快乐~）\n主要参考Seurat官方文档：\nhttps://satijalab.org/seurat/ https://satijalab.org/seurat/articles/install_v5 https://satijalab.org/seurat/articles/essential_commands https://satijalab.org/seurat/articles/seurat5_essential_commands 0. 安装 首先是安装R语言环境（linux环境），使用conda命令 1 2 3 4 5 6 conda create -n r_env -y conda activate r_env conda install -c r r-base=4.2.2 conda install conda-forge::r-seurat conda install r::r-tidyverse 其次，安装其它推荐的工具包 1 2 3 4 5 6 7 8 9 10 11 12 # rcurl使用conda安装 setRepositories(ind = 1:3, addURLs = c('https://satijalab.r-universe.dev', 'https://bnprks.r-universe.dev/')) install.packages(c(\"presto\", \"glmGamPoi\")) # conda install -c anaconda hdf5 install.packages(c(\"BPCells\")) # BiocManager::install(\"Rsamtools\") install.packages('Signac') remotes::install_github(\"satijalab/seurat-data\", quiet = TRUE) #提供示例数据集 # remotes::install_github(\"satijalab/azimuth\", quiet = TRUE) #自动注细胞类型的R包，先不安装了 remotes::install_github(\"satijalab/seurat-wrappers\", quiet = TRUE) 分析前，加载相关初始包 1 2 3 4 library(Seurat) library(dplyr) library(patchwork) library(ggplot2) 示例分析数据 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 library(SeuratData) # AvailableData() # InstallData(\"pbmc3k\") # InstallData(\"ifnb\") library(pbmc3k.SeuratData) data('pbmc3k') pbmc3k = UpdateSeuratObject(pbmc3k) pbmc3k pbmc = pbmc3k library(ifnb.SeuratData) data('ifnb') ifnb = UpdateSeuratObject(ifnb) ifnb 1. Layer结构 Seurat V4/3使用的还是v3版本的Seurat Object； Seurat V5提出了新的v5版本的Seurat Object： 主要区别之一是在Assays层级结构下，明确定义了Layer结构 每个Seurat可以包含多个Assays，每个Assay可以包含多个Layers 1.1 创建v3/5对象 读取表达矩阵\nhttps://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz\n1 2 3 4 5 6 7 pbmc.counts \u003c- Read10X(data.dir = \"filtered_gene_bc_matrices/hg19/\") class(pbmc.counts) # [1] \"dgCMatrix\" # attr(,\"package\") # [1] \"Matrix\" dim(pbmc.counts) # [1] 32738 2700 直接创建Seurat Object 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 ## v3 options(Seurat.object.assay.version = \"v3\") pbmc \u003c- CreateSeuratObject(counts = pbmc.counts) class(pbmc[[\"RNA\"]]) # [1] \"Assay\" # attr(,\"package\") # [1] \"SeuratObject\" ## v5 options(Seurat.object.assay.version = \"v5\") pbmc \u003c- CreateSeuratObject(counts = pbmc.counts) class(pbmc[[\"RNA\"]]) # [1] \"Assay5\" # attr(,\"package\") # [1] \"SeuratObject\" 先创建Assay Object，再创建Seurat Object 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 # create a v3 assay assay.v3 \u003c- CreateAssayObject(counts = pbmc.counts) class(assay.v3) # create a v5 assay assay.v5 \u003c- CreateAssay5Object(counts = pbmc.counts) class(assay.v5) # # create an assay using only normalized data # assay.v5 \u003c- CreateAssay5Object(data = log1p(pbmc.counts)) # create a Seurat object based on this assay pbmc3k_slim \u003c- CreateSeuratObject(assay.v5) pbmc3k_slim # 加入一个新的Assay（v3） pbmc3k_slim[['RNA3']] = as(object = pbmc3k_slim[[\"RNA\"]], Class = \"Assay5\") 1.2 Layers相关操作 查看 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 obj \u003c- CreateSeuratObject(counts = pbmc.counts) obj # An object of class Seurat # 32738 features across 2700 samples within 1 assay # Active assay: RNA (32738 features, 0 variable features) # 1 layer present: counts DefaultAssay(obj) # [1] \"RNA\" DefaultLayer(obj[['RNA']]) # [1] \"counts\" Layers(obj[[\"RNA\"]]) # [1] \"counts\" obj \u003c- NormalizeData(obj, verbose = FALSE) Layers(obj[[\"RNA\"]]) # [1] \"counts\" \"data\" 拆分 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 pbmc3k$batch \u003c- sample(c(\"batchA\", \"batchB\", \"batchC\"), ncol(pbmc3k), replace = TRUE) # Split layers by batch pbmc3k[[\"RNA\"]] \u003c- split(pbmc3k[[\"RNA\"]], f = pbmc3k$batch) Layers(pbmc3k[[\"RNA\"]]) # [1] \"data.batchC\" \"data.batchA\" \"data.batchB\" \"counts.batchC\" # [5] \"counts.batchA\" \"counts.batchB\" # Rejoin layers pbmc3k[[\"RNA\"]] \u003c- JoinLayers(pbmc3k[[\"RNA\"]]) Layers(pbmc3k[[\"RNA\"]]) # [1] \"data\" \"counts\" # create a Seurat object initialized with multiple layers batchA_counts \u003c- pbmc.counts[, 1:200] batchB_counts \u003c- pbmc.counts[, 201:400] batchC_counts \u003c- pbmc.counts[, 401:600] count_list \u003c- list(batchA_counts, batchB_counts, batchC_counts) names(count_list) \u003c- c(\"batchA\", \"batchB\", \"batchC\") obj \u003c- CreateSeuratObject(counts = count_list) Layers(obj[[\"RNA\"]]) # [1] \"counts.batchA\" \"counts.batchB\" \"counts.batchC\" 2. 常用操作 2.1 标准步骤 逐步处理 1 2 3 4 5 6 7 8 pbmc \u003c- NormalizeData(object = pbmc) pbmc \u003c- FindVariableFeatures(object = pbmc) pbmc \u003c- ScaleData(object = pbmc) pbmc \u003c- RunPCA(object = pbmc) pbmc \u003c- FindNeighbors(object = pbmc, dims = 1:30) pbmc \u003c- FindClusters(object = pbmc) pbmc \u003c- RunUMAP(object = pbmc, dims = 1:30) DimPlot(object = pbmc, reduction = \"umap\") SCT简化 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 # https://github.com/satijalab/seurat/issues/8183 # https://blog.csdn.net/weixin_46128755/article/details/136395843 # matrixStats版本问题会导致SCTransform运行出错，这里选择安装1.1.0版本 pbmc \u003c- SCTransform(object = pbmc) pbmc \u003c- RunPCA(object = pbmc) pbmc \u003c- FindNeighbors(object = pbmc, dims = 1:30) pbmc \u003c- FindClusters(object = pbmc) pbmc \u003c- RunUMAP(object = pbmc, dims = 1:30) DimPlot(object = pbmc, reduction = \"umap\") # 或者 pbmc \u003c- SCTransform(pbmc) %\u003e% RunPCA() %\u003e% FindNeighbors(dims = 1:30) %\u003e% FindClusters() %\u003e% RunUMAP(dims = 1:30) 2.2 查看结构 细胞/基因 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 # cell names colnames(pbmc) |\u003e head() Cells(pbmc) |\u003e head() # feature names Features(pbmc) |\u003e head() rownames(pbmc) |\u003e head() # 高变基因 # Variable feature names VariableFeatures(pbmc) |\u003e str() VariableFeatures(pbmc[[\"SCT\"]]) |\u003e str() VariableFeatures(pbmc[[\"RNA\"]]) |\u003e str() # VariableFeatures(pbmc, assay = 'SCT') var.gene.names = VariableFeatures(pbmc)[1:1000] VariableFeatures(pbmc[[\"RNA\"]]) = var.gene.names meta.data 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 colnames(pbmc@meta.data) # [1] \"orig.ident\" \"nCount_RNA\" \"nFeature_RNA\" # [4] \"seurat_annotations\" \"nCount_SCT\" \"nFeature_SCT\" # [7] \"SCT_snn_res.0.8\" \"seurat_clusters\" Idents(object = pbmc) \u003c- \"seurat_annotations\" # View cell identities, get summary table table(Idents(pbmc)) # Naive CD4 T Memory CD4 T CD14+ Mono B CD8 T FCGR3A+ Mono # 697 483 480 344 271 162 # NK DC Platelet # 155 32 14 Idents(pbmc) \u003c- \"CD4 T cells\" table(Idents(pbmc)) # 逆操作 pbmc[[\"old.ident\"]] \u003c- Idents(object = pbmc) colnames(pbmc@meta.data) # Rename identity classes pbmc \u003c- RenameIdents(object = pbmc, `CD4 T cells` = \"T Helper cells\") Idents(object = pbmc) |\u003e table() 矩阵信息 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 ## Retrieve data in an expression matrix RNA counts matrix pbmc[[\"RNA\"]]$counts |\u003e dim() # [1] 13714 2700 pbmc[[\"SCT\"]]$counts |\u003e dim() # [1] 12572 2700 LayerData(pbmc, assay = \"RNA\", layer = \"counts\") |\u003e dim() # [1] 13714 2700 # GetAssayData from Seurat v4 is still supported GetAssayData(object = pbmc, assay = \"RNA\", slot = \"counts\") |\u003e dim() # Set expression data assume new.data is a new expression matrix pbmc[[\"RNA\"]]$counts \u003c- new.data ## 降维信息 pbmc@reductions |\u003e names() # [1] \"pca\" \"umap\" pbmc[['pca']]@cell.embeddings |\u003e head() Embeddings(pbmc, reduction = \"pca\") |\u003e head() pbmc[['pca']]@feature.loadings |\u003e dim() Loadings(pbmc, reduction = \"pca\") |\u003e dim() 2.3 增删操作 提取特定feature 1 2 3 4 5 6 7 FetchData(object = pbmc, vars = c(\"PC_1\", \"nFeature_RNA\", \"MS4A1\"), layer = \"counts\") |\u003e head(3) # PC_1 nFeature_RNA MS4A1 # AAACATACAACCAC 10.165883 779 0 # AAACATTGAGCTAC 5.812925 1352 4 # AAACATTGATCAGC 8.566106 1129 0 筛选cells 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 Idents(object = pbmc) \u003c- \"seurat_annotations\" subset(x = pbmc, idents = \"B\") |\u003e dim() subset(x = pbmc, idents = \"B\", invert = TRUE) |\u003e dim() # Subset on the expression level of a gene/feature subset(x = pbmc, subset = MS4A1 \u003e 0.5) |\u003e dim() # Subset on a combination of criteria subset(x = pbmc, subset = MS4A1 \u003e 0.5 \u0026 PC_1 \u003e 5) |\u003e dim() subset(x = pbmc, subset = MS4A1 \u003e 0.5, idents = \"B\") |\u003e dim() # Subset on a value in the object meta data subset(x = pbmc, subset = groups == \"g1\") |\u003e dim() # Downsample the number of cells per identity class subset(x = pbmc, downsample = 100) |\u003e dim() 拆分与合并 在1.2小结，了解到可以按Layer拆分，但仍属于同一个Object 下面的命令可以拆分独立的多个Object 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 ## 拆分 ifnb_list \u003c- SplitObject(ifnb, split.by = \"stim\") # [1] \"list\" ifnb_list$CTRL ifnb_list$STIM ## 合并 merged_obj \u003c- merge(x = ifnb_list$CTRL, y = ifnb_list$STIM) merged_obj # An object of class Seurat # 14053 features across 13999 samples within 1 assay # Active assay: RNA (14053 features, 0 variable features) # 4 layers present: data.1, data.2, counts.1, counts.2 merged_obj[[\"RNA\"]] \u003c- JoinLayers(merged_obj[[\"RNA\"]]) merged_obj # An object of class Seurat # 14053 features across 13999 samples within 1 assay # Active assay: RNA (14053 features, 0 variable features) # 2 layers present: data, counts ##### Merge objects (with integration)：去批次效应 merged_obj \u003c- merge(x = ifnb_list$CTRL, y = ifnb_list$STIM) merged_obj \u003c- NormalizeData(merged_obj) merged_obj \u003c- FindVariableFeatures(merged_obj) merged_obj \u003c- ScaleData(merged_obj) merged_obj \u003c- RunPCA(merged_obj) #去批次整合（上面的标准流都是对于每个Layer独立操作的） merged_obj \u003c- IntegrateLayers(object = merged_obj, method = RPCAIntegration, orig.reduction = \"pca\", new.reduction = \"integrated.rpca\", verbose = FALSE) merged_obj # now that integration is complete, rejoin layers merged_obj[[\"RNA\"]] \u003c- JoinLayers(merged_obj[[\"RNA\"]]) Seurat V5版本共支持5种去批次效应的方法：CCAIntegration, RPCAIntegration, HarmonyIntegration, FastMNNIntegration, scVIIntegration，可参考：https://satijalab.org/seurat/articles/seurat5_integration\nPseudobulk计算 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 bulk \u003c- AggregateExpression(ifnb, group.by = \"seurat_annotations\", return.seurat = TRUE) Cells(bulk) # [1] \"CD14 Mono\" \"CD4 Naive T\" \"CD4 Memory T\" \"CD16 Mono\" \"B\" # [6] \"CD8 T\" \"T activated\" \"NK\" \"DC\" \"B Activated\" # [11] \"Mk\" \"pDC\" \"Eryth\" dim(bulk) # [1] 14053 13 bulk \u003c- AggregateExpression(ifnb, group.by = \"seurat_annotations\", return.seurat = FALSE) class(bulk$RNA) # [1] \"dgCMatrix\" # attr(,\"package\") # [1] \"Matrix\" # pseudobulk cells by stimulation condition AND cell type bulk \u003c- AggregateExpression(ifnb, group.by = c(\"stim\", \"seurat_annotations\"), return.seurat = TRUE) Cells(bulk) 基于Pseudobulk结果，可采用DESeq2方法进行差异基因检验。可参考：https://satijalab.org/seurat/articles/de_vignette\n2.4 可视化 主要绘图命令记录如下，基本与V4类似。示例可视化效果可参考 https://satijalab.org/seurat/articles/visualization_vignette https://lishensuo.github.io/posts/bioinfo/006%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat4%E5%9F%BA%E7%A1%80%E6%B5%81%E7%A8%8B/ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 DimPlot(object = pbmc, reduction = \"pca\") # Dimensional reduction plot, with cells colored by a quantitative feature Defaults to UMAP if available FeaturePlot(object = pbmc, features = \"MS4A1\") # Scatter plot across single cells FeatureScatter(object = pbmc, feature1 = \"MS4A1\", feature2 = \"PC_1\") FeatureScatter(object = pbmc, feature1 = \"MS4A1\", feature2 = \"CD3D\") # Scatter plot across individual features, repleaces CellPlot CellScatter(object = pbmc, cell1 = \"AGTCTACTAGGGTG\", cell2 = \"CACAGATGGTTTCT\") VariableFeaturePlot(object = pbmc) # Violin and Ridge plots VlnPlot(object = pbmc, features = c(\"LYZ\", \"CCL5\", \"IL32\")) RidgePlot(object = pbmc, feature = c(\"LYZ\", \"CCL5\", \"IL32\")) # Heatmaps (visualize scale.data slot) DimHeatmap(object = pbmc, reduction = \"pca\", cells = 200) # standard workflow var.gene.names = VariableFeatures(pbmc)[1:50] pbmc \u003c- ScaleData(pbmc, features = var.gene.names) # https://github.com/satijalab/seurat/issues/2722 DoHeatmap(object = pbmc, label = F) # heatmap with maximum of 10 cells per group DoHeatmap(pbmc, var.gene.names, cells = colnames(subset(pbmc, downsample = 10)), label = F) 绘制Dotplot时，注意分组变量不能包含NA值 1 2 3 4 5 6 7 8 9 10 11 12 table(is.na(pbmc$seurat_annotations)) # FALSE TRUE # 2638 62 Idents(pbmc) \u003c- as.character(Idents(pbmc)) pbmc$seurat_annotations_2 = pbmc$seurat_annotations pbmc$seurat_annotations_2 = as.character(pbmc$seurat_annotations) pbmc$seurat_annotations_2[is.na(pbmc$seurat_annotations_2)] = 'Other' Idents(pbmc) = pbmc$seurat_annotations_2 VlnPlot(object = pbmc, features = \"MS4A1\", split.by = \"groups\") DotPlot(object = pbmc, features = c(\"LYZ\", \"CCL5\", \"IL32\"), split.by = \"groups\") ",
  "wordCount" : "2484",
  "inLanguage": "en",
  "datePublished": "2024-09-17T00:00:00Z",
  "dateModified": "2024-09-17T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/041%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat5%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--Seurat5基础用法
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2024-09-17 00:00:00 &#43;0000 UTC&#39;&gt;2024-09-17&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2024-09-17&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2484&amp;nbsp;|&amp;nbsp;5 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#0-%e5%ae%89%e8%a3%85" aria-label="0. 安装">0. 安装</a></li>
                    <li>
                        <a href="#1-layer%e7%bb%93%e6%9e%84" aria-label="1. Layer结构">1. Layer结构</a><ul>
                            
                    <li>
                        <a href="#11-%e5%88%9b%e5%bb%bav35%e5%af%b9%e8%b1%a1" aria-label="1.1 创建v3/5对象">1.1 创建v3/5对象</a></li>
                    <li>
                        <a href="#12-layers%e7%9b%b8%e5%85%b3%e6%93%8d%e4%bd%9c" aria-label="1.2 Layers相关操作">1.2 Layers相关操作</a></li></ul>
                    </li>
                    <li>
                        <a href="#2-%e5%b8%b8%e7%94%a8%e6%93%8d%e4%bd%9c" aria-label="2. 常用操作">2. 常用操作</a><ul>
                            
                    <li>
                        <a href="#21-%e6%a0%87%e5%87%86%e6%ad%a5%e9%aa%a4" aria-label="2.1 标准步骤">2.1 标准步骤</a></li>
                    <li>
                        <a href="#22-%e6%9f%a5%e7%9c%8b%e7%bb%93%e6%9e%84" aria-label="2.2 查看结构">2.2 查看结构</a></li>
                    <li>
                        <a href="#23-%e5%a2%9e%e5%88%a0%e6%93%8d%e4%bd%9c" aria-label="2.3 增删操作">2.3 增删操作</a></li>
                    <li>
                        <a href="#24-%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="2.4 可视化">2.4 可视化</a>
                    </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>Seurat V5版本有一段时间了，由于时间原因未来得及了解。现根据其官方文档简单整理其用法，与V4比较类似的地方就不多叙述了。此外，V5的亮点之一还在于单细胞多组学的整合分析，此次就不做记录了。（PS：中秋快乐~）</p></blockquote>
<p>主要参考Seurat官方文档：</p>
<ul>
<li><a href="https://satijalab.org/seurat/">https://satijalab.org/seurat/</a></li>
<li><a href="https://satijalab.org/seurat/articles/install_v5">https://satijalab.org/seurat/articles/install_v5</a></li>
<li><a href="https://satijalab.org/seurat/articles/essential_commands">https://satijalab.org/seurat/articles/essential_commands</a></li>
<li><a href="https://satijalab.org/seurat/articles/seurat5_essential_commands">https://satijalab.org/seurat/articles/seurat5_essential_commands</a></li>
</ul>
<h1 id="0-安装">0. 安装<a hidden class="anchor" aria-hidden="true" href="#0-安装">#</a></h1>
<ul>
<li>首先是安装R语言环境（linux环境），使用conda命令</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>conda create -n r_env -y
</span></span><span style="display:flex;"><span>conda activate r_env
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>conda install -c r r-base=4.2.2
</span></span><span style="display:flex;"><span>conda install conda-forge::r-seurat
</span></span><span style="display:flex;"><span>conda install r::r-tidyverse
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>其次，安装其它推荐的工具包</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># rcurl使用conda安装</span>
</span></span><span style="display:flex;"><span>setRepositories(ind = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">3</span>, addURLs = c(<span style="color:#0ff;font-weight:bold">&#39;https://satijalab.r-universe.dev&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;https://bnprks.r-universe.dev/&#39;</span>))
</span></span><span style="display:flex;"><span>install.packages(c(<span style="color:#0ff;font-weight:bold">&#34;presto&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;glmGamPoi&#34;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># conda install -c anaconda hdf5</span>
</span></span><span style="display:flex;"><span>install.packages(c(<span style="color:#0ff;font-weight:bold">&#34;BPCells&#34;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># BiocManager::install(&#34;Rsamtools&#34;)</span>
</span></span><span style="display:flex;"><span>install.packages(<span style="color:#0ff;font-weight:bold">&#39;Signac&#39;</span>)
</span></span><span style="display:flex;"><span>remotes::install_github(<span style="color:#0ff;font-weight:bold">&#34;satijalab/seurat-data&#34;</span>, quiet = <span style="color:#fff;font-weight:bold">TRUE</span>) <span style="color:#007f7f">#提供示例数据集</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># remotes::install_github(&#34;satijalab/azimuth&#34;, quiet = TRUE) #自动注细胞类型的R包，先不安装了</span>
</span></span><span style="display:flex;"><span>remotes::install_github(<span style="color:#0ff;font-weight:bold">&#34;satijalab/seurat-wrappers&#34;</span>, quiet = <span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>分析前，加载相关初始包</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>library(dplyr)
</span></span><span style="display:flex;"><span>library(patchwork)
</span></span><span style="display:flex;"><span>library(ggplot2)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>示例分析数据</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>library(SeuratData)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AvailableData()</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># InstallData(&#34;pbmc3k&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># InstallData(&#34;ifnb&#34;)</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>library(pbmc3k.SeuratData)
</span></span><span style="display:flex;"><span>data(<span style="color:#0ff;font-weight:bold">&#39;pbmc3k&#39;</span>)
</span></span><span style="display:flex;"><span>pbmc3k = UpdateSeuratObject(pbmc3k)
</span></span><span style="display:flex;"><span>pbmc3k
</span></span><span style="display:flex;"><span>pbmc = pbmc3k
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>library(ifnb.SeuratData)
</span></span><span style="display:flex;"><span>data(<span style="color:#0ff;font-weight:bold">&#39;ifnb&#39;</span>)
</span></span><span style="display:flex;"><span>ifnb = UpdateSeuratObject(ifnb)
</span></span><span style="display:flex;"><span>ifnb
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="1-layer结构">1. Layer结构<a hidden class="anchor" aria-hidden="true" href="#1-layer结构">#</a></h1>
<ul>
<li>Seurat V4/3使用的还是v3版本的Seurat Object；</li>
<li>Seurat V5提出了新的v5版本的Seurat Object：
<ul>
<li>主要区别之一是在Assays层级结构下，明确定义了Layer结构</li>
<li>每个Seurat可以包含多个Assays，每个Assay可以包含多个Layers</li>
</ul>
</li>
</ul>
<h2 id="11-创建v35对象">1.1 创建v3/5对象<a hidden class="anchor" aria-hidden="true" href="#11-创建v35对象">#</a></h2>
<ul>
<li>
<p>读取表达矩阵</p>
<p><a href="https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz">https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz</a></p>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>pbmc.counts &lt;- Read10X(data.dir = <span style="color:#0ff;font-weight:bold">&#34;filtered_gene_bc_matrices/hg19/&#34;</span>)
</span></span><span style="display:flex;"><span>class(pbmc.counts)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;dgCMatrix&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># attr(,&#34;package&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;Matrix&#34;</span>
</span></span><span style="display:flex;"><span>dim(pbmc.counts)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 32738  2700</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>直接创建Seurat Object</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## v3</span>
</span></span><span style="display:flex;"><span>options(Seurat.object.assay.version = <span style="color:#0ff;font-weight:bold">&#34;v3&#34;</span>)
</span></span><span style="display:flex;"><span>pbmc &lt;- CreateSeuratObject(counts = pbmc.counts)
</span></span><span style="display:flex;"><span>class(pbmc[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;Assay&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># attr(,&#34;package&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;SeuratObject&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## v5</span>
</span></span><span style="display:flex;"><span>options(Seurat.object.assay.version = <span style="color:#0ff;font-weight:bold">&#34;v5&#34;</span>)
</span></span><span style="display:flex;"><span>pbmc &lt;- CreateSeuratObject(counts = pbmc.counts)
</span></span><span style="display:flex;"><span>class(pbmc[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;Assay5&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># attr(,&#34;package&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;SeuratObject&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>先创建Assay Object，再创建Seurat Object</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># create a v3 assay</span>
</span></span><span style="display:flex;"><span>assay.v3 &lt;- CreateAssayObject(counts = pbmc.counts)
</span></span><span style="display:flex;"><span>class(assay.v3)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># create a v5 assay</span>
</span></span><span style="display:flex;"><span>assay.v5 &lt;- CreateAssay5Object(counts = pbmc.counts)
</span></span><span style="display:flex;"><span>class(assay.v5)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># # create an assay using only normalized data</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># assay.v5 &lt;- CreateAssay5Object(data = log1p(pbmc.counts))</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># create a Seurat object based on this assay</span>
</span></span><span style="display:flex;"><span>pbmc3k_slim &lt;- CreateSeuratObject(assay.v5)
</span></span><span style="display:flex;"><span>pbmc3k_slim
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 加入一个新的Assay（v3） </span>
</span></span><span style="display:flex;"><span>pbmc3k_slim[[<span style="color:#0ff;font-weight:bold">&#39;RNA3&#39;</span>]] = as(object = pbmc3k_slim[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]], Class = <span style="color:#0ff;font-weight:bold">&#34;Assay5&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="12-layers相关操作">1.2 Layers相关操作<a hidden class="anchor" aria-hidden="true" href="#12-layers相关操作">#</a></h2>
<ul>
<li>查看</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>obj &lt;- CreateSeuratObject(counts = pbmc.counts)
</span></span><span style="display:flex;"><span>obj
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># An object of class Seurat </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 32738 features across 2700 samples within 1 assay </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Active assay: RNA (32738 features, 0 variable features)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  1 layer present: counts</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>DefaultAssay(obj)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;RNA&#34;</span>
</span></span><span style="display:flex;"><span>DefaultLayer(obj[[<span style="color:#0ff;font-weight:bold">&#39;RNA&#39;</span>]])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;counts&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>Layers(obj[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;counts&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>obj &lt;- NormalizeData(obj, verbose = <span style="color:#fff;font-weight:bold">FALSE</span>)
</span></span><span style="display:flex;"><span>Layers(obj[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;counts&#34; &#34;data&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>拆分</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>pbmc3k$batch &lt;- sample(c(<span style="color:#0ff;font-weight:bold">&#34;batchA&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;batchB&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;batchC&#34;</span>), ncol(pbmc3k), replace = <span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Split layers by batch</span>
</span></span><span style="display:flex;"><span>pbmc3k[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]] &lt;- split(pbmc3k[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]], f = pbmc3k$batch)
</span></span><span style="display:flex;"><span>Layers(pbmc3k[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;data.batchC&#34;   &#34;data.batchA&#34;   &#34;data.batchB&#34;   &#34;counts.batchC&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;counts.batchA&#34; &#34;counts.batchB&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Rejoin layers</span>
</span></span><span style="display:flex;"><span>pbmc3k[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]] &lt;- JoinLayers(pbmc3k[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]])
</span></span><span style="display:flex;"><span>Layers(pbmc3k[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;data&#34;   &#34;counts&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># create a Seurat object initialized with multiple layers</span>
</span></span><span style="display:flex;"><span>batchA_counts &lt;- pbmc.counts[, <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">200</span>]
</span></span><span style="display:flex;"><span>batchB_counts &lt;- pbmc.counts[, <span style="color:#ff0;font-weight:bold">201</span>:<span style="color:#ff0;font-weight:bold">400</span>]
</span></span><span style="display:flex;"><span>batchC_counts &lt;- pbmc.counts[, <span style="color:#ff0;font-weight:bold">401</span>:<span style="color:#ff0;font-weight:bold">600</span>]
</span></span><span style="display:flex;"><span>count_list &lt;- list(batchA_counts, batchB_counts, batchC_counts)
</span></span><span style="display:flex;"><span>names(count_list) &lt;- c(<span style="color:#0ff;font-weight:bold">&#34;batchA&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;batchB&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;batchC&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>obj &lt;- CreateSeuratObject(counts = count_list)
</span></span><span style="display:flex;"><span>Layers(obj[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;counts.batchA&#34; &#34;counts.batchB&#34; &#34;counts.batchC&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="2-常用操作">2. 常用操作<a hidden class="anchor" aria-hidden="true" href="#2-常用操作">#</a></h1>
<h2 id="21-标准步骤">2.1 标准步骤<a hidden class="anchor" aria-hidden="true" href="#21-标准步骤">#</a></h2>
<ul>
<li>逐步处理</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>pbmc &lt;- NormalizeData(object = pbmc)
</span></span><span style="display:flex;"><span>pbmc &lt;- FindVariableFeatures(object = pbmc)
</span></span><span style="display:flex;"><span>pbmc &lt;- ScaleData(object = pbmc)
</span></span><span style="display:flex;"><span>pbmc &lt;- RunPCA(object = pbmc)
</span></span><span style="display:flex;"><span>pbmc &lt;- FindNeighbors(object = pbmc, dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>)
</span></span><span style="display:flex;"><span>pbmc &lt;- FindClusters(object = pbmc)
</span></span><span style="display:flex;"><span>pbmc &lt;- RunUMAP(object = pbmc, dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>)
</span></span><span style="display:flex;"><span>DimPlot(object = pbmc, reduction = <span style="color:#0ff;font-weight:bold">&#34;umap&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>SCT简化</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># https://github.com/satijalab/seurat/issues/8183</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># https://blog.csdn.net/weixin_46128755/article/details/136395843</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># matrixStats版本问题会导致SCTransform运行出错，这里选择安装1.1.0版本</span>
</span></span><span style="display:flex;"><span>pbmc &lt;- SCTransform(object = pbmc)
</span></span><span style="display:flex;"><span>pbmc &lt;- RunPCA(object = pbmc)
</span></span><span style="display:flex;"><span>pbmc &lt;- FindNeighbors(object = pbmc, dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>)
</span></span><span style="display:flex;"><span>pbmc &lt;- FindClusters(object = pbmc)
</span></span><span style="display:flex;"><span>pbmc &lt;- RunUMAP(object = pbmc, dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>)
</span></span><span style="display:flex;"><span>DimPlot(object = pbmc, reduction = <span style="color:#0ff;font-weight:bold">&#34;umap&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 或者</span>
</span></span><span style="display:flex;"><span>pbmc &lt;- SCTransform(pbmc) %&gt;%
</span></span><span style="display:flex;"><span>    RunPCA() %&gt;%
</span></span><span style="display:flex;"><span>    FindNeighbors(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;%
</span></span><span style="display:flex;"><span>    FindClusters() %&gt;%
</span></span><span style="display:flex;"><span>    RunUMAP(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="22-查看结构">2.2 查看结构<a hidden class="anchor" aria-hidden="true" href="#22-查看结构">#</a></h2>
<ul>
<li>细胞/基因</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># cell names</span>
</span></span><span style="display:flex;"><span>colnames(pbmc) |&gt; head()
</span></span><span style="display:flex;"><span>Cells(pbmc) |&gt; head()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># feature names</span>
</span></span><span style="display:flex;"><span>Features(pbmc) |&gt; head()
</span></span><span style="display:flex;"><span>rownames(pbmc) |&gt; head()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 高变基因</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Variable feature names</span>
</span></span><span style="display:flex;"><span>VariableFeatures(pbmc) |&gt; str()
</span></span><span style="display:flex;"><span>VariableFeatures(pbmc[[<span style="color:#0ff;font-weight:bold">&#34;SCT&#34;</span>]]) |&gt; str()
</span></span><span style="display:flex;"><span>VariableFeatures(pbmc[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]]) |&gt; str()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># VariableFeatures(pbmc, assay = &#39;SCT&#39;) </span>
</span></span><span style="display:flex;"><span>var.gene.names = VariableFeatures(pbmc)[1:<span style="color:#ff0;font-weight:bold">1000</span>]
</span></span><span style="display:flex;"><span>VariableFeatures(pbmc[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]]) = var.gene.names
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>meta.data</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>colnames(pbmc@meta.data)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;orig.ident&#34;         &#34;nCount_RNA&#34;         &#34;nFeature_RNA&#34;      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [4] &#34;seurat_annotations&#34; &#34;nCount_SCT&#34;         &#34;nFeature_SCT&#34;      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [7] &#34;SCT_snn_res.0.8&#34;    &#34;seurat_clusters&#34;</span>
</span></span><span style="display:flex;"><span>Idents(object = pbmc) &lt;- <span style="color:#0ff;font-weight:bold">&#34;seurat_annotations&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># View cell identities, get summary table</span>
</span></span><span style="display:flex;"><span>table(Idents(pbmc))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  Naive CD4 T Memory CD4 T   CD14+ Mono            B        CD8 T FCGR3A+ Mono </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#          697          483          480          344          271          162 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#           NK           DC     Platelet </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#          155           32           14 </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>Idents(pbmc) &lt;- <span style="color:#0ff;font-weight:bold">&#34;CD4 T cells&#34;</span>
</span></span><span style="display:flex;"><span>table(Idents(pbmc))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 逆操作</span>
</span></span><span style="display:flex;"><span>pbmc[[<span style="color:#0ff;font-weight:bold">&#34;old.ident&#34;</span>]] &lt;- Idents(object = pbmc)
</span></span><span style="display:flex;"><span>colnames(pbmc@meta.data)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Rename identity classes</span>
</span></span><span style="display:flex;"><span>pbmc &lt;- RenameIdents(object = pbmc, `CD4 T cells` = <span style="color:#0ff;font-weight:bold">&#34;T Helper cells&#34;</span>)
</span></span><span style="display:flex;"><span>Idents(object = pbmc) |&gt; table()
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>矩阵信息</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## Retrieve data in an expression matrix RNA counts matrix</span>
</span></span><span style="display:flex;"><span>pbmc[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]]$counts |&gt; dim()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 13714  2700</span>
</span></span><span style="display:flex;"><span>pbmc[[<span style="color:#0ff;font-weight:bold">&#34;SCT&#34;</span>]]$counts |&gt; dim()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 12572  2700</span>
</span></span><span style="display:flex;"><span>LayerData(pbmc, assay = <span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>, layer = <span style="color:#0ff;font-weight:bold">&#34;counts&#34;</span>) |&gt; dim()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 13714  2700</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GetAssayData from Seurat v4 is still supported</span>
</span></span><span style="display:flex;"><span>GetAssayData(object = pbmc, assay = <span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>, slot = <span style="color:#0ff;font-weight:bold">&#34;counts&#34;</span>) |&gt; dim()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Set expression data assume new.data is a new expression matrix</span>
</span></span><span style="display:flex;"><span>pbmc[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]]$counts &lt;- new.data
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 降维信息</span>
</span></span><span style="display:flex;"><span>pbmc@reductions |&gt; names()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;pca&#34;  &#34;umap&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>pbmc[[<span style="color:#0ff;font-weight:bold">&#39;pca&#39;</span>]]@cell.embeddings |&gt; head()
</span></span><span style="display:flex;"><span>Embeddings(pbmc, reduction = <span style="color:#0ff;font-weight:bold">&#34;pca&#34;</span>) |&gt; head()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>pbmc[[<span style="color:#0ff;font-weight:bold">&#39;pca&#39;</span>]]@feature.loadings |&gt; dim()
</span></span><span style="display:flex;"><span>Loadings(pbmc, reduction = <span style="color:#0ff;font-weight:bold">&#34;pca&#34;</span>) |&gt; dim()
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="23-增删操作">2.3 增删操作<a hidden class="anchor" aria-hidden="true" href="#23-增删操作">#</a></h2>
<ul>
<li>提取特定feature</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>FetchData(object = pbmc, 
</span></span><span style="display:flex;"><span>          vars = c(<span style="color:#0ff;font-weight:bold">&#34;PC_1&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;nFeature_RNA&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;MS4A1&#34;</span>), 
</span></span><span style="display:flex;"><span>          layer = <span style="color:#0ff;font-weight:bold">&#34;counts&#34;</span>) |&gt; head(<span style="color:#ff0;font-weight:bold">3</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                     PC_1 nFeature_RNA MS4A1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATACAACCAC 10.165883          779     0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATTGAGCTAC  5.812925         1352     4</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATTGATCAGC  8.566106         1129     0</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>筛选cells</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>Idents(object = pbmc) &lt;- <span style="color:#0ff;font-weight:bold">&#34;seurat_annotations&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>subset(x = pbmc, idents = <span style="color:#0ff;font-weight:bold">&#34;B&#34;</span>) |&gt; dim()
</span></span><span style="display:flex;"><span>subset(x = pbmc, idents = <span style="color:#0ff;font-weight:bold">&#34;B&#34;</span>, invert = <span style="color:#fff;font-weight:bold">TRUE</span>) |&gt; dim()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Subset on the expression level of a gene/feature</span>
</span></span><span style="display:flex;"><span>subset(x = pbmc, subset = MS4A1 &gt; <span style="color:#ff0;font-weight:bold">0.5</span>) |&gt; dim()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Subset on a combination of criteria</span>
</span></span><span style="display:flex;"><span>subset(x = pbmc, subset = MS4A1 &gt; <span style="color:#ff0;font-weight:bold">0.5</span> &amp; PC_1 &gt; <span style="color:#ff0;font-weight:bold">5</span>) |&gt; dim()
</span></span><span style="display:flex;"><span>subset(x = pbmc, subset = MS4A1 &gt; <span style="color:#ff0;font-weight:bold">0.5</span>, idents = <span style="color:#0ff;font-weight:bold">&#34;B&#34;</span>) |&gt; dim()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Subset on a value in the object meta data</span>
</span></span><span style="display:flex;"><span>subset(x = pbmc, subset = groups == <span style="color:#0ff;font-weight:bold">&#34;g1&#34;</span>) |&gt; dim()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Downsample the number of cells per identity class</span>
</span></span><span style="display:flex;"><span>subset(x = pbmc, downsample = <span style="color:#ff0;font-weight:bold">100</span>) |&gt; dim()
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>拆分与合并
<ul>
<li>在1.2小结，了解到可以按Layer拆分，但仍属于同一个Object</li>
<li>下面的命令可以拆分独立的多个Object</li>
</ul>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## 拆分</span>
</span></span><span style="display:flex;"><span>ifnb_list &lt;- SplitObject(ifnb, split.by = <span style="color:#0ff;font-weight:bold">&#34;stim&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;list&#34;</span>
</span></span><span style="display:flex;"><span>ifnb_list$CTRL
</span></span><span style="display:flex;"><span>ifnb_list$STIM
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 合并</span>
</span></span><span style="display:flex;"><span>merged_obj &lt;- merge(x = ifnb_list$CTRL, y = ifnb_list$STIM)
</span></span><span style="display:flex;"><span>merged_obj
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># An object of class Seurat </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 14053 features across 13999 samples within 1 assay </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Active assay: RNA (14053 features, 0 variable features)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  4 layers present: data.1, data.2, counts.1, counts.2</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>merged_obj[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]] &lt;- JoinLayers(merged_obj[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]])
</span></span><span style="display:flex;"><span>merged_obj
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># An object of class Seurat </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 14053 features across 13999 samples within 1 assay </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Active assay: RNA (14053 features, 0 variable features)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  2 layers present: data, counts</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##### Merge objects (with integration)：去批次效应</span>
</span></span><span style="display:flex;"><span>merged_obj &lt;- merge(x = ifnb_list$CTRL, y = ifnb_list$STIM)
</span></span><span style="display:flex;"><span>merged_obj &lt;- NormalizeData(merged_obj)
</span></span><span style="display:flex;"><span>merged_obj &lt;- FindVariableFeatures(merged_obj)
</span></span><span style="display:flex;"><span>merged_obj &lt;- ScaleData(merged_obj)
</span></span><span style="display:flex;"><span>merged_obj &lt;- RunPCA(merged_obj)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#去批次整合（上面的标准流都是对于每个Layer独立操作的）</span>
</span></span><span style="display:flex;"><span>merged_obj &lt;- IntegrateLayers(object = merged_obj, method = RPCAIntegration, 
</span></span><span style="display:flex;"><span>                              orig.reduction = <span style="color:#0ff;font-weight:bold">&#34;pca&#34;</span>, new.reduction = <span style="color:#0ff;font-weight:bold">&#34;integrated.rpca&#34;</span>,
</span></span><span style="display:flex;"><span>                              verbose = <span style="color:#fff;font-weight:bold">FALSE</span>)
</span></span><span style="display:flex;"><span>merged_obj
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># now that integration is complete, rejoin layers</span>
</span></span><span style="display:flex;"><span>merged_obj[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]] &lt;- JoinLayers(merged_obj[[<span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>]])
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>Seurat V5版本共支持5种去批次效应的方法：<code>CCAIntegration, RPCAIntegration, HarmonyIntegration, FastMNNIntegration, scVIIntegration</code>，可参考：https://satijalab.org/seurat/articles/seurat5_integration</p></blockquote>
<ul>
<li>Pseudobulk计算</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>bulk &lt;- AggregateExpression(ifnb, group.by = <span style="color:#0ff;font-weight:bold">&#34;seurat_annotations&#34;</span>, return.seurat = <span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>Cells(bulk)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  [1] &#34;CD14 Mono&#34;    &#34;CD4 Naive T&#34;  &#34;CD4 Memory T&#34; &#34;CD16 Mono&#34;    &#34;B&#34;           </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  [6] &#34;CD8 T&#34;        &#34;T activated&#34;  &#34;NK&#34;           &#34;DC&#34;           &#34;B Activated&#34; </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [11] &#34;Mk&#34;           &#34;pDC&#34;          &#34;Eryth&#34;  </span>
</span></span><span style="display:flex;"><span>dim(bulk)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 14053    13</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>bulk &lt;- AggregateExpression(ifnb, group.by = <span style="color:#0ff;font-weight:bold">&#34;seurat_annotations&#34;</span>, return.seurat = <span style="color:#fff;font-weight:bold">FALSE</span>)
</span></span><span style="display:flex;"><span>class(bulk$RNA)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;dgCMatrix&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># attr(,&#34;package&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;Matrix&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># pseudobulk cells by stimulation condition AND cell type</span>
</span></span><span style="display:flex;"><span>bulk &lt;- AggregateExpression(ifnb, group.by = c(<span style="color:#0ff;font-weight:bold">&#34;stim&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;seurat_annotations&#34;</span>), return.seurat = <span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>Cells(bulk)
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>基于Pseudobulk结果，可采用DESeq2方法进行差异基因检验。可参考：https://satijalab.org/seurat/articles/de_vignette</p></blockquote>
<h2 id="24-可视化">2.4 可视化<a hidden class="anchor" aria-hidden="true" href="#24-可视化">#</a></h2>
<ul>
<li>主要绘图命令记录如下，基本与V4类似。示例可视化效果可参考
<ul>
<li><a href="https://satijalab.org/seurat/articles/visualization_vignette">https://satijalab.org/seurat/articles/visualization_vignette</a></li>
<li><a href="https://lishensuo.github.io/posts/bioinfo/006%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat4%E5%9F%BA%E7%A1%80%E6%B5%81%E7%A8%8B/">https://lishensuo.github.io/posts/bioinfo/006%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat4%E5%9F%BA%E7%A1%80%E6%B5%81%E7%A8%8B/</a></li>
</ul>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>DimPlot(object = pbmc, reduction = <span style="color:#0ff;font-weight:bold">&#34;pca&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Dimensional reduction plot, with cells colored by a quantitative feature Defaults to UMAP if available</span>
</span></span><span style="display:flex;"><span>FeaturePlot(object = pbmc, features = <span style="color:#0ff;font-weight:bold">&#34;MS4A1&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Scatter plot across single cells</span>
</span></span><span style="display:flex;"><span>FeatureScatter(object = pbmc, feature1 = <span style="color:#0ff;font-weight:bold">&#34;MS4A1&#34;</span>, feature2 = <span style="color:#0ff;font-weight:bold">&#34;PC_1&#34;</span>)
</span></span><span style="display:flex;"><span>FeatureScatter(object = pbmc, feature1 = <span style="color:#0ff;font-weight:bold">&#34;MS4A1&#34;</span>, feature2 = <span style="color:#0ff;font-weight:bold">&#34;CD3D&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Scatter plot across individual features, repleaces CellPlot</span>
</span></span><span style="display:flex;"><span>CellScatter(object = pbmc, cell1 = <span style="color:#0ff;font-weight:bold">&#34;AGTCTACTAGGGTG&#34;</span>, cell2 = <span style="color:#0ff;font-weight:bold">&#34;CACAGATGGTTTCT&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>VariableFeaturePlot(object = pbmc)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Violin and Ridge plots</span>
</span></span><span style="display:flex;"><span>VlnPlot(object = pbmc, features = c(<span style="color:#0ff;font-weight:bold">&#34;LYZ&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;CCL5&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;IL32&#34;</span>))
</span></span><span style="display:flex;"><span>RidgePlot(object = pbmc, feature = c(<span style="color:#0ff;font-weight:bold">&#34;LYZ&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;CCL5&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;IL32&#34;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Heatmaps (visualize scale.data slot)</span>
</span></span><span style="display:flex;"><span>DimHeatmap(object = pbmc, reduction = <span style="color:#0ff;font-weight:bold">&#34;pca&#34;</span>, cells = <span style="color:#ff0;font-weight:bold">200</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># standard workflow</span>
</span></span><span style="display:flex;"><span>var.gene.names = VariableFeatures(pbmc)[1:<span style="color:#ff0;font-weight:bold">50</span>]
</span></span><span style="display:flex;"><span>pbmc &lt;- ScaleData(pbmc, features = var.gene.names)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># https://github.com/satijalab/seurat/issues/2722</span>
</span></span><span style="display:flex;"><span>DoHeatmap(object = pbmc, label = <span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># heatmap with maximum of 10 cells per group</span>
</span></span><span style="display:flex;"><span>DoHeatmap(pbmc, var.gene.names, cells = colnames(subset(pbmc, downsample = <span style="color:#ff0;font-weight:bold">10</span>)), label = <span style="color:#fff;font-weight:bold">F</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>绘制Dotplot时，注意分组变量不能包含NA值</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>table(is.na(pbmc$seurat_annotations))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FALSE  TRUE </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  2638    62 </span>
</span></span><span style="display:flex;"><span>Idents(pbmc) &lt;- as.character(Idents(pbmc))
</span></span><span style="display:flex;"><span>pbmc$seurat_annotations_2 = pbmc$seurat_annotations
</span></span><span style="display:flex;"><span>pbmc$seurat_annotations_2 = as.character(pbmc$seurat_annotations)
</span></span><span style="display:flex;"><span>pbmc$seurat_annotations_2[is.na(pbmc$seurat_annotations_2)] = <span style="color:#0ff;font-weight:bold">&#39;Other&#39;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>Idents(pbmc) = pbmc$seurat_annotations_2
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>VlnPlot(object = pbmc, features = <span style="color:#0ff;font-weight:bold">&#34;MS4A1&#34;</span>, split.by = <span style="color:#0ff;font-weight:bold">&#34;groups&#34;</span>)
</span></span><span style="display:flex;"><span>DotPlot(object = pbmc, features = c(<span style="color:#0ff;font-weight:bold">&#34;LYZ&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;CCL5&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;IL32&#34;</span>), split.by = <span style="color:#0ff;font-weight:bold">&#34;groups&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/program/033python-%E5%8F%AF%E8%A7%86%E5%8C%96-sns%E7%BB%98%E5%88%B6%E7%83%AD%E5%9B%BE%E8%81%9A%E7%B1%BB%E7%83%AD%E5%9B%BE/">
    <span class="title">« Prev Page</span>
    <br>
    <span>Python-可视化-sns绘制热图&amp;聚类热图</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/program/034python-%E5%8F%AF%E8%A7%86%E5%8C%96-sns%E7%BB%98%E5%88%B6%E5%9B%9E%E5%BD%92%E7%82%B9%E5%9B%BE/">
    <span class="title">Next Page »</span>
    <br>
    <span>Python-可视化-sns绘制回归点图</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
