<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹CellPLM | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="æ–‡çŒ®-ç®—æ³•">
<meta name="description" content="
æ ‡é¢˜ CELLPLM: PRE-TRAINING OF CELL LANGUAGE MODEL BEYOND SINGLE CELLS https://openreview.net/forum?id=BKXvPDekud
å‘è¡¨ ICLR (The International Conference on Learning Representations) 2024
é€šè®¯ Jiliang Tang | Computer science and engineering department | Michigan State University | https://www.cse.msu.edu/~tangjili/
1. ç®€ä»‹
ï¼ˆ1ï¼‰scRNA-seq dataä¸natural languageçš„åŒºåˆ«">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/basic/204%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bcellplm/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/basic/204%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bcellplm/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹CellPLM" />
<meta property="og:description" content="
æ ‡é¢˜ CELLPLM: PRE-TRAINING OF CELL LANGUAGE MODEL BEYOND SINGLE CELLS https://openreview.net/forum?id=BKXvPDekud
å‘è¡¨ ICLR (The International Conference on Learning Representations) 2024
é€šè®¯ Jiliang Tang | Computer science and engineering department | Michigan State University | https://www.cse.msu.edu/~tangjili/
1. ç®€ä»‹
ï¼ˆ1ï¼‰scRNA-seq dataä¸natural languageçš„åŒºåˆ«" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/basic/204%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bcellplm/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2024-11-16T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2024-11-16T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹CellPLM"/>
<meta name="twitter:description" content="
æ ‡é¢˜ CELLPLM: PRE-TRAINING OF CELL LANGUAGE MODEL BEYOND SINGLE CELLS https://openreview.net/forum?id=BKXvPDekud
å‘è¡¨ ICLR (The International Conference on Learning Representations) 2024
é€šè®¯ Jiliang Tang | Computer science and engineering department | Michigan State University | https://www.cse.msu.edu/~tangjili/
1. ç®€ä»‹
ï¼ˆ1ï¼‰scRNA-seq dataä¸natural languageçš„åŒºåˆ«"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "åˆ†ç±»",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "ğŸ“– ç§‘ç ”åŸºç¡€ -- è¯»æ–‡çŒ®ã€çœ‹æ•™æ",
      "item": "https://lishensuo.github.io/en/posts/basic/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹CellPLM",
      "item": "https://lishensuo.github.io/en/posts/basic/204%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bcellplm/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹CellPLM",
  "name": "æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹CellPLM",
  "description": " æ ‡é¢˜ CELLPLM: PRE-TRAINING OF CELL LANGUAGE MODEL BEYOND SINGLE CELLS https://openreview.net/forum?id=BKXvPDekud\nå‘è¡¨ ICLR (The International Conference on Learning Representations) 2024\né€šè®¯ Jiliang Tang | Computer science and engineering department | Michigan State University | https://www.cse.msu.edu/~tangjili/\n1. ç®€ä»‹ ï¼ˆ1ï¼‰scRNA-seq dataä¸natural languageçš„åŒºåˆ«\n",
  "keywords": [
    "æ–‡çŒ®-ç®—æ³•"
  ],
  "articleBody": " æ ‡é¢˜ CELLPLM: PRE-TRAINING OF CELL LANGUAGE MODEL BEYOND SINGLE CELLS https://openreview.net/forum?id=BKXvPDekud\nå‘è¡¨ ICLR (The International Conference on Learning Representations) 2024\né€šè®¯ Jiliang Tang | Computer science and engineering department | Michigan State University | https://www.cse.msu.edu/~tangjili/\n1. ç®€ä»‹ ï¼ˆ1ï¼‰scRNA-seq dataä¸natural languageçš„åŒºåˆ«\nNot sequential â†’ Bag-of-genes (refer to bag-words) Cell-cell communicationåœ¨ç»†èƒçŠ¶æ€å’Œå‘è‚²è¿‡ç¨‹ä¸­çš„é‡è¦æ€§ å•ç»†èƒæ•°æ®çš„Quantityä¸Qualityè¾ƒå·® ï¼ˆ2ï¼‰CellPLMæ¨¡å‹ç‰¹ç‚¹ (single-Cell Pre-trained Language Model)\nCell language model to account for cell-cell relations é¢„è®­ç»ƒè¿‡ç¨‹ä½¿ç”¨äº†spatially-resolved transcriptomic (SRT) ç©ºé—´è½¬å½•ç»„æ•°æ® ä½¿ç”¨é«˜æ–¯æ··åˆåˆ†å¸ƒæ¨¡å‹å­¦ä¹ ç»†èƒçš„æ½œåœ¨åˆ†å¸ƒç©ºé—´ To the best of our knowledge, the proposed CellPLM is the first pre-trained transformer framework that encodes inter-cell relations, leverages spatially-resolved transcriptomic data, and adopts a reasonable prior distribution.\nï¼ˆ3ï¼‰CellPLMæ¨¡å‹çš„æ ¸å¿ƒå‡è®¾\nå¯¹äºNä¸ªç»†èƒï¼Œkä¸ªåŸºå› çš„å•ç»†èƒè¡¨è¾¾çŸ©é˜µX âˆˆ [N, k]\nå¦‚ä¸‹å¼1ï¼ŒO(i)ä»£è¡¨ç»†èƒiçš„ unobserved (masked) genes, U(i) ä»£è¡¨ç»†èƒiçš„ known (unmasked) genes å³ç»†èƒiçš„æŸä¸€åŸºå› è¡¨è¾¾çš„æ¡ä»¶æ¦‚ç‡åˆ†å¸ƒ(conditional probability distribution)å¯ç”±åŒä¸€ç»†èƒå†…å…¶å®ƒå·²çŸ¥è¡¨è¾¾çš„åŸºå› æ¨æµ‹å¾—åˆ°ã€‚ Gene tokens â†’ è¿™åœ¨è®¸å¤šç°æœ‰çš„å•ç»†èƒTransformeræ¨¡å‹å¸¸ç”¨ï¼ˆe.g. scBERT, scGPTï¼‰ å¦‚ä¸‹å¼2ï¼ŒMè¡¨ç¤ºunobserved genes from many cells, Mcè¡¨ç¤ºè¡¨è¾¾çŸ©é˜µXçš„è¡¥é›†ã€‚ å³å³é¢„æµ‹ç»†èƒiçš„æŸä¸€åŸºå› è¡¨è¾¾å¯ä»¥åŒä¸€ç»„ç»‡å†…å¤šä¸ªç»†èƒçš„known genesæ¨æµ‹å¾—åˆ° Cell tokens â†’ è¿™æ˜¯CellPLMæ¨¡å‹çš„æ ¸å¿ƒå‡è®¾ 2. é¢„è®­ç»ƒ é¢„è®­ç»ƒæ¨¡å‹åˆ†ä¸º4ä¸ªç»„æˆéƒ¨åˆ†\nï¼ˆ1ï¼‰Gene expression embedder (cell level)\nå‚è€ƒè¯è¢‹æ¨¡å‹ï¼Œå°†åŸºå› ä¿¡æ¯æ•´åˆè‡³ç»†èƒæ°´å¹³ã€‚\né¦–å…ˆå¯¹gene tokenè®¡ç®—åˆå§‹åŒ–çš„embeddingè¡¨ç¤º â†’ (gene, embedding)\nç„¶åä»¥ç»†èƒçš„åŸºå› è¡¨è¾¾(cell, gene)åˆ†åˆ«ä¸ºç³»æ•°è®¡ç®—åŠ æƒå’Œï¼Œä½œä¸ºåˆå§‹åŒ–çš„Cell Expression Embedding (N, d)\nè¡¨æ ¼çŸ©é˜µåŸºäºSeuratè¿›è¡Œäº†æ ‡å‡†çš„library size normalization and log1p transformationå¤„ç†ã€‚\nä¸€ä¸ªæ‰¹æ¬¡çš„cellsä½œä¸ºä¸€ä¸ªæ ·æœ¬ (N, k)ï¼Œdataloaderä¸€æ¬¡åªè¿­ä»£ä¸€ä¸ªæ‰¹æ¬¡ï¼ˆbatch_size = 1ï¼‰\nï¼ˆ2ï¼‰Transformer encoder\né¦–å…ˆTransformerçš„input cell embeddingç”±ä¸¤éƒ¨åˆ†ç»„æˆï¼ˆCat/Addï¼‰ã€‚é™¤äº†ä¸Šé¢çš„Expression Embedï¼Œè¿˜æœ‰Position Embed\nå¯¹äºSRTæ•°æ®ï¼Œposition embedç”±ä¸€ä¸ªFOV (fields of view) å†…çš„ç»†èƒäºŒç»´åæ ‡ç”Ÿæˆã€‚ å¯¹äºæ™®é€šscRNA-seqæ•°æ®ï¼Œç”±äºæ²¡æœ‰ç›¸å…³ä¿¡æ¯ï¼Œåˆ™é€šè¿‡éšæœºç”Ÿæˆè¿›è¡Œè¡¨ç¤ºã€‚ ç„¶åå°†ä¸Šè¿°çš„åˆå¹¶ Cell embeddingè¾“å…¥åˆ°Transformerå±‚ä¸­\né‡‡ç”¨äº†å…·æœ‰çº¿æ€§å¤æ‚åº¦çš„Flowformerè®¡ç®—æ³¨æ„åŠ›ã€‚ è¾“å‡ºä¸º(N, d) ï¼ˆ3ï¼‰Gaussian mixture latent space\nå¯¹æ³¨æ„åŠ›ç¼–ç å±‚çš„ç»†èƒè¡¨ç¤º(N, d)ï¼Œä½¿ç”¨é«˜æ–¯æ··åˆåˆ†å¸ƒæ¨¡å‹å­¦ä¹ å…¶æ½œåœ¨åˆ†å¸ƒç©ºé—´ z\nä¸€æ–¹é¢å­¦ä¹ å¾—åˆ°æ½œåœ¨åˆ†å¸ƒç©ºé—´(N, z) å¦ä¸€æ–¹é¢å¾—åˆ°çš„å˜åˆ†é‡æ„æŸå¤±ï¼ˆLatent lossï¼‰ä½œä¸ºé¢„è®­ç»ƒæŸå¤±çš„ä¸€éƒ¨åˆ† ï¼ˆ4ï¼‰Batch-aware decoder\né¢„è®­ç»ƒçš„æœ€ç»ˆè¾“å‡ºæ˜¯åŸºå› è¡¨è¾¾çŸ©é˜µ(N, k)ã€‚åœ¨è¾“å…¥åˆ°MLPé¢„æµ‹å‰ï¼Œéœ€è¦è¿›ä¸€æ­¥è€ƒè™‘å…¶å®ƒå› ç´ (Covariates)å¯¹äºè¡¨è¾¾é‡çš„å½±å“ã€‚\nCellPLMè€ƒè™‘çš„åå˜é‡å› ç´ åŒ…æ‹¬æ‰¹æ¬¡æ•ˆåº”/æµ‹åºå¹³å°/æ•°æ®é›† Loss\næœ€åå°†ç»†èƒçš„æ½œåœ¨åˆ†å¸ƒç©ºé—´(z)ä»¥åŠåå˜é‡è¡¨ç¤º(b)åˆå¹¶åï¼Œè¾“å…¥åˆ°MLPç½‘ç»œï¼Œå†ç»è‡ªç„¶æŒ‡æ•°è½¬æ¢åï¼Œå¾—åˆ°æœ€ç»ˆè¾“å‡º (N, k)\nä½¿ç”¨NBè´ŸäºŒé¡¹åˆ†å¸ƒè®¡ç®—å¯¹äºæ©ç åŸºå› çš„é¢„æµ‹å€¼ä¸çœŸå®å€¼ä¹‹é—´çš„å·®å¼‚æŸå¤±(Target loss)\næœ€ç»ˆé¢„è®­ç»ƒæŸå¤±ç”±ä¸¤éƒ¨åˆ†ç»„æˆï¼šLatent loss + Target lossï¼Œè¿›è¡Œåå‘ä¼ æ’­ï¼Œæ›´æ–°æ¨¡å‹ã€‚\né¢„è®­ç»ƒæ¨¡å‹åŒ…å«82Må‚æ•°ï¼Œå¯¹11Mç»†èƒæ•°æ®è¿›è¡Œè®­ç»ƒï¼Œ\u003c24 hours via 8 v100 GPUs\nscRNA-seq HTCA (human tumor cell atlas)ï¼š4.7M HCA (human cell atlas)ï¼š1.4M GEOï¼š2.6M SRTï¼š2.7M 3. å¾®è°ƒè®­ç»ƒ åŸºäºé¢„è®­ç»ƒæ¨¡å‹ï¼Œæ–‡ç« é‡‡ç”¨äº†å¤šè§’åº¦çš„å¾®è°ƒä»»åŠ¡ï¼Œå¹¶ä¸å…¶å®ƒå·²æœ‰æ¨¡å‹åšå¯¹æ¯”ï¼Œè¡¨æ˜CellPLMçš„ä¼˜åŠ¿ã€‚\nå¯¹äºæ¯ä¸ªå¾®è°ƒä»»åŠ¡ï¼Œæ–‡ç« åŸºæœ¬éƒ½ä»Downstream Task Datasetsï¼ŒEvaluation Metricsï¼ŒBaselinesï¼ŒFine-tuningå±•å¼€ã€‚\nCell-level tasks\nZero shotï¼šç›´æ¥ä½¿ç”¨é¢„è®­ç»ƒæ¨¡å‹ï¼Œè®¡ç®—ç»†èƒçš„åµŒå…¥è¡¨ç¤ºh (å¯ç”¨äºèšç±»åˆ†ç¾¤ç­‰) Cell Type Annotationï¼šåŸºäºLatentè¾“å‡ºçš„cell embedï¼Œè¿æ¥MLPå±‚é¢„æµ‹ç»†èƒç±»å‹ é€šè¿‡æŸ¥çœ‹ä»“åº“ä»£ç ï¼Œä¸Šè¿°ä¸¤ç§å¾®è°ƒä»»åŠ¡éƒ½æ˜¯ç›´æ¥ä½¿ç”¨çš„Latent layerè¾“å‡ºçš„ç»†èƒåµŒå…¥è¡¨ç¤ºæ‰§è¡Œç›¸åº”ä»»åŠ¡ã€‚\nGene-level tasks\nscRNA-seq denoisingï¼šå°†ä¸€éƒ¨åˆ†éé›¶è¡¨è¾¾åŸºå› ç½®æ¢ä¸º0åï¼Œè®­ç»ƒæ¨¡å‹é¢„æµ‹å…¶åŸæœ‰è¡¨è¾¾å€¼ã€‚\nSpatial Transcriptomic imputationï¼šæœ¬è´¨ä¸Šï¼Œä¹Ÿæ˜¯å°†ä¸€éƒ¨åˆ†åŸºå› ç½®æ¢ä¸º0ï¼Œ è®­ç»ƒæ¨¡å‹é¢„æµ‹åŸæœ‰è¡¨è¾¾å€¼ã€‚è®­ç»ƒè¿‡ç¨‹ä¸­ï¼Œéœ€è¦è”åˆscRNA-seqä½œä¸ºReferenceæ•°æ®ï¼Œå¸®åŠ©è®­ç»ƒã€‚\nGenetic perturbation predictionï¼šé¢„æµ‹æ‰°åŠ¨åŸºå› è¡¨è¾¾ï¼Œå…³é”®æ˜¯å¦‚ä½•åœ¨è¾“å…¥æ•°æ®ä¸­æ ‡è®°å‡ºè¢«æ‰°åŠ¨çš„åŸºå› \nFor one perturbation, we set the input of perturbed genes to be âˆ’100 to mimic the gene perturbation action. 4. å…¶å®ƒæ–¹é¢ ï¼ˆ1ï¼‰æ¶ˆèå®éªŒ\né€šè¿‡ä¸‰ä¸ªè§’åº¦çš„æ¶ˆèå®éªŒï¼Œè¯´æ˜CellPLMæ¨¡å‹componentæ‰€å‘æŒ¥çš„ä½œç”¨\næ˜¯å¦é‡‡ç”¨Mixture of Gaussian æ˜¯å¦é‡‡ç”¨Latent distribution layers æ˜¯å¦é‡‡ç”¨Transformer encoder ï¼ˆ2ï¼‰ä»£ç å€Ÿé‰´å­¦ä¹ \nç”±äºå•ç»†èƒè¡¨è¾¾çŸ©é˜µçš„ç¨€ç–æ€§ï¼Œtorchæ”¯æŒç›¸å…³é«˜æ•ˆçš„å¤„ç†æ–¹å¼ ä¾‹å¦‚ torch.sparse_csr_tensorï¼Œtorch.sparse.mmç­‰ é€šè¿‡from abc import ABC, abstractmethodï¼Œä½¿å¾—æ›´åŠ å®¹æ˜“åœ°å®ç°åŒä¸€ç±»å‹ï¼Œä¸åŒå˜ä½“çš„æ¨¡å‹component ä¾‹å¦‚ ä¸åŒçš„å¾®è°ƒ Pipeline æ¯ä¸ªæ¨¡å‹componentéƒ½ç”±ä¸€ä¸ªæ–‡ä»¶å¤¹ç»„æˆï¼Œé€šè¿‡å…¶ä¸­__init__.pyæ–‡ä»¶å‘æŒ¥å…³é”®ä½œç”¨ã€‚ Transformerï¼ŒVAEç­‰ç»å…¸æ¨¡å‹çš„å®ç°æ–¹å¼ ",
  "wordCount" : "2167",
  "inLanguage": "en",
  "datePublished": "2024-11-16T00:00:00Z",
  "dateModified": "2024-11-16T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/basic/204%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bcellplm/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="ä¸»é¡µ">
                    <span>ä¸»é¡µ</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="åˆ†ç±»">
                    <span>åˆ†ç±»</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="æ ‡ç­¾">
                    <span>æ ‡ç­¾</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="å½’æ¡£">
                    <span>å½’æ¡£</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="å…³äº">
                    <span>å…³äº</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="æœç´¢ (Alt &#43; /)" accesskey=/>
                    <span>æœç´¢</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/">åˆ†ç±»</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/basic/">ğŸ“– ç§‘ç ”åŸºç¡€ -- è¯»æ–‡çŒ®ã€çœ‹æ•™æ</a></div>
    <h1 class="post-title">
      æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹CellPLM
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2024-11-16 00:00:00 &#43;0000 UTC&#39;&gt;2024-11-16&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2024-11-16&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2167&amp;nbsp;|&amp;nbsp;5 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1-%e7%ae%80%e4%bb%8b" aria-label="1. ç®€ä»‹">1. ç®€ä»‹</a></li>
                    <li>
                        <a href="#2-%e9%a2%84%e8%ae%ad%e7%bb%83" aria-label="2. é¢„è®­ç»ƒ">2. é¢„è®­ç»ƒ</a></li>
                    <li>
                        <a href="#3-%e5%be%ae%e8%b0%83%e8%ae%ad%e7%bb%83" aria-label="3. å¾®è°ƒè®­ç»ƒ">3. å¾®è°ƒè®­ç»ƒ</a></li>
                    <li>
                        <a href="#4-%e5%85%b6%e5%ae%83%e6%96%b9%e9%9d%a2" aria-label="4. å…¶å®ƒæ–¹é¢">4. å…¶å®ƒæ–¹é¢</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p><strong>æ ‡é¢˜</strong> CELLPLM: PRE-TRAINING OF CELL LANGUAGE MODEL BEYOND SINGLE CELLS <a href="https://openreview.net/forum?id=BKXvPDekud">https://openreview.net/forum?id=BKXvPDekud</a></p>
<p><strong>å‘è¡¨</strong> ICLR (The International Conference on Learning Representations) 2024</p>
<p><strong>é€šè®¯</strong> Jiliang Tang | Computer science and engineering department | Michigan State University | <a href="https://www.cse.msu.edu/~tangjili/">https://www.cse.msu.edu/~tangjili/</a></p></blockquote>
<h1 id="1-ç®€ä»‹">1. ç®€ä»‹<a hidden class="anchor" aria-hidden="true" href="#1-ç®€ä»‹">#</a></h1>
<p><strong>ï¼ˆ1ï¼‰scRNA-seq dataä¸natural languageçš„åŒºåˆ«</strong></p>
<ul>
<li>Not sequential â†’ Bag-of-genes (refer to bag-words)</li>
<li>Cell-cell communicationåœ¨ç»†èƒçŠ¶æ€å’Œå‘è‚²è¿‡ç¨‹ä¸­çš„é‡è¦æ€§</li>
<li>å•ç»†èƒæ•°æ®çš„Quantityä¸Qualityè¾ƒå·®</li>
</ul>
<p><strong>ï¼ˆ2ï¼‰CellPLMæ¨¡å‹ç‰¹ç‚¹ (single-Cell Pre-trained Language Model)</strong></p>
<ul>
<li>Cell language model to account for cell-cell relations</li>
<li>é¢„è®­ç»ƒè¿‡ç¨‹ä½¿ç”¨äº†spatially-resolved transcriptomic (SRT) ç©ºé—´è½¬å½•ç»„æ•°æ®</li>
<li>ä½¿ç”¨é«˜æ–¯æ··åˆåˆ†å¸ƒæ¨¡å‹å­¦ä¹ ç»†èƒçš„æ½œåœ¨åˆ†å¸ƒç©ºé—´</li>
</ul>
<blockquote>
<p>To the best of our knowledge, the proposed CellPLM is the first pre-trained transformer framework that encodes inter-cell relations, leverages spatially-resolved transcriptomic data, and adopts a reasonable prior distribution.</p></blockquote>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20241116130418471.png" alt="image-20241116130418471"  />
</p>
<p><strong>ï¼ˆ3ï¼‰CellPLMæ¨¡å‹çš„æ ¸å¿ƒå‡è®¾</strong></p>
<p>å¯¹äºNä¸ªç»†èƒï¼Œkä¸ªåŸºå› çš„å•ç»†èƒè¡¨è¾¾çŸ©é˜µX âˆˆ [N, k]</p>
<ul>
<li>å¦‚ä¸‹å¼1ï¼ŒO(i)ä»£è¡¨ç»†èƒiçš„ unobserved (masked) genes, U(i) ä»£è¡¨ç»†èƒiçš„ known (unmasked) genes
<ul>
<li>å³ç»†èƒiçš„æŸä¸€åŸºå› è¡¨è¾¾çš„æ¡ä»¶æ¦‚ç‡åˆ†å¸ƒ(conditional probability distribution)å¯ç”±åŒä¸€ç»†èƒå†…å…¶å®ƒå·²çŸ¥è¡¨è¾¾çš„åŸºå› æ¨æµ‹å¾—åˆ°ã€‚</li>
<li>Gene tokens â†’ è¿™åœ¨è®¸å¤šç°æœ‰çš„å•ç»†èƒTransformeræ¨¡å‹å¸¸ç”¨ï¼ˆe.g. scBERT, scGPTï¼‰</li>
</ul>
</li>
<li>å¦‚ä¸‹å¼2ï¼ŒMè¡¨ç¤ºunobserved genes from many cells, Mcè¡¨ç¤ºè¡¨è¾¾çŸ©é˜µXçš„è¡¥é›†ã€‚
<ul>
<li>å³å³é¢„æµ‹ç»†èƒiçš„æŸä¸€åŸºå› è¡¨è¾¾å¯ä»¥åŒä¸€ç»„ç»‡å†…å¤šä¸ªç»†èƒçš„known genesæ¨æµ‹å¾—åˆ°</li>
<li>Cell tokens â†’ è¿™æ˜¯CellPLMæ¨¡å‹çš„æ ¸å¿ƒå‡è®¾</li>
</ul>
</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20241116125034284.png" alt="image-20241116125034284"  />
</p>
<h1 id="2-é¢„è®­ç»ƒ">2. é¢„è®­ç»ƒ<a hidden class="anchor" aria-hidden="true" href="#2-é¢„è®­ç»ƒ">#</a></h1>
<img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20241116130629315.png" alt="image-20241116130629315" style="zoom:67%;" />
<p>é¢„è®­ç»ƒæ¨¡å‹åˆ†ä¸º4ä¸ªç»„æˆéƒ¨åˆ†</p>
<p><strong>ï¼ˆ1ï¼‰Gene expression embedder (cell level)</strong></p>
<p>å‚è€ƒ<a href="https://zh.wikipedia.org/zh-tw/%E8%AF%8D%E8%A2%8B%E6%A8%A1%E5%9E%8B">è¯è¢‹æ¨¡å‹</a>ï¼Œå°†åŸºå› ä¿¡æ¯æ•´åˆè‡³ç»†èƒæ°´å¹³ã€‚</p>
<ul>
<li>
<p>é¦–å…ˆå¯¹gene tokenè®¡ç®—åˆå§‹åŒ–çš„embeddingè¡¨ç¤º â†’ (gene, embedding)</p>
</li>
<li>
<p>ç„¶åä»¥ç»†èƒçš„åŸºå› è¡¨è¾¾(cell, gene)åˆ†åˆ«ä¸ºç³»æ•°è®¡ç®—åŠ æƒå’Œï¼Œä½œä¸ºåˆå§‹åŒ–çš„Cell Expression Embedding (N, d)</p>
</li>
</ul>
<blockquote>
<p>è¡¨æ ¼çŸ©é˜µåŸºäºSeuratè¿›è¡Œäº†æ ‡å‡†çš„library size normalization and log1p transformationå¤„ç†ã€‚</p>
<p>ä¸€ä¸ªæ‰¹æ¬¡çš„cellsä½œä¸ºä¸€ä¸ªæ ·æœ¬ (N, k)ï¼Œdataloaderä¸€æ¬¡åªè¿­ä»£ä¸€ä¸ªæ‰¹æ¬¡ï¼ˆbatch_size = 1ï¼‰</p></blockquote>
<p><strong>ï¼ˆ2ï¼‰Transformer encoder</strong></p>
<p>é¦–å…ˆTransformerçš„input cell embeddingç”±ä¸¤éƒ¨åˆ†ç»„æˆï¼ˆCat/Addï¼‰ã€‚é™¤äº†ä¸Šé¢çš„Expression Embedï¼Œè¿˜æœ‰Position Embed</p>
<ul>
<li>å¯¹äºSRTæ•°æ®ï¼Œposition embedç”±ä¸€ä¸ªFOV (fields of view) å†…çš„ç»†èƒäºŒç»´åæ ‡ç”Ÿæˆã€‚</li>
<li>å¯¹äºæ™®é€šscRNA-seqæ•°æ®ï¼Œç”±äºæ²¡æœ‰ç›¸å…³ä¿¡æ¯ï¼Œåˆ™é€šè¿‡éšæœºç”Ÿæˆè¿›è¡Œè¡¨ç¤ºã€‚</li>
</ul>
<p>ç„¶åå°†ä¸Šè¿°çš„åˆå¹¶ Cell embeddingè¾“å…¥åˆ°Transformerå±‚ä¸­</p>
<ul>
<li>é‡‡ç”¨äº†å…·æœ‰çº¿æ€§å¤æ‚åº¦çš„Flowformerè®¡ç®—æ³¨æ„åŠ›ã€‚</li>
<li>è¾“å‡ºä¸º(N, d)</li>
</ul>
<p><strong>ï¼ˆ3ï¼‰Gaussian mixture latent space</strong></p>
<p>å¯¹æ³¨æ„åŠ›ç¼–ç å±‚çš„ç»†èƒè¡¨ç¤º(N, d)ï¼Œä½¿ç”¨é«˜æ–¯æ··åˆåˆ†å¸ƒæ¨¡å‹å­¦ä¹ å…¶æ½œåœ¨åˆ†å¸ƒç©ºé—´ z</p>
<ul>
<li>ä¸€æ–¹é¢å­¦ä¹ å¾—åˆ°æ½œåœ¨åˆ†å¸ƒç©ºé—´(N, z)</li>
<li>å¦ä¸€æ–¹é¢å¾—åˆ°çš„å˜åˆ†é‡æ„æŸå¤±ï¼ˆLatent lossï¼‰ä½œä¸ºé¢„è®­ç»ƒæŸå¤±çš„ä¸€éƒ¨åˆ†</li>
</ul>
<p><strong>ï¼ˆ4ï¼‰Batch-aware decoder</strong></p>
<p>é¢„è®­ç»ƒçš„æœ€ç»ˆè¾“å‡ºæ˜¯åŸºå› è¡¨è¾¾çŸ©é˜µ(N, k)ã€‚åœ¨è¾“å…¥åˆ°MLPé¢„æµ‹å‰ï¼Œéœ€è¦è¿›ä¸€æ­¥è€ƒè™‘å…¶å®ƒå› ç´ (Covariates)å¯¹äºè¡¨è¾¾é‡çš„å½±å“ã€‚</p>
<ul>
<li>CellPLMè€ƒè™‘çš„åå˜é‡å› ç´ åŒ…æ‹¬æ‰¹æ¬¡æ•ˆåº”/æµ‹åºå¹³å°/æ•°æ®é›†</li>
</ul>
<hr>
<p><strong>Loss</strong></p>
<p>æœ€åå°†ç»†èƒçš„æ½œåœ¨åˆ†å¸ƒç©ºé—´(z)ä»¥åŠåå˜é‡è¡¨ç¤º(b)åˆå¹¶åï¼Œè¾“å…¥åˆ°MLPç½‘ç»œï¼Œå†ç»è‡ªç„¶æŒ‡æ•°è½¬æ¢åï¼Œå¾—åˆ°æœ€ç»ˆè¾“å‡º (N, k)</p>
<p>ä½¿ç”¨NBè´ŸäºŒé¡¹åˆ†å¸ƒè®¡ç®—å¯¹äºæ©ç åŸºå› çš„é¢„æµ‹å€¼ä¸çœŸå®å€¼ä¹‹é—´çš„å·®å¼‚æŸå¤±(Target loss)</p>
<p>æœ€ç»ˆé¢„è®­ç»ƒæŸå¤±ç”±ä¸¤éƒ¨åˆ†ç»„æˆï¼šLatent loss + Target lossï¼Œè¿›è¡Œåå‘ä¼ æ’­ï¼Œæ›´æ–°æ¨¡å‹ã€‚</p>
<blockquote>
<p>é¢„è®­ç»ƒæ¨¡å‹åŒ…å«82Må‚æ•°ï¼Œå¯¹11Mç»†èƒæ•°æ®è¿›è¡Œè®­ç»ƒï¼Œ&lt;24 hours via 8 v100 GPUs</p>
<ul>
<li>scRNA-seq
<ul>
<li>HTCA (human tumor cell atlas)ï¼š4.7M</li>
<li>HCA (human cell atlas)ï¼š1.4M</li>
<li>GEOï¼š2.6M</li>
</ul>
</li>
<li>SRTï¼š2.7M</li>
</ul></blockquote>
<h1 id="3-å¾®è°ƒè®­ç»ƒ">3. å¾®è°ƒè®­ç»ƒ<a hidden class="anchor" aria-hidden="true" href="#3-å¾®è°ƒè®­ç»ƒ">#</a></h1>
<p>åŸºäºé¢„è®­ç»ƒæ¨¡å‹ï¼Œæ–‡ç« é‡‡ç”¨äº†å¤šè§’åº¦çš„å¾®è°ƒä»»åŠ¡ï¼Œå¹¶ä¸å…¶å®ƒå·²æœ‰æ¨¡å‹åšå¯¹æ¯”ï¼Œè¡¨æ˜CellPLMçš„ä¼˜åŠ¿ã€‚</p>
<p>å¯¹äºæ¯ä¸ªå¾®è°ƒä»»åŠ¡ï¼Œæ–‡ç« åŸºæœ¬éƒ½ä»Downstream Task Datasetsï¼ŒEvaluation Metricsï¼ŒBaselinesï¼ŒFine-tuningå±•å¼€ã€‚</p>
<p><strong>Cell-level tasks</strong></p>
<ul>
<li><strong>Zero shot</strong>ï¼šç›´æ¥ä½¿ç”¨é¢„è®­ç»ƒæ¨¡å‹ï¼Œè®¡ç®—ç»†èƒçš„åµŒå…¥è¡¨ç¤ºh (å¯ç”¨äºèšç±»åˆ†ç¾¤ç­‰)</li>
<li><strong>Cell Type Annotation</strong>ï¼šåŸºäºLatentè¾“å‡ºçš„cell embedï¼Œè¿æ¥MLPå±‚é¢„æµ‹ç»†èƒç±»å‹</li>
</ul>
<blockquote>
<p>é€šè¿‡æŸ¥çœ‹ä»“åº“ä»£ç ï¼Œä¸Šè¿°ä¸¤ç§å¾®è°ƒä»»åŠ¡éƒ½æ˜¯ç›´æ¥ä½¿ç”¨çš„Latent layerè¾“å‡ºçš„ç»†èƒåµŒå…¥è¡¨ç¤ºæ‰§è¡Œç›¸åº”ä»»åŠ¡ã€‚</p></blockquote>
<p><strong>Gene-level tasks</strong></p>
<ul>
<li>
<p><strong>scRNA-seq denoising</strong>ï¼šå°†ä¸€éƒ¨åˆ†éé›¶è¡¨è¾¾åŸºå› ç½®æ¢ä¸º0åï¼Œè®­ç»ƒæ¨¡å‹é¢„æµ‹å…¶åŸæœ‰è¡¨è¾¾å€¼ã€‚</p>
</li>
<li>
<p><strong>Spatial Transcriptomic imputation</strong>ï¼šæœ¬è´¨ä¸Šï¼Œä¹Ÿæ˜¯å°†ä¸€éƒ¨åˆ†åŸºå› ç½®æ¢ä¸º0ï¼Œ è®­ç»ƒæ¨¡å‹é¢„æµ‹åŸæœ‰è¡¨è¾¾å€¼ã€‚è®­ç»ƒè¿‡ç¨‹ä¸­ï¼Œéœ€è¦è”åˆscRNA-seqä½œä¸ºReferenceæ•°æ®ï¼Œå¸®åŠ©è®­ç»ƒã€‚</p>
</li>
<li>
<p><strong>Genetic perturbation prediction</strong>ï¼šé¢„æµ‹æ‰°åŠ¨åŸºå› è¡¨è¾¾ï¼Œå…³é”®æ˜¯å¦‚ä½•åœ¨è¾“å…¥æ•°æ®ä¸­æ ‡è®°å‡ºè¢«æ‰°åŠ¨çš„åŸºå› </p>
<ul>
<li>For one perturbation, we set the input of perturbed genes to be âˆ’100 to mimic the gene perturbation action.</li>
</ul>
</li>
</ul>
<h1 id="4-å…¶å®ƒæ–¹é¢">4. å…¶å®ƒæ–¹é¢<a hidden class="anchor" aria-hidden="true" href="#4-å…¶å®ƒæ–¹é¢">#</a></h1>
<p><strong>ï¼ˆ1ï¼‰æ¶ˆèå®éªŒ</strong></p>
<p>é€šè¿‡ä¸‰ä¸ªè§’åº¦çš„æ¶ˆèå®éªŒï¼Œè¯´æ˜CellPLMæ¨¡å‹componentæ‰€å‘æŒ¥çš„ä½œç”¨</p>
<ul>
<li>æ˜¯å¦é‡‡ç”¨Mixture of Gaussian</li>
<li>æ˜¯å¦é‡‡ç”¨Latent distribution layers</li>
<li>æ˜¯å¦é‡‡ç”¨Transformer encoder</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20241116222138940.png" alt="image-20241116222138940" style="zoom:67%;" />
<p><strong>ï¼ˆ2ï¼‰ä»£ç å€Ÿé‰´å­¦ä¹ </strong></p>
<ul>
<li>ç”±äºå•ç»†èƒè¡¨è¾¾çŸ©é˜µçš„ç¨€ç–æ€§ï¼Œtorchæ”¯æŒç›¸å…³é«˜æ•ˆçš„å¤„ç†æ–¹å¼
<ul>
<li>ä¾‹å¦‚ torch.sparse_csr_tensorï¼Œtorch.sparse.mmç­‰</li>
</ul>
</li>
<li>é€šè¿‡<code>from abc import ABC, abstractmethod</code>ï¼Œä½¿å¾—æ›´åŠ å®¹æ˜“åœ°å®ç°åŒä¸€ç±»å‹ï¼Œä¸åŒå˜ä½“çš„æ¨¡å‹component
<ul>
<li>ä¾‹å¦‚ ä¸åŒçš„å¾®è°ƒ Pipeline</li>
</ul>
</li>
<li>æ¯ä¸ªæ¨¡å‹componentéƒ½ç”±ä¸€ä¸ªæ–‡ä»¶å¤¹ç»„æˆï¼Œé€šè¿‡å…¶ä¸­<code>__init__.py</code>æ–‡ä»¶å‘æŒ¥å…³é”®ä½œç”¨ã€‚</li>
<li>Transformerï¼ŒVAEç­‰ç»å…¸æ¨¡å‹çš„å®ç°æ–¹å¼</li>
</ul>


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E6%96%87%E7%8C%AE-%E7%AE%97%E6%B3%95/">æ–‡çŒ®-ç®—æ³•</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/basic/203%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bscbert/">
    <span class="title">Â« Prev Page</span>
    <br>
    <span>æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹scBERT</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/basic/205%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bgeneformer/">
    <span class="title">Next Page Â»</span>
    <br>
    <span>æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹Geneformer</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>æ‚¨æ˜¯æœ¬ç«™ç¬¬ <span id="busuanzi_value_site_uv"></span> ä½è®¿é—®è€…ï¼Œæ€»æµè§ˆé‡ä¸º <span id="busuanzi_value_site_pv"></span> æ¬¡
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
