<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--MARVEL单细胞可变剪切 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
R包MARVEL是由牛津大学MRC Weatherall分子医学研究所团队开发的，用于分析单细胞水平的可变剪切事件。相关文章于2023年1月在Nucleic Acids Research期刊发表，在其github页面分享了MARVEL工具的分析流程，在此学习、记录如下。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/032%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--marvel%E5%8D%95%E7%BB%86%E8%83%9E%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/032%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--marvel%E5%8D%95%E7%BB%86%E8%83%9E%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--MARVEL单细胞可变剪切" />
<meta property="og:description" content="
R包MARVEL是由牛津大学MRC Weatherall分子医学研究所团队开发的，用于分析单细胞水平的可变剪切事件。相关文章于2023年1月在Nucleic Acids Research期刊发表，在其github页面分享了MARVEL工具的分析流程，在此学习、记录如下。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/032%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--marvel%E5%8D%95%E7%BB%86%E8%83%9E%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-03-06T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-03-06T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--MARVEL单细胞可变剪切"/>
<meta name="twitter:description" content="
R包MARVEL是由牛津大学MRC Weatherall分子医学研究所团队开发的，用于分析单细胞水平的可变剪切事件。相关文章于2023年1月在Nucleic Acids Research期刊发表，在其github页面分享了MARVEL工具的分析流程，在此学习、记录如下。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--MARVEL单细胞可变剪切",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/032%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--marvel%E5%8D%95%E7%BB%86%E8%83%9E%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--MARVEL单细胞可变剪切",
  "name": "单细胞分析工具--MARVEL单细胞可变剪切",
  "description": " R包MARVEL是由牛津大学MRC Weatherall分子医学研究所团队开发的，用于分析单细胞水平的可变剪切事件。相关文章于2023年1月在Nucleic Acids Research期刊发表，在其github页面分享了MARVEL工具的分析流程，在此学习、记录如下。\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " R包MARVEL是由牛津大学MRC Weatherall分子医学研究所团队开发的，用于分析单细胞水平的可变剪切事件。相关文章于2023年1月在Nucleic Acids Research期刊发表，在其github页面分享了MARVEL工具的分析流程，在此学习、记录如下。\nPape：https://doi.org/10.1093/nar/gkac1260 Github：https://github.com/wenweixiong/MARVEL Plate pipeline：https://wenweixiong.github.io/MARVEL_Plate.html Droplet pipeline：https://wenweixiong.github.io/MARVEL_Droplet.html 可变剪切\n关于可变剪切(Alternative splicing，AS)的基础，可以参看一篇中文教程；其中关于基因结构组成，可以参考之前的笔记。\n过去，针对Bulk RNA-seq数据的可变剪切分析工具已经有很多，例如之前学习的IsoformSwitchAnalyzeR包。MARVEL是个人目前了解到的第一个针对单细胞转录数据AS分析工具。\n单细胞数据类型\nAS分析的核心是通过检测比对到splice junction位点的read（例如，跨越两个外显子）进行分析。 目前主流的单细胞测序技术主要分为Droplet-based 10X方法与Plate-based Smart-seq方法。 10X技术由于仅对cDNA的3’端部分进行测序，只能对该区域涉及的splice junction进行分析，且无法具体分析事件类型 Smart-seq技术虽然通量低，但是全长转录本的单细胞测序原理，可以进行更为深入的AS分析。 一、Plate-based pipeline 1、上游分析 1.1 工作环境 两个同级的目录（1）basic：参考数据以及代码脚本 （2）proj1：具体的分析项目文件\n后续分析均在 proj1/ 路径下操作 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 tree basic/ -L 2 # basic/ # ├── gtf # │ ├── gencode.v31.annotation.gtf # │ ├── GRCh38.primary_assembly.genome.fa # │ ├── Log.out # │ ├── rsem_index # │ └── star_index # ├── rscript_Anno_SJ_rMATs.R # ├── rscript_bedtools_input.R # └── single_sample_upstream.sh tree proj1/ -L 2 # proj1/ # ├── SJ_phenoData.txt # └── work # ├── fq # ├── intron # ├── merge # ├── rmats # ├── rsem # └── star 1.2 conda环境 1 2 3 4 5 6 7 # 名为marvel_plate的分析环境 conda create -n marvel_plate python=3.7 conda activate marvel_plate conda install -c hcc aspera-cli conda install -c bioconda star=2.7.1a rmats=4.1.0 -y conda install -c bioconda bedtools samtools rsem -y conda install -c r r-base r-essentials r-tidyverse -y 1.2 准备数据 （1）测序数据\n官方手册分析了如下数据集的 192个单细胞数据，以其中一个演示流程 1 2 3 4 5 6 7 8 9 10 11 # https://www.ebi.ac.uk/ena/browser/view/ERR1562083 id=ERR1562083 # id=$1 mkdir -p ./work/fq/${id} ascp -QT -l 300m -P33001 \\ -i ~/miniconda3/envs/marvel_plate/etc/asperaweb_id_dsa.openssh \\ era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/${id:0:6}/00${id:0-1}/${id}/${id}_1.fastq.gz ./work/fq/${id} ascp -QT -l 300m -P33001 \\ -i ~/miniconda3/envs/marvel_plate/etc/asperaweb_id_dsa.openssh \\ era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/${id:0:6}/00${id:0-1}/${id}/${id}_2.fastq.gz ./work/fq/${id} （2）基因组数据\n在 ./basic/gtf 目录下 https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/ 1 2 3 4 5 6 7 8 9 10 11 12 13 ## STAR索引 STAR --runMode genomeGenerate \\ --genomeDir star_index \\ --genomeFastaFiles GRCh38.primary_assembly.genome.fa \\ --sjdbGTFfile gencode.v31.annotation.gtf \\ --runThreadN 40 ## RSEM索引 rsem-prepare-reference \\ --gtf gencode.v31.annotation.gtf \\ --star -p 20 \\ GRCh38.primary_assembly.genome.fa \\ rsem_index/hg38 （3）教程数据\n官方教程提供了各个分析步骤的示范文件，可供之后分析作为参考。 http://datashare.molbiol.ox.ac.uk/public/wwen/MARVEL/Tutorial/Plate/Data.zip 1 2 ls Data # GTF MARVEL rMATS R Object RSEM SJ 1.3 分析步骤 1.3.1 单样本4步分析 （1）单样本STAR比对\n将每个样本的Fastq文件使用STAR软件进行比对，产生每个splice junction count matrix\n1 2 # Reference less -SN ./Data/SJ/SJ.txt 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 mkdir -p ./work/star/${id} ref_idx_star=../basic/gtf/star_index ref_gtf=../basic/gtf/gencode.v31.annotation.gtf echo ${id} is runing...... #1st pass mode STAR --runThreadN 16 \\ --genomeDir ${ref_idx_star} \\ --readFilesCommand zcat \\ --readFilesIn ./work/fq/${id}/${id}_1.fastq.gz ./work/fq/${id}/${id}_2.fastq.gz \\ --outFileNamePrefix ./work/star/${id}/${id}. \\ --outSAMtype None #2nd pass mode STAR --runThreadN 16 \\ --genomeDir ${ref_idx_star} \\ --readFilesCommand zcat \\ --readFilesIn ./work/fq/${id}/${id}_1.fastq.gz ./work/fq/${id}/${id}_2.fastq.gz \\ --outFileNamePrefix ./work/star/${id}/${id}. \\ --sjdbFileChrStartEnd ./work/star/${id}/*SJ.out.tab \\ --sjdbGTFfile ${ref_gtf} \\ --outSAMtype BAM SortedByCoordinate \\ --outSAMattributes NH HI AS nM XS \\ --quantMode TranscriptomeSAM （2）单样本rMATs注释\n根据上述比对BAM结果，注释出每个基因(转录本)涉及的可变剪切事件类型。使用rMATs软件，可注释出5种类型：SE、RI、MXE、A5SS、A3SS。\n1 2 3 4 5 6 7 find ./Data/rMATS/ -name *featureData.txt # ./Data/rMATS/SE/SE_featureData.txt # ./Data/rMATS/A3SS/A3SS_featureData.txt # ./Data/rMATS/MXE/MXE_featureData.txt # ./Data/rMATS/RI/RI_featureData.txt # ./Data/rMATS/A5SS/A5SS_featureData.txt less -SN ./Data/rMATS/SE/SE_featureData.txt 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 mkdir -p ./work/rmats/${id} echo ${id} is running...... echo \"./work/star/${id}/${id}.Aligned.sortedByCoord.out.bam\" \u003e ./work/star/${id}/BAM_fls.txt rmats.py \\ --b1 ./work/star/${id}//BAM_fls.txt \\ --gtf ${ref_gtf} \\ --od ./work/rmats/${id} \\ --tmp ./work/rmats/${id}/tmp \\ -t paired \\ --readLength 125 \\ --variable-read-length \\ --nthread 16 \\ --statoff # readLength参考如下 # zcat ./work/fq/${id}_1.fastq.gz | awk '{if(NR%4==2) print NR\"\\t\"$0\"\\t\"length($0)}' | less （3）单样本Intron矩阵\n统计Reads对于每个内含子的比对情况，逐碱基累加作为比对结果。\n1 2 # reference less -SN Data/MARVEL/PSI/RI/Counts_by_Region.txt 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 mkdir -p ./work/intron/${id} echo ${id} is running...... samtools view -H ./work/star/${id}/${id}.Aligned.sortedByCoord.out.bam | \\ grep SQ | cut -f 2 | awk '{ sub(/^SN:/, \"\"); print;}' \u003e ./work/star/${id}/sorted_chr_in_bam.txt # 预定义R脚本，生成bedtools计算所需的两个文件：染色体大小以及内含子坐标 # https://github.com/lishensuo/utils/blob/main/marvel/rscript_bedtools_input.R Rscript ../basic/rscript_bedtools_input.R ${id} bedtools coverage \\ -g ./work/intron/${id}/hg38.chrom.sizes.txt \\ -split \\ -sorted \\ -a ./work/intron/${id}/RI_Coordinates_sorted.bed \\ -b ./work/star/${id}/${id}.Aligned.sortedByCoord.out.bam \u003e \\ ./work/intron/${id}/intron_count.txt \\ -d （4）单样本gene矩阵\n最后需要对基因表达水平进行定量。教程推荐使用RSEM软件，并进行TPM类似的标准化处理\n1 2 # reference less -SN ./Data/RSEM/TPM.txt 1 2 3 4 5 6 7 8 mkdir -p ./work/rsem/${id} rsem-calculate-expression \\ --bam \\ --paired-end \\ -p 16 \\ ./work/star/${id}/${id}.Aligned.toTranscriptome.out.bam \\ ../basic/gtf/rsem_index/hg38 \\ ./work/rsem/${id}/${id} （1）如上，由于4步的分析过程中产生的中间数据可能超过数百兆；且SMART-seq往往涉及成百上千的细胞。为节约空间，可在每个样本分析完成后，删除不必要的数据。\n1 2 3 4 5 rm ./work/fq/${id}/${id}* rm ./work/star/${id}/*bam rm -rf ./work/rmats/${id}/tmp/* rm ./work/rsem/${id}/${id}.transcript.bam rm ./work/rsem/${id}/${id}.isoforms.results （2）多样本批量后台运行方式\n1 2 3 4 5 6 7 8 9 cat ../basic/single_sample_upstream.sh chmod u+x ../basic/single_sample_upstream.sh # 单样本测试 # ../basic/single_sample_upstream.sh ERR1562273 # parallel多线程批量运行 nohup parallel -j 5 ../basic/single_sample_upstream.sh {} ::: \\ $(grep -v \"sample\" ./SJ_phenoData.txt | awk '{print $1}') \\ 1\u003e ./work/single_sample_upstream.log 2\u003e\u00261 \u0026 1.3.2 整合多样本结果 即分步整合上游4个步骤中的每个步骤的所有样本结果；均使用R语言完成\n1 2 3 4 library(tidyverse) library(parallel) sample_meta = data.table::fread(\"./SJ_phenoData.txt\",data.table=F) samples=sample_meta$sample.id step1: 汇总剪切位点矩阵 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 sj.list = mclapply(samples, function(sample){ sj_sp = data.table::fread(paste0(\"./work/star/\",sample,\"/\",sample,\".SJ.out.tab\"), data.table=F) if(nrow(sj_sp)==0) return(NULL) sj_sp = sj_sp %\u003e% dplyr::mutate(`coord.intron`=paste0(V1,\":\",V2,\":\",V3)) %\u003e% dplyr::select(`coord.intron`, V7) colnames(sj_sp)[2] = sample return(sj_sp) },mc.cores=10) names(sj.list) = samples sj.list = sj.list[!unlist(lapply(sj.list, is.null))] sj_res = sj.list[[1]] for(x in sj.list[-1]){ sj_res = dplyr::full_join(sj_res, x) } for(sample in setdiff(samples,names(sj.list))){ sj_res[,sample]=NA } dim(sj_res) write.table(sj_res, file=paste0(\"./work/merge/SJ.txt\"), quote=F, row.names=F, sep=\"\\t\") step2:整理AS注释结果 这一步最为复杂，需要理解每一种剪切事件的定义以及正负链的差异 参考官方示意：https://wenweixiong.github.io/Splicing_Nomenclature 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 # https://github.com/lishensuo/utils/blob/main/marvel/rscript_Anno_SJ_rMATs.R source(paste0(\"../basic/rscript_Anno_SJ_rMATs.R\")) AS_types = c(\"SE\", \"RI\", \"MXE\", \"A3SS\", \"A5SS\") for(AS_type in AS_types){ # AS_type = AS_types[1] print(AS_type) fls_AS = list.files(paste0(\"./work/rmats\"), pattern=paste0(\"fromGTF.\",AS_type,\".txt\"), recursive=T, full.name=T) if (AS_type==\"SE\"){ func = se_func } else if (AS_type==\"RI\"){ func = ri_func } else if (AS_type==\"MXE\"){ func = mxe_func } else if (AS_type==\"A3SS\"){ func = a3ss_func } else if (AS_type==\"A5SS\"){ func = a5ss_func } df_AS = mclapply(fls_AS, function(fls){ rmats_AS = data.table::fread(fls, data.table=F) rmats_AS = rmats_AS %\u003e% dplyr::mutate(tran_id = func(.)) %\u003e% dplyr::rename(gene_id=GeneID,gene_short_name=geneSymbol) %\u003e% dplyr::select(tran_id, gene_id, gene_short_name) return(rmats_AS) },mc.cores=10) %\u003e% do.call(rbind, .) %\u003e% dplyr::distinct() write.table(df_AS, row.names=F, sep=\"\\t\",quote=F, file=paste0(\"./work/merge/as_\",AS_type,\"_featureData.txt\")) } step3：汇总intron矩阵 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 int_mt.list = mclapply(samples, function(sample){ # sample = samples[1] int_mt = data.table::fread(paste0(\"./work/intron/\",sample,\"/intron_count.txt\")) int_mt = int_mt %\u003e% dplyr::mutate(`coord.intron`=paste(V1,V2,V3, sep=\":\")) %\u003e% dplyr::group_by(coord.intron) %\u003e% dplyr::summarize(count=sum(V5)) %\u003e% dplyr::select(coord.intron, count) %\u003e% as.data.frame() colnames(int_mt)[2] = sample return(int_mt) },mc.cores=10) intron_res = int_mt.list[[1]] for(x in int_mt.list[-1]){ intron_res = dplyr::full_join(intron_res, x) } dim(intron_res) write.table(intron_res, file=paste0(\"./work/merge/intron_count_by_region.txt\"), quote=F, row.names=F, sep=\"\\t\") 汇总gene矩阵 1 2 3 4 5 6 7 8 9 10 11 12 13 14 gene_mt.list = mclapply(samples, function(sample){ fls=paste0(\"./work/rsem/\",sample,\"/\",sample,\".genes.results\") gene_mt = data.table::fread(fls, data.table=F) %\u003e% dplyr::select(gene_id, TPM) colnames(gene_mt)[2] = sample return(gene_mt) },mc.cores=10) gene_res = gene_mt.list[[1]] for(x in gene_mt.list[-1]){ gene_res = dplyr::full_join(gene_res, x) } dim(gene_res) write.table(gene_res, file=paste0(\"./work/merge/rsem_tpm.txt\"),, quote=F, row.names=F, sep=\"\\t\") 2、下游分析 后续根据教程提供的数据学习，需要安装的主要的MARVEL包以及其它辅助分析包，具体可参考教程。\n1 2 # 建议再单独建立一个conda R环境 install.packages(\"MARVEL\") 2.1 创建marvel对象 需要准备七类文件，读入为R语言对象\n样本表型data.frame：第一列名为sample.id 剪切位点计数矩阵：第一列名为coord.intron 剪切事件类型注释列表：由数据框包装而出的list，每个数据库至少包含tran_id与gene_id 内含子计数矩阵：第一列名为coord.intron 基因表达矩阵：经标准化处理(RPKM/FPKM/TPM)，暂时不要log转换 基因类型data.frame：至少包含3列，gene_id, gene_short_name, gene_type 上游分析所用到的GTF数据 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 library(tidyverse) library(MARVEL) # (1) 细胞表型分组 df.pheno = data.table::fread(\"./SJ_phenoData.txt\",data.table=F) # (2) Gene文件 df.tpm \u003c- read.table(\"./work/merge/rsem_tpm.txt\", sep=\"\\t\", header=TRUE) dim(df.tpm) # (3) GTF文件 gtf \u003c- data.table::fread(\"../basic/gtf/gencode.v31.annotation.gtf\", sep=\"\\t\", header=FALSE, na.strings=\"NA\",quote=\"\\\"\") %\u003e% as.data.frame() # (4) Gene类型 gene.feature = subset(gtf, V3==\"gene\") df.tpm.feature \u003c- str_split(gene.feature$V9, \";\", simplify=T) %\u003e% as.data.frame() %\u003e% dplyr::mutate(gene_id=str_match(V1, '\"(.*)\"')[,2]) %\u003e% dplyr::mutate(gene_short_name=str_match(V3, '\"(.*)\"')[,2]) %\u003e% dplyr::mutate(gene_type=str_match(V2, '\"(.*)\"')[,2]) %\u003e% dplyr::select(gene_id, gene_short_name, gene_type) %\u003e% dplyr::filter(gene_id %in% df.tpm$gene_id) df.tpm.feature = df.tpm.feature[match(df.tpm$gene_id, df.tpm.feature$gene_id),] rownames(df.tpm.feature) = seq(nrow(df.tpm.feature)) # (5) SJ文件 sj = data.table::fread(\"./work/merge/SJ.txt\", sep=\"\\t\", header=TRUE, na.strings=\"NA\") %\u003e% as.data.frame() dim(sj) # (6) AS文件 df.feature.list = lapply(c(\"SE\", \"MXE\", \"RI\", \"A5SS\", \"A3SS\"), function(x){ # x = \"SE\" df.feature = read.table(paste0(\"./work/merge/as_\",x,\"_featureData.txt\"), sep=\"\\t\", header=TRUE, na.strings=\"NA\") %\u003e% dplyr::distinct(tran_id, .keep_all=TRUE) %\u003e% dplyr::left_join(df.tpm.feature) return(df.feature) }) names(df.feature.list) \u003c- c(\"SE\", \"MXE\", \"RI\", \"A5SS\", \"A3SS\") # (7) Intron文件 df.intron.counts \u003c- data.table::fread(\"./work/merge/intron_count_by_region.txt\", sep=\"\\t\", header=TRUE, na.strings=\"NA\") %\u003e% as.data.frame() dim(df.intron.counts) ## MARVEL object marvel \u003c- CreateMarvelObject(SpliceJunction=sj, SplicePheno=df.pheno, SpliceFeature=df.feature.list, IntronCounts=df.intron.counts, GeneFeature=df.tpm.feature, Exp=df.tpm, GTF=gtf ) 2.2 预处理步骤 （1）基因表达log转换\n1 2 3 4 5 marvel \u003c- TransformExpValues(MarvelObject=marvel, offset=1, transformation=\"log2\", threshold.lower=1 ) （2）检测额外AS事件\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 # 如果基因表达水平整体较低，则不考虑该基因的可变剪切事件 # min.expr:判定单个细胞是否表达基因的阈值 # min.cells: 基于上述条件，判定细胞群表达该基因的阈值 marvel \u003c- DetectEvents(MarvelObject=marvel, min.cells=50, #细胞群的表达百分比 min.expr=1, #认为细胞表达该基因的阈值 track.progress=FALSE, EventType=\"AFE\" ) marvel \u003c- DetectEvents(MarvelObject=marvel, min.cells=50, min.expr=1, track.progress=FALSE, EventType=\"ALE\" ) length(marvel$SpliceFeature) # 7 （3）计算PSI值\n计算一个基因对于特定剪切事件的PSI值，这是后续分析的基础\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 # CoverageThreshold: 支持该剪切事件的最小reads数，小于该阈值标记为NA # UnevenCoverageMultiplier: 针对SE与MXE marvel \u003c- ComputePSI(MarvelObject=marvel, CoverageThreshold=10, UnevenCoverageMultiplier=10, EventType=\"SE\" ) marvel \u003c- ComputePSI(MarvelObject=marvel, CoverageThreshold=10, UnevenCoverageMultiplier=10, EventType=\"MXE\" ) marvel \u003c- ComputePSI(MarvelObject=marvel, CoverageThreshold=10, EventType=\"RI\", thread=4 # only support RI ) marvel \u003c- ComputePSI(MarvelObject=marvel, CoverageThreshold=10, EventType=\"A5SS\" ) marvel \u003c- ComputePSI(MarvelObject=marvel, CoverageThreshold=10, EventType=\"A3SS\" ) marvel \u003c- ComputePSI(MarvelObject=marvel, CoverageThreshold=10, EventType=\"AFE\" ) marvel \u003c- ComputePSI(MarvelObject=marvel, CoverageThreshold=10, EventType=\"ALE\" ) （4）保存中间结果\n首先检查不同矩阵与对应的meta行列名信息一致 1 2 3 4 marvel \u003c- CheckAlignment(MarvelObject=marvel, level=\"SJ\") marvel \u003c- CheckAlignment(MarvelObject=marvel, level=\"splicing\") marvel \u003c- CheckAlignment(MarvelObject=marvel, level=\"gene\") marvel \u003c- CheckAlignment(MarvelObject=marvel, level=\"splicing and gene\") 可进一步选取感兴趣的细胞群 1 2 3 4 5 6 7 index.1 \u003c- which(df.pheno$cell.type %in% c(\"iPSC\", \"Endoderm\")) index.2 \u003c- which(df.pheno$qc.seq==\"pass\") index \u003c- intersect(index.1, index.2) sample.ids \u003c- df.pheno[index, \"sample.id\"] marvel \u003c- SubsetSamples(MarvelObject=marvel, sample.ids=sample.ids ) 保存中间结果，方便后续分析 1 2 3 4 save(marvel, file=\"./MARVEL2.RData\") # 后续以官方提供的示例数据学习 # load(\"../Data/MARVEL.RData\") 2.3 细胞群AS特征分析 以其中84个iPSC细胞群进行演示\n1 2 3 sample.ids = marvel$SplicePheno %\u003e% dplyr::filter(cell.type==\"iPSC\") %\u003e% dplyr::pull(sample.id) （1）AS类型占比\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 marvel \u003c- CountEvents(MarvelObject=marvel, sample.ids=sample.ids, min.cells=25 ) # 对于一个基因至少有25个细胞的AS PSI值为非缺失值，则判定发生了可变剪切事件 marvel$N.Events$Table # event_type freq pct # 1 SE 5270 40.1523810 # 3 RI 2030 15.4666667 # 6 AFE 1789 13.6304762 # 5 A3SS 1746 13.3028571 # 4 A5SS 1494 11.3828571 # 7 ALE 724 5.5161905 # 2 MXE 72 0.5485714 marvel$N.Events$Plot # ggplot绘图体系，下同 （2）AS modality\n判断细胞群中每个基因的每种可变剪切的PSI分布类型，如下可分为7类\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 marvel \u003c- AssignModality(MarvelObject=marvel, sample.ids=sample.ids, min.cells=25, seed=1 ) marvel$Modality$Results[1:5, c(\"tran_id\", \"event_type\", \"gene_id\", \"gene_short_name\", \"modality.bimodal.adj\")] table(marvel$Modality$Results$modality.bimodal.adj, marvel$Modality$Results$event_type) # A3SS A5SS AFE ALE MXE RI SE # Bimodal 20 7 4 6 0 4 17 # Excluded.Dispersed 386 484 538 189 18 783 1260 # Excluded.Primary 289 282 477 101 22 712 1049 # Included.Dispersed 563 385 372 237 12 352 1685 # Included.Primary 439 291 346 176 17 47 1184 # Middle 9 10 9 3 2 51 15 # Multimodal 40 35 43 12 1 81 60 后续有两种统计方式：仅按modality分，同时按modality与AS type分。以后者作为演示 1 2 3 4 5 6 7 8 9 10 11 12 marvel \u003c- PropModality(MarvelObject=marvel, modality.column=\"modality.bimodal.adj\", modality.type=\"extended\", event.type=c(\"SE\", \"MXE\", \"RI\", \"A5SS\", \"A3SS\", \"AFE\", \"ALE\"), across.event.type=TRUE, prop.test=\"chisq\", prop.adj=\"fdr\", xlabels.size=8 ) marvel$Modality$Prop$BarChart$Stats head(marvel$Modality$Prop$BarChart$Table) marvel$Modality$Prop$BarChart$Plot 2.4 细胞群差异分析 （1）差异基因分析\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 cell.group.g1 = marvel$SplicePheno %\u003e% dplyr::filter(cell.type==\"iPSC\") %\u003e% dplyr::pull(sample.id) # (reference) cell.group.g2 = marvel$SplicePheno %\u003e% dplyr::filter(cell.type==\"Endoderm\") %\u003e% dplyr::pull(sample.id) marvel \u003c- CompareValues(MarvelObject=marvel, cell.group.g1=cell.group.g1, cell.group.g2=cell.group.g2, min.cells=3, method=\"wilcox\", method.adjust=\"fdr\", level=\"gene\", show.progress=FALSE ) marvel$DE$Exp$Table[1:5, ] （2）差异AS分析\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 # time consuming marvel \u003c- CompareValues(MarvelObject=marvel, cell.group.g1=cell.group.g1, cell.group.g2=cell.group.g2, min.cells=25, method=c(\"ad\", \"dts\"), #同时使用两种的推荐差异方法 method.adjust=\"fdr\", level=\"splicing\", event.type=c(\"SE\", \"MXE\", \"RI\", \"A5SS\", \"A3SS\", \"ALE\", \"AFE\"), show.progress=FALSE ) names(marvel$DE$PSI$Table) head(marvel$DE$PSI$Table[[\"ad\"]]) head(marvel$DE$PSI$Table[[\"dts\"]]) （3）差异基因\u0026AS分析\n对差异可变剪切的基因执行基因表达差异分析 1 2 3 4 5 6 7 8 9 10 11 12 13 marvel \u003c- CompareValues(MarvelObject=marvel, cell.group.g1=cell.group.g1, cell.group.g2=cell.group.g2, psi.method=c(\"ad\", \"dts\"), psi.pval=c(0.10, 0.10), psi.delta=0, method.de.gene=\"wilcox\", method.adjust.de.gene=\"fdr\", downsample=FALSE, show.progress=FALSE, level=\"gene.spliced\" ) head(marvel$DE$Exp.Spliced$Table) 对于上述差异分析结果均可快速火山图，标记marker基因等\n2.5 差异AS进阶分析 （1）差异PSI分类\nExplicit：明显的变化；Implicit：有限的变化；Restricted：微小的变化 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 marvel \u003c- ModalityChange(MarvelObject=marvel, method=c(\"ad\", \"dts\"), psi.pval=c(0.10, 0.10) ) table(marvel$DE$Modality$Table$modality.change) # Explicit Implicit Restricted # 161 280 1113 head(marvel$DE$Modality$Table) marvel$DE$Modality$Plot cell.group.list \u003c- list(\"iPSC\"=cell.group.g1, \"Endoderm\"=cell.group.g2 ) tran_id \u003c- \"chr3:129171771:129171634|129171655:-@chr3:129171446:129171538\" # tran_id \u003c- \"chr6:21596763:21598248:+@chr6:21598321:21599378\" # tran_id \u003c- \"chr3:109337464:109337652:-@chr3:109333920|109333993:109333870\" marvel \u003c- PlotValues(MarvelObject=marvel, cell.group.list=cell.group.list, feature=tran_id, xlabels.size=5, level=\"splicing\", min.cells=25 ) marvel$adhocPlot$PSI （2）差异基因与AS\nCoordinated：基因高表达，相应的PSI也增高； Opposing：基因表达与PSI变化方向相反； Isoform-switching ：基因表达不变，PSI发生变化 Complex：其它复杂情况 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 marvel \u003c- IsoSwitch(MarvelObject=marvel, method=c(\"ad\", \"dts\"), psi.pval=c(0.10, 0.10), psi.delta=0, gene.pval=0.10, gene.log2fc=0.5 ) marvel$DE$Cor$Plot head(marvel$DE$Cor$Table_Raw) table(marvel$DE$Cor$Table$cor) # Coordinated Opposing Iso-Switch Complex # 192 158 331 113 ## 以Coordinated的DHX9与相应的AFE剪切事件为例 # gene df.feature \u003c- marvel$GeneFeature gene_id \u003c- df.feature[which(df.feature$gene_short_name==\"DHX9\"), \"gene_id\"] marvel \u003c- PlotValues(MarvelObject=marvel, cell.group.list=cell.group.list, feature=gene_id, maintitle=\"gene_short_name\", xlabels.size=7, level=\"gene\" ) plot.1_gene \u003c- marvel$adhocPlot$Exp # Splicing tran_id \u003c- \"chr1:182839347:182839456|182839734:182839935:+@chr1:182842545:182842677\" marvel \u003c- PlotValues(MarvelObject=marvel, cell.group.list=cell.group.list, feature=tran_id, xlabels.size=7, level=\"splicing\", min.cells=25 ) plot.1_splicing \u003c- marvel$adhocPlot$PSI library(patchwork) plot.1_gene + plot.1_splicing （3）NMD预测\nnonsense-mediated decay (NMD)无义介导的mRNA降解预测 相比对照组，观测组所剪切的外显子可能由于包含premature stop codons (PTCs)，导致编码提前结束，从而进一步使得表达量降低。 本质上是对可产生PTC的可变剪切与下调基因取交集。 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 marvel \u003c- ParseGTF(MarvelObject=marvel) marvel \u003c- FindPTC(MarvelObject=marvel, method=c(\"ad\", \"dts\"), pval=c(0.10, 0.10), delta=5 ) marvel \u003c- PropPTC(MarvelObject=marvel, xlabels.size=8, show.NovelSJ.NoCDS=FALSE, #仅展示Non PTC与PTC prop.test=\"chisq\" ) marvel$NMD$PTC.Prop$Table[1:5, ] marvel \u003c- CompareExpr(MarvelObject=marvel, xlabels.size=8 ) marvel$NMD$NMD.Expr$Plot marvel \u003c- AnnoVolcanoPlot(MarvelObject=marvel, anno=FALSE, xlabel.size=7 ) marvel$NMD$AnnoVolcanoPlot$Plot marvel$NMD$AnnoVolcanoPlot$Table 二、Drop-baed pipeline 1、上游分析 1.1 conda环境 1 2 3 4 5 conda create -n marvel_drop conda activate marvel_drop conda install -c hcc aspera-cli conda install -c bioconda STAR=2.7.8a samtools 1.2 下载数据 （1）测序数据\n官方教程分析了GSE130731的4个10X单细胞测序结果，以其中的两个作为学习的示例(SRR9008752,SRR9008753) https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130731 1 2 3 4 5 6 7 8 9 10 ascp -QT -l 300m -P33001 \\ -i ~/miniconda3/envs/marvel/etc/asperaweb_id_dsa.openssh \\ era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/SRR900/003/SRR9008752/SRR9008752_1.fastq.gz ./work/fq/ ascp -QT -l 300m -P33001 \\ -i ~/miniconda3/envs/marvel/etc/asperaweb_id_dsa.openssh \\ era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/SRR900/003/SRR9008753/SRR9008753_2.fastq.gz ./work/fq/ mv ./work/fq/SRR9008753_1.fastq.gz ./work/fq/SRR9008753_S1_L001_R1_001.fastq.gz mv ./work/fq/SRR9008753_2.fastq.gz ./work/fq/SRR9008753_S1_L001_R2_001.fastq.gz （2）基因组数据\n还是在 ./basic/gtf 文件夹下，构建starsolo版本的索引文件。\ncellranger软件及相关数据下载参考之前的笔记。\n1 2 3 4 5 6 STAR \\ --runMode genomeGenerate \\ --runThreadN 20 \\ --genomeDir ./starsolo_index \\ --genomeFastaFiles GRCh38.primary_assembly.genome.fa \\ --sjdbGTFfile gencode.v31.annotation.gtf （3）教程数据\nhttp://datashare.molbiol.ox.ac.uk/public/wwen/MARVEL/Tutorial/Droplet/Data.zip 1 2 ls Data/ # Gene_SingCellaR Gene_STARsolo GTF R Object SJ_STARsolo 1.3 比对分析 （1）cellranger\n使用cellranger软件初次比对产生bam文件，该软件用法可参考之前的笔记\n1 2 3 4 5 6 7 8 9 10 11 12 13 bin=/home/lishensuo/Softerware/cellranger-7.1.0/bin/cellranger ref=/home/lishensuo/Softerware/cellranger-7.1.0/refer/refdata-gex-GRCh38-2020-A fq_dir=/home/lishensuo/STUDY/marvel/proj2/work/fq sample=SRR9008752 ${bin} count --id=${sample} \\ --fastqs=${fq_dir} \\ --sample=${sample} \\ --localcores=10 \\ --transcriptome=${ref} mkdir -p ./work/cellranger mv ${sample} ./work/cellranger/${sample} （2）STARsolo\n使用STAR软件针对单细胞数据的模式，分析其基因表达以及剪切位点矩阵\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 # https://kb.10xgenomics.com/hc/en-us/articles/115004506263-What-is-a-barcode-whitelist- # Cell Ranger 4 and later: cellranger-x.y.z/lib/python/cellranger/barcodes/ # Cell Ranger 3 and earlier: cellranger-x.y.z/cellranger-cs/x.y.z/lib/python/cellranger/barcodes/ ls /home/lishensuo/Softerware/cellranger-7.1.0/lib/python/cellranger/barcodes/ whitelist=/home/lishensuo/Softerware/cellranger-7.1.0/lib/python/cellranger/barcodes/737K-august-2016.txt sample=SRR9008752 STAR --runThreadN 16 \\ --genomeDir ../basic/gtf/starsolo_index \\ --soloType CB_UMI_Simple \\ --readFilesIn ./work/cellranger/${sample}/outs/possorted_genome_bam.bam \\ --readFilesCommand samtools view -F 0x100 \\ --readFilesType SAM SE \\ --soloInputSAMattrBarcodeSeq CR UR \\ --soloInputSAMattrBarcodeQual CY UY \\ --soloCBwhitelist ${whitelist} \\ --soloFeatures Gene SJ mkdir ./work/star/${sample} mv Aligned.out.sam Log.final.out Log.out Log.progress.out SJ.out.tab ./Solo.out mv Solo.out ./work/star/${sample} 2、下游分析 2.1 整理比对结果 1 2 3 4 5 6 library(tidyverse) library(Seurat) library(R.utils) # ※ library(Matrix) samples=c(\"SRR9008752\",\"SRR9008753\") （1）Raw gene count\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 dir.create(\"./work/gene\") dir.create(\"./work/gene/raw\") dir.create(\"./work/gene/norm\") sce.list = lapply(samples, function(x){ print(x) fls = list.files(paste0(\"./work/star/\",x,\"/Solo.out/Gene/filtered\"), full.name=T) if(length(grep(\"gz\",fls))!=3){ for(fl in fls){ gzip(fl, remove=FALSE) } } count = Read10X(paste0(\"./work/star/\",x,\"/Solo.out/Gene/filtered\")) sce = CreateSeuratObject(count) sce = RenameCells(sce, add.cell.id=x) sce$Group = x return(sce) }) names(sce.list) = samples # 若多个样本则合并处理 sce = merge(sce.list[[1]], sce.list[-1]) table(sce$Group) count_merge = sce@assays$RNA@counts dim(count_merge) writeMM(count_merge, file = './work/gene/raw/matrix.mtx') count_merge_phe = sce@meta.data %\u003e% tibble::rownames_to_column(\"cell.id\") %\u003e% dplyr::select(cell.id) # 仅需要1列 write.table(count_merge_phe, sep=\"\\t\", row.names=F, file=\"./work/gene/raw/phenoData.txt\") count_merge_feat = data.frame(gene_short_name=rownames(sce)) write.table(count_merge_feat, sep=\"\\t\", row.names=F, file=\"./work/gene/raw/featureData.txt\") （2）Norm gene count\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 sce = NormalizeData(sce) count_merge_norm = Matrix::Matrix(exp(sce@assays$RNA@data)-1, sparse=T) # 非log转换 count_merge_norm[1:4,1:4] writeMM(count_merge_norm, file = './work/gene/norm/matrix_normalised.mtx') count_merge_norm_phe = sce@meta.data %\u003e% tibble::rownames_to_column(\"cell.id\") %\u003e% dplyr::select(cell.id, Group) # 需要细胞分组等表型信息 write.table(count_merge_norm_phe, sep=\"\\t\", row.names=F, file=\"./work/gene/norm/phenoData.txt\") count_merge_norm_feat = data.frame(gene_short_name=rownames(sce)) write.table(count_merge_norm_feat, sep=\"\\t\", row.names=F, file=\"./work/gene/norm/featureData.txt\") （3）Raw SJ count\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 dir.create(\"./work/sj/raw\") # 整理SJ meta for(x in samples){ sj_name = data.table::fread(paste0(\"./work/star/\",x,\"/Solo.out/SJ.out.tab\"), data.table=F) %\u003e% dplyr::mutate(coord.intron=paste(V1, V2, V3, sep=\":\")) %\u003e% dplyr::select(coord.intron) write.table(sj_name, sep=\"\\t\", row.names=F, col.names=F, quote=F, file=paste0(\"./work/star/\",x,\"/Solo.out/SJ/raw/features.tsv\")) } sce_sj.list = lapply(samples, function(x){ # x=\"SRR15319877\" print(x) fls = list.files(paste0(\"./work/star/\",x,\"/Solo.out/SJ/raw\"), full.name=T) if(length(grep(\"gz\",fls))!=3){ for(fl in fls){ gzip(fl, remove=FALSE) } } count = Read10X(paste0(\"./work/star/\",x,\"/Solo.out/SJ/raw\"),gene.column=1) sce_sj = CreateSeuratObject(count) sce_sj = RenameCells(sce_sj, add.cell.id=x) sce_sj = sce_sj[,colnames(sce.list[[x]])] return(sce_sj) }) sce_sj = merge(sce_sj.list[[1]], sce_sj.list[-1]) sj_count_merge = sce_sj@assays$RNA@counts writeMM(sj_count_merge, file = './work/sj/raw/matrix.mtx') sj_count_merge_phe = sce_sj@meta.data %\u003e% tibble::rownames_to_column(\"cell.id\") %\u003e% dplyr::select(cell.id) # 仅需要1列 write.table(sj_count_merge_phe, sep=\"\\t\", row.names=F, file=\"./work/sj/raw/phenoData.txt\") sj_count_merge_feat = data.frame(coord.intron=rownames(sce_sj)) %\u003e% dplyr::mutate(coord.intron=gsub(\"-\",\"_\", coord.intron)) write.table(sj_count_merge_feat, sep=\"\\t\", row.names=F, file=\"./work/sj/raw/featureData.txt\") （4）Cell dimension reduction\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 dir.create(\"./work/umap\") sce = sce %\u003e% NormalizeData() %\u003e% FindVariableFeatures() %\u003e% ScaleData() %\u003e% Seurat::RunPCA() %\u003e% Seurat::RunUMAP(dims = 1:30) umap_embed = sce@reductions$umap@cell.embeddings %\u003e% as.data.frame() %\u003e% tibble::rownames_to_column(\"cell.id\") %\u003e% dplyr::rename(x=UMAP_1, y=UMAP_2) write.table(umap_embed, row.names=F, sep=\"\\t\", file=\"./work/umap/UMAP_coordinate.txt\") 2.2 创建marvel对象 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 # (1) norm gene的三文件 df.gene.norm \u003c- readMM(\"./work/gene/norm/matrix_normalised.mtx\") df.gene.norm.pheno \u003c- read.table(\"./work/gene/norm/phenoData.txt\", sep=\"\\t\", header=TRUE) df.gene.norm.feature \u003c- read.table(\"./work/gene/norm/featureData.txt\", sep=\"\\t\", header=TRUE) # (2) raw gene的三文件 df.gene.count \u003c- readMM(\"./work/gene/raw/matrix.mtx\") df.gene.count.pheno \u003c- read.table(\"./work/gene/raw/phenoData.txt\", sep=\"\\t\", header=TRUE) df.gene.count.feature \u003c- read.table(\"./work/gene/raw/featureData.txt\", sep=\"\\t\", header=TRUE) # (3) SJ gene的三文件 df.sj.count \u003c- readMM(\"./work/sj/raw/matrix.mtx\") df.sj.count.pheno \u003c- read.table(\"./work/sj/raw/phenoData.txt\", sep=\"\\t\", header=TRUE) df.sj.count.feature \u003c- read.table(\"./work/sj/raw/featureData.txt\", sep=\"\\t\", header=TRUE) rownames(df.sj.count) = df.sj.count.feature[,1] colnames(df.sj.count) = df.sj.count.pheno[,1] # (4) 细胞降维 df.coord \u003c- read.table(\"./work/umap/UMAP_coordinate.txt\", sep=\"\\t\", header=TRUE) # (5) GTF gtf \u003c- data.table::fread(\"../basic/gtf/gencode.v31.annotation.gtf\", sep=\"\\t\", header=FALSE) %\u003e% as.data.frame() ## 由于不同来源gtf版本差异而踩的坑，需要修改如下 gtf$V9 = gsub(\"gene_type\",\"gene_biotype\",gtf$V9) gtf$V1 = gsub(\"chr\",\"\",gtf$V1) library(MARVEL) marvel \u003c- CreateMarvelObject.10x(gene.norm.matrix=df.gene.norm, gene.norm.pheno=df.gene.norm.pheno, gene.norm.feature=df.gene.norm.feature, gene.count.matrix=df.gene.count, gene.count.pheno=df.gene.count.pheno, gene.count.feature=df.gene.count.feature, sj.count.matrix=df.sj.count, sj.count.pheno=df.sj.count.pheno, sj.count.feature=df.sj.count.feature, pca=df.coord, gtf=gtf ) 2.3 预处理步骤 筛选有效的剪切位点 1 2 3 4 5 6 7 8 9 10 11 12 13 # 注释基因类型 marvel \u003c- AnnotateGenes.10x(MarvelObject=marvel) head(marvel$gene.metadata) ## SJ涉及两边分属一个蛋白编码基因 marvel \u003c- AnnotateSJ.10x(MarvelObject=marvel) #基因类型 head(marvel$sj.metadata) marvel \u003c- ValidateSJ.10x(MarvelObject=marvel) #同一基因 head(marvel$sj.metadata) marvel \u003c- FilterGenes.10x(MarvelObject=marvel, gene.type=\"protein_coding\" #蛋白基因 ) head(marvel$sj.metadata) 分析基因在细胞群的表达率以及SJ在细胞群的发生率 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 cell.group.list = split(marvel$sample.metadata$cell.id,marvel$sample.metadata$Group) ## 基因在细胞群的表达率 marvel \u003c- PlotPctExprCells.Genes.10x(MarvelObject=marvel, cell.group.g1=cell.group.list[[1]], cell.group.g2=cell.group.list[[2]], min.pct.cells=5 ) head(marvel$pct.cells.expr$Gene$Data) p1 = marvel$pct.cells.expr$Gene$Plot ## 剪切位点在不同细胞群的发生率 marvel \u003c- PlotPctExprCells.SJ.10x(MarvelObject=marvel, cell.group.g1=cell.group.list[[1]], cell.group.g2=cell.group.list[[2]], min.pct.cells.genes=5, min.pct.cells.sj=5, downsample=TRUE, downsample.pct.sj=10 # 抽样10%分析 ) head(marvel$pct.cells.expr$SJ$Data) p2 = marvel$pct.cells.expr$SJ$Plot library(patchwork) p1 + p2 保存中间结果 1 2 marvel \u003c- CheckAlignment.10x(MarvelObject=marvel) save(marvel, file=\"***.rda\") 2.4 SJ差异分析 （1）差异SJ与基因\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 ## 首先分析差异SJ marvel \u003c- CompareValues.SJ.10x(MarvelObject=marvel, cell.group.g1=cell.group.list[[1]], cell.group.g2=cell.group.list[[2]], min.pct.cells.genes=10, #至少10%细胞表达 min.pct.cells.sj=10, #至少10%细胞剪切 min.gene.norm=1.0, seed=1, n.iterations=100, downsample=TRUE, #根据样本量少的一组下采样 show.progress=FALSE ) ## 然后进一步分析差异基因 marvel \u003c- CompareValues.Genes.10x(MarvelObject=marvel, show.progress=FALSE ) head(marvel$DE$SJ$Table) （2）SJ与基因关系\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 marvel \u003c- IsoSwitch.10x(MarvelObject=marvel, pval.sj=0.05, delta.sj=5, min.gene.norm=1.0, pval.adj.gene=0.05, log2fc.gene=0.5 ) head(marvel$SJ.Gene.Cor$Data[,c(\"coord.intron\", \"gene_short_name\", \"cor.complete\")]) marvel$SJ.Gene.Cor$Proportion$Table # sj.gene.cor freq pct # 2 Coordinated 26 8.387097 # 4 Opposing 54 17.419355 # 3 Iso-Switch 226 72.903226 # 1 Complex 4 1.290323 marvel$SJ.Gene.Cor$Proportion$Plot 降维可视化 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 # Plot cell groups marvel \u003c- PlotValues.PCA.CellGroup.10x(MarvelObject=marvel, cell.group.list=cell.group.list[c(1,4)], legendtitle=\"Cell group\", type=\"tsne\" ) plot_group \u003c- marvel$adhocPlot$PCA$CellGroup # Plot gene expression marvel \u003c- PlotValues.PCA.Gene.10x(MarvelObject=marvel, gene_short_name=\"VIM\", color.gradient=c(\"grey\",\"cyan\",\"green\",\"yellow\",\"red\"), type=\"tsne\" ) plot_gene \u003c- marvel$adhocPlot$PCA$Gene # Plot PSI marvel \u003c- PlotValues.PCA.PSI.10x(MarvelObject=marvel, coord.intron=\"chr10:17235390:17235845\", min.gene.count=3, log2.transform=FALSE, color.gradient=c(\"grey\",\"cyan\",\"green\",\"yellow\",\"red\"), type=\"tsne\" ) plot_sj \u003c- marvel$adhocPlot$PCA$PSI library(patchwork) p = plot_group + plot_gene + plot_sj 2.5 特定基因分析 （1）分组表达\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 # 该基因在不同阶段的表达 marvel \u003c- adhocGene.TabulateExpression.Gene.10x(MarvelObject=marvel, cell.group.list=cell.group.list, gene_short_name=\"TPM2\", min.pct.cells=10, downsample=TRUE ) marvel$adhocGene$Expression$Gene$Table marvel$adhocGene$Expression$Gene$Plot # 该基因的不同剪切位点在不同阶段的count表达 marvel \u003c- adhocGene.TabulateExpression.PSI.10x(MarvelObject=marvel, min.pct.cells=10 ) head(marvel$adhocGene$Expression$PSI$Table) marvel$adhocGene$Expression$PSI$Plot （2）组间差异\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 # 组间两两差异 marvel \u003c- adhocGene.DE.Gene.10x(MarvelObject=marvel) marvel \u003c- adhocGene.DE.PSI.10x(MarvelObject=marvel) results \u003c- marvel$adhocGene$DE$PSI$Data head(results) # 选择特定SJ绘图 # SJ-1 # Define SJ coord.intron \u003c- results[which(results$figure.column==\"SJ-1\"), \"coord.intron\"] coord.intron \u003c- unique(coord.intron) coord.intron marvel \u003c- adhocGene.PlotDEValues.10x(MarvelObject=marvel, coord.intron=coord.intron, log2fc.gene=0.5, delta.sj=5, label.size=2, point.size=2, xmin=-2.0, xmax=2.0, ymin=-25, ymax=25 ) marvel$adhocGene$DE$VolcanoPlot$Plot 使用wiggleplotr包可视化某剪切位点在基因不同转录本的可视化 1 2 3 4 5 6 7 8 9 10 11 # http://datashare.molbiol.ox.ac.uk/public/wwen/wiggleplotr_1.18.0.tar.gz #install.packages(\"wiggleplotr_1.18.0.tar.gz\", repos = NULL, type=\"source\") marvel \u003c- adhocGene.PlotSJPosition.10x(MarvelObject=marvel, coord.intron=coord.intron, rescale_introns=FALSE, show.protein.coding.only=TRUE, anno.label.size=1.5 ) marvel$adhocGene$SJPosition$Plot ",
  "wordCount" : "6473",
  "inLanguage": "en",
  "datePublished": "2023-03-06T00:00:00Z",
  "dateModified": "2023-03-06T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/032%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--marvel%E5%8D%95%E7%BB%86%E8%83%9E%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--MARVEL单细胞可变剪切
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-03-06 00:00:00 &#43;0000 UTC&#39;&gt;2023-03-06&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-03-06&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;6473&amp;nbsp;|&amp;nbsp;13 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#%e4%b8%80plate-based-pipeline" aria-label="一、Plate-based pipeline">一、Plate-based pipeline</a><ul>
                            
                    <li>
                        <a href="#1%e4%b8%8a%e6%b8%b8%e5%88%86%e6%9e%90" aria-label="1、上游分析">1、上游分析</a><ul>
                            
                    <li>
                        <a href="#11-%e5%b7%a5%e4%bd%9c%e7%8e%af%e5%a2%83" aria-label="1.1 工作环境">1.1 工作环境</a></li>
                    <li>
                        <a href="#12-conda%e7%8e%af%e5%a2%83" aria-label="1.2 conda环境">1.2 conda环境</a></li>
                    <li>
                        <a href="#12-%e5%87%86%e5%a4%87%e6%95%b0%e6%8d%ae" aria-label="1.2 准备数据">1.2 准备数据</a></li>
                    <li>
                        <a href="#13-%e5%88%86%e6%9e%90%e6%ad%a5%e9%aa%a4" aria-label="1.3 分析步骤">1.3 分析步骤</a><ul>
                            
                    <li>
                        <a href="#131-%e5%8d%95%e6%a0%b7%e6%9c%ac4%e6%ad%a5%e5%88%86%e6%9e%90" aria-label="1.3.1 单样本4步分析">1.3.1 单样本4步分析</a></li>
                    <li>
                        <a href="#132-%e6%95%b4%e5%90%88%e5%a4%9a%e6%a0%b7%e6%9c%ac%e7%bb%93%e6%9e%9c" aria-label="1.3.2 整合多样本结果">1.3.2 整合多样本结果</a></li></ul>
                    </li></ul>
                    </li>
                    <li>
                        <a href="#2%e4%b8%8b%e6%b8%b8%e5%88%86%e6%9e%90" aria-label="2、下游分析">2、下游分析</a><ul>
                            
                    <li>
                        <a href="#21-%e5%88%9b%e5%bb%bamarvel%e5%af%b9%e8%b1%a1" aria-label="2.1 创建marvel对象">2.1 创建marvel对象</a></li>
                    <li>
                        <a href="#22-%e9%a2%84%e5%a4%84%e7%90%86%e6%ad%a5%e9%aa%a4" aria-label="2.2 预处理步骤">2.2 预处理步骤</a></li>
                    <li>
                        <a href="#23-%e7%bb%86%e8%83%9e%e7%be%a4as%e7%89%b9%e5%be%81%e5%88%86%e6%9e%90" aria-label="2.3 细胞群AS特征分析">2.3 细胞群AS特征分析</a></li>
                    <li>
                        <a href="#24-%e7%bb%86%e8%83%9e%e7%be%a4%e5%b7%ae%e5%bc%82%e5%88%86%e6%9e%90" aria-label="2.4 细胞群差异分析">2.4 细胞群差异分析</a></li>
                    <li>
                        <a href="#25-%e5%b7%ae%e5%bc%82as%e8%bf%9b%e9%98%b6%e5%88%86%e6%9e%90" aria-label="2.5 差异AS进阶分析">2.5 差异AS进阶分析</a></li></ul>
                    </li></ul>
                    </li>
                    <li>
                        <a href="#%e4%ba%8cdrop-baed-pipeline" aria-label="二、Drop-baed pipeline">二、Drop-baed pipeline</a><ul>
                            
                    <li>
                        <a href="#1%e4%b8%8a%e6%b8%b8%e5%88%86%e6%9e%90-1" aria-label="1、上游分析">1、上游分析</a><ul>
                            
                    <li>
                        <a href="#11-conda%e7%8e%af%e5%a2%83" aria-label="1.1 conda环境">1.1 conda环境</a></li>
                    <li>
                        <a href="#12-%e4%b8%8b%e8%bd%bd%e6%95%b0%e6%8d%ae" aria-label="1.2 下载数据">1.2 下载数据</a></li>
                    <li>
                        <a href="#13-%e6%af%94%e5%af%b9%e5%88%86%e6%9e%90" aria-label="1.3 比对分析">1.3 比对分析</a></li></ul>
                    </li>
                    <li>
                        <a href="#2%e4%b8%8b%e6%b8%b8%e5%88%86%e6%9e%90-1" aria-label="2、下游分析">2、下游分析</a><ul>
                            
                    <li>
                        <a href="#21-%e6%95%b4%e7%90%86%e6%af%94%e5%af%b9%e7%bb%93%e6%9e%9c" aria-label="2.1 整理比对结果">2.1 整理比对结果</a></li>
                    <li>
                        <a href="#22-%e5%88%9b%e5%bb%bamarvel%e5%af%b9%e8%b1%a1" aria-label="2.2 创建marvel对象">2.2 创建marvel对象</a></li>
                    <li>
                        <a href="#23-%e9%a2%84%e5%a4%84%e7%90%86%e6%ad%a5%e9%aa%a4" aria-label="2.3 预处理步骤">2.3 预处理步骤</a></li>
                    <li>
                        <a href="#24-sj%e5%b7%ae%e5%bc%82%e5%88%86%e6%9e%90" aria-label="2.4 SJ差异分析">2.4 SJ差异分析</a></li>
                    <li>
                        <a href="#25-%e7%89%b9%e5%ae%9a%e5%9f%ba%e5%9b%a0%e5%88%86%e6%9e%90" aria-label="2.5 特定基因分析">2.5 特定基因分析</a>
                    </li>
                </ul>
                </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>R包MARVEL是由牛津大学MRC Weatherall分子医学研究所团队开发的，用于分析单细胞水平的可变剪切事件。相关文章于2023年1月在Nucleic Acids Research期刊发表，在其github页面分享了MARVEL工具的分析流程，在此学习、记录如下。</p>
<ul>
<li>Pape：https://doi.org/10.1093/nar/gkac1260</li>
<li>Github：https://github.com/wenweixiong/MARVEL</li>
<li>Plate pipeline：https://wenweixiong.github.io/MARVEL_Plate.html</li>
<li>Droplet pipeline：https://wenweixiong.github.io/MARVEL_Droplet.html</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230226161252918.png" alt="image-20230226161252918" style="zoom:67%;" /></blockquote>
<p><strong>可变剪切</strong></p>
<ul>
<li>
<p>关于可变剪切(Alternative splicing，AS)的基础，可以参看一篇<a href="https://www.jieandze1314.com/post/cnposts/169/">中文教程</a>；其中关于基因结构组成，可以参考之前的<a href="https://lishensuo.github.io/posts/bioinfo/053%E4%BA%BA%E7%B1%BB%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%9F%BA%E7%A1%80%E7%9F%A5%E8%AF%86%E4%B8%8E%E4%B8%8B%E8%BD%BD%E6%9F%A5%E8%AF%A2/#1%e7%bc%96%e7%a0%81%e5%8c%ba%e4%b8%8e%e9%9d%9e%e7%bc%96%e7%a0%81%e5%8c%ba">笔记</a>。</p>
</li>
<li>
<p>过去，针对Bulk RNA-seq数据的可变剪切分析工具已经有很多，例如之前学习的IsoformSwitchAnalyzeR包。MARVEL是个人目前了解到的第一个针对单细胞转录数据AS分析工具。</p>
</li>
</ul>
<p><strong>单细胞数据类型</strong></p>
<ul>
<li>AS分析的核心是通过检测比对到splice junction位点的read（例如，跨越两个外显子）进行分析。</li>
<li>目前主流的单细胞测序技术主要分为<strong>Droplet-based 10X</strong>方法与<strong>Plate-based Smart-seq</strong>方法。</li>
<li>10X技术由于仅对cDNA的3&rsquo;端部分进行测序，只能对该区域涉及的splice junction进行分析，且无法具体分析事件类型</li>
<li>Smart-seq技术虽然通量低，但是全长转录本的单细胞测序原理，可以进行更为深入的AS分析。</li>
</ul>
<h1 id="一plate-based-pipeline">一、Plate-based pipeline<a hidden class="anchor" aria-hidden="true" href="#一plate-based-pipeline">#</a></h1>
<h2 id="1上游分析">1、上游分析<a hidden class="anchor" aria-hidden="true" href="#1上游分析">#</a></h2>
<h3 id="11-工作环境">1.1 工作环境<a hidden class="anchor" aria-hidden="true" href="#11-工作环境">#</a></h3>
<p>两个同级的目录（1）basic：参考数据以及代码脚本 （2）proj1：具体的分析项目文件</p>
<ul>
<li><strong>后续分析均在 <code>proj1/ </code>路径下操作</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>tree basic/ -L <span style="color:#ff0;font-weight:bold">2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># basic/</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── gtf</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── gencode.v31.annotation.gtf</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── GRCh38.primary_assembly.genome.fa</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── Log.out</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   ├── rsem_index</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># │   └── star_index</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── rscript_Anno_SJ_rMATs.R</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── rscript_bedtools_input.R</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># └── single_sample_upstream.sh</span>
</span></span><span style="display:flex;"><span>tree proj1/ -L <span style="color:#ff0;font-weight:bold">2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># proj1/</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── SJ_phenoData.txt</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># └── work</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     ├── fq</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     ├── intron</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     ├── merge</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     ├── rmats</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     ├── rsem</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     └── star</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="12-conda环境">1.2 conda环境<a hidden class="anchor" aria-hidden="true" href="#12-conda环境">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f"># 名为marvel_plate的分析环境</span>
</span></span><span style="display:flex;"><span>conda create -n marvel_plate python=3.7
</span></span><span style="display:flex;"><span>conda activate marvel_plate
</span></span><span style="display:flex;"><span>conda install -c hcc aspera-cli
</span></span><span style="display:flex;"><span>conda install -c bioconda star=2.7.1a rmats=4.1.0 -y 
</span></span><span style="display:flex;"><span>conda install -c bioconda bedtools samtools rsem -y 
</span></span><span style="display:flex;"><span>conda install -c r r-base r-essentials r-tidyverse -y
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="12-准备数据">1.2 准备数据<a hidden class="anchor" aria-hidden="true" href="#12-准备数据">#</a></h3>
<p><strong>（1）测序数据</strong></p>
<ul>
<li>官方手册分析了如下数据集的 192个单细胞数据，以其中一个演示流程</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f"># https://www.ebi.ac.uk/ena/browser/view/ERR1562083</span>
</span></span><span style="display:flex;"><span>id=ERR1562083
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># id=$1</span>
</span></span><span style="display:flex;"><span>mkdir -p ./work/fq/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>
</span></span><span style="display:flex;"><span>ascp -QT -l 300m -P33001  <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-i ~/miniconda3/envs/marvel_plate/etc/asperaweb_id_dsa.openssh   <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/<span style="color:#0ff;font-weight:bold">${</span>id:0:6<span style="color:#0ff;font-weight:bold">}</span>/00<span style="color:#0ff;font-weight:bold">${</span>id:0-1<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_1.fastq.gz  ./work/fq/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>ascp -QT -l 300m -P33001  <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-i ~/miniconda3/envs/marvel_plate/etc/asperaweb_id_dsa.openssh   <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/<span style="color:#0ff;font-weight:bold">${</span>id:0:6<span style="color:#0ff;font-weight:bold">}</span>/00<span style="color:#0ff;font-weight:bold">${</span>id:0-1<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_2.fastq.gz  ./work/fq/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（2）基因组数据</strong></p>
<blockquote>
<ul>
<li>在 ./basic/gtf 目录下</li>
<li><a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/">https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/</a></li>
</ul></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f">## STAR索引</span>
</span></span><span style="display:flex;"><span>STAR  --runMode genomeGenerate <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--genomeDir star_index <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--genomeFastaFiles GRCh38.primary_assembly.genome.fa <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--sjdbGTFfile gencode.v31.annotation.gtf <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--runThreadN <span style="color:#ff0;font-weight:bold">40</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## RSEM索引</span>
</span></span><span style="display:flex;"><span>rsem-prepare-reference <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--gtf gencode.v31.annotation.gtf <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--star  -p <span style="color:#ff0;font-weight:bold">20</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>GRCh38.primary_assembly.genome.fa <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>rsem_index/hg38
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（3）教程数据</strong></p>
<ul>
<li>官方教程提供了各个分析步骤的示范文件，可供之后分析作为参考。</li>
<li><a href="http://datashare.molbiol.ox.ac.uk/public/wwen/MARVEL/Tutorial/Plate/Data.zip">http://datashare.molbiol.ox.ac.uk/public/wwen/MARVEL/Tutorial/Plate/Data.zip</a></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>ls Data
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GTF  MARVEL  rMATS  R Object  RSEM  SJ</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="13-分析步骤">1.3 分析步骤<a hidden class="anchor" aria-hidden="true" href="#13-分析步骤">#</a></h3>
<h4 id="131-单样本4步分析">1.3.1 单样本4步分析<a hidden class="anchor" aria-hidden="true" href="#131-单样本4步分析">#</a></h4>
<p><strong>（1）单样本STAR比对</strong></p>
<blockquote>
<p>将每个样本的Fastq文件使用STAR软件进行比对，产生每个splice junction count matrix</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f"># Reference</span>
</span></span><span style="display:flex;"><span>less -SN ./Data/SJ/SJ.txt
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230305130339185.png" alt="image-20230305130339185" style="zoom: 50%;" /></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>mkdir -p ./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>
</span></span><span style="display:flex;"><span>ref_idx_star=../basic/gtf/star_index
</span></span><span style="display:flex;"><span>ref_gtf=../basic/gtf/gencode.v31.annotation.gtf
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span> is runing......
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#1st pass mode</span>
</span></span><span style="display:flex;"><span>STAR --runThreadN <span style="color:#ff0;font-weight:bold">16</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--genomeDir <span style="color:#0ff;font-weight:bold">${</span>ref_idx_star<span style="color:#0ff;font-weight:bold">}</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--readFilesCommand zcat <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--readFilesIn ./work/fq/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_1.fastq.gz ./work/fq/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_2.fastq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--outFileNamePrefix ./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>. <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--outSAMtype None
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#2nd pass mode</span>
</span></span><span style="display:flex;"><span>STAR --runThreadN <span style="color:#ff0;font-weight:bold">16</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--genomeDir <span style="color:#0ff;font-weight:bold">${</span>ref_idx_star<span style="color:#0ff;font-weight:bold">}</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--readFilesCommand zcat <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--readFilesIn ./work/fq/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_1.fastq.gz ./work/fq/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_2.fastq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--outFileNamePrefix ./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>. <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--sjdbFileChrStartEnd ./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/*SJ.out.tab <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--sjdbGTFfile <span style="color:#0ff;font-weight:bold">${</span>ref_gtf<span style="color:#0ff;font-weight:bold">}</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--outSAMtype BAM SortedByCoordinate <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--outSAMattributes NH HI AS nM XS <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--quantMode TranscriptomeSAM
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（2）单样本rMATs注释</strong></p>
<blockquote>
<p>根据上述比对BAM结果，注释出每个基因(转录本)涉及的可变剪切事件类型。使用rMATs软件，可注释出5种类型：SE、RI、MXE、A5SS、A3SS。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>find ./Data/rMATS/ -name *featureData.txt
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ./Data/rMATS/SE/SE_featureData.txt</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ./Data/rMATS/A3SS/A3SS_featureData.txt</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ./Data/rMATS/MXE/MXE_featureData.txt</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ./Data/rMATS/RI/RI_featureData.txt</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ./Data/rMATS/A5SS/A5SS_featureData.txt</span>
</span></span><span style="display:flex;"><span>less -SN ./Data/rMATS/SE/SE_featureData.txt
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230305131524024.png" alt="image-20230305131524024"  />
</p></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>mkdir -p ./work/rmats/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span> is running......
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;./work/star/</span><span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span><span style="color:#0ff;font-weight:bold">/</span><span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span><span style="color:#0ff;font-weight:bold">.Aligned.sortedByCoord.out.bam&#34;</span> &gt; ./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/BAM_fls.txt
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>rmats.py <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--b1 ./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>//BAM_fls.txt <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--gtf <span style="color:#0ff;font-weight:bold">${</span>ref_gtf<span style="color:#0ff;font-weight:bold">}</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--od ./work/rmats/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--tmp ./work/rmats/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/tmp <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-t paired <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--readLength <span style="color:#ff0;font-weight:bold">125</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--variable-read-length <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--nthread <span style="color:#ff0;font-weight:bold">16</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--statoff
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># readLength参考如下</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># zcat ./work/fq/${id}_1.fastq.gz | awk &#39;{if(NR%4==2) print NR&#34;\t&#34;$0&#34;\t&#34;length($0)}&#39; | less</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（3）单样本Intron矩阵</strong></p>
<blockquote>
<p>统计Reads对于每个内含子的比对情况，逐碱基累加作为比对结果。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f"># reference</span>
</span></span><span style="display:flex;"><span>less -SN Data/MARVEL/PSI/RI/Counts_by_Region.txt
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230305132411262.png" alt="image-20230305132411262" style="zoom:67%;" /></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>mkdir -p ./work/intron/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span> is running......
</span></span><span style="display:flex;"><span>samtools view -H ./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>.Aligned.sortedByCoord.out.bam | <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>  grep SQ | cut -f <span style="color:#ff0;font-weight:bold">2</span> | awk <span style="color:#0ff;font-weight:bold">&#39;{ sub(/^SN:/, &#34;&#34;); print;}&#39;</span> &gt; ./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/sorted_chr_in_bam.txt
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 预定义R脚本，生成bedtools计算所需的两个文件：染色体大小以及内含子坐标</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># https://github.com/lishensuo/utils/blob/main/marvel/rscript_bedtools_input.R</span>
</span></span><span style="display:flex;"><span>Rscript ../basic/rscript_bedtools_input.R <span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>bedtools coverage <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-g ./work/intron/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/hg38.chrom.sizes.txt <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-split <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-sorted <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-a ./work/intron/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/RI_Coordinates_sorted.bed <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-b ./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>.Aligned.sortedByCoord.out.bam &gt; <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>./work/intron/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/intron_count.txt <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-d
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（4）单样本gene矩阵</strong></p>
<blockquote>
<p>最后需要对基因表达水平进行定量。教程推荐使用RSEM软件，并进行TPM类似的标准化处理</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f"># reference</span>
</span></span><span style="display:flex;"><span>less -SN ./Data/RSEM/TPM.txt
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230305133226554.png" alt="image-20230305133226554" style="zoom:67%;" /></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>mkdir -p ./work/rsem/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>
</span></span><span style="display:flex;"><span>rsem-calculate-expression <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--bam <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--paired-end <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-p <span style="color:#ff0;font-weight:bold">16</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>.Aligned.toTranscriptome.out.bam <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>../basic/gtf/rsem_index/hg38 <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>./work/rsem/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>（1）如上，由于4步的分析过程中产生的中间数据可能超过数百兆；且SMART-seq往往涉及成百上千的细胞。为节约空间，可在每个样本分析完成后，删除不必要的数据。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>rm ./work/fq/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>*
</span></span><span style="display:flex;"><span>rm  ./work/star/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/*bam
</span></span><span style="display:flex;"><span>rm -rf ./work/rmats/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/tmp/*
</span></span><span style="display:flex;"><span>rm ./work/rsem/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>.transcript.bam
</span></span><span style="display:flex;"><span>rm ./work/rsem/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>.isoforms.results
</span></span></code></pre></td></tr></table>
</div>
</div><p>（2）多样本批量后台运行方式</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>cat ../basic/single_sample_upstream.sh
</span></span><span style="display:flex;"><span>chmod u+x ../basic/single_sample_upstream.sh
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 单样本测试</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ../basic/single_sample_upstream.sh ERR1562273</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># parallel多线程批量运行</span>
</span></span><span style="display:flex;"><span>nohup parallel -j <span style="color:#ff0;font-weight:bold">5</span> ../basic/single_sample_upstream.sh {} ::: <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>  <span style="color:#fff;font-weight:bold">$(</span>grep -v <span style="color:#0ff;font-weight:bold">&#34;sample&#34;</span> ./SJ_phenoData.txt | awk  <span style="color:#0ff;font-weight:bold">&#39;{print $1}&#39;</span><span style="color:#fff;font-weight:bold">)</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>  1&gt; ./work/single_sample_upstream.log 2&gt;&amp;<span style="color:#ff0;font-weight:bold">1</span> &amp;
</span></span></code></pre></td></tr></table>
</div>
</div></blockquote>
<h4 id="132-整合多样本结果">1.3.2 整合多样本结果<a hidden class="anchor" aria-hidden="true" href="#132-整合多样本结果">#</a></h4>
<p>即分步整合上游4个步骤中的每个步骤的所有样本结果；均使用R语言完成</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>library(parallel)
</span></span><span style="display:flex;"><span>sample_meta = data.table::fread(<span style="color:#0ff;font-weight:bold">&#34;./SJ_phenoData.txt&#34;</span>,data.table=<span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>samples=sample_meta$sample.id
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>step1: 汇总剪切位点矩阵</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sj.list = mclapply(samples, <span style="color:#fff;font-weight:bold">function</span>(sample){   
</span></span><span style="display:flex;"><span>     sj_sp = data.table::fread(paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/star/&#34;</span>,sample,<span style="color:#0ff;font-weight:bold">&#34;/&#34;</span>,sample,<span style="color:#0ff;font-weight:bold">&#34;.SJ.out.tab&#34;</span>), data.table=<span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>     <span style="color:#fff;font-weight:bold">if</span>(nrow(sj_sp)==<span style="color:#ff0;font-weight:bold">0</span>) <span style="color:#fff;font-weight:bold">return</span>(<span style="color:#fff;font-weight:bold">NULL</span>)
</span></span><span style="display:flex;"><span>     sj_sp = sj_sp %&gt;% 
</span></span><span style="display:flex;"><span>       dplyr::mutate(`coord.intron`=paste0(V1,<span style="color:#0ff;font-weight:bold">&#34;:&#34;</span>,V2,<span style="color:#0ff;font-weight:bold">&#34;:&#34;</span>,V3)) %&gt;% 
</span></span><span style="display:flex;"><span>       dplyr::select(`coord.intron`, V7)
</span></span><span style="display:flex;"><span>     colnames(sj_sp)[2] = sample
</span></span><span style="display:flex;"><span>     <span style="color:#fff;font-weight:bold">return</span>(sj_sp)
</span></span><span style="display:flex;"><span>},mc.cores=<span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span>names(sj.list) = samples
</span></span><span style="display:flex;"><span>sj.list = sj.list[!unlist(lapply(sj.list, is.null))]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sj_res = sj.list[[1]]
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span>(x <span style="color:#fff;font-weight:bold">in</span> sj.list[<span style="color:#ff0;font-weight:bold">-1</span>]){
</span></span><span style="display:flex;"><span>     sj_res = dplyr::full_join(sj_res, x)
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span>(sample <span style="color:#fff;font-weight:bold">in</span> setdiff(samples,names(sj.list))){
</span></span><span style="display:flex;"><span>     sj_res[,sample]=<span style="color:#fff;font-weight:bold">NA</span>
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>dim(sj_res)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>write.table(sj_res, file=paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/merge/SJ.txt&#34;</span>),
</span></span><span style="display:flex;"><span>     quote=<span style="color:#fff;font-weight:bold">F</span>, row.names=<span style="color:#fff;font-weight:bold">F</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>step2:整理AS注释结果
<ul>
<li>这一步最为复杂，需要理解每一种剪切事件的定义以及正负链的差异</li>
<li>参考官方示意：https://wenweixiong.github.io/Splicing_Nomenclature</li>
</ul>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># https://github.com/lishensuo/utils/blob/main/marvel/rscript_Anno_SJ_rMATs.R</span>
</span></span><span style="display:flex;"><span>source(paste0(<span style="color:#0ff;font-weight:bold">&#34;../basic/rscript_Anno_SJ_rMATs.R&#34;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>AS_types = c(<span style="color:#0ff;font-weight:bold">&#34;SE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;RI&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;MXE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;A3SS&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;A5SS&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span>(AS_type <span style="color:#fff;font-weight:bold">in</span> AS_types){
</span></span><span style="display:flex;"><span>     <span style="color:#007f7f"># AS_type = AS_types[1]</span>
</span></span><span style="display:flex;"><span>     print(AS_type)
</span></span><span style="display:flex;"><span>     fls_AS = list.files(paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/rmats&#34;</span>), pattern=paste0(<span style="color:#0ff;font-weight:bold">&#34;fromGTF.&#34;</span>,AS_type,<span style="color:#0ff;font-weight:bold">&#34;.txt&#34;</span>),
</span></span><span style="display:flex;"><span>          recursive=<span style="color:#fff;font-weight:bold">T</span>, full.name=<span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>     <span style="color:#fff;font-weight:bold">if</span> (AS_type==<span style="color:#0ff;font-weight:bold">&#34;SE&#34;</span>){
</span></span><span style="display:flex;"><span>          func = se_func
</span></span><span style="display:flex;"><span>     } <span style="color:#fff;font-weight:bold">else</span> <span style="color:#fff;font-weight:bold">if</span> (AS_type==<span style="color:#0ff;font-weight:bold">&#34;RI&#34;</span>){
</span></span><span style="display:flex;"><span>          func = ri_func 
</span></span><span style="display:flex;"><span>     } <span style="color:#fff;font-weight:bold">else</span> <span style="color:#fff;font-weight:bold">if</span> (AS_type==<span style="color:#0ff;font-weight:bold">&#34;MXE&#34;</span>){
</span></span><span style="display:flex;"><span>          func = mxe_func 
</span></span><span style="display:flex;"><span>     } <span style="color:#fff;font-weight:bold">else</span> <span style="color:#fff;font-weight:bold">if</span> (AS_type==<span style="color:#0ff;font-weight:bold">&#34;A3SS&#34;</span>){
</span></span><span style="display:flex;"><span>          func = a3ss_func 
</span></span><span style="display:flex;"><span>     } <span style="color:#fff;font-weight:bold">else</span> <span style="color:#fff;font-weight:bold">if</span> (AS_type==<span style="color:#0ff;font-weight:bold">&#34;A5SS&#34;</span>){
</span></span><span style="display:flex;"><span>          func = a5ss_func
</span></span><span style="display:flex;"><span>     }
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>     df_AS = mclapply(fls_AS, <span style="color:#fff;font-weight:bold">function</span>(fls){
</span></span><span style="display:flex;"><span>          rmats_AS = data.table::fread(fls, data.table=<span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>          rmats_AS = rmats_AS %&gt;%
</span></span><span style="display:flex;"><span>               dplyr::mutate(tran_id = func(.)) %&gt;% 
</span></span><span style="display:flex;"><span>               dplyr::rename(gene_id=GeneID,gene_short_name=geneSymbol) %&gt;% 
</span></span><span style="display:flex;"><span>               dplyr::select(tran_id, gene_id, gene_short_name)
</span></span><span style="display:flex;"><span>          <span style="color:#fff;font-weight:bold">return</span>(rmats_AS)
</span></span><span style="display:flex;"><span>     },mc.cores=<span style="color:#ff0;font-weight:bold">10</span>) %&gt;% do.call(rbind, .) %&gt;% 
</span></span><span style="display:flex;"><span>          dplyr::distinct()
</span></span><span style="display:flex;"><span>     write.table(df_AS, row.names=<span style="color:#fff;font-weight:bold">F</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>,quote=<span style="color:#fff;font-weight:bold">F</span>,
</span></span><span style="display:flex;"><span>          file=paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/merge/as_&#34;</span>,AS_type,<span style="color:#0ff;font-weight:bold">&#34;_featureData.txt&#34;</span>))
</span></span><span style="display:flex;"><span>}
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>step3：汇总intron矩阵</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>int_mt.list = mclapply(samples, <span style="color:#fff;font-weight:bold">function</span>(sample){
</span></span><span style="display:flex;"><span>     <span style="color:#007f7f"># sample = samples[1]</span>
</span></span><span style="display:flex;"><span>     int_mt = data.table::fread(paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/intron/&#34;</span>,sample,<span style="color:#0ff;font-weight:bold">&#34;/intron_count.txt&#34;</span>)) 
</span></span><span style="display:flex;"><span>     int_mt = int_mt %&gt;% 
</span></span><span style="display:flex;"><span>          dplyr::mutate(`coord.intron`=paste(V1,V2,V3, sep=<span style="color:#0ff;font-weight:bold">&#34;:&#34;</span>)) %&gt;% 
</span></span><span style="display:flex;"><span>          dplyr::group_by(coord.intron) %&gt;% 
</span></span><span style="display:flex;"><span>          dplyr::summarize(count=sum(V5)) %&gt;% 
</span></span><span style="display:flex;"><span>          dplyr::select(coord.intron, count) %&gt;% as.data.frame()
</span></span><span style="display:flex;"><span>     colnames(int_mt)[2] = sample
</span></span><span style="display:flex;"><span>     <span style="color:#fff;font-weight:bold">return</span>(int_mt)
</span></span><span style="display:flex;"><span>},mc.cores=<span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>intron_res = int_mt.list[[1]]
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span>(x <span style="color:#fff;font-weight:bold">in</span> int_mt.list[<span style="color:#ff0;font-weight:bold">-1</span>]){
</span></span><span style="display:flex;"><span>     intron_res = dplyr::full_join(intron_res, x)
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>dim(intron_res)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>write.table(intron_res, file=paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/merge/intron_count_by_region.txt&#34;</span>),
</span></span><span style="display:flex;"><span>     quote=<span style="color:#fff;font-weight:bold">F</span>, row.names=<span style="color:#fff;font-weight:bold">F</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>汇总gene矩阵</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>gene_mt.list = mclapply(samples, <span style="color:#fff;font-weight:bold">function</span>(sample){
</span></span><span style="display:flex;"><span>     fls=paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/rsem/&#34;</span>,sample,<span style="color:#0ff;font-weight:bold">&#34;/&#34;</span>,sample,<span style="color:#0ff;font-weight:bold">&#34;.genes.results&#34;</span>)
</span></span><span style="display:flex;"><span>     gene_mt = data.table::fread(fls, data.table=<span style="color:#fff;font-weight:bold">F</span>) %&gt;% 
</span></span><span style="display:flex;"><span>          dplyr::select(gene_id, TPM)
</span></span><span style="display:flex;"><span>     colnames(gene_mt)[2] = sample
</span></span><span style="display:flex;"><span>     <span style="color:#fff;font-weight:bold">return</span>(gene_mt)
</span></span><span style="display:flex;"><span>},mc.cores=<span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span>gene_res = gene_mt.list[[1]]
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span>(x <span style="color:#fff;font-weight:bold">in</span> gene_mt.list[<span style="color:#ff0;font-weight:bold">-1</span>]){
</span></span><span style="display:flex;"><span>     gene_res = dplyr::full_join(gene_res, x)
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>dim(gene_res)
</span></span><span style="display:flex;"><span>write.table(gene_res, file=paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/merge/rsem_tpm.txt&#34;</span>),,
</span></span><span style="display:flex;"><span>     quote=<span style="color:#fff;font-weight:bold">F</span>, row.names=<span style="color:#fff;font-weight:bold">F</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="2下游分析">2、下游分析<a hidden class="anchor" aria-hidden="true" href="#2下游分析">#</a></h2>
<p>后续根据教程提供的数据学习，需要安装的主要的MARVEL包以及其它辅助分析包，具体可参考教程。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 建议再单独建立一个conda R环境</span>
</span></span><span style="display:flex;"><span>install.packages(<span style="color:#0ff;font-weight:bold">&#34;MARVEL&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="21-创建marvel对象">2.1 创建marvel对象<a hidden class="anchor" aria-hidden="true" href="#21-创建marvel对象">#</a></h3>
<blockquote>
<p>需要准备七类文件，读入为R语言对象</p>
<ul>
<li>样本表型data.frame：第一列名为sample.id</li>
<li>剪切位点计数矩阵：第一列名为coord.intron</li>
<li>剪切事件类型注释列表：由数据框包装而出的list，每个数据库至少包含tran_id与gene_id</li>
<li>内含子计数矩阵：第一列名为coord.intron</li>
<li>基因表达矩阵：经标准化处理(RPKM/FPKM/TPM)，暂时不要log转换</li>
<li>基因类型data.frame：至少包含3列，gene_id, gene_short_name,  gene_type</li>
<li>上游分析所用到的GTF数据</li>
</ul></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">43
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">44
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">45
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">46
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">47
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">48
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">49
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">50
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">51
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">52
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">53
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">54
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">55
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">56
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">57
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>library(MARVEL)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (1) 细胞表型分组</span>
</span></span><span style="display:flex;"><span>df.pheno = data.table::fread(<span style="color:#0ff;font-weight:bold">&#34;./SJ_phenoData.txt&#34;</span>,data.table=<span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2) Gene文件</span>
</span></span><span style="display:flex;"><span>df.tpm &lt;- read.table(<span style="color:#0ff;font-weight:bold">&#34;./work/merge/rsem_tpm.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>dim(df.tpm)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (3) GTF文件</span>
</span></span><span style="display:flex;"><span>gtf &lt;- data.table::fread(<span style="color:#0ff;font-weight:bold">&#34;../basic/gtf/gencode.v31.annotation.gtf&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">FALSE</span>, na.strings=<span style="color:#0ff;font-weight:bold">&#34;NA&#34;</span>,quote=<span style="color:#0ff;font-weight:bold">&#34;\&#34;&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>     as.data.frame()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (4) Gene类型</span>
</span></span><span style="display:flex;"><span>gene.feature = subset(gtf, V3==<span style="color:#0ff;font-weight:bold">&#34;gene&#34;</span>) 
</span></span><span style="display:flex;"><span>df.tpm.feature &lt;- str_split(gene.feature$V9, <span style="color:#0ff;font-weight:bold">&#34;;&#34;</span>, simplify=<span style="color:#fff;font-weight:bold">T</span>) %&gt;% 
</span></span><span style="display:flex;"><span>     as.data.frame() %&gt;%
</span></span><span style="display:flex;"><span>     dplyr::mutate(gene_id=str_match(V1, <span style="color:#0ff;font-weight:bold">&#39;&#34;(.*)&#34;&#39;</span>)[,<span style="color:#ff0;font-weight:bold">2</span>]) %&gt;% 
</span></span><span style="display:flex;"><span>     dplyr::mutate(gene_short_name=str_match(V3, <span style="color:#0ff;font-weight:bold">&#39;&#34;(.*)&#34;&#39;</span>)[,<span style="color:#ff0;font-weight:bold">2</span>]) %&gt;%      
</span></span><span style="display:flex;"><span>     dplyr::mutate(gene_type=str_match(V2, <span style="color:#0ff;font-weight:bold">&#39;&#34;(.*)&#34;&#39;</span>)[,<span style="color:#ff0;font-weight:bold">2</span>]) %&gt;% 
</span></span><span style="display:flex;"><span>     dplyr::select(gene_id, gene_short_name, gene_type) %&gt;% 
</span></span><span style="display:flex;"><span>     dplyr::filter(gene_id %in% df.tpm$gene_id)
</span></span><span style="display:flex;"><span>df.tpm.feature = df.tpm.feature[match(df.tpm$gene_id, df.tpm.feature$gene_id),]
</span></span><span style="display:flex;"><span>rownames(df.tpm.feature) = seq(nrow(df.tpm.feature))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (5) SJ文件</span>
</span></span><span style="display:flex;"><span>sj = data.table::fread(<span style="color:#0ff;font-weight:bold">&#34;./work/merge/SJ.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>, na.strings=<span style="color:#0ff;font-weight:bold">&#34;NA&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>     as.data.frame()
</span></span><span style="display:flex;"><span>dim(sj)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (6) AS文件</span>
</span></span><span style="display:flex;"><span>df.feature.list = lapply(c(<span style="color:#0ff;font-weight:bold">&#34;SE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;MXE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;RI&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;A5SS&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;A3SS&#34;</span>), <span style="color:#fff;font-weight:bold">function</span>(x){
</span></span><span style="display:flex;"><span>     <span style="color:#007f7f"># x = &#34;SE&#34;</span>
</span></span><span style="display:flex;"><span>     df.feature = read.table(paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/merge/as_&#34;</span>,x,<span style="color:#0ff;font-weight:bold">&#34;_featureData.txt&#34;</span>), 
</span></span><span style="display:flex;"><span>          sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>, na.strings=<span style="color:#0ff;font-weight:bold">&#34;NA&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>          dplyr::distinct(tran_id, .keep_all=<span style="color:#fff;font-weight:bold">TRUE</span>) %&gt;% 
</span></span><span style="display:flex;"><span>          dplyr::left_join(df.tpm.feature)
</span></span><span style="display:flex;"><span>     <span style="color:#fff;font-weight:bold">return</span>(df.feature)
</span></span><span style="display:flex;"><span>})
</span></span><span style="display:flex;"><span>names(df.feature.list) &lt;- c(<span style="color:#0ff;font-weight:bold">&#34;SE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;MXE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;RI&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;A5SS&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;A3SS&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (7) Intron文件</span>
</span></span><span style="display:flex;"><span>df.intron.counts &lt;- data.table::fread(<span style="color:#0ff;font-weight:bold">&#34;./work/merge/intron_count_by_region.txt&#34;</span>, 
</span></span><span style="display:flex;"><span>     sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>, na.strings=<span style="color:#0ff;font-weight:bold">&#34;NA&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>     as.data.frame()
</span></span><span style="display:flex;"><span>dim(df.intron.counts)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## MARVEL object</span>
</span></span><span style="display:flex;"><span>marvel &lt;- CreateMarvelObject(SpliceJunction=sj,
</span></span><span style="display:flex;"><span>                             SplicePheno=df.pheno,
</span></span><span style="display:flex;"><span>                             SpliceFeature=df.feature.list,
</span></span><span style="display:flex;"><span>                             IntronCounts=df.intron.counts,
</span></span><span style="display:flex;"><span>                             GeneFeature=df.tpm.feature,
</span></span><span style="display:flex;"><span>                             Exp=df.tpm,
</span></span><span style="display:flex;"><span>                             GTF=gtf
</span></span><span style="display:flex;"><span>                             )
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="22-预处理步骤">2.2 预处理步骤<a hidden class="anchor" aria-hidden="true" href="#22-预处理步骤">#</a></h3>
<p><strong>（1）基因表达log转换</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- TransformExpValues(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                             offset=<span style="color:#ff0;font-weight:bold">1</span>,
</span></span><span style="display:flex;"><span>                             transformation=<span style="color:#0ff;font-weight:bold">&#34;log2&#34;</span>,
</span></span><span style="display:flex;"><span>                             threshold.lower=<span style="color:#ff0;font-weight:bold">1</span>
</span></span><span style="display:flex;"><span>                            )
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（2）检测额外AS事件</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 如果基因表达水平整体较低，则不考虑该基因的可变剪切事件</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># min.expr:判定单个细胞是否表达基因的阈值</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># min.cells: 基于上述条件，判定细胞群表达该基因的阈值</span>
</span></span><span style="display:flex;"><span>marvel &lt;- DetectEvents(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                       min.cells=<span style="color:#ff0;font-weight:bold">50</span>, <span style="color:#007f7f">#细胞群的表达百分比</span>
</span></span><span style="display:flex;"><span>                       min.expr=<span style="color:#ff0;font-weight:bold">1</span>,   <span style="color:#007f7f">#认为细胞表达该基因的阈值</span>
</span></span><span style="display:flex;"><span>                       track.progress=<span style="color:#fff;font-weight:bold">FALSE</span>,
</span></span><span style="display:flex;"><span>                       EventType=<span style="color:#0ff;font-weight:bold">&#34;AFE&#34;</span>
</span></span><span style="display:flex;"><span>                       )
</span></span><span style="display:flex;"><span>marvel &lt;- DetectEvents(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                       min.cells=<span style="color:#ff0;font-weight:bold">50</span>,
</span></span><span style="display:flex;"><span>                       min.expr=<span style="color:#ff0;font-weight:bold">1</span>,
</span></span><span style="display:flex;"><span>                       track.progress=<span style="color:#fff;font-weight:bold">FALSE</span>,
</span></span><span style="display:flex;"><span>                       EventType=<span style="color:#0ff;font-weight:bold">&#34;ALE&#34;</span>
</span></span><span style="display:flex;"><span>                       )
</span></span><span style="display:flex;"><span>length(marvel$SpliceFeature)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 7</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（3）计算PSI值</strong></p>
<p>计算一个基因对于特定剪切事件的PSI值，这是后续分析的基础</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># CoverageThreshold: 支持该剪切事件的最小reads数，小于该阈值标记为NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># UnevenCoverageMultiplier: 针对SE与MXE</span>
</span></span><span style="display:flex;"><span>marvel &lt;- ComputePSI(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                     CoverageThreshold=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>                     UnevenCoverageMultiplier=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>                     EventType=<span style="color:#0ff;font-weight:bold">&#34;SE&#34;</span>
</span></span><span style="display:flex;"><span>                     )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel &lt;- ComputePSI(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                     CoverageThreshold=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>                     UnevenCoverageMultiplier=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>                     EventType=<span style="color:#0ff;font-weight:bold">&#34;MXE&#34;</span>
</span></span><span style="display:flex;"><span>                     )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel &lt;- ComputePSI(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                     CoverageThreshold=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>                     EventType=<span style="color:#0ff;font-weight:bold">&#34;RI&#34;</span>,
</span></span><span style="display:flex;"><span>                     thread=<span style="color:#ff0;font-weight:bold">4</span>  <span style="color:#007f7f"># only support RI</span>
</span></span><span style="display:flex;"><span>                     )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel &lt;- ComputePSI(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                     CoverageThreshold=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>                     EventType=<span style="color:#0ff;font-weight:bold">&#34;A5SS&#34;</span>
</span></span><span style="display:flex;"><span>                     )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel &lt;- ComputePSI(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                     CoverageThreshold=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>                     EventType=<span style="color:#0ff;font-weight:bold">&#34;A3SS&#34;</span>
</span></span><span style="display:flex;"><span>                     )
</span></span><span style="display:flex;"><span> 
</span></span><span style="display:flex;"><span>marvel &lt;- ComputePSI(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                     CoverageThreshold=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>                     EventType=<span style="color:#0ff;font-weight:bold">&#34;AFE&#34;</span>
</span></span><span style="display:flex;"><span>                     )
</span></span><span style="display:flex;"><span>    
</span></span><span style="display:flex;"><span>marvel &lt;- ComputePSI(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                     CoverageThreshold=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>                     EventType=<span style="color:#0ff;font-weight:bold">&#34;ALE&#34;</span>
</span></span><span style="display:flex;"><span>                     )
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（4）保存中间结果</strong></p>
<ul>
<li>首先检查不同矩阵与对应的meta行列名信息一致</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- CheckAlignment(MarvelObject=marvel, level=<span style="color:#0ff;font-weight:bold">&#34;SJ&#34;</span>)
</span></span><span style="display:flex;"><span>marvel &lt;- CheckAlignment(MarvelObject=marvel, level=<span style="color:#0ff;font-weight:bold">&#34;splicing&#34;</span>)
</span></span><span style="display:flex;"><span>marvel &lt;- CheckAlignment(MarvelObject=marvel, level=<span style="color:#0ff;font-weight:bold">&#34;gene&#34;</span>)
</span></span><span style="display:flex;"><span>marvel &lt;- CheckAlignment(MarvelObject=marvel, level=<span style="color:#0ff;font-weight:bold">&#34;splicing and gene&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>可进一步选取感兴趣的细胞群</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>index.1 &lt;- which(df.pheno$cell.type %in% c(<span style="color:#0ff;font-weight:bold">&#34;iPSC&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Endoderm&#34;</span>))
</span></span><span style="display:flex;"><span>index.2 &lt;- which(df.pheno$qc.seq==<span style="color:#0ff;font-weight:bold">&#34;pass&#34;</span>)
</span></span><span style="display:flex;"><span>index &lt;- intersect(index.1, index.2)
</span></span><span style="display:flex;"><span>sample.ids &lt;- df.pheno[index, <span style="color:#0ff;font-weight:bold">&#34;sample.id&#34;</span>]
</span></span><span style="display:flex;"><span>marvel &lt;- SubsetSamples(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                        sample.ids=sample.ids
</span></span><span style="display:flex;"><span>                        )
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>保存中间结果，方便后续分析</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>save(marvel, file=<span style="color:#0ff;font-weight:bold">&#34;./MARVEL2.RData&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 后续以官方提供的示例数据学习</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># load(&#34;../Data/MARVEL.RData&#34;)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="23-细胞群as特征分析">2.3 细胞群AS特征分析<a hidden class="anchor" aria-hidden="true" href="#23-细胞群as特征分析">#</a></h3>
<p>以其中84个iPSC细胞群进行演示</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sample.ids = marvel$SplicePheno  %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::filter(cell.type==<span style="color:#0ff;font-weight:bold">&#34;iPSC&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::pull(sample.id) 
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（1）AS类型占比</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- CountEvents(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                      sample.ids=sample.ids,
</span></span><span style="display:flex;"><span>                      min.cells=<span style="color:#ff0;font-weight:bold">25</span>
</span></span><span style="display:flex;"><span>                      )
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 对于一个基因至少有25个细胞的AS PSI值为非缺失值，则判定发生了可变剪切事件</span>
</span></span><span style="display:flex;"><span>marvel$N.Events$Table
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   event_type freq        pct</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1         SE 5270 40.1523810</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3         RI 2030 15.4666667</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6        AFE 1789 13.6304762</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5       A3SS 1746 13.3028571</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4       A5SS 1494 11.3828571</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 7        ALE  724  5.5161905</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2        MXE   72  0.5485714</span>
</span></span><span style="display:flex;"><span>marvel$N.Events$Plot 
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ggplot绘图体系，下同</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230305142515213.png" alt="image-20230305142515213" style="zoom: 50%;" />
<p><strong>（2）AS modality</strong></p>
<p>判断细胞群中每个基因的每种可变剪切的PSI分布类型，如下可分为7类</p>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230305142905888.png" alt="image-20230305142905888" style="zoom: 67%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- AssignModality(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                         sample.ids=sample.ids,
</span></span><span style="display:flex;"><span>                         min.cells=<span style="color:#ff0;font-weight:bold">25</span>,
</span></span><span style="display:flex;"><span>                         seed=<span style="color:#ff0;font-weight:bold">1</span>
</span></span><span style="display:flex;"><span>                         )
</span></span><span style="display:flex;"><span>marvel$Modality$Results[1:<span style="color:#ff0;font-weight:bold">5</span>, c(<span style="color:#0ff;font-weight:bold">&#34;tran_id&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;event_type&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;gene_id&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;gene_short_name&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;modality.bimodal.adj&#34;</span>)]
</span></span><span style="display:flex;"><span>table(marvel$Modality$Results$modality.bimodal.adj, marvel$Modality$Results$event_type)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                    A3SS A5SS  AFE  ALE  MXE   RI   SE</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Bimodal              20    7    4    6    0    4   17</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Excluded.Dispersed  386  484  538  189   18  783 1260</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Excluded.Primary    289  282  477  101   22  712 1049</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Included.Dispersed  563  385  372  237   12  352 1685</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Included.Primary    439  291  346  176   17   47 1184</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Middle                9   10    9    3    2   51   15</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Multimodal           40   35   43   12    1   81   60</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>后续有两种统计方式：仅按modality分，同时按modality与AS type分。以后者作为演示</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- PropModality(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                       modality.column=<span style="color:#0ff;font-weight:bold">&#34;modality.bimodal.adj&#34;</span>,
</span></span><span style="display:flex;"><span>                       modality.type=<span style="color:#0ff;font-weight:bold">&#34;extended&#34;</span>,
</span></span><span style="display:flex;"><span>                       event.type=c(<span style="color:#0ff;font-weight:bold">&#34;SE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;MXE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;RI&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;A5SS&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;A3SS&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;AFE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;ALE&#34;</span>),
</span></span><span style="display:flex;"><span>                       across.event.type=<span style="color:#fff;font-weight:bold">TRUE</span>,
</span></span><span style="display:flex;"><span>                       prop.test=<span style="color:#0ff;font-weight:bold">&#34;chisq&#34;</span>,
</span></span><span style="display:flex;"><span>                       prop.adj=<span style="color:#0ff;font-weight:bold">&#34;fdr&#34;</span>,
</span></span><span style="display:flex;"><span>                       xlabels.size=<span style="color:#ff0;font-weight:bold">8</span>
</span></span><span style="display:flex;"><span>                       )
</span></span><span style="display:flex;"><span>marvel$Modality$Prop$BarChart$Stats
</span></span><span style="display:flex;"><span>head(marvel$Modality$Prop$BarChart$Table)
</span></span><span style="display:flex;"><span>marvel$Modality$Prop$BarChart$Plot
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230305143612067.png" alt="image-20230305143612067" style="zoom:67%;" />
<h3 id="24-细胞群差异分析">2.4 细胞群差异分析<a hidden class="anchor" aria-hidden="true" href="#24-细胞群差异分析">#</a></h3>
<p><strong>（1）差异基因分析</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>cell.group.g1 = marvel$SplicePheno  %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::filter(cell.type==<span style="color:#0ff;font-weight:bold">&#34;iPSC&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::pull(sample.id)  <span style="color:#007f7f"># (reference)</span>
</span></span><span style="display:flex;"><span>cell.group.g2 = marvel$SplicePheno  %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::filter(cell.type==<span style="color:#0ff;font-weight:bold">&#34;Endoderm&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::pull(sample.id)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel &lt;- CompareValues(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                        cell.group.g1=cell.group.g1,
</span></span><span style="display:flex;"><span>                        cell.group.g2=cell.group.g2,
</span></span><span style="display:flex;"><span>                        min.cells=<span style="color:#ff0;font-weight:bold">3</span>,
</span></span><span style="display:flex;"><span>                        method=<span style="color:#0ff;font-weight:bold">&#34;wilcox&#34;</span>,
</span></span><span style="display:flex;"><span>                        method.adjust=<span style="color:#0ff;font-weight:bold">&#34;fdr&#34;</span>,
</span></span><span style="display:flex;"><span>                        level=<span style="color:#0ff;font-weight:bold">&#34;gene&#34;</span>,
</span></span><span style="display:flex;"><span>                        show.progress=<span style="color:#fff;font-weight:bold">FALSE</span>
</span></span><span style="display:flex;"><span>                        )
</span></span><span style="display:flex;"><span>marvel$DE$Exp$Table[1:<span style="color:#ff0;font-weight:bold">5</span>, ]
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（2）差异AS分析</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># time consuming</span>
</span></span><span style="display:flex;"><span>marvel &lt;- CompareValues(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                        cell.group.g1=cell.group.g1,
</span></span><span style="display:flex;"><span>                        cell.group.g2=cell.group.g2,
</span></span><span style="display:flex;"><span>                        min.cells=<span style="color:#ff0;font-weight:bold">25</span>,
</span></span><span style="display:flex;"><span>                        method=c(<span style="color:#0ff;font-weight:bold">&#34;ad&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;dts&#34;</span>),  <span style="color:#007f7f">#同时使用两种的推荐差异方法</span>
</span></span><span style="display:flex;"><span>                        method.adjust=<span style="color:#0ff;font-weight:bold">&#34;fdr&#34;</span>,
</span></span><span style="display:flex;"><span>                        level=<span style="color:#0ff;font-weight:bold">&#34;splicing&#34;</span>,
</span></span><span style="display:flex;"><span>                        event.type=c(<span style="color:#0ff;font-weight:bold">&#34;SE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;MXE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;RI&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;A5SS&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;A3SS&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;ALE&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;AFE&#34;</span>),
</span></span><span style="display:flex;"><span>                        show.progress=<span style="color:#fff;font-weight:bold">FALSE</span>
</span></span><span style="display:flex;"><span>                        )
</span></span><span style="display:flex;"><span>names(marvel$DE$PSI$Table)
</span></span><span style="display:flex;"><span>head(marvel$DE$PSI$Table[[<span style="color:#0ff;font-weight:bold">&#34;ad&#34;</span>]])
</span></span><span style="display:flex;"><span>head(marvel$DE$PSI$Table[[<span style="color:#0ff;font-weight:bold">&#34;dts&#34;</span>]])
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（3）差异基因&amp;AS分析</strong></p>
<ul>
<li>对差异可变剪切的基因执行基因表达差异分析</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- CompareValues(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                        cell.group.g1=cell.group.g1,
</span></span><span style="display:flex;"><span>                        cell.group.g2=cell.group.g2,
</span></span><span style="display:flex;"><span>                        psi.method=c(<span style="color:#0ff;font-weight:bold">&#34;ad&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;dts&#34;</span>),
</span></span><span style="display:flex;"><span>                        psi.pval=c(<span style="color:#ff0;font-weight:bold">0.10</span>, <span style="color:#ff0;font-weight:bold">0.10</span>),
</span></span><span style="display:flex;"><span>                        psi.delta=<span style="color:#ff0;font-weight:bold">0</span>,
</span></span><span style="display:flex;"><span>                        method.de.gene=<span style="color:#0ff;font-weight:bold">&#34;wilcox&#34;</span>,
</span></span><span style="display:flex;"><span>                        method.adjust.de.gene=<span style="color:#0ff;font-weight:bold">&#34;fdr&#34;</span>,
</span></span><span style="display:flex;"><span>                        downsample=<span style="color:#fff;font-weight:bold">FALSE</span>,
</span></span><span style="display:flex;"><span>                        show.progress=<span style="color:#fff;font-weight:bold">FALSE</span>,
</span></span><span style="display:flex;"><span>                        level=<span style="color:#0ff;font-weight:bold">&#34;gene.spliced&#34;</span>
</span></span><span style="display:flex;"><span>                        )
</span></span><span style="display:flex;"><span>head(marvel$DE$Exp.Spliced$Table)
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>对于上述差异分析结果均可快速火山图，标记marker基因等</p></blockquote>
<h3 id="25-差异as进阶分析">2.5 差异AS进阶分析<a hidden class="anchor" aria-hidden="true" href="#25-差异as进阶分析">#</a></h3>
<p><strong>（1）差异PSI分类</strong></p>
<ul>
<li><strong>Explicit</strong>：明显的变化；<strong>Implicit</strong>：有限的变化；<strong>Restricted</strong>：微小的变化</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- ModalityChange(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                     method=c(<span style="color:#0ff;font-weight:bold">&#34;ad&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;dts&#34;</span>),
</span></span><span style="display:flex;"><span>                     psi.pval=c(<span style="color:#ff0;font-weight:bold">0.10</span>, <span style="color:#ff0;font-weight:bold">0.10</span>)
</span></span><span style="display:flex;"><span>                     )
</span></span><span style="display:flex;"><span>table(marvel$DE$Modality$Table$modality.change)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Explicit   Implicit Restricted </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#    161        280       1113</span>
</span></span><span style="display:flex;"><span>head(marvel$DE$Modality$Table)
</span></span><span style="display:flex;"><span>marvel$DE$Modality$Plot
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>cell.group.list &lt;- list(<span style="color:#0ff;font-weight:bold">&#34;iPSC&#34;</span>=cell.group.g1,
</span></span><span style="display:flex;"><span>                        <span style="color:#0ff;font-weight:bold">&#34;Endoderm&#34;</span>=cell.group.g2
</span></span><span style="display:flex;"><span>                        )
</span></span><span style="display:flex;"><span>tran_id &lt;- <span style="color:#0ff;font-weight:bold">&#34;chr3:129171771:129171634|129171655:-@chr3:129171446:129171538&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># tran_id &lt;- &#34;chr6:21596763:21598248:+@chr6:21598321:21599378&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># tran_id &lt;- &#34;chr3:109337464:109337652:-@chr3:109333920|109333993:109333870&#34;</span>
</span></span><span style="display:flex;"><span>marvel &lt;- PlotValues(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                     cell.group.list=cell.group.list,
</span></span><span style="display:flex;"><span>                     feature=tran_id,
</span></span><span style="display:flex;"><span>                     xlabels.size=<span style="color:#ff0;font-weight:bold">5</span>,
</span></span><span style="display:flex;"><span>                     level=<span style="color:#0ff;font-weight:bold">&#34;splicing&#34;</span>,
</span></span><span style="display:flex;"><span>                     min.cells=<span style="color:#ff0;font-weight:bold">25</span>
</span></span><span style="display:flex;"><span>                     )
</span></span><span style="display:flex;"><span>marvel$adhocPlot$PSI 
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230305150246079.png" alt="image-20230305150246079" style="zoom:67%;" />
<p><strong>（2）差异基因与AS</strong></p>
<ul>
<li><strong>Coordinated</strong>：基因高表达，相应的PSI也增高；</li>
<li><strong>Opposing</strong>：基因表达与PSI变化方向相反；</li>
<li><strong>Isoform-switching</strong> ：基因表达不变，PSI发生变化</li>
<li><strong>Complex</strong>：其它复杂情况</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- IsoSwitch(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                    method=c(<span style="color:#0ff;font-weight:bold">&#34;ad&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;dts&#34;</span>),
</span></span><span style="display:flex;"><span>                    psi.pval=c(<span style="color:#ff0;font-weight:bold">0.10</span>, <span style="color:#ff0;font-weight:bold">0.10</span>),
</span></span><span style="display:flex;"><span>                    psi.delta=<span style="color:#ff0;font-weight:bold">0</span>,
</span></span><span style="display:flex;"><span>                    gene.pval=<span style="color:#ff0;font-weight:bold">0.10</span>,
</span></span><span style="display:flex;"><span>                    gene.log2fc=<span style="color:#ff0;font-weight:bold">0.5</span>
</span></span><span style="display:flex;"><span>                    )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel$DE$Cor$Plot
</span></span><span style="display:flex;"><span>head(marvel$DE$Cor$Table_Raw)
</span></span><span style="display:flex;"><span>table(marvel$DE$Cor$Table$cor)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Coordinated    Opposing  Iso-Switch     Complex </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#         192         158         331         113</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 以Coordinated的DHX9与相应的AFE剪切事件为例</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gene</span>
</span></span><span style="display:flex;"><span>df.feature &lt;- marvel$GeneFeature
</span></span><span style="display:flex;"><span>gene_id &lt;- df.feature[which(df.feature$gene_short_name==<span style="color:#0ff;font-weight:bold">&#34;DHX9&#34;</span>), <span style="color:#0ff;font-weight:bold">&#34;gene_id&#34;</span>]
</span></span><span style="display:flex;"><span>marvel &lt;- PlotValues(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                   cell.group.list=cell.group.list,
</span></span><span style="display:flex;"><span>                   feature=gene_id,
</span></span><span style="display:flex;"><span>                   maintitle=<span style="color:#0ff;font-weight:bold">&#34;gene_short_name&#34;</span>,
</span></span><span style="display:flex;"><span>                   xlabels.size=<span style="color:#ff0;font-weight:bold">7</span>,
</span></span><span style="display:flex;"><span>                   level=<span style="color:#0ff;font-weight:bold">&#34;gene&#34;</span>
</span></span><span style="display:flex;"><span>                   )
</span></span><span style="display:flex;"><span>plot.1_gene &lt;- marvel$adhocPlot$Exp
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Splicing</span>
</span></span><span style="display:flex;"><span>tran_id &lt;- <span style="color:#0ff;font-weight:bold">&#34;chr1:182839347:182839456|182839734:182839935:+@chr1:182842545:182842677&#34;</span>
</span></span><span style="display:flex;"><span>marvel &lt;- PlotValues(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                   cell.group.list=cell.group.list,
</span></span><span style="display:flex;"><span>                   feature=tran_id,
</span></span><span style="display:flex;"><span>                   xlabels.size=<span style="color:#ff0;font-weight:bold">7</span>,
</span></span><span style="display:flex;"><span>                   level=<span style="color:#0ff;font-weight:bold">&#34;splicing&#34;</span>,
</span></span><span style="display:flex;"><span>                   min.cells=<span style="color:#ff0;font-weight:bold">25</span>
</span></span><span style="display:flex;"><span>                   )
</span></span><span style="display:flex;"><span>plot.1_splicing &lt;- marvel$adhocPlot$PSI
</span></span><span style="display:flex;"><span>library(patchwork)
</span></span><span style="display:flex;"><span>plot.1_gene + plot.1_splicing
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230305151250520.png" alt="image-20230305151250520" style="zoom: 67%;" />
<p><strong>（3）NMD预测</strong></p>
<ul>
<li>nonsense-mediated decay (NMD)无义介导的mRNA降解预测</li>
<li>相比对照组，观测组所剪切的外显子可能由于包含premature stop codons (PTCs)，导致编码提前结束，从而进一步使得表达量降低。</li>
<li>本质上是对可产生PTC的可变剪切与下调基因取交集。</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- ParseGTF(MarvelObject=marvel)
</span></span><span style="display:flex;"><span>marvel &lt;- FindPTC(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                  method=c(<span style="color:#0ff;font-weight:bold">&#34;ad&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;dts&#34;</span>),
</span></span><span style="display:flex;"><span>                  pval=c(<span style="color:#ff0;font-weight:bold">0.10</span>, <span style="color:#ff0;font-weight:bold">0.10</span>),
</span></span><span style="display:flex;"><span>                  delta=<span style="color:#ff0;font-weight:bold">5</span>
</span></span><span style="display:flex;"><span>                  )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel &lt;- PropPTC(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                  xlabels.size=<span style="color:#ff0;font-weight:bold">8</span>,
</span></span><span style="display:flex;"><span>                  show.NovelSJ.NoCDS=<span style="color:#fff;font-weight:bold">FALSE</span>, <span style="color:#007f7f">#仅展示Non PTC与PTC</span>
</span></span><span style="display:flex;"><span>                  prop.test=<span style="color:#0ff;font-weight:bold">&#34;chisq&#34;</span>
</span></span><span style="display:flex;"><span>                  )
</span></span><span style="display:flex;"><span>marvel$NMD$PTC.Prop$Table[1:<span style="color:#ff0;font-weight:bold">5</span>, ]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel &lt;- CompareExpr(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                      xlabels.size=<span style="color:#ff0;font-weight:bold">8</span>
</span></span><span style="display:flex;"><span>                      )
</span></span><span style="display:flex;"><span>marvel$NMD$NMD.Expr$Plot
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel &lt;- AnnoVolcanoPlot(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                          anno=<span style="color:#fff;font-weight:bold">FALSE</span>,
</span></span><span style="display:flex;"><span>                          xlabel.size=<span style="color:#ff0;font-weight:bold">7</span>
</span></span><span style="display:flex;"><span>                          )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel$NMD$AnnoVolcanoPlot$Plot
</span></span><span style="display:flex;"><span>marvel$NMD$AnnoVolcanoPlot$Table
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230305152637830.png" alt="image-20230305152637830"  />
</p>
<h1 id="二drop-baed-pipeline">二、Drop-baed pipeline<a hidden class="anchor" aria-hidden="true" href="#二drop-baed-pipeline">#</a></h1>
<h2 id="1上游分析-1">1、上游分析<a hidden class="anchor" aria-hidden="true" href="#1上游分析-1">#</a></h2>
<h3 id="11-conda环境">1.1 conda环境<a hidden class="anchor" aria-hidden="true" href="#11-conda环境">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>conda create -n marvel_drop
</span></span><span style="display:flex;"><span>conda activate marvel_drop
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>conda install -c hcc aspera-cli
</span></span><span style="display:flex;"><span>conda install -c bioconda STAR=2.7.8a samtools
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="12-下载数据">1.2 下载数据<a hidden class="anchor" aria-hidden="true" href="#12-下载数据">#</a></h3>
<p><strong>（1）测序数据</strong></p>
<ul>
<li>官方教程分析了GSE130731的4个10X单细胞测序结果，以其中的两个作为学习的示例(SRR9008752,SRR9008753)</li>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130731">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130731</a></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>ascp -QT -l 300m -P33001  <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-i ~/miniconda3/envs/marvel/etc/asperaweb_id_dsa.openssh   <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/SRR900/003/SRR9008752/SRR9008752_1.fastq.gz ./work/fq/
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>ascp -QT -l 300m -P33001  <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-i ~/miniconda3/envs/marvel/etc/asperaweb_id_dsa.openssh   <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/SRR900/003/SRR9008753/SRR9008753_2.fastq.gz ./work/fq/
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>mv ./work/fq/SRR9008753_1.fastq.gz ./work/fq/SRR9008753_S1_L001_R1_001.fastq.gz
</span></span><span style="display:flex;"><span>mv ./work/fq/SRR9008753_2.fastq.gz ./work/fq/SRR9008753_S1_L001_R2_001.fastq.gz
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（2）基因组数据</strong></p>
<blockquote>
<p>还是在 ./basic/gtf 文件夹下，构建starsolo版本的索引文件。</p>
<p>cellranger软件及相关数据下载参考之前的笔记。</p></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>STAR  \
</span></span><span style="display:flex;"><span>--runMode genomeGenerate \
</span></span><span style="display:flex;"><span>--runThreadN <span style="color:#ff0;font-weight:bold">20</span> \
</span></span><span style="display:flex;"><span>--genomeDir ./starsolo_index \
</span></span><span style="display:flex;"><span>--genomeFastaFiles GRCh38.primary_assembly.genome.fa  \
</span></span><span style="display:flex;"><span>--sjdbGTFfile gencode.v31.annotation.gtf
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（3）教程数据</strong></p>
<ul>
<li><a href="http://datashare.molbiol.ox.ac.uk/public/wwen/MARVEL/Tutorial/Droplet/Data.zip">http://datashare.molbiol.ox.ac.uk/public/wwen/MARVEL/Tutorial/Droplet/Data.zip</a></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>ls Data/
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Gene_SingCellaR  Gene_STARsolo  GTF  R Object  SJ_STARsolo</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="13-比对分析">1.3 比对分析<a hidden class="anchor" aria-hidden="true" href="#13-比对分析">#</a></h3>
<p><strong>（1）cellranger</strong></p>
<p>使用cellranger软件初次比对产生bam文件，该软件用法可参考之前的笔记</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>bin=/home/lishensuo/Softerware/cellranger-7.1.0/bin/cellranger 
</span></span><span style="display:flex;"><span>ref=/home/lishensuo/Softerware/cellranger-7.1.0/refer/refdata-gex-GRCh38-2020-A
</span></span><span style="display:flex;"><span>fq_dir=/home/lishensuo/STUDY/marvel/proj2/work/fq
</span></span><span style="display:flex;"><span>sample=SRR9008752
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold">${</span>bin<span style="color:#0ff;font-weight:bold">}</span> count --id=<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>            --fastqs=<span style="color:#0ff;font-weight:bold">${</span>fq_dir<span style="color:#0ff;font-weight:bold">}</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>            --sample=<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>            --localcores=<span style="color:#ff0;font-weight:bold">10</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>            --transcriptome=<span style="color:#0ff;font-weight:bold">${</span>ref<span style="color:#0ff;font-weight:bold">}</span>
</span></span><span style="display:flex;"><span>            
</span></span><span style="display:flex;"><span>mkdir -p ./work/cellranger
</span></span><span style="display:flex;"><span>mv <span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span> ./work/cellranger/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（2）STARsolo</strong></p>
<p>使用STAR软件针对单细胞数据的模式，分析其基因表达以及剪切位点矩阵</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f"># https://kb.10xgenomics.com/hc/en-us/articles/115004506263-What-is-a-barcode-whitelist-</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Cell Ranger 4 and later: cellranger-x.y.z/lib/python/cellranger/barcodes/</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Cell Ranger 3 and earlier: cellranger-x.y.z/cellranger-cs/x.y.z/lib/python/cellranger/barcodes/</span>
</span></span><span style="display:flex;"><span>ls /home/lishensuo/Softerware/cellranger-7.1.0/lib/python/cellranger/barcodes/
</span></span><span style="display:flex;"><span>whitelist=/home/lishensuo/Softerware/cellranger-7.1.0/lib/python/cellranger/barcodes/737K-august-2016.txt
</span></span><span style="display:flex;"><span>sample=SRR9008752
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>STAR --runThreadN <span style="color:#ff0;font-weight:bold">16</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>     --genomeDir ../basic/gtf/starsolo_index <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>     --soloType CB_UMI_Simple <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>     --readFilesIn ./work/cellranger/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>/outs/possorted_genome_bam.bam <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>     --readFilesCommand samtools view -F 0x100 <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>     --readFilesType SAM SE <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>     --soloInputSAMattrBarcodeSeq CR UR <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>     --soloInputSAMattrBarcodeQual CY UY <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>     --soloCBwhitelist <span style="color:#0ff;font-weight:bold">${</span>whitelist<span style="color:#0ff;font-weight:bold">}</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>     --soloFeatures Gene SJ
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>mkdir ./work/star/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>
</span></span><span style="display:flex;"><span>mv Aligned.out.sam Log.final.out  Log.out  Log.progress.out  SJ.out.tab ./Solo.out
</span></span><span style="display:flex;"><span>mv Solo.out ./work/star/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="2下游分析-1">2、下游分析<a hidden class="anchor" aria-hidden="true" href="#2下游分析-1">#</a></h2>
<h3 id="21-整理比对结果">2.1 整理比对结果<a hidden class="anchor" aria-hidden="true" href="#21-整理比对结果">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>library(R.utils) <span style="color:#007f7f"># ※</span>
</span></span><span style="display:flex;"><span>library(Matrix)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>samples=c(<span style="color:#0ff;font-weight:bold">&#34;SRR9008752&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;SRR9008753&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（1）Raw gene count</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>dir.create(<span style="color:#0ff;font-weight:bold">&#34;./work/gene&#34;</span>)
</span></span><span style="display:flex;"><span>dir.create(<span style="color:#0ff;font-weight:bold">&#34;./work/gene/raw&#34;</span>)
</span></span><span style="display:flex;"><span>dir.create(<span style="color:#0ff;font-weight:bold">&#34;./work/gene/norm&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sce.list = lapply(samples, <span style="color:#fff;font-weight:bold">function</span>(x){
</span></span><span style="display:flex;"><span>    print(x)
</span></span><span style="display:flex;"><span>    fls = list.files(paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/star/&#34;</span>,x,<span style="color:#0ff;font-weight:bold">&#34;/Solo.out/Gene/filtered&#34;</span>), full.name=<span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>    <span style="color:#fff;font-weight:bold">if</span>(length(grep(<span style="color:#0ff;font-weight:bold">&#34;gz&#34;</span>,fls))!=<span style="color:#ff0;font-weight:bold">3</span>){
</span></span><span style="display:flex;"><span>        <span style="color:#fff;font-weight:bold">for</span>(fl <span style="color:#fff;font-weight:bold">in</span> fls){ gzip(fl, remove=<span style="color:#fff;font-weight:bold">FALSE</span>) }
</span></span><span style="display:flex;"><span>    }
</span></span><span style="display:flex;"><span>    count = Read10X(paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/star/&#34;</span>,x,<span style="color:#0ff;font-weight:bold">&#34;/Solo.out/Gene/filtered&#34;</span>))
</span></span><span style="display:flex;"><span>    sce = CreateSeuratObject(count)
</span></span><span style="display:flex;"><span>    sce = RenameCells(sce, add.cell.id=x)
</span></span><span style="display:flex;"><span>    sce$Group = x
</span></span><span style="display:flex;"><span>    <span style="color:#fff;font-weight:bold">return</span>(sce)
</span></span><span style="display:flex;"><span>})
</span></span><span style="display:flex;"><span>names(sce.list) = samples
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 若多个样本则合并处理</span>
</span></span><span style="display:flex;"><span>sce = merge(sce.list[[1]], sce.list[<span style="color:#ff0;font-weight:bold">-1</span>])
</span></span><span style="display:flex;"><span>table(sce$Group)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>count_merge = sce@assays$RNA@counts
</span></span><span style="display:flex;"><span>dim(count_merge)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>writeMM(count_merge, file = <span style="color:#0ff;font-weight:bold">&#39;./work/gene/raw/matrix.mtx&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>count_merge_phe = sce@meta.data %&gt;% 
</span></span><span style="display:flex;"><span>    tibble::rownames_to_column(<span style="color:#0ff;font-weight:bold">&#34;cell.id&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::select(cell.id)  <span style="color:#007f7f"># 仅需要1列</span>
</span></span><span style="display:flex;"><span>write.table(count_merge_phe, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, row.names=<span style="color:#fff;font-weight:bold">F</span>, 
</span></span><span style="display:flex;"><span>    file=<span style="color:#0ff;font-weight:bold">&#34;./work/gene/raw/phenoData.txt&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>count_merge_feat = data.frame(gene_short_name=rownames(sce))
</span></span><span style="display:flex;"><span>write.table(count_merge_feat, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, row.names=<span style="color:#fff;font-weight:bold">F</span>, 
</span></span><span style="display:flex;"><span>    file=<span style="color:#0ff;font-weight:bold">&#34;./work/gene/raw/featureData.txt&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（2）Norm gene count</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sce = NormalizeData(sce)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>count_merge_norm = Matrix::Matrix(exp(sce@assays$RNA@data)<span style="color:#ff0;font-weight:bold">-1</span>, sparse=<span style="color:#fff;font-weight:bold">T</span>)  <span style="color:#007f7f"># 非log转换</span>
</span></span><span style="display:flex;"><span>count_merge_norm[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span>writeMM(count_merge_norm, file = <span style="color:#0ff;font-weight:bold">&#39;./work/gene/norm/matrix_normalised.mtx&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>count_merge_norm_phe = sce@meta.data %&gt;% 
</span></span><span style="display:flex;"><span>    tibble::rownames_to_column(<span style="color:#0ff;font-weight:bold">&#34;cell.id&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::select(cell.id, Group)  <span style="color:#007f7f"># 需要细胞分组等表型信息</span>
</span></span><span style="display:flex;"><span>write.table(count_merge_norm_phe, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, row.names=<span style="color:#fff;font-weight:bold">F</span>, 
</span></span><span style="display:flex;"><span>    file=<span style="color:#0ff;font-weight:bold">&#34;./work/gene/norm/phenoData.txt&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>count_merge_norm_feat = data.frame(gene_short_name=rownames(sce))
</span></span><span style="display:flex;"><span>write.table(count_merge_norm_feat, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, row.names=<span style="color:#fff;font-weight:bold">F</span>, 
</span></span><span style="display:flex;"><span>    file=<span style="color:#0ff;font-weight:bold">&#34;./work/gene/norm/featureData.txt&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（3）Raw SJ count</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>dir.create(<span style="color:#0ff;font-weight:bold">&#34;./work/sj/raw&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 整理SJ meta</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span>(x <span style="color:#fff;font-weight:bold">in</span> samples){
</span></span><span style="display:flex;"><span>sj_name = data.table::fread(paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/star/&#34;</span>,x,<span style="color:#0ff;font-weight:bold">&#34;/Solo.out/SJ.out.tab&#34;</span>), data.table=<span style="color:#fff;font-weight:bold">F</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::mutate(coord.intron=paste(V1, V2, V3, sep=<span style="color:#0ff;font-weight:bold">&#34;:&#34;</span>)) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::select(coord.intron)
</span></span><span style="display:flex;"><span>write.table(sj_name, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, row.names=<span style="color:#fff;font-weight:bold">F</span>, col.names=<span style="color:#fff;font-weight:bold">F</span>, quote=<span style="color:#fff;font-weight:bold">F</span>,
</span></span><span style="display:flex;"><span>    file=paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/star/&#34;</span>,x,<span style="color:#0ff;font-weight:bold">&#34;/Solo.out/SJ/raw/features.tsv&#34;</span>))
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sce_sj.list = lapply(samples, <span style="color:#fff;font-weight:bold">function</span>(x){
</span></span><span style="display:flex;"><span>    <span style="color:#007f7f"># x=&#34;SRR15319877&#34;</span>
</span></span><span style="display:flex;"><span>    print(x)
</span></span><span style="display:flex;"><span>    fls = list.files(paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/star/&#34;</span>,x,<span style="color:#0ff;font-weight:bold">&#34;/Solo.out/SJ/raw&#34;</span>), full.name=<span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>    <span style="color:#fff;font-weight:bold">if</span>(length(grep(<span style="color:#0ff;font-weight:bold">&#34;gz&#34;</span>,fls))!=<span style="color:#ff0;font-weight:bold">3</span>){
</span></span><span style="display:flex;"><span>        <span style="color:#fff;font-weight:bold">for</span>(fl <span style="color:#fff;font-weight:bold">in</span> fls){ gzip(fl, remove=<span style="color:#fff;font-weight:bold">FALSE</span>) }
</span></span><span style="display:flex;"><span>    }
</span></span><span style="display:flex;"><span>    count = Read10X(paste0(<span style="color:#0ff;font-weight:bold">&#34;./work/star/&#34;</span>,x,<span style="color:#0ff;font-weight:bold">&#34;/Solo.out/SJ/raw&#34;</span>),gene.column=<span style="color:#ff0;font-weight:bold">1</span>)
</span></span><span style="display:flex;"><span>    sce_sj = CreateSeuratObject(count)
</span></span><span style="display:flex;"><span>    sce_sj = RenameCells(sce_sj, add.cell.id=x)
</span></span><span style="display:flex;"><span>    sce_sj = sce_sj[,colnames(sce.list[[x]])]
</span></span><span style="display:flex;"><span>    <span style="color:#fff;font-weight:bold">return</span>(sce_sj)
</span></span><span style="display:flex;"><span>})
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sce_sj = merge(sce_sj.list[[1]], sce_sj.list[<span style="color:#ff0;font-weight:bold">-1</span>])
</span></span><span style="display:flex;"><span>sj_count_merge = sce_sj@assays$RNA@counts
</span></span><span style="display:flex;"><span>writeMM(sj_count_merge, file = <span style="color:#0ff;font-weight:bold">&#39;./work/sj/raw/matrix.mtx&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sj_count_merge_phe = sce_sj@meta.data %&gt;% 
</span></span><span style="display:flex;"><span>    tibble::rownames_to_column(<span style="color:#0ff;font-weight:bold">&#34;cell.id&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::select(cell.id)  <span style="color:#007f7f"># 仅需要1列</span>
</span></span><span style="display:flex;"><span>write.table(sj_count_merge_phe, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, row.names=<span style="color:#fff;font-weight:bold">F</span>, 
</span></span><span style="display:flex;"><span>    file=<span style="color:#0ff;font-weight:bold">&#34;./work/sj/raw/phenoData.txt&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sj_count_merge_feat = data.frame(coord.intron=rownames(sce_sj)) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::mutate(coord.intron=gsub(<span style="color:#0ff;font-weight:bold">&#34;-&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;_&#34;</span>, coord.intron))
</span></span><span style="display:flex;"><span>write.table(sj_count_merge_feat, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, row.names=<span style="color:#fff;font-weight:bold">F</span>, 
</span></span><span style="display:flex;"><span>    file=<span style="color:#0ff;font-weight:bold">&#34;./work/sj/raw/featureData.txt&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（4）Cell dimension reduction</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>dir.create(<span style="color:#0ff;font-weight:bold">&#34;./work/umap&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sce = sce %&gt;%
</span></span><span style="display:flex;"><span>  NormalizeData() %&gt;%
</span></span><span style="display:flex;"><span>  FindVariableFeatures() %&gt;%
</span></span><span style="display:flex;"><span>  ScaleData() %&gt;% 
</span></span><span style="display:flex;"><span>  Seurat::RunPCA() %&gt;% 
</span></span><span style="display:flex;"><span>  Seurat::RunUMAP(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>umap_embed = sce@reductions$umap@cell.embeddings %&gt;% 
</span></span><span style="display:flex;"><span>    as.data.frame() %&gt;% 
</span></span><span style="display:flex;"><span>    tibble::rownames_to_column(<span style="color:#0ff;font-weight:bold">&#34;cell.id&#34;</span>) %&gt;%
</span></span><span style="display:flex;"><span>    dplyr::rename(x=UMAP_1, y=UMAP_2) 
</span></span><span style="display:flex;"><span>write.table(umap_embed, row.names=<span style="color:#fff;font-weight:bold">F</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>,
</span></span><span style="display:flex;"><span>    file=<span style="color:#0ff;font-weight:bold">&#34;./work/umap/UMAP_coordinate.txt&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="22-创建marvel对象">2.2 创建marvel对象<a hidden class="anchor" aria-hidden="true" href="#22-创建marvel对象">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># (1) norm gene的三文件</span>
</span></span><span style="display:flex;"><span>df.gene.norm &lt;- readMM(<span style="color:#0ff;font-weight:bold">&#34;./work/gene/norm/matrix_normalised.mtx&#34;</span>)
</span></span><span style="display:flex;"><span>df.gene.norm.pheno &lt;- read.table(<span style="color:#0ff;font-weight:bold">&#34;./work/gene/norm/phenoData.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>df.gene.norm.feature &lt;- read.table(<span style="color:#0ff;font-weight:bold">&#34;./work/gene/norm/featureData.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2) raw gene的三文件</span>
</span></span><span style="display:flex;"><span>df.gene.count &lt;- readMM(<span style="color:#0ff;font-weight:bold">&#34;./work/gene/raw/matrix.mtx&#34;</span>)
</span></span><span style="display:flex;"><span>df.gene.count.pheno &lt;- read.table(<span style="color:#0ff;font-weight:bold">&#34;./work/gene/raw/phenoData.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>df.gene.count.feature &lt;- read.table(<span style="color:#0ff;font-weight:bold">&#34;./work/gene/raw/featureData.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (3) SJ gene的三文件</span>
</span></span><span style="display:flex;"><span>df.sj.count &lt;- readMM(<span style="color:#0ff;font-weight:bold">&#34;./work/sj/raw/matrix.mtx&#34;</span>)
</span></span><span style="display:flex;"><span>df.sj.count.pheno &lt;- read.table(<span style="color:#0ff;font-weight:bold">&#34;./work/sj/raw/phenoData.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>df.sj.count.feature  &lt;- read.table(<span style="color:#0ff;font-weight:bold">&#34;./work/sj/raw/featureData.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>rownames(df.sj.count) = df.sj.count.feature[,<span style="color:#ff0;font-weight:bold">1</span>]
</span></span><span style="display:flex;"><span>colnames(df.sj.count) = df.sj.count.pheno[,<span style="color:#ff0;font-weight:bold">1</span>]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (4) 细胞降维</span>
</span></span><span style="display:flex;"><span>df.coord &lt;- read.table(<span style="color:#0ff;font-weight:bold">&#34;./work/umap/UMAP_coordinate.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (5) GTF</span>
</span></span><span style="display:flex;"><span>gtf &lt;- data.table::fread(<span style="color:#0ff;font-weight:bold">&#34;../basic/gtf/gencode.v31.annotation.gtf&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">FALSE</span>) %&gt;% 
</span></span><span style="display:flex;"><span>     as.data.frame() 
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 由于不同来源gtf版本差异而踩的坑，需要修改如下</span>
</span></span><span style="display:flex;"><span>gtf$V9 = gsub(<span style="color:#0ff;font-weight:bold">&#34;gene_type&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;gene_biotype&#34;</span>,gtf$V9)
</span></span><span style="display:flex;"><span>gtf$V1 = gsub(<span style="color:#0ff;font-weight:bold">&#34;chr&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;&#34;</span>,gtf$V1)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>library(MARVEL)
</span></span><span style="display:flex;"><span>marvel &lt;- CreateMarvelObject.10x(gene.norm.matrix=df.gene.norm,
</span></span><span style="display:flex;"><span>                                 gene.norm.pheno=df.gene.norm.pheno,
</span></span><span style="display:flex;"><span>                                 gene.norm.feature=df.gene.norm.feature,
</span></span><span style="display:flex;"><span>                                 gene.count.matrix=df.gene.count,
</span></span><span style="display:flex;"><span>                                 gene.count.pheno=df.gene.count.pheno,
</span></span><span style="display:flex;"><span>                                 gene.count.feature=df.gene.count.feature,
</span></span><span style="display:flex;"><span>                                 sj.count.matrix=df.sj.count,
</span></span><span style="display:flex;"><span>                                 sj.count.pheno=df.sj.count.pheno,
</span></span><span style="display:flex;"><span>                                 sj.count.feature=df.sj.count.feature,
</span></span><span style="display:flex;"><span>                                 pca=df.coord,
</span></span><span style="display:flex;"><span>                                 gtf=gtf
</span></span><span style="display:flex;"><span>                                 )
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="23-预处理步骤">2.3 预处理步骤<a hidden class="anchor" aria-hidden="true" href="#23-预处理步骤">#</a></h3>
<ul>
<li>筛选有效的剪切位点</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 注释基因类型</span>
</span></span><span style="display:flex;"><span>marvel &lt;- AnnotateGenes.10x(MarvelObject=marvel)
</span></span><span style="display:flex;"><span>head(marvel$gene.metadata)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## SJ涉及两边分属一个蛋白编码基因</span>
</span></span><span style="display:flex;"><span>marvel &lt;- AnnotateSJ.10x(MarvelObject=marvel)  <span style="color:#007f7f">#基因类型</span>
</span></span><span style="display:flex;"><span>head(marvel$sj.metadata)
</span></span><span style="display:flex;"><span>marvel &lt;- ValidateSJ.10x(MarvelObject=marvel)  <span style="color:#007f7f">#同一基因</span>
</span></span><span style="display:flex;"><span>head(marvel$sj.metadata)
</span></span><span style="display:flex;"><span>marvel &lt;- FilterGenes.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                          gene.type=<span style="color:#0ff;font-weight:bold">&#34;protein_coding&#34;</span>  <span style="color:#007f7f">#蛋白基因</span>
</span></span><span style="display:flex;"><span>                          )
</span></span><span style="display:flex;"><span>head(marvel$sj.metadata)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>分析基因在细胞群的表达率以及SJ在细胞群的发生率</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>cell.group.list = split(marvel$sample.metadata$cell.id,marvel$sample.metadata$Group)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 基因在细胞群的表达率</span>
</span></span><span style="display:flex;"><span>marvel &lt;- PlotPctExprCells.Genes.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                                     cell.group.g1=cell.group.list[[1]],
</span></span><span style="display:flex;"><span>                                     cell.group.g2=cell.group.list[[2]],
</span></span><span style="display:flex;"><span>                                     min.pct.cells=<span style="color:#ff0;font-weight:bold">5</span>
</span></span><span style="display:flex;"><span>                                     )
</span></span><span style="display:flex;"><span>head(marvel$pct.cells.expr$Gene$Data)
</span></span><span style="display:flex;"><span>p1 = marvel$pct.cells.expr$Gene$Plot
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 剪切位点在不同细胞群的发生率</span>
</span></span><span style="display:flex;"><span>marvel &lt;- PlotPctExprCells.SJ.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                                  cell.group.g1=cell.group.list[[1]],
</span></span><span style="display:flex;"><span>                                  cell.group.g2=cell.group.list[[2]],
</span></span><span style="display:flex;"><span>                                  min.pct.cells.genes=<span style="color:#ff0;font-weight:bold">5</span>,
</span></span><span style="display:flex;"><span>                                  min.pct.cells.sj=<span style="color:#ff0;font-weight:bold">5</span>,
</span></span><span style="display:flex;"><span>                                  downsample=<span style="color:#fff;font-weight:bold">TRUE</span>,   
</span></span><span style="display:flex;"><span>                                  downsample.pct.sj=<span style="color:#ff0;font-weight:bold">10</span> <span style="color:#007f7f"># 抽样10%分析</span>
</span></span><span style="display:flex;"><span>                                  )
</span></span><span style="display:flex;"><span>head(marvel$pct.cells.expr$SJ$Data)
</span></span><span style="display:flex;"><span>p2 = marvel$pct.cells.expr$SJ$Plot
</span></span><span style="display:flex;"><span>library(patchwork)
</span></span><span style="display:flex;"><span>p1 + p2
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230306091952794.png" alt="image-20230306091952794" style="zoom: 33%;" />
<ul>
<li>保存中间结果</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- CheckAlignment.10x(MarvelObject=marvel)
</span></span><span style="display:flex;"><span>save(marvel, file=<span style="color:#0ff;font-weight:bold">&#34;***.rda&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="24-sj差异分析">2.4 SJ差异分析<a hidden class="anchor" aria-hidden="true" href="#24-sj差异分析">#</a></h3>
<p><strong>（1）差异SJ与基因</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## 首先分析差异SJ</span>
</span></span><span style="display:flex;"><span>marvel &lt;- CompareValues.SJ.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                               cell.group.g1=cell.group.list[[1]],
</span></span><span style="display:flex;"><span>                               cell.group.g2=cell.group.list[[2]],
</span></span><span style="display:flex;"><span>                               min.pct.cells.genes=<span style="color:#ff0;font-weight:bold">10</span>, <span style="color:#007f7f">#至少10%细胞表达</span>
</span></span><span style="display:flex;"><span>                               min.pct.cells.sj=<span style="color:#ff0;font-weight:bold">10</span>,    <span style="color:#007f7f">#至少10%细胞剪切</span>
</span></span><span style="display:flex;"><span>                               min.gene.norm=<span style="color:#ff0;font-weight:bold">1.0</span>,
</span></span><span style="display:flex;"><span>                               seed=<span style="color:#ff0;font-weight:bold">1</span>,
</span></span><span style="display:flex;"><span>                               n.iterations=<span style="color:#ff0;font-weight:bold">100</span>,
</span></span><span style="display:flex;"><span>                               downsample=<span style="color:#fff;font-weight:bold">TRUE</span>,  <span style="color:#007f7f">#根据样本量少的一组下采样</span>
</span></span><span style="display:flex;"><span>                               show.progress=<span style="color:#fff;font-weight:bold">FALSE</span>
</span></span><span style="display:flex;"><span>                               )
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 然后进一步分析差异基因</span>
</span></span><span style="display:flex;"><span>marvel &lt;- CompareValues.Genes.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                                  show.progress=<span style="color:#fff;font-weight:bold">FALSE</span>
</span></span><span style="display:flex;"><span>                                  )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>head(marvel$DE$SJ$Table)
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（2）SJ与基因关系</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>marvel &lt;- IsoSwitch.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                        pval.sj=<span style="color:#ff0;font-weight:bold">0.05</span>,
</span></span><span style="display:flex;"><span>                        delta.sj=<span style="color:#ff0;font-weight:bold">5</span>,
</span></span><span style="display:flex;"><span>                        min.gene.norm=<span style="color:#ff0;font-weight:bold">1.0</span>,
</span></span><span style="display:flex;"><span>                        pval.adj.gene=<span style="color:#ff0;font-weight:bold">0.05</span>,
</span></span><span style="display:flex;"><span>                        log2fc.gene=<span style="color:#ff0;font-weight:bold">0.5</span>
</span></span><span style="display:flex;"><span>                        )
</span></span><span style="display:flex;"><span>head(marvel$SJ.Gene.Cor$Data[,c(<span style="color:#0ff;font-weight:bold">&#34;coord.intron&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;gene_short_name&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;cor.complete&#34;</span>)])
</span></span><span style="display:flex;"><span>marvel$SJ.Gene.Cor$Proportion$Table
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   sj.gene.cor freq       pct</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 Coordinated   26  8.387097</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4    Opposing   54 17.419355</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3  Iso-Switch  226 72.903226</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1     Complex    4  1.290323</span>
</span></span><span style="display:flex;"><span>marvel$SJ.Gene.Cor$Proportion$Plot
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>降维可视化</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># Plot cell groups</span>
</span></span><span style="display:flex;"><span>marvel &lt;- PlotValues.PCA.CellGroup.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                                       cell.group.list=cell.group.list[c(<span style="color:#ff0;font-weight:bold">1</span>,<span style="color:#ff0;font-weight:bold">4</span>)],
</span></span><span style="display:flex;"><span>                                       legendtitle=<span style="color:#0ff;font-weight:bold">&#34;Cell group&#34;</span>,
</span></span><span style="display:flex;"><span>                                       type=<span style="color:#0ff;font-weight:bold">&#34;tsne&#34;</span>
</span></span><span style="display:flex;"><span>                                       )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>plot_group &lt;- marvel$adhocPlot$PCA$CellGroup
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Plot gene expression</span>
</span></span><span style="display:flex;"><span>marvel &lt;- PlotValues.PCA.Gene.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                                  gene_short_name=<span style="color:#0ff;font-weight:bold">&#34;VIM&#34;</span>,
</span></span><span style="display:flex;"><span>                                  color.gradient=c(<span style="color:#0ff;font-weight:bold">&#34;grey&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;cyan&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;green&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;yellow&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;red&#34;</span>),
</span></span><span style="display:flex;"><span>                                  type=<span style="color:#0ff;font-weight:bold">&#34;tsne&#34;</span>
</span></span><span style="display:flex;"><span>                                  )
</span></span><span style="display:flex;"><span>plot_gene &lt;- marvel$adhocPlot$PCA$Gene
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Plot PSI</span>
</span></span><span style="display:flex;"><span>marvel &lt;- PlotValues.PCA.PSI.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                                 coord.intron=<span style="color:#0ff;font-weight:bold">&#34;chr10:17235390:17235845&#34;</span>,
</span></span><span style="display:flex;"><span>                                 min.gene.count=<span style="color:#ff0;font-weight:bold">3</span>,
</span></span><span style="display:flex;"><span>                                 log2.transform=<span style="color:#fff;font-weight:bold">FALSE</span>,
</span></span><span style="display:flex;"><span>                                 color.gradient=c(<span style="color:#0ff;font-weight:bold">&#34;grey&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;cyan&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;green&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;yellow&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;red&#34;</span>),
</span></span><span style="display:flex;"><span>                                 type=<span style="color:#0ff;font-weight:bold">&#34;tsne&#34;</span>
</span></span><span style="display:flex;"><span>                                 )
</span></span><span style="display:flex;"><span>plot_sj &lt;- marvel$adhocPlot$PCA$PSI
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>library(patchwork)
</span></span><span style="display:flex;"><span>p = plot_group + plot_gene + plot_sj
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="25-特定基因分析">2.5 特定基因分析<a hidden class="anchor" aria-hidden="true" href="#25-特定基因分析">#</a></h3>
<p><strong>（1）分组表达</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 该基因在不同阶段的表达</span>
</span></span><span style="display:flex;"><span>marvel &lt;- adhocGene.TabulateExpression.Gene.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                                                cell.group.list=cell.group.list,
</span></span><span style="display:flex;"><span>                                                gene_short_name=<span style="color:#0ff;font-weight:bold">&#34;TPM2&#34;</span>,
</span></span><span style="display:flex;"><span>                                                min.pct.cells=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>                                                downsample=<span style="color:#fff;font-weight:bold">TRUE</span>
</span></span><span style="display:flex;"><span>                                                )
</span></span><span style="display:flex;"><span>marvel$adhocGene$Expression$Gene$Table
</span></span><span style="display:flex;"><span>marvel$adhocGene$Expression$Gene$Plot
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 该基因的不同剪切位点在不同阶段的count表达</span>
</span></span><span style="display:flex;"><span>marvel &lt;- adhocGene.TabulateExpression.PSI.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                                               min.pct.cells=<span style="color:#ff0;font-weight:bold">10</span>
</span></span><span style="display:flex;"><span>                                               )
</span></span><span style="display:flex;"><span>head(marvel$adhocGene$Expression$PSI$Table)
</span></span><span style="display:flex;"><span>marvel$adhocGene$Expression$PSI$Plot
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（2）组间差异</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 组间两两差异</span>
</span></span><span style="display:flex;"><span>marvel &lt;- adhocGene.DE.Gene.10x(MarvelObject=marvel)
</span></span><span style="display:flex;"><span>marvel &lt;- adhocGene.DE.PSI.10x(MarvelObject=marvel)
</span></span><span style="display:flex;"><span>results &lt;- marvel$adhocGene$DE$PSI$Data
</span></span><span style="display:flex;"><span>head(results)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 选择特定SJ绘图 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SJ-1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Define SJ</span>
</span></span><span style="display:flex;"><span>coord.intron &lt;- results[which(results$figure.column==<span style="color:#0ff;font-weight:bold">&#34;SJ-1&#34;</span>), <span style="color:#0ff;font-weight:bold">&#34;coord.intron&#34;</span>]
</span></span><span style="display:flex;"><span>coord.intron &lt;- unique(coord.intron)
</span></span><span style="display:flex;"><span>coord.intron
</span></span><span style="display:flex;"><span>marvel &lt;- adhocGene.PlotDEValues.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                                  coord.intron=coord.intron,
</span></span><span style="display:flex;"><span>                                  log2fc.gene=<span style="color:#ff0;font-weight:bold">0.5</span>,
</span></span><span style="display:flex;"><span>                                  delta.sj=<span style="color:#ff0;font-weight:bold">5</span>,
</span></span><span style="display:flex;"><span>                                  label.size=<span style="color:#ff0;font-weight:bold">2</span>,
</span></span><span style="display:flex;"><span>                                  point.size=<span style="color:#ff0;font-weight:bold">2</span>,
</span></span><span style="display:flex;"><span>                                  xmin=<span style="color:#ff0;font-weight:bold">-2.0</span>,
</span></span><span style="display:flex;"><span>                                  xmax=<span style="color:#ff0;font-weight:bold">2.0</span>,
</span></span><span style="display:flex;"><span>                                  ymin=<span style="color:#ff0;font-weight:bold">-25</span>,
</span></span><span style="display:flex;"><span>                                  ymax=<span style="color:#ff0;font-weight:bold">25</span>
</span></span><span style="display:flex;"><span>                                  )
</span></span><span style="display:flex;"><span>marvel$adhocGene$DE$VolcanoPlot$Plot
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230306092629789.png" alt="image-20230306092629789" style="zoom:67%;" />
<ul>
<li>使用wiggleplotr包可视化某剪切位点在基因不同转录本的可视化</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># http://datashare.molbiol.ox.ac.uk/public/wwen/wiggleplotr_1.18.0.tar.gz</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#install.packages(&#34;wiggleplotr_1.18.0.tar.gz&#34;, repos = NULL, type=&#34;source&#34;)</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel &lt;- adhocGene.PlotSJPosition.10x(MarvelObject=marvel,
</span></span><span style="display:flex;"><span>                                      coord.intron=coord.intron,
</span></span><span style="display:flex;"><span>                                      rescale_introns=<span style="color:#fff;font-weight:bold">FALSE</span>,
</span></span><span style="display:flex;"><span>                                      show.protein.coding.only=<span style="color:#fff;font-weight:bold">TRUE</span>,
</span></span><span style="display:flex;"><span>                                      anno.label.size=<span style="color:#ff0;font-weight:bold">1.5</span>
</span></span><span style="display:flex;"><span>                                      )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>marvel$adhocGene$SJPosition$Plot
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230306092824217.png" alt="image-20230306092824217" style="zoom: 50%;" />
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230401100858958.png" alt="image-20230401100858958" style="zoom:67%;" />


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/031%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scissor%E8%A1%A8%E5%9E%8B%E7%9B%B8%E5%85%B3%E4%BA%9A%E7%BE%A4%E9%89%B4%E5%AE%9A/">
    <span class="title">« Prev Page</span>
    <br>
    <span>单细胞分析工具--Scissor表型相关亚群鉴定</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/033%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--sccustomize%E5%8C%85%E6%94%B9%E8%BF%9Bseurat%E5%8F%AF%E8%A7%86%E5%8C%96/">
    <span class="title">Next Page »</span>
    <br>
    <span>单细胞分析工具--scCustomize包改进Seurat可视化</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
