<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--cellphonedb细胞通讯分析 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
cellphonedb是基于配受体对表达水平，分析单细胞数据中不同细胞类型间相互作用的Python工具。其于2020年在nature protocols发表，目前工具包版本以更新到3.1.0，配受体数据库已更新到4版本。如下将简单学习该软件的用法及结果可视化方法。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/026%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellphonedb%E7%BB%86%E8%83%9E%E9%80%9A%E8%AE%AF%E5%88%86%E6%9E%90/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/026%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellphonedb%E7%BB%86%E8%83%9E%E9%80%9A%E8%AE%AF%E5%88%86%E6%9E%90/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--cellphonedb细胞通讯分析" />
<meta property="og:description" content="
cellphonedb是基于配受体对表达水平，分析单细胞数据中不同细胞类型间相互作用的Python工具。其于2020年在nature protocols发表，目前工具包版本以更新到3.1.0，配受体数据库已更新到4版本。如下将简单学习该软件的用法及结果可视化方法。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/026%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellphonedb%E7%BB%86%E8%83%9E%E9%80%9A%E8%AE%AF%E5%88%86%E6%9E%90/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-01-28T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-01-28T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--cellphonedb细胞通讯分析"/>
<meta name="twitter:description" content="
cellphonedb是基于配受体对表达水平，分析单细胞数据中不同细胞类型间相互作用的Python工具。其于2020年在nature protocols发表，目前工具包版本以更新到3.1.0，配受体数据库已更新到4版本。如下将简单学习该软件的用法及结果可视化方法。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--cellphonedb细胞通讯分析",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/026%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellphonedb%E7%BB%86%E8%83%9E%E9%80%9A%E8%AE%AF%E5%88%86%E6%9E%90/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--cellphonedb细胞通讯分析",
  "name": "单细胞分析工具--cellphonedb细胞通讯分析",
  "description": " cellphonedb是基于配受体对表达水平，分析单细胞数据中不同细胞类型间相互作用的Python工具。其于2020年在nature protocols发表，目前工具包版本以更新到3.1.0，配受体数据库已更新到4版本。如下将简单学习该软件的用法及结果可视化方法。\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " cellphonedb是基于配受体对表达水平，分析单细胞数据中不同细胞类型间相互作用的Python工具。其于2020年在nature protocols发表，目前工具包版本以更新到3.1.0，配受体数据库已更新到4版本。如下将简单学习该软件的用法及结果可视化方法。\n原始论文 2020 | nature protocols | 10.1038/s41596-020-0292-x 2021 | nature genetics | 10.1038/s41588-021-00972-2 官方手册 https://github.com/ventolab/CellphoneDB https://github.com/ventolab/CellphoneDB/blob/master/Docs/RESULTS-DOCUMENTATION.md https://github.com/ventolab/CellphoneDB/blob/master/notebooks/0_prepare_your_data_from_Seurat.ipynb 扩展可视化工具 https://ktplotspy.readthedocs.io/en/latest/notebooks/tutorial.html 1、输入数据 如下图所示，目前cellphonedb可主要实现三种分析模式。第一种直接分析细胞对之间配受体表达水平，第二种方式在前者基础上进一步计算显著P值。第三种则结合了细胞类型的差异基因。\n此次主要学习第二种分析模式，需要提供两类文件：单细胞表达矩阵，细胞类型注释结果。\n单细胞表达矩阵 （1）支持格式包括：.h5ad(scannpy)，10X三文件(mtx/barcode/features)，以及txt等纯文本；对于大型单细胞数据，推荐使用前两者； （2）推荐使用标准化后的count标准矩阵，不需要scale归一化处理； （3）该工具仅支持人类基因名(ensembl, symbol)，其它物种需要同源转换。 细胞类型注释结果 （1）包含两列信息的文本文件，可支持.csv, .txt,.tsv等 （2）第一列的列名是Cell，对应表达矩阵的细胞名；第二列的列名是cell_type，表示注释细胞类型。 2、示例数据 10X官方提供的PBMC3K单细胞测序数据。 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 library(Seurat) library(tidyverse) # https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k count = Read10X(\"filtered_gene_bc_matrices/hg19/\") sce = CreateSeuratObject(count) sce = sce %\u003e% NormalizeData() %\u003e% FindVariableFeatures(nfeatures = 2000) %\u003e% ScaleData() sce = sce %\u003e% RunPCA() %\u003e% RunUMAP(dims = 1:30) %\u003e% #RunTSNE FindNeighbors(dims = 1:30) %\u003e% FindClusters(resolution = 0.1) table(sce$seurat_clusters) # 0 1 2 3 4 # 1201 684 450 351 14 ## (1) 重新保存10X三文件，prepared_10X文件夹中 dir.create(\"prepared_10x\") writeMM(sce@assays$RNA@data, file = 'prepared_10x/matrix.mtx') write(x = rownames(sce@assays$RNA@data), file = \"prepared_10x/features.tsv\") write(x = colnames(sce@assays$RNA@data), file = \"prepared_10x/barcodes.tsv\") ## (2) 保存细胞类型注释结果 sce$cell_type = paste0(\"cluster\",sce$seurat_clusters) sce$Cell = rownames(sce@meta.data) df = sce@meta.data[, c('Cell', 'cell_type')] write.table(df, file ='prepared_meta.tsv', sep = '\\t', quote = F, row.names = F) ## (3) 保存当前seurat对象，用于扩展可视化 saveRDS(sce, file = \"sce.rds\") 将上述的结果(1个文件夹，2个文件)保存到后面的cellphonedb分析环境\n3、cellphonedb分析 （1）建立cellphonedb分析环境 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 ## python版本建议使用3.7 conda create -n cellphonedb python=3.7 -y conda activate cellphonedb conda install -c r r-base rpy2 -y conda install r-pheatmap r-ggplot2 -y conda install -c conda-forge markupsafe=2.0.1 -y pip install cellphonedb cellphonedb --help # Commands: # database # method # plot # query cellphonedb method --help # Commands: # analysis # degs_analysis # statistical_analysis 如上，cellphonedb工具提供4个子命令，其中method用于细胞通讯分析的方法选择。\n后续分析选择statistical_analysis模式，用于计算P值\n（2）cellphonedb分析及结果 1 2 3 4 5 6 cellphonedb method statistical_analysis \\ --output-path pbmc_out \\ --counts-data hgnc_symbol \\ --threshold 0.1 \\ --threads 10 \\ prepared_meta.tsv prepared_10x 该命令有较多参数可以设置，具体可参考官方手册。如上所示列出了觉得比较重要的4个参数。\n--output-path ：输出文件夹名，默认为out\n--counts-data：表达矩阵基因基因名格式，默认为ensembl --threshold：只有细胞类型的相应配受体表达百分比超过阈值才会分析，默认为0.1\n--threads：多线程设置，默认为4\n1 2 ls ./pbmc_out # deconvoluted.txt means.txt pvalues.txt significant_means.txt 如上结果里， means.txt与pvalues.txt表示两两细胞类型之间配受体对的平均表达水平与显著性P值。\nsignificant_means.txt结合上述两个结果表示具有显著意义(至少在一种)的两两细胞类型之间配受体对的平均表达水平。\n如下所示，考虑到配受体对的有向性，对于同一配受体对，会分析所有的细胞类型组合(n*n)可能。\nid_cp_interaction、interacting_pair表示配受体对的代号与组成；\npartner_a、gene_a表示组成A的蛋白标识符、基因名；同理表示组成B；\nreceptor_a 表示组成A是否为受体；其它列含义可参看官方手册。\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 sig_m = read.table(\"./pbmc_out/significant_means.txt\", sep=\"\\t\", header=T, check.names=F) t(sig_m[1,]) # 1 # id_cp_interaction \"CPI-SS0E292C126\" # interacting_pair \"KLRB1_CLEC2D\" # partner_a \"simple:Q12918\" # partner_b \"simple:Q9UHP7\" # gene_a \"KLRB1\" # gene_b \"CLEC2D\" # secreted \"False\" # receptor_a \"True\" # receptor_b \"True\" # annotation_strategy \"curated\" # is_integrin \"False\" # rank \"0.04\" # cluster0|cluster0 NA # cluster0|cluster1 NA # cluster0|cluster2 NA # cluster0|cluster3 NA # cluster0|cluster4 NA # cluster1|cluster0 NA # cluster1|cluster1 NA # cluster1|cluster2 NA # cluster1|cluster3 NA # cluster1|cluster4 NA # cluster2|cluster0 \"0.291\" # cluster2|cluster1 NA # cluster2|cluster2 NA # cluster2|cluster3 NA # cluster2|cluster4 NA # cluster3|cluster0 NA # cluster3|cluster1 NA # cluster3|cluster2 NA # cluster3|cluster3 NA # cluster3|cluster4 NA # cluster4|cluster0 NA # cluster4|cluster1 NA # cluster4|cluster2 NA # cluster4|cluster3 NA # cluster4|cluster4 NA 4、结果可视化 4.1 cellphonedb可视化 细胞类型间通讯热图，需要提供cellphonedb分析结果路径以及细胞类型注释结果 1 2 3 4 5 6 7 8 9 10 cellphonedb plot heatmap_plot \\ --pvalues-path ./pbmc_out/pvalues.txt \\ --output-path ./pbmc_out \\ --count-name heatmap_count.pdf \\ --log-name heatmap_log_count.pdf \\ --count-network-name count_network.txt \\ --interaction-count-name interaction_count.txt \\ ./prepared_meta.tsv # 根据是否log转换，会产生两张热图 配受体对的点图，需要提供cellphonedb分析结果路径 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 # 默认将所有配受体对的所有细胞类型组合进行可视化 cat rows.txt # KLRB1_CLEC2D # NRG1_MS4A4A # CLEC2B_KLRF1 # NCR3_BAG6 # CD40_CD40LG # NRG1_LSR cat columns.txt # cluster0|cluster0 # cluster0|cluster1 # cluster0|cluster2 # cluster0|cluster3 # cluster0|cluster4 cellphonedb plot dot_plot \\ --means-path ./pbmc_out/means.txt \\ --pvalues-path ./pbmc_out/pvalues.txt \\ --output-path ./pbmc_out \\ --output-name plot.pdf \\ --rows rows.txt \\ --columns columns.txt 4.2 ktplotspy可视化 上述cellphonedb提供的可视化方法较为简单，难以调整绘图细节。ktplotspy工具包可针对cellphonedb分析结果提供较为丰富的绘图方案。由于依赖包冲突，需要单独为ktplotspy单独创建一个conda环境。\n（1）创建conda环境 1 2 3 4 conda create -n cellphonedb_plot python notebook -y conda install -c bioconda r-seurat bioconductor-singlecellexperiment conda install -c bioconda anndata anndata2ri pip install ktplotspy 后续在jupyter notebook中操作、分析\n（2）导入数据：seurat对象转为anndata格式，并读入cellphonedb分析结果 1 2 3 4 5 6 7 8 import os import anndata as ad import pandas as pd import ktplotspy as kpy import matplotlib.pyplot as plt import anndata2ri anndata2ri.activate() %load_ext rpy2.ipython 1 2 3 4 %%R library(Seurat) sce = readRDS(\"sce.rds\") sce 1 2 3 4 %%R -o anndata #convert the Seurat object to a SingleCellExperiment object anndata \u003c- as.SingleCellExperiment(sce) anndata 1 2 3 4 5 6 7 8 9 anndata # AnnData object with n_obs × n_vars = 2700 × 32738 # obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.1', 'seurat_clusters', 'cell_type', 'Cell', 'ident' # obsm: 'X_pca', 'X_umap' # layers: 'logcounts' means = pd.read_csv('pbmc_out/means.txt', sep = '\\t') pvals = pd.read_csv('pbmc_out/pvalues.txt', sep = '\\t') decon = pd.read_csv('pbmc_out/deconvoluted.txt', sep = '\\t') （3）绘制热图 1 2 3 4 5 6 7 8 kpy.plot_cpdb_heatmap( adata=adata, pvals=pvals, celltype_key=\"cell_type\", figsize = (5,5), # symmetrical = False, title = \"Sum of significant interactions\" ) （4）绘制点图 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 kpy.plot_cpdb( adata=adata, cell_type1=\"cluster0\", # cell_type1=\"cluster[0,1]\", # this means culster0 and cluster1 cell_type2=\".\", # this means all cell-types means=means, pvals=pvals, celltype_key=\"cell_type\", genes=[\"CD44\"], figsize = (10,5), # highlight_col = \"red\", # highlight_size = 1, 将所有显著的配受体对标识圆环的宽度相同 default_style = True, title = \"Interactions Dotplot\" ) # cell_type1与cell_type2并无先后顺序的区别 （5）绘制弦图 1 2 3 4 5 6 7 8 9 10 11 12 kpy.plot_cpdb_chord( adata=adata, cell_type1=\".\", cell_type2=\".\", means=means, pvals=pvals, deconvoluted=decon, celltype_key=\"cell_type\", genes=[\"CD44\"], edge_cmap=plt.cm.coolwarm, figsize=(6,6) ) 1 2 3 4 5 6 7 8 9 ## 可设置face_col_dict参数，修改每个条带的颜色 face_col_dict={ \"cluster0\": \"brown\", \"cluster1\": \"grey\", \"cluster2\": \"orange\", \"cluster3\": \"pink\", \"cluster4\": \"cyan\", }, edge_col_dict={\"CD44-TYROBP\": \"red\", \"CD44-LGALS9\": \"blue\"} ",
  "wordCount" : "2488",
  "inLanguage": "en",
  "datePublished": "2023-01-28T00:00:00Z",
  "dateModified": "2023-01-28T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/026%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellphonedb%E7%BB%86%E8%83%9E%E9%80%9A%E8%AE%AF%E5%88%86%E6%9E%90/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--cellphonedb细胞通讯分析
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-28 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-28&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-28&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2488&amp;nbsp;|&amp;nbsp;5 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e8%be%93%e5%85%a5%e6%95%b0%e6%8d%ae" aria-label="1、输入数据">1、输入数据</a></li>
                    <li>
                        <a href="#2%e7%a4%ba%e4%be%8b%e6%95%b0%e6%8d%ae" aria-label="2、示例数据">2、示例数据</a></li>
                    <li>
                        <a href="#3cellphonedb%e5%88%86%e6%9e%90" aria-label="3、cellphonedb分析">3、cellphonedb分析</a></li>
                    <li>
                        <a href="#4%e7%bb%93%e6%9e%9c%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="4、结果可视化">4、结果可视化</a><ul>
                            
                    <li>
                        <a href="#41-cellphonedb%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="4.1 cellphonedb可视化">4.1 cellphonedb可视化</a></li>
                    <li>
                        <a href="#42-ktplotspy%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="4.2 ktplotspy可视化">4.2 ktplotspy可视化</a>
                    </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>cellphonedb是基于配受体对表达水平，分析单细胞数据中不同细胞类型间相互作用的Python工具。其于2020年在nature protocols发表，目前工具包版本以更新到3.1.0，配受体数据库已更新到4版本。如下将简单学习该软件的用法及结果可视化方法。</p>
<ul>
<li>原始论文
<ul>
<li>2020 | nature protocols |  <a href="https://doi.org/10.1038/s41596-020-0292-x">10.1038/s41596-020-0292-x</a></li>
<li>2021 | nature genetics | <a href="https://doi.org/10.1038/s41588-021-00972-2">10.1038/s41588-021-00972-2</a></li>
</ul>
</li>
<li>官方手册
<ul>
<li><a href="https://github.com/ventolab/CellphoneDB">https://github.com/ventolab/CellphoneDB</a></li>
<li><a href="https://github.com/ventolab/CellphoneDB/blob/master/Docs/RESULTS-DOCUMENTATION.md">https://github.com/ventolab/CellphoneDB/blob/master/Docs/RESULTS-DOCUMENTATION.md</a></li>
<li><a href="https://github.com/ventolab/CellphoneDB/blob/master/notebooks/0_prepare_your_data_from_Seurat.ipynb">https://github.com/ventolab/CellphoneDB/blob/master/notebooks/0_prepare_your_data_from_Seurat.ipynb</a></li>
</ul>
</li>
<li>扩展可视化工具
<ul>
<li><a href="https://ktplotspy.readthedocs.io/en/latest/notebooks/tutorial.html">https://ktplotspy.readthedocs.io/en/latest/notebooks/tutorial.html</a></li>
</ul>
</li>
</ul>
<p><img loading="lazy" src="https://www.cellphonedb.org/images/cpdb-fast-query-explanation_2.png" alt="img"  />
</p></blockquote>
<h2 id="1输入数据">1、输入数据<a hidden class="anchor" aria-hidden="true" href="#1输入数据">#</a></h2>
<p>如下图所示，目前cellphonedb可主要实现三种分析模式。第一种直接分析细胞对之间配受体表达水平，第二种方式在前者基础上进一步计算显著P值。第三种则结合了细胞类型的差异基因。</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128095622987.png" alt="image-20230128095622987"  />
</p>
<p>此次主要学习第二种分析模式，需要提供两类文件：单细胞表达矩阵，细胞类型注释结果。</p>
<ul>
<li>单细胞表达矩阵
<ul>
<li>（1）支持格式包括：<code>.h5ad</code>(scannpy)，10X三文件(mtx/barcode/features)，以及txt等纯文本；对于大型单细胞数据，推荐使用前两者；</li>
<li>（2）推荐使用标准化后的count标准矩阵，不需要scale归一化处理；</li>
<li>（3）该工具仅支持人类基因名(ensembl, symbol)，其它物种需要同源转换。</li>
</ul>
</li>
<li>细胞类型注释结果
<ul>
<li>（1）包含两列信息的文本文件，可支持<code>.csv</code>, <code>.txt</code>,<code>.tsv</code>等</li>
<li>（2）第一列的列名是<strong>Cell</strong>，对应表达矩阵的细胞名；第二列的列名是<strong>cell_type</strong>，表示注释细胞类型。</li>
</ul>
</li>
</ul>
<h2 id="2示例数据">2、示例数据<a hidden class="anchor" aria-hidden="true" href="#2示例数据">#</a></h2>
<ul>
<li>10X官方提供的PBMC3K单细胞测序数据。</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># https://support.10xgenomics.com/single-cell-gene-expression/datasets/1.1.0/pbmc3k</span>
</span></span><span style="display:flex;"><span>count = Read10X(<span style="color:#0ff;font-weight:bold">&#34;filtered_gene_bc_matrices/hg19/&#34;</span>)
</span></span><span style="display:flex;"><span>sce = CreateSeuratObject(count)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sce = sce %&gt;%
</span></span><span style="display:flex;"><span>  NormalizeData() %&gt;%
</span></span><span style="display:flex;"><span>  FindVariableFeatures(nfeatures = <span style="color:#ff0;font-weight:bold">2000</span>) %&gt;%
</span></span><span style="display:flex;"><span>  ScaleData()
</span></span><span style="display:flex;"><span>sce = sce %&gt;% 
</span></span><span style="display:flex;"><span>  RunPCA() %&gt;% 
</span></span><span style="display:flex;"><span>  RunUMAP(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;%  <span style="color:#007f7f">#RunTSNE</span>
</span></span><span style="display:flex;"><span>  FindNeighbors(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  FindClusters(resolution = <span style="color:#ff0;font-weight:bold">0.1</span>)
</span></span><span style="display:flex;"><span>table(sce$seurat_clusters)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#    0    1    2    3    4 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1201  684  450  351   14 </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (1) 重新保存10X三文件，prepared_10X文件夹中</span>
</span></span><span style="display:flex;"><span>dir.create(<span style="color:#0ff;font-weight:bold">&#34;prepared_10x&#34;</span>)
</span></span><span style="display:flex;"><span>writeMM(sce@assays$RNA@data, file = <span style="color:#0ff;font-weight:bold">&#39;prepared_10x/matrix.mtx&#39;</span>)
</span></span><span style="display:flex;"><span>write(x = rownames(sce@assays$RNA@data), file = <span style="color:#0ff;font-weight:bold">&#34;prepared_10x/features.tsv&#34;</span>)
</span></span><span style="display:flex;"><span>write(x = colnames(sce@assays$RNA@data), file = <span style="color:#0ff;font-weight:bold">&#34;prepared_10x/barcodes.tsv&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (2) 保存细胞类型注释结果</span>
</span></span><span style="display:flex;"><span>sce$cell_type = paste0(<span style="color:#0ff;font-weight:bold">&#34;cluster&#34;</span>,sce$seurat_clusters)
</span></span><span style="display:flex;"><span>sce$Cell = rownames(sce@meta.data)
</span></span><span style="display:flex;"><span>df = sce@meta.data[, c(<span style="color:#0ff;font-weight:bold">&#39;Cell&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;cell_type&#39;</span>)]
</span></span><span style="display:flex;"><span>write.table(df, file =<span style="color:#0ff;font-weight:bold">&#39;prepared_meta.tsv&#39;</span>, 
</span></span><span style="display:flex;"><span>            sep = <span style="color:#0ff;font-weight:bold">&#39;\t&#39;</span>, quote = <span style="color:#fff;font-weight:bold">F</span>, row.names = <span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (3) 保存当前seurat对象，用于扩展可视化</span>
</span></span><span style="display:flex;"><span>saveRDS(sce, file = <span style="color:#0ff;font-weight:bold">&#34;sce.rds&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>将上述的结果(1个文件夹，2个文件)保存到后面的cellphonedb分析环境</p></blockquote>
<h2 id="3cellphonedb分析">3、cellphonedb分析<a hidden class="anchor" aria-hidden="true" href="#3cellphonedb分析">#</a></h2>
<ul>
<li>（1）建立cellphonedb分析环境</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f">## python版本建议使用3.7</span>
</span></span><span style="display:flex;"><span>conda create -n cellphonedb python=3.7 -y
</span></span><span style="display:flex;"><span>conda activate cellphonedb
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>conda install -c r r-base rpy2 -y
</span></span><span style="display:flex;"><span>conda install r-pheatmap r-ggplot2 -y
</span></span><span style="display:flex;"><span>conda install -c conda-forge markupsafe=2.0.1 -y
</span></span><span style="display:flex;"><span>pip install cellphonedb
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>cellphonedb --help
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Commands:</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   database</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   method</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   plot</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   query</span>
</span></span><span style="display:flex;"><span>cellphonedb method --help
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Commands:</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   analysis</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   degs_analysis</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   statistical_analysis</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>如上，cellphonedb工具提供4个子命令，其中method用于细胞通讯分析的方法选择。</p>
<p>后续分析选择<code>statistical_analysis</code>模式，用于计算P值</p></blockquote>
<ul>
<li>（2）cellphonedb分析及结果</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>cellphonedb method statistical_analysis  <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--output-path pbmc_out <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--counts-data hgnc_symbol  <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--threshold 0.1 <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--threads <span style="color:#ff0;font-weight:bold">10</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>prepared_meta.tsv  prepared_10x
</span></span></code></pre></td></tr></table>
</div>
</div><p>该命令有较多参数可以设置，具体可参考官方手册。如上所示列出了觉得比较重要的4个参数。</p>
<p><code>--output-path</code> ：输出文件夹名，默认为<code>out</code></p>
<p><code>--counts-data</code>：表达矩阵基因基因名格式，默认为<code>ensembl </code></p>
<p><code>--threshold</code>：只有细胞类型的相应配受体表达百分比超过阈值才会分析，默认为0.1</p>
<p><code>--threads</code>：多线程设置，默认为4</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>ls ./pbmc_out
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># deconvoluted.txt  means.txt  pvalues.txt  significant_means.txt</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>如上结果里， <strong>means.txt</strong>与<strong>pvalues.txt</strong>表示两两细胞类型之间配受体对的平均表达水平与显著性P值。</p>
<p><strong>significant_means.txt</strong>结合上述两个结果表示具有显著意义(至少在一种)的两两细胞类型之间配受体对的平均表达水平。</p>
<p>如下所示，考虑到配受体对的有向性，对于同一配受体对，会分析所有的细胞类型组合(n*n)可能。</p>
<p><code>id_cp_interaction</code>、<code>interacting_pair</code>表示配受体对的代号与组成；</p>
<p><code>partner_a</code>、<code>gene_a</code>表示组成A的蛋白标识符、基因名；同理表示组成B；</p>
<p><code>receptor_a</code> 表示组成A是否为受体；其它列含义可参看官方手册。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sig_m = read.table(<span style="color:#0ff;font-weight:bold">&#34;./pbmc_out/significant_means.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>, header=<span style="color:#fff;font-weight:bold">T</span>, check.names=<span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>t(sig_m[1,])
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                     1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># id_cp_interaction   &#34;CPI-SS0E292C126&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># interacting_pair    &#34;KLRB1_CLEC2D&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># partner_a           &#34;simple:Q12918&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># partner_b           &#34;simple:Q9UHP7&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gene_a              &#34;KLRB1&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gene_b              &#34;CLEC2D&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># secreted            &#34;False&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># receptor_a          &#34;True&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># receptor_b          &#34;True&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># annotation_strategy &#34;curated&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># is_integrin         &#34;False&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># rank                &#34;0.04&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster0|cluster0   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster0|cluster1   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster0|cluster2   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster0|cluster3   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster0|cluster4   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster1|cluster0   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster1|cluster1   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster1|cluster2   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster1|cluster3   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster1|cluster4   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster2|cluster0   &#34;0.291&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster2|cluster1   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster2|cluster2   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster2|cluster3   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster2|cluster4   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster3|cluster0   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster3|cluster1   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster3|cluster2   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster3|cluster3   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster3|cluster4   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster4|cluster0   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster4|cluster1   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster4|cluster2   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster4|cluster3   NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster4|cluster4   NA</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="4结果可视化">4、结果可视化<a hidden class="anchor" aria-hidden="true" href="#4结果可视化">#</a></h2>
<h3 id="41-cellphonedb可视化">4.1 cellphonedb可视化<a hidden class="anchor" aria-hidden="true" href="#41-cellphonedb可视化">#</a></h3>
<ul>
<li>细胞类型间通讯热图，需要提供cellphonedb分析结果路径以及细胞类型注释结果</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>cellphonedb plot heatmap_plot <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--pvalues-path ./pbmc_out/pvalues.txt <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--output-path ./pbmc_out <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--count-name heatmap_count.pdf <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--log-name heatmap_log_count.pdf <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--count-network-name count_network.txt <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--interaction-count-name interaction_count.txt <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>./prepared_meta.tsv
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 根据是否log转换，会产生两张热图</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128111731896.png" alt="image-20230128111731896" style="zoom: 67%;" />
<ul>
<li>配受体对的点图，需要提供cellphonedb分析结果路径</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f"># 默认将所有配受体对的所有细胞类型组合进行可视化</span>
</span></span><span style="display:flex;"><span>cat rows.txt
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># KLRB1_CLEC2D</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NRG1_MS4A4A</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CLEC2B_KLRF1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NCR3_BAG6</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CD40_CD40LG</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NRG1_LSR</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>cat columns.txt
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster0|cluster0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster0|cluster1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster0|cluster2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster0|cluster3</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cluster0|cluster4</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>cellphonedb plot dot_plot <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--means-path ./pbmc_out/means.txt <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--pvalues-path ./pbmc_out/pvalues.txt <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--output-path ./pbmc_out <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--output-name plot.pdf <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--rows rows.txt <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--columns columns.txt
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128113030377.png" alt="image-20230128113030377" style="zoom: 50%;" />
<h3 id="42-ktplotspy可视化">4.2 ktplotspy可视化<a hidden class="anchor" aria-hidden="true" href="#42-ktplotspy可视化">#</a></h3>
<p>上述cellphonedb提供的可视化方法较为简单，难以调整绘图细节。ktplotspy工具包可针对cellphonedb分析结果提供较为丰富的绘图方案。由于依赖包冲突，需要单独为ktplotspy单独创建一个conda环境。</p>
<ul>
<li>（1）创建conda环境</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>conda create -n cellphonedb_plot python notebook -y
</span></span><span style="display:flex;"><span>conda install -c bioconda r-seurat bioconductor-singlecellexperiment
</span></span><span style="display:flex;"><span>conda install -c bioconda anndata anndata2ri
</span></span><span style="display:flex;"><span>pip install ktplotspy
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>后续在jupyter notebook中操作、分析</p></blockquote>
<ul>
<li>（2）导入数据：seurat对象转为anndata格式，并读入cellphonedb分析结果</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> os
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> anndata <span style="color:#fff;font-weight:bold">as</span> ad
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> pandas <span style="color:#fff;font-weight:bold">as</span> pd
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> ktplotspy <span style="color:#fff;font-weight:bold">as</span> kpy
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> matplotlib.pyplot <span style="color:#fff;font-weight:bold">as</span> plt
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> anndata2ri
</span></span><span style="display:flex;"><span>anndata2ri.activate()
</span></span><span style="display:flex;"><span>%load_ext rpy2.ipython
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>%%R
</span></span><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>sce = readRDS(<span style="color:#0ff;font-weight:bold">&#34;sce.rds&#34;</span>)
</span></span><span style="display:flex;"><span>sce
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>%%R -o anndata
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#convert the Seurat object to a SingleCellExperiment object</span>
</span></span><span style="display:flex;"><span>anndata &lt;- <span style="color:#fff;font-weight:bold">as</span>.SingleCellExperiment(sce)
</span></span><span style="display:flex;"><span>anndata
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>anndata
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AnnData object with n_obs × n_vars = 2700 × 32738</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obs: &#39;orig.ident&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;RNA_snn_res.0.1&#39;, &#39;seurat_clusters&#39;, &#39;cell_type&#39;, &#39;Cell&#39;, &#39;ident&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     layers: &#39;logcounts&#39;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>means = pd.read_csv(<span style="color:#0ff;font-weight:bold">&#39;pbmc_out/means.txt&#39;</span>, sep = <span style="color:#0ff;font-weight:bold">&#39;</span><span style="color:#0ff;font-weight:bold">\t</span><span style="color:#0ff;font-weight:bold">&#39;</span>)
</span></span><span style="display:flex;"><span>pvals = pd.read_csv(<span style="color:#0ff;font-weight:bold">&#39;pbmc_out/pvalues.txt&#39;</span>, sep = <span style="color:#0ff;font-weight:bold">&#39;</span><span style="color:#0ff;font-weight:bold">\t</span><span style="color:#0ff;font-weight:bold">&#39;</span>)
</span></span><span style="display:flex;"><span>decon = pd.read_csv(<span style="color:#0ff;font-weight:bold">&#39;pbmc_out/deconvoluted.txt&#39;</span>, sep = <span style="color:#0ff;font-weight:bold">&#39;</span><span style="color:#0ff;font-weight:bold">\t</span><span style="color:#0ff;font-weight:bold">&#39;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（3）绘制热图</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>kpy.plot_cpdb_heatmap(
</span></span><span style="display:flex;"><span>        adata=adata,
</span></span><span style="display:flex;"><span>        pvals=pvals,
</span></span><span style="display:flex;"><span>        celltype_key=<span style="color:#0ff;font-weight:bold">&#34;cell_type&#34;</span>,
</span></span><span style="display:flex;"><span>        figsize = (<span style="color:#ff0;font-weight:bold">5</span>,<span style="color:#ff0;font-weight:bold">5</span>),
</span></span><span style="display:flex;"><span>        <span style="color:#007f7f"># symmetrical = False,</span>
</span></span><span style="display:flex;"><span>        title = <span style="color:#0ff;font-weight:bold">&#34;Sum of significant interactions&#34;</span>
</span></span><span style="display:flex;"><span>    )
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128125645669.png" alt="image-20230128125645669" style="zoom: 50%;" />
<ul>
<li>（4）绘制点图</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>kpy.plot_cpdb(
</span></span><span style="display:flex;"><span>        adata=adata,
</span></span><span style="display:flex;"><span>        cell_type1=<span style="color:#0ff;font-weight:bold">&#34;cluster0&#34;</span>,   
</span></span><span style="display:flex;"><span>        <span style="color:#007f7f"># cell_type1=&#34;cluster[0,1]&#34;,   #  this means culster0 and cluster1</span>
</span></span><span style="display:flex;"><span>        cell_type2=<span style="color:#0ff;font-weight:bold">&#34;.&#34;</span>, <span style="color:#007f7f"># this means all cell-types</span>
</span></span><span style="display:flex;"><span>        means=means,
</span></span><span style="display:flex;"><span>        pvals=pvals,
</span></span><span style="display:flex;"><span>        celltype_key=<span style="color:#0ff;font-weight:bold">&#34;cell_type&#34;</span>,
</span></span><span style="display:flex;"><span>        genes=[<span style="color:#0ff;font-weight:bold">&#34;CD44&#34;</span>],
</span></span><span style="display:flex;"><span>        figsize = (<span style="color:#ff0;font-weight:bold">10</span>,<span style="color:#ff0;font-weight:bold">5</span>),
</span></span><span style="display:flex;"><span>        <span style="color:#007f7f"># highlight_col = &#34;red&#34;,</span>
</span></span><span style="display:flex;"><span>        <span style="color:#007f7f"># highlight_size = 1,  将所有显著的配受体对标识圆环的宽度相同</span>
</span></span><span style="display:flex;"><span>        default_style = <span style="color:#fff;font-weight:bold">True</span>,
</span></span><span style="display:flex;"><span>        title = <span style="color:#0ff;font-weight:bold">&#34;Interactions Dotplot&#34;</span>
</span></span><span style="display:flex;"><span>    )
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># cell_type1与cell_type2并无先后顺序的区别</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128130715829.png" alt="image-20230128130715829" style="zoom:50%;" />
<ul>
<li>（5）绘制弦图</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>kpy.plot_cpdb_chord(
</span></span><span style="display:flex;"><span>        adata=adata,
</span></span><span style="display:flex;"><span>        cell_type1=<span style="color:#0ff;font-weight:bold">&#34;.&#34;</span>,
</span></span><span style="display:flex;"><span>        cell_type2=<span style="color:#0ff;font-weight:bold">&#34;.&#34;</span>,
</span></span><span style="display:flex;"><span>        means=means,
</span></span><span style="display:flex;"><span>        pvals=pvals,
</span></span><span style="display:flex;"><span>        deconvoluted=decon,
</span></span><span style="display:flex;"><span>        celltype_key=<span style="color:#0ff;font-weight:bold">&#34;cell_type&#34;</span>,
</span></span><span style="display:flex;"><span>        genes=[<span style="color:#0ff;font-weight:bold">&#34;CD44&#34;</span>],
</span></span><span style="display:flex;"><span>        edge_cmap=plt.cm.coolwarm,
</span></span><span style="display:flex;"><span>        figsize=(<span style="color:#ff0;font-weight:bold">6</span>,<span style="color:#ff0;font-weight:bold">6</span>)
</span></span><span style="display:flex;"><span>    )
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128130923711.png" alt="image-20230128130923711" style="zoom: 50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f">## 可设置face_col_dict参数，修改每个条带的颜色</span>
</span></span><span style="display:flex;"><span>face_col_dict={
</span></span><span style="display:flex;"><span>    <span style="color:#0ff;font-weight:bold">&#34;cluster0&#34;</span>: <span style="color:#0ff;font-weight:bold">&#34;brown&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#0ff;font-weight:bold">&#34;cluster1&#34;</span>: <span style="color:#0ff;font-weight:bold">&#34;grey&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#0ff;font-weight:bold">&#34;cluster2&#34;</span>: <span style="color:#0ff;font-weight:bold">&#34;orange&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#0ff;font-weight:bold">&#34;cluster3&#34;</span>: <span style="color:#0ff;font-weight:bold">&#34;pink&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#0ff;font-weight:bold">&#34;cluster4&#34;</span>: <span style="color:#0ff;font-weight:bold">&#34;cyan&#34;</span>,
</span></span><span style="display:flex;"><span>},
</span></span><span style="display:flex;"><span>edge_col_dict={<span style="color:#0ff;font-weight:bold">&#34;CD44-TYROBP&#34;</span>: <span style="color:#0ff;font-weight:bold">&#34;red&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;CD44-LGALS9&#34;</span>: <span style="color:#0ff;font-weight:bold">&#34;blue&#34;</span>}
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/program/021python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6-numpy%E6%95%B0%E7%BB%84torch%E5%BC%A0%E9%87%8F/">
    <span class="title">« Prev Page</span>
    <br>
    <span>Python数据科学-(2)Numpy数组(torch张量)</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/program/022python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6-pandas%E8%A1%A8%E6%A0%BC/">
    <span class="title">Next Page »</span>
    <br>
    <span>Python数据科学-(3)Pandas表格</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
