<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>gseapy富集分析与可视化 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="转录组">
<meta name="description" content="参考

https://gseapy.readthedocs.io/en/latest/index.html
示例数据来自 https://github.com/zqfang/GSEApy/tree/master/tests/data



1
2
3
4
5
6
7
8


import os
import pandas as pd
import scanpy as sc

import gseapy as gp
from gseapy import barplot, dotplot

import matplotlib.pyplot as plt


0. 基因集获取

Enrichr API supported organisms:  Human, Mouse, Yeast, Fly, Fish, Worm



 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32


names = gp.get_library_name(organism =  &#34;Human&#34;)
len(names)
# 217

[name for name in names if &#34;GO&#34; in name]
# [&#39;GO_Biological_Process_2021&#39;,
#  &#39;GO_Biological_Process_2023&#39;,
#  &#39;GO_Biological_Process_2025&#39;,
#  &#39;GO_Cellular_Component_2021&#39;,
#  &#39;GO_Cellular_Component_2023&#39;,
#  &#39;GO_Cellular_Component_2025&#39;,
#  &#39;GO_Molecular_Function_2021&#39;,
#  &#39;GO_Molecular_Function_2023&#39;,
#  &#39;GO_Molecular_Function_2025&#39;,
#  &#39;SynGO_2022&#39;,
#  &#39;SynGO_2024&#39;]
[name for name in names if &#34;KEGG&#34; in name]
# [&#39;KEGG_2013&#39;,
#  &#39;KEGG_2015&#39;,
#  &#39;KEGG_2016&#39;,
#  &#39;KEGG_2019_Human&#39;,
#  &#39;KEGG_2019_Mouse&#39;,
#  &#39;KEGG_2021_Human&#39;]
[name for name in names if &#34;MSigDB&#34; in name]
# [&#39;MSigDB_Computational&#39;, &#39;MSigDB_Hallmark_2020&#39;, &#39;MSigDB_Oncogenic_Signatures&#39;]


kegg = gp.get_library(name=&#39;KEGG_2021_Human&#39;, organism=&#39;Human&#39;)
type(kegg) # dict

kegg[&#39;AMPK signaling pathway&#39;][:4]
# [&#39;RAB2A&#39;, &#39;PPP2R1A&#39;, &#39;TSC2&#39;, &#39;TSC1&#39;]





 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17


# 还支持对MSigDB基因集数据库的获取
from gseapy import Msigdb
msig = Msigdb()

# all available versions
msig.list_dbver() 

# all types of gene sets given a dbver
msig.list_category(dbver=&#34;2025.1.Hs&#34;)

# all gene sets of one type
gmt = msig.get_gmt(category=&#39;h.all&#39;, dbver=&#34;2025.1.Hs&#34;)

type(gmt)   # dict
gmt.keys()  # hallmark pathway names

gmt[&#34;HALLMARK_ADIPOGENESIS&#34;] # gene list



1. ORA富集
1.1 分析

Online analysis



 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47


## 1) 两个最重要的参数 （后同）
# gene_list
gene_list=&#34;tests/data/gene_list.txt&#34;         # file path, one gene per row
gene_list= [&#34;RAB2A&#34;,&#34;PRKAB2&#34;,&#34;PRKAA1&#34;]    # list of genes

# gene_sets (from enrichr API)
gene_sets=&#39;KEGG_2016&#39;
gene_sets=[&#39;KEGG_2021_Human&#39;,&#39;GO_Biological_Process_2025&#39;]


gene_list = pd.read_csv(&#34;tests/data/gene_list.txt&#34;, header=None)[0].tolist()
gene_list[:4]
# [&#39;IGKV4-1&#39;, &#39;CD55&#39;, &#39;IGKC&#39;, &#39;PPFIBP1&#39;]

enr = gp.enrichr(gene_list=gene_list, # or &#34;./tests/data/gene_list.txt&#34;,
                 gene_sets=[&#39;GO_Biological_Process_2025&#39;],
                 organism=&#39;human&#39;,
                 outdir=None, 
                )


enr.results.head(5)
# Index([&#39;Gene_set&#39;, &#39;Term&#39;, &#39;Overlap&#39;, &#39;P-value&#39;, &#39;Adjusted P-value&#39;,
#        &#39;Old P-value&#39;, &#39;Old Adjusted P-value&#39;, &#39;Odds Ratio&#39;, &#39;Combined Score&#39;,
#        &#39;Genes&#39;],
#       dtype=&#39;object&#39;)


## 2) 设置结果导出路径
enr = gp.enrichr(gene_list=gene_list, 
                 gene_sets=[&#39;GO_Biological_Process_2025&#39;],
                 organism=&#39;human&#39;,
                 outdir=&#34;enrichr_kegg&#34;, 
                )
os.listdir(&#34;enrichr_kegg&#34;)
# [&#39;GO_Biological_Process_2025.human.enrichr.reports.txt&#39;,
#  &#39;GO_Biological_Process_2025.human.enrichr.reports.pdf&#39;]



## 3) 自定义Background genes (P值会存在一定差异)
enr = gp.enrichr(gene_list=gene_list, 
                 gene_sets=[&#39;GO_Biological_Process_2025&#39;],
                 organism=&#39;human&#39;,
                 background=&#34;tests/data/background.txt&#34;,  # 自定义背景基因 / or list object
                 outdir=None, 
                )



Offline analysis  (本地提供通路基因集数据)



 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11


from gseapy.parser import read_gmt

gmt_dict = read_gmt(&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;)
gmt_dict.keys()

enr2 = gp.enrich(gene_list=gene_list, # or gene_list=glist
                 gene_sets=[&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;], 
                 background=None, 
                 outdir=None,
                 verbose=True)
enr2.results.head(5)


1.2 可视化


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27


# ofname 设置图片的导出路径，若为None则不保存。也支持后续使用plt语法进一步调整图片

ax = dotplot(enr.results,
              column=&#34;Adjusted P-value&#34;,
              x=&#39;Gene_set&#39;, # set x axis, so you could do a multi-sample/library comparsion
              size=5,
              cutoff=0.05, 
              top_term=10,
              figsize=(3, 5),
              title = &#34;GO_BP&#34;,
              xticklabels_rot=45, # rotate xtick labels
              show_ring=True, # set to False to revmove outer ring
              marker=&#39;o&#39;,
              ofname=&#39;./plot/p1.png&#39;   
             )


ax = barplot(enr.results,
              column=&#34;Adjusted P-value&#34;,
              group=&#39;Gene_set&#39;, # set group, so you could do a multi-sample/library comparsion
              size=10,
              top_term=10,
              figsize=(3,5),
              #color=[&#39;darkred&#39;, &#39;darkblue&#39;] # set colors for group
              color = {&#39;GO_Biological_Process_2025&#39;: &#39;salmon&#39;},
              ofname=&#39;./plot/p2.png&#39;   
             )





1
2
3


ax = dotplot(enr.res2d, title=&#39;Enrichr&#39;,cmap=&#39;viridis_r&#39;, size=10,  figsize=(3,5), ofname=&#39;./plot/p3.png&#39;)

ax = barplot(enr.res2d, title=&#39;Enrichr&#39;,cmap=&#39;viridis_r&#39;, size=4, figsize=(4, 5), color=&#39;darkred&#39;, ofname=&#39;./plot/p4.png&#39;)  # 差别不大



2. Prerank富集
基于差异分析结果的GSEA">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/061gseapy%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/061gseapy%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="gseapy富集分析与可视化" />
<meta property="og:description" content="参考

https://gseapy.readthedocs.io/en/latest/index.html
示例数据来自 https://github.com/zqfang/GSEApy/tree/master/tests/data



1
2
3
4
5
6
7
8


import os
import pandas as pd
import scanpy as sc

import gseapy as gp
from gseapy import barplot, dotplot

import matplotlib.pyplot as plt


0. 基因集获取

Enrichr API supported organisms:  Human, Mouse, Yeast, Fly, Fish, Worm



 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32


names = gp.get_library_name(organism =  &#34;Human&#34;)
len(names)
# 217

[name for name in names if &#34;GO&#34; in name]
# [&#39;GO_Biological_Process_2021&#39;,
#  &#39;GO_Biological_Process_2023&#39;,
#  &#39;GO_Biological_Process_2025&#39;,
#  &#39;GO_Cellular_Component_2021&#39;,
#  &#39;GO_Cellular_Component_2023&#39;,
#  &#39;GO_Cellular_Component_2025&#39;,
#  &#39;GO_Molecular_Function_2021&#39;,
#  &#39;GO_Molecular_Function_2023&#39;,
#  &#39;GO_Molecular_Function_2025&#39;,
#  &#39;SynGO_2022&#39;,
#  &#39;SynGO_2024&#39;]
[name for name in names if &#34;KEGG&#34; in name]
# [&#39;KEGG_2013&#39;,
#  &#39;KEGG_2015&#39;,
#  &#39;KEGG_2016&#39;,
#  &#39;KEGG_2019_Human&#39;,
#  &#39;KEGG_2019_Mouse&#39;,
#  &#39;KEGG_2021_Human&#39;]
[name for name in names if &#34;MSigDB&#34; in name]
# [&#39;MSigDB_Computational&#39;, &#39;MSigDB_Hallmark_2020&#39;, &#39;MSigDB_Oncogenic_Signatures&#39;]


kegg = gp.get_library(name=&#39;KEGG_2021_Human&#39;, organism=&#39;Human&#39;)
type(kegg) # dict

kegg[&#39;AMPK signaling pathway&#39;][:4]
# [&#39;RAB2A&#39;, &#39;PPP2R1A&#39;, &#39;TSC2&#39;, &#39;TSC1&#39;]





 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17


# 还支持对MSigDB基因集数据库的获取
from gseapy import Msigdb
msig = Msigdb()

# all available versions
msig.list_dbver() 

# all types of gene sets given a dbver
msig.list_category(dbver=&#34;2025.1.Hs&#34;)

# all gene sets of one type
gmt = msig.get_gmt(category=&#39;h.all&#39;, dbver=&#34;2025.1.Hs&#34;)

type(gmt)   # dict
gmt.keys()  # hallmark pathway names

gmt[&#34;HALLMARK_ADIPOGENESIS&#34;] # gene list



1. ORA富集
1.1 分析

Online analysis



 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47


## 1) 两个最重要的参数 （后同）
# gene_list
gene_list=&#34;tests/data/gene_list.txt&#34;         # file path, one gene per row
gene_list= [&#34;RAB2A&#34;,&#34;PRKAB2&#34;,&#34;PRKAA1&#34;]    # list of genes

# gene_sets (from enrichr API)
gene_sets=&#39;KEGG_2016&#39;
gene_sets=[&#39;KEGG_2021_Human&#39;,&#39;GO_Biological_Process_2025&#39;]


gene_list = pd.read_csv(&#34;tests/data/gene_list.txt&#34;, header=None)[0].tolist()
gene_list[:4]
# [&#39;IGKV4-1&#39;, &#39;CD55&#39;, &#39;IGKC&#39;, &#39;PPFIBP1&#39;]

enr = gp.enrichr(gene_list=gene_list, # or &#34;./tests/data/gene_list.txt&#34;,
                 gene_sets=[&#39;GO_Biological_Process_2025&#39;],
                 organism=&#39;human&#39;,
                 outdir=None, 
                )


enr.results.head(5)
# Index([&#39;Gene_set&#39;, &#39;Term&#39;, &#39;Overlap&#39;, &#39;P-value&#39;, &#39;Adjusted P-value&#39;,
#        &#39;Old P-value&#39;, &#39;Old Adjusted P-value&#39;, &#39;Odds Ratio&#39;, &#39;Combined Score&#39;,
#        &#39;Genes&#39;],
#       dtype=&#39;object&#39;)


## 2) 设置结果导出路径
enr = gp.enrichr(gene_list=gene_list, 
                 gene_sets=[&#39;GO_Biological_Process_2025&#39;],
                 organism=&#39;human&#39;,
                 outdir=&#34;enrichr_kegg&#34;, 
                )
os.listdir(&#34;enrichr_kegg&#34;)
# [&#39;GO_Biological_Process_2025.human.enrichr.reports.txt&#39;,
#  &#39;GO_Biological_Process_2025.human.enrichr.reports.pdf&#39;]



## 3) 自定义Background genes (P值会存在一定差异)
enr = gp.enrichr(gene_list=gene_list, 
                 gene_sets=[&#39;GO_Biological_Process_2025&#39;],
                 organism=&#39;human&#39;,
                 background=&#34;tests/data/background.txt&#34;,  # 自定义背景基因 / or list object
                 outdir=None, 
                )



Offline analysis  (本地提供通路基因集数据)



 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11


from gseapy.parser import read_gmt

gmt_dict = read_gmt(&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;)
gmt_dict.keys()

enr2 = gp.enrich(gene_list=gene_list, # or gene_list=glist
                 gene_sets=[&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;], 
                 background=None, 
                 outdir=None,
                 verbose=True)
enr2.results.head(5)


1.2 可视化


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27


# ofname 设置图片的导出路径，若为None则不保存。也支持后续使用plt语法进一步调整图片

ax = dotplot(enr.results,
              column=&#34;Adjusted P-value&#34;,
              x=&#39;Gene_set&#39;, # set x axis, so you could do a multi-sample/library comparsion
              size=5,
              cutoff=0.05, 
              top_term=10,
              figsize=(3, 5),
              title = &#34;GO_BP&#34;,
              xticklabels_rot=45, # rotate xtick labels
              show_ring=True, # set to False to revmove outer ring
              marker=&#39;o&#39;,
              ofname=&#39;./plot/p1.png&#39;   
             )


ax = barplot(enr.results,
              column=&#34;Adjusted P-value&#34;,
              group=&#39;Gene_set&#39;, # set group, so you could do a multi-sample/library comparsion
              size=10,
              top_term=10,
              figsize=(3,5),
              #color=[&#39;darkred&#39;, &#39;darkblue&#39;] # set colors for group
              color = {&#39;GO_Biological_Process_2025&#39;: &#39;salmon&#39;},
              ofname=&#39;./plot/p2.png&#39;   
             )





1
2
3


ax = dotplot(enr.res2d, title=&#39;Enrichr&#39;,cmap=&#39;viridis_r&#39;, size=10,  figsize=(3,5), ofname=&#39;./plot/p3.png&#39;)

ax = barplot(enr.res2d, title=&#39;Enrichr&#39;,cmap=&#39;viridis_r&#39;, size=4, figsize=(4, 5), color=&#39;darkred&#39;, ofname=&#39;./plot/p4.png&#39;)  # 差别不大



2. Prerank富集
基于差异分析结果的GSEA" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/061gseapy%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2025-07-22T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2025-07-22T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="gseapy富集分析与可视化"/>
<meta name="twitter:description" content="参考

https://gseapy.readthedocs.io/en/latest/index.html
示例数据来自 https://github.com/zqfang/GSEApy/tree/master/tests/data



1
2
3
4
5
6
7
8


import os
import pandas as pd
import scanpy as sc

import gseapy as gp
from gseapy import barplot, dotplot

import matplotlib.pyplot as plt


0. 基因集获取

Enrichr API supported organisms:  Human, Mouse, Yeast, Fly, Fish, Worm



 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32


names = gp.get_library_name(organism =  &#34;Human&#34;)
len(names)
# 217

[name for name in names if &#34;GO&#34; in name]
# [&#39;GO_Biological_Process_2021&#39;,
#  &#39;GO_Biological_Process_2023&#39;,
#  &#39;GO_Biological_Process_2025&#39;,
#  &#39;GO_Cellular_Component_2021&#39;,
#  &#39;GO_Cellular_Component_2023&#39;,
#  &#39;GO_Cellular_Component_2025&#39;,
#  &#39;GO_Molecular_Function_2021&#39;,
#  &#39;GO_Molecular_Function_2023&#39;,
#  &#39;GO_Molecular_Function_2025&#39;,
#  &#39;SynGO_2022&#39;,
#  &#39;SynGO_2024&#39;]
[name for name in names if &#34;KEGG&#34; in name]
# [&#39;KEGG_2013&#39;,
#  &#39;KEGG_2015&#39;,
#  &#39;KEGG_2016&#39;,
#  &#39;KEGG_2019_Human&#39;,
#  &#39;KEGG_2019_Mouse&#39;,
#  &#39;KEGG_2021_Human&#39;]
[name for name in names if &#34;MSigDB&#34; in name]
# [&#39;MSigDB_Computational&#39;, &#39;MSigDB_Hallmark_2020&#39;, &#39;MSigDB_Oncogenic_Signatures&#39;]


kegg = gp.get_library(name=&#39;KEGG_2021_Human&#39;, organism=&#39;Human&#39;)
type(kegg) # dict

kegg[&#39;AMPK signaling pathway&#39;][:4]
# [&#39;RAB2A&#39;, &#39;PPP2R1A&#39;, &#39;TSC2&#39;, &#39;TSC1&#39;]





 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17


# 还支持对MSigDB基因集数据库的获取
from gseapy import Msigdb
msig = Msigdb()

# all available versions
msig.list_dbver() 

# all types of gene sets given a dbver
msig.list_category(dbver=&#34;2025.1.Hs&#34;)

# all gene sets of one type
gmt = msig.get_gmt(category=&#39;h.all&#39;, dbver=&#34;2025.1.Hs&#34;)

type(gmt)   # dict
gmt.keys()  # hallmark pathway names

gmt[&#34;HALLMARK_ADIPOGENESIS&#34;] # gene list



1. ORA富集
1.1 分析

Online analysis



 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47


## 1) 两个最重要的参数 （后同）
# gene_list
gene_list=&#34;tests/data/gene_list.txt&#34;         # file path, one gene per row
gene_list= [&#34;RAB2A&#34;,&#34;PRKAB2&#34;,&#34;PRKAA1&#34;]    # list of genes

# gene_sets (from enrichr API)
gene_sets=&#39;KEGG_2016&#39;
gene_sets=[&#39;KEGG_2021_Human&#39;,&#39;GO_Biological_Process_2025&#39;]


gene_list = pd.read_csv(&#34;tests/data/gene_list.txt&#34;, header=None)[0].tolist()
gene_list[:4]
# [&#39;IGKV4-1&#39;, &#39;CD55&#39;, &#39;IGKC&#39;, &#39;PPFIBP1&#39;]

enr = gp.enrichr(gene_list=gene_list, # or &#34;./tests/data/gene_list.txt&#34;,
                 gene_sets=[&#39;GO_Biological_Process_2025&#39;],
                 organism=&#39;human&#39;,
                 outdir=None, 
                )


enr.results.head(5)
# Index([&#39;Gene_set&#39;, &#39;Term&#39;, &#39;Overlap&#39;, &#39;P-value&#39;, &#39;Adjusted P-value&#39;,
#        &#39;Old P-value&#39;, &#39;Old Adjusted P-value&#39;, &#39;Odds Ratio&#39;, &#39;Combined Score&#39;,
#        &#39;Genes&#39;],
#       dtype=&#39;object&#39;)


## 2) 设置结果导出路径
enr = gp.enrichr(gene_list=gene_list, 
                 gene_sets=[&#39;GO_Biological_Process_2025&#39;],
                 organism=&#39;human&#39;,
                 outdir=&#34;enrichr_kegg&#34;, 
                )
os.listdir(&#34;enrichr_kegg&#34;)
# [&#39;GO_Biological_Process_2025.human.enrichr.reports.txt&#39;,
#  &#39;GO_Biological_Process_2025.human.enrichr.reports.pdf&#39;]



## 3) 自定义Background genes (P值会存在一定差异)
enr = gp.enrichr(gene_list=gene_list, 
                 gene_sets=[&#39;GO_Biological_Process_2025&#39;],
                 organism=&#39;human&#39;,
                 background=&#34;tests/data/background.txt&#34;,  # 自定义背景基因 / or list object
                 outdir=None, 
                )



Offline analysis  (本地提供通路基因集数据)



 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11


from gseapy.parser import read_gmt

gmt_dict = read_gmt(&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;)
gmt_dict.keys()

enr2 = gp.enrich(gene_list=gene_list, # or gene_list=glist
                 gene_sets=[&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;], 
                 background=None, 
                 outdir=None,
                 verbose=True)
enr2.results.head(5)


1.2 可视化


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27


# ofname 设置图片的导出路径，若为None则不保存。也支持后续使用plt语法进一步调整图片

ax = dotplot(enr.results,
              column=&#34;Adjusted P-value&#34;,
              x=&#39;Gene_set&#39;, # set x axis, so you could do a multi-sample/library comparsion
              size=5,
              cutoff=0.05, 
              top_term=10,
              figsize=(3, 5),
              title = &#34;GO_BP&#34;,
              xticklabels_rot=45, # rotate xtick labels
              show_ring=True, # set to False to revmove outer ring
              marker=&#39;o&#39;,
              ofname=&#39;./plot/p1.png&#39;   
             )


ax = barplot(enr.results,
              column=&#34;Adjusted P-value&#34;,
              group=&#39;Gene_set&#39;, # set group, so you could do a multi-sample/library comparsion
              size=10,
              top_term=10,
              figsize=(3,5),
              #color=[&#39;darkred&#39;, &#39;darkblue&#39;] # set colors for group
              color = {&#39;GO_Biological_Process_2025&#39;: &#39;salmon&#39;},
              ofname=&#39;./plot/p2.png&#39;   
             )





1
2
3


ax = dotplot(enr.res2d, title=&#39;Enrichr&#39;,cmap=&#39;viridis_r&#39;, size=10,  figsize=(3,5), ofname=&#39;./plot/p3.png&#39;)

ax = barplot(enr.res2d, title=&#39;Enrichr&#39;,cmap=&#39;viridis_r&#39;, size=4, figsize=(4, 5), color=&#39;darkred&#39;, ofname=&#39;./plot/p4.png&#39;)  # 差别不大



2. Prerank富集
基于差异分析结果的GSEA"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "gseapy富集分析与可视化",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/061gseapy%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "gseapy富集分析与可视化",
  "name": "gseapy富集分析与可视化",
  "description": "参考\nhttps://gseapy.readthedocs.io/en/latest/index.html 示例数据来自 https://github.com/zqfang/GSEApy/tree/master/tests/data 1 2 3 4 5 6 7 8 import os import pandas as pd import scanpy as sc import gseapy as gp from gseapy import barplot, dotplot import matplotlib.pyplot as plt 0. 基因集获取 Enrichr API supported organisms: Human, Mouse, Yeast, Fly, Fish, Worm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 names = gp.get_library_name(organism = \u0026#34;Human\u0026#34;) len(names) # 217 [name for name in names if \u0026#34;GO\u0026#34; in name] # [\u0026#39;GO_Biological_Process_2021\u0026#39;, # \u0026#39;GO_Biological_Process_2023\u0026#39;, # \u0026#39;GO_Biological_Process_2025\u0026#39;, # \u0026#39;GO_Cellular_Component_2021\u0026#39;, # \u0026#39;GO_Cellular_Component_2023\u0026#39;, # \u0026#39;GO_Cellular_Component_2025\u0026#39;, # \u0026#39;GO_Molecular_Function_2021\u0026#39;, # \u0026#39;GO_Molecular_Function_2023\u0026#39;, # \u0026#39;GO_Molecular_Function_2025\u0026#39;, # \u0026#39;SynGO_2022\u0026#39;, # \u0026#39;SynGO_2024\u0026#39;] [name for name in names if \u0026#34;KEGG\u0026#34; in name] # [\u0026#39;KEGG_2013\u0026#39;, # \u0026#39;KEGG_2015\u0026#39;, # \u0026#39;KEGG_2016\u0026#39;, # \u0026#39;KEGG_2019_Human\u0026#39;, # \u0026#39;KEGG_2019_Mouse\u0026#39;, # \u0026#39;KEGG_2021_Human\u0026#39;] [name for name in names if \u0026#34;MSigDB\u0026#34; in name] # [\u0026#39;MSigDB_Computational\u0026#39;, \u0026#39;MSigDB_Hallmark_2020\u0026#39;, \u0026#39;MSigDB_Oncogenic_Signatures\u0026#39;] kegg = gp.get_library(name=\u0026#39;KEGG_2021_Human\u0026#39;, organism=\u0026#39;Human\u0026#39;) type(kegg) # dict kegg[\u0026#39;AMPK signaling pathway\u0026#39;][:4] # [\u0026#39;RAB2A\u0026#39;, \u0026#39;PPP2R1A\u0026#39;, \u0026#39;TSC2\u0026#39;, \u0026#39;TSC1\u0026#39;] 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 # 还支持对MSigDB基因集数据库的获取 from gseapy import Msigdb msig = Msigdb() # all available versions msig.list_dbver() # all types of gene sets given a dbver msig.list_category(dbver=\u0026#34;2025.1.Hs\u0026#34;) # all gene sets of one type gmt = msig.get_gmt(category=\u0026#39;h.all\u0026#39;, dbver=\u0026#34;2025.1.Hs\u0026#34;) type(gmt) # dict gmt.keys() # hallmark pathway names gmt[\u0026#34;HALLMARK_ADIPOGENESIS\u0026#34;] # gene list 1. ORA富集 1.1 分析 Online analysis 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 ## 1) 两个最重要的参数 （后同） # gene_list gene_list=\u0026#34;tests/data/gene_list.txt\u0026#34; # file path, one gene per row gene_list= [\u0026#34;RAB2A\u0026#34;,\u0026#34;PRKAB2\u0026#34;,\u0026#34;PRKAA1\u0026#34;] # list of genes # gene_sets (from enrichr API) gene_sets=\u0026#39;KEGG_2016\u0026#39; gene_sets=[\u0026#39;KEGG_2021_Human\u0026#39;,\u0026#39;GO_Biological_Process_2025\u0026#39;] gene_list = pd.read_csv(\u0026#34;tests/data/gene_list.txt\u0026#34;, header=None)[0].tolist() gene_list[:4] # [\u0026#39;IGKV4-1\u0026#39;, \u0026#39;CD55\u0026#39;, \u0026#39;IGKC\u0026#39;, \u0026#39;PPFIBP1\u0026#39;] enr = gp.enrichr(gene_list=gene_list, # or \u0026#34;./tests/data/gene_list.txt\u0026#34;, gene_sets=[\u0026#39;GO_Biological_Process_2025\u0026#39;], organism=\u0026#39;human\u0026#39;, outdir=None, ) enr.results.head(5) # Index([\u0026#39;Gene_set\u0026#39;, \u0026#39;Term\u0026#39;, \u0026#39;Overlap\u0026#39;, \u0026#39;P-value\u0026#39;, \u0026#39;Adjusted P-value\u0026#39;, # \u0026#39;Old P-value\u0026#39;, \u0026#39;Old Adjusted P-value\u0026#39;, \u0026#39;Odds Ratio\u0026#39;, \u0026#39;Combined Score\u0026#39;, # \u0026#39;Genes\u0026#39;], # dtype=\u0026#39;object\u0026#39;) ## 2) 设置结果导出路径 enr = gp.enrichr(gene_list=gene_list, gene_sets=[\u0026#39;GO_Biological_Process_2025\u0026#39;], organism=\u0026#39;human\u0026#39;, outdir=\u0026#34;enrichr_kegg\u0026#34;, ) os.listdir(\u0026#34;enrichr_kegg\u0026#34;) # [\u0026#39;GO_Biological_Process_2025.human.enrichr.reports.txt\u0026#39;, # \u0026#39;GO_Biological_Process_2025.human.enrichr.reports.pdf\u0026#39;] ## 3) 自定义Background genes (P值会存在一定差异) enr = gp.enrichr(gene_list=gene_list, gene_sets=[\u0026#39;GO_Biological_Process_2025\u0026#39;], organism=\u0026#39;human\u0026#39;, background=\u0026#34;tests/data/background.txt\u0026#34;, # 自定义背景基因 / or list object outdir=None, ) Offline analysis (本地提供通路基因集数据) 1 2 3 4 5 6 7 8 9 10 11 from gseapy.parser import read_gmt gmt_dict = read_gmt(\u0026#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt\u0026#34;) gmt_dict.keys() enr2 = gp.enrich(gene_list=gene_list, # or gene_list=glist gene_sets=[\u0026#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt\u0026#34;], background=None, outdir=None, verbose=True) enr2.results.head(5) 1.2 可视化 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 # ofname 设置图片的导出路径，若为None则不保存。也支持后续使用plt语法进一步调整图片 ax = dotplot(enr.results, column=\u0026#34;Adjusted P-value\u0026#34;, x=\u0026#39;Gene_set\u0026#39;, # set x axis, so you could do a multi-sample/library comparsion size=5, cutoff=0.05, top_term=10, figsize=(3, 5), title = \u0026#34;GO_BP\u0026#34;, xticklabels_rot=45, # rotate xtick labels show_ring=True, # set to False to revmove outer ring marker=\u0026#39;o\u0026#39;, ofname=\u0026#39;./plot/p1.png\u0026#39; ) ax = barplot(enr.results, column=\u0026#34;Adjusted P-value\u0026#34;, group=\u0026#39;Gene_set\u0026#39;, # set group, so you could do a multi-sample/library comparsion size=10, top_term=10, figsize=(3,5), #color=[\u0026#39;darkred\u0026#39;, \u0026#39;darkblue\u0026#39;] # set colors for group color = {\u0026#39;GO_Biological_Process_2025\u0026#39;: \u0026#39;salmon\u0026#39;}, ofname=\u0026#39;./plot/p2.png\u0026#39; ) 1 2 3 ax = dotplot(enr.res2d, title=\u0026#39;Enrichr\u0026#39;,cmap=\u0026#39;viridis_r\u0026#39;, size=10, figsize=(3,5), ofname=\u0026#39;./plot/p3.png\u0026#39;) ax = barplot(enr.res2d, title=\u0026#39;Enrichr\u0026#39;,cmap=\u0026#39;viridis_r\u0026#39;, size=4, figsize=(4, 5), color=\u0026#39;darkred\u0026#39;, ofname=\u0026#39;./plot/p4.png\u0026#39;) # 差别不大 2. Prerank富集 基于差异分析结果的GSEA\n",
  "keywords": [
    "转录组"
  ],
  "articleBody": "参考\nhttps://gseapy.readthedocs.io/en/latest/index.html 示例数据来自 https://github.com/zqfang/GSEApy/tree/master/tests/data 1 2 3 4 5 6 7 8 import os import pandas as pd import scanpy as sc import gseapy as gp from gseapy import barplot, dotplot import matplotlib.pyplot as plt 0. 基因集获取 Enrichr API supported organisms: Human, Mouse, Yeast, Fly, Fish, Worm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 names = gp.get_library_name(organism = \"Human\") len(names) # 217 [name for name in names if \"GO\" in name] # ['GO_Biological_Process_2021', # 'GO_Biological_Process_2023', # 'GO_Biological_Process_2025', # 'GO_Cellular_Component_2021', # 'GO_Cellular_Component_2023', # 'GO_Cellular_Component_2025', # 'GO_Molecular_Function_2021', # 'GO_Molecular_Function_2023', # 'GO_Molecular_Function_2025', # 'SynGO_2022', # 'SynGO_2024'] [name for name in names if \"KEGG\" in name] # ['KEGG_2013', # 'KEGG_2015', # 'KEGG_2016', # 'KEGG_2019_Human', # 'KEGG_2019_Mouse', # 'KEGG_2021_Human'] [name for name in names if \"MSigDB\" in name] # ['MSigDB_Computational', 'MSigDB_Hallmark_2020', 'MSigDB_Oncogenic_Signatures'] kegg = gp.get_library(name='KEGG_2021_Human', organism='Human') type(kegg) # dict kegg['AMPK signaling pathway'][:4] # ['RAB2A', 'PPP2R1A', 'TSC2', 'TSC1'] 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 # 还支持对MSigDB基因集数据库的获取 from gseapy import Msigdb msig = Msigdb() # all available versions msig.list_dbver() # all types of gene sets given a dbver msig.list_category(dbver=\"2025.1.Hs\") # all gene sets of one type gmt = msig.get_gmt(category='h.all', dbver=\"2025.1.Hs\") type(gmt) # dict gmt.keys() # hallmark pathway names gmt[\"HALLMARK_ADIPOGENESIS\"] # gene list 1. ORA富集 1.1 分析 Online analysis 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 ## 1) 两个最重要的参数 （后同） # gene_list gene_list=\"tests/data/gene_list.txt\" # file path, one gene per row gene_list= [\"RAB2A\",\"PRKAB2\",\"PRKAA1\"] # list of genes # gene_sets (from enrichr API) gene_sets='KEGG_2016' gene_sets=['KEGG_2021_Human','GO_Biological_Process_2025'] gene_list = pd.read_csv(\"tests/data/gene_list.txt\", header=None)[0].tolist() gene_list[:4] # ['IGKV4-1', 'CD55', 'IGKC', 'PPFIBP1'] enr = gp.enrichr(gene_list=gene_list, # or \"./tests/data/gene_list.txt\", gene_sets=['GO_Biological_Process_2025'], organism='human', outdir=None, ) enr.results.head(5) # Index(['Gene_set', 'Term', 'Overlap', 'P-value', 'Adjusted P-value', # 'Old P-value', 'Old Adjusted P-value', 'Odds Ratio', 'Combined Score', # 'Genes'], # dtype='object') ## 2) 设置结果导出路径 enr = gp.enrichr(gene_list=gene_list, gene_sets=['GO_Biological_Process_2025'], organism='human', outdir=\"enrichr_kegg\", ) os.listdir(\"enrichr_kegg\") # ['GO_Biological_Process_2025.human.enrichr.reports.txt', # 'GO_Biological_Process_2025.human.enrichr.reports.pdf'] ## 3) 自定义Background genes (P值会存在一定差异) enr = gp.enrichr(gene_list=gene_list, gene_sets=['GO_Biological_Process_2025'], organism='human', background=\"tests/data/background.txt\", # 自定义背景基因 / or list object outdir=None, ) Offline analysis (本地提供通路基因集数据) 1 2 3 4 5 6 7 8 9 10 11 from gseapy.parser import read_gmt gmt_dict = read_gmt(\"tests/data/c2.cp.kegg.v7.5.1.symbols.gmt\") gmt_dict.keys() enr2 = gp.enrich(gene_list=gene_list, # or gene_list=glist gene_sets=[\"tests/data/c2.cp.kegg.v7.5.1.symbols.gmt\"], background=None, outdir=None, verbose=True) enr2.results.head(5) 1.2 可视化 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 # ofname 设置图片的导出路径，若为None则不保存。也支持后续使用plt语法进一步调整图片 ax = dotplot(enr.results, column=\"Adjusted P-value\", x='Gene_set', # set x axis, so you could do a multi-sample/library comparsion size=5, cutoff=0.05, top_term=10, figsize=(3, 5), title = \"GO_BP\", xticklabels_rot=45, # rotate xtick labels show_ring=True, # set to False to revmove outer ring marker='o', ofname='./plot/p1.png' ) ax = barplot(enr.results, column=\"Adjusted P-value\", group='Gene_set', # set group, so you could do a multi-sample/library comparsion size=10, top_term=10, figsize=(3,5), #color=['darkred', 'darkblue'] # set colors for group color = {'GO_Biological_Process_2025': 'salmon'}, ofname='./plot/p2.png' ) 1 2 3 ax = dotplot(enr.res2d, title='Enrichr',cmap='viridis_r', size=10, figsize=(3,5), ofname='./plot/p3.png') ax = barplot(enr.res2d, title='Enrichr',cmap='viridis_r', size=4, figsize=(4, 5), color='darkred', ofname='./plot/p4.png') # 差别不大 2. Prerank富集 基于差异分析结果的GSEA\n2.1 分析 输入的是降序排列基因 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 # pd.DataFrame: index为基因, 列一般为差异倍数(降序) rnk = pd.read_csv(\"./tests/data/temp.rnk\", header=None, index_col=0, sep=\"\\t\") type(rnk) # pandas.core.frame.DataFrame rnk.shape # (22922, 1) rnk.head() # 1 # 0 # ATXN1 16.456753 # UBQLN4 13.989493 # CALM1 13.745533 # DLG4 12.796588 # MRE11A 12.787631 pre_res = gp.prerank(rnk=rnk, # or \"./tests/data/temp.rnk\" gene_sets=['GO_Biological_Process_2025'], threads=4, min_size=5, max_size=1000, permutation_num=1000, # reduce number to speed up testing outdir=None, # don't write to disk seed=6, verbose=True, # see what's going on behind the scenes ) pre_res.res2d.head(5) # Index(['Name', 'Term', 'ES', 'NES', 'NOM p-val', 'FDR q-val', 'FWER p-val', # 'Tag %', 'Gene %', 'Lead_genes'], # dtype='object') 2.2 可视化 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 terms = pre_res.res2d.Term # Series terms[:3] # 0 GO_Biological_Process_2025__Primary miRNA Proc... # 1 GO_Biological_Process_2025__Cytoplasmic Transl... # 2 GO_Biological_Process_2025__Regulation of Extr... axs = pre_res.plot(terms=terms[1], figsize=(4, 6), ofname=\"./plot/p5.png\") axs = pre_res.plot(terms=terms[1:5], #legend_kws={'loc': (1.2, 0)}, # set the legend loc show_ranking=True, # whether to show the second yaxis figsize=(3,4), ofname=\"./plot/p6.pdf\" ) 1 2 3 4 5 6 7 ax = dotplot(pre_res.res2d, column=\"FDR q-val\", title='Prerank dotplot', cmap=plt.cm.viridis, size=6, # adjust dot size figsize=(4,5), cutoff=0.25, show_ring=False, ofname=\"./plot/p7.pdf\") 3. GSEA富集 基于分组表达矩阵的GSEA分析 (经典)\n3.1 分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 adata = sc.read_h5ad(\"tests/data/ifnb.h5ad\") adata.obs.head() adata.layers['counts'] = adata.X sc.pp.normalize_total(adata, target_sum=1e4) sc.pp.log1p(adata) adata.obs.groupby('seurat_annotations')['stim'].value_counts() # set STIM as class 0, CTRL as class 1, to make categorical adata.obs['stim'] = pd.Categorical(adata.obs['stim'], categories=[\"STIM\", \"CTRL\"], ordered=True) # 排序 indices = adata.obs.sort_values(['seurat_annotations', 'stim']).index adata = adata[indices,:] # 取其中一种细胞类型的两个分组 bdata = adata[adata.obs.seurat_annotations == \"CD14 Mono\"].copy() bdata # AnnData object with n_obs × n_vars = 4362 × 14053 # obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'stim', 'seurat_annotations' # var: 'features' # uns: 'log1p' # layers: 'counts' res = gp.gsea(data=bdata.to_df().T, # row -\u003e genes, column-\u003e samples gene_sets=\"GO_Biological_Process_2025\", cls=bdata.obs.stim, permutation_num=1000, permutation_type='phenotype', outdir=None, method='s2n', # signal_to_noise threads= 16) res.res2d.head(10) # Index(['Name', 'Term', 'ES', 'NES', 'NOM p-val', 'FDR q-val', 'FWER p-val', # 'Tag %', 'Gene %', 'Lead_genes'], # dtype='object') 3.2 可视化 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 i = 7 genes = res.res2d.Lead_genes.iloc[i].split(\";\") ax = gp.heatmap(df = res.heatmat.loc[genes], z_score=None, title=res.res2d.Term.iloc[i], figsize=(6,5), cmap=plt.cm.viridis, xticklabels=False, ofname=\"./plot/p8.png\" ) term = res.res2d.Term # gp.gseaplot(res.ranking, term=term[i], **res.results[term[i]]) axs = res.plot(terms=term[:5], ofname=\"./plot/p9.png\") 4. ssGSEA富集 基于表达矩阵分析每个样本\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 exp_dat = pd.read_csv(\"tests/extdata/Leukemia_hgu95av2.trim.txt\", sep=\"\\t\", index_col=0) exp_dat = exp_dat.drop(columns=[\"NAME\"]) exp_dat.iloc[:4,:4] # ALL_1 ALL_2 ALL_3 ALL_4 # Gene # MAPK3 1633.6 2455.0 866.0 1000.0 # TIE1 284.4 159.0 173.0 216.0 # CYP2C19 285.8 114.0 429.0 -43.0 # CXCR5 -126.6 -388.0 143.0 -915.0 import gseapy as gp # txt, gct file input ss = gp.ssgsea(#data='./tests/extdata/Leukemia_hgu95av2.trim.txt', data=exp_dat, gene_sets='./tests/extdata/h.all.v7.0.symbols.gmt', outdir=None, sample_norm_method='rank', # choose 'custom' will only use the raw value of `data` no_plot=True) ss.res2d.head() # Name Term ES NES # 0 ALL_2 HALLMARK_MYC_TARGETS_V1 3393.823575 0.707975 # 1 ALL_12 HALLMARK_MYC_TARGETS_V1 3385.626111 0.706265 # 2 AML_11 HALLMARK_MYC_TARGETS_V1 3359.186716 0.700749 nes = ss.res2d.pivot(index='Term', columns='Name', values='NES') nes.iloc[:,:3].head() # Name ALL_1 ALL_10 ALL_11 # Term # HALLMARK_ADIPOGENESIS 0.287384 0.274548 0.290059 # HALLMARK_ALLOGRAFT_REJECTION 0.06177 0.028062 0.096589 # HALLMARK_ANDROGEN_RESPONSE 0.133453 0.113911 0.193074 ",
  "wordCount" : "1367",
  "inLanguage": "en",
  "datePublished": "2025-07-22T00:00:00Z",
  "dateModified": "2025-07-22T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/061gseapy%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      gseapy富集分析与可视化
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2025-07-22 00:00:00 &#43;0000 UTC&#39;&gt;2025-07-22&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2025-07-22&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1367&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#0-%e5%9f%ba%e5%9b%a0%e9%9b%86%e8%8e%b7%e5%8f%96" aria-label="0. 基因集获取">0. 基因集获取</a></li>
                    <li>
                        <a href="#1-ora%e5%af%8c%e9%9b%86" aria-label="1. ORA富集">1. ORA富集</a><ul>
                            
                    <li>
                        <a href="#11-%e5%88%86%e6%9e%90" aria-label="1.1 分析">1.1 分析</a></li>
                    <li>
                        <a href="#12-%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="1.2 可视化">1.2 可视化</a></li></ul>
                    </li>
                    <li>
                        <a href="#2-prerank%e5%af%8c%e9%9b%86" aria-label="2. Prerank富集">2. Prerank富集</a><ul>
                            
                    <li>
                        <a href="#21-%e5%88%86%e6%9e%90" aria-label="2.1 分析">2.1 分析</a></li>
                    <li>
                        <a href="#22-%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="2.2 可视化">2.2 可视化</a></li></ul>
                    </li>
                    <li>
                        <a href="#3-gsea%e5%af%8c%e9%9b%86" aria-label="3. GSEA富集">3. GSEA富集</a><ul>
                            
                    <li>
                        <a href="#31-%e5%88%86%e6%9e%90" aria-label="3.1 分析">3.1 分析</a></li>
                    <li>
                        <a href="#32-%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="3.2 可视化">3.2 可视化</a></li></ul>
                    </li>
                    <li>
                        <a href="#4-ssgsea%e5%af%8c%e9%9b%86" aria-label="4. ssGSEA富集">4. ssGSEA富集</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><p>参考</p>
<ul>
<li><a href="https://gseapy.readthedocs.io/en/latest/index.html">https://gseapy.readthedocs.io/en/latest/index.html</a></li>
<li>示例数据来自 <a href="https://github.com/zqfang/GSEApy/tree/master/tests/data">https://github.com/zqfang/GSEApy/tree/master/tests/data</a></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> os
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> pandas <span style="color:#fff;font-weight:bold">as</span> pd
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> scanpy <span style="color:#fff;font-weight:bold">as</span> sc
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> gseapy <span style="color:#fff;font-weight:bold">as</span> gp
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">from</span> gseapy <span style="color:#fff;font-weight:bold">import</span> barplot, dotplot
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> matplotlib.pyplot <span style="color:#fff;font-weight:bold">as</span> plt
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="0-基因集获取">0. 基因集获取<a hidden class="anchor" aria-hidden="true" href="#0-基因集获取">#</a></h1>
<ul>
<li>Enrichr API supported organisms:  Human, Mouse, Yeast, Fly, Fish, Worm</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>names = gp.get_library_name(organism =  <span style="color:#0ff;font-weight:bold">&#34;Human&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">len</span>(names)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 217</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>[name <span style="color:#fff;font-weight:bold">for</span> name in names <span style="color:#fff;font-weight:bold">if</span> <span style="color:#0ff;font-weight:bold">&#34;GO&#34;</span> in name]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;GO_Biological_Process_2021&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;GO_Biological_Process_2023&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;GO_Biological_Process_2025&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;GO_Cellular_Component_2021&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;GO_Cellular_Component_2023&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;GO_Cellular_Component_2025&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;GO_Molecular_Function_2021&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;GO_Molecular_Function_2023&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;GO_Molecular_Function_2025&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;SynGO_2022&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;SynGO_2024&#39;]</span>
</span></span><span style="display:flex;"><span>[name <span style="color:#fff;font-weight:bold">for</span> name in names <span style="color:#fff;font-weight:bold">if</span> <span style="color:#0ff;font-weight:bold">&#34;KEGG&#34;</span> in name]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;KEGG_2013&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;KEGG_2015&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;KEGG_2016&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;KEGG_2019_Human&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;KEGG_2019_Mouse&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;KEGG_2021_Human&#39;]</span>
</span></span><span style="display:flex;"><span>[name <span style="color:#fff;font-weight:bold">for</span> name in names <span style="color:#fff;font-weight:bold">if</span> <span style="color:#0ff;font-weight:bold">&#34;MSigDB&#34;</span> in name]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;MSigDB_Computational&#39;, &#39;MSigDB_Hallmark_2020&#39;, &#39;MSigDB_Oncogenic_Signatures&#39;]</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>kegg = gp.get_library(name=<span style="color:#0ff;font-weight:bold">&#39;KEGG_2021_Human&#39;</span>, organism=<span style="color:#0ff;font-weight:bold">&#39;Human&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">type</span>(kegg) <span style="color:#007f7f"># dict</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>kegg[<span style="color:#0ff;font-weight:bold">&#39;AMPK signaling pathway&#39;</span>][:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;RAB2A&#39;, &#39;PPP2R1A&#39;, &#39;TSC2&#39;, &#39;TSC1&#39;]</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f"># 还支持对MSigDB基因集数据库的获取</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">from</span> gseapy <span style="color:#fff;font-weight:bold">import</span> Msigdb
</span></span><span style="display:flex;"><span>msig = Msigdb()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># all available versions</span>
</span></span><span style="display:flex;"><span>msig.list_dbver() 
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># all types of gene sets given a dbver</span>
</span></span><span style="display:flex;"><span>msig.list_category(dbver=<span style="color:#0ff;font-weight:bold">&#34;2025.1.Hs&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># all gene sets of one type</span>
</span></span><span style="display:flex;"><span>gmt = msig.get_gmt(category=<span style="color:#0ff;font-weight:bold">&#39;h.all&#39;</span>, dbver=<span style="color:#0ff;font-weight:bold">&#34;2025.1.Hs&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">type</span>(gmt)   <span style="color:#007f7f"># dict</span>
</span></span><span style="display:flex;"><span>gmt.keys()  <span style="color:#007f7f"># hallmark pathway names</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gmt[<span style="color:#0ff;font-weight:bold">&#34;HALLMARK_ADIPOGENESIS&#34;</span>] <span style="color:#007f7f"># gene list</span>
</span></span></code></pre></td></tr></table>
</div>
</div></blockquote>
<h1 id="1-ora富集">1. ORA富集<a hidden class="anchor" aria-hidden="true" href="#1-ora富集">#</a></h1>
<h2 id="11-分析">1.1 分析<a hidden class="anchor" aria-hidden="true" href="#11-分析">#</a></h2>
<ul>
<li>Online analysis</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">43
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">44
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">45
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">46
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">47
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f">## 1) 两个最重要的参数 （后同）</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gene_list</span>
</span></span><span style="display:flex;"><span>gene_list=<span style="color:#0ff;font-weight:bold">&#34;tests/data/gene_list.txt&#34;</span>         <span style="color:#007f7f"># file path, one gene per row</span>
</span></span><span style="display:flex;"><span>gene_list= [<span style="color:#0ff;font-weight:bold">&#34;RAB2A&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;PRKAB2&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;PRKAA1&#34;</span>]    <span style="color:#007f7f"># list of genes</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gene_sets (from enrichr API)</span>
</span></span><span style="display:flex;"><span>gene_sets=<span style="color:#0ff;font-weight:bold">&#39;KEGG_2016&#39;</span>
</span></span><span style="display:flex;"><span>gene_sets=[<span style="color:#0ff;font-weight:bold">&#39;KEGG_2021_Human&#39;</span>,<span style="color:#0ff;font-weight:bold">&#39;GO_Biological_Process_2025&#39;</span>]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gene_list = pd.read_csv(<span style="color:#0ff;font-weight:bold">&#34;tests/data/gene_list.txt&#34;</span>, header=<span style="color:#fff;font-weight:bold">None</span>)[<span style="color:#ff0;font-weight:bold">0</span>].tolist()
</span></span><span style="display:flex;"><span>gene_list[:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;IGKV4-1&#39;, &#39;CD55&#39;, &#39;IGKC&#39;, &#39;PPFIBP1&#39;]</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>enr = gp.enrichr(gene_list=gene_list, <span style="color:#007f7f"># or &#34;./tests/data/gene_list.txt&#34;,</span>
</span></span><span style="display:flex;"><span>                 gene_sets=[<span style="color:#0ff;font-weight:bold">&#39;GO_Biological_Process_2025&#39;</span>],
</span></span><span style="display:flex;"><span>                 organism=<span style="color:#0ff;font-weight:bold">&#39;human&#39;</span>,
</span></span><span style="display:flex;"><span>                 outdir=<span style="color:#fff;font-weight:bold">None</span>, 
</span></span><span style="display:flex;"><span>                )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>enr.results.head(<span style="color:#ff0;font-weight:bold">5</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Index([&#39;Gene_set&#39;, &#39;Term&#39;, &#39;Overlap&#39;, &#39;P-value&#39;, &#39;Adjusted P-value&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;Old P-value&#39;, &#39;Old Adjusted P-value&#39;, &#39;Odds Ratio&#39;, &#39;Combined Score&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;Genes&#39;],</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#       dtype=&#39;object&#39;)</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 2) 设置结果导出路径</span>
</span></span><span style="display:flex;"><span>enr = gp.enrichr(gene_list=gene_list, 
</span></span><span style="display:flex;"><span>                 gene_sets=[<span style="color:#0ff;font-weight:bold">&#39;GO_Biological_Process_2025&#39;</span>],
</span></span><span style="display:flex;"><span>                 organism=<span style="color:#0ff;font-weight:bold">&#39;human&#39;</span>,
</span></span><span style="display:flex;"><span>                 outdir=<span style="color:#0ff;font-weight:bold">&#34;enrichr_kegg&#34;</span>, 
</span></span><span style="display:flex;"><span>                )
</span></span><span style="display:flex;"><span>os.listdir(<span style="color:#0ff;font-weight:bold">&#34;enrichr_kegg&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;GO_Biological_Process_2025.human.enrichr.reports.txt&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;GO_Biological_Process_2025.human.enrichr.reports.pdf&#39;]</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 3) 自定义Background genes (P值会存在一定差异)</span>
</span></span><span style="display:flex;"><span>enr = gp.enrichr(gene_list=gene_list, 
</span></span><span style="display:flex;"><span>                 gene_sets=[<span style="color:#0ff;font-weight:bold">&#39;GO_Biological_Process_2025&#39;</span>],
</span></span><span style="display:flex;"><span>                 organism=<span style="color:#0ff;font-weight:bold">&#39;human&#39;</span>,
</span></span><span style="display:flex;"><span>                 background=<span style="color:#0ff;font-weight:bold">&#34;tests/data/background.txt&#34;</span>,  <span style="color:#007f7f"># 自定义背景基因 / or list object</span>
</span></span><span style="display:flex;"><span>                 outdir=<span style="color:#fff;font-weight:bold">None</span>, 
</span></span><span style="display:flex;"><span>                )
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>Offline analysis  (本地提供通路基因集数据)</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">from</span> gseapy.parser <span style="color:#fff;font-weight:bold">import</span> read_gmt
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gmt_dict = read_gmt(<span style="color:#0ff;font-weight:bold">&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;</span>)
</span></span><span style="display:flex;"><span>gmt_dict.keys()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>enr2 = gp.enrich(gene_list=gene_list, <span style="color:#007f7f"># or gene_list=glist</span>
</span></span><span style="display:flex;"><span>                 gene_sets=[<span style="color:#0ff;font-weight:bold">&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;</span>], 
</span></span><span style="display:flex;"><span>                 background=<span style="color:#fff;font-weight:bold">None</span>, 
</span></span><span style="display:flex;"><span>                 outdir=<span style="color:#fff;font-weight:bold">None</span>,
</span></span><span style="display:flex;"><span>                 verbose=<span style="color:#fff;font-weight:bold">True</span>)
</span></span><span style="display:flex;"><span>enr2.results.head(<span style="color:#ff0;font-weight:bold">5</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="12-可视化">1.2 可视化<a hidden class="anchor" aria-hidden="true" href="#12-可视化">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f"># ofname 设置图片的导出路径，若为None则不保存。也支持后续使用plt语法进一步调整图片</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>ax = dotplot(enr.results,
</span></span><span style="display:flex;"><span>              column=<span style="color:#0ff;font-weight:bold">&#34;Adjusted P-value&#34;</span>,
</span></span><span style="display:flex;"><span>              x=<span style="color:#0ff;font-weight:bold">&#39;Gene_set&#39;</span>, <span style="color:#007f7f"># set x axis, so you could do a multi-sample/library comparsion</span>
</span></span><span style="display:flex;"><span>              size=<span style="color:#ff0;font-weight:bold">5</span>,
</span></span><span style="display:flex;"><span>              cutoff=<span style="color:#ff0;font-weight:bold">0.05</span>, 
</span></span><span style="display:flex;"><span>              top_term=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>              figsize=(<span style="color:#ff0;font-weight:bold">3</span>, <span style="color:#ff0;font-weight:bold">5</span>),
</span></span><span style="display:flex;"><span>              title = <span style="color:#0ff;font-weight:bold">&#34;GO_BP&#34;</span>,
</span></span><span style="display:flex;"><span>              xticklabels_rot=<span style="color:#ff0;font-weight:bold">45</span>, <span style="color:#007f7f"># rotate xtick labels</span>
</span></span><span style="display:flex;"><span>              show_ring=<span style="color:#fff;font-weight:bold">True</span>, <span style="color:#007f7f"># set to False to revmove outer ring</span>
</span></span><span style="display:flex;"><span>              marker=<span style="color:#0ff;font-weight:bold">&#39;o&#39;</span>,
</span></span><span style="display:flex;"><span>              ofname=<span style="color:#0ff;font-weight:bold">&#39;./plot/p1.png&#39;</span>   
</span></span><span style="display:flex;"><span>             )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>ax = barplot(enr.results,
</span></span><span style="display:flex;"><span>              column=<span style="color:#0ff;font-weight:bold">&#34;Adjusted P-value&#34;</span>,
</span></span><span style="display:flex;"><span>              group=<span style="color:#0ff;font-weight:bold">&#39;Gene_set&#39;</span>, <span style="color:#007f7f"># set group, so you could do a multi-sample/library comparsion</span>
</span></span><span style="display:flex;"><span>              size=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>              top_term=<span style="color:#ff0;font-weight:bold">10</span>,
</span></span><span style="display:flex;"><span>              figsize=(<span style="color:#ff0;font-weight:bold">3</span>,<span style="color:#ff0;font-weight:bold">5</span>),
</span></span><span style="display:flex;"><span>              <span style="color:#007f7f">#color=[&#39;darkred&#39;, &#39;darkblue&#39;] # set colors for group</span>
</span></span><span style="display:flex;"><span>              color = {<span style="color:#0ff;font-weight:bold">&#39;GO_Biological_Process_2025&#39;</span>: <span style="color:#0ff;font-weight:bold">&#39;salmon&#39;</span>},
</span></span><span style="display:flex;"><span>              ofname=<span style="color:#0ff;font-weight:bold">&#39;./plot/p2.png&#39;</span>   
</span></span><span style="display:flex;"><span>             )
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01image-20250723083652809.png" alt="image-20250723083652809" style="zoom:50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>ax = dotplot(enr.res2d, title=<span style="color:#0ff;font-weight:bold">&#39;Enrichr&#39;</span>,cmap=<span style="color:#0ff;font-weight:bold">&#39;viridis_r&#39;</span>, size=<span style="color:#ff0;font-weight:bold">10</span>,  figsize=(<span style="color:#ff0;font-weight:bold">3</span>,<span style="color:#ff0;font-weight:bold">5</span>), ofname=<span style="color:#0ff;font-weight:bold">&#39;./plot/p3.png&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>ax = barplot(enr.res2d, title=<span style="color:#0ff;font-weight:bold">&#39;Enrichr&#39;</span>,cmap=<span style="color:#0ff;font-weight:bold">&#39;viridis_r&#39;</span>, size=<span style="color:#ff0;font-weight:bold">4</span>, figsize=(<span style="color:#ff0;font-weight:bold">4</span>, <span style="color:#ff0;font-weight:bold">5</span>), color=<span style="color:#0ff;font-weight:bold">&#39;darkred&#39;</span>, ofname=<span style="color:#0ff;font-weight:bold">&#39;./plot/p4.png&#39;</span>)  <span style="color:#007f7f"># 差别不大</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01image-20250723084005190.png" alt="image-20250723084005190" style="zoom:50%;" />
<h1 id="2-prerank富集">2. Prerank富集<a hidden class="anchor" aria-hidden="true" href="#2-prerank富集">#</a></h1>
<p>基于差异分析结果的GSEA</p>
<h2 id="21-分析">2.1 分析<a hidden class="anchor" aria-hidden="true" href="#21-分析">#</a></h2>
<ul>
<li>输入的是降序排列基因</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f"># pd.DataFrame: index为基因, 列一般为差异倍数(降序)</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>rnk = pd.read_csv(<span style="color:#0ff;font-weight:bold">&#34;./tests/data/temp.rnk&#34;</span>, header=<span style="color:#fff;font-weight:bold">None</span>, index_col=<span style="color:#ff0;font-weight:bold">0</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;</span><span style="color:#0ff;font-weight:bold">\t</span><span style="color:#0ff;font-weight:bold">&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">type</span>(rnk)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># pandas.core.frame.DataFrame</span>
</span></span><span style="display:flex;"><span>rnk.shape <span style="color:#007f7f"># (22922, 1)</span>
</span></span><span style="display:flex;"><span>rnk.head()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                 1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 0                </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ATXN1   16.456753</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># UBQLN4  13.989493</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CALM1   13.745533</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DLG4    12.796588</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MRE11A  12.787631</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>pre_res = gp.prerank(rnk=rnk, <span style="color:#007f7f"># or &#34;./tests/data/temp.rnk&#34;</span>
</span></span><span style="display:flex;"><span>                     gene_sets=[<span style="color:#0ff;font-weight:bold">&#39;GO_Biological_Process_2025&#39;</span>],
</span></span><span style="display:flex;"><span>                     threads=<span style="color:#ff0;font-weight:bold">4</span>,
</span></span><span style="display:flex;"><span>                     min_size=<span style="color:#ff0;font-weight:bold">5</span>,
</span></span><span style="display:flex;"><span>                     max_size=<span style="color:#ff0;font-weight:bold">1000</span>,
</span></span><span style="display:flex;"><span>                     permutation_num=<span style="color:#ff0;font-weight:bold">1000</span>, <span style="color:#007f7f"># reduce number to speed up testing</span>
</span></span><span style="display:flex;"><span>                     outdir=<span style="color:#fff;font-weight:bold">None</span>, <span style="color:#007f7f"># don&#39;t write to disk</span>
</span></span><span style="display:flex;"><span>                     seed=<span style="color:#ff0;font-weight:bold">6</span>,
</span></span><span style="display:flex;"><span>                     verbose=<span style="color:#fff;font-weight:bold">True</span>, <span style="color:#007f7f"># see what&#39;s going on behind the scenes</span>
</span></span><span style="display:flex;"><span>                    )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>pre_res.res2d.head(<span style="color:#ff0;font-weight:bold">5</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Index([&#39;Name&#39;, &#39;Term&#39;, &#39;ES&#39;, &#39;NES&#39;, &#39;NOM p-val&#39;, &#39;FDR q-val&#39;, &#39;FWER p-val&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;Tag %&#39;, &#39;Gene %&#39;, &#39;Lead_genes&#39;],</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#       dtype=&#39;object&#39;)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="22-可视化">2.2 可视化<a hidden class="anchor" aria-hidden="true" href="#22-可视化">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>terms = pre_res.res2d.Term  <span style="color:#007f7f"># Series</span>
</span></span><span style="display:flex;"><span>terms[:<span style="color:#ff0;font-weight:bold">3</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 0    GO_Biological_Process_2025__Primary miRNA Proc...</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1    GO_Biological_Process_2025__Cytoplasmic Transl...</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2    GO_Biological_Process_2025__Regulation of Extr...</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>axs = pre_res.plot(terms=terms[<span style="color:#ff0;font-weight:bold">1</span>], figsize=(<span style="color:#ff0;font-weight:bold">4</span>, <span style="color:#ff0;font-weight:bold">6</span>), ofname=<span style="color:#0ff;font-weight:bold">&#34;./plot/p5.png&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>axs = pre_res.plot(terms=terms[<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">5</span>],
</span></span><span style="display:flex;"><span>                   <span style="color:#007f7f">#legend_kws={&#39;loc&#39;: (1.2, 0)}, # set the legend loc</span>
</span></span><span style="display:flex;"><span>                   show_ranking=<span style="color:#fff;font-weight:bold">True</span>, <span style="color:#007f7f"># whether to show the second yaxis</span>
</span></span><span style="display:flex;"><span>                   figsize=(<span style="color:#ff0;font-weight:bold">3</span>,<span style="color:#ff0;font-weight:bold">4</span>),
</span></span><span style="display:flex;"><span>                   ofname=<span style="color:#0ff;font-weight:bold">&#34;./plot/p6.pdf&#34;</span>
</span></span><span style="display:flex;"><span>                  )
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01image-20250723085011851.png" alt="image-20250723085011851" style="zoom: 50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-py" data-lang="py"><span style="display:flex;"><span>ax = dotplot(pre_res.res2d,
</span></span><span style="display:flex;"><span>             column=<span style="color:#0ff;font-weight:bold">&#34;FDR q-val&#34;</span>,
</span></span><span style="display:flex;"><span>             title=<span style="color:#0ff;font-weight:bold">&#39;Prerank dotplot&#39;</span>,
</span></span><span style="display:flex;"><span>             cmap=plt.cm.viridis,
</span></span><span style="display:flex;"><span>             size=<span style="color:#ff0;font-weight:bold">6</span>, <span style="color:#007f7f"># adjust dot size</span>
</span></span><span style="display:flex;"><span>             figsize=(<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">5</span>), cutoff=<span style="color:#ff0;font-weight:bold">0.25</span>, show_ring=<span style="color:#fff;font-weight:bold">False</span>,
</span></span><span style="display:flex;"><span>             ofname=<span style="color:#0ff;font-weight:bold">&#34;./plot/p7.pdf&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01image-20250723085109849.png" alt="image-20250723085109849" style="zoom:50%;" />
<h1 id="3-gsea富集">3. GSEA富集<a hidden class="anchor" aria-hidden="true" href="#3-gsea富集">#</a></h1>
<p>基于分组表达矩阵的GSEA分析 (经典)</p>
<h2 id="31-分析">3.1 分析<a hidden class="anchor" aria-hidden="true" href="#31-分析">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>adata = sc.read_h5ad(<span style="color:#0ff;font-weight:bold">&#34;tests/data/ifnb.h5ad&#34;</span>) 
</span></span><span style="display:flex;"><span>adata.obs.head()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>adata.layers[<span style="color:#0ff;font-weight:bold">&#39;counts&#39;</span>] = adata.X
</span></span><span style="display:flex;"><span>sc.pp.normalize_total(adata, target_sum=<span style="color:#ff0;font-weight:bold">1e4</span>)
</span></span><span style="display:flex;"><span>sc.pp.log1p(adata)
</span></span><span style="display:flex;"><span>adata.obs.groupby(<span style="color:#0ff;font-weight:bold">&#39;seurat_annotations&#39;</span>)[<span style="color:#0ff;font-weight:bold">&#39;stim&#39;</span>].value_counts()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># set STIM as class 0, CTRL as class 1, to make categorical</span>
</span></span><span style="display:flex;"><span>adata.obs[<span style="color:#0ff;font-weight:bold">&#39;stim&#39;</span>] = pd.Categorical(adata.obs[<span style="color:#0ff;font-weight:bold">&#39;stim&#39;</span>], categories=[<span style="color:#0ff;font-weight:bold">&#34;STIM&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;CTRL&#34;</span>], ordered=<span style="color:#fff;font-weight:bold">True</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 排序</span>
</span></span><span style="display:flex;"><span>indices = adata.obs.sort_values([<span style="color:#0ff;font-weight:bold">&#39;seurat_annotations&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;stim&#39;</span>]).index
</span></span><span style="display:flex;"><span>adata = adata[indices,:]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 取其中一种细胞类型的两个分组</span>
</span></span><span style="display:flex;"><span>bdata = adata[adata.obs.seurat_annotations == <span style="color:#0ff;font-weight:bold">&#34;CD14 Mono&#34;</span>].copy()
</span></span><span style="display:flex;"><span>bdata
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AnnData object with n_obs × n_vars = 4362 × 14053</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obs: &#39;orig.ident&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;stim&#39;, &#39;seurat_annotations&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     var: &#39;features&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     uns: &#39;log1p&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     layers: &#39;counts&#39;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>res = gp.gsea(data=bdata.to_df().T, <span style="color:#007f7f"># row -&gt; genes, column-&gt; samples</span>
</span></span><span style="display:flex;"><span>        gene_sets=<span style="color:#0ff;font-weight:bold">&#34;GO_Biological_Process_2025&#34;</span>,
</span></span><span style="display:flex;"><span>        <span style="color:#fff;font-weight:bold">cls</span>=bdata.obs.stim,
</span></span><span style="display:flex;"><span>        permutation_num=<span style="color:#ff0;font-weight:bold">1000</span>,
</span></span><span style="display:flex;"><span>        permutation_type=<span style="color:#0ff;font-weight:bold">&#39;phenotype&#39;</span>,
</span></span><span style="display:flex;"><span>        outdir=<span style="color:#fff;font-weight:bold">None</span>,
</span></span><span style="display:flex;"><span>        method=<span style="color:#0ff;font-weight:bold">&#39;s2n&#39;</span>, <span style="color:#007f7f"># signal_to_noise</span>
</span></span><span style="display:flex;"><span>        threads= <span style="color:#ff0;font-weight:bold">16</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>res.res2d.head(<span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Index([&#39;Name&#39;, &#39;Term&#39;, &#39;ES&#39;, &#39;NES&#39;, &#39;NOM p-val&#39;, &#39;FDR q-val&#39;, &#39;FWER p-val&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;Tag %&#39;, &#39;Gene %&#39;, &#39;Lead_genes&#39;],</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#       dtype=&#39;object&#39;)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="32-可视化">3.2 可视化<a hidden class="anchor" aria-hidden="true" href="#32-可视化">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>i = <span style="color:#ff0;font-weight:bold">7</span>
</span></span><span style="display:flex;"><span>genes = res.res2d.Lead_genes.iloc[i].split(<span style="color:#0ff;font-weight:bold">&#34;;&#34;</span>)
</span></span><span style="display:flex;"><span>ax = gp.heatmap(df = res.heatmat.loc[genes],
</span></span><span style="display:flex;"><span>           z_score=<span style="color:#fff;font-weight:bold">None</span>,
</span></span><span style="display:flex;"><span>           title=res.res2d.Term.iloc[i],
</span></span><span style="display:flex;"><span>           figsize=(<span style="color:#ff0;font-weight:bold">6</span>,<span style="color:#ff0;font-weight:bold">5</span>),
</span></span><span style="display:flex;"><span>           cmap=plt.cm.viridis,
</span></span><span style="display:flex;"><span>           xticklabels=<span style="color:#fff;font-weight:bold">False</span>,
</span></span><span style="display:flex;"><span>           ofname=<span style="color:#0ff;font-weight:bold">&#34;./plot/p8.png&#34;</span>
</span></span><span style="display:flex;"><span>           )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>term = res.res2d.Term
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gp.gseaplot(res.ranking, term=term[i], **res.results[term[i]])</span>
</span></span><span style="display:flex;"><span>axs = res.plot(terms=term[:<span style="color:#ff0;font-weight:bold">5</span>], ofname=<span style="color:#0ff;font-weight:bold">&#34;./plot/p9.png&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01image-20250723085607612.png" alt="image-20250723085607612" style="zoom:50%;" />
<h1 id="4-ssgsea富集">4. ssGSEA富集<a hidden class="anchor" aria-hidden="true" href="#4-ssgsea富集">#</a></h1>
<p>基于表达矩阵分析每个样本</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>exp_dat = pd.read_csv(<span style="color:#0ff;font-weight:bold">&#34;tests/extdata/Leukemia_hgu95av2.trim.txt&#34;</span>, sep=<span style="color:#0ff;font-weight:bold">&#34;</span><span style="color:#0ff;font-weight:bold">\t</span><span style="color:#0ff;font-weight:bold">&#34;</span>, index_col=<span style="color:#ff0;font-weight:bold">0</span>)
</span></span><span style="display:flex;"><span>exp_dat = exp_dat.drop(columns=[<span style="color:#0ff;font-weight:bold">&#34;NAME&#34;</span>])
</span></span><span style="display:flex;"><span>exp_dat.iloc[:<span style="color:#ff0;font-weight:bold">4</span>,:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#           ALL_1   ALL_2  ALL_3   ALL_4</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Gene                                  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MAPK3    1633.6  2455.0  866.0  1000.0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TIE1      284.4   159.0  173.0   216.0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CYP2C19   285.8   114.0  429.0   -43.0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CXCR5    -126.6  -388.0  143.0  -915.0</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> gseapy <span style="color:#fff;font-weight:bold">as</span> gp
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># txt, gct file input</span>
</span></span><span style="display:flex;"><span>ss = gp.ssgsea(<span style="color:#007f7f">#data=&#39;./tests/extdata/Leukemia_hgu95av2.trim.txt&#39;,</span>
</span></span><span style="display:flex;"><span>               data=exp_dat,
</span></span><span style="display:flex;"><span>               gene_sets=<span style="color:#0ff;font-weight:bold">&#39;./tests/extdata/h.all.v7.0.symbols.gmt&#39;</span>,
</span></span><span style="display:flex;"><span>               outdir=<span style="color:#fff;font-weight:bold">None</span>,
</span></span><span style="display:flex;"><span>               sample_norm_method=<span style="color:#0ff;font-weight:bold">&#39;rank&#39;</span>, <span style="color:#007f7f"># choose &#39;custom&#39; will only use the raw value of `data`</span>
</span></span><span style="display:flex;"><span>               no_plot=<span style="color:#fff;font-weight:bold">True</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>ss.res2d.head()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#      Name                     Term           ES       NES</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 0   ALL_2  HALLMARK_MYC_TARGETS_V1  3393.823575  0.707975</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1  ALL_12  HALLMARK_MYC_TARGETS_V1  3385.626111  0.706265</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2  AML_11  HALLMARK_MYC_TARGETS_V1  3359.186716  0.700749</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>nes = ss.res2d.pivot(index=<span style="color:#0ff;font-weight:bold">&#39;Term&#39;</span>, columns=<span style="color:#0ff;font-weight:bold">&#39;Name&#39;</span>, values=<span style="color:#0ff;font-weight:bold">&#39;NES&#39;</span>)
</span></span><span style="display:flex;"><span>nes.iloc[:,:<span style="color:#ff0;font-weight:bold">3</span>].head()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Name                             ALL_1    ALL_10    ALL_11</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Term                                                      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># HALLMARK_ADIPOGENESIS         0.287384  0.274548  0.290059</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># HALLMARK_ALLOGRAFT_REJECTION   0.06177  0.028062  0.096589</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># HALLMARK_ANDROGEN_RESPONSE    0.133453  0.113911  0.193074</span>
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">转录组</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/060%E4%BB%8Ernaseq%E7%9A%84fastq.gz%E6%8F%90%E5%8F%96%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5/">
    <span class="title">« Prev Page</span>
    <br>
    <span>从RNAseq的fastq.gz提取表达矩阵</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/075%E7%94%9F%E4%BF%A1%E7%9B%B8%E5%85%B3%E7%BD%91%E7%AB%99%E6%95%B0%E6%8D%AE%E5%BA%93%E9%9B%86%E9%94%A6/">
    <span class="title">Next Page »</span>
    <br>
    <span>生信相关网站数据库集锦</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
