<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>从RNAseq的fastq.gz提取表达矩阵 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="转录组">
<meta name="description" content="
RNA-seq数据比对流程主要分为三步（1）整理数据；（2）质控；（3）比对。其中每一步都涉及到若干软件的用法，如下简单整理基本的分析流程。

示例数据：GSE158623中6个样本的RNA-seq测序结果(human)，对应SRR12720999~SRR12721004">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/060%E4%BB%8Ernaseq%E7%9A%84fastq.gz%E6%8F%90%E5%8F%96%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/060%E4%BB%8Ernaseq%E7%9A%84fastq.gz%E6%8F%90%E5%8F%96%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="从RNAseq的fastq.gz提取表达矩阵" />
<meta property="og:description" content="
RNA-seq数据比对流程主要分为三步（1）整理数据；（2）质控；（3）比对。其中每一步都涉及到若干软件的用法，如下简单整理基本的分析流程。

示例数据：GSE158623中6个样本的RNA-seq测序结果(human)，对应SRR12720999~SRR12721004" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/060%E4%BB%8Ernaseq%E7%9A%84fastq.gz%E6%8F%90%E5%8F%96%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-01-29T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-01-29T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="从RNAseq的fastq.gz提取表达矩阵"/>
<meta name="twitter:description" content="
RNA-seq数据比对流程主要分为三步（1）整理数据；（2）质控；（3）比对。其中每一步都涉及到若干软件的用法，如下简单整理基本的分析流程。

示例数据：GSE158623中6个样本的RNA-seq测序结果(human)，对应SRR12720999~SRR12721004"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "从RNAseq的fastq.gz提取表达矩阵",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/060%E4%BB%8Ernaseq%E7%9A%84fastq.gz%E6%8F%90%E5%8F%96%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "从RNAseq的fastq.gz提取表达矩阵",
  "name": "从RNAseq的fastq.gz提取表达矩阵",
  "description": " RNA-seq数据比对流程主要分为三步（1）整理数据；（2）质控；（3）比对。其中每一步都涉及到若干软件的用法，如下简单整理基本的分析流程。\n示例数据：GSE158623中6个样本的RNA-seq测序结果(human)，对应SRR12720999~SRR12721004\n",
  "keywords": [
    "转录组"
  ],
  "articleBody": " RNA-seq数据比对流程主要分为三步（1）整理数据；（2）质控；（3）比对。其中每一步都涉及到若干软件的用法，如下简单整理基本的分析流程。\n示例数据：GSE158623中6个样本的RNA-seq测序结果(human)，对应SRR12720999~SRR12721004\nhttps://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158623 https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA665998 1 2 3 4 5 6 7 cat SraAccList.txt SRR12720999 SRR12721000 SRR12721001 SRR12721002 SRR12721003 SRR12721004 0、搭建conda环境 根据需要，搭建两个conda分析环境\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 ### (1) 下载数据 conda activate download #ascp的aspera高速下载 conda install -c hcc aspera-cli #prefech下载sra文件 conda install -c bioconda sra-tools # 基因组下载商店 # conda install refgenie # QC质控 conda install -c bioconda trim-galore multiqc ### (2) 比对分析 conda activate fq_map #不同类型的比对软件 conda install -c bioconda hisat2 conda install -c bioconda star=2.7.1a conda install -c bioconda bwa conda install -c bioconda bowtie2 conda install -c bioconda salmon=1.5.2 # 汇总比对结果 conda install -c bioconda multiqc # 基因组下载商店 # conda install refgenie # sam2bam转换 conda install -c bioconda samtools # 基因表达定量 conda install -c bioconda subread 如上，默认下载软件的最新版本即可，但结合尝试、探索，star与salmon比对软件的版本信息必须与构建对应索引文件的软件版本一致(refgenie构建)。因此安装了上述指定的版本。\n1、下载数据 （1）下载公共测序数据：参考之前笔记，有aspera/prefetch/wget等三种下载方式，如下为aspera的下载代码\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 # 批量生成下载链接 # era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/SRR166/009/SRR1663609/SRR1663609_1.fastq.gz touch ascp.link cat SraAccList.txt | while read id do echo \"era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/${id:0:6}/0${id:0-2}/${id}/${id}_1.fastq.gz\" \u003e\u003e ascp.link echo \"era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/${id:0:6}/0${id:0-2}/${id}/${id}_2.fastq.gz\" \u003e\u003e ascp.link done #ascp高速下载 cat ascp.link |while read sample do ascp -QT -l 300m -P33001 \\ -i ~/miniconda3/envs/download/etc/asperaweb_id_dsa.openssh \\ $sample . done （2）下载基因组相关参考数据：主要包括基因组序列fasta、基因组注释文件gtf、各类比对软件的基因组索引文件。\n参考之前笔记，上述文件均可通过refgenie工具下载。如下为下载代码\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 refgenie init -c ~/refgenie/genome_config.yaml refgenie listr -c ~/refgenie/genome_config.yaml refgenie listr -g hg38 -c ~/refgenie/genome_config.yaml #参考基因组 refgenie pull hg38/fasta -c ~/refgenie/genome_config.yaml refgenie pull hg38_cdna/fasta -c ~/refgenie/genome_config.yaml #参考注释信息 refgenie pull hg38/gencode_gtf -c ~/refgenie/genome_config.yaml #比对软件的索引文件 refgenie pull hg38/bowtie2_index -c ~/refgenie/genome_config.yaml refgenie pull hg38/bwa_index -c ~/refgenie/genome_config.yaml refgenie pull hg38/star_index -c ~/refgenie/genome_config.yaml refgenie pull hg38/hisat2_index -c ~/refgenie/genome_config.yaml refgenie pull hg38_cdna/salmon_index -c ~/refgenie/genome_config.yaml #列出本地已经下载的数据 refgenie list -c ~/refgenie/genome_config.yaml 2、测序数据质控 主要目的是为了过滤测序数据中的低质量reads。\n相关RNAseq质控软件有很多，如下以trim_glaore软件用法为例。\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 # (1)以其中一个作为示例 pare_dir=/home/data/****/mapping fq1=${pare_dir}/raw/ebi/SRR12720999_1.fastq.gz fq2=${pare_dir}/raw/ebi/SRR12720999_2.fastq.gz trim_galore -j 8 -q 25 --phred33 --length 36 \\ -paired -o ${pare_dir}/trim \\ $fq1 $fq2 # (2)批量分析 cat ${pare_dir}/SraAccList.txt | while read id do echo $id trim_galore -j 8 -q 25 --phred33 --length 36 \\ -paired -o ${pare_dir}/trim \\ ${pare_dir}/raw/ebi/${id}_1.fastq.gz \\ ${pare_dir}/raw/ebi/${id}_2.fastq.gz done # (3)查看质控之后的fatsq.gz质量 fastqc $(ls ${pare_dir}/trim/*gz) -o ${pare_dir}/trim/ -t 10 multiqc ./ -n trim_multiqc_report.html 3、各类比对软件 RNAseq比对软件有很多，常见的包括hisat2、star、bowtie2、bwa以及salmon； 命令调用均主要包括参数：参考基因组索引文件、 测序数据、输出结果名，以及线程数 比对结果通常是.sam文件，一般需要转为.bam格式，最后提取出count表达数据； （1）以其中一对fatsq.gz文件为例，总结各个比对软件的用法\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 conda activate fq_map pare_dir=/home/data/****/mapping fq1=${pare_dir}/trim/SRR12720999_1_val_1.fq.gz fq2=${pare_dir}/trim/SRR12720999_2_val_2.fq.gz #hisat2 ref_idx_hisat2=$(refgenie seek hg38/hisat2_index -c ~/refgenie/genome_config.yaml) time hisat2 -t -p 10 -x $ref_idx_hisat2 -1 $fq1 -2 $fq2 -S test.sam #STAR ref_idx_star=$(refgenie seek hg38/star_index -c ~/refgenie/genome_config.yaml) time STAR --genomeDir $ref_idx_star \\ --runThreadN 10 --readFilesCommand zcat \\ --readFilesIn $fq1 $fq2 \\ --outSAMtype SAM --outFileNamePrefix test #Bowtie2 ref_idx_bowtie2=$(refgenie seek hg38/bowtie2_index -c ~/refgenie/genome_config.yaml) time bowtie2 -p 10 -x $ref_idx_bowtie2 -1 $fq1 -2 $fq2 -S test.sam #BWA ref_idx_bwa=$(refgenie seek hg38/bwa_index -c ~/refgenie/genome_config.yaml) time bwa mem -t 10 $ref_idx_bwa $fq1 $fq2 -o test.sam #salmon ref_idx_salmon=$(refgenie seek hg38_cdna/salmon_index -c ~/refgenie/genome_config.yaml) time salmon quant -i $ref_idx_salmon -l A \\ -1 $fq1 -2 $fq2 \\ -p 10 -o test_quant （2）基于全部数据，总结每个比对软件的批量比对全流程用法\n[1] hisat2 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 #1、定义变量 ref_idx_hisat2=$(refgenie seek hg38/hisat2_index -c ~/refgenie/genome_config.yaml) ref_gtf=$(refgenie seek hg38/gencode_gtf -c ~/refgenie/genome_config.yaml) pare_dir=/home/data/****/mapping #2、批量比对 cat ${pare_dir}/SraAccList.txt | while read id do echo $id #首先进行比对，生成sam文件 echo \"Start hisat mapping....\" hisat2 -p 10 -x $ref_idx_hisat2 \\ -1 ${pare_dir}/trim/${id}_1_val_1.fq.gz \\ -2 ${pare_dir}/trim/${id}_2_val_2.fq.gz \\ -S ${pare_dir}/hisat2/${id}.sam #然后sam转bam,同时删除内存较大的bam echo \"Start sam2bam....\" samtools view -S ${pare_dir}/hisat2/${id}.sam \\ -@ 10 -b \u003e ${pare_dir}/hisat2/${id}.bam rm ${pare_dir}/hisat2/${id}.sam done #3、featureCounts提取表达信息 featureCounts -p -T 10 -t exon -g gene_name -a $ref_gtf -o hisat2_exp_counts.txt *.bam #第1列为基因名，第7至最后1列为各个样本的表达信息 #4、各个样本比对率概况 multiqc ./ -n hisat2_multiqc_report.html [2] STAR 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 #1、定义变量 ref_idx_star=$(refgenie seek hg38/star_index -c ~/refgenie/genome_config.yaml) ref_gtf=$(refgenie seek hg38/gencode_gtf -c ~/refgenie/genome_config.yaml) pare_dir=/home/data/****/mapping #2、批量比对 # 尝试换用for循环，本质都一样 for sample in $(cat ${pare_dir}/SraAccList.txt) do echo \"START sample ${sample}\" echo \"Start STAR mapping...\" #首先进行比对，生成sam文件 STAR --genomeDir $ref_idx_star \\ --runThreadN 10 --readFilesCommand zcat \\ --readFilesIn ${pare_dir}/trim/${sample}_1_val_1.fq.gz ${pare_dir}/trim/${sample}_2_val_2.fq.gz \\ --outSAMtype SAM --outFileNamePrefix ${sample} echo \"Start sam2bam...\" #然后sam转bam,同时删除内存较大的bam samtools view -S ${pare_dir}/star/${sample}Aligned.out.sam \\ -@ 10 -b \u003e ${pare_dir}/star/${sample}.bam rm ${pare_dir}/star/${sample}Aligned.out.sam done #3、featureCounts提取表达信息 featureCounts -p -T 10 -t exon -g gene_name -a $ref_gtf -o star_exp_counts.txt *.bam #第1列为基因名，第7至最后1列为各个样本的表达信息 #4、各个样本比对率概况 multiqc ./ -n star_multiqc_report.html [3] Bowtie2 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 #1、定义变量 ref_idx_bowtie2=$(refgenie seek hg38/bowtie2_index -c ~/refgenie/genome_config.yaml) ref_gtf=$(refgenie seek hg38/gencode_gtf -c ~/refgenie/genome_config.yaml) pare_dir=/home/data/****/mapping #2、批量比对 for sample in $(cat ${pare_dir}/SraAccList.txt) do echo \"START sample ${sample}\" echo \"Start Bowtie2 mapping...\" bowtie2 -p 10 -x $ref_idx_bowtie2 \\ -1 ${pare_dir}/trim/${sample}_1_val_1.fq.gz \\ -2 ${pare_dir}/trim/${sample}_2_val_2.fq.gz \\ -S ${pare_dir}/bowtie2/${sample}.sam echo \"Start sam2bam...\" samtools view -S ${pare_dir}/bowtie2/${sample}.sam \\ -@ 10 -b \u003e ${pare_dir}/bowtie2/${sample}.bam rm ${pare_dir}/bowtie2/${sample}.sam done #3、featureCounts提取表达信息 featureCounts -p -T 10 -t exon -g gene_name -a $ref_gtf -o bowtie2_exp_counts.txt *.bam #第1列为基因名，第7至最后1列为各个样本的表达信息 #4、各个样本比对率概况 multiqc ./ -n bowtie2_multiqc_report.html [4] BWA 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 #1、定义变量 ref_idx_bwa=$(refgenie seek hg38/bwa_index -c ~/refgenie/genome_config.yaml) ref_gtf=$(refgenie seek hg38/gencode_gtf -c ~/refgenie/genome_config.yaml) pare_dir=/home/data/****/mapping #2、批量比对 for sample in $(cat ${pare_dir}/SraAccList.txt) do echo \"START sample ${sample}\" echo \"Start BWA mapping...\" bwa mem -t 10 $ref_idx_bwa \\ ${pare_dir}/trim/${sample}_1_val_1.fq.gz \\ ${pare_dir}/trim/${sample}_2_val_2.fq.gz \\ -S -o ${pare_dir}/bwa/${sample}.sam echo \"Start sam2bam...\" samtools view -S ${pare_dir}/bwa/${sample}.sam \\ -@ 10 -b \u003e ${pare_dir}/bwa/${sample}.bam rm ${pare_dir}/bwa/${sample}.sam done #3、featureCounts提取表达信息 featureCounts -p -T 10 -t exon -g gene_name -a $ref_gtf -o bwa_exp_counts.txt *.bam #第1列为基因名，第7至最后1列为各个样本的表达信息 #4、各个样本比对率概况 multiqc ./ -n bwa_multiqc_report.html [5] salmon\nsalmon的比对结果并非sam文件，需要配合相关R包tximport提取表达矩阵\n1 2 3 4 5 6 7 8 9 10 11 12 13 #1、定义变量 refgenie list -g hg38_cdna -c ~/refgenie/genome_config.yaml ref_idx_salmon=$(refgenie seek hg38_cdna/salmon_index -c ~/refgenie/genome_config.yaml) pare_dir=/home/data/****/mapping #2、批量比对 cat ${pare_dir}/SraAccList.txt | while read id do echo $id salmon quant -i $ref_idx_salmon -l A \\ -1 ${pare_dir}/trim/${id}_1_val_1.fq.gz \\ -2 ${pare_dir}/trim/${id}_2_val_2.fq.gz \\ -p 10 -o ${pare_dir}/salmon/${id}_quant done 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 library(tximport) library(readr) library(biomaRt) ensembl = useMart(\"ensembl\",dataset=\"hsapiens_gene_ensembl\") attributes = listAttributes(ensembl) attributes[1:5,] # library(httr) # httr::set_config(config(ssl_verifypeer = 0L)) gene_ids \u003c- getBM(attributes= c(\"hgnc_symbol\",\"ensembl_transcript_id\"), mart= ensembl) gene_ids = gene_ids[!duplicated(gene_ids[,2]),]\tcolnames(gene_ids) = c(\"gene_id\",\"tx_id\") gene_ids = gene_ids[,c(\"tx_id\",\"gene_id\")] files \u003c- list.files(pattern = '*sf',recursive = T, full.names=T) txi \u003c- tximport(files, type = \"salmon\", tx2gene = gene_ids, ignoreTxVersion = T, ignoreAfterBar=T) class(txi) names(txi) head(txi$length) head(txi$counts) srrs = stringr::str_extract(files, \"SRR[:digit:]+\") salmon_expr \u003c- txi$counts salmon_expr \u003c- apply(salmon_expr, 2, as.integer) rownames(salmon_expr) \u003c- rownames(txi$counts) colnames(salmon_expr) \u003c- srrs save(salmon_expr, file=\"./salmon_expr.rda\") ",
  "wordCount" : "2347",
  "inLanguage": "en",
  "datePublished": "2023-01-29T00:00:00Z",
  "dateModified": "2023-01-29T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/060%E4%BB%8Ernaseq%E7%9A%84fastq.gz%E6%8F%90%E5%8F%96%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      从RNAseq的fastq.gz提取表达矩阵
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-29 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-29&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-29&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2347&amp;nbsp;|&amp;nbsp;5 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#0%e6%90%ad%e5%bb%baconda%e7%8e%af%e5%a2%83" aria-label="0、搭建conda环境">0、搭建conda环境</a></li>
                    <li>
                        <a href="#1%e4%b8%8b%e8%bd%bd%e6%95%b0%e6%8d%ae" aria-label="1、下载数据">1、下载数据</a></li>
                    <li>
                        <a href="#2%e6%b5%8b%e5%ba%8f%e6%95%b0%e6%8d%ae%e8%b4%a8%e6%8e%a7" aria-label="2、测序数据质控">2、测序数据质控</a></li>
                    <li>
                        <a href="#3%e5%90%84%e7%b1%bb%e6%af%94%e5%af%b9%e8%bd%af%e4%bb%b6" aria-label="3、各类比对软件">3、各类比对软件</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>RNA-seq数据比对流程主要分为三步（1）整理数据；（2）质控；（3）比对。其中每一步都涉及到若干软件的用法，如下简单整理基本的分析流程。</p>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-c3ca775f4c1f19d6.png" alt="image.png" style="zoom:67%;" /></blockquote>
<p>示例数据：GSE158623中6个样本的RNA-seq测序结果(human)，对应SRR12720999~SRR12721004</p>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158623">https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE158623</a></li>
<li><a href="https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA665998">https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA665998</a></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>cat  SraAccList.txt
</span></span><span style="display:flex;"><span>SRR12720999
</span></span><span style="display:flex;"><span>SRR12721000
</span></span><span style="display:flex;"><span>SRR12721001
</span></span><span style="display:flex;"><span>SRR12721002
</span></span><span style="display:flex;"><span>SRR12721003
</span></span><span style="display:flex;"><span>SRR12721004
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="0搭建conda环境">0、搭建conda环境<a hidden class="anchor" aria-hidden="true" href="#0搭建conda环境">#</a></h2>
<p>根据需要，搭建两个conda分析环境</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">### (1) 下载数据</span>
</span></span><span style="display:flex;"><span>conda activate download
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#ascp的aspera高速下载</span>
</span></span><span style="display:flex;"><span>conda install -c hcc aspera-cli
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#prefech下载sra文件</span>
</span></span><span style="display:flex;"><span>conda install -c bioconda sra-tools 
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 基因组下载商店</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># conda install refgenie</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># QC质控</span>
</span></span><span style="display:flex;"><span>conda install -c bioconda trim-galore multiqc
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">### (2) 比对分析</span>
</span></span><span style="display:flex;"><span>conda activate fq_map
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#不同类型的比对软件</span>
</span></span><span style="display:flex;"><span>conda install -c bioconda hisat2
</span></span><span style="display:flex;"><span>conda install -c bioconda star=<span style="color:#ff0;font-weight:bold">2.7.1</span>a
</span></span><span style="display:flex;"><span>conda install -c bioconda bwa
</span></span><span style="display:flex;"><span>conda install -c bioconda bowtie2
</span></span><span style="display:flex;"><span>conda install -c bioconda salmon=<span style="color:#ff0;font-weight:bold">1.5.2</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 汇总比对结果</span>
</span></span><span style="display:flex;"><span>conda install -c bioconda multiqc
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 基因组下载商店</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># conda install refgenie</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># sam2bam转换</span>
</span></span><span style="display:flex;"><span>conda install -c bioconda samtools
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 基因表达定量</span>
</span></span><span style="display:flex;"><span>conda install -c bioconda subread
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>如上，默认下载软件的最新版本即可，但结合尝试、探索，star与salmon比对软件的版本信息必须与构建对应索引文件的软件版本一致(refgenie构建)。因此安装了上述指定的版本。</p></blockquote>
<h2 id="1下载数据">1、下载数据<a hidden class="anchor" aria-hidden="true" href="#1下载数据">#</a></h2>
<p>（1）下载公共测序数据：参考之前笔记，有aspera/prefetch/wget等三种下载方式，如下为aspera的下载代码</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 批量生成下载链接</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/SRR166/009/SRR1663609/SRR1663609_1.fastq.gz</span>
</span></span><span style="display:flex;"><span>touch ascp.link
</span></span><span style="display:flex;"><span>cat SraAccList.txt | <span style="color:#fff;font-weight:bold">while</span> read id 
</span></span><span style="display:flex;"><span>do
</span></span><span style="display:flex;"><span>echo <span style="color:#0ff;font-weight:bold">&#34;era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/${id:0:6}/0${id:0-2}/${id}/${id}_1.fastq.gz&#34;</span> &gt;&gt; ascp.link
</span></span><span style="display:flex;"><span>echo <span style="color:#0ff;font-weight:bold">&#34;era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/${id:0:6}/0${id:0-2}/${id}/${id}_2.fastq.gz&#34;</span> &gt;&gt; ascp.link
</span></span><span style="display:flex;"><span>done
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#ascp高速下载</span>
</span></span><span style="display:flex;"><span>cat ascp.link |<span style="color:#fff;font-weight:bold">while</span> read sample
</span></span><span style="display:flex;"><span>do
</span></span><span style="display:flex;"><span>ascp -QT -l <span style="color:#ff0;font-weight:bold">300</span>m -P33001  \
</span></span><span style="display:flex;"><span>-i ~/miniconda3/envs/download/etc/asperaweb_id_dsa.openssh   \
</span></span><span style="display:flex;"><span>$sample  .
</span></span><span style="display:flex;"><span>done
</span></span></code></pre></td></tr></table>
</div>
</div><p>（2）下载基因组相关参考数据：主要包括基因组序列fasta、基因组注释文件gtf、各类比对软件的基因组索引文件。</p>
<p>参考之前笔记，上述文件均可通过refgenie工具下载。如下为下载代码</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>refgenie init -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>refgenie listr -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>refgenie listr -g hg38 -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#参考基因组</span>
</span></span><span style="display:flex;"><span>refgenie pull hg38/fasta -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>refgenie pull hg38_cdna/fasta -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#参考注释信息</span>
</span></span><span style="display:flex;"><span>refgenie pull hg38/gencode_gtf -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#比对软件的索引文件</span>
</span></span><span style="display:flex;"><span>refgenie pull hg38/bowtie2_index -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>refgenie pull hg38/bwa_index -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>refgenie pull hg38/star_index -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>refgenie pull hg38/hisat2_index -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>refgenie pull hg38_cdna/salmon_index -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#列出本地已经下载的数据</span>
</span></span><span style="display:flex;"><span>refgenie list -c ~/refgenie/genome_config.yaml
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="2测序数据质控">2、测序数据质控<a hidden class="anchor" aria-hidden="true" href="#2测序数据质控">#</a></h2>
<p>主要目的是为了过滤测序数据中的低质量reads。</p>
<p>相关RNAseq质控软件有很多，如下以trim_glaore软件用法为例。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f"># (1)以其中一个作为示例</span>
</span></span><span style="display:flex;"><span>pare_dir=/home/data/****/mapping
</span></span><span style="display:flex;"><span>fq1=<span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/raw/ebi/SRR12720999_1.fastq.gz
</span></span><span style="display:flex;"><span>fq2=<span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/raw/ebi/SRR12720999_2.fastq.gz
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>trim_galore -j <span style="color:#ff0;font-weight:bold">8</span> -q <span style="color:#ff0;font-weight:bold">25</span> --phred33 --length <span style="color:#ff0;font-weight:bold">36</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-paired -o <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>$fq1 $fq2
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2)批量分析</span>
</span></span><span style="display:flex;"><span>cat <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/SraAccList.txt | <span style="color:#fff;font-weight:bold">while</span> <span style="color:#fff;font-weight:bold">read</span> id
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">do</span> 
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">echo</span> $id
</span></span><span style="display:flex;"><span>trim_galore -j <span style="color:#ff0;font-weight:bold">8</span> -q <span style="color:#ff0;font-weight:bold">25</span> --phred33 --length <span style="color:#ff0;font-weight:bold">36</span> <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-paired -o <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span><span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/raw/ebi/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_1.fastq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span><span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/raw/ebi/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_2.fastq.gz
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (3)查看质控之后的fatsq.gz质量</span>
</span></span><span style="display:flex;"><span>fastqc <span style="color:#fff;font-weight:bold">$(</span>ls <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/*gz<span style="color:#fff;font-weight:bold">)</span> -o <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/ -t <span style="color:#ff0;font-weight:bold">10</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>multiqc ./ -n trim_multiqc_report.html
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-4827a4f32908a156.png" alt=""  />
</p>
<h2 id="3各类比对软件">3、各类比对软件<a hidden class="anchor" aria-hidden="true" href="#3各类比对软件">#</a></h2>
<ul>
<li>RNAseq比对软件有很多，常见的包括hisat2、star、bowtie2、bwa以及salmon；
<ul>
<li>命令调用均主要包括参数：参考基因组索引文件、 测序数据、输出结果名，以及线程数</li>
</ul>
</li>
<li>比对结果通常是<code>.sam</code>文件，一般需要转为<code>.bam</code>格式，最后提取出count表达数据；</li>
</ul>
<p>（1）以其中一对fatsq.gz文件为例，总结各个比对软件的用法</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-3f1e444d51732516.png" alt=""  />
</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>conda activate fq_map
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>pare_dir=/home/data/****/mapping
</span></span><span style="display:flex;"><span>fq1=<span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/SRR12720999_1_val_1.fq.gz
</span></span><span style="display:flex;"><span>fq2=<span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/SRR12720999_2_val_2.fq.gz
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#hisat2</span>
</span></span><span style="display:flex;"><span>ref_idx_hisat2=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/hisat2_index -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">time</span> hisat2 -t -p <span style="color:#ff0;font-weight:bold">10</span> -x $ref_idx_hisat2 -1 $fq1 -2 $fq2 -S test.sam
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#STAR</span>
</span></span><span style="display:flex;"><span>ref_idx_star=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/star_index -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">time</span> STAR --genomeDir $ref_idx_star  <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--runThreadN <span style="color:#ff0;font-weight:bold">10</span> --readFilesCommand zcat <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--readFilesIn $fq1  $fq2 <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>--outSAMtype SAM --outFileNamePrefix <span style="color:#fff;font-weight:bold">test</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#Bowtie2</span>
</span></span><span style="display:flex;"><span>ref_idx_bowtie2=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/bowtie2_index -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">time</span> bowtie2 -p <span style="color:#ff0;font-weight:bold">10</span> -x $ref_idx_bowtie2 -1 $fq1 -2 $fq2 -S test.sam
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#BWA</span>
</span></span><span style="display:flex;"><span>ref_idx_bwa=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/bwa_index -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">time</span> bwa mem -t <span style="color:#ff0;font-weight:bold">10</span> $ref_idx_bwa $fq1 $fq2 -o test.sam
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#salmon</span>
</span></span><span style="display:flex;"><span>ref_idx_salmon=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38_cdna/salmon_index -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">time</span> salmon quant -i $ref_idx_salmon -l A <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-1 $fq1 -2 $fq2 <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-p <span style="color:#ff0;font-weight:bold">10</span> -o test_quant
</span></span></code></pre></td></tr></table>
</div>
</div><p>（2）基于全部数据，总结每个比对软件的批量比对全流程用法</p>
<ul>
<li><strong>[1] hisat2</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f">#1、定义变量</span>
</span></span><span style="display:flex;"><span>ref_idx_hisat2=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/hisat2_index -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span>ref_gtf=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/gencode_gtf -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span>pare_dir=/home/data/****/mapping
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#2、批量比对</span>
</span></span><span style="display:flex;"><span>cat <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/SraAccList.txt | <span style="color:#fff;font-weight:bold">while</span> <span style="color:#fff;font-weight:bold">read</span> id
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">do</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">echo</span> $id
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#首先进行比对，生成sam文件</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;Start hisat mapping....&#34;</span>
</span></span><span style="display:flex;"><span>hisat2 -p <span style="color:#ff0;font-weight:bold">10</span> -x $ref_idx_hisat2 <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-1 <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_1_val_1.fq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-2 <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_2_val_2.fq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-S <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/hisat2/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>.sam
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#然后sam转bam,同时删除内存较大的bam</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;Start sam2bam....&#34;</span>
</span></span><span style="display:flex;"><span>samtools view -S <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/hisat2/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>.sam <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-@ <span style="color:#ff0;font-weight:bold">10</span> -b &gt; <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/hisat2/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>.bam
</span></span><span style="display:flex;"><span>rm <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/hisat2/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>.sam
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#3、featureCounts提取表达信息</span>
</span></span><span style="display:flex;"><span>featureCounts -p -T <span style="color:#ff0;font-weight:bold">10</span>  -t exon -g gene_name -a $ref_gtf -o hisat2_exp_counts.txt *.bam
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#第1列为基因名，第7至最后1列为各个样本的表达信息</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#4、各个样本比对率概况</span>
</span></span><span style="display:flex;"><span>multiqc ./ -n hisat2_multiqc_report.html
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li><strong>[2] STAR</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f">#1、定义变量</span>
</span></span><span style="display:flex;"><span>ref_idx_star=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/star_index -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span>ref_gtf=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/gencode_gtf -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span>pare_dir=/home/data/****/mapping
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#2、批量比对</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 尝试换用for循环，本质都一样</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span> sample in <span style="color:#fff;font-weight:bold">$(</span>cat <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/SraAccList.txt<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">do</span>
</span></span><span style="display:flex;"><span>	<span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;START sample </span><span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span><span style="color:#0ff;font-weight:bold">&#34;</span>
</span></span><span style="display:flex;"><span>	<span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;Start STAR mapping...&#34;</span>
</span></span><span style="display:flex;"><span>	<span style="color:#007f7f">#首先进行比对，生成sam文件</span>
</span></span><span style="display:flex;"><span>	STAR --genomeDir $ref_idx_star <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	--runThreadN <span style="color:#ff0;font-weight:bold">10</span> --readFilesCommand zcat <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	--readFilesIn <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>_1_val_1.fq.gz <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>_2_val_2.fq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	--outSAMtype SAM 	--outFileNamePrefix <span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>
</span></span><span style="display:flex;"><span>	<span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;Start sam2bam...&#34;</span>
</span></span><span style="display:flex;"><span>	<span style="color:#007f7f">#然后sam转bam,同时删除内存较大的bam</span>
</span></span><span style="display:flex;"><span>	samtools view -S <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/star/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>Aligned.out.sam <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	-@ <span style="color:#ff0;font-weight:bold">10</span> -b &gt; <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/star/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>.bam
</span></span><span style="display:flex;"><span>	rm <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/star/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>Aligned.out.sam
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#3、featureCounts提取表达信息</span>
</span></span><span style="display:flex;"><span>featureCounts -p -T <span style="color:#ff0;font-weight:bold">10</span>  -t exon -g gene_name -a $ref_gtf -o star_exp_counts.txt *.bam
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#第1列为基因名，第7至最后1列为各个样本的表达信息</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#4、各个样本比对率概况</span>
</span></span><span style="display:flex;"><span>multiqc ./ -n star_multiqc_report.html
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li><strong>[3] Bowtie2</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f">#1、定义变量</span>
</span></span><span style="display:flex;"><span>ref_idx_bowtie2=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/bowtie2_index -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span>ref_gtf=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/gencode_gtf -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span>pare_dir=/home/data/****/mapping
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#2、批量比对</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span> sample in <span style="color:#fff;font-weight:bold">$(</span>cat <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/SraAccList.txt<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">do</span>
</span></span><span style="display:flex;"><span>	<span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;START sample </span><span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span><span style="color:#0ff;font-weight:bold">&#34;</span>
</span></span><span style="display:flex;"><span>	<span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;Start Bowtie2 mapping...&#34;</span>
</span></span><span style="display:flex;"><span>	bowtie2 -p <span style="color:#ff0;font-weight:bold">10</span> -x $ref_idx_bowtie2 <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	-1 <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>_1_val_1.fq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	-2 <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>_2_val_2.fq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	-S <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/bowtie2/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>.sam
</span></span><span style="display:flex;"><span>	
</span></span><span style="display:flex;"><span>	<span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;Start sam2bam...&#34;</span>
</span></span><span style="display:flex;"><span>	samtools view -S <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/bowtie2/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>.sam <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	-@ <span style="color:#ff0;font-weight:bold">10</span> -b &gt; <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/bowtie2/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>.bam
</span></span><span style="display:flex;"><span>	rm <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/bowtie2/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>.sam
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#3、featureCounts提取表达信息</span>
</span></span><span style="display:flex;"><span>featureCounts -p -T <span style="color:#ff0;font-weight:bold">10</span>  -t exon -g gene_name -a $ref_gtf -o bowtie2_exp_counts.txt *.bam
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#第1列为基因名，第7至最后1列为各个样本的表达信息</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#4、各个样本比对率概况</span>
</span></span><span style="display:flex;"><span>multiqc ./ -n bowtie2_multiqc_report.html
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li><strong>[4] BWA</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f">#1、定义变量</span>
</span></span><span style="display:flex;"><span>ref_idx_bwa=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/bwa_index -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span>ref_gtf=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38/gencode_gtf -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span>pare_dir=/home/data/****/mapping
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#2、批量比对</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span> sample in <span style="color:#fff;font-weight:bold">$(</span>cat <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/SraAccList.txt<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">do</span>
</span></span><span style="display:flex;"><span>	<span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;START sample </span><span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span><span style="color:#0ff;font-weight:bold">&#34;</span>
</span></span><span style="display:flex;"><span>	<span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;Start BWA mapping...&#34;</span>
</span></span><span style="display:flex;"><span>	bwa mem -t <span style="color:#ff0;font-weight:bold">10</span> $ref_idx_bwa <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	<span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>_1_val_1.fq.gz  <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	<span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>_2_val_2.fq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	-S -o <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/bwa/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>.sam
</span></span><span style="display:flex;"><span>	
</span></span><span style="display:flex;"><span>	<span style="color:#fff;font-weight:bold">echo</span> <span style="color:#0ff;font-weight:bold">&#34;Start sam2bam...&#34;</span>
</span></span><span style="display:flex;"><span>	samtools view -S <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/bwa/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>.sam <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>	-@ <span style="color:#ff0;font-weight:bold">10</span> -b &gt; <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/bwa/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>.bam
</span></span><span style="display:flex;"><span>	rm <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/bwa/<span style="color:#0ff;font-weight:bold">${</span>sample<span style="color:#0ff;font-weight:bold">}</span>.sam
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#3、featureCounts提取表达信息</span>
</span></span><span style="display:flex;"><span>featureCounts -p -T <span style="color:#ff0;font-weight:bold">10</span>  -t exon -g gene_name -a $ref_gtf -o bwa_exp_counts.txt *.bam
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#第1列为基因名，第7至最后1列为各个样本的表达信息</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#4、各个样本比对率概况</span>
</span></span><span style="display:flex;"><span>multiqc ./ -n bwa_multiqc_report.html
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>
<p><strong>[5] salmon</strong></p>
<p>salmon的比对结果并非sam文件，需要配合相关R包tximport提取表达矩阵</p>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f">#1、定义变量</span>
</span></span><span style="display:flex;"><span>refgenie list -g hg38_cdna -c ~/refgenie/genome_config.yaml
</span></span><span style="display:flex;"><span>ref_idx_salmon=<span style="color:#fff;font-weight:bold">$(</span>refgenie seek hg38_cdna/salmon_index -c ~/refgenie/genome_config.yaml<span style="color:#fff;font-weight:bold">)</span>
</span></span><span style="display:flex;"><span>pare_dir=/home/data/****/mapping
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#2、批量比对</span>
</span></span><span style="display:flex;"><span>cat <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/SraAccList.txt | <span style="color:#fff;font-weight:bold">while</span> <span style="color:#fff;font-weight:bold">read</span> id
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">do</span> 
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">echo</span> $id
</span></span><span style="display:flex;"><span>salmon quant -i $ref_idx_salmon -l A <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-1 <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_1_val_1.fq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-2 <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/trim/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_2_val_2.fq.gz <span style="color:#0ff;font-weight:bold">\
</span></span></span><span style="display:flex;"><span><span style="color:#0ff;font-weight:bold"></span>-p <span style="color:#ff0;font-weight:bold">10</span> -o <span style="color:#0ff;font-weight:bold">${</span>pare_dir<span style="color:#0ff;font-weight:bold">}</span>/salmon/<span style="color:#0ff;font-weight:bold">${</span>id<span style="color:#0ff;font-weight:bold">}</span>_quant
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">done</span>
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(tximport)
</span></span><span style="display:flex;"><span>library(readr)
</span></span><span style="display:flex;"><span>library(biomaRt)
</span></span><span style="display:flex;"><span>ensembl = useMart(<span style="color:#0ff;font-weight:bold">&#34;ensembl&#34;</span>,dataset=<span style="color:#0ff;font-weight:bold">&#34;hsapiens_gene_ensembl&#34;</span>)
</span></span><span style="display:flex;"><span>attributes = listAttributes(ensembl)
</span></span><span style="display:flex;"><span>attributes[1:<span style="color:#ff0;font-weight:bold">5</span>,]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># library(httr)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># httr::set_config(config(ssl_verifypeer = 0L))</span>
</span></span><span style="display:flex;"><span>gene_ids &lt;- getBM(attributes= c(<span style="color:#0ff;font-weight:bold">&#34;hgnc_symbol&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;ensembl_transcript_id&#34;</span>), 
</span></span><span style="display:flex;"><span>                   mart= ensembl)
</span></span><span style="display:flex;"><span>gene_ids = gene_ids[!duplicated(gene_ids[,<span style="color:#ff0;font-weight:bold">2</span>]),]	
</span></span><span style="display:flex;"><span>colnames(gene_ids) = c(<span style="color:#0ff;font-weight:bold">&#34;gene_id&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;tx_id&#34;</span>)
</span></span><span style="display:flex;"><span>gene_ids = gene_ids[,c(<span style="color:#0ff;font-weight:bold">&#34;tx_id&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;gene_id&#34;</span>)]
</span></span><span style="display:flex;"><span>files &lt;- list.files(pattern = <span style="color:#0ff;font-weight:bold">&#39;*sf&#39;</span>,recursive = <span style="color:#fff;font-weight:bold">T</span>, full.names=<span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>txi &lt;- tximport(files, type = <span style="color:#0ff;font-weight:bold">&#34;salmon&#34;</span>, tx2gene = gene_ids, ignoreTxVersion = <span style="color:#fff;font-weight:bold">T</span>, ignoreAfterBar=<span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>class(txi)
</span></span><span style="display:flex;"><span>names(txi)
</span></span><span style="display:flex;"><span>head(txi$length)
</span></span><span style="display:flex;"><span>head(txi$counts)
</span></span><span style="display:flex;"><span>srrs = stringr::str_extract(files, <span style="color:#0ff;font-weight:bold">&#34;SRR[:digit:]+&#34;</span>)
</span></span><span style="display:flex;"><span>salmon_expr &lt;- txi$counts
</span></span><span style="display:flex;"><span>salmon_expr &lt;- apply(salmon_expr, <span style="color:#ff0;font-weight:bold">2</span>, as.integer)
</span></span><span style="display:flex;"><span>rownames(salmon_expr) &lt;- rownames(txi$counts)
</span></span><span style="display:flex;"><span>colnames(salmon_expr) &lt;- srrs
</span></span><span style="display:flex;"><span>save(salmon_expr,  file=<span style="color:#0ff;font-weight:bold">&#34;./salmon_expr.rda&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">转录组</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/074music%E5%8C%85%E6%A0%B9%E6%8D%AEscrnaseq%E9%A2%84%E6%B5%8Bbulk%E7%BB%86%E8%83%9E%E7%BB%84%E6%88%90/">
    <span class="title">« Prev Page</span>
    <br>
    <span>MuSiC包根据scRNAseq预测Bulk细胞组成</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/061gseapy%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/">
    <span class="title">Next Page »</span>
    <br>
    <span>gseapy富集分析与可视化</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
