<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--DoubletFinder识别双细胞 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
R包DoubletFinder可用于检测基于Droplet单细胞测序技术(10X)的双细胞。如下简单学习识别原理以用法。

原始论文：https://doi.org/10.1016/j.cels.2019.03.003
官方手册：https://github.com/chris-mcginnis-ucsf/DoubletFinder
视频教程：https://www.youtube.com/watch?v=NqvAS4HgmrE



1
2


remotes::install_github(&#39;chris-mcginnis-ucsf/DoubletFinder&#39;)
library(DoubletFinder)



Bug in Seurat v4: could not find function &ldquo;paramSweep_v3&rdquo;
https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/184">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/027%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--doubletfinder%E8%AF%86%E5%88%AB%E5%8F%8C%E7%BB%86%E8%83%9E/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/027%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--doubletfinder%E8%AF%86%E5%88%AB%E5%8F%8C%E7%BB%86%E8%83%9E/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--DoubletFinder识别双细胞" />
<meta property="og:description" content="
R包DoubletFinder可用于检测基于Droplet单细胞测序技术(10X)的双细胞。如下简单学习识别原理以用法。

原始论文：https://doi.org/10.1016/j.cels.2019.03.003
官方手册：https://github.com/chris-mcginnis-ucsf/DoubletFinder
视频教程：https://www.youtube.com/watch?v=NqvAS4HgmrE



1
2


remotes::install_github(&#39;chris-mcginnis-ucsf/DoubletFinder&#39;)
library(DoubletFinder)



Bug in Seurat v4: could not find function &ldquo;paramSweep_v3&rdquo;
https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/184" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/027%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--doubletfinder%E8%AF%86%E5%88%AB%E5%8F%8C%E7%BB%86%E8%83%9E/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-01-28T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-01-28T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--DoubletFinder识别双细胞"/>
<meta name="twitter:description" content="
R包DoubletFinder可用于检测基于Droplet单细胞测序技术(10X)的双细胞。如下简单学习识别原理以用法。

原始论文：https://doi.org/10.1016/j.cels.2019.03.003
官方手册：https://github.com/chris-mcginnis-ucsf/DoubletFinder
视频教程：https://www.youtube.com/watch?v=NqvAS4HgmrE



1
2


remotes::install_github(&#39;chris-mcginnis-ucsf/DoubletFinder&#39;)
library(DoubletFinder)



Bug in Seurat v4: could not find function &ldquo;paramSweep_v3&rdquo;
https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/184"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--DoubletFinder识别双细胞",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/027%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--doubletfinder%E8%AF%86%E5%88%AB%E5%8F%8C%E7%BB%86%E8%83%9E/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--DoubletFinder识别双细胞",
  "name": "单细胞分析工具--DoubletFinder识别双细胞",
  "description": " R包DoubletFinder可用于检测基于Droplet单细胞测序技术(10X)的双细胞。如下简单学习识别原理以用法。\n原始论文：https://doi.org/10.1016/j.cels.2019.03.003 官方手册：https://github.com/chris-mcginnis-ucsf/DoubletFinder 视频教程：https://www.youtube.com/watch?v=NqvAS4HgmrE 1 2 remotes::install_github(\u0026#39;chris-mcginnis-ucsf/DoubletFinder\u0026#39;) library(DoubletFinder) Bug in Seurat v4: could not find function \u0026ldquo;paramSweep_v3\u0026rdquo; https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/184\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " R包DoubletFinder可用于检测基于Droplet单细胞测序技术(10X)的双细胞。如下简单学习识别原理以用法。\n原始论文：https://doi.org/10.1016/j.cels.2019.03.003 官方手册：https://github.com/chris-mcginnis-ucsf/DoubletFinder 视频教程：https://www.youtube.com/watch?v=NqvAS4HgmrE 1 2 remotes::install_github('chris-mcginnis-ucsf/DoubletFinder') library(DoubletFinder) Bug in Seurat v4: could not find function “paramSweep_v3” https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/184\n1 remotes::install_github('https://github.com/ekernf01/DoubletFinder', force = T) 1、原理简介 如下示意图可分为4个步骤 （1）根据随机两个单细胞的平均表达水平模拟出若干artificial doublet；\n（2）从模拟双细胞的角度，计算出这些模拟双细胞的最近邻细胞群；\n（3）从真实单细胞数据角度，计算每个细胞与模拟双细胞群的距离\"pANN\";\n（4）最后通过期望的双细胞比例，设置\"pANN\"阈值，划分出双细胞群。\n在上述计算过程中涉及3个重要参数 （1）pN：模拟出的artificial doublet数量。不同取值对识别结果影响不大，默认为0.25。\n（2）pK：计算每个双细胞的最近邻居细胞数量，不同取值对识别结果影响很大，需结合数据选最佳值。\n（3）nExp：期望的双细胞群比例。如下图所示(10X v3.1)，不同测序数据体量的预期占比不同。\n例如对于10000个细胞的单细胞表达矩阵， 当pN=0.25、pK=0.1、nExp=0.08，表示如下含义：\n（1）随机模拟出3333个双细胞(总共13333)；（2）计算每个模拟双细胞的1333个最近邻细胞；\n（3）最终期望识别出800个双细胞。\n双细胞群根据组成细胞的转录相似性可分为如下两大类\n（1）heterotypic doublets ：doublets formed from transcriptionally-distinct cell states – but is insensitive to homotypic doublets\n（2）homotypic doublets ：doublets formed from transcriptionally-similar cell states\n作者认为上述识别算法适合于检测heterotypic doublets，因此在计算过程中有必要考虑homotypic doublets所占比例的影响。\n2、计算过程 示例数据：10X官方提供的PBMC8K测序数据 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 # https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc8k untar(\"pbmc8k_filtered_gene_bc_matrices.tar.gz\") counts = Read10X(\"filtered_gene_bc_matrices/GRCh38/\") sce = CreateSeuratObject(counts) dim(sce) # [1] 33694 8381 ## (1) 预处理：质控过滤 + 降维分群 library(Seurat) library(tidyverse) sce = PercentageFeatureSet(sce, \"^MT-\", col.name = \"percent_mito\") sce = sce %\u003e% subset(., nFeature_RNA \u003e 500) %\u003e% subset(., nCount_RNA \u003e 800) %\u003e% subset(., percent_mito \u003c 10) sce = sce %\u003e% NormalizeData() %\u003e% FindVariableFeatures() %\u003e% ScaleData() sce = sce %\u003e% RunPCA() %\u003e% RunUMAP(dims = 1:30) %\u003e% #RunTSNE FindNeighbors(dims = 1:30) %\u003e% FindClusters(resolution = 0.1) table(sce$seurat_clusters) # 0 1 2 3 4 5 6 7 # 2330 2039 1230 1142 970 339 213 108 DimPlot(sce, reduction = \"umap\", group.by = \"orig.ident\") 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 ## (2) DoubletFinder分析 library(DoubletFinder) # 首先获得最佳的pK值 sweep.res.list \u003c- paramSweep_v3(sce, PCs = 1:30, sct = FALSE) sweep.stats \u003c- summarizeSweep(sweep.res.list, GT = FALSE) bcmvn \u003c- find.pK(sweep.stats) pk_best = bcmvn %\u003e% dplyr::arrange(desc(BCmetric)) %\u003e% dplyr::pull(pK) %\u003e% .[1] %\u003e% as.character() %\u003e% as.numeric() # [1] 0.28 ggplot(bcmvn, aes(x=pK, y=BCmetric, group=1)) + geom_point() + geom_line() 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 # 然后估算出双细胞群中，homotypic doublets的比例(optional) annotations \u003c- sce$seurat_clusters homotypic.prop \u003c- modelHomotypic(annotations) # [1] 0.1928876 # 根据10X V3.1,对于8K+的单细胞数据，期望的双细胞占比为7%左右 nExp_poi \u003c- round(0.07*nrow(sce@meta.data)) #586 nExp_poi.adj \u003c- round(nExp_poi*(1-homotypic.prop)) #473 # 最后根据上述参数(pN=0.25, pK=0.28, nExp=0.07), 识别出双细胞 sce \u003c- doubletFinder_v3(sce, PCs = 1:30, pN = 0.25, pK = pk_best, nExp = nExp_poi, # nExp = nExp_poi.adj, reuse.pANN = FALSE, sct = FALSE) head(sce@meta.data[,c(\"pANN_0.25_0.28_586\", \"DF.classifications_0.25_0.28_586\")]) # pANN_0.25_0.28_586 DF.classifications_0.25_0.28_586 # AAACCTGAGCATCATC-1 0.21824 Singlet # AAACCTGAGCTAACTC-1 0.35488 Singlet # AAACCTGAGCTAGTGG-1 0.18624 Singlet # AAACCTGCACATTAGC-1 0.10816 Singlet # AAACCTGCACTGTTAG-1 0.34528 Singlet # AAACCTGCATAGTAAG-1 0.35904 Singlet DimPlot(sce, reduction = \"umap\", group.by = \"DF.classifications_0.25_0.28_586\") VlnPlot(sce, group.by = \"DF.classifications_0.25_0.28_586\", features = c(\"nCount_RNA\", \"nFeature_RNA\"), pt.size = 0, ncol = 2) ",
  "wordCount" : "1340",
  "inLanguage": "en",
  "datePublished": "2023-01-28T00:00:00Z",
  "dateModified": "2023-01-28T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/027%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--doubletfinder%E8%AF%86%E5%88%AB%E5%8F%8C%E7%BB%86%E8%83%9E/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--DoubletFinder识别双细胞
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-28 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-28&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-28&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1340&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e5%8e%9f%e7%90%86%e7%ae%80%e4%bb%8b" aria-label="1、原理简介">1、原理简介</a></li>
                    <li>
                        <a href="#2%e8%ae%a1%e7%ae%97%e8%bf%87%e7%a8%8b" aria-label="2、计算过程">2、计算过程</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>R包DoubletFinder可用于检测基于Droplet单细胞测序技术(10X)的双细胞。如下简单学习识别原理以用法。</p>
<ul>
<li>原始论文：https://doi.org/10.1016/j.cels.2019.03.003</li>
<li>官方手册：https://github.com/chris-mcginnis-ucsf/DoubletFinder</li>
<li>视频教程：https://www.youtube.com/watch?v=NqvAS4HgmrE</li>
</ul></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>remotes::install_github(<span style="color:#0ff;font-weight:bold">&#39;chris-mcginnis-ucsf/DoubletFinder&#39;</span>)
</span></span><span style="display:flex;"><span>library(DoubletFinder)
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>Bug in Seurat v4: could not find function &ldquo;paramSweep_v3&rdquo;
<a href="https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/184">https://github.com/chris-mcginnis-ucsf/DoubletFinder/issues/184</a></p></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>remotes::install_github(<span style="color:#0ff;font-weight:bold">&#39;https://github.com/ekernf01/DoubletFinder&#39;</span>, force = <span style="color:#fff;font-weight:bold">T</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="1原理简介">1、原理简介<a hidden class="anchor" aria-hidden="true" href="#1原理简介">#</a></h2>
<ul>
<li>如下示意图可分为4个步骤</li>
</ul>
<p>（1）根据随机两个单细胞的平均表达水平模拟出若干artificial doublet；</p>
<p>（2）从模拟双细胞的角度，计算出这些模拟双细胞的最近邻细胞群；</p>
<p>（3）从真实单细胞数据角度，计算每个细胞与模拟双细胞群的距离&quot;pANN&quot;;</p>
<p>（4）最后通过期望的双细胞比例，设置&quot;pANN&quot;阈值，划分出双细胞群。</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128150550760.png" alt="image-20230128150550760"  />
</p>
<ul>
<li>在上述计算过程中涉及3个重要参数</li>
</ul>
<p>（1）pN：模拟出的artificial doublet数量。不同取值对识别结果影响不大，默认为0.25。</p>
<p>（2）pK：计算每个双细胞的最近邻居细胞数量，不同取值对识别结果影响很大，需结合数据选最佳值。</p>
<p>（3）nExp：期望的双细胞群比例。如下图所示(10X v3.1)，不同测序数据体量的预期占比不同。</p>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128152528719.png" alt="image-20230128152528719" style="zoom:80%;" />
<blockquote>
<p>例如对于10000个细胞的单细胞表达矩阵， 当pN=0.25、pK=0.1、nExp=0.08，表示如下含义：</p>
<p>（1）随机模拟出3333个双细胞(总共13333)；（2）计算每个模拟双细胞的1333个最近邻细胞；</p>
<p>（3）最终期望识别出800个双细胞。</p></blockquote>
<ul>
<li>
<p>双细胞群根据组成细胞的转录相似性可分为如下两大类</p>
<p>（1）<strong>heterotypic doublets</strong> ：doublets formed from transcriptionally-distinct cell states &ndash; but is insensitive to homotypic doublets</p>
<p>（2）<strong>homotypic doublets</strong> ：doublets formed from transcriptionally-similar cell states</p>
</li>
</ul>
<p>作者认为上述识别算法适合于检测heterotypic doublets，因此在计算过程中有必要考虑homotypic doublets所占比例的影响。</p>
<h2 id="2计算过程">2、计算过程<a hidden class="anchor" aria-hidden="true" href="#2计算过程">#</a></h2>
<ul>
<li>示例数据：10X官方提供的PBMC8K测序数据</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc8k</span>
</span></span><span style="display:flex;"><span>untar(<span style="color:#0ff;font-weight:bold">&#34;pbmc8k_filtered_gene_bc_matrices.tar.gz&#34;</span>)
</span></span><span style="display:flex;"><span>counts = Read10X(<span style="color:#0ff;font-weight:bold">&#34;filtered_gene_bc_matrices/GRCh38/&#34;</span>)
</span></span><span style="display:flex;"><span>sce = CreateSeuratObject(counts)
</span></span><span style="display:flex;"><span>dim(sce)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 33694  8381</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (1) 预处理：质控过滤 + 降维分群</span>
</span></span><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sce = PercentageFeatureSet(sce, <span style="color:#0ff;font-weight:bold">&#34;^MT-&#34;</span>, col.name = <span style="color:#0ff;font-weight:bold">&#34;percent_mito&#34;</span>)
</span></span><span style="display:flex;"><span>sce = sce %&gt;% 
</span></span><span style="display:flex;"><span>  subset(., nFeature_RNA &gt; <span style="color:#ff0;font-weight:bold">500</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  subset(., nCount_RNA &gt; <span style="color:#ff0;font-weight:bold">800</span>) %&gt;%   
</span></span><span style="display:flex;"><span>  subset(., percent_mito &lt; <span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span>sce = sce %&gt;%
</span></span><span style="display:flex;"><span>  NormalizeData() %&gt;%
</span></span><span style="display:flex;"><span>  FindVariableFeatures() %&gt;%
</span></span><span style="display:flex;"><span>  ScaleData()
</span></span><span style="display:flex;"><span>sce = sce %&gt;% 
</span></span><span style="display:flex;"><span>  RunPCA() %&gt;% 
</span></span><span style="display:flex;"><span>  RunUMAP(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;%  <span style="color:#007f7f">#RunTSNE</span>
</span></span><span style="display:flex;"><span>  FindNeighbors(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  FindClusters(resolution = <span style="color:#ff0;font-weight:bold">0.1</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>table(sce$seurat_clusters)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#    0    1    2    3    4    5    6    7 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2330 2039 1230 1142  970  339  213  108</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>DimPlot(sce, reduction = <span style="color:#0ff;font-weight:bold">&#34;umap&#34;</span>, group.by = <span style="color:#0ff;font-weight:bold">&#34;orig.ident&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128160039740.png" alt="image-20230128160039740" style="zoom: 50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## (2) DoubletFinder分析</span>
</span></span><span style="display:flex;"><span>library(DoubletFinder)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 首先获得最佳的pK值</span>
</span></span><span style="display:flex;"><span>sweep.res.list &lt;- paramSweep_v3(sce, PCs = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>, sct = <span style="color:#fff;font-weight:bold">FALSE</span>)
</span></span><span style="display:flex;"><span>sweep.stats &lt;- summarizeSweep(sweep.res.list, GT = <span style="color:#fff;font-weight:bold">FALSE</span>)
</span></span><span style="display:flex;"><span>bcmvn &lt;- find.pK(sweep.stats)
</span></span><span style="display:flex;"><span>pk_best = bcmvn %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::arrange(desc(BCmetric)) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::pull(pK) %&gt;% 
</span></span><span style="display:flex;"><span>  .[1] %&gt;% as.character() %&gt;% as.numeric()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 0.28</span>
</span></span><span style="display:flex;"><span>ggplot(bcmvn, aes(x=pK, y=BCmetric, group=<span style="color:#ff0;font-weight:bold">1</span>)) + 
</span></span><span style="display:flex;"><span>  geom_point() +
</span></span><span style="display:flex;"><span>  geom_line()
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128160711436.png" alt="image-20230128160711436" style="zoom:50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 然后估算出双细胞群中，homotypic doublets的比例(optional)</span>
</span></span><span style="display:flex;"><span>annotations &lt;- sce$seurat_clusters
</span></span><span style="display:flex;"><span>homotypic.prop &lt;- modelHomotypic(annotations)  
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 0.1928876</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 根据10X V3.1,对于8K+的单细胞数据，期望的双细胞占比为7%左右</span>
</span></span><span style="display:flex;"><span>nExp_poi &lt;- round(<span style="color:#ff0;font-weight:bold">0.07</span>*nrow(sce@meta.data))        <span style="color:#007f7f">#586</span>
</span></span><span style="display:flex;"><span>nExp_poi.adj &lt;- round(nExp_poi*(<span style="color:#ff0;font-weight:bold">1</span>-homotypic.prop)) <span style="color:#007f7f">#473</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 最后根据上述参数(pN=0.25, pK=0.28, nExp=0.07), 识别出双细胞</span>
</span></span><span style="display:flex;"><span>sce &lt;- doubletFinder_v3(sce, PCs = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>, 
</span></span><span style="display:flex;"><span>                        pN = <span style="color:#ff0;font-weight:bold">0.25</span>, pK = pk_best, nExp = nExp_poi, <span style="color:#007f7f"># nExp = nExp_poi.adj, </span>
</span></span><span style="display:flex;"><span>                        reuse.pANN = <span style="color:#fff;font-weight:bold">FALSE</span>, sct = <span style="color:#fff;font-weight:bold">FALSE</span>)
</span></span><span style="display:flex;"><span>head(sce@meta.data[,c(<span style="color:#0ff;font-weight:bold">&#34;pANN_0.25_0.28_586&#34;</span>,
</span></span><span style="display:flex;"><span>                      <span style="color:#0ff;font-weight:bold">&#34;DF.classifications_0.25_0.28_586&#34;</span>)])
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                    pANN_0.25_0.28_586 DF.classifications_0.25_0.28_586</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCTGAGCATCATC-1            0.21824                          Singlet</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCTGAGCTAACTC-1            0.35488                          Singlet</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCTGAGCTAGTGG-1            0.18624                          Singlet</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCTGCACATTAGC-1            0.10816                          Singlet</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCTGCACTGTTAG-1            0.34528                          Singlet</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCTGCATAGTAAG-1            0.35904                          Singlet</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>DimPlot(sce, reduction = <span style="color:#0ff;font-weight:bold">&#34;umap&#34;</span>, 
</span></span><span style="display:flex;"><span>        group.by = <span style="color:#0ff;font-weight:bold">&#34;DF.classifications_0.25_0.28_586&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>VlnPlot(sce, group.by = <span style="color:#0ff;font-weight:bold">&#34;DF.classifications_0.25_0.28_586&#34;</span>, 
</span></span><span style="display:flex;"><span>        features = c(<span style="color:#0ff;font-weight:bold">&#34;nCount_RNA&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;nFeature_RNA&#34;</span>), 
</span></span><span style="display:flex;"><span>        pt.size = <span style="color:#ff0;font-weight:bold">0</span>, ncol = <span style="color:#ff0;font-weight:bold">2</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128161303842.png" alt="image-20230128161303842" style="zoom: 50%;" />
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230128161330809.png" alt="image-20230128161330809" style="zoom: 50%;" />

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/program/022python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6-pandas%E8%A1%A8%E6%A0%BC/">
    <span class="title">« Prev Page</span>
    <br>
    <span>Python数据科学-(3)Pandas表格</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/program/023python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6-matplotlib%E5%8F%AF%E8%A7%86%E5%8C%96/">
    <span class="title">Next Page »</span>
    <br>
    <span>Python数据科学-(4)Matplotlib可视化</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
