<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>使用Seurat包分析空间转录组数据 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
单细胞数据分析常用的Seurat包也集成了空间转录组的分析流程。
https://satijalab.org/seurat/articles/spatial_vignette.html。
空间转录组可以简单理解为在2D组织切面的多个采样spot进行测序，同时记录下每个spot在切片中的2维坐标位置。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/015%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/015%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="使用Seurat包分析空间转录组数据" />
<meta property="og:description" content="
单细胞数据分析常用的Seurat包也集成了空间转录组的分析流程。
https://satijalab.org/seurat/articles/spatial_vignette.html。
空间转录组可以简单理解为在2D组织切面的多个采样spot进行测序，同时记录下每个spot在切片中的2维坐标位置。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/015%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-06-05T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2022-06-05T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="使用Seurat包分析空间转录组数据"/>
<meta name="twitter:description" content="
单细胞数据分析常用的Seurat包也集成了空间转录组的分析流程。
https://satijalab.org/seurat/articles/spatial_vignette.html。
空间转录组可以简单理解为在2D组织切面的多个采样spot进行测序，同时记录下每个spot在切片中的2维坐标位置。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "使用Seurat包分析空间转录组数据",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/015%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "使用Seurat包分析空间转录组数据",
  "name": "使用Seurat包分析空间转录组数据",
  "description": " 单细胞数据分析常用的Seurat包也集成了空间转录组的分析流程。\nhttps://satijalab.org/seurat/articles/spatial_vignette.html。\n空间转录组可以简单理解为在2D组织切面的多个采样spot进行测序，同时记录下每个spot在切片中的2维坐标位置。\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " 单细胞数据分析常用的Seurat包也集成了空间转录组的分析流程。\nhttps://satijalab.org/seurat/articles/spatial_vignette.html。\n空间转录组可以简单理解为在2D组织切面的多个采样spot进行测序，同时记录下每个spot在切片中的2维坐标位置。\n在Seurat流程数据分析时可以将一个spot视为一个细胞（实际上还没有达到单细胞分辨率），然后基本可按照一般单细胞数据分析流程。\n0、数据准备 来自10X官方提供的小鼠脑部空间转录组测序数据，分为anterior与posterior两个切片。\n以其中一张切片所包含的数据为例，可分为两部分数据。\n（1）每个spot的基因表达测序数据，基本类似单细胞测序结果格式可以是三元组或者h5格式等\n1 2 3 4 5 6 7 8 list.files(\"posterior/\", recursive = TRUE) # [1] \"filtered_feature_bc_matrix.h5\" # [2] \"filtered_feature_bc_matrix/barcodes.tsv.gz\" # [3] \"filtered_feature_bc_matrix/features.tsv.gz\" # [4] \"filtered_feature_bc_matrix/matrix.mtx.gz\" # [5] \"spatial/scalefactors_json.json\" # [6] \"spatial/tissue_lowres_image.png\" # [7] \"spatial/tissue_positions_list.csv\" （2）每个spot在组织切片中的位置信息\n首先是组织切片图片(tissue_lowres_image.png)，需要查看确认一下图片的分辨率。如下示例图片为600×600 然后是每个spot在实际测序过程中的位置信息。(tissue_positions_list.csv) 第一列：spot的名字，对应于测序数据的spot ID 第二列：1/0，表示是否为组织测序区域 第三、四列：每个spot的相对位置坐标 第五、六列：每个spot在组织图片中的绝对位置（像素单位） 1 2 3 4 5 6 7 spot_meta = read.csv(\"posterior/spatial/tissue_positions_list.csv\", header = FALSE) head(spot_meta) #tissue row col imagerow imagecol # V1 V2 V3 V4 V5 V6 # 1 ACGCCTGACACGCGCT-1 0 0 0 1419 1432 # 2 TACCGATCCAACACTT-1 0 1 1 1538 1501 # 3 ATTAAAGCGGACGAGC-1 0 0 2 1419 1570 最后是一个json文件(scalefactors_json.json)，用于记录tissue_positions_list.csv所记录的spot绝对坐标与tissue_lowres_image.png图片的比例关系。以及spot的直径大小。 例如 (1419, 1432)*0.0516=(73.2204, 73.8912)即为在tissue_lowres_image.png中的坐标位置。 1 2 3 4 { \"fiducial_diameter_fullres\": 144.59793566029023, \"tissue_lowres_scalef\": 0.051635113 } https://github.com/satijalab/seurat/issues/4993\nThere are four scaling factors that 10X provides, but we generally only use the lowres factor. If you’re using your own, non-10X data, simply create a fake scalefactors object with scalefactors, making sure that you use the correct factor for lowres. If you coordinates match 1:1 with your image, then your lowres scale factor would be 1\n1、创建Seurat对象 思路：先根据spot基因表达数据创建Seurat对象，然后加入组织图片以及spot的位置信息。\n1.1 分步创建 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 ##(1)构建传统seurat对象 sce = Read10X(\"posterior/filtered_feature_bc_matrix/\") %\u003e% CreateSeuratObject(assay = \"Spot\") #将assay名设置为Spatial # sce \u003c- Read10X_h5(\"posterior/filtered_feature_bc_matrix.h5\") %\u003e% # CreateSeuratObject(assay = \"Spatial\") dim(sce) # [1] 32285 3355 # 3355个spot的32285个基因的表达 sce@assays$Spot@counts[1:4,1:4] # 4 x 4 sparse Matrix of class \"dgCMatrix\" # AAACAAGTATCTCCCA-1 AAACACCAATAACTGC-1 AAACAGAGCGACTCCT-1 AAACAGCTTTCAGAAG-1 # Xkr4 . . . . # Gm1992 . . . . # Gm19938 . 1 . . # Gm37381 . . . . head(sce@meta.data) # orig.ident nCount_Spot nFeature_Spot # AAACAAGTATCTCCCA-1 SeuratProject 9195 3089 # AAACACCAATAACTGC-1 SeuratProject 33655 6468 # AAACAGAGCGACTCCT-1 SeuratProject 19619 5245 ##(2)加入image信息 image \u003c- Read10X_Image(image.dir = \"posterior/spatial/\") image@image %\u003e% dim() # [1] 600 600 3 #image \u003c- image[Cells(x = sce)] image@coordinates %\u003e% head() # tissue row col imagerow imagecol # TGGGACCATTGGGAGT-1 1 7 5 2257 1776 # CCGGTGCGAGTGATAG-1 1 7 7 2257 1914 # TAGCCAGAGGGTCCGG-1 1 7 9 2257 2052 image@scale.factors #储存json信息 DefaultAssay(image) \u003c- \"Spot\" DefaultAssay(image) sce[[\"posterior\"]] \u003c- image sce # An object of class Seurat # 32285 features across 3355 samples within 1 assay # Active assay: Spot (32285 features, 0 variable features) # 1 image present: posterior 1.2 一步创建 Load10X_Spatial()函数支持在准备好相应的文件基础上，一步创建出包含image信息的seurat对象\n测序表达数据的h5文件\n1 # [1] \"filtered_feature_bc_matrix.h5\" 图片及位置信息的spatial文件夹\n1 2 list.files(\"posterior/spatial/\") # [1] \"scalefactors_json.json\" \"tissue_lowres_image.png\" \"tissue_positions_list.csv\" 1 2 3 4 5 sce = Load10X_Spatial(data.dir = \"./posterior/\",slice = \"posterior\") # An object of class Seurat # 32285 features across 3355 samples within 1 assay # Active assay: Spatial (32285 features, 0 variable features) # 1 image present: posterior 2、分析流程与可视化 之后的分析可基本参考Seurat单细胞数据分析流程，完成spot的过滤、标准化、聚类分群、差异分析等；就不过多介绍了。\n主要会在可视化方面有空间转录组自身独特的展示方式。\n2.1 常规分析流程 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 ##标准化 sce \u003c- SCTransform(sce,assay = \"Spot\") ##降维 sce \u003c- RunPCA(sce) sce \u003c- RunUMAP(sce, dims = 1:30, label = T) ##聚类分群 sce \u003c- FindNeighbors(sce,dims = 1:20) sce \u003c- FindClusters(sce,resolution = 0.1) table(sce$seurat_clusters) # 0 1 2 3 4 5 6 # 1161 659 483 396 235 230 191 ##差异分析 dif \u003c- FindAllMarkers(sce, assay = \"Spot\", only.pos = T) sig.dif \u003c- dif%\u003e%group_by(cluster) %\u003e% top_n(n = 5,wt = avg_log2FC) genes \u003c- unique(sig.dif$gene) 2.2 可视化 这里主要记录空间转录组自身独特的展示方式：以测序组织图片为背景，展示spot的聚类信息或者特定基因在每个spot的表达信息。\n1 2 3 4 5 6 7 SpatialDimPlot(sce, alpha = 0.5) SpatialDimPlot(sce, cells.highlight = CellsByIdentities(sce,idents = \"1\")) SpatialDimPlot(sce,cells.highlight = CellsByIdentities(sce,idents = c(\"0\",\"2\")), facet.highlight = T) SpatialDimPlot(sce,cells.highlight = CellsByIdentities(sce,idents = c(\"0\",\"2\",\"3\")), facet.highlight = T) 1 LinkedDimPlot(sce,reduction = \"umap\") 1 2 SpatialFeaturePlot(sce,features = genes[1:2],interactive = F) SpatialFeaturePlot(sce,features = genes[1],pt.size=1.5,alpha=c(0.1,1),interactive = T) 1 LinkedFeaturePlot(sce, feature = \"Stx1a\") 3、合并多个切片分析 类似合并多个单细胞测序样本分析，此时每个Seurat对象都要有自己的image信息。必要时也可以使用批次校正等工具。\n1 2 3 4 5 6 7 8 9 10 11 anterior1 \u003c- Load10X_Spatial(data.dir = \"./Anterior\",slice = \"anterior\") posterior1 \u003c- Load10X_Spatial(data.dir = \"./posterior/\",slice = \"posterior\") dim(anterior1) dim(posterior1) st \u003c- merge(anterior1,posterior1,add.cell.ids = c(\"ante\",\"post\")) #st \u003c- merge(a,y=c(b,c,d),add.cell.ids = c(\"a\",\"b\",\"c\",\"d\"),project = \"four_merged\") SpatialDimPlot(st,label = T,label.size = 3, images = \"anterior\") SpatialDimPlot(st,label = T,label.size = 3, images = \"posterior\") SpatialDimPlot(st,label = T,label.size = 3, images = c(\"anterior\",\"posterior\")) ",
  "wordCount" : "1831",
  "inLanguage": "en",
  "datePublished": "2022-06-05T00:00:00Z",
  "dateModified": "2022-06-05T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/015%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      使用Seurat包分析空间转录组数据
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2022-06-05 00:00:00 &#43;0000 UTC&#39;&gt;2022-06-05&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-06-05&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1831&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#0%e6%95%b0%e6%8d%ae%e5%87%86%e5%a4%87" aria-label="0、数据准备">0、数据准备</a></li>
                    <li>
                        <a href="#1%e5%88%9b%e5%bb%baseurat%e5%af%b9%e8%b1%a1" aria-label="1、创建Seurat对象">1、创建Seurat对象</a><ul>
                            
                    <li>
                        <a href="#11-%e5%88%86%e6%ad%a5%e5%88%9b%e5%bb%ba" aria-label="1.1 分步创建">1.1 分步创建</a></li>
                    <li>
                        <a href="#12-%e4%b8%80%e6%ad%a5%e5%88%9b%e5%bb%ba" aria-label="1.2 一步创建">1.2 一步创建</a></li></ul>
                    </li>
                    <li>
                        <a href="#2%e5%88%86%e6%9e%90%e6%b5%81%e7%a8%8b%e4%b8%8e%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="2、分析流程与可视化">2、分析流程与可视化</a><ul>
                            
                    <li>
                        <a href="#21-%e5%b8%b8%e8%a7%84%e5%88%86%e6%9e%90%e6%b5%81%e7%a8%8b" aria-label="2.1 常规分析流程">2.1 常规分析流程</a></li>
                    <li>
                        <a href="#22-%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="2.2 可视化">2.2 可视化</a></li></ul>
                    </li>
                    <li>
                        <a href="#3%e5%90%88%e5%b9%b6%e5%a4%9a%e4%b8%aa%e5%88%87%e7%89%87%e5%88%86%e6%9e%90" aria-label="3、合并多个切片分析">3、合并多个切片分析</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>单细胞数据分析常用的Seurat包也集成了空间转录组的分析流程。</p>
<p><a href="https://satijalab.org/seurat/articles/spatial_vignette.html">https://satijalab.org/seurat/articles/spatial_vignette.html</a>。</p>
<p>空间转录组可以简单理解为在2D组织切面的多个采样spot进行测序，同时记录下每个spot在切片中的2维坐标位置。</p>
<p>在Seurat流程数据分析时可以将一个spot视为一个细胞（实际上还没有达到单细胞分辨率），然后基本可按照一般单细胞数据分析流程。</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/v2-dc7627df0cf3b3277ec872549ee735c9_b.jpg" alt="探索基因表达的空间信息—— 空间转录组学（Spatial Transcriptomics） - 知乎"  />
</p></blockquote>
<h1 id="0数据准备">0、数据准备<a hidden class="anchor" aria-hidden="true" href="#0数据准备">#</a></h1>
<p>来自10X官方提供的小鼠脑部空间转录组测序数据，分为anterior与posterior两个切片。</p>
<p>以其中一张切片所包含的数据为例，可分为两部分数据。</p>
<p>（1）每个spot的基因表达测序数据，基本类似单细胞测序结果格式可以是三元组或者h5格式等</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>list.files(<span style="color:#0ff;font-weight:bold">&#34;posterior/&#34;</span>, recursive = <span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;filtered_feature_bc_matrix.h5&#34;             </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [2] &#34;filtered_feature_bc_matrix/barcodes.tsv.gz&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [3] &#34;filtered_feature_bc_matrix/features.tsv.gz&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [4] &#34;filtered_feature_bc_matrix/matrix.mtx.gz&#34;  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;spatial/scalefactors_json.json&#34;            </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [6] &#34;spatial/tissue_lowres_image.png&#34;           </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [7] &#34;spatial/tissue_positions_list.csv&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>（2）每个spot在组织切片中的位置信息</p>
<ul>
<li>首先是组织切片图片(tissue_lowres_image.png)，需要查看确认一下图片的分辨率。如下示例图片为600×600</li>
<li>然后是每个spot在实际测序过程中的位置信息。(tissue_positions_list.csv)
<ul>
<li>第一列：spot的名字，对应于测序数据的spot ID</li>
<li>第二列：1/0，表示是否为组织测序区域</li>
<li>第三、四列：每个spot的相对位置坐标</li>
<li>第五、六列：每个spot在组织图片中的绝对位置（像素单位）</li>
</ul>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>spot_meta = read.csv(<span style="color:#0ff;font-weight:bold">&#34;posterior/spatial/tissue_positions_list.csv&#34;</span>,
</span></span><span style="display:flex;"><span>                     header = <span style="color:#fff;font-weight:bold">FALSE</span>)
</span></span><span style="display:flex;"><span>head(spot_meta)    <span style="color:#007f7f">#tissue row col imagerow imagecol</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                   V1 V2 V3 V4   V5   V6</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 ACGCCTGACACGCGCT-1  0  0  0 1419 1432</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 TACCGATCCAACACTT-1  0  1  1 1538 1501</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 ATTAAAGCGGACGAGC-1  0  0  2 1419 1570</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>最后是一个json文件(scalefactors_json.json)，用于记录tissue_positions_list.csv所记录的spot绝对坐标与tissue_lowres_image.png图片的比例关系。以及spot的直径大小。
<ul>
<li>例如 (1419, 1432)*0.0516=(73.2204, 73.8912)即为在tissue_lowres_image.png中的坐标位置。</li>
</ul>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-json" data-lang="json"><span style="display:flex;"><span>{
</span></span><span style="display:flex;"><span><span style="font-weight:bold">&#34;fiducial_diameter_fullres&#34;</span>: <span style="color:#ff0;font-weight:bold">144.59793566029023</span>, 
</span></span><span style="display:flex;"><span><span style="font-weight:bold">&#34;tissue_lowres_scalef&#34;</span>: <span style="color:#ff0;font-weight:bold">0.051635113</span>
</span></span><span style="display:flex;"><span>}
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p><a href="https://github.com/satijalab/seurat/issues/4993">https://github.com/satijalab/seurat/issues/4993</a></p>
<p>There are four scaling factors that 10X provides, but we generally <strong>only use the <code>lowres</code> factor.</strong> If you&rsquo;re using your own, <u>non-10X data</u>, simply create a fake <code>scalefactors</code> object with <code>scalefactors</code>, making sure that you use the correct factor for <code>lowres</code>. If you coordinates match 1:1 with your image, then your lowres scale factor would be 1</p></blockquote>
<h1 id="1创建seurat对象">1、创建Seurat对象<a hidden class="anchor" aria-hidden="true" href="#1创建seurat对象">#</a></h1>
<p>思路：先根据spot基因表达数据创建Seurat对象，然后加入组织图片以及spot的位置信息。</p>
<h2 id="11-分步创建">1.1 分步创建<a hidden class="anchor" aria-hidden="true" href="#11-分步创建">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1)构建传统seurat对象</span>
</span></span><span style="display:flex;"><span>sce = Read10X(<span style="color:#0ff;font-weight:bold">&#34;posterior/filtered_feature_bc_matrix/&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  CreateSeuratObject(assay = <span style="color:#0ff;font-weight:bold">&#34;Spot&#34;</span>)     <span style="color:#007f7f">#将assay名设置为Spatial</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># sce &lt;- Read10X_h5(&#34;posterior/filtered_feature_bc_matrix.h5&#34;) %&gt;% </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   CreateSeuratObject(assay = &#34;Spatial&#34;)</span>
</span></span><span style="display:flex;"><span>dim(sce)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 32285  3355</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3355个spot的32285个基因的表达</span>
</span></span><span style="display:flex;"><span>sce@assays$Spot@counts[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 x 4 sparse Matrix of class &#34;dgCMatrix&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#         AAACAAGTATCTCCCA-1 AAACACCAATAACTGC-1 AAACAGAGCGACTCCT-1 AAACAGCTTTCAGAAG-1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Xkr4                     .                  .                  .                  .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Gm1992                   .                  .                  .                  .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Gm19938                  .                  1                  .                  .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Gm37381                  .                  .                  .                  .</span>
</span></span><span style="display:flex;"><span>head(sce@meta.data)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                       orig.ident nCount_Spot nFeature_Spot</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACAAGTATCTCCCA-1 SeuratProject        9195          3089</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACACCAATAACTGC-1 SeuratProject       33655          6468</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACAGAGCGACTCCT-1 SeuratProject       19619          5245</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2)加入image信息</span>
</span></span><span style="display:flex;"><span>image &lt;- Read10X_Image(image.dir = <span style="color:#0ff;font-weight:bold">&#34;posterior/spatial/&#34;</span>)
</span></span><span style="display:flex;"><span>image@image %&gt;% dim()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 600 600   3</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#image &lt;- image[Cells(x = sce)] </span>
</span></span><span style="display:flex;"><span>image@coordinates %&gt;% head()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                    tissue row col imagerow imagecol</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TGGGACCATTGGGAGT-1      1   7   5     2257     1776</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CCGGTGCGAGTGATAG-1      1   7   7     2257     1914</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TAGCCAGAGGGTCCGG-1      1   7   9     2257     2052</span>
</span></span><span style="display:flex;"><span>image@scale.factors  <span style="color:#007f7f">#储存json信息</span>
</span></span><span style="display:flex;"><span>DefaultAssay(image) &lt;- <span style="color:#0ff;font-weight:bold">&#34;Spot&#34;</span>  
</span></span><span style="display:flex;"><span>DefaultAssay(image)
</span></span><span style="display:flex;"><span>sce[[<span style="color:#0ff;font-weight:bold">&#34;posterior&#34;</span>]] &lt;- image
</span></span><span style="display:flex;"><span>sce
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># An object of class Seurat </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 32285 features across 3355 samples within 1 assay </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Active assay: Spot (32285 features, 0 variable features)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 image present: posterior</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="12-一步创建">1.2 一步创建<a hidden class="anchor" aria-hidden="true" href="#12-一步创建">#</a></h2>
<p><code>Load10X_Spatial()</code>函数支持在准备好相应的文件基础上，一步创建出包含image信息的seurat对象</p>
<ul>
<li>
<p>测序表达数据的h5文件</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;filtered_feature_bc_matrix.h5&#34;   </span>
</span></span></code></pre></td></tr></table>
</div>
</div></li>
<li>
<p>图片及位置信息的spatial文件夹</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>list.files(<span style="color:#0ff;font-weight:bold">&#34;posterior/spatial/&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;scalefactors_json.json&#34;    &#34;tissue_lowres_image.png&#34;   &#34;tissue_positions_list.csv&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sce = Load10X_Spatial(data.dir = <span style="color:#0ff;font-weight:bold">&#34;./posterior/&#34;</span>,slice = <span style="color:#0ff;font-weight:bold">&#34;posterior&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># An object of class Seurat </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 32285 features across 3355 samples within 1 assay </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Active assay: Spatial (32285 features, 0 variable features)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 image present: posterior</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="2分析流程与可视化">2、分析流程与可视化<a hidden class="anchor" aria-hidden="true" href="#2分析流程与可视化">#</a></h1>
<p>之后的分析可基本参考Seurat单细胞数据分析流程，完成spot的过滤、标准化、聚类分群、差异分析等；就不过多介绍了。</p>
<p>主要会在可视化方面有空间转录组自身独特的展示方式。</p>
<h2 id="21-常规分析流程">2.1 常规分析流程<a hidden class="anchor" aria-hidden="true" href="#21-常规分析流程">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##标准化</span>
</span></span><span style="display:flex;"><span>sce &lt;- SCTransform(sce,assay = <span style="color:#0ff;font-weight:bold">&#34;Spot&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##降维</span>
</span></span><span style="display:flex;"><span>sce &lt;- RunPCA(sce)
</span></span><span style="display:flex;"><span>sce &lt;- RunUMAP(sce, dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>, label = <span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##聚类分群</span>
</span></span><span style="display:flex;"><span>sce &lt;- FindNeighbors(sce,dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">20</span>)
</span></span><span style="display:flex;"><span>sce &lt;- FindClusters(sce,resolution = <span style="color:#ff0;font-weight:bold">0.1</span>)
</span></span><span style="display:flex;"><span>table(sce$seurat_clusters)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#    0    1    2    3    4    5    6 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1161  659  483  396  235  230  191</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##差异分析</span>
</span></span><span style="display:flex;"><span>dif &lt;- FindAllMarkers(sce, assay = <span style="color:#0ff;font-weight:bold">&#34;Spot&#34;</span>, only.pos = <span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>sig.dif &lt;- dif%&gt;%group_by(cluster) %&gt;% top_n(n = <span style="color:#ff0;font-weight:bold">5</span>,wt = avg_log2FC)
</span></span><span style="display:flex;"><span>genes &lt;- unique(sig.dif$gene)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="22-可视化">2.2 可视化<a hidden class="anchor" aria-hidden="true" href="#22-可视化">#</a></h2>
<p>这里主要记录空间转录组自身独特的展示方式：以测序组织图片为背景，展示spot的聚类信息或者特定基因在每个spot的表达信息。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>SpatialDimPlot(sce, alpha = <span style="color:#ff0;font-weight:bold">0.5</span>)
</span></span><span style="display:flex;"><span>SpatialDimPlot(sce,
</span></span><span style="display:flex;"><span>               cells.highlight = CellsByIdentities(sce,idents = <span style="color:#0ff;font-weight:bold">&#34;1&#34;</span>))
</span></span><span style="display:flex;"><span>SpatialDimPlot(sce,cells.highlight = CellsByIdentities(sce,idents = c(<span style="color:#0ff;font-weight:bold">&#34;0&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;2&#34;</span>)),
</span></span><span style="display:flex;"><span>               facet.highlight = <span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>SpatialDimPlot(sce,cells.highlight = CellsByIdentities(sce,idents = c(<span style="color:#0ff;font-weight:bold">&#34;0&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;2&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;3&#34;</span>)),
</span></span><span style="display:flex;"><span>               facet.highlight = <span style="color:#fff;font-weight:bold">T</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220605090831040.png" alt="image-20220605090831040" style="zoom: 50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-fallback" data-lang="fallback"><span style="display:flex;"><span>LinkedDimPlot(sce,reduction = &#34;umap&#34;)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220605091318443.png" alt="image-20220605091318443" style="zoom: 33%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>SpatialFeaturePlot(sce,features = genes[1:<span style="color:#ff0;font-weight:bold">2</span>],interactive = <span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>SpatialFeaturePlot(sce,features = genes[1],pt.size=<span style="color:#ff0;font-weight:bold">1.5</span>,alpha=c(<span style="color:#ff0;font-weight:bold">0.1</span>,<span style="color:#ff0;font-weight:bold">1</span>),interactive = <span style="color:#fff;font-weight:bold">T</span>) 
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220605091105406.png" alt="image-20220605091105406" style="zoom: 67%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>LinkedFeaturePlot(sce, feature = <span style="color:#0ff;font-weight:bold">&#34;Stx1a&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220605091459725.png" alt="image-20220605091459725" style="zoom: 33%;" />
<h1 id="3合并多个切片分析">3、合并多个切片分析<a hidden class="anchor" aria-hidden="true" href="#3合并多个切片分析">#</a></h1>
<p>类似合并多个单细胞测序样本分析，此时每个Seurat对象都要有自己的image信息。必要时也可以使用批次校正等工具。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>anterior1 &lt;- Load10X_Spatial(data.dir = <span style="color:#0ff;font-weight:bold">&#34;./Anterior&#34;</span>,slice = <span style="color:#0ff;font-weight:bold">&#34;anterior&#34;</span>)
</span></span><span style="display:flex;"><span>posterior1 &lt;- Load10X_Spatial(data.dir = <span style="color:#0ff;font-weight:bold">&#34;./posterior/&#34;</span>,slice = <span style="color:#0ff;font-weight:bold">&#34;posterior&#34;</span>)
</span></span><span style="display:flex;"><span>dim(anterior1)
</span></span><span style="display:flex;"><span>dim(posterior1)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>st &lt;- merge(anterior1,posterior1,add.cell.ids = c(<span style="color:#0ff;font-weight:bold">&#34;ante&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;post&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#st &lt;- merge(a,y=c(b,c,d),add.cell.ids = c(&#34;a&#34;,&#34;b&#34;,&#34;c&#34;,&#34;d&#34;),project = &#34;four_merged&#34;)</span>
</span></span><span style="display:flex;"><span>SpatialDimPlot(st,label = <span style="color:#fff;font-weight:bold">T</span>,label.size = <span style="color:#ff0;font-weight:bold">3</span>, images = <span style="color:#0ff;font-weight:bold">&#34;anterior&#34;</span>)
</span></span><span style="display:flex;"><span>SpatialDimPlot(st,label = <span style="color:#fff;font-weight:bold">T</span>,label.size = <span style="color:#ff0;font-weight:bold">3</span>, images = <span style="color:#0ff;font-weight:bold">&#34;posterior&#34;</span>)
</span></span><span style="display:flex;"><span>SpatialDimPlot(st,label = <span style="color:#fff;font-weight:bold">T</span>,label.size = <span style="color:#ff0;font-weight:bold">3</span>, images = c(<span style="color:#0ff;font-weight:bold">&#34;anterior&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;posterior&#34;</span>))
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/program/012python-tdigest%E5%88%86%E4%BD%8D%E6%95%B0%E7%BB%9F%E8%AE%A1/">
    <span class="title">« Prev Page</span>
    <br>
    <span>Python-tdigest分位数统计</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/016%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--hdwgcna%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C/">
    <span class="title">Next Page »</span>
    <br>
    <span>单细胞分析工具--hdWGCNA共表达网络</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
