<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>文献--髓鞘脱失与再生模型小鼠snRNAseq数据分析 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="文献-组学">
<meta name="description" content="
题目：Transcriptomic atlas and interaction networks of brain cells in mouse CNS demyelination and remyelination
期刊|日期：Cell Report | April 2023
DOI：https://doi.org/10.1016/j.celrep.2023.112293">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/basic/176%E6%96%87%E7%8C%AE--%E9%AB%93%E9%9E%98%E8%84%B1%E5%A4%B1%E4%B8%8E%E5%86%8D%E7%94%9F%E6%A8%A1%E5%9E%8B%E5%B0%8F%E9%BC%A0snrnaseq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/basic/176%E6%96%87%E7%8C%AE--%E9%AB%93%E9%9E%98%E8%84%B1%E5%A4%B1%E4%B8%8E%E5%86%8D%E7%94%9F%E6%A8%A1%E5%9E%8B%E5%B0%8F%E9%BC%A0snrnaseq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="文献--髓鞘脱失与再生模型小鼠snRNAseq数据分析" />
<meta property="og:description" content="
题目：Transcriptomic atlas and interaction networks of brain cells in mouse CNS demyelination and remyelination
期刊|日期：Cell Report | April 2023
DOI：https://doi.org/10.1016/j.celrep.2023.112293" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/basic/176%E6%96%87%E7%8C%AE--%E9%AB%93%E9%9E%98%E8%84%B1%E5%A4%B1%E4%B8%8E%E5%86%8D%E7%94%9F%E6%A8%A1%E5%9E%8B%E5%B0%8F%E9%BC%A0snrnaseq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-04-17T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-04-17T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="文献--髓鞘脱失与再生模型小鼠snRNAseq数据分析"/>
<meta name="twitter:description" content="
题目：Transcriptomic atlas and interaction networks of brain cells in mouse CNS demyelination and remyelination
期刊|日期：Cell Report | April 2023
DOI：https://doi.org/10.1016/j.celrep.2023.112293"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 科研基础 -- 读文献、看教材",
      "item": "https://lishensuo.github.io/en/posts/basic/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "文献--髓鞘脱失与再生模型小鼠snRNAseq数据分析",
      "item": "https://lishensuo.github.io/en/posts/basic/176%E6%96%87%E7%8C%AE--%E9%AB%93%E9%9E%98%E8%84%B1%E5%A4%B1%E4%B8%8E%E5%86%8D%E7%94%9F%E6%A8%A1%E5%9E%8B%E5%B0%8F%E9%BC%A0snrnaseq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "文献--髓鞘脱失与再生模型小鼠snRNAseq数据分析",
  "name": "文献--髓鞘脱失与再生模型小鼠snRNAseq数据分析",
  "description": " 题目：Transcriptomic atlas and interaction networks of brain cells in mouse CNS demyelination and remyelination\n期刊|日期：Cell Report | April 2023\nDOI：https://doi.org/10.1016/j.celrep.2023.112293\n",
  "keywords": [
    "文献-组学"
  ],
  "articleBody": " 题目：Transcriptomic atlas and interaction networks of brain cells in mouse CNS demyelination and remyelination\n期刊|日期：Cell Report | April 2023\nDOI：https://doi.org/10.1016/j.celrep.2023.112293\n简介：髓鞘脱失是MS(多发性硬化症)发生和发展的主要病理生理机制之一。来自美国华盛顿大学医学院的Marco Colonna团队对小鼠髓鞘脱失与再生模型进行单核细胞测序分析，对多种脑内细胞特异的分子特异变化进行了探索与解读。测序数据已上传到 GSE20477。\n1、单核细胞测序整体分析 （1）测序组织分组\n正常小鼠（n=2），简称normal\n髓鞘脱失模型小鼠（n=3），简称Demye\n髓鞘再生模型小鼠（n=3），简称Remye\n（2）测序初步结果\n58079个细胞，19个细胞群，根据marker gene注释为7大类细胞类型\n同一细胞类型在3组的比例变化\n同一细胞类型在3组的两两表的差异基因\n2、Oli少突胶质细胞分析 将Oli分为8个亚群\n分析Demye组与normal比例相差较大的亚群及其对应功能，再分析这些亚群在Remye的变化\n关联外部数据分析\nDemye与Remye内亚群的发育轨迹 与AD的Oli合并分析发现独特与共有的亚型 基于人的MS的Oli差异基因表现部分Remye特征 发现IL-33在不同组的Oli表达变化，及其同源受体ST2在Ast/Micro的表达变化\n3、Ast星形胶质细胞分析 首先分析了Demye， Remye的marker gene及其功能 将Ast进一步分为9个亚群，分析每组相关的亚群及其功能 结合MS相关的Ast marker，分析与Demye， Remye状态下的Ast marker关系 4、其它细胞类型分析 针对其它细胞类型，分析思路与上面类似，概述如下\n（1）Micro小胶质细胞\n将Micro分为9个亚群（下图） 分析每组相关的亚群及其功能 （2）Vasc血管细胞等\n同样进行亚群分析，对于同一亚群进行Demye与 Remye间差异分析 对两个内皮细胞亚群之间的特征分 （3）神经元细胞分析\n分别对兴奋与抑制性神经元进行亚群分析 同一亚群在Demye与 Remye间差异分析 5、NicheNet细胞通讯分析 将Oli作为受体细胞，在normal与Demye差异表达变化下，分析来自其它细胞类型的可能配体 将上述分析也将Ast与Micro作为受体细胞同样做一遍，发现了若干重要的配受体对，进行深入研究 6、TERM2基因敲除的影响 同样对TERM2基因敲除的3组进行单细胞测序，细胞类型注释 分析Demye中每一细胞类型WT与敲除组的差异；对Remye进行同样分析 对Micro进行亚群注释，比较Demye中WT与敲除组的亚群比例差异；对Remye进行同样分析 对Oli同样进行上述分析 ",
  "wordCount" : "1115",
  "inLanguage": "en",
  "datePublished": "2023-04-17T00:00:00Z",
  "dateModified": "2023-04-17T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/basic/176%E6%96%87%E7%8C%AE--%E9%AB%93%E9%9E%98%E8%84%B1%E5%A4%B1%E4%B8%8E%E5%86%8D%E7%94%9F%E6%A8%A1%E5%9E%8B%E5%B0%8F%E9%BC%A0snrnaseq%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/basic/">📖 科研基础 -- 读文献、看教材</a></div>
    <h1 class="post-title">
      文献--髓鞘脱失与再生模型小鼠snRNAseq数据分析
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-04-17 00:00:00 &#43;0000 UTC&#39;&gt;2023-04-17&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-04-17&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1115&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e5%8d%95%e6%a0%b8%e7%bb%86%e8%83%9e%e6%b5%8b%e5%ba%8f%e6%95%b4%e4%bd%93%e5%88%86%e6%9e%90" aria-label="1、单核细胞测序整体分析">1、单核细胞测序整体分析</a></li>
                    <li>
                        <a href="#2oli%e5%b0%91%e7%aa%81%e8%83%b6%e8%b4%a8%e7%bb%86%e8%83%9e%e5%88%86%e6%9e%90" aria-label="2、Oli少突胶质细胞分析">2、Oli少突胶质细胞分析</a></li>
                    <li>
                        <a href="#3ast%e6%98%9f%e5%bd%a2%e8%83%b6%e8%b4%a8%e7%bb%86%e8%83%9e%e5%88%86%e6%9e%90" aria-label="3、Ast星形胶质细胞分析">3、Ast星形胶质细胞分析</a></li>
                    <li>
                        <a href="#4%e5%85%b6%e5%ae%83%e7%bb%86%e8%83%9e%e7%b1%bb%e5%9e%8b%e5%88%86%e6%9e%90" aria-label="4、其它细胞类型分析">4、其它细胞类型分析</a></li>
                    <li>
                        <a href="#5nichenet%e7%bb%86%e8%83%9e%e9%80%9a%e8%ae%af%e5%88%86%e6%9e%90" aria-label="5、NicheNet细胞通讯分析">5、NicheNet细胞通讯分析</a></li>
                    <li>
                        <a href="#6term2%e5%9f%ba%e5%9b%a0%e6%95%b2%e9%99%a4%e7%9a%84%e5%bd%b1%e5%93%8d" aria-label="6、TERM2基因敲除的影响">6、TERM2基因敲除的影响</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p><strong>题目</strong>：Transcriptomic atlas and interaction networks of brain cells in mouse CNS demyelination and remyelination</p>
<p><strong>期刊|日期</strong>：Cell Report | April 2023</p>
<p><strong>DOI</strong>：https://doi.org/10.1016/j.celrep.2023.112293</p>
<p><strong>简介</strong>：髓鞘脱失是MS(多发性硬化症)发生和发展的主要病理生理机制之一。来自美国华盛顿大学医学院的Marco Colonna团队对小鼠髓鞘脱失与再生模型进行单核细胞测序分析，对多种脑内细胞特异的分子特异变化进行了探索与解读。测序数据已上传到 GSE20477。</p>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230423173021568.png" alt="image-20230423173021568" style="zoom:67%;" /></blockquote>
<h3 id="1单核细胞测序整体分析">1、单核细胞测序整体分析<a hidden class="anchor" aria-hidden="true" href="#1单核细胞测序整体分析">#</a></h3>
<p><strong>（1）测序组织分组</strong></p>
<ul>
<li>
<p>正常小鼠（n=2），简称normal</p>
</li>
<li>
<p>髓鞘脱失模型小鼠（n=3），简称Demye</p>
</li>
<li>
<p>髓鞘再生模型小鼠（n=3），简称Remye</p>
</li>
</ul>
<p><strong>（2）测序初步结果</strong></p>
<ul>
<li>
<p>58079个细胞，19个细胞群，根据marker gene注释为7大类细胞类型</p>
</li>
<li>
<p>同一细胞类型在3组的比例变化</p>
</li>
<li>
<p>同一细胞类型在3组的两两表的差异基因</p>
</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230423173153836.png" alt="image-20230423173153836" style="zoom:67%;" />
<h3 id="2oli少突胶质细胞分析">2、Oli少突胶质细胞分析<a hidden class="anchor" aria-hidden="true" href="#2oli少突胶质细胞分析">#</a></h3>
<ul>
<li>
<p>将Oli分为8个亚群</p>
</li>
<li>
<p>分析Demye组与normal比例相差较大的亚群及其对应功能，再分析这些亚群在Remye的变化</p>
</li>
<li>
<p>关联外部数据分析</p>
<ul>
<li>Demye与Remye内亚群的发育轨迹</li>
<li>与AD的Oli合并分析发现独特与共有的亚型</li>
<li>基于人的MS的Oli差异基因表现部分Remye特征</li>
</ul>
</li>
<li>
<p>发现IL-33在不同组的Oli表达变化，及其同源受体ST2在Ast/Micro的表达变化</p>
</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230423173322771.png" alt="image-20230423173322771" style="zoom:67%;" />
<h3 id="3ast星形胶质细胞分析">3、Ast星形胶质细胞分析<a hidden class="anchor" aria-hidden="true" href="#3ast星形胶质细胞分析">#</a></h3>
<ul>
<li>首先分析了Demye， Remye的marker gene及其功能</li>
<li>将Ast进一步分为9个亚群，分析每组相关的亚群及其功能</li>
<li>结合MS相关的Ast marker，分析与Demye， Remye状态下的Ast marker关系</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230423173420886.png" alt="image-20230423173420886" style="zoom: 67%;" />
<h3 id="4其它细胞类型分析">4、其它细胞类型分析<a hidden class="anchor" aria-hidden="true" href="#4其它细胞类型分析">#</a></h3>
<blockquote>
<p>针对其它细胞类型，分析思路与上面类似，概述如下</p></blockquote>
<p><strong>（1）Micro小胶质细胞</strong></p>
<ul>
<li>将Micro分为9个亚群（下图）</li>
<li>分析每组相关的亚群及其功能</li>
</ul>
<p><strong>（2）Vasc血管细胞等</strong></p>
<ul>
<li>同样进行亚群分析，对于同一亚群进行Demye与 Remye间差异分析</li>
<li>对两个内皮细胞亚群之间的特征分</li>
</ul>
<p><strong>（3）神经元细胞分析</strong></p>
<ul>
<li>分别对兴奋与抑制性神经元进行亚群分析</li>
<li>同一亚群在Demye与 Remye间差异分析</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230423173955580.png" alt="image-20230423173955580" style="zoom:67%;" />
<h3 id="5nichenet细胞通讯分析">5、NicheNet细胞通讯分析<a hidden class="anchor" aria-hidden="true" href="#5nichenet细胞通讯分析">#</a></h3>
<ul>
<li>将Oli作为受体细胞，在normal与Demye差异表达变化下，分析来自其它细胞类型的可能配体</li>
<li>将上述分析也将Ast与Micro作为受体细胞同样做一遍，发现了若干重要的配受体对，进行深入研究</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230423174102920.png" alt="image-20230423174102920" style="zoom:67%;" />
<h3 id="6term2基因敲除的影响">6、TERM2基因敲除的影响<a hidden class="anchor" aria-hidden="true" href="#6term2基因敲除的影响">#</a></h3>
<ul>
<li>同样对TERM2基因敲除的3组进行单细胞测序，细胞类型注释</li>
<li>分析Demye中每一细胞类型WT与敲除组的差异；对Remye进行同样分析</li>
<li>对Micro进行亚群注释，比较Demye中WT与敲除组的亚群比例差异；对Remye进行同样分析</li>
<li>对Oli同样进行上述分析</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230423174151699.png" alt="image-20230423174151699" style="zoom:67%;" />


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E6%96%87%E7%8C%AE-%E7%BB%84%E5%AD%A6/">文献-组学</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/basic/175%E6%96%87%E7%8C%AE--ccrcc%E5%8D%95%E7%BB%86%E8%83%9E%E8%BD%AC%E5%BD%95%E7%BB%84/">
    <span class="title">« Prev Page</span>
    <br>
    <span>文献--ccRCC单细胞转录组</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/basic/177%E6%96%87%E7%8C%AE--%E5%9F%BA%E4%BA%8E%E6%B3%9B%E7%99%8Cscrnaseq%E7%9A%84t%E7%BB%86%E8%83%9E%E5%9B%BE%E8%B0%B1%E6%95%B4%E5%90%88%E5%88%86%E6%9E%90/">
    <span class="title">Next Page »</span>
    <br>
    <span>文献--基于泛癌scRNAseq的T细胞图谱整合分析</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
