<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>å•ç»†èƒåˆ†æå·¥å…·--ç»†èƒå‘¨æœŸåˆ†æ | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
A cell cycle is a series of events that takes place in a cell as it grows and divides.å³æè¿°ç»†èƒç”Ÿé•¿ã€åˆ†è£‚æ•´ä¸ªè¿‡ç¨‹ä¸­ç»†èƒå˜åŒ–è¿‡ç¨‹ã€‚æœ€é‡è¦çš„ä¸¤ä¸ªç‰¹ç‚¹å°±æ˜¯DNAå¤åˆ¶ã€åˆ†è£‚æˆä¸¤ä¸ªä¸€æ ·çš„å­ç»†èƒã€‚å¯¹äºå•ç»†èƒè½¬å½•ç»„æ•°æ®ï¼Œå¯æ ¹æ®ç›¸å…³markeråŸºå› çš„è¡¨è¾¾æ°´å¹³åˆ¤æ–­æ¯ä¸€ä¸ªç»†èƒæ‰€å¤„çš„ç»†èƒå‘¨æœŸçŠ¶æ€ã€‚æ­¤å¤–åœ¨åˆ†æå¤šæ•°æ®é›†é—´æ‰¹æ¬¡æ•ˆåº”æ—¶ï¼Œå¯æ ¹æ®æ¯ä¸ªæ•°æ®é›†ä¸­å„ä¸ªç»†èƒå‘¨æœŸæ¯”ä¾‹è¿›è¡Œåˆ¤æ–­ä¸æ ¡æ­£ã€‚">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/020%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F%E5%88%86%E6%9E%90/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/020%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F%E5%88%86%E6%9E%90/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="å•ç»†èƒåˆ†æå·¥å…·--ç»†èƒå‘¨æœŸåˆ†æ" />
<meta property="og:description" content="
A cell cycle is a series of events that takes place in a cell as it grows and divides.å³æè¿°ç»†èƒç”Ÿé•¿ã€åˆ†è£‚æ•´ä¸ªè¿‡ç¨‹ä¸­ç»†èƒå˜åŒ–è¿‡ç¨‹ã€‚æœ€é‡è¦çš„ä¸¤ä¸ªç‰¹ç‚¹å°±æ˜¯DNAå¤åˆ¶ã€åˆ†è£‚æˆä¸¤ä¸ªä¸€æ ·çš„å­ç»†èƒã€‚å¯¹äºå•ç»†èƒè½¬å½•ç»„æ•°æ®ï¼Œå¯æ ¹æ®ç›¸å…³markeråŸºå› çš„è¡¨è¾¾æ°´å¹³åˆ¤æ–­æ¯ä¸€ä¸ªç»†èƒæ‰€å¤„çš„ç»†èƒå‘¨æœŸçŠ¶æ€ã€‚æ­¤å¤–åœ¨åˆ†æå¤šæ•°æ®é›†é—´æ‰¹æ¬¡æ•ˆåº”æ—¶ï¼Œå¯æ ¹æ®æ¯ä¸ªæ•°æ®é›†ä¸­å„ä¸ªç»†èƒå‘¨æœŸæ¯”ä¾‹è¿›è¡Œåˆ¤æ–­ä¸æ ¡æ­£ã€‚" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/020%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F%E5%88%86%E6%9E%90/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-01-22T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-01-22T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="å•ç»†èƒåˆ†æå·¥å…·--ç»†èƒå‘¨æœŸåˆ†æ"/>
<meta name="twitter:description" content="
A cell cycle is a series of events that takes place in a cell as it grows and divides.å³æè¿°ç»†èƒç”Ÿé•¿ã€åˆ†è£‚æ•´ä¸ªè¿‡ç¨‹ä¸­ç»†èƒå˜åŒ–è¿‡ç¨‹ã€‚æœ€é‡è¦çš„ä¸¤ä¸ªç‰¹ç‚¹å°±æ˜¯DNAå¤åˆ¶ã€åˆ†è£‚æˆä¸¤ä¸ªä¸€æ ·çš„å­ç»†èƒã€‚å¯¹äºå•ç»†èƒè½¬å½•ç»„æ•°æ®ï¼Œå¯æ ¹æ®ç›¸å…³markeråŸºå› çš„è¡¨è¾¾æ°´å¹³åˆ¤æ–­æ¯ä¸€ä¸ªç»†èƒæ‰€å¤„çš„ç»†èƒå‘¨æœŸçŠ¶æ€ã€‚æ­¤å¤–åœ¨åˆ†æå¤šæ•°æ®é›†é—´æ‰¹æ¬¡æ•ˆåº”æ—¶ï¼Œå¯æ ¹æ®æ¯ä¸ªæ•°æ®é›†ä¸­å„ä¸ªç»†èƒå‘¨æœŸæ¯”ä¾‹è¿›è¡Œåˆ¤æ–­ä¸æ ¡æ­£ã€‚"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "åˆ†ç±»",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "å•ç»†èƒåˆ†æå·¥å…·--ç»†èƒå‘¨æœŸåˆ†æ",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/020%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F%E5%88%86%E6%9E%90/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "å•ç»†èƒåˆ†æå·¥å…·--ç»†èƒå‘¨æœŸåˆ†æ",
  "name": "å•ç»†èƒåˆ†æå·¥å…·--ç»†èƒå‘¨æœŸåˆ†æ",
  "description": " A cell cycle is a series of events that takes place in a cell as it grows and divides.å³æè¿°ç»†èƒç”Ÿé•¿ã€åˆ†è£‚æ•´ä¸ªè¿‡ç¨‹ä¸­ç»†èƒå˜åŒ–è¿‡ç¨‹ã€‚æœ€é‡è¦çš„ä¸¤ä¸ªç‰¹ç‚¹å°±æ˜¯DNAå¤åˆ¶ã€åˆ†è£‚æˆä¸¤ä¸ªä¸€æ ·çš„å­ç»†èƒã€‚å¯¹äºå•ç»†èƒè½¬å½•ç»„æ•°æ®ï¼Œå¯æ ¹æ®ç›¸å…³markeråŸºå› çš„è¡¨è¾¾æ°´å¹³åˆ¤æ–­æ¯ä¸€ä¸ªç»†èƒæ‰€å¤„çš„ç»†èƒå‘¨æœŸçŠ¶æ€ã€‚æ­¤å¤–åœ¨åˆ†æå¤šæ•°æ®é›†é—´æ‰¹æ¬¡æ•ˆåº”æ—¶ï¼Œå¯æ ¹æ®æ¯ä¸ªæ•°æ®é›†ä¸­å„ä¸ªç»†èƒå‘¨æœŸæ¯”ä¾‹è¿›è¡Œåˆ¤æ–­ä¸æ ¡æ­£ã€‚\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " A cell cycle is a series of events that takes place in a cell as it grows and divides.å³æè¿°ç»†èƒç”Ÿé•¿ã€åˆ†è£‚æ•´ä¸ªè¿‡ç¨‹ä¸­ç»†èƒå˜åŒ–è¿‡ç¨‹ã€‚æœ€é‡è¦çš„ä¸¤ä¸ªç‰¹ç‚¹å°±æ˜¯DNAå¤åˆ¶ã€åˆ†è£‚æˆä¸¤ä¸ªä¸€æ ·çš„å­ç»†èƒã€‚å¯¹äºå•ç»†èƒè½¬å½•ç»„æ•°æ®ï¼Œå¯æ ¹æ®ç›¸å…³markeråŸºå› çš„è¡¨è¾¾æ°´å¹³åˆ¤æ–­æ¯ä¸€ä¸ªç»†èƒæ‰€å¤„çš„ç»†èƒå‘¨æœŸçŠ¶æ€ã€‚æ­¤å¤–åœ¨åˆ†æå¤šæ•°æ®é›†é—´æ‰¹æ¬¡æ•ˆåº”æ—¶ï¼Œå¯æ ¹æ®æ¯ä¸ªæ•°æ®é›†ä¸­å„ä¸ªç»†èƒå‘¨æœŸæ¯”ä¾‹è¿›è¡Œåˆ¤æ–­ä¸æ ¡æ­£ã€‚\nä¸€ã€ç»†èƒå‘¨æœŸé˜¶æ®µ ç»†èƒå‘¨æœŸç›¸å…³åŸºç¡€çŸ¥è¯†å…·ä½“å¯å‚è€ƒå¦‚ä¸‹é“¾æ¥ï¼š\nhttps://teachmephysiology.com/biochemistry/cell-growth-death/cell-cycle/ https://www.genome.gov/genetics-glossary/Cell-Cycle ç»†èƒå‘¨æœŸ _ æœç´¢ç»“æœ_å“”å“©å“”å“©_Bilibili\nç®€å•æ¥è¯´ï¼Œä¸€èˆ¬å¯åˆ†æˆ4ä¸ªé˜¶æ®µ\nG1(gap1)ï¼šCell increases in size(Cellular contents duplicated) S(synthesis) ï¼šDNA replication, each of the 46 chromosomes (23 pairs) is replicated by the cell G2(gap2)ï¼šCell grows moreï¼Œorganelles and proteins develop in preparation for cell division M(mitosis)ï¼šâ€˜Oldâ€™ cell partitions the two copies of the genetic material into the two daughter cells. And the cell cycle can begin again. åœ¨å¤„ç†å•ç»†èƒæ•°æ®æ—¶ï¼Œåˆ†æç»†èƒå‘¨æœŸå¯å¯¹äºç”Ÿç‰©æœºåˆ¶æ¢ç´¢æˆ–è€…æ½œåœ¨æ‰¹æ¬¡æ•ˆåº”åˆ¤æ–­æä¾›å¦ä¸€ç§è§’åº¦çš„è§è§£ã€‚ç”±äºç»†èƒå‘¨æœŸä¹Ÿæ˜¯é€šè¿‡cell cycle related protein è°ƒæ§ï¼Œå³æ¯ä¸ªé˜¶æ®µæœ‰æ˜¾è‘—çš„markeråŸºå› ï¼›é€šè¿‡åˆ†æç»†èƒå‘¨æœŸæœ‰å…³åŸºå› çš„è¡¨è¾¾æƒ…å†µï¼Œå¯ä»¥å¯¹ç»†èƒæ‰€å¤„å‘¨æœŸé˜¶æ®µè¿›è¡Œæ³¨é‡Šã€‚\nNoteï¼šåœ¨å•ç»†èƒå‘¨æœŸåˆ†ææ—¶ï¼Œé€šå¸¸åªè€ƒè™‘ä¸‰ä¸ªé˜¶æ®µï¼šG1ã€Sã€G2Mã€‚(å³æŠŠG2å’ŒMå½“åšä¸€ä¸ªphase)\nå¦‚ä¸‹ç®€å•å­¦ä¹ åˆ†æå•ç»†èƒç»†èƒå‘¨æœŸçš„ä¸¤ç§æ–¹æ³•ï¼Œç¤ºä¾‹æ•°æ®é›†æ¥è‡ª10Xçš„PBMCæ ·æœ¬\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 library(Seurat) library(tidyverse) # download.file(\"https://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz\", # \"pbmc3k_filtered_gene_bc_matrices.tar.gz\") # untar(\"pbmc3k_filtered_gene_bc_matrices.tar.gz\") pbmc.data = Read10X(data.dir = \"filtered_gene_bc_matrices/hg19/\") pbmc = CreateSeuratObject(counts = pbmc.data, project = \"pbmc3k\") pbmc = NormalizeData(pbmc) %\u003e% FindVariableFeatures() pbmc = ScaleData(pbmc) %\u003e% RunPCA() %\u003e% RunUMAP(dims = 1:30) DimPlot(pbmc, reduction = \"pca\") äºŒã€SeuratåŒ…åˆ†æ SeuratåŒ…ç»†èƒå‘¨æœŸåˆ†æçš„åŸºæœ¬åŸç†ç±»ä¼¼æ¨¡å—æ‰“åˆ†ã€‚ é¦–å…ˆè¯¥åŒ…å®šä¹‰äº†SæœŸä¸G2MæœŸçš„åŸºå› æ¨¡å—ï¼Œåˆ†åˆ«å¯¹æ¯ä¸ªå•ç»†èƒè¿›è¡Œæ‰“åˆ†ã€‚ è‹¥SæœŸåŸºå› æ¨¡å—åˆ†æ•°é«˜ï¼Œåˆ™è®¤ä¸ºç»†èƒå¤„äºSæœŸï¼›åŒç†åˆ†æG2MæœŸã€‚è‹¥äºŒè€…åˆ†æ•°å‡è¾ƒä½ï¼Œåˆ™è®¤ä¸ºæ˜¯G1æœŸã€‚ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 str(cc.genes) # List of 2 # $ s.genes : chr [1:43] \"MCM5\" \"PCNA\" \"TYMS\" \"FEN1\" ... # $ g2m.genes: chr [1:54] \"HMGB2\" \"CDK1\" \"NUSAP1\" \"UBE2C\" ... ## CellCycleScoring()å‡½æ•°åˆ†æç»†èƒå‘¨æœŸ pbmc = CellCycleScoring(pbmc, s.features = cc.genes$s.genes, g2m.features = cc.genes$g2m.genes) head(pbmc@meta.data[,c(\"S.Score\", \"G2M.Score\", \"Phase\")]) # S.Score G2M.Score Phase # AAACATACAACCAC-1 0.07721641 -0.027546191 S # AAACATTGAGCTAC-1 -0.02723808 -0.038481308 G1 # AAACATTGATCAGC-1 -0.01832618 0.069915003 G2M # AAACCGTGCTTCCG-1 0.01548134 0.008207127 S # AAACCGTGTATGCG-1 -0.05922341 0.031726222 G2M # AAACGCACTGGTAC-1 -0.05420895 -0.063251247 G1 table(pbmc$Phase) # G1 G2M S # 1437 544 719 ç»“æœå¯è§†åŒ– 1 2 3 4 5 6 7 8 plot(pbmc$S.Score, pbmc$G2M.Score, col=factor(pbmc$Phase), main=\"CellCycleScoring\") legend(\"topleft\", inset=0.05, title = \"cell cycle\", c(\"G1\",\"S\",\"G2M\"), pch = c(1), col=c(\"black\",\"green\",\"red\")) 1 2 DimPlot(pbmc, reduction = \"pca\", group.by = \"Phase\") SeuratåŒ…æä¾›çš„cc.genesä¸ºäººç±»åŸºå› åã€‚å¦‚æœéœ€åˆ†æå°é¼ æ•°æ®ï¼Œå¯å°†å…¶è½¬æ¢ä¸ºé¼ æºåŸºå› åã€‚\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 ## äººå·¥è½¬æ¢ï¼ˆè€ƒéªŒç½‘ç»œï¼‰ convertHumanGeneList \u003c- function(x){ require(\"biomaRt\") human = useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\") mouse = useMart(\"ensembl\", dataset = \"mmusculus_gene_ensembl\") genesV2 = getLDS(attributes = c(\"hgnc_symbol\"), filters = \"hgnc_symbol\", values = x , mart = human, attributesL = c(\"mgi_symbol\"), martL = mouse, uniqueRows=T) humanx \u003c- unique(genesV2[, 2]) print(head(humanx)) return(humanx) } m.s.genes \u003c- convertHumanGeneList(cc.genes$s.genes) m.g2m.genes \u003c- convertHumanGeneList(cc.genes$g2m.genes) ## githubä¸‹è½½ # https://github.com/satijalab/seurat/issues/462 ä¸‰ã€scranåŒ…åˆ†æ scranåŒ…ç»†èƒå‘¨æœŸåˆ†æçš„åŸºæœ¬åŸç†æ˜¯ gene pairè¡¨è¾¾ç‰¹å¾æ¯”è¾ƒ é¦–å…ˆè¯¥åŒ…åˆ†åˆ«æä¾›äº†ä¸‰ç§ç»†èƒå‘¨æœŸçš„gene pairè¡¨è¾¾ç‰¹å¾ï¼Œå…¶ä¸­firståˆ—è¡¨è¾¾æ°´å¹³å‡é«˜äºsecondåˆ—ã€‚ ç„¶ååˆ†ææ¯ä¸ªå•ç»†èƒçš„ç›¸åº”gene pairè¡¨è¾¾æ°´å¹³ï¼Œæ¯”è¾ƒæœ€ç¬¦åˆå“ªä¸€ç±»ç»†èƒå‘¨æœŸçš„ç‰¹å¾ã€‚ 1 2 3 4 5 6 7 8 9 10 11 12 13 library(scran) hs.pairs = readRDS(system.file(\"exdata\", \"human_cycle_markers.rds\", package=\"scran\")) # mm.pairs = readRDS(system.file(\"exdata\", \"mouse_cycle_markers.rds\", package=\"scran\")) names(hs.pairs) # [1] \"G1\" \"S\" \"G2M\" head(hs.pairs$G2M) # first second # 1 ENSG00000100519 ENSG00000065135 # 2 ENSG00000100519 ENSG00000080345 # 3 ENSG00000100519 ENSG00000151914 # 4 ENSG00000100519 ENSG00000085224 # 5 ENSG00000100519 ENSG00000116133 # 6 ENSG00000100519 ENSG00000143924 å¦‚ä¸Šç¤ºæ„ï¼Œç‰¹å¾æ•°æ®çš„åŸºå› åä¸ºENSEMBLçš„IDæ ¼å¼ã€‚å¯æ ¹æ®éœ€è¦å¯¹å•ç»†èƒè¡¨è¾¾çŸ©é˜µçš„è¡Œåè¿›è¡Œè½¬æ¢ã€‚\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 mt_count = pbmc@assays$RNA@counts mt_count[1:4,1:4] # 4 x 4 sparse Matrix of class \"dgCMatrix\" # AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1 # MIR1302-10 . . . . # FAM138A . . . . # OR4F5 . . . . # RP11-34P13.7 . . . . str(rownames(mt_count)) # chr [1:32738] \"MIR1302-10\" \"FAM138A\" \"OR4F5\" \"RP11-34P13.7\" \"RP11-34P13.8\" ... ## IDè½¬æ¢ library(org.Hs.eg.db) gene_ids = AnnotationDbi::select(org.Hs.eg.db, keys=rownames(mt_count), columns=c(\"ENSEMBL\"), keytype=\"SYMBOL\") gene_ids = gene_ids %\u003e% na.omit() %\u003e% dplyr::distinct(SYMBOL, .keep_all = T) %\u003e% dplyr::distinct(ENSEMBL, .keep_all = T) mt_count = mt_count[gene_ids$SYMBOL,] rownames(mt_count) = gene_ids$ENSEMBL dim(mt_count) # [1] 18893 2700 ç»†èƒå‘¨æœŸåˆ†æ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 assignments = scran::cyclone(mt_count, pairs = hs.pairs) names(assignments) # [1] \"phases\" \"scores\" \"normalized.scores\" dim(assignments$normalized.scores) # [1] 2700 3 assignments$normalized.scores[1:3,] # G1 S G2M # 1 0.8118324 0.1865242 0.001643385 # 2 0.4174428 0.5427604 0.039796782 # 3 0.7157215 0.2842785 0.000000000 table(assignments$phases) # G1 G2M S # 2551 60 89 ç»“æœå¯è§†åŒ– 1 2 3 4 5 6 7 scores = t(assignments$scores) colnames(scores)=1:ncol(scores) library(pheatmap) pheatmap(scores, show_colnames = F, annotation_col=data.frame(phase=assignments$phases, row.names = 1:ncol(scores))) æœ€åæ¯”è¾ƒä¸€ä¸‹ä¸¤ç§æ–¹æ³•æ³¨é‡Šç»“æœçš„å·®å¼‚\n1 2 3 4 5 6 7 8 9 res_seurat = pbmc$Phase res_scran = assignments$phases tb = table(res_seurat, res_scran) # res_scran # res_seurat G1 G2M S # G1 1350 35 52 # G2M 505 17 22 # S 696 8 15 gplots::balloonplot(tb) ",
  "wordCount" : "1533",
  "inLanguage": "en",
  "datePublished": "2023-01-22T00:00:00Z",
  "dateModified": "2023-01-22T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/020%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F%E5%88%86%E6%9E%90/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="ä¸»é¡µ">
                    <span>ä¸»é¡µ</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="åˆ†ç±»">
                    <span>åˆ†ç±»</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="æ ‡ç­¾">
                    <span>æ ‡ç­¾</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="å½’æ¡£">
                    <span>å½’æ¡£</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="å…³äº">
                    <span>å…³äº</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="æœç´¢ (Alt &#43; /)" accesskey=/>
                    <span>æœç´¢</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/">åˆ†ç±»</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰</a></div>
    <h1 class="post-title">
      å•ç»†èƒåˆ†æå·¥å…·--ç»†èƒå‘¨æœŸåˆ†æ
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-22 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-22&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-22&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1533&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#%e4%b8%80%e7%bb%86%e8%83%9e%e5%91%a8%e6%9c%9f%e9%98%b6%e6%ae%b5" aria-label="ä¸€ã€ç»†èƒå‘¨æœŸé˜¶æ®µ">ä¸€ã€ç»†èƒå‘¨æœŸé˜¶æ®µ</a></li>
                    <li>
                        <a href="#%e4%ba%8cseurat%e5%8c%85%e5%88%86%e6%9e%90" aria-label="äºŒã€SeuratåŒ…åˆ†æ">äºŒã€SeuratåŒ…åˆ†æ</a></li>
                    <li>
                        <a href="#%e4%b8%89scran%e5%8c%85%e5%88%86%e6%9e%90" aria-label="ä¸‰ã€scranåŒ…åˆ†æ">ä¸‰ã€scranåŒ…åˆ†æ</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>A cell cycle is a series of events that takes place in a cell as it grows and divides.å³æè¿°ç»†èƒç”Ÿé•¿ã€åˆ†è£‚æ•´ä¸ªè¿‡ç¨‹ä¸­ç»†èƒå˜åŒ–è¿‡ç¨‹ã€‚æœ€é‡è¦çš„ä¸¤ä¸ªç‰¹ç‚¹å°±æ˜¯DNAå¤åˆ¶ã€åˆ†è£‚æˆä¸¤ä¸ªä¸€æ ·çš„å­ç»†èƒã€‚å¯¹äºå•ç»†èƒè½¬å½•ç»„æ•°æ®ï¼Œå¯æ ¹æ®ç›¸å…³markeråŸºå› çš„è¡¨è¾¾æ°´å¹³åˆ¤æ–­æ¯ä¸€ä¸ªç»†èƒæ‰€å¤„çš„ç»†èƒå‘¨æœŸçŠ¶æ€ã€‚æ­¤å¤–åœ¨åˆ†æå¤šæ•°æ®é›†é—´æ‰¹æ¬¡æ•ˆåº”æ—¶ï¼Œå¯æ ¹æ®æ¯ä¸ªæ•°æ®é›†ä¸­å„ä¸ªç»†èƒå‘¨æœŸæ¯”ä¾‹è¿›è¡Œåˆ¤æ–­ä¸æ ¡æ­£ã€‚</p></blockquote>
<h2 id="ä¸€ç»†èƒå‘¨æœŸé˜¶æ®µ">ä¸€ã€ç»†èƒå‘¨æœŸé˜¶æ®µ<a hidden class="anchor" aria-hidden="true" href="#ä¸€ç»†èƒå‘¨æœŸé˜¶æ®µ">#</a></h2>
<p>ç»†èƒå‘¨æœŸç›¸å…³åŸºç¡€çŸ¥è¯†å…·ä½“å¯å‚è€ƒå¦‚ä¸‹é“¾æ¥ï¼š</p>
<blockquote>
<p><a href="https://teachmephysiology.com/biochemistry/cell-growth-death/cell-cycle/">https://teachmephysiology.com/biochemistry/cell-growth-death/cell-cycle/</a>
<a href="https://www.genome.gov/genetics-glossary/Cell-Cycle">https://www.genome.gov/genetics-glossary/Cell-Cycle</a>
<a href="https://search.bilibili.com/all?keyword=%E7%BB%86%E8%83%9E%E5%91%A8%E6%9C%9F&amp;from_source=webtop_search&amp;spm_id_from=333.851">ç»†èƒå‘¨æœŸ _ æœç´¢ç»“æœ_å“”å“©å“”å“©_Bilibili</a></p></blockquote>
<p>ç®€å•æ¥è¯´ï¼Œä¸€èˆ¬å¯åˆ†æˆ4ä¸ªé˜¶æ®µ</p>
<ul>
<li><strong>G1(gap1)</strong>ï¼šCell increases in size(Cellular contents duplicated)</li>
<li><strong>S(synthesis)</strong> ï¼šDNA replication, each of the 46 chromosomes (23 pairs) is replicated by the cell</li>
<li><strong>G2(gap2)</strong>ï¼šCell grows moreï¼Œorganelles and proteins develop in preparation for cell division</li>
<li><strong>M(mitosis)</strong>ï¼š&lsquo;Old&rsquo; cell partitions the two copies of the genetic material into the two daughter cells.  And the cell cycle can begin again.</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-c27d90b93e73eca9.png" alt="cell cycle" style="zoom: 50%;" />
<p>åœ¨å¤„ç†å•ç»†èƒæ•°æ®æ—¶ï¼Œåˆ†æç»†èƒå‘¨æœŸå¯å¯¹äºç”Ÿç‰©æœºåˆ¶æ¢ç´¢æˆ–è€…æ½œåœ¨æ‰¹æ¬¡æ•ˆåº”åˆ¤æ–­æä¾›å¦ä¸€ç§è§’åº¦çš„è§è§£ã€‚ç”±äºç»†èƒå‘¨æœŸä¹Ÿæ˜¯é€šè¿‡cell cycle related protein è°ƒæ§ï¼Œå³æ¯ä¸ªé˜¶æ®µæœ‰æ˜¾è‘—çš„markeråŸºå› ï¼›é€šè¿‡åˆ†æç»†èƒå‘¨æœŸæœ‰å…³åŸºå› çš„è¡¨è¾¾æƒ…å†µï¼Œå¯ä»¥å¯¹ç»†èƒæ‰€å¤„å‘¨æœŸé˜¶æ®µè¿›è¡Œæ³¨é‡Šã€‚</p>
<blockquote>
<p>Noteï¼šåœ¨å•ç»†èƒå‘¨æœŸåˆ†ææ—¶ï¼Œé€šå¸¸åªè€ƒè™‘ä¸‰ä¸ªé˜¶æ®µï¼šG1ã€Sã€G2Mã€‚(å³æŠŠG2å’ŒMå½“åšä¸€ä¸ªphase)</p></blockquote>
<p>å¦‚ä¸‹ç®€å•å­¦ä¹ åˆ†æå•ç»†èƒç»†èƒå‘¨æœŸçš„ä¸¤ç§æ–¹æ³•ï¼Œç¤ºä¾‹æ•°æ®é›†æ¥è‡ª10Xçš„PBMCæ ·æœ¬</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># download.file(&#34;https://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz&#34;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#               &#34;pbmc3k_filtered_gene_bc_matrices.tar.gz&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># untar(&#34;pbmc3k_filtered_gene_bc_matrices.tar.gz&#34;)</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>pbmc.data = Read10X(data.dir = <span style="color:#0ff;font-weight:bold">&#34;filtered_gene_bc_matrices/hg19/&#34;</span>)
</span></span><span style="display:flex;"><span>pbmc = CreateSeuratObject(counts = pbmc.data, project = <span style="color:#0ff;font-weight:bold">&#34;pbmc3k&#34;</span>)
</span></span><span style="display:flex;"><span>pbmc = NormalizeData(pbmc) %&gt;% 
</span></span><span style="display:flex;"><span>  FindVariableFeatures()
</span></span><span style="display:flex;"><span>pbmc = ScaleData(pbmc) %&gt;% 
</span></span><span style="display:flex;"><span>  RunPCA() %&gt;% 
</span></span><span style="display:flex;"><span>  RunUMAP(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>)
</span></span><span style="display:flex;"><span>DimPlot(pbmc, reduction = <span style="color:#0ff;font-weight:bold">&#34;pca&#34;</span>)  
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230122103725712.png" alt="image-20230122103725712" style="zoom: 50%;" />
<h2 id="äºŒseuratåŒ…åˆ†æ">äºŒã€SeuratåŒ…åˆ†æ<a hidden class="anchor" aria-hidden="true" href="#äºŒseuratåŒ…åˆ†æ">#</a></h2>
<ul>
<li>SeuratåŒ…ç»†èƒå‘¨æœŸåˆ†æçš„åŸºæœ¬åŸç†ç±»ä¼¼æ¨¡å—æ‰“åˆ†ã€‚</li>
<li>é¦–å…ˆè¯¥åŒ…å®šä¹‰äº†SæœŸä¸G2MæœŸçš„åŸºå› æ¨¡å—ï¼Œåˆ†åˆ«å¯¹æ¯ä¸ªå•ç»†èƒè¿›è¡Œæ‰“åˆ†ã€‚</li>
<li>è‹¥SæœŸåŸºå› æ¨¡å—åˆ†æ•°é«˜ï¼Œåˆ™è®¤ä¸ºç»†èƒå¤„äºSæœŸï¼›åŒç†åˆ†æG2MæœŸã€‚è‹¥äºŒè€…åˆ†æ•°å‡è¾ƒä½ï¼Œåˆ™è®¤ä¸ºæ˜¯G1æœŸã€‚</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>str(cc.genes)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># List of 2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   $ s.genes  : chr [1:43] &#34;MCM5&#34; &#34;PCNA&#34; &#34;TYMS&#34; &#34;FEN1&#34; ...</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   $ g2m.genes: chr [1:54] &#34;HMGB2&#34; &#34;CDK1&#34; &#34;NUSAP1&#34; &#34;UBE2C&#34; ...</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## CellCycleScoring()å‡½æ•°åˆ†æç»†èƒå‘¨æœŸ</span>
</span></span><span style="display:flex;"><span>pbmc = CellCycleScoring(pbmc,
</span></span><span style="display:flex;"><span>                        s.features = cc.genes$s.genes,
</span></span><span style="display:flex;"><span>                        g2m.features = cc.genes$g2m.genes)
</span></span><span style="display:flex;"><span>head(pbmc@meta.data[,c(<span style="color:#0ff;font-weight:bold">&#34;S.Score&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;G2M.Score&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Phase&#34;</span>)])
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                      S.Score    G2M.Score Phase</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATACAACCAC-1  0.07721641 -0.027546191     S</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATTGAGCTAC-1 -0.02723808 -0.038481308    G1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATTGATCAGC-1 -0.01832618  0.069915003   G2M</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCGTGCTTCCG-1  0.01548134  0.008207127     S</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCGTGTATGCG-1 -0.05922341  0.031726222   G2M</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACGCACTGGTAC-1 -0.05420895 -0.063251247    G1</span>
</span></span><span style="display:flex;"><span>table(pbmc$Phase)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   G1  G2M    S </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1437  544  719 </span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>ç»“æœå¯è§†åŒ–</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>plot(pbmc$S.Score,
</span></span><span style="display:flex;"><span>     pbmc$G2M.Score,
</span></span><span style="display:flex;"><span>     col=factor(pbmc$Phase),
</span></span><span style="display:flex;"><span>     main=<span style="color:#0ff;font-weight:bold">&#34;CellCycleScoring&#34;</span>)
</span></span><span style="display:flex;"><span>legend(<span style="color:#0ff;font-weight:bold">&#34;topleft&#34;</span>, inset=<span style="color:#ff0;font-weight:bold">0.05</span>,
</span></span><span style="display:flex;"><span>       title = <span style="color:#0ff;font-weight:bold">&#34;cell cycle&#34;</span>,  
</span></span><span style="display:flex;"><span>       c(<span style="color:#0ff;font-weight:bold">&#34;G1&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;S&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;G2M&#34;</span>), 
</span></span><span style="display:flex;"><span>       pch = c(<span style="color:#ff0;font-weight:bold">1</span>), col=c(<span style="color:#0ff;font-weight:bold">&#34;black&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;green&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;red&#34;</span>))
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230122104215078.png" alt="image-20230122104215078" style="zoom:50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>DimPlot(pbmc, reduction = <span style="color:#0ff;font-weight:bold">&#34;pca&#34;</span>, 
</span></span><span style="display:flex;"><span>        group.by = <span style="color:#0ff;font-weight:bold">&#34;Phase&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230122104322930.png" alt="image-20230122104322930" style="zoom:67%;" />
<blockquote>
<p>SeuratåŒ…æä¾›çš„<code>cc.genes</code>ä¸ºäººç±»åŸºå› åã€‚å¦‚æœéœ€åˆ†æå°é¼ æ•°æ®ï¼Œå¯å°†å…¶è½¬æ¢ä¸ºé¼ æºåŸºå› åã€‚</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## äººå·¥è½¬æ¢ï¼ˆè€ƒéªŒç½‘ç»œï¼‰</span>
</span></span><span style="display:flex;"><span>convertHumanGeneList &lt;- <span style="color:#fff;font-weight:bold">function</span>(x){
</span></span><span style="display:flex;"><span>  require(<span style="color:#0ff;font-weight:bold">&#34;biomaRt&#34;</span>)
</span></span><span style="display:flex;"><span>  human = useMart(<span style="color:#0ff;font-weight:bold">&#34;ensembl&#34;</span>, dataset = <span style="color:#0ff;font-weight:bold">&#34;hsapiens_gene_ensembl&#34;</span>)
</span></span><span style="display:flex;"><span>  mouse = useMart(<span style="color:#0ff;font-weight:bold">&#34;ensembl&#34;</span>, dataset = <span style="color:#0ff;font-weight:bold">&#34;mmusculus_gene_ensembl&#34;</span>)
</span></span><span style="display:flex;"><span>  genesV2 = getLDS(attributes = c(<span style="color:#0ff;font-weight:bold">&#34;hgnc_symbol&#34;</span>), filters = <span style="color:#0ff;font-weight:bold">&#34;hgnc_symbol&#34;</span>, values = x , mart = human, attributesL = c(<span style="color:#0ff;font-weight:bold">&#34;mgi_symbol&#34;</span>), martL = mouse, uniqueRows=<span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>  humanx &lt;- unique(genesV2[, <span style="color:#ff0;font-weight:bold">2</span>])
</span></span><span style="display:flex;"><span>  print(head(humanx))
</span></span><span style="display:flex;"><span>  <span style="color:#fff;font-weight:bold">return</span>(humanx)
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>m.s.genes &lt;- convertHumanGeneList(cc.genes$s.genes)
</span></span><span style="display:flex;"><span>m.g2m.genes &lt;- convertHumanGeneList(cc.genes$g2m.genes)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## githubä¸‹è½½</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># https://github.com/satijalab/seurat/issues/462</span>
</span></span></code></pre></td></tr></table>
</div>
</div></blockquote>
<h2 id="ä¸‰scranåŒ…åˆ†æ">ä¸‰ã€scranåŒ…åˆ†æ<a hidden class="anchor" aria-hidden="true" href="#ä¸‰scranåŒ…åˆ†æ">#</a></h2>
<ul>
<li>scranåŒ…ç»†èƒå‘¨æœŸåˆ†æçš„åŸºæœ¬åŸç†æ˜¯ gene pairè¡¨è¾¾ç‰¹å¾æ¯”è¾ƒ</li>
<li>é¦–å…ˆè¯¥åŒ…åˆ†åˆ«æä¾›äº†ä¸‰ç§ç»†èƒå‘¨æœŸçš„gene pairè¡¨è¾¾ç‰¹å¾ï¼Œå…¶ä¸­firståˆ—è¡¨è¾¾æ°´å¹³å‡é«˜äºsecondåˆ—ã€‚</li>
<li>ç„¶ååˆ†ææ¯ä¸ªå•ç»†èƒçš„ç›¸åº”gene pairè¡¨è¾¾æ°´å¹³ï¼Œæ¯”è¾ƒæœ€ç¬¦åˆå“ªä¸€ç±»ç»†èƒå‘¨æœŸçš„ç‰¹å¾ã€‚</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(scran)
</span></span><span style="display:flex;"><span>hs.pairs = readRDS(system.file(<span style="color:#0ff;font-weight:bold">&#34;exdata&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;human_cycle_markers.rds&#34;</span>, package=<span style="color:#0ff;font-weight:bold">&#34;scran&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># mm.pairs = readRDS(system.file(&#34;exdata&#34;, &#34;mouse_cycle_markers.rds&#34;, package=&#34;scran&#34;))</span>
</span></span><span style="display:flex;"><span>names(hs.pairs)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;G1&#34;  &#34;S&#34;   &#34;G2M&#34;</span>
</span></span><span style="display:flex;"><span>head(hs.pairs$G2M)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#             first          second</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 ENSG00000100519 ENSG00000065135</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 ENSG00000100519 ENSG00000080345</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 ENSG00000100519 ENSG00000151914</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 ENSG00000100519 ENSG00000085224</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5 ENSG00000100519 ENSG00000116133</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6 ENSG00000100519 ENSG00000143924</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>å¦‚ä¸Šç¤ºæ„ï¼Œç‰¹å¾æ•°æ®çš„åŸºå› åä¸ºENSEMBLçš„IDæ ¼å¼ã€‚å¯æ ¹æ®éœ€è¦å¯¹å•ç»†èƒè¡¨è¾¾çŸ©é˜µçš„è¡Œåè¿›è¡Œè½¬æ¢ã€‚</p></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>mt_count = pbmc@assays$RNA@counts
</span></span><span style="display:flex;"><span>mt_count[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 x 4 sparse Matrix of class &#34;dgCMatrix&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#              AAACATACAACCAC-1 AAACATTGAGCTAC-1 AAACATTGATCAGC-1 AAACCGTGCTTCCG-1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MIR1302-10                  .                .                .                .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FAM138A                     .                .                .                .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># OR4F5                       .                .                .                .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># RP11-34P13.7                .                .                .                .</span>
</span></span><span style="display:flex;"><span>str(rownames(mt_count))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># chr [1:32738] &#34;MIR1302-10&#34; &#34;FAM138A&#34; &#34;OR4F5&#34; &#34;RP11-34P13.7&#34; &#34;RP11-34P13.8&#34; ...</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## IDè½¬æ¢</span>
</span></span><span style="display:flex;"><span>library(org.Hs.eg.db)
</span></span><span style="display:flex;"><span>gene_ids = AnnotationDbi::select(org.Hs.eg.db, 
</span></span><span style="display:flex;"><span>                                 keys=rownames(mt_count), 
</span></span><span style="display:flex;"><span>                                 columns=c(<span style="color:#0ff;font-weight:bold">&#34;ENSEMBL&#34;</span>), 
</span></span><span style="display:flex;"><span>                                 keytype=<span style="color:#0ff;font-weight:bold">&#34;SYMBOL&#34;</span>)
</span></span><span style="display:flex;"><span>gene_ids = gene_ids %&gt;% 
</span></span><span style="display:flex;"><span>  na.omit() %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::distinct(SYMBOL, .keep_all = <span style="color:#fff;font-weight:bold">T</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::distinct(ENSEMBL, .keep_all = <span style="color:#fff;font-weight:bold">T</span>) 
</span></span><span style="display:flex;"><span>mt_count = mt_count[gene_ids$SYMBOL,]
</span></span><span style="display:flex;"><span>rownames(mt_count) = gene_ids$ENSEMBL
</span></span><span style="display:flex;"><span>dim(mt_count)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 18893  2700</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>ç»†èƒå‘¨æœŸåˆ†æ</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>assignments = scran::cyclone(mt_count,
</span></span><span style="display:flex;"><span>                      pairs = hs.pairs)
</span></span><span style="display:flex;"><span>names(assignments)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;phases&#34;            &#34;scores&#34;            &#34;normalized.scores&#34;</span>
</span></span><span style="display:flex;"><span>dim(assignments$normalized.scores)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 2700    3</span>
</span></span><span style="display:flex;"><span>assignments$normalized.scores[1:<span style="color:#ff0;font-weight:bold">3</span>,]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#          G1         S         G2M</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 0.8118324 0.1865242 0.001643385</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 0.4174428 0.5427604 0.039796782</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 0.7157215 0.2842785 0.000000000</span>
</span></span><span style="display:flex;"><span>table(assignments$phases)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   G1  G2M    S </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2551   60   89 </span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>ç»“æœå¯è§†åŒ–</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>scores = t(assignments$scores)
</span></span><span style="display:flex;"><span>colnames(scores)=<span style="color:#ff0;font-weight:bold">1</span>:ncol(scores)
</span></span><span style="display:flex;"><span>library(pheatmap)
</span></span><span style="display:flex;"><span>pheatmap(scores,
</span></span><span style="display:flex;"><span>         show_colnames = <span style="color:#fff;font-weight:bold">F</span>,
</span></span><span style="display:flex;"><span>         annotation_col=data.frame(phase=assignments$phases,
</span></span><span style="display:flex;"><span>                                   row.names = <span style="color:#ff0;font-weight:bold">1</span>:ncol(scores)))
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230122112355699.png" alt="image-20230122112355699" style="zoom: 50%;" />
<hr>
<p>æœ€åæ¯”è¾ƒä¸€ä¸‹ä¸¤ç§æ–¹æ³•æ³¨é‡Šç»“æœçš„å·®å¼‚</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>res_seurat = pbmc$Phase
</span></span><span style="display:flex;"><span>res_scran = assignments$phases
</span></span><span style="display:flex;"><span>tb = table(res_seurat, res_scran)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#           res_scran</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># res_seurat   G1  G2M    S</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        G1  1350   35   52</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        G2M  505   17   22</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        S    696    8   15</span>
</span></span><span style="display:flex;"><span>gplots::balloonplot(tb)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230122113111502.png" alt="image-20230122113111502" style="zoom:67%;" />


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/program/015python-lambda%E5%87%BD%E6%95%B0%E8%A1%A8%E8%BE%BE%E5%BC%8F/">
    <span class="title">Â« Prev Page</span>
    <br>
    <span>Python-lambdaå‡½æ•°è¡¨è¾¾å¼</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/program/016python-threading%E5%A4%9A%E7%BA%BF%E7%A8%8B/">
    <span class="title">Next Page Â»</span>
    <br>
    <span>Python-threadingå¤šçº¿ç¨‹</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>æ‚¨æ˜¯æœ¬ç«™ç¬¬ <span id="busuanzi_value_site_uv"></span> ä½è®¿é—®è€…ï¼Œæ€»æµè§ˆé‡ä¸º <span id="busuanzi_value_site_pv"></span> æ¬¡
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
