<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰ | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="">
<meta name="description" content="ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰ - Li&#39;s Bioinfo-Blog">
<meta name="author" content="">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" type="application/rss+xml" href="https://lishensuo.github.io/en/posts/bioinfo/index.xml">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰" />
<meta property="og:description" content="æç”³é”çš„åšå®¢ï¼Œè®°å½•ç”Ÿä¿¡å­¦ä¹ è¿‡ç¨‹ã€‚R/Python/Shell" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰"/>
<meta name="twitter:description" content="æç”³é”çš„åšå®¢ï¼Œè®°å½•ç”Ÿä¿¡å­¦ä¹ è¿‡ç¨‹ã€‚R/Python/Shell"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "åˆ†ç±»",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }
  ]
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="list" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="ä¸»é¡µ">
                    <span>ä¸»é¡µ</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="åˆ†ç±»">
                    <span>åˆ†ç±»</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="æ ‡ç­¾">
                    <span>æ ‡ç­¾</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="å½’æ¡£">
                    <span>å½’æ¡£</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="å…³äº">
                    <span>å…³äº</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="æœç´¢ (Alt &#43; /)" accesskey=/>
                    <span>æœç´¢</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main"> 
<header class="page-header"><div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/">åˆ†ç±»</a></div>
  <h1>ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰</h1>
</header>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>WGCNAåŸºå› åŠ æƒå…±è¡¨è¾¾ç½‘ç»œåˆ†æ
			</h2>
		</header>
		<section class="entry-content">
			<p>1ã€å…³äºWGCNAåŸç† 1.1 å»ºç«‹å…±è¡¨è¾¾ç½‘ç»œ åœ¨åŸºå› å…±è¡¨è¾¾ç½‘ç»œä¸­ï¼ŒèŠ‚ç‚¹nodeä»£è¡¨åŸºå› ï¼Œè¾¹edgeä»£è¡¨ä¸¤ä¸ªåŸºå› é—´å…±è¡¨è¾¾å…³ç³»ã€‚
è‹¥ä¸€ä¸ªåŸºå› åŒæ—¶ä¸å¤šä¸ªåŸºå› å­˜åœ¨ç›¸å…³æ€§ï¼Œç§°ä¸ºhubåŸºå› ã€‚ è‹¥ä¸€ç¾¤åŸºå› å­˜åœ¨é«˜åº¦äº’ç›¸ç›¸å…³ï¼Œç§°ä¸ºmoduleã€‚ åŸºå› å…±è¡¨è¾¾ç½‘ç»œçš„å±•ç¤ºå½¢å¼ä¸€èˆ¬ä¸º n Ã— n é‚»æ¥çŸ©é˜µadjacency matrix(nä¸ªåŸºå› )
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-25 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-25&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-05-06&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;5111&amp;nbsp;|&amp;nbsp;11 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to WGCNAåŸºå› åŠ æƒå…±è¡¨è¾¾ç½‘ç»œåˆ†æ" href="https://lishensuo.github.io/en/posts/bioinfo/071wgcna%E5%9F%BA%E5%9B%A0%E5%8A%A0%E6%9D%83%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C%E5%88%86%E6%9E%90/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>GSVAåŒ…å•æ ·æœ¬å¯Œé›†åˆ†æ
			</h2>
		</header>
		<section class="entry-content">
			<p>1ã€GSVAå‡½æ•° 1 2 3 4 5 6 7 # BiocManager::install(&#34;GSVA&#34;) library(GSVA) ?gsva gsva(expr = , #metrixæ ¼å¼è¡¨è¾¾çŸ©é˜µ(è¡Œ--åŸºå› ï¼Œåˆ—--æ ·æœ¬) gset.idx.list = , #listæ ¼å¼åŸºå› é›† method=c(&#34;gsva&#34;, &#34;ssgsea&#34;, &#34;zscore&#34;, &#34;plage&#34;), # defaul:gsva kcdf=c(&#34;Gaussian&#34;, &#34;Poisson&#34;, &#34;none&#34;)) # default:Gaussian å¦‚ä¸Šæ‰€ç¤ºï¼Œgsva()å‡½æ•°æœ‰4ä¸ªå…³é”®å‚æ•°ã€‚
å‰ä¸¤ä¸ªå‚æ•°åˆ†åˆ«æä¾›è¡¨è¾¾çŸ©é˜µä¸åŸºå› é›†ï¼Œå…¶ä¸­åŸºå› åéœ€è¦ä¸ºç›¸åŒçš„æ ¼å¼ã€‚ methodå‚æ•°è®¾ç½®æ–¹æ³•ï¼Œé»˜è®¤ä¸ºgsva kcdfå‚æ•°è®¾ç½®åˆ†å¸ƒæ¨¡æ‹Ÿç±»å‹ï¼Œé»˜è®¤ä¸ºGaussianï¼Œé€‚åˆäºarrayæ•°æ®ä»¥åŠlogæ ‡å‡†åŒ–(CPM/RPKM/TPMç­‰)çš„RNA-seqæ•°æ®ã€‚Poissonåˆ†å¸ƒé€‚åˆRNA-seqçš„åŸå§‹countçŸ©é˜µã€‚ 2ã€ç¤ºä¾‹æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 ## (1) è¡¨è¾¾çŸ©é˜µ # BiocManager::install(&#34;GSVAdata&#34;) library(GSVAdata) data(commonPickrellHuang) ### arrayçŸ©é˜µ array_mt = exprs(huangArrayRMAnoBatchCommon_eset) dim(array_mt) # [1] 11508 36 array_mt[1:4,1:4] # NA19099 NA18523 NA19144 NA19137 # 8567 8.370526 8.544890 8.270395 8.585984 # 23139 7.536228 7.232781 7.469420 7.801986 # 7580 6.325102 6.405961 6.510249 6.285510 # 55619 9.134495 9.049938 9.704679 9.285202 ### countçŸ©é˜µ count_mt = exprs(pickrellCountsArgonneCQNcommon_eset) dim(count_mt) # [1] 11508 36 count_mt[1:4,1:4] # NA19099 NA18523 NA19144 NA19137 # 8567 326 209 318 343 # 23139 255 169 245 361 # 7580 72 69 124 76 # 55619 487 590 678 540 ## (2) åŸºå› é›† library(tidyverse) library(msigdbr) hm_H = msigdbr(species = &#34;Homo sapiens&#34;, category = &#34;H&#34;, subcategory = NULL) %&gt;% as.data.frame() %&gt;% dplyr::select(gs_cat, gs_subcat, gs_name, entrez_gene) hm_H.list = split(hm_H$entrez_gene, hm_H$gs_name) str(head(hm_H.list)) # List of 6 # $ HALLMARK_ADIPOGENESIS : int [1:210] 19 11194 10449 33 34 35 47 50 51 112 ... # $ HALLMARK_ALLOGRAFT_REJECTION: int [1:335] 16 6059 10006 43 92 207 322 567 567 586 ... # $ HALLMARK_ANDROGEN_RESPONSE : int [1:102] 10257 11057 2181 87 9510 11047 9590 207 220 56172 ... # $ HALLMARK_ANGIOGENESIS : int [1:36] 350 351 894 1281 1290 6372 2260 11167 3685 182 ... # $ HALLMARK_APICAL_JUNCTION : int [1:231] 58 60 70 71 72 87 88 89 81 81 ... # $ HALLMARK_APICAL_SURFACE : int [1:46] 102 79602 79602 84632 9465 351 50617 5205 2683 672 ... 3ã€GSVAåˆ†æ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 ## (1) array &#43; gsva res_gsva = gsva(array_mt, hm_H.list) class(res_gsva) # [1] &#34;matrix&#34; &#34;array&#34; dim(res_gsva) # [1] 50 36 res_gsva[1:4,1:4] # NA19099 NA18523 NA19144 NA19137 # HALLMARK_ADIPOGENESIS -0.21147644 0.08157479 0.004672524 -0.1656925 # HALLMARK_ALLOGRAFT_REJECTION -0.32620137 -0.44105423 -0.452996166 -0.4288909 # HALLMARK_ANDROGEN_RESPONSE 0.02056377 0.09123573 -0.135386552 -0.2198959 # HALLMARK_ANGIOGENESIS -0.02435541 -0.18282244 0.281235763 0.5114103 ## (2) count &#43; ssgsea res_gsva = gsva(count_mt, hm_H.list, method = &#34;ssgsea&#34;, kcdf = &#34;Poisson&#34;) class(res_gsva) # [1] &#34;matrix&#34; &#34;array&#34; dim(res_gsva) # [1] 50 36 res_gsva[1:4,1:4] # NA19099 NA18523 NA19144 NA19137 # HALLMARK_ADIPOGENESIS 0.3288768 0.32899026 0.2893928 0.3297140 # HALLMARK_ALLOGRAFT_REJECTION 0.5018248 0.49183093 0.4672342 0.4531009 # HALLMARK_ANDROGEN_RESPONSE 0.3210905 0.28289473 0.2818011 0.2845855 # HALLMARK_ANGIOGENESIS 0.1794361 0.07705857 0.2076328 0.1276449 4ã€limmaå·®å¼‚åˆ†æ ä»¥ä¸Šé¢array &#43; gsvaçš„åˆ†æç»“æœä¸ºä¾‹ï¼Œéšæœºæ¨¡æ‹Ÿåˆ†ç»„è¿›è¡Œlimmaå·®å¼‚åˆ†æ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 ## æ¨¡æ‹Ÿåˆ†ç»„ meta = data.frame(sample=colnames(array_mt), group=rep(c(&#34;G1&#34;,&#34;G2&#34;), each=18)) head(meta) # sample group # 1 NA19099 G1 # 2 NA18523 G1 # 3 NA19144 G1 # 4 NA19137 G1 # 5 NA18861 G1 # 6 NA19116 G1 ## limmaå·®å¼‚åˆ†æ library(limma) mod &lt;- model.matrix(~ factor(meta$group)) colnames(mod) &lt;- c(&#34;G1&#34;, &#34;G2&#34;) fit &lt;- lmFit(res_gsva, mod) fit &lt;- eBayes(fit) tt &lt;- topTable(fit, coef=2, n=Inf) head(tt) # logFC AveExpr t P.Value adj.P.Val B # HALLMARK_COMPLEMENT 0.08684907 -0.10731154 2.495909 0.01676849 0.3324198 -3.059104 # HALLMARK_HEME_METABOLISM 0.07942485 -0.04250442 2.238926 0.03076698 0.3324198 -3.486635 # HALLMARK_IL2_STAT5_SIGNALING 0.07839683 -0.09276742 2.193525 0.03411795 0.3324198 -3.558694 # HALLMARK_MITOTIC_SPINDLE -0.10721385 -0.07656961 -2.057416 0.04617883 0.3324198 -3.768076 # HALLMARK_IL6_JAK_STAT3_SIGNALING 0.09342746 -0.20104143 2.026107 0.04943145 0.3324198 -3.814790 # HALLMARK_INFLAMMATORY_RESPONSE 0.10211588 -0.10571870 1.996939 0.05263925 0.3324198 -3.857810 </p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-15 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-15&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-15&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;872&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to GSVAåŒ…å•æ ·æœ¬å¯Œé›†åˆ†æ" href="https://lishensuo.github.io/en/posts/bioinfo/058gsva%E5%8C%85%E5%8D%95%E6%A0%B7%E6%9C%AC%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>xCellä¸CIBERSORTç­‰å…ç–«æµ¸æ¶¦åˆ†æ
			</h2>
		</header>
		<section class="entry-content">
			<p>xCell xCellåŒ… 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 devtools::install_github(&#39;dviraran/xCell&#39;) library(xCell) data(&#34;xCell.data&#34;) ##æŸ¥çœ‹æ”¯æŒçš„64ç§ç»†èƒç±»å‹ï¼ŒåŒä¸‹å›¾ colnames(xCell.data$spill$K) ##é¢„æµ‹å‡½æ•°çš„å…³é”®å‚æ•°è§£é‡Š ?xCellAnalysis() # expr = äº¤ä»£è¡¨è¾¾çŸ©é˜µï¼› ##å¦‚æœæ˜¯arrayï¼Œä¸éœ€è¦é¢å¤–æ ‡å‡†åŒ–ï¼›å¦‚æœæ˜¯RNAseqï¼Œéœ€è¦TPM/FPKM/TPMã€‚ ##å¯¹äºåŸºå› IDæ ¼å¼éœ€è¦æ˜¯symbolæ ¼å¼ã€‚ # rnaseq = TRUE æ•°æ®æ˜¯å¦ä¸ºRNAseqæ•°æ®ï¼Œå¦‚æœæ˜¯èŠ¯ç‰‡æ•°æ®è®¾ç½®ä¸ºFALSE # cell.types.use = NULL æä¾›ä¸€ä¸ªå­—ç¬¦ä¸²ï¼Œè¯´æ˜æƒ³è¦é¢„æµ‹64ç§ç»†èƒä¸­çš„å“ªäº›ç»†èƒç±»å‹ # parallel.sz = 4 è°ƒç”¨çš„çº¿ç¨‹æ•°ï¼Œé»˜è®¤ä¸º4 NOTEï¼š
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-25 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-25&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-02-18&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2554&amp;nbsp;|&amp;nbsp;6 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to xCellä¸CIBERSORTç­‰å…ç–«æµ¸æ¶¦åˆ†æ" href="https://lishensuo.github.io/en/posts/bioinfo/072xcell%E4%B8%8Ecibersort%E7%AD%89%E5%85%8D%E7%96%AB%E6%B5%B8%E6%B6%A6%E5%88%86%E6%9E%90/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>ClusterGVisåŒ…ç»˜åˆ¶åŸºå› è¡¨è¾¾çŸ©é˜µçƒ­å›¾
			</h2>
		</header>
		<section class="entry-content">
			<p> ClusterGVisåŒ…æ˜¯ä¸­å›½è¯ç§‘å¤§å­¦Jun Zhangåšå£«å¼€å‘çš„ç³»åˆ—å¯è§†åŒ–å·¥å…·åŒ…ä¹‹ä¸€ï¼Œå¯ä»¥åŸºå› è¡¨è¾¾çŸ©é˜µè¿›è¡Œé«˜çº§çš„çƒ­å›¾å¯è§†åŒ–åˆ†æã€‚å¦‚ä¸‹æ ¹æ®å…¶githubä»¥åŠå¾®ä¿¡æ•™ç¨‹ç®€å•æ•´ç†ä¸€ä¸‹è‡ªå·±æ„Ÿå…´è¶£çš„ç”¨æ³•ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-03-19 00:00:00 &#43;0000 UTC&#39;&gt;2023-03-19&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-03-19&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1261&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to ClusterGVisåŒ…ç»˜åˆ¶åŸºå› è¡¨è¾¾çŸ©é˜µçƒ­å›¾" href="https://lishensuo.github.io/en/posts/bioinfo/073clustergvis%E5%8C%85%E7%BB%98%E5%88%B6%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5%E7%83%AD%E5%9B%BE/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>survivalåŒ…ç”Ÿå­˜åˆ†æåŠglmnetåŒ…lassoå›å½’
			</h2>
		</header>
		<section class="entry-content">
			<p> ç”Ÿå­˜åˆ†æ(survival analysis)çš„ä¸»è¦ç›®çš„æ˜¯å‘ç°ä¸æ‚£è€…ç”Ÿå­˜äº‹ä»¶ç›¸å…³çš„æŒ‡æ ‡å› ç´ ï¼Œä¾‹å¦‚å¹´é¾„æ€§åˆ«ã€åŸºå› è¡¨è¾¾/çªå˜ç­‰ã€‚å¦‚ä¸‹å­¦ä¹ ç›¸å…³åŸºç¡€çŸ¥è¯†åŠå‡ ç§å¸¸è§çš„ç”Ÿå­˜åˆ†ææ–¹æ³•ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-28 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-28&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-29&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;3794&amp;nbsp;|&amp;nbsp;8 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to survivalåŒ…ç”Ÿå­˜åˆ†æåŠglmnetåŒ…lassoå›å½’" href="https://lishensuo.github.io/en/posts/bioinfo/059survival%E5%8C%85%E7%94%9F%E5%AD%98%E5%88%86%E6%9E%90%E5%8F%8Aglmnet%E5%8C%85lasso%E5%9B%9E%E5%BD%92/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>MuSiCåŒ…æ ¹æ®scRNAseqé¢„æµ‹Bulkç»†èƒç»„æˆ
			</h2>
		</header>
		<section class="entry-content">
			<p> MuSiC(MUlti-Subject SIngle Cell deconvolution)æ˜¯æ¥è‡ªå®¾å¤•æ³•å°¼äºšå¤§å­¦Biostatistics, Epidemiology and Informaticsç³»çš„Mingyao Liè¯¾é¢˜ç»„äº2019å¹´å‘è¡¨äºNature Communicationçš„ä¸€ä¸ªå·¥å…·RåŒ…ï¼Œå¯æ ¹æ®å•ç»†èƒè½¬å½•ç»„ä¿¡æ¯æ¨æµ‹Bulk RNA-seqç»†èƒç»„æˆã€‚è€Œåï¼Œè¯¥å›¢é˜Ÿåˆäº2022å¹´åœ¨Briefing in bioinformaticså‘è¡¨äº†æ‰©å±•ç‰ˆæœ¬MuSiC2ï¼Œå¯ä»¥è€ƒè™‘æ›´å¤æ‚çš„åœºæ™¯ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-03-26 00:00:00 &#43;0000 UTC&#39;&gt;2023-03-26&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-03-26&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1799&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to MuSiCåŒ…æ ¹æ®scRNAseqé¢„æµ‹Bulkç»†èƒç»„æˆ" href="https://lishensuo.github.io/en/posts/bioinfo/074music%E5%8C%85%E6%A0%B9%E6%8D%AEscrnaseq%E9%A2%84%E6%B5%8Bbulk%E7%BB%86%E8%83%9E%E7%BB%84%E6%88%90/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>ä»RNAseqçš„fastq.gzæå–è¡¨è¾¾çŸ©é˜µ
			</h2>
		</header>
		<section class="entry-content">
			<p> RNA-seqæ•°æ®æ¯”å¯¹æµç¨‹ä¸»è¦åˆ†ä¸ºä¸‰æ­¥ï¼ˆ1ï¼‰æ•´ç†æ•°æ®ï¼›ï¼ˆ2ï¼‰è´¨æ§ï¼›ï¼ˆ3ï¼‰æ¯”å¯¹ã€‚å…¶ä¸­æ¯ä¸€æ­¥éƒ½æ¶‰åŠåˆ°è‹¥å¹²è½¯ä»¶çš„ç”¨æ³•ï¼Œå¦‚ä¸‹ç®€å•æ•´ç†åŸºæœ¬çš„åˆ†ææµç¨‹ã€‚
ç¤ºä¾‹æ•°æ®ï¼šGSE158623ä¸­6ä¸ªæ ·æœ¬çš„RNA-seqæµ‹åºç»“æœ(human)ï¼Œå¯¹åº”SRR12720999~SRR12721004
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-29 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-29&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-29&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2347&amp;nbsp;|&amp;nbsp;5 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to ä»RNAseqçš„fastq.gzæå–è¡¨è¾¾çŸ©é˜µ" href="https://lishensuo.github.io/en/posts/bioinfo/060%E4%BB%8Ernaseq%E7%9A%84fastq.gz%E6%8F%90%E5%8F%96%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>gseapyå¯Œé›†åˆ†æä¸å¯è§†åŒ–
			</h2>
		</header>
		<section class="entry-content">
			<p>å‚è€ƒ
https://gseapy.readthedocs.io/en/latest/index.html ç¤ºä¾‹æ•°æ®æ¥è‡ª https://github.com/zqfang/GSEApy/tree/master/tests/data 1 2 3 4 5 6 7 8 import os import pandas as pd import scanpy as sc import gseapy as gp from gseapy import barplot, dotplot import matplotlib.pyplot as plt 0. åŸºå› é›†è·å– Enrichr API supported organisms: Human, Mouse, Yeast, Fly, Fish, Worm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 names = gp.get_library_name(organism = &#34;Human&#34;) len(names) # 217 [name for name in names if &#34;GO&#34; in name] # [&#39;GO_Biological_Process_2021&#39;, # &#39;GO_Biological_Process_2023&#39;, # &#39;GO_Biological_Process_2025&#39;, # &#39;GO_Cellular_Component_2021&#39;, # &#39;GO_Cellular_Component_2023&#39;, # &#39;GO_Cellular_Component_2025&#39;, # &#39;GO_Molecular_Function_2021&#39;, # &#39;GO_Molecular_Function_2023&#39;, # &#39;GO_Molecular_Function_2025&#39;, # &#39;SynGO_2022&#39;, # &#39;SynGO_2024&#39;] [name for name in names if &#34;KEGG&#34; in name] # [&#39;KEGG_2013&#39;, # &#39;KEGG_2015&#39;, # &#39;KEGG_2016&#39;, # &#39;KEGG_2019_Human&#39;, # &#39;KEGG_2019_Mouse&#39;, # &#39;KEGG_2021_Human&#39;] [name for name in names if &#34;MSigDB&#34; in name] # [&#39;MSigDB_Computational&#39;, &#39;MSigDB_Hallmark_2020&#39;, &#39;MSigDB_Oncogenic_Signatures&#39;] kegg = gp.get_library(name=&#39;KEGG_2021_Human&#39;, organism=&#39;Human&#39;) type(kegg) # dict kegg[&#39;AMPK signaling pathway&#39;][:4] # [&#39;RAB2A&#39;, &#39;PPP2R1A&#39;, &#39;TSC2&#39;, &#39;TSC1&#39;] 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 # è¿˜æ”¯æŒå¯¹MSigDBåŸºå› é›†æ•°æ®åº“çš„è·å– from gseapy import Msigdb msig = Msigdb() # all available versions msig.list_dbver() # all types of gene sets given a dbver msig.list_category(dbver=&#34;2025.1.Hs&#34;) # all gene sets of one type gmt = msig.get_gmt(category=&#39;h.all&#39;, dbver=&#34;2025.1.Hs&#34;) type(gmt) # dict gmt.keys() # hallmark pathway names gmt[&#34;HALLMARK_ADIPOGENESIS&#34;] # gene list 1. ORAå¯Œé›† 1.1 åˆ†æ Online analysis 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 ## 1) ä¸¤ä¸ªæœ€é‡è¦çš„å‚æ•° ï¼ˆååŒï¼‰ # gene_list gene_list=&#34;tests/data/gene_list.txt&#34; # file path, one gene per row gene_list= [&#34;RAB2A&#34;,&#34;PRKAB2&#34;,&#34;PRKAA1&#34;] # list of genes # gene_sets (from enrichr API) gene_sets=&#39;KEGG_2016&#39; gene_sets=[&#39;KEGG_2021_Human&#39;,&#39;GO_Biological_Process_2025&#39;] gene_list = pd.read_csv(&#34;tests/data/gene_list.txt&#34;, header=None)[0].tolist() gene_list[:4] # [&#39;IGKV4-1&#39;, &#39;CD55&#39;, &#39;IGKC&#39;, &#39;PPFIBP1&#39;] enr = gp.enrichr(gene_list=gene_list, # or &#34;./tests/data/gene_list.txt&#34;, gene_sets=[&#39;GO_Biological_Process_2025&#39;], organism=&#39;human&#39;, outdir=None, ) enr.results.head(5) # Index([&#39;Gene_set&#39;, &#39;Term&#39;, &#39;Overlap&#39;, &#39;P-value&#39;, &#39;Adjusted P-value&#39;, # &#39;Old P-value&#39;, &#39;Old Adjusted P-value&#39;, &#39;Odds Ratio&#39;, &#39;Combined Score&#39;, # &#39;Genes&#39;], # dtype=&#39;object&#39;) ## 2) è®¾ç½®ç»“æœå¯¼å‡ºè·¯å¾„ enr = gp.enrichr(gene_list=gene_list, gene_sets=[&#39;GO_Biological_Process_2025&#39;], organism=&#39;human&#39;, outdir=&#34;enrichr_kegg&#34;, ) os.listdir(&#34;enrichr_kegg&#34;) # [&#39;GO_Biological_Process_2025.human.enrichr.reports.txt&#39;, # &#39;GO_Biological_Process_2025.human.enrichr.reports.pdf&#39;] ## 3) è‡ªå®šä¹‰Background genes (På€¼ä¼šå­˜åœ¨ä¸€å®šå·®å¼‚) enr = gp.enrichr(gene_list=gene_list, gene_sets=[&#39;GO_Biological_Process_2025&#39;], organism=&#39;human&#39;, background=&#34;tests/data/background.txt&#34;, # è‡ªå®šä¹‰èƒŒæ™¯åŸºå›  / or list object outdir=None, ) Offline analysis (æœ¬åœ°æä¾›é€šè·¯åŸºå› é›†æ•°æ®) 1 2 3 4 5 6 7 8 9 10 11 from gseapy.parser import read_gmt gmt_dict = read_gmt(&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;) gmt_dict.keys() enr2 = gp.enrich(gene_list=gene_list, # or gene_list=glist gene_sets=[&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;], background=None, outdir=None, verbose=True) enr2.results.head(5) 1.2 å¯è§†åŒ– 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 # ofname è®¾ç½®å›¾ç‰‡çš„å¯¼å‡ºè·¯å¾„ï¼Œè‹¥ä¸ºNoneåˆ™ä¸ä¿å­˜ã€‚ä¹Ÿæ”¯æŒåç»­ä½¿ç”¨pltè¯­æ³•è¿›ä¸€æ­¥è°ƒæ•´å›¾ç‰‡ ax = dotplot(enr.results, column=&#34;Adjusted P-value&#34;, x=&#39;Gene_set&#39;, # set x axis, so you could do a multi-sample/library comparsion size=5, cutoff=0.05, top_term=10, figsize=(3, 5), title = &#34;GO_BP&#34;, xticklabels_rot=45, # rotate xtick labels show_ring=True, # set to False to revmove outer ring marker=&#39;o&#39;, ofname=&#39;./plot/p1.png&#39; ) ax = barplot(enr.results, column=&#34;Adjusted P-value&#34;, group=&#39;Gene_set&#39;, # set group, so you could do a multi-sample/library comparsion size=10, top_term=10, figsize=(3,5), #color=[&#39;darkred&#39;, &#39;darkblue&#39;] # set colors for group color = {&#39;GO_Biological_Process_2025&#39;: &#39;salmon&#39;}, ofname=&#39;./plot/p2.png&#39; ) 1 2 3 ax = dotplot(enr.res2d, title=&#39;Enrichr&#39;,cmap=&#39;viridis_r&#39;, size=10, figsize=(3,5), ofname=&#39;./plot/p3.png&#39;) ax = barplot(enr.res2d, title=&#39;Enrichr&#39;,cmap=&#39;viridis_r&#39;, size=4, figsize=(4, 5), color=&#39;darkred&#39;, ofname=&#39;./plot/p4.png&#39;) # å·®åˆ«ä¸å¤§ 2. Prerankå¯Œé›† åŸºäºå·®å¼‚åˆ†æç»“æœçš„GSEA
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2025-07-22 00:00:00 &#43;0000 UTC&#39;&gt;2025-07-22&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2025-07-22&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1367&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to gseapyå¯Œé›†åˆ†æä¸å¯è§†åŒ–" href="https://lishensuo.github.io/en/posts/bioinfo/061gseapy%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>ç”Ÿä¿¡ç›¸å…³ç½‘ç«™æ•°æ®åº“é›†é”¦
			</h2>
		</header>
		<section class="entry-content">
			<p>1ã€HPA HPAï¼šthe Human Protein Atlasï¼Œç”±2003å¹´æ¥è‡ªç‘å…¸çš„ç§‘ç ”æœºæ„å‘èµ·ï¼Œæ—¨åœ¨ç»˜åˆ¶ç»¼åˆæ€§äººç±»è›‹ç™½è´¨å›¾è°±ã€‚ https://www.proteinatlas.org/ è›‹ç™½è´¨è¡¨è¾¾æ•°æ®åº“ï¼Œå¸¸è§ç”¨é€”åŒ…æ‹¬ï¼š ï¼ˆ1ï¼‰æŸ¥çœ‹ç‰¹å®šåŸºå› åœ¨ä¸åŒç»„ç»‡ã€è„‘åŒºï¼Œç»†èƒç±»å‹ï¼Œç»„ç»‡ç»†èƒç±»å‹ï¼Œç–¾ç—…ï¼ˆç™Œç—‡ï¼‰ï¼Œå…ç–«ç»†èƒï¼Œè‚¿ç˜¤ç»†èƒç³»ç­‰è¡¨è¾¾æƒ…å†µã€‚ ï¼ˆ2ï¼‰æ•°æ®æŒ–æ˜ç±»æ–‡ç« å¸¸ä½¿ç”¨å…¶è¿›è¡Œæ¯”è¾ƒåŸºå› åœ¨è‚¿ç˜¤éƒ¨ä½ä¸ç›¸åº”æ­£å¸¸éƒ¨ä½çš„è›‹ç™½æ°´å¹³è¡¨è¾¾å·®å¼‚ã€‚ 2ã€ENCORI ENCORIï¼ŒThe Encyclopedia of RNA Interactomesï¼Œç”±ä¸­å±±å¤§å­¦ç”Ÿå‘½ç§‘å­¦å­¦é™¢å±ˆè‰¯é¹„å›¢é˜Ÿå¼€å‘ï¼Œäº2014å¹´å‘è¡¨äºNucleic Acids Resã€‚ https://starbase.sysu.edu.cn/index.php miRNAã€lncRNAã€RBPç­‰å¤šç»´ç›¸äº’ä½œç”¨ç½‘ç»œï¼Œå¸¸è§ç”¨é€”åŒ…æ‹¬ miRNA/RBPçš„é¶æ ‡æŸ¥è¯¢(mRNA,lncRNAâ€¦) RNA interaction, ceRNA network TCGAè‚¿ç˜¤çš„å·®å¼‚è¡¨è¾¾ï¼Œç”Ÿå­˜åˆ†æï¼Œç›¸å…³æ€§ç­‰ 3ã€TCIA TCIAï¼ŒThe Cancer Immunome Databaseï¼Œç”±æ¥è‡ªå¥¥åœ°åˆ©çš„å› æ–¯å¸ƒé²å…‹å¤§å­¦åŒ»å­¦é™¢Zlatko Trajanoskiå›¢é˜Ÿå¼€å‘ï¼Œäº2017å¹´å‘è¡¨äºCell Reports https://tcia.at/home 20ç§solid cancerçš„å…ç–«ç›¸å…³åˆ†æï¼Œä¾‹å¦‚ å…ç–«åŸºå› è¡¨è¾¾ã€ç»†èƒæµ¸æ¶¦ã€è‚¿ç˜¤äºšå…‹éš†ç­‰ äº®ç‚¹ä¹‹ä¸€æ˜¯æå‡ºImmunophenoscoreæŒ‡æ ‡ç”¨äºé¢„æµ‹å…ç–«æ²»ç–—å“åº” https://github.com/icbi-lab/Immunophenogram æ ·æœ¬ä¸´åºŠä¿¡æ¯ä¹Ÿæ•´ç†çš„è¾ƒä¸ºå®Œæ•´ ...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-04-02 00:00:00 &#43;0000 UTC&#39;&gt;2023-04-02&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-04-02&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;832&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to ç”Ÿä¿¡ç›¸å…³ç½‘ç«™æ•°æ®åº“é›†é”¦" href="https://lishensuo.github.io/en/posts/bioinfo/075%E7%94%9F%E4%BF%A1%E7%9B%B8%E5%85%B3%E7%BD%91%E7%AB%99%E6%95%B0%E6%8D%AE%E5%BA%93%E9%9B%86%E9%94%A6/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>TCGAçš„SNVæ•°æ®ä¸‹è½½ä¸maftoolså¯è§†åŒ–
			</h2>
		</header>
		<section class="entry-content">
			<p>1ã€TCGAbiolinksä¸‹è½½æ•°æ® ä½¿ç”¨TCGAbiolinksä¸‹è½½ç‰¹å®šè‚¿ç˜¤ç±»å‹çš„SNVæ•°æ® https://bioconductor.org/packages/release/bioc/vignettes/TCGAbiolinks/inst/doc/mutation.html 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 library(TCGAbiolinks) query &lt;- GDCquery( project = &#34;TCGA-CHOL&#34;, data.category = &#34;Simple Nucleotide Variation&#34;, access = &#34;open&#34;, legacy = FALSE, data.type = &#34;Masked Somatic Mutation&#34;) GDCdownload(query) maf &lt;- GDCprepare(query) dim(maf) # [1] 3764 141 ## (1) å› åç»­éœ€è¦ï¼Œä¿®æ”¹Tumor_Sample_Barcodeåˆ— maf$long_Barcode = maf$Tumor_Sample_Barcode maf$Tumor_Sample_Barcode = substr(maf$Tumor_Sample_Barcode,1,12) length(unique(maf$Tumor_Sample_Barcode)) # 51 ## (2) è¯»å–ä¸´åºŠç”Ÿå­˜æ•°æ® clinical = readxl::read_xlsx(&#34;TCGA_Pan_Cancer_Clinical_Data_mmc1.xlsx&#34;) clinical_sle = clinical %&gt;% dplyr::filter(type==&#34;CHOL&#34;) %&gt;% dplyr::select(bcr_patient_barcode, OS, OS.time, clinical_stage) %&gt;% dplyr::rename(Tumor_Sample_Barcode=bcr_patient_barcode) dim(clinical_sle) # 45 2ã€maftoolså¯è§†åŒ– https://bioconductor.org/packages/release/bioc/vignettes/maftools/inst/doc/maftools.html https://www.jieandze1314.com/post/cnposts/237/ 1 2 3 4 5 6 7 8 9 library(maftools) maf_obj = read.maf(maf = maf, clinicalData = clinical_sle) #æ¯ä¸ªæ ·æœ¬çš„çªå˜æƒ…å†µç»Ÿè®¡ getSampleSummary(maf_obj) #æ¯ä¸ªåŸºå› çš„çªå˜ç±»å‹ç»Ÿè®¡ getGeneSummary(maf_obj) 2.1æ¦‚æ‹¬å›¾ 1 2 3 plotmafSummary(maf = maf_obj, rmOutlier = TRUE, addStat = &#39;median&#39;, dashboard = TRUE, titvRaw = FALSE) 2.2 åŸºå› ç€‘å¸ƒå›¾ 1 2 3 4 5 6 7 8 9 10 11 12 13 oncoplot(maf = maf_obj, top = 10) # å¦‚ä¸‹å›¾ ## æ·»åŠ ä¸´åºŠæ³¨é‡Š oncoplot(maf = maf_obj, top = 10, clinicalFeatures = c(&#39;clinical_stage&#39;,&#34;OS&#34;), draw_titv = TRUE) ## é€‰æ‹©ç‰¹å®šåŸºå› é›† set.seed(42) genes = sample(maf_obj@data$Hugo_Symbol,10) oncoplot(maf = maf_obj, genes = genes, clinicalFeatures = c(&#39;clinical_stage&#39;,&#34;OS&#34;), draw_titv = TRUE) 2.3 è½¬æ¢é¢ å€’ç»Ÿè®¡ 1 2 3 4 5 # Transition è½¬æ¢ ï¼š å˜Œå‘¤(AG)æˆ–å˜§å•¶(CT)å†…éƒ¨ä¹‹é—´è½¬æ¢ # Transversions é¢ å€’ï¼šå˜Œå‘¤ä¸å˜§å•¶é—´äº’ç›¸è½¬æ¢ maf_obj.titv = titv(maf = maf_obj, plot = FALSE, useSyn = TRUE) plotTiTv(res = maf_obj.titv) 2.4 åŸºå› å¯¹çªå˜ç»Ÿè®¡ 1 2 3 4 5 6 7 8 # green: co-occuring # yellow: mutually exclusive somaticInteractions(maf = maf_obj, top = 25, pvalue = c(0.05, 0.1)) set.seed(42) genes = sample(maf_obj@data$Hugo_Symbol,25) somaticInteractions(maf = maf_obj, genes = genes , pvalue = c(0.05, 0.1)) 2.5 ç”Ÿå­˜åˆ†æ æ ¹æ®ç‰¹å®šåŸºå› æ˜¯å¦çªå˜å°†ç—…äººåˆ†æˆWTä¸Mutantä¸¤ç»„ 1 2 3 4 5 6 7 8 9 10 11 mafSurvival(maf = maf_obj, genes = &#39;TP53&#39;, time = &#39;OS.time&#39;, Status = &#39;OS&#39;) # Group medianTime N # 1: Mutant 732 4 # 2: WT 650 41 ## æå–ä¿¡æ¯ # maf_obj@clinical.data %&gt;% # dplyr::mutate(Group=ifelse(Tumor_Sample_Barcode %in% # subset(maf_obj@data, Hugo_Symbol==&#34;TP53&#34;)$Barcode, # &#34;Mutant&#34;,&#34;WT&#34;)) 2.6 åŸºå› å¯¹çš„ç”Ÿå­˜ç›¸å…³æ€§ 1 2 3 4 5 6 7 8 9 10 prog_geneset = survGroup(maf = maf_obj, top = 200, geneSetSize = 2, time = &#34;OS.time&#34;, Status = &#34;OS&#34;, verbose = FALSE,minSamples = 3) prog_geneset # Gene_combination P_value hr WT Mutant # 1: PBRM1_PLXNA4 0.243 2.36e&#43;00 42 3 # 2: PBRM1_PCLO 0.294 3.46e-01 42 3 # 3: PBRM1_TP53 0.320 3.64e-01 42 3 mafSurvGroup(maf = maf_obj, geneSet = c(&#34;PBRM1&#34;, &#34;PLXNA4&#34;), time = &#34;OS.time&#34;, Status = &#34;OS&#34;) </p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-04-09 00:00:00 &#43;0000 UTC&#39;&gt;2023-04-09&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-04-09&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;608&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to TCGAçš„SNVæ•°æ®ä¸‹è½½ä¸maftoolså¯è§†åŒ–" href="https://lishensuo.github.io/en/posts/bioinfo/076tcga%E7%9A%84snv%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD%E4%B8%8Emaftools%E5%8F%AF%E8%A7%86%E5%8C%96/"></a>
</article>
<footer class="page-footer">
  <nav class="pagination">
    <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/page/5/">Â« Prev Page</a>
    <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/page/7/">Next Page Â»</a>
  </nav>
</footer>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>æ‚¨æ˜¯æœ¬ç«™ç¬¬ <span id="busuanzi_value_site_uv"></span> ä½è®¿é—®è€…ï¼Œæ€»æµè§ˆé‡ä¸º <span id="busuanzi_value_site_pv"></span> æ¬¡
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
