<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹scGPT | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="æ–‡çŒ®-ç®—æ³•">
<meta name="description" content="
æ–‡çŒ®ï¼š scGPT: toward building a foundation model for single-cell multi-omics using generative AI
æ—¶é—´ï¼š2024 Feb. (Published)
æœŸåˆŠï¼šNature Method
DOIï¼šhttps://doi.org/10.1038/s41592-024-02201-0">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/basic/202%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bscgpt/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/basic/202%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bscgpt/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹scGPT" />
<meta property="og:description" content="
æ–‡çŒ®ï¼š scGPT: toward building a foundation model for single-cell multi-omics using generative AI
æ—¶é—´ï¼š2024 Feb. (Published)
æœŸåˆŠï¼šNature Method
DOIï¼šhttps://doi.org/10.1038/s41592-024-02201-0" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/basic/202%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bscgpt/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2024-09-29T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2024-09-29T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹scGPT"/>
<meta name="twitter:description" content="
æ–‡çŒ®ï¼š scGPT: toward building a foundation model for single-cell multi-omics using generative AI
æ—¶é—´ï¼š2024 Feb. (Published)
æœŸåˆŠï¼šNature Method
DOIï¼šhttps://doi.org/10.1038/s41592-024-02201-0"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "åˆ†ç±»",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "ğŸ“– ç§‘ç ”åŸºç¡€ -- è¯»æ–‡çŒ®ã€çœ‹æ•™æ",
      "item": "https://lishensuo.github.io/en/posts/basic/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹scGPT",
      "item": "https://lishensuo.github.io/en/posts/basic/202%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bscgpt/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹scGPT",
  "name": "æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹scGPT",
  "description": " æ–‡çŒ®ï¼š scGPT: toward building a foundation model for single-cell multi-omics using generative AI\næ—¶é—´ï¼š2024 Feb. (Published)\næœŸåˆŠï¼šNature Method\nDOIï¼šhttps://doi.org/10.1038/s41592-024-02201-0\n",
  "keywords": [
    "æ–‡çŒ®-ç®—æ³•"
  ],
  "articleBody": " æ–‡çŒ®ï¼š scGPT: toward building a foundation model for single-cell multi-omics using generative AI\næ—¶é—´ï¼š2024 Feb. (Published)\næœŸåˆŠï¼šNature Method\nDOIï¼šhttps://doi.org/10.1038/s41592-024-02201-0\nGithubï¼šhttps://github.com/bowang-lab/scGPT\n1. ç®€ä»‹ 1.1 å…³äºä½œè€… é€šè®¯ä½œè€…ï¼šBo Wangï¼Œå¤šä¼¦å¤šå¤§å­¦\nhttps://wanglab.ai/ https://github.com/bowang-lab/ https://scholar.google.com/citations?hl=en\u0026user=37FDILIAAAAJ 1.2 æ–‡çŒ®æ¦‚è¿° å—å¯å‘äºGPTç­‰åŸºäºtransformerçš„å¤§æ¨¡å‹ï¼ŒWangå›¢é˜ŸåŸºäºå¤§é‡å•ç»†èƒè½¬å½•ç»„æ•°æ®ï¼Œæ„å»ºäº†scGPT foundation modelï¼Œå¯ç”¨äºå¤šç§å¸¸è§çš„å•ç»†èƒä¸‹æ¸¸åˆ†æä»»åŠ¡ï¼Œæ€§èƒ½ä¼˜äºå…¶å®ƒå·²æœ‰çš„åˆ†æå·¥å…·ã€‚\næ ¸å¿ƒå‡è®¾ï¼š\nåœ¨NLPé¢†åŸŸä¸­ï¼Œä¸€ä¸ªå¥å­ç”±å¤šä¸ªå•è¯ç»„æˆã€‚ç»è¿‡é¢„è®­ç»ƒçš„è‡ªç›‘ç£å­¦ä¹ ï¼Œfoundation modelå¯ä»¥æå–è¾“å…¥å¥å­ä¸­æ¯ä¸ªå•è¯(æˆ–è€…å¥å­æ•´ä½“)çš„é«˜çº§è¡¨ç¤º(Embedding)ï¼› è€Œåœ¨scRNA-seqä¸­ï¼Œä¸€ä¸ªç»†èƒå¯ä»¥è®¤ä¸ºç”±æ‰€æœ‰åŸºå› çš„ä¸åŒç¨‹åº¦è¡¨è¾¾å®šä¹‰ã€‚ç»è¿‡é¢„è®­ç»ƒçš„è‡ªç›‘ç£å­¦ä¹ ï¼Œfoundation modelå¯ä»¥æå–è¾“å…¥å•ç»†èƒè¡¨è¾¾æ•°æ®çš„æ¯ä¸ªåŸºå› (æˆ–è€…ç»†èƒæ•´ä½“)çš„é«˜çº§è¡¨ç¤º(Embedding)ã€‚ åœ¨Pre-trained foundation modelsåŸºç¡€ä¸Šï¼Œå°†æå–çš„åŸºå› /ç»†èƒEmbeddingåœ¨å…·ä½“çš„å•ç»†èƒä¸‹æ¸¸åˆ†æä»»åŠ¡ä¸­è¿›è¡ŒäºŒæ¬¡å¾®è°ƒ(fine-tune)ï¼Œå‘æŒ¥å¤§æ•°æ®ç”Ÿæˆçš„å¼ºå¤§ä¼˜åŠ¿ã€‚\n2. Pretrainé¢„è®­ç»ƒ Pre-train foundation modelçš„æ ¸å¿ƒåŠŸèƒ½å°±æ˜¯ä»¥ä¸€ä¸ªå•ç»†èƒçš„æ•°æ®ä¸ºè¾“å…¥ï¼Œå°†æå–çš„Gene/Cell Embeddingä¸ºè¾“å‡ºã€‚\n2.0 æ•°æ®è§„æ¨¡ æ–‡ç« ä»CELLxGENEç½‘ç«™æ”¶é›†äº†33Mä¸ªæ­£å¸¸ç»†èƒçš„scRNA-seqæ•°æ®ç”¨äºè®­ç»ƒFoundation model å…¶ä¸­æ¥è‡ªBrainã€Bloodç­‰å™¨å®˜çš„ç»†èƒæ•°æœ€å¤š 2.1 åˆå§‹åŒ–è¾“å…¥ scGPTçš„åŸå§‹è¾“å…¥æ•°æ®é€šå¸¸ä¸ºNä¸ªç»†èƒï¼ŒGä¸ªåŸºå› çš„Countè¡¨è¾¾çŸ©é˜µã€‚æ­¤æ—¶ï¼Œéœ€è¦å¯¹æ¯ä¸ªç»†èƒçš„åŸºå› æ•°æ®è¿›è¡Œåˆå§‹åŒ–å¤„ç†ï¼Œä½œä¸ºscGPTçš„æ ‡å‡†è¾“å…¥ï¼› æ¦‚æ‹¬æ¥è¯´ï¼Œåˆ†åˆ«ä»æ¯ä¸ªåŸºå› çš„ä¸‰ä¸ªè§’åº¦è¿›è¡ŒDç»´åµŒå…¥ç¼–ç  (1Ã—D)ï¼Œç„¶åå†è¿›è¡ŒçŸ©é˜µåŠ æ³•ï¼ˆä»ä¸º1Ã—Dï¼‰ï¼Œæœ€ç»ˆå°†æ‰€æœ‰åŸºå› åˆå¹¶å¾—åˆ°ï¼ˆMÃ—Dï¼‰çš„ç»†èƒç‰¹å¾çŸ©é˜µã€‚ å› ä¸ºç»†èƒçš„åŸºå› ä¼šæ ¹æ®æƒ…å†µè¿›è¡Œé€‰æ‹©ï¼Œå¹¶ä¸”è€ƒè™‘åˆ°ç­‰ç‰¹æ®Šè¯å…ƒï¼Œæ‰€ä»¥è¿™é‡Œå¹¶ä¸æ˜¯GÃ—D\nï¼ˆ1ï¼‰Gene token\næ‰€æœ‰åŸºå› (è¯å…ƒ)æ„æˆçš„è¯è¡¨ä¸­ï¼Œæ¯ä¸ªåŸºå› éƒ½æœ‰ä¸€ä¸ªå¯¹åº”çš„æ•´æ•°æ ‡è¯†ç¬¦ï¼ˆInteger Identifierï¼‰ æ­¤å¤–è¿˜æœ‰å…¶å®ƒç‰¹æ®Šçš„è¯å…ƒï¼Œä¾‹å¦‚, ç­‰ è¯å…ƒé€šå¸¸æ”¾åœ¨ç»†èƒMä¸ªè¯å…ƒä¸­çš„ç¬¬ä¸€ä¸ªï¼Œä¸è¡¨ç¤ºç‰¹å®šåŸºå› ï¼Œè€Œç”¨äºcell representation è¯å…ƒç”¨äºè¡¥é•¿è‡³å›ºå®šé•¿åº¦M(e.g. æœ‰äº›æƒ…å†µä¸è€ƒè™‘è¡¨è¾¾å€¼ä¸º0çš„åŸºå› æ—¶) é‡‡ç”¨PyTorch embedding layerå°†æŠ•å°„æˆDç»´çš„Embedding ï¼ˆ2ï¼‰Gene expression\nå•ç»†èƒçš„countè¡¨è¾¾å€¼åœ¨ä¸åŒæµ‹åºèƒŒæ™¯çš„æ¡ä»¶ä¸‹ï¼Œä¸å…·æœ‰å¯æ¯”æ€§ã€‚ scGPTé‡‡ç”¨binåˆ†æ®µå¤„ç†ï¼Œå°†å•ä¸ªç»†èƒiçš„éé›¶åŸºå› è¡¨è¾¾å€¼ï¼Œåˆ†ä¸ºkä¸ªbinsï¼Œåˆ†æ•°åˆ†åˆ«æ˜¯1..kã€‚ ä¾‹å¦‚ï¼Œè‹¥åŸºå› binæ‰“åˆ†ä¸ºkï¼Œè¡¨ç¤ºå…¶è¡¨è¾¾å€¼å¤„äºæœ€é«˜çš„binèŒƒå›´å†…ï¼› æ¯ä¸ªç»†èƒåˆ†binçš„é˜ˆå€¼æ ‡å‡†éƒ½ä¼šä¸å°½ç›¸åŒã€‚ é‡‡ç”¨MLPï¼Œè½¬æ¢ä¸ºDç»´çš„Embedding ï¼ˆ3ï¼‰Condition token\nè®°å½•åŸºå› é¢å¤–çš„Condition tokenï¼Œä¾‹å¦‚æ˜¯å¦ä¸ºperturbated genes é‡‡ç”¨PyTorch embedding layerå°†æŠ•å°„æˆDç»´çš„Embedding ç»¼ä¸Šï¼Œä¸€ä¸ªç»†èƒ(i)çš„åˆå§‹åŒ–è¾“å…¥h (M Ã— D)çš„è®¡ç®—æ–¹å¼å¦‚ä¸‹å›¾æ‰€ç¤ºã€‚å…¶ä¸­ç¬¬ä¸€åˆ—é€šå¸¸è¡¨ç¤ºä¸ºä½œä¸ºCell representationï¼›å…¶å®ƒåˆ—ï¼ˆé™¤ç­‰ç‰¹æ®Šå­—ç¬¦å¤–ï¼Œå‡è¡¨ç¤ºGene Embeddingã€‚\n2.2 è‡ªç›‘ç£è®­ç»ƒ ï¼ˆ1ï¼‰Transformerå—\nFoundation modelä¸»è¦ç”± 12 (l)ä¸ªTransformerå—ç»„æˆï¼Œæ¯ä¸ªå—åˆ™å‡é‡‡ç”¨äº†å¤šå¤´ï¼ˆ8ï¼‰è‡ªæ³¨æ„åŠ›æœºåˆ¶ï¼› ç”±äºè¿™é‡Œçš„Mé€šå¸¸è¾ƒå¤§ï¼Œå³ä¸€ä¸ªç»†èƒè€ƒè™‘æ•°åƒä¸Šä¸‡ä¸ªåŸºå› ï¼ˆå¯ä»¥ç†è§£ä¸ºç‰¹åˆ«é•¿çš„å¥å­ï¼‰ï¼ŒscGPTé‡‡ç”¨äº†FlashAttentionç®—æ³•ç”¨ä»¥åŠ é€Ÿè‡ªæ³¨æ„åŠ›è®¡ç®—ã€‚\næ­¤å¤–ï¼Œå¯¹äºæ¯ä¸ªç»†èƒï¼Œåªæœ‰è¡¨è¾¾å€¼é0çš„åŸºå› å‚ä¸é¢„è®­ç»ƒè¿‡ç¨‹ï¼Œä»¥æé«˜é€Ÿåº¦ã€‚\nä½œè€…ä¹Ÿæ¨èäº†å…¶å®ƒé«˜æ•ˆè®¡ç®—çš„Transformerå˜ä½“ï¼ŒåŒ…æ‹¬linear complexity (Linformer)ï¼ŒKernelized Self-Attention (KSA)ã€‚\nï¼ˆ2ï¼‰è‡ªç›‘ç£ä»»åŠ¡\nscGPTè‡ªç›‘ç£ä»»åŠ¡çš„æ ¸å¿ƒæ˜¯é¢„æµ‹æ©ç (masked)åŸºå› çš„è¡¨è¾¾å€¼æ°´å¹³ï¼Œé‡‡ç”¨MSEæŸå¤±å‡½æ•°ã€‚ å…·ä½“åˆ†ä¸ºå¦‚ä¸‹ä¸¤ä¸ªå­ä»»åŠ¡ï¼š Gene-promptï¼šåŸºäºå·²çŸ¥è¡¨è¾¾å€¼çš„Gene Embeddingï¼Œé¢„æµ‹æœªçŸ¥è¡¨è¾¾å€¼Geneçš„expression valueï¼› Cell-promptï¼šåŸºäº cell representationé¢„æµ‹å…¨åŸºå› çš„expression valueã€‚ ç„¶åï¼Œå°†ä¸Šè¿°ä¸¤ç§æ¨¡å¼çš„æŸå¤±lossç›¸åŠ åï¼Œå†è®¡ç®—æ¢¯åº¦å¹¶æ›´æ–°æ¨¡å‹å‚æ•°ã€‚ ï¼ˆ3ï¼‰Masked Attention\nè‡ªæ³¨æ„è®¡ç®—è¿‡ç¨‹ä¸ºï¼šå°†masked gene Embedding( without expression) ä½œä¸ºqueryï¼Œè®¡ç®—ä¸å…¶å®ƒknown gene Embeddingï¼ˆä¹ŸåŒ…æ‹¬è‡ªå·±ï¼‰çš„æ³¨æ„åŠ›æƒé‡åï¼Œå†è¿›è¡ŒåŠ æƒè¿ç®—å¾—åˆ°è¾“å‡ºï¼› ä¸GPTæ¨¡å‹æ‰€å¤„ç†çš„æ–‡æœ¬å¥ä¸åŒä¹‹å¤„åœ¨äºï¼šthe non-sequential nature of the genes in one cell. ä¸ºæ­¤ä½œè€…è®¾è®¡äº†masked attentionæ³¨æ„åŠ›è®¡ç®—æ–¹å¼ï¼› å¦‚ä¸Šå…¬å¼ï¼Œåœ¨ä¸è€ƒè™‘Amaskçš„æƒ…å†µä¸‹ï¼Œä¸ºæ ‡å‡†çš„è‡ªæ³¨æ„åŠ›è®¡ç®—æ–¹å¼ã€‚è€ŒAmaskå¯å‚è€ƒå¦‚ä¸‹å…¬å¼ï¼Œä»¥åŠä¸‹å›¾å·¦Aï¼ˆè¡Œè¡¨ç¤ºqueryï¼Œåˆ—è¡¨ç¤ºkeyï¼‰ç†è§£ã€‚æ¯ä¸€ä¸ªqueryï¼ˆiï¼‰è®¡ç®—ä¸å…¶ä»–è¯å…ƒæ³¨æ„åŠ›ï¼ˆåŒ…æ‹¬ä¸å®ƒè‡ªå·±ï¼‰æ—¶ï¼š è‹¥key ï¼ˆjï¼‰ä¸æ˜¯unknown expression geneï¼Œåˆ™ä¸º0ï¼› è‹¥i = jæ—¶ï¼Œä¸”jæ˜¯unknown expression geneï¼Œåˆ™ä¸º0ï¼ˆè‡ªå·±ä¸è‡ªå·±çš„æ³¨æ„åŠ›è®¡ç®—ï¼‰ å…¶å®ƒæƒ…å†µä¸‹ï¼Œåˆ™ä¸ºè´Ÿæ— ç©·ï¼ˆå¯¹åº”å›¾ä¸­çš„æ·±è“è‰²å•å…ƒæ ¼ï¼‰ æ¯ä¸€è¡Œ(i)è¡¨ç¤ºä¸€ä¸ªqueryè¯å…ƒä¸è¯¥ç»†èƒæ‰€æœ‰è¯å…ƒçš„æ³¨æ„åŠ›è®¡ç®—ã€‚ a = 0è¡¨ç¤ºä¸äº§ç”Ÿä»»ä½•å½±å“ï¼Œa = -infè¡¨ç¤ºå°†query(i)å¯¹äºkey(j)çš„æ³¨æ„åŠ›ç½®æ¢ä¸º-infï¼Œç»softmaxè½¬æ¢ååˆ™å˜ä¸º0ã€‚ The rule of thumb for scGPT attention masking is to only allow attention computation between embeddings of the â€˜known genesâ€™ and the query gene itself.\nåœ¨ä¸€ä¸ªcellçš„å…¨éƒ¨masked geneè¡¨è¾¾å€¼é¢„æµ‹çš„è¿‡ç¨‹ä¸­ï¼Œé‡‡ç”¨å¤šè½®ï¼ˆkï¼‰è¿­ä»£é¢„æµ‹çš„æ€è·¯ï¼ˆå¦‚ä¸Šå›¾å³ï¼Œç±»æ¯”GPTçš„è‡ªå›å½’è®­ç»ƒï¼‰ åœ¨ç¬¬ä¸€è½®ä¸­ï¼Œå¯¹äºæ‰€æœ‰masked tokençš„é¢„æµ‹ï¼Œå°†å…¶ä¸­1/kä¸ªhigh prediction confidenceçš„tokenæ ‡æ³¨ä¸ºknown genesã€‚ ç„¶ååœ¨æ–°ä¸€è½®è¿­ä»£ä¸­ï¼Œé‡å¤ä¸Šä¸€æ­¥éª¤ï¼Œç›´è‡³é¢„æµ‹å‡ºæ‰€æœ‰çš„masked tokenã€‚ å€¼å¾—æ³¨æ„çš„æ˜¯ï¼Œæ–‡ç« å¹¶æ²¡æœ‰è¯´æ˜prediction confidenceæ˜¯å¦‚ä½•è®¡ç®—çš„ã€‚ä¸ªäººç†è§£æœ¬è´¨è¿˜æ˜¯åŸºäºä¸çœŸå®è¡¨è¾¾å€¼ä¹‹é—´çš„è¯¯å·®ã€‚ In each generation iteration, scGPT predicts the gene expression values of a new set of genes, and these genes in turn become the â€˜known genesâ€™ in the next iteration for attention computation.\n2.3 å¤šæ‰¹æ¬¡ä¸å¤šæ¨¡æ€è¡¨ç¤º åœ¨å¤šæ‰¹æ¬¡ã€å¤šç»„å­¦åˆå¹¶ä¸‹æ¸¸ä»»åŠ¡ä¸­ï¼Œéœ€è¦é¢å¤–çš„token Embeddingä»¥è¡¨ç¤ºå¿…è¦çš„æ‰¹æ¬¡å’Œç»„å­¦ä¿¡æ¯ï¼Œä¾›æ¨¡å‹å­¦ä¹ ï¼› ä½†æ˜¯scGPTåœ¨Transformerçš„pre-trainè¿‡ç¨‹ä¸­ï¼Œå¹¶æœªåŠ å…¥ç›¸å…³ä¿¡æ¯ã€‚è€Œæ˜¯åœ¨foundation modelçš„è¾“å‡ºç»“æœä¸­å†è¿›è¡Œæ‹¼æ¥æ“ä½œï¼Œå³æ¨¡å‹åœ¨Pre-trainåï¼ŒFine-tuneå‰å¼•å…¥æ‰¹æ¬¡å’Œæ¨¡æ€ä¿¡æ¯ï¼Œä»¥å¸Œæœ›åœ¨å¾®è°ƒè¿‡ç¨‹ä¸­å­¦ä¹ åˆ°ç›¸å…³ä¿¡æ¯ã€‚ è¿™æ ·åšçš„ä¸»è¦åŸå› æ˜¯ï¼šå¦‚æœæ‰¹æ¬¡å’Œæ¨¡æ€ä¿¡æ¯åœ¨è¾“å…¥é˜¶æ®µè¢«å¼•å…¥ï¼ŒTransformerçš„è‡ªæ³¨æ„åŠ›æœºåˆ¶å¯èƒ½ä¼šè¿‡åº¦å…³æ³¨åŒä¸€æ¨¡æ€æˆ–æ‰¹æ¬¡å†…çš„ç‰¹å¾ï¼Œå¯¼è‡´æ¨¡å‹å¿½ç•¥è·¨æ¨¡æ€æˆ–æ‰¹æ¬¡çš„é‡è¦å…³è”ã€‚ This is to prevent the transformer from amplifying the attention within features of the same modalities while underestimating those of different modalities.\nmodality tokens (tm)ï¼šè¡¨ç¤ºè¯å…ƒå¯¹åº”çš„featureæ˜¯Gene/Protein/Peakä¸­çš„å“ªä¸€ç§ï¼› batch tokens (tb)ï¼šè¡¨ç¤ºtokenæ˜¯å¦æ¥è‡ªä¸€ä¸ªæ‰¹æ¬¡ï¼Œé€šå¸¸å¯¹äºç»†èƒæ¥è¯´çš„ã€‚å› æ­¤ä¸€ä¸ªç»†èƒçš„batch tokenéƒ½æ˜¯ä¸€æ ·çš„ã€‚ å¦‚ä¸‹å…¬å¼ï¼Œè¡¨ç¤ºåŒæ—¶å­˜åœ¨å¤šæ‰¹æ¬¡ä»¥åŠå¤šç»„å­¦çš„æƒ…å†µ å¦‚ä¸‹å…¬å¼ï¼Œè¡¨ç¤ºåŒç§ç»„å­¦ï¼Œå­˜åœ¨å¤šæ‰¹æ¬¡çš„æƒ…å†µ 3. Fine-tuneå¾®è°ƒ 3.1 ç»†èƒç±»å‹æ³¨é‡Š ï¼ˆ1ï¼‰Method\nå¯¹äºFoundation modelæå–çš„æ¯ä¸ªç»†èƒCell Embedding () æ„å»ºä¸€ä¸ªMLPåˆ†ç±»å™¨ï¼Œç”¨ä»¥é¢„æµ‹ç»†èƒçš„ç±»åˆ«ï¼Œå¹¶ä½¿ç”¨äº¤å‰ç†µä½œä¸ºæŸå¤±å‡½æ•°ã€‚\né¦–å…ˆä½¿ç”¨ä¸€ä¸ªæ ‡æ³¨ç»†èƒç±»å‹çš„æ•°æ®é›†ï¼ˆreference setï¼‰è¿›è¡ŒFine-tuneï¼Œç„¶åå†ä½¿ç”¨ä¸€ä¸ªHeld-outæ•°æ®é›†éªŒè¯\nè¾“å…¥æ•°æ®å‰çš„é¢„å¤„ç†ï¼š\ncommon set between foundation model and the input data Gene expression: normalization â†’ log1 â†’ bin All gene as input (include zero expression) ï¼ˆ2ï¼‰Result\nåœ¨Human pancreas (èƒ°è…º) dataset\nå¯¹äºæ¯ç§ç»†èƒçš„é¢„æµ‹Precisionéƒ½è¾¾åˆ°0.8ä»¥ä¸Š (é¢„æµ‹ä¸ºè¯¥ç»†èƒç±»å‹ä¸­ï¼Œå®é™…ä¸ºè¯¥ç»†èƒç±»å‹çš„æ¯”ä¾‹) ä»¥Human immune cellsè¿›è¡Œfine-tuneï¼Œé¢„æµ‹å¤šå‘æ€§ç¡¬åŒ–ç—‡(MS)çš„ç»†èƒç±»å‹\nå¹³å‡å‡†ç¡®ç‡Accuracyå¯ä»¥è¾¾åˆ°0.85å·¦å³ ä½¿ç”¨6ç§è‚¿ç˜¤ç»†èƒç±»å‹ä½œä¸ºfine-tuneï¼Œé¢„æµ‹3ç§å…¶å®ƒè‚¿ç˜¤çš„ç»†èƒç±»å‹\nåœ¨è‚¿ç˜¤å¾®ç¯å¢ƒç»†èƒç±»å‹é¢„æµ‹åŒæ ·è¡¨ç°è‰¯å¥½ ä¸TOSICAã€scBERTæ¨¡å‹è¿›è¡Œäº†æ¯”è¾ƒï¼Œå‡è¡¨ç°å‡ºä¸€å®šçš„ä¼˜åŠ¿\næ­¤å¤–ï¼Œä½œè€…è¿˜æå‡ºäº†Reference mappingæ³¨é‡Šæ–¹æ³•ï¼šåœ¨Zero/Fine-tuneæ¨¡å‹æ—¶ï¼Œè·å¾—æ ‡ç­¾ç»†èƒç±»å‹çš„Cell Embeddingã€‚ç„¶åå†è®¡ç®—å‡ºQuery cellsçš„Embeddingã€‚æœ€åæ®æ­¤ï¼Œè®¡ç®—å‡ºæ¯ä¸ªQuery cellæœ€ç›¸è¿‘çš„Reference cellsï¼Œä»è€Œæ³¨é‡Šå…¶ç»†èƒç±»å‹ï¼ˆKNNï¼‰ã€‚\n3.2 åŸºå› æ‰°åŠ¨é¢„æµ‹ ï¼ˆ1ï¼‰Method\nå¯¹äºFoundation modelæå–çš„æ¯ä¸ªç»†èƒçš„Gene Embeddingï¼Œå°†ä¸€éƒ¨åˆ†åŸºå› çš„è¡¨è¾¾å€¼æ©ç ï¼Œä»è€Œè¿›è¡ŒFine-tuneå¾®è°ƒè®­ç»ƒï¼Œå¹¶ä»¥MSEä½œä¸ºæŸå¤±å‡½æ•°ï¼› åœ¨Perturbation prediction taskï¼š Input: Gene expressionä¸ºControl cell expression, Condition tokenæ ‡æ³¨ç›¸åº”ä½ç½®Geneæ˜¯å¦è¢«Perturbate Output: Post-perturbation expression è¾“å…¥æ•°æ®å‰çš„é¢„å¤„ç†ï¼š only select HVGs for training Log1p expression instead of binned values ï¼ˆoutputç›¸åŒï¼‰ ï¼ˆ2ï¼‰Result\nåˆ†åˆ«åŸºäºä¸‰ä¸ªå•ç»†èƒCRISPRæ•°æ®é›†ï¼Œæ¯”è¾ƒäº†scGPTä¸Linearã€GEARSæ¨¡å‹çš„æ€§èƒ½è¡¨ç° Adamson: 87 one-gene perturbationsï¼› Replogle: 2823 one-gene perturbationsï¼› Norman: 131 two-gene and 105 one-gene perturbations å¯¹äºæ¯ä¸ªæ•°æ®é›†ï¼Œä½¿ç”¨ä¸€éƒ¨åˆ†çš„æ‰°åŠ¨æ•°æ®è¿›è¡Œfine-tuneï¼Œå†å¯¹å…¶ä½™unseen geneçš„æ‰°åŠ¨æ•°æ®ä½œä¸ºtest è¯„ä»·æŒ‡æ ‡ä¸ºè®¡ç®—é¢„æµ‹ä¸çœŸå®çš„post-perturbation expressionçš„ç›¸å…³æ€§ (åŸºäºå…¨éƒ¨åŸºå› ï¼Œæˆ–è€…æ˜¯å‰20ä¸ªå½±å“æœ€æ˜¾è‘—çš„åŸºå› ) ç»“æœå‘ç°scGPTæ¨¡å‹ç›¸æ¯”äºå…¶å®ƒä¸¤ç§æ¨¡å‹ï¼Œæé«˜äº†5â€“20% ï¼ˆ3ï¼‰In silicon reverse perturbation prediction\nç®€å•ç†è§£ï¼Œæ ¹æ®perturbation expressionçš„ç»“æœåå‘é¢„æµ‹æ˜¯æœ€æœ‰å¯èƒ½å“ªä¸ªåŸºå› è¢«æ‰°åŠ¨ï¼› å‚è€ƒä½œè€…åœ¨Github issue ï¼ˆhttps://github.com/bowang-lab/scGPT/issues/87ï¼‰çš„è§£ç­”ï¼š å…¶Fine-tuneæ­¥éª¤å…¶å®ä¸ä¸Šé¢Perturbation prediction taskåŸºæœ¬ä¸€è‡´ï¼› ä½¿ç”¨æ•°æ®é›†çš„ä¸€éƒ¨åˆ†è¿›è¡Œå¾®è°ƒåï¼Œå†é¢„æµ‹æ‰€æœ‰ç›¸å…³åŸºå› æ‰°åŠ¨çš„è¡¨è¾¾ç»“æœï¼› æœ€åä½¿ç”¨çœŸå®çš„Query perturbationä¸Šè¿°çš„é¢„æµ‹ç»“æœè¿›è¡ŒKNNå…³è”åˆ†æã€‚ The model was fine-tuned in the same way as in the â€œforwardâ€ perturbation prediction. It used a subset of the dataset as we illustrated in Figure 3F. The reverse perturbation task utilized the model in a different way. To summarize, the result cell states of all possible perturbations were predicted by the fine-tuned model, and then an actual sequenced cell state can query all the predicted cell states in a nearest neighbor search manner, so that the retrieved neighbors indicate the possible origin perturbations.\n3.3 å¤šæ‰¹æ¬¡/ç»„å­¦æ•´åˆ ï¼ˆ1ï¼‰Methodï¼ˆMulti-batchï¼‰\næ ¸å¿ƒç›®æ ‡ï¼šå¯¹äºå¤šæ‰¹æ¬¡çš„scRNA-seqæ•°æ®ï¼Œä¼˜åŒ–ä¸åŒæ‰¹æ¬¡ä¸­æ¯ä¸ªç»†èƒçš„cell representation\ncorrect batch effects while preserving biological variance è¾“å…¥æ•°æ®çš„é¢„å¤„ç†\nCommon set between foundation model and the input data\nGene expression: normalizationâ€“log1â€“bin\nAll genes as input (include zero expression)\nå¦‚2.3ä¸­æ‰€è¿°ï¼Œåœ¨å¾®è°ƒå‰ï¼Œè¿˜éœ€è¦åœ¨Foundation modelçš„è¾“å‡ºç»“æœä¸­ï¼Œè¡¥å……ç»†èƒçš„æ‰¹æ¬¡ä¿¡æ¯ã€‚\nå…·ä½“åœ¨Fine-tuneè®­ç»ƒä¸­ï¼Œè®¾è®¡äº†å¯¹å¤šä¸ªç›®æ ‡å‡½æ•°è¿›è¡ŒæŸå¤±è®¡ç®—ï¼Œä»¥å…±åŒç”¨äºæ¨¡å‹ä¼˜åŒ–\n1ï¼‰GEPï¼šåŸºå› è¡¨è¾¾é¢„æµ‹ï¼Œå‚è€ƒ3.2ï¼›\n2ï¼‰GEPC, Gene expression prediction for cell modelingï¼Œå³ä¸ºäº†ä¼˜åŒ–Cell Embedding(hc)ï¼Œè€Œé¢„æµ‹åŸºå› è¡¨è¾¾ã€‚\n3ï¼‰ECS, Elastic cell similarityï¼šåŸºäºä¸€ä¸ªmini-batchçš„ä¸¤ä¸ªç»†èƒçš„ç›¸ä¼¼åº¦è®¡ç®—æŸå¤±å‡½æ•°ã€‚ä½¿å¾—é«˜äºæŸä¸ªé˜ˆå€¼çš„ä¸¤ä¸ªç»†èƒæ›´ç›¸ä¼¼ï¼Œä½äºçš„åˆ™æ›´è¿œç¦»ã€‚\n4ï¼‰Domain adaptation via reverse back propagationï¼šä¸»è¦ä½œç”¨æ˜¯ä½¿å¾—ç”Ÿæˆçš„cell representationè¡¨ç¤ºç»è¿‡MLPåˆ†ç±»å™¨æ— æ³•é¢„æµ‹å…¶æ­£ç¡®çš„batchä¿¡æ¯ï¼ˆç±»ä¼¼äºGAN å¯¹æŠ—ç¥ç»ç½‘ç»œï¼‰ã€‚æœ¬è´¨ä¸Šè¿˜æ˜¯å»ºç«‹MLPé¢„æµ‹Batchçš„åˆ†ç±»å™¨ï¼Œä½†æ˜¯è®¡ç®—æ¢¯åº¦åï¼Œè¿›è¡Œåå‘æ›´æ–°å‚æ•°ã€‚\nï¼ˆ2ï¼‰Result\nåŸºäºä¸‰ä¸ªæ•°æ®é›†ï¼Œæ¯”è¾ƒäº†scGPTä¸scVIï¼ŒSeuratï¼ŒHarmonyä¸‰ä¸ªå·¥å…·å»æ‰¹æ¬¡çš„æ•ˆæœ\nCOVID-19: 18 batches\nPBMC-10K: 2 batches\nPerirhinal cortex: 2 batches\nè¯„ä»·æŒ‡æ ‡ï¼šåŸºäºä¸‰ä¸ªæŒ‡æ ‡ï¼ˆNMI/ARI/ASWï¼‰çš„AvgBIOï¼Œå€¼è¶Šé«˜è¡¨æ˜å»æ‰¹æ¬¡æ•ˆæœè¶Šå¥½ã€‚\nscGPTå‡ä¼˜äºå…¶å®ƒä¸‰ç§å·¥å…·çš„æ•ˆæœã€‚ ï¼ˆ3ï¼‰Multi-omics integration\nMethod\nç›®æ ‡æ˜¯ä½¿å¾—ç›¸åŒç»†èƒç±»å‹çš„ä¸åŒç»„å­¦æ•°æ®èƒ½å¤Ÿæœ‰ç›¸ä¼¼çš„Cell Embeddingï¼Œä»è€Œåœ¨èšç±»æ—¶æ¯”è¾ƒæ¥è¿‘ã€‚ scGPTä¸»è¦è€ƒè™‘äº†3ç§ç»„å­¦ï¼Œåˆ†åˆ«æ˜¯scRNA-seqã€scATAC-seqï¼ŒSingle-cell proteomicsã€‚å¯¹æ­¤ï¼Œæœ‰ä¸¤ç§æ•°æ®å½¢å¼ï¼š paired settingï¼šä¸€ç¾¤ç»†èƒåŒæ—¶æµ‹å¤šç§ç»„å­¦ mosaic settingï¼šä¸€ç¾¤ç»†èƒæµ‹ä¸€ç§ç»„å­¦ï¼Œå¦ä¸€ç¾¤æµ‹å¦ä¸€ç»„ã€‚ å¯¹äºscRNA-seqå¯ä»¥ç›´æ¥ç»§æ‰¿ç¬¬2èŠ‚è®­ç»ƒçš„Foundation modelï¼Œå¯¹äºå…¶å®ƒä¸¤ç»„éœ€è¦é‡å¤´è®­ç»ƒã€‚ åœ¨Pre-trainä¹‹åï¼Œå¦‚2.3å°èŠ‚éœ€è¦æ‹¼æ¥è¡¨ç¤ºå¤šæ¨¡æ€çš„tokenï¼ˆå¦‚æœæ¶‰åŠå¤šæ‰¹æ¬¡ï¼Œä¹Ÿè¦æ·»åŠ batch tokenï¼‰ Fine-tune ç›®æ ‡å‡½æ•°åŒ…æ‹¬GEPï¼ŒGEPCï¼ˆå¦‚æœæ¶‰åŠå¤šæ‰¹æ¬¡ï¼Œè¿˜éœ€è¦æ·»åŠ DARï¼‰ Result\nå¯¹äºMultiome PBMC (paired)æ•°æ®é›†ï¼Œä¸scGLUEï¼ŒSeuratè¿›è¡Œäº†æ¯”è¾ƒï¼Œå¦‚ä¸‹å›¾æ‰€ç¤º\nRNA-seqã€ATAC å¯¹äºBone marrow mononuclear cells (paired)æ•°æ®é›†ï¼Œä¸Seuratè¿›è¡Œäº†æ¯”è¾ƒ\nRNA-seqã€Protein 9wä¸ªç»†èƒï¼Œ12ä¸ªdonor(multiple batches)ï¼Œ48ç§ç»†èƒç±»å‹ å¯¹äºASAP human PBMC (mosaic)æ•°æ®é›†ï¼Œä¸scMoMatè¿›è¡Œäº†æ¯”è¾ƒ\n4ä¸ªæ‰¹æ¬¡ï¼Œ3ç§ç»„å­¦ 3.4 åŸºå› è°ƒæ§ç½‘ç»œæ„å»º ï¼ˆ1ï¼‰Method\næœ¬è´¨ä¸Šæ˜¯åŸºäºFoundation modelæˆ–è€…Fine-tune modelæå–çš„Gene Embeddingè®¡ç®—ä¸¤ä¸¤åŸºå› é—´çš„ç›¸å…³æ€§ï¼Œç”¨æ¥æ„å»ºgene similarity networkã€‚\nZero-shot settingæ¨¡å¼ä¸‹ï¼Œç›´æ¥ä½¿ç”¨Foundation modelè¾“å‡ºçš„Gene Embeddingè®¡ç®—ï¼›\nFine-tuned settingæ¨¡å¼ä¸‹ï¼Œä½¿ç”¨ç‰¹å®šæ•°æ®é›†å¾®è°ƒåè¾“å‡ºçš„Gene Embeddingè®¡ç®—\nä½¿ç”¨integration taskè¿›è¡Œå¾®è°ƒï¼Œä»¥å­¦ä¹ ç‰¹å®šæ•°æ®é›†ç›¸å…³çš„Gene Embedding ï¼ˆä¸‹åŒï¼‰\nåŸºäºæ³¨æ„åŠ›æœºåˆ¶çš„target geneé‰´å®šï¼šAttention mapå¯ä»¥åæ˜ å‡ºåŸºå› ä¹‹é—´çš„ç›¸äº’å½±å“ã€‚å…¶ä¸­ï¼Œæ¯ä¸€åˆ—(column)ï¼Œè¡¨ç¤ºè¿™ä¸ªgeneï¼ˆåˆ—åï¼‰å¯¹æ‰€æœ‰query geneçš„influenceã€‚\nä½¿ç”¨perturbation datasetså¯ä»¥å¸®åŠ©æ¨æµ‹perturbating geneçš„targetï¼Œå¦‚ä¸‹å›¾æ‰€ç¤º\né¦–å…ˆï¼ŒåŸºäºcontrol cellå¾—åˆ°çš„attention mapï¼› ç„¶åï¼ŒåŸºäºperturbation expressionå¾—åˆ°çš„attention mapï¼› è®¡ç®—ä¸Šè¿°äºŒè€…å¯¹çš„å·®å€¼ï¼Œå¯ä»¥æ¨æµ‹ç‰¹å®šåŸºå› å¹²æ‰°å‰åï¼Œå½±å“æœ€å¤§çš„gene ï¼ˆ2ï¼‰Result\næ–‡ç« é¦–å…ˆå¯¹zero-shotæ¨¡å¼è¿›è¡Œäº†æ¢ç´¢ï¼Œç„¶ååœ¨å¯¹ä¸€ä¸ªhuman immune datasetè¿›è¡Œäº†Fine-tuneåæ„å»ºGRNï¼Œå‡å‘ç°äº†å…·æœ‰ç”Ÿç‰©å­¦æ„ä¹‰çš„è°ƒæ§ç½‘ç»œä»¥åŠGene cluster (Leiden)ï¼Œå¹¶è¿›è¡Œäº†é€šè·¯å¯Œé›†åˆ†æç­‰ã€‚ æ¥ä¸‹æ¥ï¼Œä½œè€…åŸºäºPre-train scGPT blood modelï¼Œä½¿ç”¨äº†Adamson CRISPRæ•°æ®é›†(87 CRISPR inference on leukemia cells)è¿›è¡Œäº†å¾®è°ƒã€‚ ä¾‹å¦‚ä¸‹å›¾ï¼Œé€šè¿‡æ¯”è¾ƒDDIT3æ‰°åŠ¨å‰åçš„difference attention score, é‰´å®šäº†å…¶å½±å“æœ€å¤§çš„çš„Top20/100ä¸ªåŸºå› ã€‚ 4. æ¨¡å‹å½±å“å› ç´  ç»¼ä¸Šï¼ŒscGPT Foundation modelåœ¨å¤„ç†ä¸‹æ¸¸åˆ†æä»»åŠ¡æ—¶ï¼Œç›¸æ¯”äºå…¶å®ƒå·²æœ‰çš„å•ç»†èƒå·¥å…·å‡è¡¨ç°å‡ºæ˜æ˜¾çš„ä¼˜åŠ¿ï¼› æœ€åï¼Œæ–‡ç« è®¨è®ºäº†ä¸¤ä¸ªå¯èƒ½å½±å“Pre-train modelæ€§èƒ½è¡¨ç°çš„å› ç´ ã€‚ 4.1 è®­ç»ƒæ ·æœ¬é‡ é¦–å…ˆï¼Œæ–‡ç« ç ”ç©¶äº†è®­ç»ƒæ ·æœ¬é‡å¯¹äºé¢„è®­ç»ƒæ¨¡å‹çš„å½±å“ï¼› å¦‚ä¸‹å›¾ï¼Œç»“æœå‘ç°æ•°æ®é‡æœ€å¤šæ—¶ï¼ˆä»30Kåˆ°33Mï¼‰ï¼Œæ¨¡å‹åœ¨å¤šç§ä¸‹æ¸¸ä»»åŠ¡ä¸­çš„è¡¨ç°è¶Šå¥½ï¼› As larger and more diverse datasets become available, we can anticipate further improvements in model performance, advancing our understanding of cellular processes. 4.2 è®­ç»ƒæ ·æœ¬ç±»å‹ æ–‡ç« è¿›ä¸€æ­¥ç ”ç©¶äº†ç»†èƒæ¥æºå¯¹äºæ¨¡å‹çš„å½±å“ï¼› åœ¨COVID-19æ•°æ®é›†çš„å¤šæ‰¹æ¬¡åˆå¹¶ä»»åŠ¡ä¸­ï¼Œç»“æœå‘ç°ï¼š ä½¿ç”¨å…¨éƒ¨33Mæ•°æ®ä¸ä»…ä½¿ç”¨Bloodæ•°æ®çš„Foundation modelçš„æ€§èƒ½æ¯”è¾ƒæ¥è¿‘ï¼› Lungæ¥æºçš„ç»†èƒé‡å°½ç®¡åªæœ‰2.1Mï¼Œè¿œå°‘äºBrainç»†èƒé‡(13.2M)ï¼Œä½†å‰è€…æ€§èƒ½æ˜æ˜¾è¡¨ç°æ›´ä¼˜ã€‚ This emphasizes the importance of aligning the cellular context in pretraining with the target dataset for superior results in downstream tasks. åç»­å°†å‚è€ƒscGPTå·¥å…·çš„æ•™ç¨‹æ‰‹å†ŒåŠæºä»£ç ï¼Œå­¦ä¹ å…¶æ„å»ºç»†èŠ‚ä¸æ•°æ®å¤„ç†æ–¹å¼ç­‰ã€‚\n",
  "wordCount" : "6652",
  "inLanguage": "en",
  "datePublished": "2024-09-29T00:00:00Z",
  "dateModified": "2024-09-29T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/basic/202%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bscgpt/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="ä¸»é¡µ">
                    <span>ä¸»é¡µ</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="åˆ†ç±»">
                    <span>åˆ†ç±»</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="æ ‡ç­¾">
                    <span>æ ‡ç­¾</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="å½’æ¡£">
                    <span>å½’æ¡£</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="å…³äº">
                    <span>å…³äº</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="æœç´¢ (Alt &#43; /)" accesskey=/>
                    <span>æœç´¢</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/">åˆ†ç±»</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/basic/">ğŸ“– ç§‘ç ”åŸºç¡€ -- è¯»æ–‡çŒ®ã€çœ‹æ•™æ</a></div>
    <h1 class="post-title">
      æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹scGPT
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2024-09-29 00:00:00 &#43;0000 UTC&#39;&gt;2024-09-29&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2024-09-29&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;6652&amp;nbsp;|&amp;nbsp;14 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1-%e7%ae%80%e4%bb%8b" aria-label="1. ç®€ä»‹">1. ç®€ä»‹</a><ul>
                            
                    <li>
                        <a href="#11-%e5%85%b3%e4%ba%8e%e4%bd%9c%e8%80%85" aria-label="1.1 å…³äºä½œè€…">1.1 å…³äºä½œè€…</a></li>
                    <li>
                        <a href="#12-%e6%96%87%e7%8c%ae%e6%a6%82%e8%bf%b0" aria-label="1.2 æ–‡çŒ®æ¦‚è¿°">1.2 æ–‡çŒ®æ¦‚è¿°</a></li></ul>
                    </li>
                    <li>
                        <a href="#2-pretrain%e9%a2%84%e8%ae%ad%e7%bb%83" aria-label="2. Pretrainé¢„è®­ç»ƒ">2. Pretrainé¢„è®­ç»ƒ</a><ul>
                            
                    <li>
                        <a href="#20-%e6%95%b0%e6%8d%ae%e8%a7%84%e6%a8%a1" aria-label="2.0 æ•°æ®è§„æ¨¡">2.0 æ•°æ®è§„æ¨¡</a></li>
                    <li>
                        <a href="#21-%e5%88%9d%e5%a7%8b%e5%8c%96%e8%be%93%e5%85%a5" aria-label="2.1 åˆå§‹åŒ–è¾“å…¥">2.1 åˆå§‹åŒ–è¾“å…¥</a></li>
                    <li>
                        <a href="#22-%e8%87%aa%e7%9b%91%e7%9d%a3%e8%ae%ad%e7%bb%83" aria-label="2.2 è‡ªç›‘ç£è®­ç»ƒ">2.2 è‡ªç›‘ç£è®­ç»ƒ</a></li>
                    <li>
                        <a href="#23-%e5%a4%9a%e6%89%b9%e6%ac%a1%e4%b8%8e%e5%a4%9a%e6%a8%a1%e6%80%81%e8%a1%a8%e7%a4%ba" aria-label="2.3 å¤šæ‰¹æ¬¡ä¸å¤šæ¨¡æ€è¡¨ç¤º">2.3 å¤šæ‰¹æ¬¡ä¸å¤šæ¨¡æ€è¡¨ç¤º</a></li></ul>
                    </li>
                    <li>
                        <a href="#3-fine-tune%e5%be%ae%e8%b0%83" aria-label="3. Fine-tuneå¾®è°ƒ">3. Fine-tuneå¾®è°ƒ</a><ul>
                            
                    <li>
                        <a href="#31-%e7%bb%86%e8%83%9e%e7%b1%bb%e5%9e%8b%e6%b3%a8%e9%87%8a" aria-label="3.1 ç»†èƒç±»å‹æ³¨é‡Š">3.1 ç»†èƒç±»å‹æ³¨é‡Š</a></li>
                    <li>
                        <a href="#32-%e5%9f%ba%e5%9b%a0%e6%89%b0%e5%8a%a8%e9%a2%84%e6%b5%8b" aria-label="3.2 åŸºå› æ‰°åŠ¨é¢„æµ‹">3.2 åŸºå› æ‰°åŠ¨é¢„æµ‹</a></li>
                    <li>
                        <a href="#33-%e5%a4%9a%e6%89%b9%e6%ac%a1%e7%bb%84%e5%ad%a6%e6%95%b4%e5%90%88" aria-label="3.3 å¤šæ‰¹æ¬¡/ç»„å­¦æ•´åˆ">3.3 å¤šæ‰¹æ¬¡/ç»„å­¦æ•´åˆ</a></li>
                    <li>
                        <a href="#34-%e5%9f%ba%e5%9b%a0%e8%b0%83%e6%8e%a7%e7%bd%91%e7%bb%9c%e6%9e%84%e5%bb%ba" aria-label="3.4 åŸºå› è°ƒæ§ç½‘ç»œæ„å»º">3.4 åŸºå› è°ƒæ§ç½‘ç»œæ„å»º</a></li></ul>
                    </li>
                    <li>
                        <a href="#4-%e6%a8%a1%e5%9e%8b%e5%bd%b1%e5%93%8d%e5%9b%a0%e7%b4%a0" aria-label="4. æ¨¡å‹å½±å“å› ç´ ">4. æ¨¡å‹å½±å“å› ç´ </a><ul>
                            
                    <li>
                        <a href="#41-%e8%ae%ad%e7%bb%83%e6%a0%b7%e6%9c%ac%e9%87%8f" aria-label="4.1 è®­ç»ƒæ ·æœ¬é‡">4.1 è®­ç»ƒæ ·æœ¬é‡</a></li>
                    <li>
                        <a href="#42-%e8%ae%ad%e7%bb%83%e6%a0%b7%e6%9c%ac%e7%b1%bb%e5%9e%8b" aria-label="4.2 è®­ç»ƒæ ·æœ¬ç±»å‹">4.2 è®­ç»ƒæ ·æœ¬ç±»å‹</a>
                    </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p><strong>æ–‡çŒ®</strong>ï¼š scGPT: toward building a foundation model for single-cell multi-omics using generative AI</p>
<p><strong>æ—¶é—´</strong>ï¼š2024 Feb. (Published)</p>
<p><strong>æœŸåˆŠ</strong>ï¼šNature Method</p>
<p><strong>DOI</strong>ï¼šhttps://doi.org/10.1038/s41592-024-02201-0</p>
<p><strong>Github</strong>ï¼šhttps://github.com/bowang-lab/scGPT</p></blockquote>
<h1 id="1-ç®€ä»‹">1. ç®€ä»‹<a hidden class="anchor" aria-hidden="true" href="#1-ç®€ä»‹">#</a></h1>
<h2 id="11-å…³äºä½œè€…">1.1 å…³äºä½œè€…<a hidden class="anchor" aria-hidden="true" href="#11-å…³äºä½œè€…">#</a></h2>
<p>é€šè®¯ä½œè€…ï¼š<strong>Bo Wang</strong>ï¼Œå¤šä¼¦å¤šå¤§å­¦</p>
<ul>
<li><a href="https://wanglab.ai/">https://wanglab.ai/</a></li>
<li><a href="https://github.com/bowang-lab/">https://github.com/bowang-lab/</a></li>
<li><a href="https://scholar.google.com/citations?hl=en&amp;user=37FDILIAAAAJ">https://scholar.google.com/citations?hl=en&user=37FDILIAAAAJ</a></li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928131111066.png" alt="image-20240928131111066"  />
</p>
<h2 id="12-æ–‡çŒ®æ¦‚è¿°">1.2 æ–‡çŒ®æ¦‚è¿°<a hidden class="anchor" aria-hidden="true" href="#12-æ–‡çŒ®æ¦‚è¿°">#</a></h2>
<ul>
<li>
<p>å—å¯å‘äºGPTç­‰åŸºäºtransformerçš„å¤§æ¨¡å‹ï¼ŒWangå›¢é˜ŸåŸºäºå¤§é‡å•ç»†èƒè½¬å½•ç»„æ•°æ®ï¼Œæ„å»ºäº†scGPT foundation modelï¼Œå¯ç”¨äºå¤šç§å¸¸è§çš„å•ç»†èƒä¸‹æ¸¸åˆ†æä»»åŠ¡ï¼Œæ€§èƒ½ä¼˜äºå…¶å®ƒå·²æœ‰çš„åˆ†æå·¥å…·ã€‚</p>
</li>
<li>
<p>æ ¸å¿ƒå‡è®¾ï¼š</p>
<ul>
<li>åœ¨NLPé¢†åŸŸä¸­ï¼Œä¸€ä¸ªå¥å­ç”±å¤šä¸ªå•è¯ç»„æˆã€‚ç»è¿‡é¢„è®­ç»ƒçš„è‡ªç›‘ç£å­¦ä¹ ï¼Œfoundation modelå¯ä»¥æå–è¾“å…¥å¥å­ä¸­æ¯ä¸ªå•è¯(æˆ–è€…å¥å­æ•´ä½“)çš„é«˜çº§è¡¨ç¤º(Embedding)ï¼›</li>
<li>è€Œåœ¨scRNA-seqä¸­ï¼Œä¸€ä¸ªç»†èƒå¯ä»¥è®¤ä¸ºç”±æ‰€æœ‰åŸºå› çš„ä¸åŒç¨‹åº¦è¡¨è¾¾å®šä¹‰ã€‚ç»è¿‡é¢„è®­ç»ƒçš„è‡ªç›‘ç£å­¦ä¹ ï¼Œfoundation modelå¯ä»¥æå–è¾“å…¥å•ç»†èƒè¡¨è¾¾æ•°æ®çš„æ¯ä¸ªåŸºå› (æˆ–è€…ç»†èƒæ•´ä½“)çš„é«˜çº§è¡¨ç¤º(Embedding)ã€‚</li>
</ul>
</li>
<li>
<p>åœ¨Pre-trained foundation modelsåŸºç¡€ä¸Šï¼Œå°†æå–çš„åŸºå› /ç»†èƒEmbeddingåœ¨å…·ä½“çš„å•ç»†èƒä¸‹æ¸¸åˆ†æä»»åŠ¡ä¸­è¿›è¡ŒäºŒæ¬¡å¾®è°ƒ(fine-tune)ï¼Œå‘æŒ¥å¤§æ•°æ®ç”Ÿæˆçš„å¼ºå¤§ä¼˜åŠ¿ã€‚</p>
</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928131642668.png" alt="image-20240928131642668"  />
</p>
<h1 id="2-pretrainé¢„è®­ç»ƒ">2. Pretrainé¢„è®­ç»ƒ<a hidden class="anchor" aria-hidden="true" href="#2-pretrainé¢„è®­ç»ƒ">#</a></h1>
<blockquote>
<p>Pre-train foundation modelçš„æ ¸å¿ƒåŠŸèƒ½å°±æ˜¯ä»¥ä¸€ä¸ªå•ç»†èƒçš„æ•°æ®ä¸ºè¾“å…¥ï¼Œå°†æå–çš„Gene/Cell Embeddingä¸ºè¾“å‡ºã€‚</p></blockquote>
<h2 id="20-æ•°æ®è§„æ¨¡">2.0 æ•°æ®è§„æ¨¡<a hidden class="anchor" aria-hidden="true" href="#20-æ•°æ®è§„æ¨¡">#</a></h2>
<ul>
<li>æ–‡ç« ä»CELLxGENEç½‘ç«™æ”¶é›†äº†33Mä¸ªæ­£å¸¸ç»†èƒçš„scRNA-seqæ•°æ®ç”¨äºè®­ç»ƒFoundation model</li>
<li>å…¶ä¸­æ¥è‡ªBrainã€Bloodç­‰å™¨å®˜çš„ç»†èƒæ•°æœ€å¤š</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240929091250859.png" alt="image-20240929091250859"  />
</p>
<h2 id="21-åˆå§‹åŒ–è¾“å…¥">2.1 åˆå§‹åŒ–è¾“å…¥<a hidden class="anchor" aria-hidden="true" href="#21-åˆå§‹åŒ–è¾“å…¥">#</a></h2>
<ul>
<li>scGPTçš„åŸå§‹è¾“å…¥æ•°æ®é€šå¸¸ä¸ºNä¸ªç»†èƒï¼ŒGä¸ªåŸºå› çš„Countè¡¨è¾¾çŸ©é˜µã€‚æ­¤æ—¶ï¼Œéœ€è¦å¯¹æ¯ä¸ªç»†èƒçš„åŸºå› æ•°æ®è¿›è¡Œåˆå§‹åŒ–å¤„ç†ï¼Œä½œä¸ºscGPTçš„æ ‡å‡†è¾“å…¥ï¼›</li>
<li>æ¦‚æ‹¬æ¥è¯´ï¼Œåˆ†åˆ«ä»æ¯ä¸ªåŸºå› çš„ä¸‰ä¸ªè§’åº¦è¿›è¡ŒDç»´åµŒå…¥ç¼–ç  (1Ã—D)ï¼Œç„¶åå†è¿›è¡ŒçŸ©é˜µåŠ æ³•ï¼ˆä»ä¸º1Ã—Dï¼‰ï¼Œæœ€ç»ˆå°†æ‰€æœ‰åŸºå› åˆå¹¶å¾—åˆ°ï¼ˆMÃ—Dï¼‰çš„ç»†èƒç‰¹å¾çŸ©é˜µã€‚</li>
</ul>
<blockquote>
<p>å› ä¸ºç»†èƒçš„åŸºå› ä¼šæ ¹æ®æƒ…å†µè¿›è¡Œé€‰æ‹©ï¼Œå¹¶ä¸”è€ƒè™‘åˆ°&lt;cls&gt;ç­‰ç‰¹æ®Šè¯å…ƒï¼Œæ‰€ä»¥è¿™é‡Œå¹¶ä¸æ˜¯GÃ—D</p></blockquote>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928150405475.png" alt="image-20240928150405475"  />
</p>
<p><strong>ï¼ˆ1ï¼‰Gene token</strong></p>
<ul>
<li>æ‰€æœ‰åŸºå› (è¯å…ƒ)æ„æˆçš„è¯è¡¨ä¸­ï¼Œæ¯ä¸ªåŸºå› éƒ½æœ‰ä¸€ä¸ªå¯¹åº”çš„æ•´æ•°æ ‡è¯†ç¬¦ï¼ˆInteger Identifierï¼‰</li>
<li>æ­¤å¤–è¿˜æœ‰å…¶å®ƒç‰¹æ®Šçš„è¯å…ƒï¼Œä¾‹å¦‚&lt;cls&gt;, &lt;pad&gt;ç­‰
<ul>
<li>&lt;cls&gt;è¯å…ƒé€šå¸¸æ”¾åœ¨ç»†èƒMä¸ªè¯å…ƒä¸­çš„ç¬¬ä¸€ä¸ªï¼Œä¸è¡¨ç¤ºç‰¹å®šåŸºå› ï¼Œè€Œç”¨äºcell representation</li>
<li>&lt;pad&gt;è¯å…ƒç”¨äºè¡¥é•¿è‡³å›ºå®šé•¿åº¦M(e.g. æœ‰äº›æƒ…å†µä¸è€ƒè™‘è¡¨è¾¾å€¼ä¸º0çš„åŸºå› æ—¶)</li>
</ul>
</li>
<li>é‡‡ç”¨PyTorch embedding layerå°†æŠ•å°„æˆDç»´çš„Embedding</li>
</ul>
<p><strong>ï¼ˆ2ï¼‰Gene expression</strong></p>
<ul>
<li>å•ç»†èƒçš„countè¡¨è¾¾å€¼åœ¨ä¸åŒæµ‹åºèƒŒæ™¯çš„æ¡ä»¶ä¸‹ï¼Œä¸å…·æœ‰å¯æ¯”æ€§ã€‚</li>
<li>scGPTé‡‡ç”¨binåˆ†æ®µå¤„ç†ï¼Œå°†å•ä¸ªç»†èƒiçš„éé›¶åŸºå› è¡¨è¾¾å€¼ï¼Œåˆ†ä¸ºkä¸ªbinsï¼Œåˆ†æ•°åˆ†åˆ«æ˜¯1..kã€‚
<ul>
<li>ä¾‹å¦‚ï¼Œè‹¥åŸºå› binæ‰“åˆ†ä¸ºkï¼Œè¡¨ç¤ºå…¶è¡¨è¾¾å€¼å¤„äºæœ€é«˜çš„binèŒƒå›´å†…ï¼›</li>
<li>æ¯ä¸ªç»†èƒåˆ†binçš„é˜ˆå€¼æ ‡å‡†éƒ½ä¼šä¸å°½ç›¸åŒã€‚</li>
</ul>
</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928144527642.png" alt="image-20240928144527642"  />
</p>
<ul>
<li>é‡‡ç”¨MLPï¼Œè½¬æ¢ä¸ºDç»´çš„Embedding</li>
</ul>
<p><strong>ï¼ˆ3ï¼‰Condition token</strong></p>
<ul>
<li>è®°å½•åŸºå› é¢å¤–çš„Condition tokenï¼Œä¾‹å¦‚æ˜¯å¦ä¸ºperturbated genes</li>
<li>é‡‡ç”¨PyTorch embedding layerå°†æŠ•å°„æˆDç»´çš„Embedding</li>
</ul>
<p>ç»¼ä¸Šï¼Œä¸€ä¸ªç»†èƒ(i)çš„åˆå§‹åŒ–è¾“å…¥h (M Ã— D)çš„è®¡ç®—æ–¹å¼å¦‚ä¸‹å›¾æ‰€ç¤ºã€‚å…¶ä¸­ç¬¬ä¸€åˆ—é€šå¸¸è¡¨ç¤ºä¸º&lt;cls&gt;ä½œä¸ºCell representationï¼›å…¶å®ƒåˆ—ï¼ˆé™¤&lt;pad&gt;ç­‰ç‰¹æ®Šå­—ç¬¦å¤–ï¼Œå‡è¡¨ç¤ºGene Embeddingã€‚</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928143902797.png" alt="image-20240928143902797"  />
</p>
<h2 id="22-è‡ªç›‘ç£è®­ç»ƒ">2.2 è‡ªç›‘ç£è®­ç»ƒ<a hidden class="anchor" aria-hidden="true" href="#22-è‡ªç›‘ç£è®­ç»ƒ">#</a></h2>
<p><strong>ï¼ˆ1ï¼‰Transformerå—</strong></p>
<ul>
<li>Foundation modelä¸»è¦ç”± 12 (<code>l</code>)ä¸ªTransformerå—ç»„æˆï¼Œæ¯ä¸ªå—åˆ™å‡é‡‡ç”¨äº†å¤šå¤´ï¼ˆ8ï¼‰è‡ªæ³¨æ„åŠ›æœºåˆ¶ï¼›</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928152927329.png" alt="image-20240928152927329"  />
</p>
<ul>
<li>
<p>ç”±äºè¿™é‡Œçš„Mé€šå¸¸è¾ƒå¤§ï¼Œå³ä¸€ä¸ªç»†èƒè€ƒè™‘æ•°åƒä¸Šä¸‡ä¸ªåŸºå› ï¼ˆå¯ä»¥ç†è§£ä¸ºç‰¹åˆ«é•¿çš„å¥å­ï¼‰ï¼ŒscGPTé‡‡ç”¨äº†FlashAttentionç®—æ³•ç”¨ä»¥åŠ é€Ÿè‡ªæ³¨æ„åŠ›è®¡ç®—ã€‚</p>
<blockquote>
<p>æ­¤å¤–ï¼Œå¯¹äºæ¯ä¸ªç»†èƒï¼Œåªæœ‰è¡¨è¾¾å€¼é0çš„åŸºå› å‚ä¸é¢„è®­ç»ƒè¿‡ç¨‹ï¼Œä»¥æé«˜é€Ÿåº¦ã€‚</p></blockquote>
</li>
<li>
<p>ä½œè€…ä¹Ÿæ¨èäº†å…¶å®ƒé«˜æ•ˆè®¡ç®—çš„Transformerå˜ä½“ï¼ŒåŒ…æ‹¬linear complexity (Linformer)ï¼ŒKernelized Self-Attention (KSA)ã€‚</p>
</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928151221844.png" alt="image-20240928151221844" style="zoom:67%;" />
<p><strong>ï¼ˆ2ï¼‰è‡ªç›‘ç£ä»»åŠ¡</strong></p>
<ul>
<li>scGPTè‡ªç›‘ç£ä»»åŠ¡çš„æ ¸å¿ƒæ˜¯é¢„æµ‹æ©ç (masked)åŸºå› çš„è¡¨è¾¾å€¼æ°´å¹³ï¼Œé‡‡ç”¨MSEæŸå¤±å‡½æ•°ã€‚</li>
<li>å…·ä½“åˆ†ä¸ºå¦‚ä¸‹ä¸¤ä¸ªå­ä»»åŠ¡ï¼š
<ul>
<li>Gene-promptï¼šåŸºäºå·²çŸ¥è¡¨è¾¾å€¼çš„Gene Embeddingï¼Œé¢„æµ‹æœªçŸ¥è¡¨è¾¾å€¼Geneçš„expression valueï¼›</li>
<li>Cell-promptï¼šåŸºäº&lt;cls&gt; cell representationé¢„æµ‹å…¨åŸºå› çš„expression valueã€‚</li>
</ul>
</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928154530333.png" alt="image-20240928154530333"  />
</p>
<ul>
<li>ç„¶åï¼Œå°†ä¸Šè¿°ä¸¤ç§æ¨¡å¼çš„æŸå¤±lossç›¸åŠ åï¼Œå†è®¡ç®—æ¢¯åº¦å¹¶æ›´æ–°æ¨¡å‹å‚æ•°ã€‚</li>
</ul>
<p><strong>ï¼ˆ3ï¼‰Masked Attention</strong></p>
<ul>
<li>è‡ªæ³¨æ„è®¡ç®—è¿‡ç¨‹ä¸ºï¼šå°†masked gene Embedding( without expression) ä½œä¸ºqueryï¼Œè®¡ç®—ä¸å…¶å®ƒknown gene Embeddingï¼ˆä¹ŸåŒ…æ‹¬è‡ªå·±ï¼‰çš„æ³¨æ„åŠ›æƒé‡åï¼Œå†è¿›è¡ŒåŠ æƒè¿ç®—å¾—åˆ°è¾“å‡ºï¼›</li>
<li>ä¸GPTæ¨¡å‹æ‰€å¤„ç†çš„æ–‡æœ¬å¥ä¸åŒä¹‹å¤„åœ¨äºï¼šthe <strong>non-sequential</strong> nature of the genes in one cell. ä¸ºæ­¤ä½œè€…è®¾è®¡äº†masked attentionæ³¨æ„åŠ›è®¡ç®—æ–¹å¼ï¼›</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928155607273.png" alt="image-20240928155607273"  />
</p>
<ul>
<li>å¦‚ä¸Šå…¬å¼ï¼Œåœ¨ä¸è€ƒè™‘<strong>A</strong>maskçš„æƒ…å†µä¸‹ï¼Œä¸ºæ ‡å‡†çš„è‡ªæ³¨æ„åŠ›è®¡ç®—æ–¹å¼ã€‚è€Œ<strong>A</strong>maskå¯å‚è€ƒå¦‚ä¸‹å…¬å¼ï¼Œä»¥åŠä¸‹å›¾å·¦Aï¼ˆè¡Œè¡¨ç¤ºqueryï¼Œåˆ—è¡¨ç¤ºkeyï¼‰ç†è§£ã€‚æ¯ä¸€ä¸ªqueryï¼ˆiï¼‰è®¡ç®—ä¸å…¶ä»–è¯å…ƒæ³¨æ„åŠ›ï¼ˆåŒ…æ‹¬ä¸å®ƒè‡ªå·±ï¼‰æ—¶ï¼š
<ul>
<li>è‹¥key ï¼ˆjï¼‰ä¸æ˜¯unknown expression geneï¼Œåˆ™ä¸º0ï¼›</li>
<li>è‹¥i = jæ—¶ï¼Œä¸”jæ˜¯unknown expression geneï¼Œåˆ™ä¸º0ï¼ˆè‡ªå·±ä¸è‡ªå·±çš„æ³¨æ„åŠ›è®¡ç®—ï¼‰</li>
<li>å…¶å®ƒæƒ…å†µä¸‹ï¼Œåˆ™ä¸ºè´Ÿæ— ç©·ï¼ˆå¯¹åº”å›¾ä¸­çš„æ·±è“è‰²å•å…ƒæ ¼ï¼‰</li>
<li>æ¯ä¸€è¡Œ(i)è¡¨ç¤ºä¸€ä¸ªqueryè¯å…ƒä¸è¯¥ç»†èƒæ‰€æœ‰è¯å…ƒçš„æ³¨æ„åŠ›è®¡ç®—ã€‚</li>
</ul>
</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928160244274.png" alt="image-20240928160244274" style="zoom:67%;" />
<ul>
<li>a = 0è¡¨ç¤ºä¸äº§ç”Ÿä»»ä½•å½±å“ï¼Œa = -infè¡¨ç¤ºå°†query(i)å¯¹äºkey(j)çš„æ³¨æ„åŠ›ç½®æ¢ä¸º-infï¼Œç»softmaxè½¬æ¢ååˆ™å˜ä¸º0ã€‚</li>
</ul>
<blockquote>
<p>The rule of thumb for scGPT attention masking is to only allow attention computation between embeddings of the â€˜known genesâ€™ and the query gene itself.</p></blockquote>
<img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928160540771.png" alt="image-20240928160540771" style="zoom: 50%;" />
<ul>
<li>åœ¨ä¸€ä¸ªcellçš„å…¨éƒ¨masked geneè¡¨è¾¾å€¼é¢„æµ‹çš„è¿‡ç¨‹ä¸­ï¼Œé‡‡ç”¨å¤šè½®ï¼ˆkï¼‰è¿­ä»£é¢„æµ‹çš„æ€è·¯ï¼ˆå¦‚ä¸Šå›¾å³ï¼Œç±»æ¯”GPTçš„è‡ªå›å½’è®­ç»ƒï¼‰
<ul>
<li>åœ¨ç¬¬ä¸€è½®ä¸­ï¼Œå¯¹äºæ‰€æœ‰masked tokençš„é¢„æµ‹ï¼Œå°†å…¶ä¸­1/kä¸ªhigh prediction confidenceçš„tokenæ ‡æ³¨ä¸ºknown genesã€‚</li>
<li>ç„¶ååœ¨æ–°ä¸€è½®è¿­ä»£ä¸­ï¼Œé‡å¤ä¸Šä¸€æ­¥éª¤ï¼Œç›´è‡³é¢„æµ‹å‡ºæ‰€æœ‰çš„masked tokenã€‚</li>
<li>å€¼å¾—æ³¨æ„çš„æ˜¯ï¼Œæ–‡ç« å¹¶æ²¡æœ‰è¯´æ˜prediction confidenceæ˜¯å¦‚ä½•è®¡ç®—çš„ã€‚ä¸ªäººç†è§£æœ¬è´¨è¿˜æ˜¯åŸºäºä¸çœŸå®è¡¨è¾¾å€¼ä¹‹é—´çš„è¯¯å·®ã€‚</li>
</ul>
</li>
</ul>
<blockquote>
<p>In each generation iteration, scGPT predicts the gene expression values of a new set of genes, and these genes in turn become the â€˜known genesâ€™ in the next iteration for attention computation.</p></blockquote>
<h2 id="23-å¤šæ‰¹æ¬¡ä¸å¤šæ¨¡æ€è¡¨ç¤º">2.3 å¤šæ‰¹æ¬¡ä¸å¤šæ¨¡æ€è¡¨ç¤º<a hidden class="anchor" aria-hidden="true" href="#23-å¤šæ‰¹æ¬¡ä¸å¤šæ¨¡æ€è¡¨ç¤º">#</a></h2>
<ul>
<li>åœ¨å¤šæ‰¹æ¬¡ã€å¤šç»„å­¦åˆå¹¶ä¸‹æ¸¸ä»»åŠ¡ä¸­ï¼Œéœ€è¦é¢å¤–çš„token Embeddingä»¥è¡¨ç¤ºå¿…è¦çš„æ‰¹æ¬¡å’Œç»„å­¦ä¿¡æ¯ï¼Œä¾›æ¨¡å‹å­¦ä¹ ï¼›</li>
<li>ä½†æ˜¯scGPTåœ¨Transformerçš„pre-trainè¿‡ç¨‹ä¸­ï¼Œå¹¶æœªåŠ å…¥ç›¸å…³ä¿¡æ¯ã€‚è€Œæ˜¯åœ¨foundation modelçš„è¾“å‡ºç»“æœä¸­å†è¿›è¡Œæ‹¼æ¥æ“ä½œï¼Œå³æ¨¡å‹åœ¨Pre-trainåï¼ŒFine-tuneå‰å¼•å…¥æ‰¹æ¬¡å’Œæ¨¡æ€ä¿¡æ¯ï¼Œä»¥å¸Œæœ›åœ¨å¾®è°ƒè¿‡ç¨‹ä¸­å­¦ä¹ åˆ°ç›¸å…³ä¿¡æ¯ã€‚</li>
<li>è¿™æ ·åšçš„ä¸»è¦åŸå› æ˜¯ï¼šå¦‚æœæ‰¹æ¬¡å’Œæ¨¡æ€ä¿¡æ¯åœ¨è¾“å…¥é˜¶æ®µè¢«å¼•å…¥ï¼ŒTransformerçš„è‡ªæ³¨æ„åŠ›æœºåˆ¶å¯èƒ½ä¼šè¿‡åº¦å…³æ³¨åŒä¸€æ¨¡æ€æˆ–æ‰¹æ¬¡å†…çš„ç‰¹å¾ï¼Œå¯¼è‡´æ¨¡å‹å¿½ç•¥è·¨æ¨¡æ€æˆ–æ‰¹æ¬¡çš„é‡è¦å…³è”ã€‚</li>
</ul>
<blockquote>
<p>This is to prevent the transformer from amplifying the attention within features of the same modalities while underestimating those of different modalities.</p></blockquote>
<ul>
<li>modality tokens (tm)ï¼šè¡¨ç¤ºè¯å…ƒå¯¹åº”çš„featureæ˜¯Gene/Protein/Peakä¸­çš„å“ªä¸€ç§ï¼›</li>
<li>batch tokens (tb)ï¼šè¡¨ç¤ºtokenæ˜¯å¦æ¥è‡ªä¸€ä¸ªæ‰¹æ¬¡ï¼Œé€šå¸¸å¯¹äºç»†èƒæ¥è¯´çš„ã€‚å› æ­¤ä¸€ä¸ªç»†èƒçš„batch tokenéƒ½æ˜¯ä¸€æ ·çš„ã€‚</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240924205846107.png" alt="image-20240924205846107"  />
</p>
<ul>
<li>å¦‚ä¸‹å…¬å¼ï¼Œè¡¨ç¤ºåŒæ—¶å­˜åœ¨å¤šæ‰¹æ¬¡ä»¥åŠå¤šç»„å­¦çš„æƒ…å†µ</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240924211913127.png" alt="image-20240924211913127"  />
</p>
<ul>
<li>å¦‚ä¸‹å…¬å¼ï¼Œè¡¨ç¤ºåŒç§ç»„å­¦ï¼Œå­˜åœ¨å¤šæ‰¹æ¬¡çš„æƒ…å†µ</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240924212230354.png" alt="image-20240924212230354"  />
</p>
<h1 id="3-fine-tuneå¾®è°ƒ">3. Fine-tuneå¾®è°ƒ<a hidden class="anchor" aria-hidden="true" href="#3-fine-tuneå¾®è°ƒ">#</a></h1>
<h2 id="31-ç»†èƒç±»å‹æ³¨é‡Š">3.1 ç»†èƒç±»å‹æ³¨é‡Š<a hidden class="anchor" aria-hidden="true" href="#31-ç»†èƒç±»å‹æ³¨é‡Š">#</a></h2>
<p><strong>ï¼ˆ1ï¼‰Method</strong></p>
<ul>
<li>
<p>å¯¹äºFoundation modelæå–çš„æ¯ä¸ªç»†èƒ<strong>Cell Embedding (&lt;cls&gt;)</strong> æ„å»ºä¸€ä¸ªMLPåˆ†ç±»å™¨ï¼Œç”¨ä»¥é¢„æµ‹ç»†èƒçš„ç±»åˆ«ï¼Œå¹¶ä½¿ç”¨äº¤å‰ç†µä½œä¸ºæŸå¤±å‡½æ•°ã€‚</p>
</li>
<li>
<p>é¦–å…ˆä½¿ç”¨ä¸€ä¸ªæ ‡æ³¨ç»†èƒç±»å‹çš„æ•°æ®é›†ï¼ˆreference setï¼‰è¿›è¡ŒFine-tuneï¼Œç„¶åå†ä½¿ç”¨ä¸€ä¸ªHeld-outæ•°æ®é›†éªŒè¯</p>
</li>
<li>
<p>è¾“å…¥æ•°æ®å‰çš„é¢„å¤„ç†ï¼š</p>
<ul>
<li>common set between foundation model and the input data</li>
<li>Gene expression: normalization â†’ log1 â†’ bin</li>
<li>All gene as input (include zero expression)</li>
</ul>
</li>
</ul>
<p><strong>ï¼ˆ2ï¼‰Result</strong></p>
<ul>
<li>
<p>åœ¨Human pancreas (èƒ°è…º) dataset</p>
<ul>
<li>å¯¹äºæ¯ç§ç»†èƒçš„é¢„æµ‹Precisionéƒ½è¾¾åˆ°0.8ä»¥ä¸Š (é¢„æµ‹ä¸ºè¯¥ç»†èƒç±»å‹ä¸­ï¼Œå®é™…ä¸ºè¯¥ç»†èƒç±»å‹çš„æ¯”ä¾‹)</li>
</ul>
</li>
<li>
<p>ä»¥Human immune cellsè¿›è¡Œfine-tuneï¼Œé¢„æµ‹å¤šå‘æ€§ç¡¬åŒ–ç—‡(MS)çš„ç»†èƒç±»å‹</p>
<ul>
<li>å¹³å‡å‡†ç¡®ç‡Accuracyå¯ä»¥è¾¾åˆ°0.85å·¦å³</li>
</ul>
</li>
<li>
<p>ä½¿ç”¨6ç§è‚¿ç˜¤ç»†èƒç±»å‹ä½œä¸ºfine-tuneï¼Œé¢„æµ‹3ç§å…¶å®ƒè‚¿ç˜¤çš„ç»†èƒç±»å‹</p>
<ul>
<li>åœ¨è‚¿ç˜¤å¾®ç¯å¢ƒç»†èƒç±»å‹é¢„æµ‹åŒæ ·è¡¨ç°è‰¯å¥½</li>
</ul>
</li>
<li>
<p>ä¸TOSICAã€scBERTæ¨¡å‹è¿›è¡Œäº†æ¯”è¾ƒï¼Œå‡è¡¨ç°å‡ºä¸€å®šçš„ä¼˜åŠ¿</p>
</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240926100118451.png" alt="image-20240926100118451"  />
</p>
<blockquote>
<p>æ­¤å¤–ï¼Œä½œè€…è¿˜æå‡ºäº†Reference mappingæ³¨é‡Šæ–¹æ³•ï¼šåœ¨Zero/Fine-tuneæ¨¡å‹æ—¶ï¼Œè·å¾—æ ‡ç­¾ç»†èƒç±»å‹çš„Cell Embeddingã€‚ç„¶åå†è®¡ç®—å‡ºQuery cellsçš„Embeddingã€‚æœ€åæ®æ­¤ï¼Œè®¡ç®—å‡ºæ¯ä¸ªQuery cellæœ€ç›¸è¿‘çš„Reference cellsï¼Œä»è€Œæ³¨é‡Šå…¶ç»†èƒç±»å‹ï¼ˆKNNï¼‰ã€‚</p></blockquote>
<h2 id="32-åŸºå› æ‰°åŠ¨é¢„æµ‹">3.2 åŸºå› æ‰°åŠ¨é¢„æµ‹<a hidden class="anchor" aria-hidden="true" href="#32-åŸºå› æ‰°åŠ¨é¢„æµ‹">#</a></h2>
<p><strong>ï¼ˆ1ï¼‰Method</strong></p>
<ul>
<li>å¯¹äºFoundation modelæå–çš„æ¯ä¸ªç»†èƒçš„Gene Embeddingï¼Œå°†ä¸€éƒ¨åˆ†åŸºå› çš„è¡¨è¾¾å€¼æ©ç ï¼Œä»è€Œè¿›è¡ŒFine-tuneå¾®è°ƒè®­ç»ƒï¼Œå¹¶ä»¥MSEä½œä¸ºæŸå¤±å‡½æ•°ï¼›</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928201056193.png" alt="image-20240928201056193" style="zoom:80%;" />
<ul>
<li>åœ¨Perturbation prediction taskï¼š
<ul>
<li>Input: Gene expressionä¸ºControl cell expression, Condition tokenæ ‡æ³¨ç›¸åº”ä½ç½®Geneæ˜¯å¦è¢«Perturbate</li>
<li>Output: Post-perturbation expression</li>
</ul>
</li>
<li>è¾“å…¥æ•°æ®å‰çš„é¢„å¤„ç†ï¼š
<ul>
<li>only select HVGs for training</li>
<li>Log1p expression instead of binned values ï¼ˆoutputç›¸åŒï¼‰</li>
</ul>
</li>
</ul>
<p><strong>ï¼ˆ2ï¼‰Result</strong></p>
<ul>
<li>åˆ†åˆ«åŸºäºä¸‰ä¸ªå•ç»†èƒCRISPRæ•°æ®é›†ï¼Œæ¯”è¾ƒäº†scGPTä¸Linearã€GEARSæ¨¡å‹çš„æ€§èƒ½è¡¨ç°
<ul>
<li>Adamson: 87 one-gene perturbationsï¼›</li>
<li>Replogle: 2823 one-gene perturbationsï¼›</li>
<li>Norman: 131 two-gene and 105 one-gene perturbations</li>
</ul>
</li>
<li>å¯¹äºæ¯ä¸ªæ•°æ®é›†ï¼Œä½¿ç”¨ä¸€éƒ¨åˆ†çš„æ‰°åŠ¨æ•°æ®è¿›è¡Œfine-tuneï¼Œå†å¯¹å…¶ä½™unseen geneçš„æ‰°åŠ¨æ•°æ®ä½œä¸ºtest</li>
<li>è¯„ä»·æŒ‡æ ‡ä¸ºè®¡ç®—é¢„æµ‹ä¸çœŸå®çš„post-perturbation expressionçš„ç›¸å…³æ€§ (åŸºäºå…¨éƒ¨åŸºå› ï¼Œæˆ–è€…æ˜¯å‰20ä¸ªå½±å“æœ€æ˜¾è‘—çš„åŸºå› )
<ul>
<li>ç»“æœå‘ç°scGPTæ¨¡å‹ç›¸æ¯”äºå…¶å®ƒä¸¤ç§æ¨¡å‹ï¼Œæé«˜äº†5â€“20%</li>
</ul>
</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928203804558.png" alt="image-20240928203804558"  />
</p>
<p><strong>ï¼ˆ3ï¼‰In silicon reverse perturbation prediction</strong></p>
<ul>
<li>ç®€å•ç†è§£ï¼Œæ ¹æ®perturbation expressionçš„ç»“æœåå‘é¢„æµ‹æ˜¯æœ€æœ‰å¯èƒ½å“ªä¸ªåŸºå› è¢«æ‰°åŠ¨ï¼›</li>
<li>å‚è€ƒä½œè€…åœ¨Github issue ï¼ˆhttps://github.com/bowang-lab/scGPT/issues/87ï¼‰çš„è§£ç­”ï¼š
<ul>
<li>å…¶Fine-tuneæ­¥éª¤å…¶å®ä¸ä¸Šé¢Perturbation prediction taskåŸºæœ¬ä¸€è‡´ï¼›</li>
<li>ä½¿ç”¨æ•°æ®é›†çš„ä¸€éƒ¨åˆ†è¿›è¡Œå¾®è°ƒåï¼Œå†é¢„æµ‹æ‰€æœ‰ç›¸å…³åŸºå› æ‰°åŠ¨çš„è¡¨è¾¾ç»“æœï¼›</li>
<li>æœ€åä½¿ç”¨çœŸå®çš„Query perturbationä¸Šè¿°çš„é¢„æµ‹ç»“æœè¿›è¡ŒKNNå…³è”åˆ†æã€‚</li>
</ul>
</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928205214389.png" alt="image-20240928205214389"  />
</p>
<blockquote>
<p>The model was fine-tuned in the same way as in the &ldquo;forward&rdquo; perturbation prediction. It used a subset of the dataset as we illustrated in Figure 3F. The reverse perturbation task utilized the model in a different way. To summarize, the result cell states of <strong>all possible perturbations were predicted by the fine-tuned model</strong>, and then <strong>an actual sequenced cell state can query</strong> all the predicted cell states in a nearest neighbor search manner, so that the retrieved neighbors indicate the possible origin perturbations.</p></blockquote>
<h2 id="33-å¤šæ‰¹æ¬¡ç»„å­¦æ•´åˆ">3.3 å¤šæ‰¹æ¬¡/ç»„å­¦æ•´åˆ<a hidden class="anchor" aria-hidden="true" href="#33-å¤šæ‰¹æ¬¡ç»„å­¦æ•´åˆ">#</a></h2>
<p><strong>ï¼ˆ1ï¼‰Methodï¼ˆMulti-batchï¼‰</strong></p>
<ul>
<li>
<p>æ ¸å¿ƒç›®æ ‡ï¼šå¯¹äºå¤šæ‰¹æ¬¡çš„scRNA-seqæ•°æ®ï¼Œä¼˜åŒ–ä¸åŒæ‰¹æ¬¡ä¸­æ¯ä¸ªç»†èƒçš„cell representation</p>
<ul>
<li>correct batch effects while preserving biological variance</li>
</ul>
</li>
<li>
<p>è¾“å…¥æ•°æ®çš„é¢„å¤„ç†</p>
<ul>
<li>
<p>Common set between foundation model and the input data</p>
</li>
<li>
<p>Gene expression: normalization&ndash;log1&ndash;bin</p>
</li>
<li>
<p>All genes as input (include zero expression)</p>
</li>
<li>
<p>å¦‚2.3ä¸­æ‰€è¿°ï¼Œåœ¨å¾®è°ƒå‰ï¼Œè¿˜éœ€è¦åœ¨Foundation modelçš„è¾“å‡ºç»“æœä¸­ï¼Œè¡¥å……ç»†èƒçš„æ‰¹æ¬¡ä¿¡æ¯ã€‚</p>
</li>
</ul>
</li>
<li>
<p>å…·ä½“åœ¨Fine-tuneè®­ç»ƒä¸­ï¼Œè®¾è®¡äº†å¯¹å¤šä¸ªç›®æ ‡å‡½æ•°è¿›è¡ŒæŸå¤±è®¡ç®—ï¼Œä»¥å…±åŒç”¨äºæ¨¡å‹ä¼˜åŒ–</p>
</li>
</ul>
<p>1ï¼‰GEPï¼šåŸºå› è¡¨è¾¾é¢„æµ‹ï¼Œå‚è€ƒ3.2ï¼›</p>
<p>2ï¼‰GEPC, Gene expression prediction for cell modelingï¼Œå³ä¸ºäº†ä¼˜åŒ–Cell Embedding(hc)ï¼Œè€Œé¢„æµ‹åŸºå› è¡¨è¾¾ã€‚</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928212510674.png" alt="image-20240928212510674"  />
</p>
<p>3ï¼‰ECS, Elastic cell similarityï¼šåŸºäºä¸€ä¸ªmini-batchçš„ä¸¤ä¸ªç»†èƒçš„ç›¸ä¼¼åº¦è®¡ç®—æŸå¤±å‡½æ•°ã€‚ä½¿å¾—é«˜äºæŸä¸ªé˜ˆå€¼çš„ä¸¤ä¸ªç»†èƒæ›´ç›¸ä¼¼ï¼Œä½äºçš„åˆ™æ›´è¿œç¦»ã€‚</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240928213604728.png" alt="image-20240928213604728"  />
</p>
<p>4ï¼‰Domain adaptation via reverse back propagationï¼šä¸»è¦ä½œç”¨æ˜¯ä½¿å¾—ç”Ÿæˆçš„cell representationè¡¨ç¤ºç»è¿‡MLPåˆ†ç±»å™¨æ— æ³•é¢„æµ‹å…¶æ­£ç¡®çš„batchä¿¡æ¯ï¼ˆç±»ä¼¼äºGAN å¯¹æŠ—ç¥ç»ç½‘ç»œï¼‰ã€‚æœ¬è´¨ä¸Šè¿˜æ˜¯å»ºç«‹MLPé¢„æµ‹Batchçš„åˆ†ç±»å™¨ï¼Œä½†æ˜¯è®¡ç®—æ¢¯åº¦åï¼Œè¿›è¡Œåå‘æ›´æ–°å‚æ•°ã€‚</p>
<p><strong>ï¼ˆ2ï¼‰Result</strong></p>
<ul>
<li>
<p>åŸºäºä¸‰ä¸ªæ•°æ®é›†ï¼Œæ¯”è¾ƒäº†scGPTä¸scVIï¼ŒSeuratï¼ŒHarmonyä¸‰ä¸ªå·¥å…·å»æ‰¹æ¬¡çš„æ•ˆæœ</p>
<ul>
<li>
<p>COVID-19: 18 batches</p>
</li>
<li>
<p>PBMC-10K: 2 batches</p>
</li>
<li>
<p>Perirhinal cortex: 2 batches</p>
</li>
</ul>
</li>
<li>
<p>è¯„ä»·æŒ‡æ ‡ï¼šåŸºäºä¸‰ä¸ªæŒ‡æ ‡ï¼ˆNMI/ARI/ASWï¼‰çš„AvgBIOï¼Œå€¼è¶Šé«˜è¡¨æ˜å»æ‰¹æ¬¡æ•ˆæœè¶Šå¥½ã€‚</p>
<ul>
<li>scGPTå‡ä¼˜äºå…¶å®ƒä¸‰ç§å·¥å…·çš„æ•ˆæœã€‚</li>
</ul>
</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240926125712463.png" alt="image-20240926125712463"  />
</p>
<p><strong>ï¼ˆ3ï¼‰Multi-omics integration</strong></p>
<p><strong>Method</strong></p>
<ul>
<li>ç›®æ ‡æ˜¯ä½¿å¾—ç›¸åŒç»†èƒç±»å‹çš„ä¸åŒç»„å­¦æ•°æ®èƒ½å¤Ÿæœ‰ç›¸ä¼¼çš„Cell Embeddingï¼Œä»è€Œåœ¨èšç±»æ—¶æ¯”è¾ƒæ¥è¿‘ã€‚</li>
<li>scGPTä¸»è¦è€ƒè™‘äº†3ç§ç»„å­¦ï¼Œåˆ†åˆ«æ˜¯scRNA-seqã€scATAC-seqï¼ŒSingle-cell proteomicsã€‚å¯¹æ­¤ï¼Œæœ‰ä¸¤ç§æ•°æ®å½¢å¼ï¼š
<ul>
<li>paired settingï¼šä¸€ç¾¤ç»†èƒåŒæ—¶æµ‹å¤šç§ç»„å­¦</li>
<li>mosaic settingï¼šä¸€ç¾¤ç»†èƒæµ‹ä¸€ç§ç»„å­¦ï¼Œå¦ä¸€ç¾¤æµ‹å¦ä¸€ç»„ã€‚</li>
</ul>
</li>
<li>å¯¹äºscRNA-seqå¯ä»¥ç›´æ¥ç»§æ‰¿ç¬¬2èŠ‚è®­ç»ƒçš„Foundation modelï¼Œå¯¹äºå…¶å®ƒä¸¤ç»„éœ€è¦é‡å¤´è®­ç»ƒã€‚</li>
<li>åœ¨Pre-trainä¹‹åï¼Œå¦‚2.3å°èŠ‚éœ€è¦æ‹¼æ¥è¡¨ç¤ºå¤šæ¨¡æ€çš„tokenï¼ˆå¦‚æœæ¶‰åŠå¤šæ‰¹æ¬¡ï¼Œä¹Ÿè¦æ·»åŠ batch tokenï¼‰</li>
<li>Fine-tune ç›®æ ‡å‡½æ•°åŒ…æ‹¬GEPï¼ŒGEPCï¼ˆå¦‚æœæ¶‰åŠå¤šæ‰¹æ¬¡ï¼Œè¿˜éœ€è¦æ·»åŠ DARï¼‰</li>
</ul>
<p><strong>Result</strong></p>
<ul>
<li>
<p>å¯¹äºMultiome PBMC (paired)æ•°æ®é›†ï¼Œä¸scGLUEï¼ŒSeuratè¿›è¡Œäº†æ¯”è¾ƒï¼Œå¦‚ä¸‹å›¾æ‰€ç¤º</p>
<ul>
<li>RNA-seqã€ATAC</li>
</ul>
</li>
<li>
<p>å¯¹äºBone marrow mononuclear cells (paired)æ•°æ®é›†ï¼Œä¸Seuratè¿›è¡Œäº†æ¯”è¾ƒ</p>
<ul>
<li>RNA-seqã€Protein</li>
<li>9wä¸ªç»†èƒï¼Œ12ä¸ªdonor(multiple batches)ï¼Œ48ç§ç»†èƒç±»å‹</li>
</ul>
</li>
<li>
<p>å¯¹äºASAP human PBMC (mosaic)æ•°æ®é›†ï¼Œä¸scMoMatè¿›è¡Œäº†æ¯”è¾ƒ</p>
<ul>
<li>4ä¸ªæ‰¹æ¬¡ï¼Œ3ç§ç»„å­¦</li>
</ul>
</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240926132021275.png" alt="image-20240926132021275"  />
</p>
<h2 id="34-åŸºå› è°ƒæ§ç½‘ç»œæ„å»º">3.4 åŸºå› è°ƒæ§ç½‘ç»œæ„å»º<a hidden class="anchor" aria-hidden="true" href="#34-åŸºå› è°ƒæ§ç½‘ç»œæ„å»º">#</a></h2>
<p><strong>ï¼ˆ1ï¼‰Method</strong></p>
<ul>
<li>
<p>æœ¬è´¨ä¸Šæ˜¯åŸºäºFoundation modelæˆ–è€…Fine-tune modelæå–çš„Gene Embeddingè®¡ç®—ä¸¤ä¸¤åŸºå› é—´çš„ç›¸å…³æ€§ï¼Œç”¨æ¥æ„å»ºgene similarity networkã€‚</p>
<ul>
<li>
<p>Zero-shot settingæ¨¡å¼ä¸‹ï¼Œç›´æ¥ä½¿ç”¨Foundation modelè¾“å‡ºçš„Gene Embeddingè®¡ç®—ï¼›</p>
</li>
<li>
<p>Fine-tuned settingæ¨¡å¼ä¸‹ï¼Œä½¿ç”¨ç‰¹å®šæ•°æ®é›†å¾®è°ƒåè¾“å‡ºçš„Gene Embeddingè®¡ç®—</p>
<blockquote>
<p>ä½¿ç”¨integration taskè¿›è¡Œå¾®è°ƒï¼Œä»¥å­¦ä¹ ç‰¹å®šæ•°æ®é›†ç›¸å…³çš„Gene Embedding ï¼ˆä¸‹åŒï¼‰</p></blockquote>
</li>
</ul>
</li>
<li>
<p>åŸºäºæ³¨æ„åŠ›æœºåˆ¶çš„target geneé‰´å®šï¼šAttention mapå¯ä»¥åæ˜ å‡ºåŸºå› ä¹‹é—´çš„ç›¸äº’å½±å“ã€‚å…¶ä¸­ï¼Œæ¯ä¸€åˆ—(column)ï¼Œè¡¨ç¤ºè¿™ä¸ªgeneï¼ˆåˆ—åï¼‰å¯¹æ‰€æœ‰query geneçš„influenceã€‚</p>
</li>
<li>
<p>ä½¿ç”¨perturbation datasetså¯ä»¥å¸®åŠ©æ¨æµ‹perturbating geneçš„targetï¼Œå¦‚ä¸‹å›¾æ‰€ç¤º</p>
<ul>
<li>é¦–å…ˆï¼ŒåŸºäºcontrol cellå¾—åˆ°çš„attention mapï¼›</li>
<li>ç„¶åï¼ŒåŸºäºperturbation expressionå¾—åˆ°çš„attention mapï¼›</li>
<li>è®¡ç®—ä¸Šè¿°äºŒè€…å¯¹çš„å·®å€¼ï¼Œå¯ä»¥æ¨æµ‹ç‰¹å®šåŸºå› å¹²æ‰°å‰åï¼Œå½±å“æœ€å¤§çš„gene</li>
</ul>
</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240926134630079.png" alt="image-20240926134630079"  />
</p>
<p><strong>ï¼ˆ2ï¼‰Result</strong></p>
<ul>
<li>æ–‡ç« é¦–å…ˆå¯¹zero-shotæ¨¡å¼è¿›è¡Œäº†æ¢ç´¢ï¼Œç„¶ååœ¨å¯¹ä¸€ä¸ªhuman immune datasetè¿›è¡Œäº†Fine-tuneåæ„å»ºGRNï¼Œå‡å‘ç°äº†å…·æœ‰ç”Ÿç‰©å­¦æ„ä¹‰çš„è°ƒæ§ç½‘ç»œä»¥åŠGene cluster (Leiden)ï¼Œå¹¶è¿›è¡Œäº†é€šè·¯å¯Œé›†åˆ†æç­‰ã€‚</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240929081715151.png" alt="image-20240929081715151"  />
</p>
<ul>
<li>æ¥ä¸‹æ¥ï¼Œä½œè€…åŸºäºPre-train scGPT blood modelï¼Œä½¿ç”¨äº†Adamson CRISPRæ•°æ®é›†(87 CRISPR inference on leukemia cells)è¿›è¡Œäº†å¾®è°ƒã€‚</li>
<li>ä¾‹å¦‚ä¸‹å›¾ï¼Œé€šè¿‡æ¯”è¾ƒDDIT3æ‰°åŠ¨å‰åçš„difference attention score, é‰´å®šäº†å…¶å½±å“æœ€å¤§çš„çš„Top20/100ä¸ªåŸºå› ã€‚</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240929082629291.png" alt="image-20240929082629291"  />
</p>
<h1 id="4-æ¨¡å‹å½±å“å› ç´ ">4. æ¨¡å‹å½±å“å› ç´ <a hidden class="anchor" aria-hidden="true" href="#4-æ¨¡å‹å½±å“å› ç´ ">#</a></h1>
<ul>
<li>ç»¼ä¸Šï¼ŒscGPT Foundation modelåœ¨å¤„ç†ä¸‹æ¸¸åˆ†æä»»åŠ¡æ—¶ï¼Œç›¸æ¯”äºå…¶å®ƒå·²æœ‰çš„å•ç»†èƒå·¥å…·å‡è¡¨ç°å‡ºæ˜æ˜¾çš„ä¼˜åŠ¿ï¼›</li>
<li>æœ€åï¼Œæ–‡ç« è®¨è®ºäº†ä¸¤ä¸ªå¯èƒ½å½±å“Pre-train modelæ€§èƒ½è¡¨ç°çš„å› ç´ ã€‚</li>
</ul>
<h2 id="41-è®­ç»ƒæ ·æœ¬é‡">4.1 è®­ç»ƒæ ·æœ¬é‡<a hidden class="anchor" aria-hidden="true" href="#41-è®­ç»ƒæ ·æœ¬é‡">#</a></h2>
<ul>
<li>é¦–å…ˆï¼Œæ–‡ç« ç ”ç©¶äº†è®­ç»ƒæ ·æœ¬é‡å¯¹äºé¢„è®­ç»ƒæ¨¡å‹çš„å½±å“ï¼›</li>
<li>å¦‚ä¸‹å›¾ï¼Œç»“æœå‘ç°æ•°æ®é‡æœ€å¤šæ—¶ï¼ˆä»30Kåˆ°33Mï¼‰ï¼Œæ¨¡å‹åœ¨å¤šç§ä¸‹æ¸¸ä»»åŠ¡ä¸­çš„è¡¨ç°è¶Šå¥½ï¼›</li>
<li>As larger and more diverse datasets become available, we can anticipate further improvements in model performance, advancing our understanding of cellular processes.</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240929085432250.png" alt="image-20240929085432250"  />
</p>
<h2 id="42-è®­ç»ƒæ ·æœ¬ç±»å‹">4.2 è®­ç»ƒæ ·æœ¬ç±»å‹<a hidden class="anchor" aria-hidden="true" href="#42-è®­ç»ƒæ ·æœ¬ç±»å‹">#</a></h2>
<ul>
<li>æ–‡ç« è¿›ä¸€æ­¥ç ”ç©¶äº†ç»†èƒæ¥æºå¯¹äºæ¨¡å‹çš„å½±å“ï¼›</li>
<li>åœ¨COVID-19æ•°æ®é›†çš„å¤šæ‰¹æ¬¡åˆå¹¶ä»»åŠ¡ä¸­ï¼Œç»“æœå‘ç°ï¼š
<ul>
<li>ä½¿ç”¨å…¨éƒ¨33Mæ•°æ®ä¸ä»…ä½¿ç”¨Bloodæ•°æ®çš„Foundation modelçš„æ€§èƒ½æ¯”è¾ƒæ¥è¿‘ï¼›</li>
<li>Lungæ¥æºçš„ç»†èƒé‡å°½ç®¡åªæœ‰2.1Mï¼Œè¿œå°‘äºBrainç»†èƒé‡(13.2M)ï¼Œä½†å‰è€…æ€§èƒ½æ˜æ˜¾è¡¨ç°æ›´ä¼˜ã€‚</li>
</ul>
</li>
<li>This emphasizes the importance of aligning the cellular context in pretraining with the target dataset for superior results in downstream tasks.</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20240929085123045.png" alt="image-20240929085123045"  />
</p>
<blockquote>
<p>åç»­å°†å‚è€ƒscGPTå·¥å…·çš„æ•™ç¨‹æ‰‹å†ŒåŠæºä»£ç ï¼Œå­¦ä¹ å…¶æ„å»ºç»†èŠ‚ä¸æ•°æ®å¤„ç†æ–¹å¼ç­‰ã€‚</p></blockquote>


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E6%96%87%E7%8C%AE-%E7%AE%97%E6%B3%95/">æ–‡çŒ®-ç®—æ³•</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/202gene2vec%E7%AE%97%E6%B3%95%E6%A0%B9%E6%8D%AE%E5%9F%BA%E5%9B%A0%E5%AF%B9%E8%AE%A1%E7%AE%97%E5%9F%BA%E5%9B%A0%E8%A1%A8%E7%A4%BA/">
    <span class="title">Â« Prev Page</span>
    <br>
    <span>Gene2vecç®—æ³•æ ¹æ®åŸºå› å¯¹è®¡ç®—åŸºå› è¡¨ç¤º</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/basic/203%E6%96%87%E7%8C%AE--%E5%8D%95%E7%BB%86%E8%83%9E%E7%BB%84%E5%AD%A6%E5%A4%A7%E6%A8%A1%E5%9E%8B%E4%B9%8Bscbert/">
    <span class="title">Next Page Â»</span>
    <br>
    <span>æ–‡çŒ®--å•ç»†èƒç»„å­¦å¤§æ¨¡å‹ä¹‹scBERT</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>æ‚¨æ˜¯æœ¬ç«™ç¬¬ <span id="busuanzi_value_site_uv"></span> ä½è®¿é—®è€…ï¼Œæ€»æµè§ˆé‡ä¸º <span id="busuanzi_value_site_pv"></span> æ¬¡
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
