<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--scVI去除批次效应 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="scVI (single-cell variational inference)是2018年发表的一项单细胞分析工具。它主要基于VAE变分自编码器的思想，计算去除批次效应后的细胞低维空间表示。而后于2022年被整合到单细胞组学分析工具包scvi-tool中。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/043%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scvi%E5%8E%BB%E9%99%A4%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/043%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scvi%E5%8E%BB%E9%99%A4%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--scVI去除批次效应" />
<meta property="og:description" content="scVI (single-cell variational inference)是2018年发表的一项单细胞分析工具。它主要基于VAE变分自编码器的思想，计算去除批次效应后的细胞低维空间表示。而后于2022年被整合到单细胞组学分析工具包scvi-tool中。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/043%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scvi%E5%8E%BB%E9%99%A4%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2024-11-05T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2024-11-05T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--scVI去除批次效应"/>
<meta name="twitter:description" content="scVI (single-cell variational inference)是2018年发表的一项单细胞分析工具。它主要基于VAE变分自编码器的思想，计算去除批次效应后的细胞低维空间表示。而后于2022年被整合到单细胞组学分析工具包scvi-tool中。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--scVI去除批次效应",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/043%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scvi%E5%8E%BB%E9%99%A4%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--scVI去除批次效应",
  "name": "单细胞分析工具--scVI去除批次效应",
  "description": "scVI (single-cell variational inference)是2018年发表的一项单细胞分析工具。它主要基于VAE变分自编码器的思想，计算去除批次效应后的细胞低维空间表示。而后于2022年被整合到单细胞组学分析工具包scvi-tool中。\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": "scVI (single-cell variational inference)是2018年发表的一项单细胞分析工具。它主要基于VAE变分自编码器的思想，计算去除批次效应后的细胞低维空间表示。而后于2022年被整合到单细胞组学分析工具包scvi-tool中。\nDeep generative modeling for single-cell transcriptomics, https://www.nature.com/articles/s41592-018-0229-2 A Python library for probabilistic analysis of single-cell omics data, https://www.nature.com/articles/s41587-021-01206-w Github： https://github.com/scverse/scvi-tools 上图为18年发表文章的Figure，简单理解如下\n观测/输入数据\nXn,1: Cell N的Gene 1的观测表达值\nSn：Cell N的批次标签\n潜变量：\nZn：Cell N的低维表示空间，即我们最关心的结果。\nln：Cell N的library size，可以理解为UMI counts\n变分后验(Variational posterior)表示对观测数据 (X,S) 所假设的潜在变量 (Z, l) 的真实后验分布的近似（编码器）。在VAE中，通常将其建模为高斯分布，其参数（均值和方差）由神经网络输出。然后生成模型部分（解码器）负责从潜在空间中生成新的数据样本X，应与真实数据相似。\n上述过程的可优化损失主要来自两方面：（1）重构损失，即解码器的输出与输入X的差异；（2）通过KL散度计算变分后验分布p(z|x)与先验分布p(z)的差异，主要确保潜在空间的分布结构合理。\n详细解释参考：https://docs.scvi-tools.org/en/stable/user_guide/models/scvi.html\n1. 安装 https://docs.scvi-tools.org/en/stable/tutorials/notebooks/quick_start/api_overview.html https://docs.scvi-tools.org/en/stable/tutorials/notebooks/scrna/harmonization.html#integration-with-scanvi 1 2 3 4 5 6 7 8 9 conda create -n scvi python=3.10 -y conda activate scvi # conda install scvi-tools -c conda-forge pip install -U scvi-tools -i https://pypi.tuna.tsinghua.edu.cn/simple pip install -U scanpy -i https://pypi.tuna.tsinghua.edu.cn/simple # 在具体分析过程中，还遇到部分缺少依赖包的报错，根据提示安装即可 # 例如：scib-metrics leidenalg pymde scib_metrics 2. 初始设置 1 2 3 4 5 6 7 8 9 10 scvi.settings.seed = 0 print(\"Last run with scvi-tools version:\", scvi.__version__) # 1.2.0 sc.set_figure_params(figsize=(6, 6), frameon=False) sns.set_theme() torch.set_float32_matmul_precision(\"high\") save_dir = \"./data/\" %config InlineBackend.print_figure_kwargs={\"facecolor\": \"w\"} %config InlineBackend.figure_format=\"retina\" 3. 读取数据及预处理 1 2 3 4 5 6 7 8 9 10 # adata = scvi.data.heart_cell_atlas_subsampled(save_path=save_dir.name) adata = sc.read(os.path.join(save_dir, \"hca_subsampled_20k.h5ad\")) remove = [\"doublets\", \"NotAssigned\"] keep = [c not in remove for c in adata.obs.cell_type.values] adata = adata[keep, :].copy() adata # AnnData object with n_obs × n_vars = 18641 × 26662 # obs: 'NRP', 'age_group', 'cell_source', 'cell_type', 'donor', 'gender', 'n_counts', 'n_genes', 'percent_mito', 'percent_ribo', 'region', 'sample', 'scrublet_score', 'source', 'type', 'version', 'cell_states', 'Used' # var: 'gene_ids-Harvard-Nuclei', 'feature_types-Harvard-Nuclei', 'gene_ids-Sanger-Nuclei', 'feature_types-Sanger-Nuclei', 'gene_ids-Sanger-Cells', 'feature_types-Sanger-Cells', 'gene_ids-Sanger-CD45', 'feature_types-Sanger-CD45', 'n_counts' # uns: 'cell_type_colors' 质控/标准化/相关备份 1 2 3 4 5 6 7 sc.pp.filter_genes(adata, min_counts=3) adata.layers[\"counts\"] = adata.X.copy() # 保存备份整个对象的状态.X 矩阵的副本 sc.pp.normalize_total(adata, target_sum=1e4) sc.pp.log1p(adata) adata.raw = adata #保存备份整个对象的状态 鉴定高变基因 1 2 3 4 5 6 7 8 sc.pp.highly_variable_genes( adata, n_top_genes=1200, subset=True, layer=\"counts\", flavor=\"seurat_v3\", batch_key=\"cell_source\", ) 3. 计算潜变量 实例化模型 1 2 3 4 5 6 7 8 9 10 11 12 13 14 scvi.model.SCVI.setup_anndata( adata, layer=\"counts\", categorical_covariate_keys=[\"cell_source\", \"donor\"], continuous_covariate_keys=[\"percent_mito\", \"percent_ribo\"], ) model = scvi.model.SCVI(adata) #可在此调整n_latent等参数 model # SCVI model with the following parameters: # n_hidden: 128, n_latent: 10, n_layers: 1, dropout_rate: 0.1, dispersion: gene, gene_likelihood: zinb, # latent_distribution: normal. # Training status: Not Trained # Model's adata is minified?: False VAE属于自监督学习算法，scVI基于Pytorch实现的神经网络架构，支持GPU加速 1 2 3 4 5 6 model.train() # model_dir = os.path.join(save_dir, \"scvi_model\") # model.save(model_dir, overwrite=True) # model = scvi.model.SCVI.load(model_dir, adata=adata) 计算潜变量结果 1 2 3 4 5 6 SCVI_LATENT_KEY = \"X_scVI\" latent = model.get_latent_representation() adata.obsm[SCVI_LATENT_KEY] = latent latent.shape # (18641, 10) 计算 normalized (decoded) gene expression. 1 2 3 4 adata_subset = adata[adata.obs.cell_type == \"Fibroblast\"] latent_subset = model.get_latent_representation(adata_subset) # (2446, 10) denoised = model.get_normalized_expression(adata_subset, library_size=1e4, ) # reconstructed output denoised.iloc[:5, :5] 4. 可视化 1 2 3 4 5 6 7 8 9 10 11 12 13 SCVI_NORMALIZED_KEY = \"scvi_normalized\" adata.layers[SCVI_NORMALIZED_KEY] = model.get_normalized_expression(library_size=10e4) sc.pp.neighbors(adata, use_rep=SCVI_LATENT_KEY) sc.tl.umap(adata, min_dist=0.3) sc.pl.umap( adata, color=[\"donor\", \"cell_source\"], ncols=2, frameon=False, ) 1 2 3 # neighbors were already computed using scVI SCVI_CLUSTERS_KEY = \"leiden_scVI\" sc.tl.leiden(adata, key_added=SCVI_CLUSTERS_KEY, resolution=0.5) 5. 鉴定差异基因 1 2 3 4 5 6 7 8 9 10 11 # one vs one de_df = model.differential_expression( groupby=\"cell_type\", group1=\"Endothelial\", group2=\"Fibroblast\" ) de_df.head() # one vs rest de_df = model.differential_expression( groupby=\"cell_type\", ) de_df.head() 6. SCANVI去批次 Now, we assume that all of our data is annotated. This can lead to a more accurate integration result when using scANVI 如果已经有注释好的细胞类型，使用scANVI去批次会更有效。 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 adata.obs.cell_type.unique().tolist() # ['Myeloid', # 'Ventricular_Cardiomyocyte', # 'Fibroblast', # 'Endothelial', # 'Adipocytes', # 'Pericytes', # 'Atrial_Cardiomyocyte', # 'Smooth_muscle_cells', # 'Neuronal', # 'Lymphoid', # 'Mesothelial'] scanvi_model = scvi.model.SCANVI.from_scvi_model( model, adata=adata, labels_key=\"cell_type\", unlabeled_category=\"Unknown\", ) scanvi_model.train() SCANVI_LATENT_KEY = \"X_scANVI\" adata.obsm[SCANVI_LATENT_KEY] = scanvi_model.get_latent_representation(adata) adata.obsm[SCANVI_LATENT_KEY].shape # (18641, 10) 1 2 # 传统PCA sc.tl.pca(adata) 使用scib_metrics评估三种降维的效果 https://scib-metrics.readthedocs.io/en/stable/notebooks/lung_example.html 1 2 3 4 5 6 7 8 9 10 11 12 from scib_metrics.benchmark import Benchmarker bm = Benchmarker( adata, batch_key=\"cell_source\", label_key=\"cell_type\", embedding_obsm_keys=[\"X_pca\", SCVI_LATENT_KEY, SCANVI_LATENT_KEY], n_jobs=-1, ) bm.benchmark() bm.plot_results_table(min_max_scale=False) ",
  "wordCount" : "1342",
  "inLanguage": "en",
  "datePublished": "2024-11-05T00:00:00Z",
  "dateModified": "2024-11-05T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/043%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scvi%E5%8E%BB%E9%99%A4%E6%89%B9%E6%AC%A1%E6%95%88%E5%BA%94/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--scVI去除批次效应
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2024-11-05 00:00:00 &#43;0000 UTC&#39;&gt;2024-11-05&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2024-11-05&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1342&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1-%e5%ae%89%e8%a3%85" aria-label="1. 安装">1. 安装</a></li>
                    <li>
                        <a href="#2-%e5%88%9d%e5%a7%8b%e8%ae%be%e7%bd%ae" aria-label="2. 初始设置">2. 初始设置</a></li>
                    <li>
                        <a href="#3-%e8%af%bb%e5%8f%96%e6%95%b0%e6%8d%ae%e5%8f%8a%e9%a2%84%e5%a4%84%e7%90%86" aria-label="3. 读取数据及预处理">3. 读取数据及预处理</a></li>
                    <li>
                        <a href="#3-%e8%ae%a1%e7%ae%97%e6%bd%9c%e5%8f%98%e9%87%8f" aria-label="3. 计算潜变量">3. 计算潜变量</a></li>
                    <li>
                        <a href="#4-%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="4. 可视化">4. 可视化</a></li>
                    <li>
                        <a href="#5-%e9%89%b4%e5%ae%9a%e5%b7%ae%e5%bc%82%e5%9f%ba%e5%9b%a0" aria-label="5. 鉴定差异基因">5. 鉴定差异基因</a></li>
                    <li>
                        <a href="#6-scanvi%e5%8e%bb%e6%89%b9%e6%ac%a1" aria-label="6. SCANVI去批次">6. SCANVI去批次</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><p>scVI (single-cell variational inference)是2018年发表的一项单细胞分析工具。它主要基于VAE变分自编码器的思想，计算去除批次效应后的细胞低维空间表示。而后于2022年被整合到单细胞组学分析工具包scvi-tool中。</p>
<blockquote>
<ul>
<li>Deep generative modeling for single-cell transcriptomics, <a href="https://www.nature.com/articles/s41592-018-0229-2">https://www.nature.com/articles/s41592-018-0229-2</a></li>
<li>A Python library for probabilistic analysis of single-cell omics data, <a href="https://www.nature.com/articles/s41587-021-01206-w">https://www.nature.com/articles/s41587-021-01206-w</a></li>
<li>Github： <a href="https://github.com/scverse/scvi-tools">https://github.com/scverse/scvi-tools</a></li>
</ul></blockquote>
<p><img loading="lazy" src="C:%5cUsers%5cxiaoxin%5cAppData%5cRoaming%5cTypora%5ctypora-user-images%5cimage-20241103145224602.png" alt="image-20241103145224602"  />
</p>
<p>上图为18年发表文章的Figure，简单理解如下</p>
<p>观测/输入数据</p>
<ul>
<li>
<p><strong>Xn,1</strong>: Cell N的Gene 1的观测表达值</p>
</li>
<li>
<p><strong>Sn</strong>：Cell N的批次标签</p>
</li>
</ul>
<p>潜变量：</p>
<ul>
<li>
<p><strong>Zn</strong>：Cell N的低维表示空间，即我们最关心的结果。</p>
</li>
<li>
<p><strong>ln</strong>：Cell N的library size，可以理解为UMI counts</p>
</li>
</ul>
<p>变分后验(Variational posterior)表示对观测数据 (X,S) 所假设的潜在变量 (Z, l) 的真实后验分布的近似（编码器）。在VAE中，通常将其建模为高斯分布，其参数（均值和方差）由神经网络输出。然后生成模型部分（解码器）负责从潜在空间中生成新的数据样本<strong>X</strong>，应与真实数据相似。</p>
<p>上述过程的可优化损失主要来自两方面：（1）重构损失，即解码器的输出与输入X的差异；（2）通过KL散度计算变分后验分布p(z|x)与先验分布p(z)的差异，主要确保潜在空间的分布结构合理。</p>
<p>详细解释参考：https://docs.scvi-tools.org/en/stable/user_guide/models/scvi.html</p>
<h2 id="1-安装">1. 安装<a hidden class="anchor" aria-hidden="true" href="#1-安装">#</a></h2>
<ul>
<li><a href="https://docs.scvi-tools.org/en/stable/tutorials/notebooks/quick_start/api_overview.html">https://docs.scvi-tools.org/en/stable/tutorials/notebooks/quick_start/api_overview.html</a></li>
<li><a href="https://docs.scvi-tools.org/en/stable/tutorials/notebooks/scrna/harmonization.html#integration-with-scanvi">https://docs.scvi-tools.org/en/stable/tutorials/notebooks/scrna/harmonization.html#integration-with-scanvi</a></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>conda create -n scvi python=3.10 -y
</span></span><span style="display:flex;"><span>conda activate scvi
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># conda install scvi-tools -c conda-forge</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>pip install -U scvi-tools -i https://pypi.tuna.tsinghua.edu.cn/simple
</span></span><span style="display:flex;"><span>pip install -U scanpy -i https://pypi.tuna.tsinghua.edu.cn/simple
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 在具体分析过程中，还遇到部分缺少依赖包的报错，根据提示安装即可</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 例如：scib-metrics leidenalg pymde scib_metrics</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="2-初始设置">2. 初始设置<a hidden class="anchor" aria-hidden="true" href="#2-初始设置">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>scvi.settings.seed = <span style="color:#ff0;font-weight:bold">0</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">print</span>(<span style="color:#0ff;font-weight:bold">&#34;Last run with scvi-tools version:&#34;</span>, scvi.__version__) <span style="color:#007f7f"># 1.2.0</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sc.set_figure_params(figsize=(<span style="color:#ff0;font-weight:bold">6</span>, <span style="color:#ff0;font-weight:bold">6</span>), frameon=<span style="color:#fff;font-weight:bold">False</span>)
</span></span><span style="display:flex;"><span>sns.set_theme()
</span></span><span style="display:flex;"><span>torch.set_float32_matmul_precision(<span style="color:#0ff;font-weight:bold">&#34;high&#34;</span>)
</span></span><span style="display:flex;"><span>save_dir = <span style="color:#0ff;font-weight:bold">&#34;./data/&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>%config InlineBackend.print_figure_kwargs={<span style="color:#0ff;font-weight:bold">&#34;facecolor&#34;</span>: <span style="color:#0ff;font-weight:bold">&#34;w&#34;</span>}
</span></span><span style="display:flex;"><span>%config InlineBackend.figure_format=<span style="color:#0ff;font-weight:bold">&#34;retina&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="3-读取数据及预处理">3. 读取数据及预处理<a hidden class="anchor" aria-hidden="true" href="#3-读取数据及预处理">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f"># adata = scvi.data.heart_cell_atlas_subsampled(save_path=save_dir.name)</span>
</span></span><span style="display:flex;"><span>adata = sc.read(os.path.join(save_dir, <span style="color:#0ff;font-weight:bold">&#34;hca_subsampled_20k.h5ad&#34;</span>))
</span></span><span style="display:flex;"><span>remove = [<span style="color:#0ff;font-weight:bold">&#34;doublets&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;NotAssigned&#34;</span>]
</span></span><span style="display:flex;"><span>keep = [c not in remove <span style="color:#fff;font-weight:bold">for</span> c in adata.obs.cell_type.values]
</span></span><span style="display:flex;"><span>adata = adata[keep, :].copy()
</span></span><span style="display:flex;"><span>adata
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AnnData object with n_obs × n_vars = 18641 × 26662</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obs: &#39;NRP&#39;, &#39;age_group&#39;, &#39;cell_source&#39;, &#39;cell_type&#39;, &#39;donor&#39;, &#39;gender&#39;, &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;percent_mito&#39;, &#39;percent_ribo&#39;, &#39;region&#39;, &#39;sample&#39;, &#39;scrublet_score&#39;, &#39;source&#39;, &#39;type&#39;, &#39;version&#39;, &#39;cell_states&#39;, &#39;Used&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     var: &#39;gene_ids-Harvard-Nuclei&#39;, &#39;feature_types-Harvard-Nuclei&#39;, &#39;gene_ids-Sanger-Nuclei&#39;, &#39;feature_types-Sanger-Nuclei&#39;, &#39;gene_ids-Sanger-Cells&#39;, &#39;feature_types-Sanger-Cells&#39;, &#39;gene_ids-Sanger-CD45&#39;, &#39;feature_types-Sanger-CD45&#39;, &#39;n_counts&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     uns: &#39;cell_type_colors&#39;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>质控/标准化/相关备份</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>sc.pp.filter_genes(adata, min_counts=<span style="color:#ff0;font-weight:bold">3</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>adata.layers[<span style="color:#0ff;font-weight:bold">&#34;counts&#34;</span>] = adata.X.copy()  <span style="color:#007f7f"># 保存备份整个对象的状态.X 矩阵的副本</span>
</span></span><span style="display:flex;"><span>sc.pp.normalize_total(adata, target_sum=<span style="color:#ff0;font-weight:bold">1e4</span>)
</span></span><span style="display:flex;"><span>sc.pp.log1p(adata)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>adata.raw = adata <span style="color:#007f7f">#保存备份整个对象的状态</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>鉴定高变基因</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>sc.pp.highly_variable_genes(
</span></span><span style="display:flex;"><span>    adata,
</span></span><span style="display:flex;"><span>    n_top_genes=<span style="color:#ff0;font-weight:bold">1200</span>,
</span></span><span style="display:flex;"><span>    subset=<span style="color:#fff;font-weight:bold">True</span>,
</span></span><span style="display:flex;"><span>    layer=<span style="color:#0ff;font-weight:bold">&#34;counts&#34;</span>,
</span></span><span style="display:flex;"><span>    flavor=<span style="color:#0ff;font-weight:bold">&#34;seurat_v3&#34;</span>,
</span></span><span style="display:flex;"><span>    batch_key=<span style="color:#0ff;font-weight:bold">&#34;cell_source&#34;</span>,
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="3-计算潜变量">3. 计算潜变量<a hidden class="anchor" aria-hidden="true" href="#3-计算潜变量">#</a></h2>
<ul>
<li>实例化模型</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>scvi.model.SCVI.setup_anndata(
</span></span><span style="display:flex;"><span>    adata,
</span></span><span style="display:flex;"><span>    layer=<span style="color:#0ff;font-weight:bold">&#34;counts&#34;</span>,
</span></span><span style="display:flex;"><span>    categorical_covariate_keys=[<span style="color:#0ff;font-weight:bold">&#34;cell_source&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;donor&#34;</span>],
</span></span><span style="display:flex;"><span>    continuous_covariate_keys=[<span style="color:#0ff;font-weight:bold">&#34;percent_mito&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;percent_ribo&#34;</span>],
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>model = scvi.model.SCVI(adata) <span style="color:#007f7f">#可在此调整n_latent等参数</span>
</span></span><span style="display:flex;"><span>model
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SCVI model with the following parameters: </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># n_hidden: 128, n_latent: 10, n_layers: 1, dropout_rate: 0.1, dispersion: gene, gene_likelihood: zinb, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># latent_distribution: normal.</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Training status: Not Trained</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Model&#39;s adata is minified?: False</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>VAE属于自监督学习算法，scVI基于Pytorch实现的神经网络架构，支持GPU加速</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>model.train()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># model_dir = os.path.join(save_dir, &#34;scvi_model&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># model.save(model_dir, overwrite=True)</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># model = scvi.model.SCVI.load(model_dir, adata=adata)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>计算潜变量结果</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>SCVI_LATENT_KEY = <span style="color:#0ff;font-weight:bold">&#34;X_scVI&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>latent = model.get_latent_representation()
</span></span><span style="display:flex;"><span>adata.obsm[SCVI_LATENT_KEY] = latent
</span></span><span style="display:flex;"><span>latent.shape
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (18641, 10)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>计算 normalized (decoded) gene expression.</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>adata_subset = adata[adata.obs.cell_type == <span style="color:#0ff;font-weight:bold">&#34;Fibroblast&#34;</span>]
</span></span><span style="display:flex;"><span>latent_subset = model.get_latent_representation(adata_subset) <span style="color:#007f7f"># (2446, 10)</span>
</span></span><span style="display:flex;"><span>denoised = model.get_normalized_expression(adata_subset, library_size=<span style="color:#ff0;font-weight:bold">1e4</span>, ) <span style="color:#007f7f"># reconstructed output</span>
</span></span><span style="display:flex;"><span>denoised.iloc[:<span style="color:#ff0;font-weight:bold">5</span>, :<span style="color:#ff0;font-weight:bold">5</span>]
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="4-可视化">4. 可视化<a hidden class="anchor" aria-hidden="true" href="#4-可视化">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>SCVI_NORMALIZED_KEY = <span style="color:#0ff;font-weight:bold">&#34;scvi_normalized&#34;</span>
</span></span><span style="display:flex;"><span>adata.layers[SCVI_NORMALIZED_KEY] = model.get_normalized_expression(library_size=<span style="color:#ff0;font-weight:bold">10e4</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sc.pp.neighbors(adata, use_rep=SCVI_LATENT_KEY)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sc.tl.umap(adata, min_dist=<span style="color:#ff0;font-weight:bold">0.3</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sc.pl.umap(
</span></span><span style="display:flex;"><span>    adata,
</span></span><span style="display:flex;"><span>    color=[<span style="color:#0ff;font-weight:bold">&#34;donor&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;cell_source&#34;</span>],
</span></span><span style="display:flex;"><span>    ncols=<span style="color:#ff0;font-weight:bold">2</span>,
</span></span><span style="display:flex;"><span>    frameon=<span style="color:#fff;font-weight:bold">False</span>,
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="C:\Users\xiaoxin\AppData\Roaming\Typora\typora-user-images\image-20241103161347836.png" alt="image-20241103161347836" style="zoom: 67%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f"># neighbors were already computed using scVI</span>
</span></span><span style="display:flex;"><span>SCVI_CLUSTERS_KEY = <span style="color:#0ff;font-weight:bold">&#34;leiden_scVI&#34;</span>
</span></span><span style="display:flex;"><span>sc.tl.leiden(adata, key_added=SCVI_CLUSTERS_KEY, resolution=<span style="color:#ff0;font-weight:bold">0.5</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="C:\Users\xiaoxin\AppData\Roaming\Typora\typora-user-images\image-20241103161510785.png" alt="image-20241103161510785" style="zoom:50%;" />
<h2 id="5-鉴定差异基因">5. 鉴定差异基因<a hidden class="anchor" aria-hidden="true" href="#5-鉴定差异基因">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f"># one vs one</span>
</span></span><span style="display:flex;"><span>de_df = model.differential_expression(
</span></span><span style="display:flex;"><span>    groupby=<span style="color:#0ff;font-weight:bold">&#34;cell_type&#34;</span>, group1=<span style="color:#0ff;font-weight:bold">&#34;Endothelial&#34;</span>, group2=<span style="color:#0ff;font-weight:bold">&#34;Fibroblast&#34;</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>de_df.head()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># one vs rest</span>
</span></span><span style="display:flex;"><span>de_df = model.differential_expression(
</span></span><span style="display:flex;"><span>    groupby=<span style="color:#0ff;font-weight:bold">&#34;cell_type&#34;</span>,
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>de_df.head()
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="6-scanvi去批次">6. SCANVI去批次<a hidden class="anchor" aria-hidden="true" href="#6-scanvi去批次">#</a></h2>
<ul>
<li>Now, we assume that all of our data is <strong>annotated</strong>. This can lead to a <strong>more accurate integration</strong> result when using scANVI</li>
<li>如果已经有注释好的细胞类型，使用scANVI去批次会更有效。</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>adata.obs.cell_type.unique().tolist()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;Myeloid&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;Ventricular_Cardiomyocyte&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;Fibroblast&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;Endothelial&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;Adipocytes&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;Pericytes&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;Atrial_Cardiomyocyte&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;Smooth_muscle_cells&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;Neuronal&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;Lymphoid&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;Mesothelial&#39;]</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>scanvi_model = scvi.model.SCANVI.from_scvi_model(
</span></span><span style="display:flex;"><span>    model,
</span></span><span style="display:flex;"><span>    adata=adata,
</span></span><span style="display:flex;"><span>    labels_key=<span style="color:#0ff;font-weight:bold">&#34;cell_type&#34;</span>,
</span></span><span style="display:flex;"><span>    unlabeled_category=<span style="color:#0ff;font-weight:bold">&#34;Unknown&#34;</span>,
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>scanvi_model.train()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>SCANVI_LATENT_KEY = <span style="color:#0ff;font-weight:bold">&#34;X_scANVI&#34;</span>
</span></span><span style="display:flex;"><span>adata.obsm[SCANVI_LATENT_KEY] = scanvi_model.get_latent_representation(adata)
</span></span><span style="display:flex;"><span>adata.obsm[SCANVI_LATENT_KEY].shape
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (18641, 10)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-py" data-lang="py"><span style="display:flex;"><span><span style="color:#007f7f"># 传统PCA</span>
</span></span><span style="display:flex;"><span>sc.tl.pca(adata)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>使用scib_metrics评估三种降维的效果</li>
<li><a href="https://scib-metrics.readthedocs.io/en/stable/notebooks/lung_example.html">https://scib-metrics.readthedocs.io/en/stable/notebooks/lung_example.html</a></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">from</span> scib_metrics.benchmark <span style="color:#fff;font-weight:bold">import</span> Benchmarker
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>bm = Benchmarker(
</span></span><span style="display:flex;"><span>    adata,
</span></span><span style="display:flex;"><span>    batch_key=<span style="color:#0ff;font-weight:bold">&#34;cell_source&#34;</span>,
</span></span><span style="display:flex;"><span>    label_key=<span style="color:#0ff;font-weight:bold">&#34;cell_type&#34;</span>,
</span></span><span style="display:flex;"><span>    embedding_obsm_keys=[<span style="color:#0ff;font-weight:bold">&#34;X_pca&#34;</span>, SCVI_LATENT_KEY, SCANVI_LATENT_KEY],
</span></span><span style="display:flex;"><span>    n_jobs=-<span style="color:#ff0;font-weight:bold">1</span>,
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>bm.benchmark()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>bm.plot_results_table(min_max_scale=<span style="color:#fff;font-weight:bold">False</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="C:%5cUsers%5cxiaoxin%5cAppData%5cRoaming%5cTypora%5ctypora-user-images%5cimage-20241103163422118.png" alt="image-20241103163422118"  />
</p>


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/042%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cellxgene%E6%95%B0%E6%8D%AE%E5%BA%93/">
    <span class="title">« Prev Page</span>
    <br>
    <span>单细胞分析工具--CELLxGENE数据库</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/038%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--sccoda%E7%BB%86%E8%83%9E%E7%BB%84%E6%88%90%E6%AF%94%E8%BE%83/">
    <span class="title">Next Page »</span>
    <br>
    <span>单细胞分析工具--sccoda细胞组成比较</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
