<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>clusterProfileråŒ…å¯Œé›†åˆ†æä¸å¯è§†åŒ– | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="è½¬å½•ç»„">
<meta name="description" content="1ã€èƒŒæ™¯çŸ¥è¯†
ï¼ˆ1ï¼‰ä¸¤ç§å¯Œé›†åˆ†æ


åŸºäºè¶…å‡ ä½•æ£€éªŒçš„ORA(over representation analysis)å¯Œé›†åˆ†æ

â‘  å‡è®¾å¯¹è½¬å½•ç»„åˆ†ç»„æµ‹åºçš„10000ä¸ªåŸºå› è¡¨è¾¾æ•°æ®è¿›è¡Œå·®å¼‚åˆ†æï¼›
â‘¡ æŒ‰ç…§è§„å®šcutoffé˜ˆå€¼(logFC/Pvalue)ç­›é€‰å¾—åˆ°100ä¸ªå·®å¼‚è¡¨è¾¾åŸºå› ï¼›">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/056clusterprofiler%E5%8C%85%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/056clusterprofiler%E5%8C%85%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="clusterProfileråŒ…å¯Œé›†åˆ†æä¸å¯è§†åŒ–" />
<meta property="og:description" content="1ã€èƒŒæ™¯çŸ¥è¯†
ï¼ˆ1ï¼‰ä¸¤ç§å¯Œé›†åˆ†æ


åŸºäºè¶…å‡ ä½•æ£€éªŒçš„ORA(over representation analysis)å¯Œé›†åˆ†æ

â‘  å‡è®¾å¯¹è½¬å½•ç»„åˆ†ç»„æµ‹åºçš„10000ä¸ªåŸºå› è¡¨è¾¾æ•°æ®è¿›è¡Œå·®å¼‚åˆ†æï¼›
â‘¡ æŒ‰ç…§è§„å®šcutoffé˜ˆå€¼(logFC/Pvalue)ç­›é€‰å¾—åˆ°100ä¸ªå·®å¼‚è¡¨è¾¾åŸºå› ï¼›" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/056clusterprofiler%E5%8C%85%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-01-14T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-01-14T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="clusterProfileråŒ…å¯Œé›†åˆ†æä¸å¯è§†åŒ–"/>
<meta name="twitter:description" content="1ã€èƒŒæ™¯çŸ¥è¯†
ï¼ˆ1ï¼‰ä¸¤ç§å¯Œé›†åˆ†æ


åŸºäºè¶…å‡ ä½•æ£€éªŒçš„ORA(over representation analysis)å¯Œé›†åˆ†æ

â‘  å‡è®¾å¯¹è½¬å½•ç»„åˆ†ç»„æµ‹åºçš„10000ä¸ªåŸºå› è¡¨è¾¾æ•°æ®è¿›è¡Œå·®å¼‚åˆ†æï¼›
â‘¡ æŒ‰ç…§è§„å®šcutoffé˜ˆå€¼(logFC/Pvalue)ç­›é€‰å¾—åˆ°100ä¸ªå·®å¼‚è¡¨è¾¾åŸºå› ï¼›"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "åˆ†ç±»",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "clusterProfileråŒ…å¯Œé›†åˆ†æä¸å¯è§†åŒ–",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/056clusterprofiler%E5%8C%85%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "clusterProfileråŒ…å¯Œé›†åˆ†æä¸å¯è§†åŒ–",
  "name": "clusterProfileråŒ…å¯Œé›†åˆ†æä¸å¯è§†åŒ–",
  "description": "1ã€èƒŒæ™¯çŸ¥è¯† ï¼ˆ1ï¼‰ä¸¤ç§å¯Œé›†åˆ†æ åŸºäºè¶…å‡ ä½•æ£€éªŒçš„ORA(over representation analysis)å¯Œé›†åˆ†æ â‘  å‡è®¾å¯¹è½¬å½•ç»„åˆ†ç»„æµ‹åºçš„10000ä¸ªåŸºå› è¡¨è¾¾æ•°æ®è¿›è¡Œå·®å¼‚åˆ†æï¼›\nâ‘¡ æŒ‰ç…§è§„å®šcutoffé˜ˆå€¼(logFC/Pvalue)ç­›é€‰å¾—åˆ°100ä¸ªå·®å¼‚è¡¨è¾¾åŸºå› ï¼›\n",
  "keywords": [
    "è½¬å½•ç»„"
  ],
  "articleBody": "1ã€èƒŒæ™¯çŸ¥è¯† ï¼ˆ1ï¼‰ä¸¤ç§å¯Œé›†åˆ†æ åŸºäºè¶…å‡ ä½•æ£€éªŒçš„ORA(over representation analysis)å¯Œé›†åˆ†æ â‘  å‡è®¾å¯¹è½¬å½•ç»„åˆ†ç»„æµ‹åºçš„10000ä¸ªåŸºå› è¡¨è¾¾æ•°æ®è¿›è¡Œå·®å¼‚åˆ†æï¼›\nâ‘¡ æŒ‰ç…§è§„å®šcutoffé˜ˆå€¼(logFC/Pvalue)ç­›é€‰å¾—åˆ°100ä¸ªå·®å¼‚è¡¨è¾¾åŸºå› ï¼›\nâ‘¢ å¯¹äºç‰¹å®šé€šè·¯ï¼ŒåŒ…å«500ä¸ªåŸºå› ï¼›å…¶ä¸­æœ‰30ä¸ªå±äºå·®å¼‚åŸºå› \nâ‘£ å·®å¼‚åŸºå› æ˜¯å¦å¯Œé›†åˆ°è¯¥é€šè·¯ï¼Œå³è¡¨ç¤º30/70ä¸470/9430ç›¸æ¯”ï¼Œæ˜¯å¦å…·æœ‰æ˜¾è‘—æ„ä¹‰ã€‚\n1 2 3 4 # å¦‚ä¸Šå…¬å¼ï¼šN=10000ï¼ŒM=500ï¼Œn=100ï¼Œk=30 d \u003c- data.frame(non_DEG=c(470, 9430), DEG=c(30, 70)) row.names(d) \u003c- c(\"In set\", \"not in category\") fisher.test(d, alternative = \"less\") å¦‚ä¸Šè®¡ç®—ç»“æœï¼Œpå€¼è¶Šå°(\u003c0.05)ï¼Œè¯´æ˜å¾—åˆ°çš„å·®å¼‚åŸºå› æ˜¯æ˜¾è‘—å¯Œé›†åˆ°è¿™ä¸ªé€šè·¯åŸºå› é›†ä¸Šçš„ã€‚ç„¶åå°±å¯ä»¥è¿™ä¸ªåŸºå› é›†çš„ç”Ÿç‰©å­¦æ„ä¹‰è®²è¿°åç»­çš„æ•…äº‹äº†ã€‚\nåŸºäºç½®æ¢æ£€éªŒçš„GSEA(gene set enrichment analysis)å¯Œé›†åˆ†æ â‘  å‡è®¾å¯¹è½¬å½•ç»„åˆ†ç»„æµ‹åºçš„10000ä¸ªåŸºå› è¡¨è¾¾æ•°æ®è¿›è¡Œå·®å¼‚åˆ†æï¼Œä¸€èˆ¬æŒ‰ç…§å·®å¼‚å€æ•°ä»é«˜åˆ°ä½è¿›è¡Œæ’åºï¼›\nâ‘¡ å‡è®¾å¯¹äºåŒ…å«500ä¸ªåŸºå› çš„ç‰¹å®šé€šè·¯ï¼Œéå†ä¸Šè¿°10000ä¸ªåŸºå› çš„æ’åºåˆ—è¡¨ï¼›\nâ‘¢ åœ¨éå†è¿‡ç¨‹ä¸­ï¼Œå¦‚åŸºå› å±äºé€šè·¯é›†åˆ™åŠ åˆ†(hit)ï¼Œä¸å±äºåˆ™å‡åˆ†ï¼›åŠ å‡åˆ†çš„æƒé‡ä¸å·®å¼‚å€æ•°å…ˆå…³ã€‚\nâ‘£ ES(enrichment score)è¡¨ç¤ºéå†è¿‡ç¨‹ä¸­äº§ç”Ÿçš„ç»å¯¹å€¼æœ€å¤§å€¼ã€‚ESå¤§äº0ï¼Œè¡¨æ˜é€šè·¯é›†å¯Œé›†åˆ°æ’åºåˆ—è¡¨çš„é¡¶éƒ¨ï¼›åä¹‹åˆ™å¯Œé›†åˆ°æ’åºåˆ—è¡¨çš„åº•éƒ¨ã€‚\nâ‘¤ éšæœºé€‰æ‹©500ä¸ªåŸºå› è‹¥å¹²æ¬¡ï¼Œè®¡ç®—ESèƒŒæ™¯åˆ†å¸ƒï¼Œä»è€Œè®¡ç®—ç›¸åº”çš„På€¼ã€‚\nï¼ˆ2ï¼‰é€šè·¯åŸºå› é›† å¸¸è§çš„é€šè·¯åŸºå› é›†åŒ…æ‹¬GOã€KEGGã€Reactomeç­‰ã€‚MsigDBæ•°æ®åº“åˆ™ç»¼åˆæ”¶é›†äº†å¤šç§é€šè·¯åŸºå› é›†ã€‚4\nGO\nGene Ontology åŸºå› æœ¬ä½“è®º http://www.geneontology.org/ GO åˆ†åˆ«å®šä¹‰äº†ä¸‰ç§ç±»å‹çš„åŸºå› é›†ï¼šâ‘ ç»†èƒç»„æˆï¼ˆcellular componentï¼ŒCCï¼‰ï¼›â‘¡ç”Ÿç‰©è¿‡ç¨‹ï¼ˆbiological processï¼ŒBPï¼‰ï¼›â‘¢åˆ†å­åŠŸèƒ½ï¼ˆMolecular Functionï¼ŒMFï¼‰ã€‚è¯¦è§http://geneontology.org/docs/ontology-documentation/ ä¸¾ä¸ªä¾‹å­æ¥è¯´ï¼šåŸºå› A(å…¶è›‹ç™½äº§ç‰©æ˜¯cytochrome c(ç»†èƒè‰²ç´ c)ï¼‰ï¼Œè¯¥åŸºå› çš„å…·æœ‰æ°§åŒ–è¿˜åŸæ´»æ€§ï¼ˆmolecular functionï¼‰ï¼›å‚ä¸æ°§åŒ–ç£·é…¸åŒ–ç”Ÿç‰©å­¦è¿‡ç¨‹(molecular function)ï¼›è€Œå‘æŒ¥ä½œç”¨çš„ä½ç½®ä½äºç»†èƒçº¿ç²’ä½“åŸºè´¨(cellular component) KEGG\nKyoto Encyclopedia of Genes and Genomes äº¬éƒ½åŸºå› ä¸åŸºå› ç»„ç™¾ç§‘å…¨ä¹¦ https://www.genome.jp/kegg/ KEGGæ˜¯ç³»ç»Ÿåˆ†æåŸºå› åŠŸèƒ½ï¼Œè”ç³»åŸºå› ç»„ä¿¡æ¯å’ŒåŠŸèƒ½ä¿¡æ¯ï¼Œä»¥æ¢ç´¢åˆ†å­ç›¸äº’ä½œç”¨å’Œååº”ç½‘ç»œçš„çŸ¥è¯†åº“ï¼ˆäººå·¥ç»˜åˆ¶ï¼‰,å…¶ä¸­åŒ…å«æœ‰å¤§é‡çš„é€šè·¯ï¼ˆPATHWAYï¼‰å›¾ï¼Œæ¶‰åŠmetabolism, genetic and environmental information processing, cellular processes, organismal systems, human diseases, and drug developmentä¸ƒå¤§ç±»ã€‚ Reactome https://reactome.org/\nWikiPathway https://www.wikipathways.org/\nMsigdb\nMolecular Signatures Database åˆ†å­ç‰¹å¾æ•°æ®åº“ https://www.gsea-msigdb.org/gsea/msigdb/index.jsp æˆªæ­¢ç›®å‰è¯¥æ•°æ®åº“æ”¶é›†æœ‰32284ä¸ªåŸºå› é›†ï¼Œå¯åˆ†ä¸º9å¤§ç±» â€” H: hallmark gene setsï¼ŒC1: positional gene setsï¼ŒC2: curated gene setsï¼ŒC3: regulatory target gene setsï¼ŒC4: computational gene setsï¼ŒC5: ontology gene setsï¼ŒC6: oncogenic signature gene setsï¼ŒC7: immunologic signature gene setsï¼ŒC8: cell type signature gene sets æ­¤å¤–è¿˜æœ‰å…¶å®ƒç±»å‹é€šè·¯åŸºå› é›†ï¼Œä¾‹å¦‚ç–¾ç—…åŸºå› é›†DOï¼Œå°±ä¸ä¸€ä¸€åˆ—ä¸¾äº†ã€‚\nï¼ˆ3ï¼‰clusterProfileråŒ… RåŒ…clusterProfilerç”±å—æ–¹åŒ»ç§‘å¤§å­¦ä½™å…‰åˆ›å›¢é˜Ÿå¼€å‘ï¼Œæ˜¯ä¸€ä¸ªå¸¸ç”¨çš„å¯Œé›†åˆ†æå·¥å…·åŒ…ï¼Œç›®å‰å·²æ›´æ–°åˆ°4.xç³»åˆ—ç‰ˆæœ¬ã€‚è¯¥åŒ…åŒ…å«äº†å¯Œé›†åˆ†æç›¸å…³çš„æ–¹æ–¹é¢é¢å‡½æ•°åŠŸèƒ½ï¼Œä»ä¸Šæ¸¸åˆ†æåˆ°ä¸‹æ¸¸å¯è§†åŒ–ç­‰ä¸€åº”ä¿±å…¨ã€‚åç»­ç¬”è®°ä¸»è¦å‚è€ƒå®˜æ–¹æ•™ç¨‹å­¦ä¹ è¯¥åŒ…çš„åŸºç¡€ç”¨æ³•ã€‚\nDOI: 10.1016/j.xinn.2021.100141 Tutorialï¼šhttps://yulab-smu.top/biomedical-knowledge-mining-book/index.html 2ã€å¯Œé›†åˆ†æ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 ## åŠ è½½åŒ… library(tidyverse) library(clusterProfiler) library(org.Hs.eg.db) ## ç¤ºä¾‹åŸºå› æ•°æ® # (1) é™åºå·®å¼‚åŸºå› åˆ—è¡¨ç”¨äºGSEAåˆ†æ data(geneList, package=\"DOSE\") str(geneList) # Named num [1:12495] 4.57 4.51 4.42 4.14 3.88 ... # - attr(*, \"names\")= chr [1:12495] \"4312\" \"8318\" \"10874\" \"55143\" ... gene_list = geneList # (2) æ˜¾è‘—å·®å¼‚åŸºå› é›†ç”¨äºORAåˆ†æ gene_deg = names(geneList)[abs(geneList) \u003e 2] head(gene_deg) # [1] \"4312\" \"8318\" \"10874\" \"55143\" \"55388\" \"991\" éœ€è¦ä½¿ç”¨ENTREZIDçš„åŸºå› IDã€‚å¦‚æœæ˜¯å…¶å®ƒæ ¼å¼ï¼Œå¯ä½¿ç”¨clusterProfiler::bitr()å‡½æ•°æˆ–è€…ä¹‹å‰ç¬”è®°æåˆ°çš„æ–¹å¼è¿›è¡Œè½¬æ¢ã€‚\n1 2 3 4 5 6 7 8 9 10 keytypes(org.Hs.eg.db) # [1] \"ACCNUM\" \"ALIAS\" \"ENSEMBL\" \"ENSEMBLPROT\" \"ENSEMBLTRANS\" # [6] \"ENTREZID\" \"ENZYME\" \"EVIDENCE\" \"EVIDENCEALL\" \"GENENAME\" # [11] \"GENETYPE\" \"GO\" \"GOALL\" \"IPI\" \"MAP\" # [16] \"OMIM\" \"ONTOLOGY\" \"ONTOLOGYALL\" \"PATH\" \"PFAM\" # [21] \"PMID\" \"PROSITE\" \"REFSEQ\" \"SYMBOL\" \"UCSCKG\" # [26] \"UNIPROT\" ids \u003c- bitr(x, fromType=\"SYMBOL\", toType=\"ENTREZID\", OrgDb=\"org.Hs.eg.db\") ï¼ˆ1ï¼‰GO ORA 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 ego \u003c- enrichGO(gene = gene_deg, OrgDb = org.Hs.eg.db, ont = \"CC\", # \"BP\",\"MF\",\"ALL\" pAdjustMethod = \"BH\", pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE) t(ego@result[1,]) # GO:0005819 # ID \"GO:0005819\" # Description \"spindle\" # GeneRatio \"26/201\" # BgRatio \"426/19869\" # pvalue \"2.149144e-13\" # p.adjust \"6.339976e-11\" # qvalue \"5.610398e-11\" # geneID \"CDCA8/CDC20/KIF23/CENPE/ASPM/DLGAP5/SKA1/NUSAP1/TPX2/TACC3/NEK2/CDK1/MAD2L1/KIF18A/BIRC5/KIF11/TRAT1/TTK/AURKB/PRC1/KIFC1/KIF18B/KIF20A/AURKA/CCNB1/KIF4A\" # Count \"26\" ## *è¡¥å……ï¼šå¦‚æœè¾“å…¥åŸºå› é›†æ˜¯SYMBOLæ ¼å¼ï¼Œå¯ä»¥è®¾ç½®å‚æ•°keyType = \"SYMBOL\"ç›´æ¥åˆ†æã€‚ä½†åªæœ‰GOæ”¯æŒã€‚ GO å»å†—ä½™ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 ## (1) GOSemSimåŒ…å¯ä»¥è®¡ç®—GO termç›¸ä¼¼åº¦ library(GOSemSim) hsGO \u003c- godata('org.Hs.eg.db', ont=\"MF\") goSim(\"GO:0004022\", \"GO:0005515\", semData=hsGO, measure=\"Jiang\") go1 = c(\"GO:0004022\",\"GO:0004024\",\"GO:0004174\") go2 = c(\"GO:0009055\",\"GO:0005515\") mgoSim(go1, go2, semData=hsGO, measure=\"Wang\", combine=NULL) # pair mgoSim(go1, go2, semData=hsGO, measure=\"Wang\", combine=\"BMA\") # whole # åŸºå› (set)ç›¸ä¼¼åº¦ GOSemSim::geneSim(\"241\", \"251\", semData=hsGO, measure=\"Wang\", combine=\"BMA\") mgeneSim(genes=c(\"835\", \"5261\",\"241\", \"994\"), semData=hsGO, measure=\"Wang\",verbose=FALSE) # pair-wise gs1 \u003c- c(\"835\", \"5261\",\"241\", \"994\", \"514\", \"533\") gs2 \u003c- c(\"578\",\"582\", \"400\", \"409\", \"411\") clusterSim(gs1, gs2, semData=hsGO, measure=\"Wang\", combine=\"BMA\") mclusterSim(list(gs1, gs2, gs3), semData=hsGO, measure=\"Wang\", combine=\"BMA\") ## (2) å¯Œé›†ç»“æœå»å†—ä½™ dim(ego) # [1] 23 9 ego_sim \u003c- clusterProfiler::simplify(ego, cutoff=0.7, measure = \"Wang\", by=\"p.adjust\", select_fun=min) dim(ego_sim) # [1] 8 9 # åŒæ ·ä¹Ÿæ”¯æŒå¯¹GSEAç»“æœçš„å»å†—ä½™å¤„ç† GSEA 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 ## GSEA gseago \u003c- gseGO(geneList = gene_list, OrgDb = org.Hs.eg.db, ont = \"CC\", minGSSize = 100, maxGSSize = 500, pvalueCutoff = 0.05, verbose = FALSE) gseago = setReadable(gseago, OrgDb = org.Hs.eg.db) t(gseago@result[1,]) # GO:0000775 # ID \"GO:0000775\" # Description \"chromosome, centromeric region\" # setSize \"188\" # enrichmentScore \"0.5970689\" # NES \"2.676869\" # pvalue \"1e-10\" # p.adjust \"1.62e-09\" # qvalue \"1.021053e-09\" # rank \"530\" # leading_edge \"tags=20%, list=4%, signal=19%\" # core_enrichment \"CDCA8/CENPE/NDC80/TOP2A/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/EZH2/TTK/NCAPG/AURKB/CCNB1/KIF2C/PLK1/BUB1B/ZWINT/CENPU/SPC25/SPAG5/DSCC1/CENPI/OIP5/HELLS/NCAPD2/CENPA/BUB1/CENPF/ZWILCH/CEBPB/H2BC11\" ï¼ˆ2ï¼‰KEGG ORA 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 # R.utils::setOption(\"clusterProfiler.download.method\",\"auto\") ekg \u003c- enrichKEGG(gene = gene_deg, organism = 'hsa', pvalueCutoff = 0.05) ekg \u003c- setReadable(ekg, OrgDb = org.Hs.eg.db, keyType = \"ENTREZID\") t(ekg@result[1,]) # hsa04110 # ID \"hsa04110\" # Description \"Cell cycle\" # GeneRatio \"11/94\" # BgRatio \"127/8218\" # pvalue \"1.809545e-07\" # p.adjust \"3.781948e-05\" # qvalue \"3.695281e-05\" # geneID \"CDC45/CDC20/CCNB2/CCNA2/CDK1/MAD2L1/TTK/CHEK1/CCNB1/MCM5/PTTG1\" # Count \"11\" browseKEGG(ekg, 'hsa04110') library(\"pathview\") pathview(gene.data = gene_list, pathway.id = \"hsa04110\", species = \"hsa\", limit = list(gene=max(abs(geneList)), cpd=1)) å½“å‡ºç°æŠ¥é”™ç±»ä¼¼äºNo gene can be mapped when using enrichKEGGï¼Œä¸”æ£€æŸ¥è¾“å…¥åŸºå› æ— è¯¯åï¼Œä¸€ä¸ªå¯èƒ½çš„è§£å†³æ–¹æ¡ˆæ˜¯è‡ªå·±æ„å»ºæœ¬åœ°KEGGæ•°æ®åº“ã€‚å‚è€ƒhttps://github.com/YuLab-SMU/clusterProfiler/issues/561ä¸­undo6411è§£ç­”ã€‚æˆ–è€…ä¹Ÿå¯ä»¥ç›´æ¥ä»MsigDBæ•°æ®åº“ä¸‹è½½ï¼Œå†ä½¿ç”¨GSEA()å‡½æ•°ã€‚\nGSEA 1 2 3 4 5 6 7 gseakg \u003c- gseKEGG(geneList = gene_list, organism = 'hsa', minGSSize = 120, pvalueCutoff = 0.05, verbose = FALSE) gseakg \u003c- setReadable(gseakg, OrgDb = org.Hs.eg.db, keyType = \"ENTREZID\") t(gseakg@result[1,]) é’ˆå¯¹å…¶å®ƒé€šè·¯é›†çš„å¯Œé›†åˆ†æå‡½æ•°\né€šè·¯é›† ORAå‡½æ•° GSEAå‡½æ•° WikiPathway enrichWP gseWP Reactome enrichPathway gsePathway DO(disease ontology) enrichDO gseDO NCG(Network of Cancer Gene) enrichNCG gseNCG DisGeNET(disease gene network) enrichDGN gseDGN Mesh(åŒ»å­¦ä¸»é¢˜è¯) enrichMeSH gseMeSH ï¼ˆ3ï¼‰è‡ªå®šä¹‰åŸºå› é›† clusterProfileråŒ…æä¾›çš„enricher()ä¸GSEA()å‡½æ•°å¯å®ç°å¯¹è‡ªå®šä¹‰åŸºå› é›†è¿›è¡Œå¯Œé›†åˆ†æ ä¸»è¦é€šè¿‡TERM2GENEå‚æ•°æä¾›åŸºå› é›†æ•°æ®æ¡†ï¼Œä¸€åˆ—åä¸ºtermä»£è¡¨é€šè·¯åï¼›ä¸€åˆ—ä¸ºgeneä»£è¡¨ç»„æˆåŸºå›  å¦‚ä¸‹ä»¥MsigDBæ•°æ®åº“ä¸ºä¾‹ï¼Œå­¦ä¹ ç”¨æ³•ã€‚è¯¥åº“çš„è¯¦ç»†ä»‹ç»å‚è§ç¬”è®° 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 # install.packages(\"msigdbr\") library(msigdbr) hm_df = msigdbr(species = \"Homo sapiens\") %\u003e% as.data.frame() head(hm_df) colnames(hm_df) # [1] \"gs_cat\" \"gs_subcat\" \"gs_name\" \"gene_symbol\" # [5] \"entrez_gene\" \"ensembl_gene\" \"human_gene_symbol\" \"human_entrez_gene\" # [9] \"human_ensembl_gene\" \"gs_id\" \"gs_pmid\" \"gs_geoid\" # [13] \"gs_exact_source\" \"gs_url\" \"gs_description\" unique(hm_df$gs_cat) # [1] \"C3\" \"C2\" \"C8\" \"C6\" \"C7\" \"C4\" \"C1\" \"C5\" \"H\" unique(hm_df$gs_subcat) # [1] \"MIR:MIR_Legacy\" \"TFT:TFT_Legacy\" \"CGP\" \"TFT:GTRD\" # [5] \"\" \"VAX\" \"CP:BIOCARTA\" \"CGN\" # [9] \"GO:BP\" \"GO:CC\" \"IMMUNESIGDB\" \"GO:MF\" # [13] \"HPO\" \"CP:KEGG\" \"MIR:MIRDB\" \"CM\" # [17] \"CP\" \"CP:PID\" \"CP:REACTOME\" \"CP:WIKIPATHWAYS\" hm_H = msigdbr(species = \"Homo sapiens\", category = \"H\", subcategory = NULL) %\u003e% as.data.frame() %\u003e% dplyr::select(gs_cat, gs_subcat, gs_name, gene_symbol) hm_H = hm_H %\u003e% dplyr::select(gs_name, gene_symbol) %\u003e% dplyr::rename(\"term\" = \"gs_name\", \"gene\" = \"gene_symbol\") head(hm_H) # term gene # 1 HALLMARK_ADIPOGENESIS ABCA1 # 2 HALLMARK_ADIPOGENESIS ABCB8 # 3 HALLMARK_ADIPOGENESIS ACAA2 # 4 HALLMARK_ADIPOGENESIS ACADL # 5 HALLMARK_ADIPOGENESIS ACADM # 6 HALLMARK_ADIPOGENESIS ACADS å¯Œé›†åˆ†æ 1 2 3 4 5 6 7 8 9 10 # æ³¨æ„ä¸Šè¿°é€‰æ‹©çš„æ˜¯Symbolæ ¼å¼åŸºå› ï¼Œè¾“å…¥åŸºå› ä¹Ÿéœ€è¦ä¸ä¹‹åŒ¹é… res_go \u003c- enricher(gene_deg2, TERM2GENE = hm_H) dim(res_go) dim(res_go@result) res_go_df = res_go@result res_gsea \u003c- GSEA(gene_list2, TERM2GENE = hm_H) dim(res_gsea) dim(res_gsea@result) res_gsea_df = res_gsea@result 3ã€å¯è§†åŒ– ä»¥ä¸Šé¢åˆ†æçš„egoä¸gseagoä¸ºä¾‹ 1 2 library(enrichplot) # é™¤ç¬¬ä¸€ä¸ªbarplotå‡½æ•°ï¼Œå…¶ä½™å‡åŸºäºggplotç»˜å›¾ä½“ç³» ï¼ˆ1ï¼‰æŸ±çŠ¶å›¾/ç‚¹å›¾ 1 2 3 4 5 6 7 8 9 ## barplot (RåŸºç¡€å‡½æ•°) barplot(ego, showCategory=10) # æŒ‡å®šGOé€šè·¯é›† pw2show = c(\"GO:0005819\",\"GO:0072686\",\"GO:0098687\",\"GO:0000779\") barplot(ego, showCategory=ego@result$Description[which(rownames(ego@result) %in% pw2show)]) # é’ˆå¯¹ont = \"ALL\"çš„æƒ…å†µ barplot(ego, split = \"ONTOLOGY\") + facet_grid(ONTOLOGY~., scale = \"free\") 1 2 3 4 5 dotplot(ego) ## é»˜è®¤å¦‚ä¸‹å¯¹åº”å…³ç³»,å¯è®¾ç½®å‚æ•°ä¿®æ”¹è¿™äº›å…³ç³»ï¼š # é¢œè‰²æ˜ å°„På€¼ # å¤§å°æ˜ å°„äº¤é›†åŸºå› æ•°(å·®å¼‚åŸºå› ä¸é€šè·¯åŸºå› é›†) # æ¨ªè½´è¡¨ç¤ºæ¯”ä¾‹(count/geneset) 1 2 edox \u003c- pairwise_termsim(ego) treeplot(edox, cluster.params = list(method=\"ward.D\", n = 4, label_words_n = 2)) ï¼ˆ2ï¼‰GSEAæ‰“åˆ† å•ä¸ªé€šè·¯GSEA 1 2 3 4 5 gseaplot(gseago, geneSetID = 1, title = gseago$Description[1]) # by = \"preranked\"åªå±•ç¤ºä¸ŠåŠéƒ¨åˆ† # by = \"runningScore\"åªå±•ç¤ºä¸‹åŠéƒ¨åˆ† å¤šæ¡é€šè·¯ç»“æœ 1 2 3 gseaplot2(gseago, geneSetID = 1:3, pvalue_table = T) ï¼ˆ3ï¼‰é€šè·¯ä¸åŸºå›  çƒ­å›¾ 1 2 3 heatplot(ego, foldChange=gene_list, showCategory = 5) ç½‘ç»œå›¾ 1 2 3 4 5 6 library(patchwork) p1 = cnetplot(ego, foldChange=gene_list,showCategory = 3) p2 = cnetplot(ego, foldChange=gene_list,showCategory = 3, circular = T) p3 = cnetplot(ego, foldChange=gene_list,showCategory = 3, colorEdge = T, node_label=\"gene\") p4 = cnetplot(ego, foldChange=gene_list,showCategory = 3, layout = \"gem\") (p1 + p2) / (p3 + p4) ï¼ˆ4ï¼‰é€šè·¯ä¸é€šè·¯ 1 2 3 4 5 ego_pair \u003c- enrichplot::pairwise_termsim(ego) emapplot(ego_pair, showCategory = 15, layout=\"kk\", cex_category=1.5,min_edge = 0.8) gseago_pair \u003c- enrichplot::pairwise_termsim(gseago) emapplot(gseago_pair, showCategory = 15, layout=\"kk\", min_edge = 0.1, color = \"NES\") 4ã€å¤šåŸºå› é›†åˆ†æ clusterProfileråŒ…çš„compareCluster()å‡½æ•°æä¾›äº†é’ˆå¯¹å¤šç»„å®éªŒè®¾è®¡å¾—åˆ°çš„è‹¥å¹²å·®å¼‚åŸºå› åˆ—è¡¨çš„åŒæ—¶å¯Œé›†åˆ†æ(ORA)ç»“æœï¼ŒåŠå‹å¥½çš„å¯è§†åŒ–ã€‚ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 ## (1) ç›´æ¥å¤šåŸºå› é›†çš„å¯Œé›†é€šè·¯æ¯”è¾ƒ data(gcSample) str(gcSample[5:8]) #å–å4ä¸ªå·®å¼‚åŸºå› åˆ—è¡¨åšæ¼”ç¤º # List of 4 # $ X5: chr [1:929] \"5982\" \"7318\" \"6352\" \"2101\" ... # $ X6: chr [1:585] \"5337\" \"9295\" \"4035\" \"811\" ... # $ X7: chr [1:582] \"2621\" \"2665\" \"5690\" \"3608\" ... # $ X8: chr [1:237] \"2665\" \"4735\" \"1327\" \"3192\" ... ## ä¸¤ä¸ªä¸»è¦å‚æ•° # å¤šä¸ªå·®å¼‚åŸºå› åˆ—è¡¨ # å¯Œé›†åˆ†æmethod (One of \"groupGO\", \"enrichGO\", \"enrichKEGG\", \"enrichDO\" or \"enrichPathway\") ck \u003c- compareCluster(geneCluster = gcSample[5:8], fun = enrichKEGG) ck \u003c- setReadable(ck, OrgDb = org.Hs.eg.db, keyType=\"ENTREZID\") table(ck@compareClusterResult$Cluster) # X5 X6 X7 X8 # 10 1 15 19 dotplot(ck) 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 ## (2) å‡å¦‚åˆ†åˆ«ç”¨ä¸¤ç§è¯ç‰©Aã€Bè¿›è¡Œå¹²æ‰°ç»†èƒç³»ï¼Œç»1hã€3håå¾—åˆ°4ç»„å·®å¼‚åŸºå›  group_A_1h=gcSample[[2]] group_A_3h=gcSample[[3]] group_B_1h=gcSample[[4]] group_B_3h=gcSample[[5]] design=c(rep(\"A_1h\",length(group_A_1h)), rep(\"A_3h\",length(group_A_3h)), rep(\"B_1h\",length(group_B_1h)), rep(\"B_3h\",length(group_B_3h))) mydf = data.frame(Drug = stringr::str_split(design,\"_\",simplify = T)[,1], Time = stringr::str_split(design,\"_\",simplify = T)[,2], DEG = c(group_A_1h,group_A_3h,group_B_1h,group_B_3h)) # Drug Time DEG # 1 A 1h 23450 # 2 A 1h 5160 # 3 A 1h 7126 formula_res \u003c- compareCluster(DEG~Time+Drug, data=mydf, fun=\"enrichKEGG\") table(formula_res@compareClusterResult$Cluster) # 1h.A 1h.B 3h.A 3h.B # 3 18 3 10 dotplot(formula_res) + dotplot(formula_res, x=\"Time\") + facet_grid(~Drug) ",
  "wordCount" : "3645",
  "inLanguage": "en",
  "datePublished": "2023-01-14T00:00:00Z",
  "dateModified": "2023-01-14T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/056clusterprofiler%E5%8C%85%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="ä¸»é¡µ">
                    <span>ä¸»é¡µ</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="åˆ†ç±»">
                    <span>åˆ†ç±»</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="æ ‡ç­¾">
                    <span>æ ‡ç­¾</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="å½’æ¡£">
                    <span>å½’æ¡£</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="å…³äº">
                    <span>å…³äº</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="æœç´¢ (Alt &#43; /)" accesskey=/>
                    <span>æœç´¢</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/">åˆ†ç±»</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰</a></div>
    <h1 class="post-title">
      clusterProfileråŒ…å¯Œé›†åˆ†æä¸å¯è§†åŒ–
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-14 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-14&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-14&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;3645&amp;nbsp;|&amp;nbsp;8 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e8%83%8c%e6%99%af%e7%9f%a5%e8%af%86" aria-label="1ã€èƒŒæ™¯çŸ¥è¯†">1ã€èƒŒæ™¯çŸ¥è¯†</a><ul>
                            
                    <li>
                        <a href="#1%e4%b8%a4%e7%a7%8d%e5%af%8c%e9%9b%86%e5%88%86%e6%9e%90" aria-label="ï¼ˆ1ï¼‰ä¸¤ç§å¯Œé›†åˆ†æ">ï¼ˆ1ï¼‰ä¸¤ç§å¯Œé›†åˆ†æ</a></li>
                    <li>
                        <a href="#2%e9%80%9a%e8%b7%af%e5%9f%ba%e5%9b%a0%e9%9b%86" aria-label="ï¼ˆ2ï¼‰é€šè·¯åŸºå› é›†">ï¼ˆ2ï¼‰é€šè·¯åŸºå› é›†</a></li>
                    <li>
                        <a href="#3clusterprofiler%e5%8c%85" aria-label="ï¼ˆ3ï¼‰clusterProfileråŒ…">ï¼ˆ3ï¼‰clusterProfileråŒ…</a></li></ul>
                    </li>
                    <li>
                        <a href="#2%e5%af%8c%e9%9b%86%e5%88%86%e6%9e%90" aria-label="2ã€å¯Œé›†åˆ†æ">2ã€å¯Œé›†åˆ†æ</a><ul>
                            
                    <li>
                        <a href="#1go" aria-label="ï¼ˆ1ï¼‰GO">ï¼ˆ1ï¼‰GO</a></li>
                    <li>
                        <a href="#2kegg" aria-label="ï¼ˆ2ï¼‰KEGG">ï¼ˆ2ï¼‰KEGG</a></li>
                    <li>
                        <a href="#3%e8%87%aa%e5%ae%9a%e4%b9%89%e5%9f%ba%e5%9b%a0%e9%9b%86" aria-label="ï¼ˆ3ï¼‰è‡ªå®šä¹‰åŸºå› é›†">ï¼ˆ3ï¼‰è‡ªå®šä¹‰åŸºå› é›†</a></li></ul>
                    </li>
                    <li>
                        <a href="#3%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="3ã€å¯è§†åŒ–">3ã€å¯è§†åŒ–</a><ul>
                            
                    <li>
                        <a href="#1%e6%9f%b1%e7%8a%b6%e5%9b%be%e7%82%b9%e5%9b%be" aria-label="ï¼ˆ1ï¼‰æŸ±çŠ¶å›¾/ç‚¹å›¾">ï¼ˆ1ï¼‰æŸ±çŠ¶å›¾/ç‚¹å›¾</a></li>
                    <li>
                        <a href="#2gsea%e6%89%93%e5%88%86" aria-label="ï¼ˆ2ï¼‰GSEAæ‰“åˆ†">ï¼ˆ2ï¼‰GSEAæ‰“åˆ†</a></li>
                    <li>
                        <a href="#3%e9%80%9a%e8%b7%af%e4%b8%8e%e5%9f%ba%e5%9b%a0" aria-label="ï¼ˆ3ï¼‰é€šè·¯ä¸åŸºå› ">ï¼ˆ3ï¼‰é€šè·¯ä¸åŸºå› </a></li>
                    <li>
                        <a href="#4%e9%80%9a%e8%b7%af%e4%b8%8e%e9%80%9a%e8%b7%af" aria-label="ï¼ˆ4ï¼‰é€šè·¯ä¸é€šè·¯">ï¼ˆ4ï¼‰é€šè·¯ä¸é€šè·¯</a></li></ul>
                    </li>
                    <li>
                        <a href="#4%e5%a4%9a%e5%9f%ba%e5%9b%a0%e9%9b%86%e5%88%86%e6%9e%90" aria-label="4ã€å¤šåŸºå› é›†åˆ†æ">4ã€å¤šåŸºå› é›†åˆ†æ</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><h2 id="1èƒŒæ™¯çŸ¥è¯†">1ã€èƒŒæ™¯çŸ¥è¯†<a hidden class="anchor" aria-hidden="true" href="#1èƒŒæ™¯çŸ¥è¯†">#</a></h2>
<h3 id="1ä¸¤ç§å¯Œé›†åˆ†æ">ï¼ˆ1ï¼‰ä¸¤ç§å¯Œé›†åˆ†æ<a hidden class="anchor" aria-hidden="true" href="#1ä¸¤ç§å¯Œé›†åˆ†æ">#</a></h3>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-2db430f93bc01a2d.png" alt="img" style="zoom: 33%;" />
<ul>
<li>åŸºäºè¶…å‡ ä½•æ£€éªŒçš„ORA(over representation analysis)å¯Œé›†åˆ†æ</li>
</ul>
<p>â‘  å‡è®¾å¯¹è½¬å½•ç»„åˆ†ç»„æµ‹åºçš„<strong>10000</strong>ä¸ªåŸºå› è¡¨è¾¾æ•°æ®è¿›è¡Œå·®å¼‚åˆ†æï¼›</p>
<p>â‘¡ æŒ‰ç…§è§„å®šcutoffé˜ˆå€¼(logFC/Pvalue)ç­›é€‰å¾—åˆ°<strong>100</strong>ä¸ªå·®å¼‚è¡¨è¾¾åŸºå› ï¼›</p>
<p>â‘¢ å¯¹äºç‰¹å®šé€šè·¯ï¼ŒåŒ…å«<strong>500</strong>ä¸ªåŸºå› ï¼›å…¶ä¸­æœ‰<strong>30</strong>ä¸ªå±äºå·®å¼‚åŸºå› </p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-5513efd1d8b2fe28.png" alt="img"  />
</p>
<p>â‘£ å·®å¼‚åŸºå› æ˜¯å¦å¯Œé›†åˆ°è¯¥é€šè·¯ï¼Œå³è¡¨ç¤º30/70ä¸470/9430ç›¸æ¯”ï¼Œæ˜¯å¦å…·æœ‰æ˜¾è‘—æ„ä¹‰ã€‚</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-2126ef63ab6d216b.png" alt="img"  />
</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># å¦‚ä¸Šå…¬å¼ï¼šN=10000ï¼ŒM=500ï¼Œn=100ï¼Œk=30</span>
</span></span><span style="display:flex;"><span>d &lt;- data.frame(non_DEG=c(<span style="color:#ff0;font-weight:bold">470</span>, <span style="color:#ff0;font-weight:bold">9430</span>), DEG=c(<span style="color:#ff0;font-weight:bold">30</span>, <span style="color:#ff0;font-weight:bold">70</span>))
</span></span><span style="display:flex;"><span>row.names(d) &lt;- c(<span style="color:#0ff;font-weight:bold">&#34;In set&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;not in category&#34;</span>)
</span></span><span style="display:flex;"><span>fisher.test(d, alternative = <span style="color:#0ff;font-weight:bold">&#34;less&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-e3834febf7457d72.png" alt="img"  />
</p>
<p>å¦‚ä¸Šè®¡ç®—ç»“æœï¼Œpå€¼è¶Šå°(&lt;0.05)ï¼Œè¯´æ˜å¾—åˆ°çš„å·®å¼‚åŸºå› æ˜¯æ˜¾è‘—å¯Œé›†åˆ°è¿™ä¸ªé€šè·¯åŸºå› é›†ä¸Šçš„ã€‚ç„¶åå°±å¯ä»¥è¿™ä¸ªåŸºå› é›†çš„ç”Ÿç‰©å­¦æ„ä¹‰è®²è¿°åç»­çš„æ•…äº‹äº†ã€‚</p>
<ul>
<li>åŸºäºç½®æ¢æ£€éªŒçš„GSEA(gene set enrichment analysis)å¯Œé›†åˆ†æ</li>
</ul>
<p>â‘  å‡è®¾å¯¹è½¬å½•ç»„åˆ†ç»„æµ‹åºçš„<strong>10000</strong>ä¸ªåŸºå› è¡¨è¾¾æ•°æ®è¿›è¡Œå·®å¼‚åˆ†æï¼Œä¸€èˆ¬æŒ‰ç…§å·®å¼‚å€æ•°ä»é«˜åˆ°ä½è¿›è¡Œæ’åºï¼›</p>
<p>â‘¡ å‡è®¾å¯¹äºåŒ…å«500ä¸ªåŸºå› çš„ç‰¹å®šé€šè·¯ï¼Œéå†ä¸Šè¿°10000ä¸ªåŸºå› çš„æ’åºåˆ—è¡¨ï¼›</p>
<p>â‘¢ åœ¨éå†è¿‡ç¨‹ä¸­ï¼Œå¦‚åŸºå› å±äºé€šè·¯é›†åˆ™åŠ åˆ†(hit)ï¼Œä¸å±äºåˆ™å‡åˆ†ï¼›åŠ å‡åˆ†çš„æƒé‡ä¸å·®å¼‚å€æ•°å…ˆå…³ã€‚</p>
<p>â‘£ ES(enrichment score)è¡¨ç¤ºéå†è¿‡ç¨‹ä¸­äº§ç”Ÿçš„ç»å¯¹å€¼æœ€å¤§å€¼ã€‚ESå¤§äº0ï¼Œè¡¨æ˜é€šè·¯é›†å¯Œé›†åˆ°æ’åºåˆ—è¡¨çš„é¡¶éƒ¨ï¼›åä¹‹åˆ™å¯Œé›†åˆ°æ’åºåˆ—è¡¨çš„åº•éƒ¨ã€‚</p>
<p>â‘¤ éšæœºé€‰æ‹©500ä¸ªåŸºå› è‹¥å¹²æ¬¡ï¼Œè®¡ç®—ESèƒŒæ™¯åˆ†å¸ƒï¼Œä»è€Œè®¡ç®—ç›¸åº”çš„På€¼ã€‚</p>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/zpq0370595180001.jpeg" alt="Gene set enrichment analysis: A knowledge-based approach for interpreting  genome-wide expression profiles | PNAS" style="zoom:33%;" />
<h3 id="2é€šè·¯åŸºå› é›†">ï¼ˆ2ï¼‰é€šè·¯åŸºå› é›†<a hidden class="anchor" aria-hidden="true" href="#2é€šè·¯åŸºå› é›†">#</a></h3>
<p>å¸¸è§çš„é€šè·¯åŸºå› é›†åŒ…æ‹¬GOã€KEGGã€Reactomeç­‰ã€‚MsigDBæ•°æ®åº“åˆ™ç»¼åˆæ”¶é›†äº†å¤šç§é€šè·¯åŸºå› é›†ã€‚4</p>
<ul>
<li>
<p>GO</p>
<ul>
<li>Gene Ontology åŸºå› æœ¬ä½“è®º</li>
<li><a href="https://links.jianshu.com/go?to=http%3A%2F%2Fwww.geneontology.org%2F">http://www.geneontology.org/ </a></li>
<li>GO åˆ†åˆ«å®šä¹‰äº†ä¸‰ç§ç±»å‹çš„åŸºå› é›†ï¼šâ‘ ç»†èƒç»„æˆï¼ˆcellular componentï¼ŒCCï¼‰ï¼›â‘¡ç”Ÿç‰©è¿‡ç¨‹ï¼ˆbiological processï¼ŒBPï¼‰ï¼›â‘¢åˆ†å­åŠŸèƒ½ï¼ˆMolecular Functionï¼ŒMFï¼‰ã€‚è¯¦è§<a href="https://links.jianshu.com/go?to=http%3A%2F%2Fgeneontology.org%2Fdocs%2Fontology-documentation%2F">http://geneontology.org/docs/ontology-documentation/</a></li>
<li>ä¸¾ä¸ªä¾‹å­æ¥è¯´ï¼šåŸºå› A(å…¶è›‹ç™½äº§ç‰©æ˜¯cytochrome c(ç»†èƒè‰²ç´ c)ï¼‰ï¼Œè¯¥åŸºå› çš„å…·æœ‰æ°§åŒ–è¿˜åŸæ´»æ€§ï¼ˆmolecular functionï¼‰ï¼›å‚ä¸æ°§åŒ–ç£·é…¸åŒ–ç”Ÿç‰©å­¦è¿‡ç¨‹(molecular function)ï¼›è€Œå‘æŒ¥ä½œç”¨çš„ä½ç½®ä½äºç»†èƒçº¿ç²’ä½“åŸºè´¨(cellular component)</li>
</ul>
</li>
<li>
<p>KEGG</p>
<ul>
<li>Kyoto Encyclopedia of Genes and Genomes äº¬éƒ½åŸºå› ä¸åŸºå› ç»„ç™¾ç§‘å…¨ä¹¦</li>
<li><a href="https://links.jianshu.com/go?to=https%3A%2F%2Fwww.genome.jp%2Fkegg%2F">https://www.genome.jp/kegg/</a></li>
<li>KEGGæ˜¯ç³»ç»Ÿåˆ†æåŸºå› åŠŸèƒ½ï¼Œè”ç³»åŸºå› ç»„ä¿¡æ¯å’ŒåŠŸèƒ½ä¿¡æ¯ï¼Œä»¥æ¢ç´¢åˆ†å­ç›¸äº’ä½œç”¨å’Œååº”ç½‘ç»œçš„çŸ¥è¯†åº“ï¼ˆäººå·¥ç»˜åˆ¶ï¼‰,å…¶ä¸­åŒ…å«æœ‰å¤§é‡çš„é€šè·¯ï¼ˆPATHWAYï¼‰å›¾ï¼Œæ¶‰åŠmetabolism, genetic and environmental information processing, cellular processes, organismal systems, human diseases, and drug developmentä¸ƒå¤§ç±»ã€‚</li>
</ul>
</li>
<li>
<p>Reactome  <a href="https://reactome.org/">https://reactome.org/</a></p>
</li>
<li>
<p>WikiPathway <a href="https://www.wikipathways.org/">https://www.wikipathways.org/</a></p>
</li>
<li>
<p>Msigdb</p>
<ul>
<li>Molecular Signatures Database åˆ†å­ç‰¹å¾æ•°æ®åº“</li>
<li><a href="https://www.gsea-msigdb.org/gsea/msigdb/index.jsp">https://www.gsea-msigdb.org/gsea/msigdb/index.jsp</a></li>
<li>æˆªæ­¢ç›®å‰è¯¥æ•°æ®åº“æ”¶é›†æœ‰32284ä¸ªåŸºå› é›†ï¼Œå¯åˆ†ä¸º9å¤§ç±» &mdash; H: hallmark gene setsï¼ŒC1: positional gene setsï¼ŒC2: curated gene setsï¼ŒC3: regulatory target gene setsï¼ŒC4: computational gene setsï¼ŒC5: ontology gene setsï¼ŒC6: oncogenic signature gene setsï¼ŒC7: immunologic signature gene setsï¼ŒC8: cell type signature gene sets</li>
</ul>
</li>
</ul>
<blockquote>
<p>æ­¤å¤–è¿˜æœ‰å…¶å®ƒç±»å‹é€šè·¯åŸºå› é›†ï¼Œä¾‹å¦‚ç–¾ç—…åŸºå› é›†DOï¼Œå°±ä¸ä¸€ä¸€åˆ—ä¸¾äº†ã€‚</p></blockquote>
<h3 id="3clusterprofileråŒ…">ï¼ˆ3ï¼‰clusterProfileråŒ…<a hidden class="anchor" aria-hidden="true" href="#3clusterprofileråŒ…">#</a></h3>
<p>RåŒ…clusterProfilerç”±å—æ–¹åŒ»ç§‘å¤§å­¦ä½™å…‰åˆ›å›¢é˜Ÿå¼€å‘ï¼Œæ˜¯ä¸€ä¸ªå¸¸ç”¨çš„å¯Œé›†åˆ†æå·¥å…·åŒ…ï¼Œç›®å‰å·²æ›´æ–°åˆ°4.xç³»åˆ—ç‰ˆæœ¬ã€‚è¯¥åŒ…åŒ…å«äº†å¯Œé›†åˆ†æç›¸å…³çš„æ–¹æ–¹é¢é¢å‡½æ•°åŠŸèƒ½ï¼Œä»ä¸Šæ¸¸åˆ†æåˆ°ä¸‹æ¸¸å¯è§†åŒ–ç­‰ä¸€åº”ä¿±å…¨ã€‚åç»­ç¬”è®°ä¸»è¦å‚è€ƒå®˜æ–¹æ•™ç¨‹å­¦ä¹ è¯¥åŒ…çš„åŸºç¡€ç”¨æ³•ã€‚</p>
<ul>
<li>DOI: <a href="https://doi.org/10.1016/j.xinn.2021.100141">10.1016/j.xinn.2021.100141</a></li>
<li>Tutorialï¼šhttps://yulab-smu.top/biomedical-knowledge-mining-book/index.html</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230113092610681.png" alt="image-20230113092610681"  />
</p>
<h2 id="2å¯Œé›†åˆ†æ">2ã€å¯Œé›†åˆ†æ<a hidden class="anchor" aria-hidden="true" href="#2å¯Œé›†åˆ†æ">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## åŠ è½½åŒ…</span>
</span></span><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>library(clusterProfiler)
</span></span><span style="display:flex;"><span>library(org.Hs.eg.db)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## ç¤ºä¾‹åŸºå› æ•°æ®</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (1) é™åºå·®å¼‚åŸºå› åˆ—è¡¨ç”¨äºGSEAåˆ†æ</span>
</span></span><span style="display:flex;"><span>data(geneList, package=<span style="color:#0ff;font-weight:bold">&#34;DOSE&#34;</span>)
</span></span><span style="display:flex;"><span>str(geneList)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Named num [1:12495] 4.57 4.51 4.42 4.14 3.88 ...</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># - attr(*, &#34;names&#34;)= chr [1:12495] &#34;4312&#34; &#34;8318&#34; &#34;10874&#34; &#34;55143&#34; ...</span>
</span></span><span style="display:flex;"><span>gene_list = geneList
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2) æ˜¾è‘—å·®å¼‚åŸºå› é›†ç”¨äºORAåˆ†æ</span>
</span></span><span style="display:flex;"><span>gene_deg = names(geneList)[abs(geneList) &gt; <span style="color:#ff0;font-weight:bold">2</span>]
</span></span><span style="display:flex;"><span>head(gene_deg)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;4312&#34;  &#34;8318&#34;  &#34;10874&#34; &#34;55143&#34; &#34;55388&#34; &#34;991&#34; </span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>éœ€è¦ä½¿ç”¨ENTREZIDçš„åŸºå› IDã€‚å¦‚æœæ˜¯å…¶å®ƒæ ¼å¼ï¼Œå¯ä½¿ç”¨<code>clusterProfiler::bitr()</code>å‡½æ•°æˆ–è€…ä¹‹å‰ç¬”è®°æåˆ°çš„æ–¹å¼è¿›è¡Œè½¬æ¢ã€‚</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>keytypes(org.Hs.eg.db)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;ACCNUM&#34;       &#34;ALIAS&#34;        &#34;ENSEMBL&#34;      &#34;ENSEMBLPROT&#34;  &#34;ENSEMBLTRANS&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [6] &#34;ENTREZID&#34;     &#34;ENZYME&#34;       &#34;EVIDENCE&#34;     &#34;EVIDENCEALL&#34;  &#34;GENENAME&#34;    </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [11] &#34;GENETYPE&#34;     &#34;GO&#34;           &#34;GOALL&#34;        &#34;IPI&#34;          &#34;MAP&#34;         </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [16] &#34;OMIM&#34;         &#34;ONTOLOGY&#34;     &#34;ONTOLOGYALL&#34;  &#34;PATH&#34;         &#34;PFAM&#34;        </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [21] &#34;PMID&#34;         &#34;PROSITE&#34;      &#34;REFSEQ&#34;       &#34;SYMBOL&#34;       &#34;UCSCKG&#34;      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [26] &#34;UNIPROT&#34;     </span>
</span></span><span style="display:flex;"><span>ids &lt;- bitr(x, fromType=<span style="color:#0ff;font-weight:bold">&#34;SYMBOL&#34;</span>, 
</span></span><span style="display:flex;"><span>            toType=<span style="color:#0ff;font-weight:bold">&#34;ENTREZID&#34;</span>, 
</span></span><span style="display:flex;"><span>            OrgDb=<span style="color:#0ff;font-weight:bold">&#34;org.Hs.eg.db&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div></blockquote>
<h3 id="1go">ï¼ˆ1ï¼‰GO<a hidden class="anchor" aria-hidden="true" href="#1go">#</a></h3>
<ul>
<li><strong>ORA</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>ego &lt;- enrichGO(gene          = gene_deg,
</span></span><span style="display:flex;"><span>                OrgDb         = org.Hs.eg.db,
</span></span><span style="display:flex;"><span>                ont           = <span style="color:#0ff;font-weight:bold">&#34;CC&#34;</span>, <span style="color:#007f7f"># &#34;BP&#34;,&#34;MF&#34;,&#34;ALL&#34;</span>
</span></span><span style="display:flex;"><span>                pAdjustMethod = <span style="color:#0ff;font-weight:bold">&#34;BH&#34;</span>,
</span></span><span style="display:flex;"><span>                pvalueCutoff  = <span style="color:#ff0;font-weight:bold">0.01</span>,
</span></span><span style="display:flex;"><span>                qvalueCutoff  = <span style="color:#ff0;font-weight:bold">0.05</span>,
</span></span><span style="display:flex;"><span>                readable      = <span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>t(ego@result[1,])
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#              GO:0005819                                                                                                                                                 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ID          &#34;GO:0005819&#34;                                                                                                                                               </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Description &#34;spindle&#34;                                                                                                                                                  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GeneRatio   &#34;26/201&#34;                                                                                                                                                   </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># BgRatio     &#34;426/19869&#34;                                                                                                                                                </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># pvalue      &#34;2.149144e-13&#34;                                                                                                                                             </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># p.adjust    &#34;6.339976e-11&#34;                                                                                                                                             </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># qvalue      &#34;5.610398e-11&#34;                                                                                                                                             </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># geneID      &#34;CDCA8/CDC20/KIF23/CENPE/ASPM/DLGAP5/SKA1/NUSAP1/TPX2/TACC3/NEK2/CDK1/MAD2L1/KIF18A/BIRC5/KIF11/TRAT1/TTK/AURKB/PRC1/KIFC1/KIF18B/KIF20A/AURKA/CCNB1/KIF4A&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Count       &#34;26&#34;  </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## *è¡¥å……ï¼šå¦‚æœè¾“å…¥åŸºå› é›†æ˜¯SYMBOLæ ¼å¼ï¼Œå¯ä»¥è®¾ç½®å‚æ•°keyType = &#34;SYMBOL&#34;ç›´æ¥åˆ†æã€‚ä½†åªæœ‰GOæ”¯æŒã€‚</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li><strong>GO å»å†—ä½™</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## (1) GOSemSimåŒ…å¯ä»¥è®¡ç®—GO termç›¸ä¼¼åº¦</span>
</span></span><span style="display:flex;"><span>library(GOSemSim)
</span></span><span style="display:flex;"><span>hsGO &lt;- godata(<span style="color:#0ff;font-weight:bold">&#39;org.Hs.eg.db&#39;</span>, ont=<span style="color:#0ff;font-weight:bold">&#34;MF&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>goSim(<span style="color:#0ff;font-weight:bold">&#34;GO:0004022&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;GO:0005515&#34;</span>, semData=hsGO, measure=<span style="color:#0ff;font-weight:bold">&#34;Jiang&#34;</span>)
</span></span><span style="display:flex;"><span>go1 = c(<span style="color:#0ff;font-weight:bold">&#34;GO:0004022&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;GO:0004024&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;GO:0004174&#34;</span>)
</span></span><span style="display:flex;"><span>go2 = c(<span style="color:#0ff;font-weight:bold">&#34;GO:0009055&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;GO:0005515&#34;</span>)
</span></span><span style="display:flex;"><span>mgoSim(go1, go2, semData=hsGO, measure=<span style="color:#0ff;font-weight:bold">&#34;Wang&#34;</span>, combine=<span style="color:#fff;font-weight:bold">NULL</span>)  <span style="color:#007f7f"># pair</span>
</span></span><span style="display:flex;"><span>mgoSim(go1, go2, semData=hsGO, measure=<span style="color:#0ff;font-weight:bold">&#34;Wang&#34;</span>, combine=<span style="color:#0ff;font-weight:bold">&#34;BMA&#34;</span>) <span style="color:#007f7f"># whole</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># åŸºå› (set)ç›¸ä¼¼åº¦</span>
</span></span><span style="display:flex;"><span>GOSemSim::geneSim(<span style="color:#0ff;font-weight:bold">&#34;241&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;251&#34;</span>, semData=hsGO, measure=<span style="color:#0ff;font-weight:bold">&#34;Wang&#34;</span>, combine=<span style="color:#0ff;font-weight:bold">&#34;BMA&#34;</span>)
</span></span><span style="display:flex;"><span>mgeneSim(genes=c(<span style="color:#0ff;font-weight:bold">&#34;835&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;5261&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;241&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;994&#34;</span>),
</span></span><span style="display:flex;"><span>         semData=hsGO, measure=<span style="color:#0ff;font-weight:bold">&#34;Wang&#34;</span>,verbose=<span style="color:#fff;font-weight:bold">FALSE</span>)  <span style="color:#007f7f"># pair-wise</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gs1 &lt;- c(<span style="color:#0ff;font-weight:bold">&#34;835&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;5261&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;241&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;994&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;514&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;533&#34;</span>)
</span></span><span style="display:flex;"><span>gs2 &lt;- c(<span style="color:#0ff;font-weight:bold">&#34;578&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;582&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;400&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;409&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;411&#34;</span>)
</span></span><span style="display:flex;"><span>clusterSim(gs1, gs2, semData=hsGO, measure=<span style="color:#0ff;font-weight:bold">&#34;Wang&#34;</span>, combine=<span style="color:#0ff;font-weight:bold">&#34;BMA&#34;</span>)
</span></span><span style="display:flex;"><span>mclusterSim(list(gs1, gs2, gs3), semData=hsGO, measure=<span style="color:#0ff;font-weight:bold">&#34;Wang&#34;</span>, combine=<span style="color:#0ff;font-weight:bold">&#34;BMA&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (2) å¯Œé›†ç»“æœå»å†—ä½™</span>
</span></span><span style="display:flex;"><span>dim(ego)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 23  9</span>
</span></span><span style="display:flex;"><span>ego_sim &lt;- clusterProfiler::simplify(ego, cutoff=<span style="color:#ff0;font-weight:bold">0.7</span>, measure = <span style="color:#0ff;font-weight:bold">&#34;Wang&#34;</span>,
</span></span><span style="display:flex;"><span>                                     by=<span style="color:#0ff;font-weight:bold">&#34;p.adjust&#34;</span>, select_fun=min)
</span></span><span style="display:flex;"><span>dim(ego_sim)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 8 9</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># åŒæ ·ä¹Ÿæ”¯æŒå¯¹GSEAç»“æœçš„å»å†—ä½™å¤„ç†</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li><strong>GSEA</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## GSEA</span>
</span></span><span style="display:flex;"><span>gseago &lt;- gseGO(geneList     = gene_list,
</span></span><span style="display:flex;"><span>                OrgDb        = org.Hs.eg.db,
</span></span><span style="display:flex;"><span>                ont          = <span style="color:#0ff;font-weight:bold">&#34;CC&#34;</span>,
</span></span><span style="display:flex;"><span>                minGSSize    = <span style="color:#ff0;font-weight:bold">100</span>,
</span></span><span style="display:flex;"><span>                maxGSSize    = <span style="color:#ff0;font-weight:bold">500</span>,
</span></span><span style="display:flex;"><span>                pvalueCutoff = <span style="color:#ff0;font-weight:bold">0.05</span>,
</span></span><span style="display:flex;"><span>                verbose      = <span style="color:#fff;font-weight:bold">FALSE</span>)
</span></span><span style="display:flex;"><span>gseago = setReadable(gseago, OrgDb = org.Hs.eg.db)
</span></span><span style="display:flex;"><span>t(gseago@result[1,])
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                  GO:0000775                                                                                                                                                                                                                  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ID              &#34;GO:0000775&#34;                                                                                                                                                                                                                </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Description     &#34;chromosome, centromeric region&#34;                                                                                                                                                                                            </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># setSize         &#34;188&#34;                                                                                                                                                                                                                       </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># enrichmentScore &#34;0.5970689&#34;                                                                                                                                                                                                                 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NES             &#34;2.676869&#34;                                                                                                                                                                                                                  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># pvalue          &#34;1e-10&#34;                                                                                                                                                                                                                     </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># p.adjust        &#34;1.62e-09&#34;                                                                                                                                                                                                                  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># qvalue          &#34;1.021053e-09&#34;                                                                                                                                                                                                              </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># rank            &#34;530&#34;                                                                                                                                                                                                                       </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># leading_edge    &#34;tags=20%, list=4%, signal=19%&#34;                                                                                                                                                                                             </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># core_enrichment &#34;CDCA8/CENPE/NDC80/TOP2A/HJURP/SKA1/NEK2/CENPM/CENPN/ERCC6L/MAD2L1/KIF18A/CDT1/BIRC5/EZH2/TTK/NCAPG/AURKB/CCNB1/KIF2C/PLK1/BUB1B/ZWINT/CENPU/SPC25/SPAG5/DSCC1/CENPI/OIP5/HELLS/NCAPD2/CENPA/BUB1/CENPF/ZWILCH/CEBPB/H2BC11&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="2kegg">ï¼ˆ2ï¼‰KEGG<a hidden class="anchor" aria-hidden="true" href="#2kegg">#</a></h3>
<ul>
<li><strong>ORA</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># R.utils::setOption(&#34;clusterProfiler.download.method&#34;,&#34;auto&#34;)</span>
</span></span><span style="display:flex;"><span>ekg &lt;- enrichKEGG(gene         = gene_deg,
</span></span><span style="display:flex;"><span>                  organism     = <span style="color:#0ff;font-weight:bold">&#39;hsa&#39;</span>,
</span></span><span style="display:flex;"><span>                  pvalueCutoff = <span style="color:#ff0;font-weight:bold">0.05</span>)
</span></span><span style="display:flex;"><span>ekg &lt;- setReadable(ekg, OrgDb = org.Hs.eg.db, keyType = <span style="color:#0ff;font-weight:bold">&#34;ENTREZID&#34;</span>)
</span></span><span style="display:flex;"><span>t(ekg@result[1,])
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#              hsa04110                                                        </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ID          &#34;hsa04110&#34;                                                      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Description &#34;Cell cycle&#34;                                                    </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GeneRatio   &#34;11/94&#34;                                                         </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># BgRatio     &#34;127/8218&#34;                                                      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># pvalue      &#34;1.809545e-07&#34;                                                  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># p.adjust    &#34;3.781948e-05&#34;                                                  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># qvalue      &#34;3.695281e-05&#34;                                                  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># geneID      &#34;CDC45/CDC20/CCNB2/CCNA2/CDK1/MAD2L1/TTK/CHEK1/CCNB1/MCM5/PTTG1&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Count       &#34;11&#34; </span>
</span></span><span style="display:flex;"><span>browseKEGG(ekg, <span style="color:#0ff;font-weight:bold">&#39;hsa04110&#39;</span>)
</span></span><span style="display:flex;"><span>library(<span style="color:#0ff;font-weight:bold">&#34;pathview&#34;</span>)
</span></span><span style="display:flex;"><span>pathview(gene.data  = gene_list,
</span></span><span style="display:flex;"><span>         pathway.id = <span style="color:#0ff;font-weight:bold">&#34;hsa04110&#34;</span>,
</span></span><span style="display:flex;"><span>         species    = <span style="color:#0ff;font-weight:bold">&#34;hsa&#34;</span>,
</span></span><span style="display:flex;"><span>         limit      = list(gene=max(abs(geneList)), cpd=<span style="color:#ff0;font-weight:bold">1</span>))
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>å½“å‡ºç°æŠ¥é”™ç±»ä¼¼äºNo gene can be mapped when using enrichKEGGï¼Œä¸”æ£€æŸ¥è¾“å…¥åŸºå› æ— è¯¯åï¼Œä¸€ä¸ªå¯èƒ½çš„è§£å†³æ–¹æ¡ˆæ˜¯è‡ªå·±æ„å»ºæœ¬åœ°KEGGæ•°æ®åº“ã€‚å‚è€ƒhttps://github.com/YuLab-SMU/clusterProfiler/issues/561ä¸­undo6411è§£ç­”ã€‚æˆ–è€…ä¹Ÿå¯ä»¥ç›´æ¥ä»MsigDBæ•°æ®åº“ä¸‹è½½ï¼Œå†ä½¿ç”¨<code>GSEA()</code>å‡½æ•°ã€‚</p></blockquote>
<ul>
<li><strong>GSEA</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>gseakg &lt;- gseKEGG(geneList     = gene_list,
</span></span><span style="display:flex;"><span>                  organism     = <span style="color:#0ff;font-weight:bold">&#39;hsa&#39;</span>,
</span></span><span style="display:flex;"><span>                  minGSSize    = <span style="color:#ff0;font-weight:bold">120</span>,
</span></span><span style="display:flex;"><span>                  pvalueCutoff = <span style="color:#ff0;font-weight:bold">0.05</span>,
</span></span><span style="display:flex;"><span>                  verbose      = <span style="color:#fff;font-weight:bold">FALSE</span>)
</span></span><span style="display:flex;"><span>gseakg &lt;- setReadable(gseakg, OrgDb = org.Hs.eg.db, keyType = <span style="color:#0ff;font-weight:bold">&#34;ENTREZID&#34;</span>)
</span></span><span style="display:flex;"><span>t(gseakg@result[1,])
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>é’ˆå¯¹å…¶å®ƒé€šè·¯é›†çš„å¯Œé›†åˆ†æå‡½æ•°</p>
<table>
  <thead>
      <tr>
          <th style="text-align: center">é€šè·¯é›†</th>
          <th style="text-align: center">ORAå‡½æ•°</th>
          <th style="text-align: center">GSEAå‡½æ•°</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td style="text-align: center">WikiPathway</td>
          <td style="text-align: center">enrichWP</td>
          <td style="text-align: center">gseWP</td>
      </tr>
      <tr>
          <td style="text-align: center">Reactome</td>
          <td style="text-align: center">enrichPathway</td>
          <td style="text-align: center">gsePathway</td>
      </tr>
      <tr>
          <td style="text-align: center">DO(disease ontology)</td>
          <td style="text-align: center">enrichDO</td>
          <td style="text-align: center">gseDO</td>
      </tr>
      <tr>
          <td style="text-align: center">NCG(Network of Cancer Gene)</td>
          <td style="text-align: center">enrichNCG</td>
          <td style="text-align: center">gseNCG</td>
      </tr>
      <tr>
          <td style="text-align: center">DisGeNET(disease gene network)</td>
          <td style="text-align: center">enrichDGN</td>
          <td style="text-align: center">gseDGN</td>
      </tr>
      <tr>
          <td style="text-align: center">Mesh(åŒ»å­¦ä¸»é¢˜è¯)</td>
          <td style="text-align: center">enrichMeSH</td>
          <td style="text-align: center">gseMeSH</td>
      </tr>
  </tbody>
</table></blockquote>
<h3 id="3è‡ªå®šä¹‰åŸºå› é›†">ï¼ˆ3ï¼‰è‡ªå®šä¹‰åŸºå› é›†<a hidden class="anchor" aria-hidden="true" href="#3è‡ªå®šä¹‰åŸºå› é›†">#</a></h3>
<ul>
<li>clusterProfileråŒ…æä¾›çš„<code>enricher()</code>ä¸<code>GSEA()</code>å‡½æ•°å¯å®ç°å¯¹è‡ªå®šä¹‰åŸºå› é›†è¿›è¡Œå¯Œé›†åˆ†æ</li>
<li>ä¸»è¦é€šè¿‡TERM2GENEå‚æ•°æä¾›åŸºå› é›†æ•°æ®æ¡†ï¼Œä¸€åˆ—åä¸ºtermä»£è¡¨é€šè·¯åï¼›ä¸€åˆ—ä¸ºgeneä»£è¡¨ç»„æˆåŸºå› </li>
<li>å¦‚ä¸‹ä»¥MsigDBæ•°æ®åº“ä¸ºä¾‹ï¼Œå­¦ä¹ ç”¨æ³•ã€‚è¯¥åº“çš„è¯¦ç»†ä»‹ç»å‚è§<a href="https://lishensuo.github.io/posts/bioinfo/152%E6%95%B0%E6%8D%AE%E5%BA%93--%E9%80%9A%E8%B7%AF%E5%9F%BA%E5%9B%A0%E9%9B%86msigdb/">ç¬”è®°</a></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># install.packages(&#34;msigdbr&#34;)</span>
</span></span><span style="display:flex;"><span>library(msigdbr)
</span></span><span style="display:flex;"><span>hm_df = msigdbr(species = <span style="color:#0ff;font-weight:bold">&#34;Homo sapiens&#34;</span>) %&gt;% as.data.frame()
</span></span><span style="display:flex;"><span>head(hm_df)
</span></span><span style="display:flex;"><span>colnames(hm_df)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;gs_cat&#34;             &#34;gs_subcat&#34;          &#34;gs_name&#34;            &#34;gene_symbol&#34;       </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;entrez_gene&#34;        &#34;ensembl_gene&#34;       &#34;human_gene_symbol&#34;  &#34;human_entrez_gene&#34; </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [9] &#34;human_ensembl_gene&#34; &#34;gs_id&#34;              &#34;gs_pmid&#34;            &#34;gs_geoid&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [13] &#34;gs_exact_source&#34;    &#34;gs_url&#34;             &#34;gs_description&#34;   </span>
</span></span><span style="display:flex;"><span>unique(hm_df$gs_cat)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;C3&#34; &#34;C2&#34; &#34;C8&#34; &#34;C6&#34; &#34;C7&#34; &#34;C4&#34; &#34;C1&#34; &#34;C5&#34; &#34;H&#34; </span>
</span></span><span style="display:flex;"><span>unique(hm_df$gs_subcat)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;MIR:MIR_Legacy&#34;  &#34;TFT:TFT_Legacy&#34;  &#34;CGP&#34;             &#34;TFT:GTRD&#34;       </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;&#34;                &#34;VAX&#34;             &#34;CP:BIOCARTA&#34;     &#34;CGN&#34;            </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [9] &#34;GO:BP&#34;           &#34;GO:CC&#34;           &#34;IMMUNESIGDB&#34;     &#34;GO:MF&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [13] &#34;HPO&#34;             &#34;CP:KEGG&#34;         &#34;MIR:MIRDB&#34;       &#34;CM&#34;             </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [17] &#34;CP&#34;              &#34;CP:PID&#34;          &#34;CP:REACTOME&#34;     &#34;CP:WIKIPATHWAYS&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>hm_H = msigdbr(species = <span style="color:#0ff;font-weight:bold">&#34;Homo sapiens&#34;</span>,
</span></span><span style="display:flex;"><span>        category = <span style="color:#0ff;font-weight:bold">&#34;H&#34;</span>,
</span></span><span style="display:flex;"><span>        subcategory = <span style="color:#fff;font-weight:bold">NULL</span>) %&gt;% as.data.frame() %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::select(gs_cat, gs_subcat, gs_name, gene_symbol)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>hm_H = hm_H %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::select(gs_name, gene_symbol) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::rename(<span style="color:#0ff;font-weight:bold">&#34;term&#34;</span> = <span style="color:#0ff;font-weight:bold">&#34;gs_name&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;gene&#34;</span> = <span style="color:#0ff;font-weight:bold">&#34;gene_symbol&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>head(hm_H)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                    term  gene</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1 HALLMARK_ADIPOGENESIS ABCA1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2 HALLMARK_ADIPOGENESIS ABCB8</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 HALLMARK_ADIPOGENESIS ACAA2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 HALLMARK_ADIPOGENESIS ACADL</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5 HALLMARK_ADIPOGENESIS ACADM</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6 HALLMARK_ADIPOGENESIS ACADS</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li><strong>å¯Œé›†åˆ†æ</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># æ³¨æ„ä¸Šè¿°é€‰æ‹©çš„æ˜¯Symbolæ ¼å¼åŸºå› ï¼Œè¾“å…¥åŸºå› ä¹Ÿéœ€è¦ä¸ä¹‹åŒ¹é…</span>
</span></span><span style="display:flex;"><span>res_go &lt;- enricher(gene_deg2, TERM2GENE = hm_H)
</span></span><span style="display:flex;"><span>dim(res_go)
</span></span><span style="display:flex;"><span>dim(res_go@result)
</span></span><span style="display:flex;"><span>res_go_df = res_go@result
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>res_gsea &lt;- GSEA(gene_list2, TERM2GENE = hm_H)
</span></span><span style="display:flex;"><span>dim(res_gsea)
</span></span><span style="display:flex;"><span>dim(res_gsea@result)
</span></span><span style="display:flex;"><span>res_gsea_df = res_gsea@result
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="3å¯è§†åŒ–">3ã€å¯è§†åŒ–<a hidden class="anchor" aria-hidden="true" href="#3å¯è§†åŒ–">#</a></h2>
<ul>
<li>ä»¥ä¸Šé¢åˆ†æçš„egoä¸gseagoä¸ºä¾‹</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(enrichplot)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># é™¤ç¬¬ä¸€ä¸ªbarplotå‡½æ•°ï¼Œå…¶ä½™å‡åŸºäºggplotç»˜å›¾ä½“ç³»</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="1æŸ±çŠ¶å›¾ç‚¹å›¾">ï¼ˆ1ï¼‰æŸ±çŠ¶å›¾/ç‚¹å›¾<a hidden class="anchor" aria-hidden="true" href="#1æŸ±çŠ¶å›¾ç‚¹å›¾">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## barplot (RåŸºç¡€å‡½æ•°)</span>
</span></span><span style="display:flex;"><span>barplot(ego, showCategory=<span style="color:#ff0;font-weight:bold">10</span>) 
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># æŒ‡å®šGOé€šè·¯é›†</span>
</span></span><span style="display:flex;"><span>pw2show = c(<span style="color:#0ff;font-weight:bold">&#34;GO:0005819&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;GO:0072686&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;GO:0098687&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;GO:0000779&#34;</span>)
</span></span><span style="display:flex;"><span>barplot(ego, 
</span></span><span style="display:flex;"><span>        showCategory=ego@result$Description[which(rownames(ego@result) %in% pw2show)]) 
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># é’ˆå¯¹ont = &#34;ALL&#34;çš„æƒ…å†µ</span>
</span></span><span style="display:flex;"><span>barplot(ego, split = <span style="color:#0ff;font-weight:bold">&#34;ONTOLOGY&#34;</span>) +
</span></span><span style="display:flex;"><span>  facet_grid(ONTOLOGY~., scale = <span style="color:#0ff;font-weight:bold">&#34;free&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230114080237928.png" alt="image-20230114080237928" style="zoom:50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>dotplot(ego)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## é»˜è®¤å¦‚ä¸‹å¯¹åº”å…³ç³»,å¯è®¾ç½®å‚æ•°ä¿®æ”¹è¿™äº›å…³ç³»ï¼š</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># é¢œè‰²æ˜ å°„På€¼</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># å¤§å°æ˜ å°„äº¤é›†åŸºå› æ•°(å·®å¼‚åŸºå› ä¸é€šè·¯åŸºå› é›†)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># æ¨ªè½´è¡¨ç¤ºæ¯”ä¾‹(count/geneset)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230114081856363.png" alt="image-20230114081856363" style="zoom:50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>edox &lt;- pairwise_termsim(ego)
</span></span><span style="display:flex;"><span>treeplot(edox, cluster.params = list(method=<span style="color:#0ff;font-weight:bold">&#34;ward.D&#34;</span>, n = <span style="color:#ff0;font-weight:bold">4</span>, label_words_n = <span style="color:#ff0;font-weight:bold">2</span>))
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230617155433789.png" alt="image-20230617155433789" style="zoom:40%;" />
<h3 id="2gseaæ‰“åˆ†">ï¼ˆ2ï¼‰GSEAæ‰“åˆ†<a hidden class="anchor" aria-hidden="true" href="#2gseaæ‰“åˆ†">#</a></h3>
<ul>
<li>å•ä¸ªé€šè·¯GSEA</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>gseaplot(gseago, 
</span></span><span style="display:flex;"><span>         geneSetID = <span style="color:#ff0;font-weight:bold">1</span>, 
</span></span><span style="display:flex;"><span>         title = gseago$Description[1])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># by = &#34;preranked&#34;åªå±•ç¤ºä¸ŠåŠéƒ¨åˆ†</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># by = &#34;runningScore&#34;åªå±•ç¤ºä¸‹åŠéƒ¨åˆ†</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230114083100951.png" alt="image-20230114083100951" style="zoom: 33%;" />
<ul>
<li>å¤šæ¡é€šè·¯ç»“æœ</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>gseaplot2(gseago, 
</span></span><span style="display:flex;"><span>          geneSetID = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">3</span>, 
</span></span><span style="display:flex;"><span>          pvalue_table = <span style="color:#fff;font-weight:bold">T</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230114083247332.png" alt="image-20230114083247332" style="zoom:33%;" />
<h3 id="3é€šè·¯ä¸åŸºå› ">ï¼ˆ3ï¼‰é€šè·¯ä¸åŸºå› <a hidden class="anchor" aria-hidden="true" href="#3é€šè·¯ä¸åŸºå› ">#</a></h3>
<ul>
<li>çƒ­å›¾</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>heatplot(ego, 
</span></span><span style="display:flex;"><span>         foldChange=gene_list,
</span></span><span style="display:flex;"><span>         showCategory = <span style="color:#ff0;font-weight:bold">5</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230114083736463.png" alt="image-20230114083736463" style="zoom: 50%;" />
<ul>
<li>ç½‘ç»œå›¾</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(patchwork)
</span></span><span style="display:flex;"><span>p1 = cnetplot(ego, foldChange=gene_list,showCategory = <span style="color:#ff0;font-weight:bold">3</span>)
</span></span><span style="display:flex;"><span>p2 = cnetplot(ego, foldChange=gene_list,showCategory = <span style="color:#ff0;font-weight:bold">3</span>, circular = <span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>p3 = cnetplot(ego, foldChange=gene_list,showCategory = <span style="color:#ff0;font-weight:bold">3</span>, colorEdge = <span style="color:#fff;font-weight:bold">T</span>, node_label=<span style="color:#0ff;font-weight:bold">&#34;gene&#34;</span>)
</span></span><span style="display:flex;"><span>p4 = cnetplot(ego, foldChange=gene_list,showCategory = <span style="color:#ff0;font-weight:bold">3</span>, layout = <span style="color:#0ff;font-weight:bold">&#34;gem&#34;</span>)
</span></span><span style="display:flex;"><span>(p1 + p2) / (p3 + p4)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230114083957103.png" alt="image-20230114083957103" style="zoom: 33%;" />
<h3 id="4é€šè·¯ä¸é€šè·¯">ï¼ˆ4ï¼‰é€šè·¯ä¸é€šè·¯<a hidden class="anchor" aria-hidden="true" href="#4é€šè·¯ä¸é€šè·¯">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>ego_pair &lt;- enrichplot::pairwise_termsim(ego)
</span></span><span style="display:flex;"><span>emapplot(ego_pair, showCategory = <span style="color:#ff0;font-weight:bold">15</span>, layout=<span style="color:#0ff;font-weight:bold">&#34;kk&#34;</span>, cex_category=<span style="color:#ff0;font-weight:bold">1.5</span>,min_edge = <span style="color:#ff0;font-weight:bold">0.8</span>) 
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gseago_pair &lt;- enrichplot::pairwise_termsim(gseago)
</span></span><span style="display:flex;"><span>emapplot(gseago_pair, showCategory = <span style="color:#ff0;font-weight:bold">15</span>, layout=<span style="color:#0ff;font-weight:bold">&#34;kk&#34;</span>, min_edge = <span style="color:#ff0;font-weight:bold">0.1</span>, color = <span style="color:#0ff;font-weight:bold">&#34;NES&#34;</span>) 
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230114084348140.png" alt="image-20230114084348140" style="zoom: 50%;" />
<h2 id="4å¤šåŸºå› é›†åˆ†æ">4ã€å¤šåŸºå› é›†åˆ†æ<a hidden class="anchor" aria-hidden="true" href="#4å¤šåŸºå› é›†åˆ†æ">#</a></h2>
<ul>
<li>clusterProfileråŒ…çš„<code>compareCluster()</code>å‡½æ•°æä¾›äº†é’ˆå¯¹å¤šç»„å®éªŒè®¾è®¡å¾—åˆ°çš„è‹¥å¹²å·®å¼‚åŸºå› åˆ—è¡¨çš„åŒæ—¶å¯Œé›†åˆ†æ(ORA)ç»“æœï¼ŒåŠå‹å¥½çš„å¯è§†åŒ–ã€‚</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## (1) ç›´æ¥å¤šåŸºå› é›†çš„å¯Œé›†é€šè·¯æ¯”è¾ƒ</span>
</span></span><span style="display:flex;"><span>data(gcSample)
</span></span><span style="display:flex;"><span>str(gcSample[5:<span style="color:#ff0;font-weight:bold">8</span>]) <span style="color:#007f7f">#å–å4ä¸ªå·®å¼‚åŸºå› åˆ—è¡¨åšæ¼”ç¤º</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># List of 4</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># $ X5: chr [1:929] &#34;5982&#34; &#34;7318&#34; &#34;6352&#34; &#34;2101&#34; ...</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># $ X6: chr [1:585] &#34;5337&#34; &#34;9295&#34; &#34;4035&#34; &#34;811&#34; ...</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># $ X7: chr [1:582] &#34;2621&#34; &#34;2665&#34; &#34;5690&#34; &#34;3608&#34; ...</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># $ X8: chr [1:237] &#34;2665&#34; &#34;4735&#34; &#34;1327&#34; &#34;3192&#34; ...</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## ä¸¤ä¸ªä¸»è¦å‚æ•°</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># å¤šä¸ªå·®å¼‚åŸºå› åˆ—è¡¨</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># å¯Œé›†åˆ†æmethod (One of &#34;groupGO&#34;, &#34;enrichGO&#34;, &#34;enrichKEGG&#34;, &#34;enrichDO&#34; or &#34;enrichPathway&#34;)</span>
</span></span><span style="display:flex;"><span>ck &lt;- compareCluster(geneCluster = gcSample[5:<span style="color:#ff0;font-weight:bold">8</span>], fun = enrichKEGG)
</span></span><span style="display:flex;"><span>ck &lt;- setReadable(ck, OrgDb = org.Hs.eg.db, keyType=<span style="color:#0ff;font-weight:bold">&#34;ENTREZID&#34;</span>)
</span></span><span style="display:flex;"><span>table(ck@compareClusterResult$Cluster)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># X5 X6 X7 X8 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 10  1 15 19</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>dotplot(ck)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-c61a485f495595e7.png" alt="img" style="zoom:33%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## (2) å‡å¦‚åˆ†åˆ«ç”¨ä¸¤ç§è¯ç‰©Aã€Bè¿›è¡Œå¹²æ‰°ç»†èƒç³»ï¼Œç»1hã€3håå¾—åˆ°4ç»„å·®å¼‚åŸºå› </span>
</span></span><span style="display:flex;"><span>group_A_1h=gcSample[[2]]
</span></span><span style="display:flex;"><span>group_A_3h=gcSample[[3]]
</span></span><span style="display:flex;"><span>group_B_1h=gcSample[[4]]
</span></span><span style="display:flex;"><span>group_B_3h=gcSample[[5]]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>design=c(rep(<span style="color:#0ff;font-weight:bold">&#34;A_1h&#34;</span>,length(group_A_1h)),
</span></span><span style="display:flex;"><span>          rep(<span style="color:#0ff;font-weight:bold">&#34;A_3h&#34;</span>,length(group_A_3h)),
</span></span><span style="display:flex;"><span>          rep(<span style="color:#0ff;font-weight:bold">&#34;B_1h&#34;</span>,length(group_B_1h)),
</span></span><span style="display:flex;"><span>          rep(<span style="color:#0ff;font-weight:bold">&#34;B_3h&#34;</span>,length(group_B_3h)))
</span></span><span style="display:flex;"><span>mydf = data.frame(Drug = stringr::str_split(design,<span style="color:#0ff;font-weight:bold">&#34;_&#34;</span>,simplify = <span style="color:#fff;font-weight:bold">T</span>)[,<span style="color:#ff0;font-weight:bold">1</span>],
</span></span><span style="display:flex;"><span>                  Time = stringr::str_split(design,<span style="color:#0ff;font-weight:bold">&#34;_&#34;</span>,simplify = <span style="color:#fff;font-weight:bold">T</span>)[,<span style="color:#ff0;font-weight:bold">2</span>],
</span></span><span style="display:flex;"><span>                  DEG = c(group_A_1h,group_A_3h,group_B_1h,group_B_3h))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   Drug Time   DEG</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1    A   1h 23450</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2    A   1h  5160</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3    A   1h  7126</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>formula_res &lt;- compareCluster(DEG~Time+Drug, data=mydf, fun=<span style="color:#0ff;font-weight:bold">&#34;enrichKEGG&#34;</span>)
</span></span><span style="display:flex;"><span>table(formula_res@compareClusterResult$Cluster)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1h.A 1h.B 3h.A 3h.B </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   3   18    3   10</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>dotplot(formula_res) + 
</span></span><span style="display:flex;"><span>  dotplot(formula_res, x=<span style="color:#0ff;font-weight:bold">&#34;Time&#34;</span>) + facet_grid(~Drug)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-e31c156de2160992.png" alt="img" style="zoom:50%;" />


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">è½¬å½•ç»„</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/055gsea%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7/">
    <span class="title">Â« Prev Page</span>
    <br>
    <span>GSEAå¯Œé›†åˆ†æå·¥å…·</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/057%E5%AF%8C%E9%9B%86%E9%80%9A%E8%B7%AF%E5%8F%AF%E8%A7%86%E5%8C%96%E6%96%B9%E5%BC%8F/">
    <span class="title">Next Page Â»</span>
    <br>
    <span>å¯Œé›†é€šè·¯å¯è§†åŒ–æ–¹å¼</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>æ‚¨æ˜¯æœ¬ç«™ç¬¬ <span id="busuanzi_value_site_uv"></span> ä½è®¿é—®è€…ï¼Œæ€»æµè§ˆé‡ä¸º <span id="busuanzi_value_site_pv"></span> æ¬¡
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
