<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>人类基因组基础知识与下载查询 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="生信基础">
<meta name="description" content="

一、基因组大小
（1）人类基因组主要由细胞核的23对染色体组成(核基因组)，还包括线粒体中的小DNA分子(线粒体基因组)。
（2）单倍体基因组大概有30亿个碱基对组成，具体到每个染色体的碱基对长度与基因数量如下所示(参照UCSC的hg38)。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/053%E4%BA%BA%E7%B1%BB%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%9F%BA%E7%A1%80%E7%9F%A5%E8%AF%86%E4%B8%8E%E4%B8%8B%E8%BD%BD%E6%9F%A5%E8%AF%A2/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/053%E4%BA%BA%E7%B1%BB%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%9F%BA%E7%A1%80%E7%9F%A5%E8%AF%86%E4%B8%8E%E4%B8%8B%E8%BD%BD%E6%9F%A5%E8%AF%A2/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="人类基因组基础知识与下载查询" />
<meta property="og:description" content="

一、基因组大小
（1）人类基因组主要由细胞核的23对染色体组成(核基因组)，还包括线粒体中的小DNA分子(线粒体基因组)。
（2）单倍体基因组大概有30亿个碱基对组成，具体到每个染色体的碱基对长度与基因数量如下所示(参照UCSC的hg38)。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/053%E4%BA%BA%E7%B1%BB%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%9F%BA%E7%A1%80%E7%9F%A5%E8%AF%86%E4%B8%8E%E4%B8%8B%E8%BD%BD%E6%9F%A5%E8%AF%A2/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-05-28T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2022-05-28T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="人类基因组基础知识与下载查询"/>
<meta name="twitter:description" content="

一、基因组大小
（1）人类基因组主要由细胞核的23对染色体组成(核基因组)，还包括线粒体中的小DNA分子(线粒体基因组)。
（2）单倍体基因组大概有30亿个碱基对组成，具体到每个染色体的碱基对长度与基因数量如下所示(参照UCSC的hg38)。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "人类基因组基础知识与下载查询",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/053%E4%BA%BA%E7%B1%BB%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%9F%BA%E7%A1%80%E7%9F%A5%E8%AF%86%E4%B8%8E%E4%B8%8B%E8%BD%BD%E6%9F%A5%E8%AF%A2/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "人类基因组基础知识与下载查询",
  "name": "人类基因组基础知识与下载查询",
  "description": "\r一、基因组大小 （1）人类基因组主要由细胞核的23对染色体组成(核基因组)，还包括线粒体中的小DNA分子(线粒体基因组)。\n（2）单倍体基因组大概有30亿个碱基对组成，具体到每个染色体的碱基对长度与基因数量如下所示(参照UCSC的hg38)。\n",
  "keywords": [
    "生信基础"
  ],
  "articleBody": "\r一、基因组大小 （1）人类基因组主要由细胞核的23对染色体组成(核基因组)，还包括线粒体中的小DNA分子(线粒体基因组)。\n（2）单倍体基因组大概有30亿个碱基对组成，具体到每个染色体的碱基对长度与基因数量如下所示(参照UCSC的hg38)。\n（3）在这30亿个碱基中，仅有1.5%的区域是2~3w个可编码蛋白质的基因。\n平均基因长度有10Kbp左右，但是不同基因的长度区别很大 其余98.5%的区域为非编码区，包括各种调控基因表达的DNA原件，同时也可以转录为non-coding RNA，发挥潜在作用。 二、基因组fasta文件 人类基因组序列信息通常以染色体为单位保存为fasta文件中。\n（1）参考基因组一般保存为纯文本格式，即直接记录“A”、“T”、“C”、“G”这样的 ASCII 码字符。而1个 ASCII 字符，大小是 1B，所以，如果按纯文本保存 30亿个字母(单链)，就是30亿字母 = 3,000,000,000 B = 3 GB。\n（2）理论上应该只有25条序列信息(Chr1:22,X,Y,ChrM)。但是在实际下载的文件中，序列数量远远不止这些。如下以UCSC里的hg38版本为例，有455条序列信息。\n除了上述25条序列外，其它序列主要是scaffolds，可大致分为ChrUn, Random, Alt三类。\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz gunzip hg38.fa.gz #提取染色体id grep \"^\u003e\" hg38.fa \u003e chr.id wc -l chr.id #455 chr.id head chr.id # \u003echr1 # \u003echr10 # \u003echr11 # \u003echr11_KI270721v1_random # \u003echr12 # \u003echr13 # \u003echr14 # \u003echr14_GL000009v2_random # \u003echr14_GL000225v1_random # \u003echr14_KI270722v1_random 人类基因组主要通过二代测序技术获得，将测得的短Read片段拼接、组装成基因组的染色体序列，需要经历contigs与scaffolds两个过程\n（1）contigs是依靠read间的重叠拼接的序列(a few kbp long)，特点是不含有N碱基；\n（2）scaffolds则主要依靠read pairs关系进一步拼接contigs，特点是会产生N碱基(a few hundred kbp)；\n（3）最终再由scaffolds拼接成染色体序列。\n2.1 Unlocalized scaffolds(*****random) 知道这些scaffolds在哪条染色体上，但不知道其在染色体的具体位置及方向\nformat: chr{chromosome number orname}_{sequence_accession}v{sequence_version}_random\n1 2 3 4 5 6 7 8 9 10 grep \"random\" chr.id \u003e chr.random wc -l chr.random #42 chr.random head chr.random # \u003echr11_KI270721v1_random # \u003echr14_GL000009v2_random # \u003echr14_GL000225v1_random # \u003echr14_KI270722v1_random # \u003echr14_GL000194v1_random # \u003echr14_KI270723v1_random 2.2 Unplaced scaffolds(chrUn******) 不知道这条scaffolds的所属染色体信息 format: chrUn_{sequence_accession}v{sequence_version} 1 2 3 4 5 6 7 8 9 10 grep \"chrUn\" chr.id \u003e chr.chrUn wc -l chr.chrUn #127 chr.chrUn head chr.chrUn # \u003echrUn_KI270302v1 # \u003echrUn_KI270304v1 # \u003echrUn_KI270303v1 # \u003echrUn_KI270305v1 # \u003echrUn_KI270322v1 # \u003echrUn_KI270320v1 2.3 Alternate loci scaffolds(*****alt) 简单理解：参考基因组存在的主要依据是人类99.9%的序列是一致的。但是会存在一些序列在不同人群中不一致。例如49%人群该基因组特定位置为序列A，而49%人群则为序列B，都是正常的。但拿其中一种作为参考基因组都可能不太合适，因此标记出Alternate loci scaffolds。 format: chr{chromosome number or name}_{sequence_accession}v{sequence_version}_alt Alternate loci scaffolds为hg38版本基因组新添类型Sequence，此前hg19版本还没有。 1 2 3 4 5 6 7 8 9 10 grep \"alt\" chr.id \u003e chr.alt wc -l chr.alt #261 chr.alt head chr.alt # \u003echr1_KI270762v1_alt # \u003echr1_KI270766v1_alt # \u003echr1_KI270760v1_alt # \u003echr1_KI270765v1_alt # \u003echr1_GL383518v1_alt # \u003echr1_GL383519v1_alt 注意：以上具体的chromosome name均为ucsc的hg版本，与GRCh38略有差异，但基本也是这几种类型sequence。\n三、基因结构 1、编码区与非编码区 基因 DNA可分为编码区和非编码区：编码区可转录为 mRNA 并最终翻译成蛋白质；非编码区上具有基因表达的调控元件。\n（1）编码区：在转录的过程，从DNA编码区的5’端开始转录生成preRNA，然后进一步加工修饰剪切得到成熟mRNA，进行后续的翻译。\n外显子：在preRNA 经过剪切或修饰后、被保留的DNA部分。 起始密码子(Start Codon)：是mRNA上开始合成蛋白质的密码子，也是第一个被核糖体翻译的mRNA密码子。通常为AUG。 终止密码子(Stop Codon)：代表核糖体翻译的终止。通常为UAG、UAA、UGA。 UTR(untranslated region)：属于外显子部分。即外显子也包含非编码区域。 位于5’端的UTR称为5’UTR，如下图橙色标注，从mRNA起点的甲基化鸟嘌呤核苷酸帽延伸至AUG起始密码子。 与5’UTR上游的第一个碱基相对应DNA链上的碱基称为TSS(Transcription start sites) 位于3’端的UTR称为3’UTR，为mRNA的结束部分，如下图红色标注，从编码区末端的终止密码子延伸至多聚A尾巴（Poly-A）。 与3’UTR下游的最后一个碱基相对应DNA链上的碱基称为TTS(Transcription termination sites) CDS(Coding sequence)：包括mRNA中从5’UTR后的起始密码子开始到3’UTR前的终止密码子的实际编码蛋白序列 ORF(Open reading frame)：从一个起始密码子开始到一个终止密码子结束的一段序列 注意与CDS定义上的区别。CDS一定是ORF，ORF不一定是CDS Transcript转录本：一条基因通过可变剪切机制可转录形成的一种或多种可供编码蛋白质的成熟的mRNA。 在最终成熟mRNA的上下游两端修饰有特殊的结构，分别是5’ cap帽子与 3’ Poly-A尾巴。 Poly-A尾是mRNA’区别于其它non-coding RNA的主要标志 （2）非编码区：位于DNA序列上下游的non-coding区域含有调控基因表达原件\n启动子Promoter：DNA分子上能与RNA聚合酶结合并形成转录起始复合体的区域，位于TSS位点上游的100~1000碱基序列。 真核生物启动子通常会包括一个TATA盒，比较接近TSS区域(50个碱基对以内) 终止子Terminator：给RNA聚合酶提供转录终止信号的DNA序列 增强子Enhancer：可以蛋白质结合的小段DNA，从而加强所调控基因的转录作用。可以位于靶基因的上游/下游，距离也可以很远。 2、基因组注释文件gtf 第一列：染色体ID 第三列：region类型 第四列：region的起始位点坐标（染色体的第一个碱基序号为1） 第五列：region的终止位点坐标 第七列：正负链信息（+forward/-reverse） 正链：表示实际mRNA序列与坐标对应DNA序列相同，除了（U/T变换） 负链：表示实际mRNA序列与坐标对应DNA序列的互补链相同，除了（U/T变换），也需要从5’到3’端读取。 第八列：对于实际编码区coding region，表示在这个CDS区域开始编码新密码子的起始index{0,1,2} 第九列：region的属性键值对。 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 # wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.refGene.gtf.gz gtf = data.table::fread(\"hg38.refGene.gtf.gz\") head(gtf,3) # V1 V2 V3 V4 V5 V6 V7 V8 # 1: chr1 refGene transcript 11874 14409 . + . # 2: chr1 refGene exon 11874 12227 . + . # 3: chr1 refGene exon 12613 12721 . + . # V9 # 1: gene_id \"DDX11L1\"; transcript_id \"NR_046018\"; gene_name \"DDX11L1\"; # 2: gene_id \"DDX11L1\"; transcript_id \"NR_046018\"; exon_number \"1\"; exon_id \"NR_046018.1\"; gene_name \"DDX11L1\"; # 3: gene_id \"DDX11L1\"; transcript_id \"NR_046018\"; exon_number \"2\"; exon_id \"NR_046018.2\"; gene_name \"DDX11L1\"; table(gtf$V3) # 3UTR 5UTR CDS exon start_codon stop_codon transcript # 67901 121178 655401 836541 64026 64026 84787 ## 一般来说，每个exon region都对应一个CDS region，除了假基因Pseudogenes。 不同平台机构给出的gtf注释文件略有差异，但每列含义基本相同。发现NCBI提供的gff3文件注释信息最全，包括非编码RNA，蛋白质编码RNA、假基因等\n1 2 3 4 5 6 7 8 9 library(AnnoProbe) IDs \u003c- c(\"DDX11L1\", \"MIR6859-1\", \"OR4G4P\", \"OR4F5\") ID_type = \"SYMBOL\" annoGene(IDs, ID_type) # SYMBOL biotypes ENSEMBL chr start end # 1 DDX11L1 transcribed_unprocessed_pseudogene ENSG00000223972 chr1 11869 14409 # 3 MIR6859-1 miRNA ENSG00000278267 chr1 17369 17436 # 7 OR4G4P unprocessed_pseudogene ENSG00000268020 chr1 52473 53312 # 9 OR4F5 protein_coding ENSG00000186092 chr1 65419 71585 四、下载参考基因组 目前常用的基因组版本为GRCh38/37，hg38/19，前者可通过NCBI/Ensembl下载，后者可通过UCSC网站下载。\n如下图所示GRCh38可认为等同于hg38；GRCh37可认为等同于hg19。\n以下载GRCh38/hg38为例，如下\n4.1 NCBI https://www.ncbi.nlm.nih.gov/projects/genome/guide/human/ 1 wget -c ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_genomic.fna.gz 4.2 ensembl ensembl的release-103版本可以认为等于GRCh38 https://asia.ensembl.org/Homo_sapiens/Info/Index http://ftp.ensembl.org/pub/release-103/fasta/homo_sapiens/dna/ https://www.ensembl.org/info/data/ftp/index.html 还包含有cDNA转录本序列 1 wget -c http://ftp.ensembl.org/pub/release-103/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.toplevel.fa.gz 4.3 UCSC http://hgdownload.soe.ucsc.edu/downloads.html http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/ 1 wget -c http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz 也可以使用之前学习的Refgenie工具下载上述提到的gtf注释文件\n五、TxDb与GRanges对象查询gtf TxDb系列包是将物种基因组的gtf注释信息封装为TxDb对象，以供使用者方便的查询。\n1 2 3 4 5 6 7 8 9 10 11 12 13 #BiocManager::install(\"TxDb.Hsapiens.UCSC.hg38.knownGene\") library(TxDb.Hsapiens.UCSC.hg38.knownGene) txdb \u003c- TxDb.Hsapiens.UCSC.hg38.knownGene class(txdb) # [1] \"TxDb\" # attr(,\"package\") # [1] \"GenomicFeatures\" seqlevels(txdb) \u003c- paste0(\"chr\",c(1:22,\"X\",\"Y\",\"M\")) seqlevels(txdb) # [1] \"chr1\" \"chr2\" \"chr3\" \"chr4\" \"chr5\" \"chr6\" \"chr7\" \"chr8\" \"chr9\" \"chr10\" \"chr11\" # [12] \"chr12\" \"chr13\" \"chr14\" \"chr15\" \"chr16\" \"chr17\" \"chr18\" \"chr19\" \"chr20\" \"chr21\" \"chr22\" # [23] \"chrX\" \"chrY\" \"chrM\" #seqlevels(txdb) \u003c- seqlevels0(txdb) txdb对象支持4种region fearture，共计22种的属性注释。\n具体使用时可提供keytypes() 所支持的7种信息之一，查询所匹配的其它（22种内）注释信息。\n其中GENEID为Entrez基因ID格式，TXID为Ensemble转录本ID格式\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 columns(txdb) # Gene : \"GENEID\" # CDS : \"CDSCHROM\" \"CDSEND\" \"CDSID\" \"CDSNAME\" \"CDSPHASE\" \"CDSSTART\" \"CDSSTRAND\" # EXON : \"EXONCHROM\" \"EXONEND\" \"EXONID\" \"EXONNAME\" \"EXONRANK\" \"EXONSTART\" \"EXONSTRAND\" # TX : \"TXCHROM\" \"TXEND\" \"TXID\" \"TXNAME\" \"TXSTART\" \"TXSTRAND\" keytypes(txdb) # [1] \"GENEID\" \"CDSID\" \"CDSNAME\" \"EXONID\" \"EXONNAME\" \"TXID\" \"TXNAME\" keys = \"348\" #APOE AnnotationDbi::select(txdb, keys = keys, keytype=\"GENEID\", columns=columns(txdb)) AnnotationDbi::select(txdb, keys = keys, keytype=\"GENEID\", columns=c(\"TXID\",\"TXSTART\",\"TXEND\")) # 'select()' returned 1:many mapping between keys and columns # GENEID TXID TXSTART TXEND # 1 348 201573 44905791 44908944 # 2 348 201574 44905796 44907326 # 3 348 201575 44905796 44909393 # 4 348 201576 44905812 44909025 # 5 348 201577 44906360 44908954 GenomicFeatures包提供了genes(),transcripts(), exons(), cds(), promoters()函数用于提取特定feature region为GRange对象。\n其中可通过filter=list()参数进行筛选符合注释信息的region，可包括\n“gene_id”, “tx_id”, “tx_name”, “tx_chrom”, “tx_strand”, “exon_id”, “exon_name”, “exon_chrom”, “exon_strand”, “exon_rank”. “cds_id”, “cds_name”, “cds_chrom”, “cds_strand”\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 GR = transcripts(txdb, filter=list(gene_id=\"348\")) GR$tx_id # GRanges object with 5 ranges and 2 metadata columns: # seqnames ranges strand | tx_id tx_name # | # [1] chr19 44905791-44908944 + | 201573 ENST00000446996.5 # [2] chr19 44905796-44907326 + | 201574 ENST00000485628.2 # [3] chr19 44905796-44909393 + | 201575 ENST00000252486.9 # [4] chr19 44905812-44909025 + | 201576 ENST00000434152.5 # [5] chr19 44906360-44908954 + | 201577 ENST00000425718.1 # ------- # seqinfo: 25 sequences (1 circular) from hg38 genome ##相关操作 ranges(GR) width(GR) start(GR) end(GR) mcols(GR) GR$tx_id GR[strand(GR) == \"+\"] GR[1:4] GRange对象结合BSgenome系列包获取指定feature region的序列信息\n1 2 3 4 library(BSgenome.Hsapiens.UCSC.hg38) BSgenome = BSgenome.Hsapiens.UCSC.hg38 GR.seq \u003c- getSeq(BSgenome,GR) GR.seq_df = as.data.frame(GR.seq) ",
  "wordCount" : "3802",
  "inLanguage": "en",
  "datePublished": "2022-05-28T00:00:00Z",
  "dateModified": "2022-05-28T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/053%E4%BA%BA%E7%B1%BB%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%9F%BA%E7%A1%80%E7%9F%A5%E8%AF%86%E4%B8%8E%E4%B8%8B%E8%BD%BD%E6%9F%A5%E8%AF%A2/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      人类基因组基础知识与下载查询
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2022-05-28 00:00:00 &#43;0000 UTC&#39;&gt;2022-05-28&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-05-28&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;3802&amp;nbsp;|&amp;nbsp;8 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#%e4%b8%80%e5%9f%ba%e5%9b%a0%e7%bb%84%e5%a4%a7%e5%b0%8f" aria-label="一、基因组大小">一、基因组大小</a></li>
                    <li>
                        <a href="#%e4%ba%8c%e5%9f%ba%e5%9b%a0%e7%bb%84fasta%e6%96%87%e4%bb%b6" aria-label="二、基因组fasta文件">二、基因组fasta文件</a><ul>
                            <ul>
                            
                    <li>
                        <a href="#21-unlocalized-scaffoldsrandom" aria-label="2.1 Unlocalized scaffolds(*****random)">2.1 Unlocalized scaffolds(*****random)</a></li>
                    <li>
                        <a href="#22-unplaced-scaffoldschrun" aria-label="2.2 Unplaced scaffolds(chrUn******)">2.2 Unplaced scaffolds(chrUn******)</a></li>
                    <li>
                        <a href="#23--alternate-loci-scaffoldsalt" aria-label="2.3  Alternate loci scaffolds(*****alt)">2.3  Alternate loci scaffolds(*****alt)</a></li></ul>
                        </ul>
                    </li>
                    <li>
                        <a href="#%e4%b8%89%e5%9f%ba%e5%9b%a0%e7%bb%93%e6%9e%84" aria-label="三、基因结构">三、基因结构</a><ul>
                            
                    <li>
                        <a href="#1%e7%bc%96%e7%a0%81%e5%8c%ba%e4%b8%8e%e9%9d%9e%e7%bc%96%e7%a0%81%e5%8c%ba" aria-label="1、编码区与非编码区">1、编码区与非编码区</a></li>
                    <li>
                        <a href="#2%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b3%a8%e9%87%8a%e6%96%87%e4%bb%b6gtf" aria-label="2、基因组注释文件gtf">2、基因组注释文件gtf</a></li></ul>
                    </li>
                    <li>
                        <a href="#%e5%9b%9b%e4%b8%8b%e8%bd%bd%e5%8f%82%e8%80%83%e5%9f%ba%e5%9b%a0%e7%bb%84" aria-label="四、下载参考基因组">四、下载参考基因组</a><ul>
                            
                    <li>
                        <a href="#41-ncbi" aria-label="4.1 NCBI">4.1 NCBI</a></li>
                    <li>
                        <a href="#42-ensembl" aria-label="4.2 ensembl">4.2 ensembl</a></li>
                    <li>
                        <a href="#43-ucsc" aria-label="4.3 UCSC">4.3 UCSC</a></li></ul>
                    </li>
                    <li>
                        <a href="#%e4%ba%94txdb%e4%b8%8egranges%e5%af%b9%e8%b1%a1%e6%9f%a5%e8%af%a2gtf" aria-label="五、TxDb与GRanges对象查询gtf">五、<code>TxDb</code>与<code>GRanges</code>对象查询gtf</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><p><img loading="lazy" src="https://www.edinformatics.com/math_science/human_chromosomes2.jpg" alt="How many genes are in the human genome?"  />
</p>
<h1 id="一基因组大小">一、基因组大小<a hidden class="anchor" aria-hidden="true" href="#一基因组大小">#</a></h1>
<p>（1）人类基因组主要由细胞核的23对染色体组成(核基因组)，还包括线粒体中的小DNA分子(线粒体基因组)。</p>
<p>（2）单倍体基因组大概有30亿个碱基对组成，具体到每个染色体的碱基对长度与基因数量如下所示(参照UCSC的hg38)。</p>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220528084330069.png" alt="image-20220528084330069" style="zoom: 50%;" />
<p>（3）在这30亿个碱基中，仅有1.5%的区域是2~3w个可编码蛋白质的基因。</p>
<ul>
<li>平均基因长度有10Kbp左右，但是不同基因的长度区别很大</li>
<li>其余98.5%的区域为非编码区，包括各种调控基因表达的DNA原件，同时也可以转录为non-coding RNA，发挥潜在作用。</li>
</ul>
<h1 id="二基因组fasta文件">二、基因组fasta文件<a hidden class="anchor" aria-hidden="true" href="#二基因组fasta文件">#</a></h1>
<p>人类基因组序列信息通常以染色体为单位保存为fasta文件中。</p>
<p>（1）参考基因组一般保存为纯文本格式，即直接记录“A”、“T”、“C”、“G”这样的 ASCII 码字符。而1个 ASCII 字符，大小是 1B，所以，如果按纯文本保存 30亿个字母(单链)，就是30亿字母 = 3,000,000,000 B = 3 GB。</p>
<p>（2）理论上应该只有25条序列信息(Chr1:22,X,Y,ChrM)。但是在实际下载的文件中，序列数量远远不止这些。如下以UCSC里的hg38版本为例，有455条序列信息。</p>
<p>除了上述25条序列外，其它序列主要是scaffolds，可大致分为<strong>ChrUn, Random, Alt</strong>三类。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
</span></span><span style="display:flex;"><span>gunzip hg38.fa.gz
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#提取染色体id</span>
</span></span><span style="display:flex;"><span>grep <span style="color:#0ff;font-weight:bold">&#34;^&gt;&#34;</span> hg38.fa &gt; chr.id
</span></span><span style="display:flex;"><span>wc -l chr.id
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#455 chr.id</span>
</span></span><span style="display:flex;"><span>head chr.id
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr10</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr11</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr11_KI270721v1_random</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr12</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr13</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr14</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr14_GL000009v2_random</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr14_GL000225v1_random</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr14_KI270722v1_random</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>人类基因组主要通过二代测序技术获得，将测得的短Read片段拼接、组装成基因组的染色体序列，需要经历contigs与scaffolds两个过程</p>
<ul>
<li>
<p>（1）contigs是依靠read间的重叠拼接的序列(a few kbp long)，特点是不含有N碱基；</p>
</li>
<li>
<p>（2）scaffolds则主要依靠read pairs关系进一步拼接contigs，特点是会产生N碱基(a few hundred kbp)；</p>
</li>
<li>
<p>（3）最终再由scaffolds拼接成染色体序列。</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-5332ad63e19b3a8a.png" alt="read→chromosomes"  />
</p>
</li>
</ul>
<h3 id="21-unlocalized-scaffoldsrandom">2.1 Unlocalized scaffolds(*****random)<a hidden class="anchor" aria-hidden="true" href="#21-unlocalized-scaffoldsrandom">#</a></h3>
<ul>
<li>
<p>知道这些scaffolds在哪条染色体上，但不知道其在染色体的具体位置及方向</p>
</li>
<li>
<p>format: chr{chromosome number orname}_{sequence_accession}v{sequence_version}_random</p>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>grep <span style="color:#0ff;font-weight:bold">&#34;random&#34;</span> chr.id &gt; chr.random
</span></span><span style="display:flex;"><span>wc -l chr.random
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#42 chr.random</span>
</span></span><span style="display:flex;"><span>head chr.random
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr11_KI270721v1_random</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr14_GL000009v2_random</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr14_GL000225v1_random</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr14_KI270722v1_random</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr14_GL000194v1_random</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr14_KI270723v1_random</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="22-unplaced-scaffoldschrun">2.2 Unplaced scaffolds(chrUn******)<a hidden class="anchor" aria-hidden="true" href="#22-unplaced-scaffoldschrun">#</a></h3>
<ul>
<li>不知道这条scaffolds的所属染色体信息</li>
<li>format: chrUn_{sequence_accession}v{sequence_version}</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>grep <span style="color:#0ff;font-weight:bold">&#34;chrUn&#34;</span> chr.id &gt; chr.chrUn
</span></span><span style="display:flex;"><span>wc -l chr.chrUn
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#127 chr.chrUn</span>
</span></span><span style="display:flex;"><span>head chr.chrUn
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chrUn_KI270302v1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chrUn_KI270304v1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chrUn_KI270303v1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chrUn_KI270305v1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chrUn_KI270322v1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chrUn_KI270320v1</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="23--alternate-loci-scaffoldsalt">2.3  Alternate loci scaffolds(*****alt)<a hidden class="anchor" aria-hidden="true" href="#23--alternate-loci-scaffoldsalt">#</a></h3>
<ul>
<li>简单理解：参考基因组存在的主要依据是人类99.9%的序列是一致的。但是会存在一些序列在不同人群中不一致。例如49%人群该基因组特定位置为序列A，而49%人群则为序列B，都是正常的。但拿其中一种作为参考基因组都可能不太合适，因此标记出Alternate loci scaffolds。</li>
<li>format: chr{chromosome number or name}_{sequence_accession}v{sequence_version}_alt</li>
<li>Alternate loci scaffolds为hg38版本基因组新添类型Sequence，此前hg19版本还没有。</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>grep <span style="color:#0ff;font-weight:bold">&#34;alt&#34;</span> chr.id &gt; chr.alt
</span></span><span style="display:flex;"><span>wc -l chr.alt
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#261 chr.alt</span>
</span></span><span style="display:flex;"><span>head chr.alt
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr1_KI270762v1_alt</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr1_KI270766v1_alt</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr1_KI270760v1_alt</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr1_KI270765v1_alt</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr1_GL383518v1_alt</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &gt;chr1_GL383519v1_alt</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>注意：以上具体的chromosome name均为ucsc的hg版本，与GRCh38略有差异，但基本也是这几种类型sequence。</p></blockquote>
<h1 id="三基因结构">三、基因结构<a hidden class="anchor" aria-hidden="true" href="#三基因结构">#</a></h1>
<h2 id="1编码区与非编码区">1、编码区与非编码区<a hidden class="anchor" aria-hidden="true" href="#1编码区与非编码区">#</a></h2>
<p>基因 DNA可分为编码区和非编码区：编码区可转录为 mRNA 并最终翻译成蛋白质；非编码区上具有基因表达的调控元件。</p>
<p>（1）编码区：在转录的过程，从DNA编码区的5&rsquo;端开始转录生成preRNA，然后进一步加工修饰剪切得到成熟mRNA，进行后续的翻译。</p>
<ul>
<li>外显子：在preRNA 经过剪切或修饰后、被保留的DNA部分。</li>
<li>起始密码子(Start Codon)：是mRNA上开始合成蛋白质的密码子，也是第一个被核糖体翻译的mRNA密码子。通常为AUG。</li>
<li>终止密码子(Stop Codon)：代表核糖体翻译的终止。通常为UAG、UAA、UGA。</li>
<li>UTR(untranslated region)：属于外显子部分。即外显子也包含非编码区域。
<ul>
<li>位于5&rsquo;端的UTR称为5&rsquo;UTR，如下图橙色标注，从mRNA起点的甲基化鸟嘌呤核苷酸帽延伸至AUG起始密码子。
<ul>
<li>与5&rsquo;UTR上游的第一个碱基相对应DNA链上的碱基称为TSS(Transcription start sites)</li>
</ul>
</li>
<li>位于3&rsquo;端的UTR称为3&rsquo;UTR，为mRNA的结束部分，如下图红色标注，从编码区末端的终止密码子延伸至多聚A尾巴（Poly-A）。
<ul>
<li>与3&rsquo;UTR下游的最后一个碱基相对应DNA链上的碱基称为TTS(Transcription termination sites)</li>
</ul>
</li>
</ul>
</li>
<li>CDS(Coding sequence)：<strong>包括</strong>mRNA中从5&rsquo;UTR后的起始密码子开始到3&rsquo;UTR前的终止密码子的实际编码蛋白序列
<ul>
<li>ORF(Open reading frame)：从一个起始密码子开始到一个终止密码子结束的一段序列
<ul>
<li>注意与CDS定义上的区别。CDS一定是ORF，ORF不一定是CDS</li>
</ul>
</li>
<li>Transcript转录本：一条基因通过可变剪切机制可转录形成的一种或多种可供编码蛋白质的成熟的mRNA。</li>
</ul>
</li>
<li>在最终成熟mRNA的上下游两端修饰有特殊的结构，分别是5&rsquo; cap帽子与 3&rsquo; Poly-A尾巴。
<ul>
<li>Poly-A尾是mRNA&rsquo;区别于其它non-coding RNA的主要标志</li>
</ul>
</li>
</ul>
<p><img loading="lazy" src="https://o.quizlet.com/N2v3.CoCt2zyodCYhg1vMg_b.jpg" alt="Eukaryotic Gene Structure Diagram | Quizlet"  />
</p>
<p>（2）非编码区：位于DNA序列上下游的non-coding区域含有调控基因表达原件</p>
<ul>
<li>启动子Promoter：DNA分子上能与RNA聚合酶结合并形成转录起始复合体的区域，位于TSS位点上游的100~1000碱基序列。
<ul>
<li>真核生物启动子通常会包括一个TATA盒，比较接近TSS区域(50个碱基对以内)</li>
</ul>
</li>
<li>终止子Terminator：给RNA聚合酶提供转录终止信号的DNA序列</li>
<li>增强子Enhancer：可以蛋白质结合的小段DNA，从而加强所调控基因的转录作用。可以位于靶基因的上游/下游，距离也可以很远。</li>
</ul>
<h2 id="2基因组注释文件gtf">2、基因组注释文件gtf<a hidden class="anchor" aria-hidden="true" href="#2基因组注释文件gtf">#</a></h2>
<ul>
<li>第一列：染色体ID</li>
<li>第三列：region类型</li>
<li>第四列：region的起始位点坐标（染色体的第一个碱基序号为1）</li>
<li>第五列：region的终止位点坐标</li>
<li>第七列：正负链信息（+forward/-reverse）
<ul>
<li>正链：表示实际mRNA序列与坐标对应DNA序列相同，除了（U/T变换）</li>
<li>负链：表示实际mRNA序列与坐标对应DNA序列的互补链相同，除了（U/T变换），也需要从5&rsquo;到3&rsquo;端读取。</li>
</ul>
</li>
<li>第八列：对于实际编码区coding region，表示在这个CDS区域开始编码新密码子的起始index{0,1,2}</li>
<li>第九列：region的属性键值对。</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.refGene.gtf.gz</span>
</span></span><span style="display:flex;"><span>gtf = data.table::fread(<span style="color:#0ff;font-weight:bold">&#34;hg38.refGene.gtf.gz&#34;</span>)
</span></span><span style="display:flex;"><span>head(gtf,<span style="color:#ff0;font-weight:bold">3</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#      V1      V2         V3    V4    V5 V6 V7 V8</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1: chr1 refGene transcript 11874 14409  .  +  .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2: chr1 refGene       exon 11874 12227  .  +  .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3: chr1 refGene       exon 12613 12721  .  +  .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                                                                                                            V9</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1:                                        gene_id &#34;DDX11L1&#34;; transcript_id &#34;NR_046018&#34;;  gene_name &#34;DDX11L1&#34;;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2: gene_id &#34;DDX11L1&#34;; transcript_id &#34;NR_046018&#34;; exon_number &#34;1&#34;; exon_id &#34;NR_046018.1&#34;; gene_name &#34;DDX11L1&#34;;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3: gene_id &#34;DDX11L1&#34;; transcript_id &#34;NR_046018&#34;; exon_number &#34;2&#34;; exon_id &#34;NR_046018.2&#34;; gene_name &#34;DDX11L1&#34;;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>table(gtf$V3)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  3UTR        5UTR         CDS        exon start_codon  stop_codon  transcript </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 67901      121178      655401      836541       64026       64026       84787</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 一般来说，每个exon region都对应一个CDS region，除了假基因Pseudogenes。</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>不同平台机构给出的gtf注释文件略有差异，但每列含义基本相同。发现NCBI提供的gff3文件注释信息最全，包括非编码RNA，蛋白质编码RNA、假基因等</p></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(AnnoProbe)
</span></span><span style="display:flex;"><span>IDs &lt;- c(<span style="color:#0ff;font-weight:bold">&#34;DDX11L1&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;MIR6859-1&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;OR4G4P&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;OR4F5&#34;</span>)
</span></span><span style="display:flex;"><span>ID_type = <span style="color:#0ff;font-weight:bold">&#34;SYMBOL&#34;</span>
</span></span><span style="display:flex;"><span>annoGene(IDs, ID_type)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#      SYMBOL                           biotypes         ENSEMBL  chr start   end</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1   DDX11L1 transcribed_unprocessed_pseudogene ENSG00000223972 chr1 11869 14409</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 MIR6859-1                              miRNA ENSG00000278267 chr1 17369 17436</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 7    OR4G4P             unprocessed_pseudogene ENSG00000268020 chr1 52473 53312</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 9     OR4F5                     protein_coding ENSG00000186092 chr1 65419 71585</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="四下载参考基因组">四、下载参考基因组<a hidden class="anchor" aria-hidden="true" href="#四下载参考基因组">#</a></h1>
<p>目前常用的基因组版本为GRCh38/37，hg38/19，前者可通过NCBI/Ensembl下载，后者可通过UCSC网站下载。</p>
<p>如下图所示GRCh38可认为等同于hg38；GRCh37可认为等同于hg19。</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-f2cb46c60b043a9c.png" alt="human genome version"  />
</p>
<p>以下载GRCh38/hg38为例，如下</p>
<h2 id="41-ncbi">4.1 NCBI<a hidden class="anchor" aria-hidden="true" href="#41-ncbi">#</a></h2>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/projects/genome/guide/human/">https://www.ncbi.nlm.nih.gov/projects/genome/guide/human/</a></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>wget -c ftp://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_genomic.fna.gz
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-b88aeae504d8b7a4.png" alt="NCBI" style="zoom: 67%;" />
<h2 id="42-ensembl">4.2 ensembl<a hidden class="anchor" aria-hidden="true" href="#42-ensembl">#</a></h2>
<ul>
<li>ensembl的release-103版本可以认为等于GRCh38</li>
<li><a href="https://asia.ensembl.org/Homo_sapiens/Info/Index">https://asia.ensembl.org/Homo_sapiens/Info/Index</a></li>
<li><a href="http://ftp.ensembl.org/pub/release-103/fasta/homo_sapiens/dna/">http://ftp.ensembl.org/pub/release-103/fasta/homo_sapiens/dna/</a></li>
<li><a href="https://www.ensembl.org/info/data/ftp/index.html">https://www.ensembl.org/info/data/ftp/index.html</a> 还包含有cDNA转录本序列</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>wget -c http://ftp.ensembl.org/pub/release-103/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.toplevel.fa.gz
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://upload-images.jianshu.io/upload_images/20354525-dd4f4a5e7332bb50.png?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240" alt="ensembl" style="zoom: 67%;" />
<h2 id="43-ucsc">4.3 UCSC<a hidden class="anchor" aria-hidden="true" href="#43-ucsc">#</a></h2>
<ul>
<li><a href="http://hgdownload.soe.ucsc.edu/downloads.html">http://hgdownload.soe.ucsc.edu/downloads.html</a></li>
<li><a href="http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/">http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/</a></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>wget -c http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-ac3ca81ec73a0d6b.png" alt="UCSC" style="zoom:67%;" />
<blockquote>
<p>也可以使用之前学习的Refgenie工具下载上述提到的gtf注释文件</p></blockquote>
<h1 id="五txdb与granges对象查询gtf">五、<code>TxDb</code>与<code>GRanges</code>对象查询gtf<a hidden class="anchor" aria-hidden="true" href="#五txdb与granges对象查询gtf">#</a></h1>
<p>TxDb系列包是将物种基因组的gtf注释信息封装为TxDb对象，以供使用者方便的查询。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">#BiocManager::install(&#34;TxDb.Hsapiens.UCSC.hg38.knownGene&#34;)</span>
</span></span><span style="display:flex;"><span>library(TxDb.Hsapiens.UCSC.hg38.knownGene)
</span></span><span style="display:flex;"><span>txdb &lt;- TxDb.Hsapiens.UCSC.hg38.knownGene
</span></span><span style="display:flex;"><span>class(txdb)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;TxDb&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># attr(,&#34;package&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;GenomicFeatures&#34;</span>
</span></span><span style="display:flex;"><span>seqlevels(txdb) &lt;- paste0(<span style="color:#0ff;font-weight:bold">&#34;chr&#34;</span>,c(<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">22</span>,<span style="color:#0ff;font-weight:bold">&#34;X&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;Y&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;M&#34;</span>))
</span></span><span style="display:flex;"><span>seqlevels(txdb)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;chr1&#34;  &#34;chr2&#34;  &#34;chr3&#34;  &#34;chr4&#34;  &#34;chr5&#34;  &#34;chr6&#34;  &#34;chr7&#34;  &#34;chr8&#34;  &#34;chr9&#34;  &#34;chr10&#34; &#34;chr11&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [12] &#34;chr12&#34; &#34;chr13&#34; &#34;chr14&#34; &#34;chr15&#34; &#34;chr16&#34; &#34;chr17&#34; &#34;chr18&#34; &#34;chr19&#34; &#34;chr20&#34; &#34;chr21&#34; &#34;chr22&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [23] &#34;chrX&#34;  &#34;chrY&#34;  &#34;chrM&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#seqlevels(txdb) &lt;- seqlevels0(txdb)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p>txdb对象支持4种region fearture，共计22种的属性注释。</p>
<p>具体使用时可提供<code>keytypes()</code> 所支持的7种信息之一，查询所匹配的其它（22种内）注释信息。</p>
<p>其中GENEID为Entrez基因ID格式，TXID为Ensemble转录本ID格式</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>columns(txdb)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Gene : &#34;GENEID&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CDS  : &#34;CDSCHROM&#34; &#34;CDSEND&#34;  &#34;CDSID&#34; &#34;CDSNAME&#34; &#34;CDSPHASE&#34;  &#34;CDSSTART&#34;  &#34;CDSSTRAND&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># EXON : &#34;EXONCHROM&#34; &#34;EXONEND&#34; &#34;EXONID&#34; &#34;EXONNAME&#34; &#34;EXONRANK&#34; &#34;EXONSTART&#34; &#34;EXONSTRAND&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TX   : &#34;TXCHROM&#34;  &#34;TXEND&#34; &#34;TXID&#34; &#34;TXNAME&#34; &#34;TXSTART&#34; &#34;TXSTRAND&#34; </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>keytypes(txdb)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;GENEID&#34;  &#34;CDSID&#34;    &#34;CDSNAME&#34;  &#34;EXONID&#34;   &#34;EXONNAME&#34;  &#34;TXID&#34;    &#34;TXNAME&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>keys = <span style="color:#0ff;font-weight:bold">&#34;348&#34;</span>  <span style="color:#007f7f">#APOE</span>
</span></span><span style="display:flex;"><span>AnnotationDbi::select(txdb, keys = keys, keytype=<span style="color:#0ff;font-weight:bold">&#34;GENEID&#34;</span>, 
</span></span><span style="display:flex;"><span>                      columns=columns(txdb))
</span></span><span style="display:flex;"><span>AnnotationDbi::select(txdb, keys = keys, keytype=<span style="color:#0ff;font-weight:bold">&#34;GENEID&#34;</span>, 
</span></span><span style="display:flex;"><span>                      columns=c(<span style="color:#0ff;font-weight:bold">&#34;TXID&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TXSTART&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TXEND&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &#39;select()&#39; returned 1:many mapping between keys and columns</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   GENEID   TXID  TXSTART    TXEND</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1    348 201573 44905791 44908944</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2    348 201574 44905796 44907326</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3    348 201575 44905796 44909393</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4    348 201576 44905812 44909025</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5    348 201577 44906360 44908954</span>
</span></span></code></pre></td></tr></table>
</div>
</div><p><code>GenomicFeatures</code>包提供了<code>genes()</code>,<code>transcripts()</code>, <code>exons()</code>, <code>cds()</code>,  <code>promoters()</code>函数用于提取特定feature region为<code>GRange</code>对象。</p>
<p>其中可通过<code>filter=list()</code>参数进行筛选符合注释信息的region，可包括</p>
<blockquote>
<p>&ldquo;gene_id&rdquo;,
&ldquo;tx_id&rdquo;, &ldquo;tx_name&rdquo;, &ldquo;tx_chrom&rdquo;, &ldquo;tx_strand&rdquo;,
&ldquo;exon_id&rdquo;, &ldquo;exon_name&rdquo;, &ldquo;exon_chrom&rdquo;, &ldquo;exon_strand&rdquo;, &ldquo;exon_rank&rdquo;.
&ldquo;cds_id&rdquo;, &ldquo;cds_name&rdquo;, &ldquo;cds_chrom&rdquo;, &ldquo;cds_strand&rdquo;</p></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>GR = transcripts(txdb, filter=list(gene_id=<span style="color:#0ff;font-weight:bold">&#34;348&#34;</span>))
</span></span><span style="display:flex;"><span>GR$tx_id
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GRanges object with 5 ranges and 2 metadata columns:</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   seqnames            ranges strand |     tx_id           tx_name</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;       &lt;character&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1]    chr19 44905791-44908944      + |    201573 ENST00000446996.5</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [2]    chr19 44905796-44907326      + |    201574 ENST00000485628.2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [3]    chr19 44905796-44909393      + |    201575 ENST00000252486.9</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [4]    chr19 44905812-44909025      + |    201576 ENST00000434152.5</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5]    chr19 44906360-44908954      + |    201577 ENST00000425718.1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># -------</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   seqinfo: 25 sequences (1 circular) from hg38 genome</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##相关操作</span>
</span></span><span style="display:flex;"><span>ranges(GR)
</span></span><span style="display:flex;"><span>width(GR)
</span></span><span style="display:flex;"><span>start(GR)
</span></span><span style="display:flex;"><span>end(GR)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>mcols(GR)
</span></span><span style="display:flex;"><span>GR$tx_id
</span></span><span style="display:flex;"><span>GR[strand(GR) == <span style="color:#0ff;font-weight:bold">&#34;+&#34;</span>]
</span></span><span style="display:flex;"><span>GR[1:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span></code></pre></td></tr></table>
</div>
</div><p><code>GRange</code>对象结合BSgenome系列包获取指定feature region的序列信息</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(BSgenome.Hsapiens.UCSC.hg38)
</span></span><span style="display:flex;"><span>BSgenome = BSgenome.Hsapiens.UCSC.hg38
</span></span><span style="display:flex;"><span>GR.seq &lt;- getSeq(BSgenome,GR)
</span></span><span style="display:flex;"><span>GR.seq_df = as.data.frame(GR.seq)
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220528205745522.png" alt="image-20220528205745522"  />
</p>


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80/">生信基础</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/052%E4%BD%BF%E7%94%A8clusterprofiler%E4%B8%8B%E8%BD%BDgokegg%E9%80%9A%E8%B7%AF%E5%9F%BA%E5%9B%A0/">
    <span class="title">« Prev Page</span>
    <br>
    <span>使用clusterProfiler下载GO&amp;KEGG通路基因</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/054%E5%9F%BA%E5%9B%A0-%E8%9B%8B%E7%99%BD-%E5%8C%96%E5%90%88%E7%89%A9id%E8%BD%AC%E6%8D%A2/">
    <span class="title">Next Page »</span>
    <br>
    <span>基因-蛋白-化合物ID转换</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
