<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>UCSCXenaShiny包肿瘤数据分析可视化 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="多组学">
<meta name="description" content="
UCSCXenaShiny是基于集成了多种肿瘤数据库的UCSCXena平台，进行数据下载、分析、可视化的Shiny工具（以及同名R包），由上海科技大学王诗翔博士等共同开发；于2021年6月发表于Bioinformatics。下面主要学习其R包的相关函数，了解其核心功能。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/079ucscxenashiny%E5%8C%85%E8%82%BF%E7%98%A4%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E5%8F%AF%E8%A7%86%E5%8C%96/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/079ucscxenashiny%E5%8C%85%E8%82%BF%E7%98%A4%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E5%8F%AF%E8%A7%86%E5%8C%96/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="UCSCXenaShiny包肿瘤数据分析可视化" />
<meta property="og:description" content="
UCSCXenaShiny是基于集成了多种肿瘤数据库的UCSCXena平台，进行数据下载、分析、可视化的Shiny工具（以及同名R包），由上海科技大学王诗翔博士等共同开发；于2021年6月发表于Bioinformatics。下面主要学习其R包的相关函数，了解其核心功能。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/079ucscxenashiny%E5%8C%85%E8%82%BF%E7%98%A4%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E5%8F%AF%E8%A7%86%E5%8C%96/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-05-12T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-05-12T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="UCSCXenaShiny包肿瘤数据分析可视化"/>
<meta name="twitter:description" content="
UCSCXenaShiny是基于集成了多种肿瘤数据库的UCSCXena平台，进行数据下载、分析、可视化的Shiny工具（以及同名R包），由上海科技大学王诗翔博士等共同开发；于2021年6月发表于Bioinformatics。下面主要学习其R包的相关函数，了解其核心功能。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "UCSCXenaShiny包肿瘤数据分析可视化",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/079ucscxenashiny%E5%8C%85%E8%82%BF%E7%98%A4%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E5%8F%AF%E8%A7%86%E5%8C%96/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "UCSCXenaShiny包肿瘤数据分析可视化",
  "name": "UCSCXenaShiny包肿瘤数据分析可视化",
  "description": " UCSCXenaShiny是基于集成了多种肿瘤数据库的UCSCXena平台，进行数据下载、分析、可视化的Shiny工具（以及同名R包），由上海科技大学王诗翔博士等共同开发；于2021年6月发表于Bioinformatics。下面主要学习其R包的相关函数，了解其核心功能。\n",
  "keywords": [
    "多组学"
  ],
  "articleBody": " UCSCXenaShiny是基于集成了多种肿瘤数据库的UCSCXena平台，进行数据下载、分析、可视化的Shiny工具（以及同名R包），由上海科技大学王诗翔博士等共同开发；于2021年6月发表于Bioinformatics。下面主要学习其R包的相关函数，了解其核心功能。\nhttps://doi.org/10.1093/bioinformatics/btab561 https://github.com/openbiox/UCSCXenaShiny https://shiny.hiplot.cn/ucsc-xena-shiny/ 1、数据关系 1.1 Hub/Cohort等\nHub：不同的大型项目，例如TCGA、GTEx Cohort：不同的队列，可以表示一独立的研究，例如一种肿瘤 Subtype：数据类型，例如表达矩阵/表型/突变数据 Dataset：最终的数据实体，例如不同标准化方式的表达矩阵 所有的数据关系具体可通过如下查询 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 data_stat = UCSCXenaTools::XenaGenerate() ## hosts → cohorts → datasets data_hosts = data_stat@hosts # 11 data_cohorts = data_stat@cohorts # 159 data_datasets = data_stat@datasets # 1681 data_meta = UCSCXenaTools::XenaData t(data_meta[1,]) # [,1] # XenaHosts \"https://ucscpublic.xenahubs.net\" # XenaHostNames \"publicHub\" # XenaCohorts \"Breast Cancer Cell Lines (Neve 2006)\" # XenaDatasets \"ucsfNeve_public/ucsfNeveExp_genomicMatrix\" # SampleCount \"51\" # DataSubtype \"gene expression\" # Label \"Neve Cell Line gene expression\" # Type \"genomicMatrix\" # AnatomicalOrigin \"Breast\" # SampleType \"cell line\" # Tags \"cell lines,breast cancer\" # ProbeMap \"probeMap/affyU133_ucscGenomeBrowser_hg18.probeMap\" # LongTitle \"Cell Line Gene Expression (Neve et al. Cancer Cell 2006)\" # Citation \"Cancer Cell. 2006 Dec;10(6):515-27.\" # Version \"2011-11-01\" # Unit NA # Platform NA table(data_meta$Type) # clinicalMatrix genomicMatrix genomicSegment mutationVector # 345 981 152 203 clinicalMatrix主要包括survival data与phenotype data；\ngenomicMatrix包括所有可以表示为：行名是molecule，列名是sample的矩阵数据，例如基因表达矩阵等\ngenomicSegment主要针对CNV数据；mutationVector主要针对SNV数据\n1.2 常用Hub\nTCGA与TARGET均是由NIH与NCI等机构完成，二者均涉及多种癌症的多组学分析，而后者更专注于少年儿童肿瘤研究； GTEx主要关注正常组织的相关测序数据； CCLE是针对癌症细胞系的多组学研究以及药物敏感性的相关数据； PCAWG是ICGC的子项目之一，也是一项大规模的癌症基因组学研究； TOIL是一个数据计算框架，对上述数据进行单独/整合分析，例如TCGA+TARGET+GETx 1.3 两个R包\nUCSCXenaShiny网站主要以两个R包作为底层基础：\n（1）UCSCXenaTools包是从UCSCXena下载数据/查询数据的工具包；\nhttps://github.com/ropensci/UCSCXenaTools （2）UCSCXenaShiny包基于获取到的数据进行多种生信分析以及可视化。下面主要学习下这个R包的系列函数\nhttps://openbiox.github.io/UCSCXenaShiny/reference/index.html 2、UCSCXenaShiny包 2.1 内置数据 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 TCGA.organ # TCGA缩写 tcga_gtex # 样本所属的组织，分组(肿瘤/正常) tcga_clinical # 肿瘤所属样本年龄性别，肿瘤分期等 tcga_subtypes # 肿瘤亚型 tcga_surv # 肿瘤的4种生存数据 ## 肿瘤其它的不同属性 tcga_genome_instability tcga_purity tcga_tmb ccle_info # 细胞系基本信息 toil_info # toil样本基本信息（TCGA+TARGET+GTEX） pcawg_info #pcawg肿瘤基本信息（包括肿瘤分析，样本OS生存资料等） ## ccle、pcawg的肿瘤纯度信息 ccle_absolute pcawg_purity 此外还有部分数据由于体量大等原因放在了Zenodo数据库，相关分析用到时可另行下载到R包安装路径的extdata文件夹内。\n2.2 查询单个数据 查询任一ense (Genomic) Matrix Dataset（上述1.1）的某个molecule/signature的样本(表达)数据 1 2 3 4 5 6 query_molecule_value() query_molecule_value(dataset=, molecule=) dataset \u003c- \"ccle/CCLE_copynumber_byGene_2013-12-03\" x \u003c- query_molecule_value(dataset, \"TP53\") head(x) 直接查询3大肿瘤Hub(toil/ccle/pcawg)中的某个molecule/signature的样本(表达)数据 1 2 3 4 5 6 7 8 9 query_pancan_value() query_pancan_value(molecule=, data_type= c(\"mRNA\", \"transcript\", \"protein\", \"mutation\", \"cnv\", \"cnv_gistic2\", \"methylation\", \"miRNA\", \"fusion\", \"promoter\", \"APOBEC\"), database= c(\"toil\", \"ccle\", \"pcawg\")) query_pancan_value(\"KRAS\") query_pancan_value(\"KRAS\", database = \"ccle\") query_pancan_value(\"KRAS\", database = \"pcawg\") 2.3 肿瘤相关可视化 这是Shiny网页版的Quick PanCan Analysis模块的分析相关函数\n（1）差异基因分析可视化\n1 2 3 4 vis_toil_TvsN() vis_toil_TvsN_cancer() vis_pcawg_dist() （2）基因表达相关性\n1 2 3 4 5 6 vis_gene_cor() # TCGA vis_gene_cor_cancer() vis_ccle_gene_cor() vis_pcawg_gene_cor() （3）基因表达与肿瘤指标的相关性\n1 2 3 4 5 vis_gene_TIL_cor() vis_gene_immune_cor() vis_gene_stemness_cor() vis_gene_tmb_cor() vis_gene_msi_cor() （4）生存相关分析\n1 2 3 4 5 6 7 8 9 10 11 12 # survival分析 ## step1：获取数据：分子表达+生存资料 sur_data = tcga_surv_get(item=, TCGA_cohort=, profile=) sur_data = tcga_surv_get(\"TP53\", \"BRCA\", \"mRNA\") ## step2 绘图可视化：默认取最佳分组，可自定义阈值 tcga_surv_plot( sur_data, time = \"OS.time\", status = \"OS\") # 单变量Cox回归 vis_unicox_tree() vis_pcawg_unicox_tree() （5）药物敏感度相关分析\n1 2 3 4 5 6 7 # 在多种肿瘤细胞系中，基因表达与药物敏感性的相关性分析 analyze_gene_drug_response_asso(\"TP53\") vis_gene_drug_response_asso(\"TP53\") # 在多种肿瘤细胞系中，根据某基因高低表达分组，药物敏感性的差异 data = analyze_gene_drug_response_diff(\"TP53\") vis_gene_drug_response_diff(\"TP53\") 共涉及20余个药物。\n2.4 一般分析 这是Shiny网页版的General Analysis模块的分析相关函数\n1 2 3 4 5 6 7 8 9 10 11 # 两两相关性可视化（点图） vis_identifier_cor() # 多个间相关性（热图） vis_identifier_multi_cor() # 分组表达差异 vis_identifier_grp_comparison() # 分组生存分析 vis_identifier_grp_surv() ",
  "wordCount" : "1745",
  "inLanguage": "en",
  "datePublished": "2023-05-12T00:00:00Z",
  "dateModified": "2023-05-12T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/079ucscxenashiny%E5%8C%85%E8%82%BF%E7%98%A4%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90%E5%8F%AF%E8%A7%86%E5%8C%96/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      UCSCXenaShiny包肿瘤数据分析可视化
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-05-12 00:00:00 &#43;0000 UTC&#39;&gt;2023-05-12&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-05-12&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1745&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e6%95%b0%e6%8d%ae%e5%85%b3%e7%b3%bb" aria-label="1、数据关系">1、数据关系</a></li>
                    <li>
                        <a href="#2ucscxenashiny%e5%8c%85" aria-label="2、UCSCXenaShiny包">2、UCSCXenaShiny包</a><ul>
                            
                    <li>
                        <a href="#21-%e5%86%85%e7%bd%ae%e6%95%b0%e6%8d%ae" aria-label="2.1 内置数据">2.1 内置数据</a></li>
                    <li>
                        <a href="#22-%e6%9f%a5%e8%af%a2%e5%8d%95%e4%b8%aa%e6%95%b0%e6%8d%ae" aria-label="2.2 查询单个数据">2.2 查询单个数据</a></li>
                    <li>
                        <a href="#23-%e8%82%bf%e7%98%a4%e7%9b%b8%e5%85%b3%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="2.3 肿瘤相关可视化">2.3 肿瘤相关可视化</a></li>
                    <li>
                        <a href="#24-%e4%b8%80%e8%88%ac%e5%88%86%e6%9e%90" aria-label="2.4 一般分析">2.4 一般分析</a>
                    </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>UCSCXenaShiny是基于集成了多种肿瘤数据库的UCSCXena平台，进行数据下载、分析、可视化的Shiny工具（以及同名R包），由上海科技大学王诗翔博士等共同开发；于2021年6月发表于Bioinformatics。下面主要学习其R包的相关函数，了解其核心功能。</p>
<ul>
<li><a href="https://doi.org/10.1093/bioinformatics/btab561">https://doi.org/10.1093/bioinformatics/btab561</a></li>
<li><a href="https://github.com/openbiox/UCSCXenaShiny">https://github.com/openbiox/UCSCXenaShiny</a></li>
<li><a href="https://shiny.hiplot.cn/ucsc-xena-shiny/">https://shiny.hiplot.cn/ucsc-xena-shiny/</a></li>
</ul></blockquote>
<h3 id="1数据关系">1、数据关系<a hidden class="anchor" aria-hidden="true" href="#1数据关系">#</a></h3>
<p><strong>1.1 Hub/Cohort等</strong></p>
<ul>
<li>Hub：不同的大型项目，例如TCGA、GTEx
<ul>
<li>Cohort：不同的队列，可以表示一独立的研究，例如一种肿瘤
<ul>
<li>Subtype：数据类型，例如表达矩阵/表型/突变数据
<ul>
<li>Dataset：最终的数据实体，例如不同标准化方式的表达矩阵</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230511195620921.png" alt="image-20230511195620921" style="zoom: 67%;" />
<ul>
<li>所有的数据关系具体可通过如下查询</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>data_stat = UCSCXenaTools::XenaGenerate()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## hosts → cohorts → datasets</span>
</span></span><span style="display:flex;"><span>data_hosts = data_stat@hosts       <span style="color:#007f7f"># 11</span>
</span></span><span style="display:flex;"><span>data_cohorts = data_stat@cohorts   <span style="color:#007f7f"># 159</span>
</span></span><span style="display:flex;"><span>data_datasets = data_stat@datasets <span style="color:#007f7f"># 1681</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>data_meta = UCSCXenaTools::XenaData
</span></span><span style="display:flex;"><span>t(data_meta[1,])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [,1]                                                      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># XenaHosts        &#34;https://ucscpublic.xenahubs.net&#34;                         </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># XenaHostNames    &#34;publicHub&#34;                                               </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># XenaCohorts      &#34;Breast Cancer Cell Lines (Neve 2006)&#34;                    </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># XenaDatasets     &#34;ucsfNeve_public/ucsfNeveExp_genomicMatrix&#34;               </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SampleCount      &#34;51&#34;                                                      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DataSubtype      &#34;gene expression&#34;                                         </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Label            &#34;Neve Cell Line gene expression&#34;                          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Type             &#34;genomicMatrix&#34;                                           </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AnatomicalOrigin &#34;Breast&#34;                                                  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SampleType       &#34;cell line&#34;                                               </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Tags             &#34;cell lines,breast cancer&#34;                                </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ProbeMap         &#34;probeMap/affyU133_ucscGenomeBrowser_hg18.probeMap&#34;       </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># LongTitle        &#34;Cell Line Gene Expression (Neve et al. Cancer Cell 2006)&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Citation         &#34;Cancer Cell. 2006 Dec;10(6):515-27.&#34;                     </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Version          &#34;2011-11-01&#34;                                              </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Unit             NA                                                        </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Platform         NA         </span>
</span></span><span style="display:flex;"><span>table(data_meta$Type)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># clinicalMatrix  genomicMatrix genomicSegment mutationVector </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            345            981            152            203</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p><strong>clinicalMatrix</strong>主要包括survival data与phenotype data；</p>
<p><strong>genomicMatrix</strong>包括所有可以表示为：行名是molecule，列名是sample的矩阵数据，例如基因表达矩阵等</p>
<p>genomicSegment主要针对CNV数据；mutationVector主要针对SNV数据</p></blockquote>
<p><strong>1.2 常用Hub</strong></p>
<ul>
<li>TCGA与TARGET均是由NIH与NCI等机构完成，二者均涉及多种癌症的多组学分析，而后者更专注于少年儿童肿瘤研究；</li>
<li>GTEx主要关注正常组织的相关测序数据；</li>
<li>CCLE是针对癌症细胞系的多组学研究以及药物敏感性的相关数据；</li>
<li>PCAWG是ICGC的子项目之一，也是一项大规模的癌症基因组学研究；</li>
<li>TOIL是一个数据计算框架，对上述数据进行单独/整合分析，例如TCGA+TARGET+GETx</li>
</ul>
<p><img loading="lazy" src="https://xena.ucsc.edu/static/6bb3020cbbfe4b4cb51276c8d2776643/dd45a/PublicDataLogos.png" alt="UCSC Xena"  />
</p>
<p><strong>1.3 两个R包</strong></p>
<p>UCSCXenaShiny网站主要以两个R包作为底层基础：</p>
<p>（1）UCSCXenaTools包是从UCSCXena下载数据/查询数据的工具包；</p>
<ul>
<li><a href="https://github.com/ropensci/UCSCXenaTools">https://github.com/ropensci/UCSCXenaTools</a></li>
</ul>
<p>（2）UCSCXenaShiny包基于获取到的数据进行多种生信分析以及可视化。下面主要学习下这个R包的系列函数</p>
<ul>
<li><a href="https://openbiox.github.io/UCSCXenaShiny/reference/index.html">https://openbiox.github.io/UCSCXenaShiny/reference/index.html</a></li>
</ul>
<h3 id="2ucscxenashiny包">2、UCSCXenaShiny包<a hidden class="anchor" aria-hidden="true" href="#2ucscxenashiny包">#</a></h3>
<h4 id="21-内置数据">2.1 内置数据<a hidden class="anchor" aria-hidden="true" href="#21-内置数据">#</a></h4>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>TCGA.organ   <span style="color:#007f7f"># TCGA缩写</span>
</span></span><span style="display:flex;"><span>tcga_gtex <span style="color:#007f7f"># 样本所属的组织，分组(肿瘤/正常)</span>
</span></span><span style="display:flex;"><span>tcga_clinical  <span style="color:#007f7f"># 肿瘤所属样本年龄性别，肿瘤分期等</span>
</span></span><span style="display:flex;"><span>tcga_subtypes <span style="color:#007f7f"># 肿瘤亚型</span>
</span></span><span style="display:flex;"><span>tcga_surv <span style="color:#007f7f"># 肿瘤的4种生存数据</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 肿瘤其它的不同属性</span>
</span></span><span style="display:flex;"><span>tcga_genome_instability
</span></span><span style="display:flex;"><span>tcga_purity
</span></span><span style="display:flex;"><span>tcga_tmb
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>ccle_info <span style="color:#007f7f"># 细胞系基本信息</span>
</span></span><span style="display:flex;"><span>toil_info <span style="color:#007f7f"># toil样本基本信息（TCGA+TARGET+GTEX）</span>
</span></span><span style="display:flex;"><span>pcawg_info <span style="color:#007f7f">#pcawg肿瘤基本信息（包括肿瘤分析，样本OS生存资料等）</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## ccle、pcawg的肿瘤纯度信息</span>
</span></span><span style="display:flex;"><span>ccle_absolute
</span></span><span style="display:flex;"><span>pcawg_purity
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>此外还有部分数据由于体量大等原因放在了Zenodo数据库，相关分析用到时可另行下载到R包安装路径的extdata文件夹内。</p></blockquote>
<h4 id="22-查询单个数据">2.2 查询单个数据<a hidden class="anchor" aria-hidden="true" href="#22-查询单个数据">#</a></h4>
<ul>
<li>查询任一ense (Genomic) Matrix Dataset（上述1.1）的某个molecule/signature的样本(表达)数据</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>query_molecule_value()
</span></span><span style="display:flex;"><span>query_molecule_value(dataset=, molecule=)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>dataset &lt;- <span style="color:#0ff;font-weight:bold">&#34;ccle/CCLE_copynumber_byGene_2013-12-03&#34;</span>
</span></span><span style="display:flex;"><span>x &lt;- query_molecule_value(dataset, <span style="color:#0ff;font-weight:bold">&#34;TP53&#34;</span>)
</span></span><span style="display:flex;"><span>head(x)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>直接查询3大肿瘤Hub(toil/ccle/pcawg)中的某个molecule/signature的样本(表达)数据</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>query_pancan_value() 
</span></span><span style="display:flex;"><span>query_pancan_value(molecule=, 
</span></span><span style="display:flex;"><span>  data_type= c(<span style="color:#0ff;font-weight:bold">&#34;mRNA&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;transcript&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;protein&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;mutation&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;cnv&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;cnv_gistic2&#34;</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#0ff;font-weight:bold">&#34;methylation&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;miRNA&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;fusion&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;promoter&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;APOBEC&#34;</span>),
</span></span><span style="display:flex;"><span>  database= c(<span style="color:#0ff;font-weight:bold">&#34;toil&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;ccle&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;pcawg&#34;</span>)) 
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>query_pancan_value(<span style="color:#0ff;font-weight:bold">&#34;KRAS&#34;</span>)
</span></span><span style="display:flex;"><span>query_pancan_value(<span style="color:#0ff;font-weight:bold">&#34;KRAS&#34;</span>, database = <span style="color:#0ff;font-weight:bold">&#34;ccle&#34;</span>)
</span></span><span style="display:flex;"><span>query_pancan_value(<span style="color:#0ff;font-weight:bold">&#34;KRAS&#34;</span>, database = <span style="color:#0ff;font-weight:bold">&#34;pcawg&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h4 id="23-肿瘤相关可视化">2.3 肿瘤相关可视化<a hidden class="anchor" aria-hidden="true" href="#23-肿瘤相关可视化">#</a></h4>
<p>这是Shiny网页版的<code>Quick PanCan Analysis</code>模块的分析相关函数</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230512210649206.png" alt="image-20230512210649206"  />
</p>
<p><strong>（1）差异基因分析可视化</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>vis_toil_TvsN()
</span></span><span style="display:flex;"><span>vis_toil_TvsN_cancer()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>vis_pcawg_dist()
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（2）基因表达相关性</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>vis_gene_cor()  <span style="color:#007f7f"># TCGA</span>
</span></span><span style="display:flex;"><span>vis_gene_cor_cancer() 
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>vis_ccle_gene_cor()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>vis_pcawg_gene_cor()
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（3）基因表达与肿瘤指标的相关性</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>vis_gene_TIL_cor()
</span></span><span style="display:flex;"><span>vis_gene_immune_cor()
</span></span><span style="display:flex;"><span>vis_gene_stemness_cor()
</span></span><span style="display:flex;"><span>vis_gene_tmb_cor()
</span></span><span style="display:flex;"><span>vis_gene_msi_cor()
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（4）生存相关分析</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># survival分析</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## step1：获取数据：分子表达+生存资料</span>
</span></span><span style="display:flex;"><span>sur_data = tcga_surv_get(item=, TCGA_cohort=, profile=)
</span></span><span style="display:flex;"><span>sur_data = tcga_surv_get(<span style="color:#0ff;font-weight:bold">&#34;TP53&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;BRCA&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;mRNA&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## step2 绘图可视化：默认取最佳分组，可自定义阈值</span>
</span></span><span style="display:flex;"><span>tcga_surv_plot(
</span></span><span style="display:flex;"><span>  sur_data,
</span></span><span style="display:flex;"><span>  time = <span style="color:#0ff;font-weight:bold">&#34;OS.time&#34;</span>, status = <span style="color:#0ff;font-weight:bold">&#34;OS&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 单变量Cox回归</span>
</span></span><span style="display:flex;"><span>vis_unicox_tree()
</span></span><span style="display:flex;"><span>vis_pcawg_unicox_tree()
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>（5）药物敏感度相关分析</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 在多种肿瘤细胞系中，基因表达与药物敏感性的相关性分析</span>
</span></span><span style="display:flex;"><span>analyze_gene_drug_response_asso(<span style="color:#0ff;font-weight:bold">&#34;TP53&#34;</span>)
</span></span><span style="display:flex;"><span>vis_gene_drug_response_asso(<span style="color:#0ff;font-weight:bold">&#34;TP53&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 在多种肿瘤细胞系中，根据某基因高低表达分组，药物敏感性的差异</span>
</span></span><span style="display:flex;"><span>data = analyze_gene_drug_response_diff(<span style="color:#0ff;font-weight:bold">&#34;TP53&#34;</span>)
</span></span><span style="display:flex;"><span>vis_gene_drug_response_diff(<span style="color:#0ff;font-weight:bold">&#34;TP53&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>共涉及20余个药物。</p></blockquote>
<h4 id="24-一般分析">2.4 一般分析<a hidden class="anchor" aria-hidden="true" href="#24-一般分析">#</a></h4>
<p>这是Shiny网页版的<code>General Analysis</code>模块的分析相关函数</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230512212857662.png" alt="image-20230512212857662"  />
</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 两两相关性可视化（点图）</span>
</span></span><span style="display:flex;"><span>vis_identifier_cor()  
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 多个间相关性（热图）</span>
</span></span><span style="display:flex;"><span>vis_identifier_multi_cor()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 分组表达差异</span>
</span></span><span style="display:flex;"><span>vis_identifier_grp_comparison()  
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 分组生存分析</span>
</span></span><span style="display:flex;"><span>vis_identifier_grp_surv()
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E5%A4%9A%E7%BB%84%E5%AD%A6/">多组学</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/078%E8%AF%BA%E6%A8%A1%E5%9B%BE%E6%88%96%E5%88%97%E7%BA%BF%E5%9B%BE%E7%BB%98%E5%88%B6%E5%88%86%E6%9E%90/">
    <span class="title">« Prev Page</span>
    <br>
    <span>诺模图或列线图绘制分析</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/080enhancedvolcano%E5%8C%85%E7%BB%98%E5%88%B6%E7%81%AB%E5%B1%B1%E5%9B%BE/">
    <span class="title">Next Page »</span>
    <br>
    <span>EnhancedVolcano包绘制火山图</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
