<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰ | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="">
<meta name="description" content="ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰ - Li&#39;s Bioinfo-Blog">
<meta name="author" content="">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" type="application/rss+xml" href="https://lishensuo.github.io/en/posts/bioinfo/index.xml">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰" />
<meta property="og:description" content="æç”³é”çš„åšå®¢ï¼Œè®°å½•ç”Ÿä¿¡å­¦ä¹ è¿‡ç¨‹ã€‚R/Python/Shell" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰"/>
<meta name="twitter:description" content="æç”³é”çš„åšå®¢ï¼Œè®°å½•ç”Ÿä¿¡å­¦ä¹ è¿‡ç¨‹ã€‚R/Python/Shell"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "åˆ†ç±»",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }
  ]
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="list" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="ä¸»é¡µ">
                    <span>ä¸»é¡µ</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="åˆ†ç±»">
                    <span>åˆ†ç±»</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="æ ‡ç­¾">
                    <span>æ ‡ç­¾</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="å½’æ¡£">
                    <span>å½’æ¡£</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="å…³äº">
                    <span>å…³äº</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="æœç´¢ (Alt &#43; /)" accesskey=/>
                    <span>æœç´¢</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main"> 
<header class="page-header"><div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/">åˆ†ç±»</a></div>
  <h1>ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰</h1>
</header>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>ä»GEOä¸‹è½½èŠ¯ç‰‡æˆ–RNAseqæµ‹åºæ•°æ®
			</h2>
		</header>
		<section class="entry-content">
			<p> æŒ–æ˜GEOæ•°æ®æ—¶ï¼Œä¸»è¦ä¸€æ–¹é¢æ˜¯ä¸‹è½½GEOçš„æµ‹åºæ•°æ®ï¼ˆåŒ…æ‹¬åŸºå› èŠ¯ç‰‡arrayä¸RNAseqä¸¤ç±»ï¼‰çš„è¡¨è¾¾çŸ©é˜µã€‚åŒæ—¶ä¼šæ¶‰åŠåˆ°ä¸€äº›ç»†èŠ‚é—®é¢˜ï¼Œä¾‹å¦‚arrayèŠ¯ç‰‡IDè½¬æ¢ã€æ ·æœ¬metaä¿¡æ¯ç­‰ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-30 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-30&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-04-30&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1791&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to ä»GEOä¸‹è½½èŠ¯ç‰‡æˆ–RNAseqæµ‹åºæ•°æ®" href="https://lishensuo.github.io/en/posts/bioinfo/048%E4%BB%8Egeo%E4%B8%8B%E8%BD%BD%E8%8A%AF%E7%89%87%E6%88%96rnaseq%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AE/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>ä¸‹è½½æµ‹åºæ•°æ®SRRä¸fastq.gzæ–¹å¼
			</h2>
		</header>
		<section class="entry-content">
			<p>1ã€å‡†å¤‡condaç¯å¢ƒä¸è½¯ä»¶ 1 2 3 4 5 6 7 8 # å‡†å¤‡downloadç¯å¢ƒ conda create -n download conda activate download # å®‰è£…è½¯ä»¶ conda install -c hcc aspera-cli conda install -c bioconda sra-tools conda install -c conda-forge pigz ä»¥SRR13911909ä¸ºä¾‹ 2ã€asperaä¸‹è½½ https://www.ebi.ac.uk/ena/browser/view/SRR13911909 2.1 ä¸‹è½½é“¾æ¥ 1 2 3 4 5 # era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/SRR139/009/SRR13911909/SRR13911909_1.fastq.gz # era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/SRR139/009/SRR13911909/SRR13911909_2.fastq.gz # è§‚å¯Ÿä¸Šè¿°é“¾æ¥è§„å¾‹åï¼Œå¯ä»¥è‡ªåŠ¨ç”Ÿæˆä¸‹è½½é“¾æ¥ era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/${id:0:6}/00${id:0-1}/${id}/${id}_1.fastq.gz era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/${id:0:6}/00${id:0-1}/${id}/${id}_2.fastq.gz 2.2 asperaä¸‹è½½ 1 2 3 4 5 6 7 8 9 10 11 12 13 id=SRR13911909 ascp -QT -l 300m -P33001 \ -i ~/miniconda3/envs/download/etc/asperaweb_id_dsa.openssh \ era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/${id:0:6}/00${id:0-1}/${id}/${id}_1.fastq.gz . ascp -QT -l 300m -P33001 \ -i ~/miniconda3/envs/download/etc/asperaweb_id_dsa.openssh \ era-fasp@fasp.sra.ebi.ac.uk:/vol1/fastq/${id:0:6}/00${id:0-1}/${id}/${id}_2.fastq.gz . #æœ€åä¸€ä¸ªç‚¹è¡¨ç¤ºä¸‹è½½æ–‡ä»¶çš„å‚¨å­˜è·¯å¾„ #å¦‚æœæ˜¯å…¶å®ƒç¯å¢ƒï¼Œå°†downloadæ›¿æ¢ä¸ºå¯¹åº”ç¯å¢ƒåå³å¯ #å¦‚æœæ˜¯baseç¯å¢ƒï¼š ~/miniconda3/etc/asperaweb_id_dsa.openssh ä½¿ç”¨asperaå¯ä»¥è¾¾åˆ°ç™¾å…†çš„é€Ÿåº¦ï¼Œå»ºè®®é¦–å…ˆå°è¯•ã€‚ä½†æœ€è¿‘è¯•äº†å‡ æ¬¡ï¼Œå®¹æ˜“æŠ¥é”™ï¼Œä¸ç¨³å®š(æŠ¥é”™å†…å®¹å¦‚ä¸‹)ï¼›æœ‰æ—¶å€™å¯ä»¥ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-16 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-16&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-05-23&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;559&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to ä¸‹è½½æµ‹åºæ•°æ®SRRä¸fastq.gzæ–¹å¼" href="https://lishensuo.github.io/en/posts/bioinfo/049%E4%B8%8B%E8%BD%BD%E6%B5%8B%E5%BA%8F%E6%95%B0%E6%8D%AEsrr%E4%B8%8Efastq.gz%E6%96%B9%E5%BC%8F/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>Refgenieä¸‹è½½å‚è€ƒåŸºå› ç»„
			</h2>
		</header>
		<section class="entry-content">
			<p>refgenieï¼šå‚è€ƒåŸºå› ç»„ï¼ˆé˜¿æ‹‰ä¸ï¼‰å•†åº— http://refgenie.databio.org/ Here we provide a web interface and a RESTful API to access genome assets for popular reference genome assemblies.
è¯¥å¹³å°ç”±ä½äºå¼—å‰å°¼äºšå¤§å­¦å…¬å…±å«ç”ŸåŸºå› ç»„å­¦ä¸­å¿ƒçš„è®¡ç®—ç”Ÿç‰©å­¦å’Œç”Ÿç‰©ä¿¡æ¯å­¦ç ”ç©¶å°ç»„(Sheffield lab of computational biology)å»ºç«‹ã€‚ä¸Šæ¬¡ä¿®æ”¹/æ›´æ–°æ˜¯2021å¹´11æœˆã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-16 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-16&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-04-16&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1243&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to Refgenieä¸‹è½½å‚è€ƒåŸºå› ç»„" href="https://lishensuo.github.io/en/posts/bioinfo/050refgenie%E4%B8%8B%E8%BD%BD%E5%8F%82%E8%80%83%E5%9F%BA%E5%9B%A0%E7%BB%84/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>STRINGdbåŒ…ä¸‹è½½è›‹ç™½PPIæ•°æ®
			</h2>
		</header>
		<section class="entry-content">
			<p> 1 2 3 4 5 if (!require(&#34;BiocManager&#34;, quietly = TRUE)) install.packages(&#34;BiocManager&#34;) BiocManager::install(&#34;STRINGdb&#34;) library(STRINGdb) 1ã€å®šä¹‰è¦ä½¿ç”¨çš„STRINGç‰ˆæœ¬ã€ç‰©ç§ï¼Œä»¥åŠPPIé˜ˆå€¼åˆ†æ•° 1 2 3 4 string_db &lt;- STRINGdb$new(version=&#34;11&#34;, species=9606, score_threshold=200, input_directory=&#34;&#34;) 2ã€ç¤ºä¾‹åŸºå›  1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 data(diff_exp_example1) genes = diff_exp_example1[1:50,] head(genes) # pvalue logFC gene # 1 0.0001018 3.333461 VSTM2L # 2 0.0001392 3.822383 TBC1D2 # 3 0.0001720 3.306056 LENG9 # 4 0.0001739 3.024605 TMEM27 # 5 0.0001990 3.854414 LOC100506014 # 6 0.0002393 3.082052 TSPAN1 ###åŸºå› ååŒ¹é…protein ID #ç¬¬ä¸€ä¸ªå‚æ•°æ˜¯data.frame; ç¬¬äºŒä¸ªå‚æ•°æ˜¯åŸºå› æ‰€åœ¨åˆ—çš„åˆ—å genes_mapped &lt;- string_db$map(genes, &#34;gene&#34; ) #Warning: we couldn&#39;t map to STRING 30% of your identifiers head(genes_mapped) # gene pvalue logFC STRING_id # 1 VSTM2L 0.0001018 3.333461 9606.ENSP00000362560 # 2 TBC1D2 0.0001392 3.822383 9606.ENSP00000481721 # 3 LENG9 0.0001720 3.306056 9606.ENSP00000479355 # 4 TMEM27 0.0001739 3.024605 9606.ENSP00000369699 # 40 LOC100506014 0.0001990 3.854414 &lt;NA&gt; # 5 TSPAN1 0.0002393 3.082052 9606.ENSP00000361072 #string_db$plot_network(genes_mapped$STRING_id) 3ã€ä¸‹è½½è¿™äº›åŸºå› é—´çš„äº’ä½œå…³ç³» 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 ppi = string_db$get_interactions(genes_mapped$STRING_id) %&gt;% distinct() ppi = ppi %&gt;% dplyr::left_join(genes_mapped[,c(1,4)], by=c(&#39;from&#39;=&#39;STRING_id&#39;)) %&gt;% #åˆ—æ ‡åºå·æ ¹æ®å…·ä½“æƒ…å†µè€Œå®š dplyr::rename(Gene1=gene) %&gt;% ##åˆ—åæ ¹æ®å…·ä½“æƒ…å†µè€Œå®š dplyr::left_join(genes_mapped[,c(1,4)], by=c(&#39;to&#39;=&#39;STRING_id&#39;)) %&gt;% dplyr::rename(Gene2=gene) %&gt;% dplyr::select(Gene1, Gene2, combined_score) head(ppi) # Gene1 Gene2 combined_score # 1 C3 TYROBP 240 # 2 ABCA12 GRHL3 308 # 3 FAM189A1 TM4SF20 400 # 4 ABCA12 NIPAL4 824 # 5 GRHL3 NIPAL4 275 # 6 GRHL3 IGDCC4 238 </p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-16 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-16&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-04-16&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;340&amp;nbsp;|&amp;nbsp;1 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to STRINGdbåŒ…ä¸‹è½½è›‹ç™½PPIæ•°æ®" href="https://lishensuo.github.io/en/posts/bioinfo/051stringdb%E5%8C%85%E4%B8%8B%E8%BD%BD%E8%9B%8B%E7%99%BDppi%E6%95%B0%E6%8D%AE/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>ä½¿ç”¨clusterProfilerä¸‹è½½GO&amp;KEGGé€šè·¯åŸºå› 
			</h2>
		</header>
		<section class="entry-content">
			<p>1ã€GO 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 GO_data &lt;- clusterProfiler:::get_GO_data(&#34;org.Hs.eg.db&#34;, &#34;ALL&#34;, &#34;SYMBOL&#34;) names(GO_data) # [1] &#34;PATHID2NAME&#34; &#34;EXTID2PATHID&#34; &#34;GO2ONT&#34; &#34;PATHID2EXTID&#34; ###ï¼ˆ1ï¼‰GO termçš„ç»„æˆåŸºå›  class(GO_data$PATHID2EXTID) #[1] &#34;list&#34; GO_data$PATHID2EXTID[1] # $`GO:0000002` # [1] &#34;PARP1&#34; &#34;SLC25A4&#34; &#34;DNA2&#34; &#34;TYMP&#34; &#34;LIG3&#34; &#34;MEF2A&#34; # [7] &#34;MPV17&#34; &#34;OPA1&#34; &#34;TOP3A&#34; &#34;TP53&#34; &#34;LONP1&#34; &#34;AKT3&#34; # [13] &#34;PPARGC1A&#34; &#34;POLG2&#34; &#34;SLC25A36&#34; &#34;PIF1&#34; &#34;SESN2&#34; &#34;SLC25A33&#34; # [19] &#34;MGME1&#34; &#34;PRIMPOL&#34; &#34;STOX1&#34; ###ï¼ˆ2ï¼‰åŸºå› æ‰€æ¶‰åŠçš„é€šè·¯ class(GO_data$EXTID2PATHID) #[1] &#34;list&#34; GO_data$EXTID2PATHID[1] # $A1BG # [1] &#34;GO:0001775&#34; &#34;GO:0002252&#34; &#34;GO:0002263&#34; &#34;GO:0002274&#34; &#34;GO:0002275&#34; &#34;GO:0002283&#34; &#34;GO:0002366&#34; # [8] &#34;GO:0002376&#34; &#34;GO:0002443&#34; &#34;GO:0002444&#34; &#34;GO:0002446&#34; &#34;GO:0002576&#34; &#34;GO:0003674&#34; &#34;GO:0005575&#34; # ... ###ï¼ˆ3ï¼‰GO termçš„åå­— class(GO_data$PATHID2NAME) #[1] &#34;character&#34; GO_data$PATHID2NAME[1] # GO:0000001 # &#34;mitochondrion inheritance&#34; ###ï¼ˆ4ï¼‰GO termçš„ç±»åˆ« class(GO_data$GO2ONT) #[1] &#34;character&#34; GO_data$GO2ONT[1] # GO:0000002 # &#34;BP&#34; table(GO_data$GO2ONT) # BP CC MF # 16013 1981 4755 library(tidyverse) go_name = reshape2::melt(GO_data$PATHID2NAME) %&gt;% rownames_to_column(&#34;ID&#34;) %&gt;% dplyr::rename(&#34;Name&#34;=&#34;value&#34;) go_type = reshape2::melt(GO_data$GO2ONT) %&gt;% rownames_to_column(&#34;ID&#34;) %&gt;% dplyr::rename(&#34;Type&#34;=&#34;value&#34;) go_info = inner_join(go_name, go_type) %&gt;% dplyr::mutate(GSEA=toupper(gsub(&#34; &#34;,&#34;_&#34;,paste0(&#34;GO&#34;,Type,&#34; &#34;,Name)))) dim(go_info) head(go_info) table(rownames(brca_enrich_kegg) %in% go_info$GSEA) # ID Name Type # 1 GO:0000002 mitochondrial genome maintenance BP # 2 GO:0000003 reproduction BP # 3 GO:0000009 alpha-1,6-mannosyltransferase activity MF # 4 GO:0000010 trans-hexaprenyltranstransferase activity MF # 5 GO:0000012 single strand break repair BP # 6 GO:0000014 single-stranded DNA endodeoxyribonuclease activity MF # GSEA # 1 GOBP_MITOCHONDRIAL_GENOME_MAINTENANCE # 2 GOBP_REPRODUCTION # 3 GOMF_ALPHA-1,6-MANNOSYLTRANSFERASE_ACTIVITY # 4 GOMF_TRANS-HEXAPRENYLTRANSTRANSFERASE_ACTIVITY # 5 GOBP_SINGLE_STRAND_BREAK_REPAIR # 6 GOMF_SINGLE-STRANDED_DNA_ENDODEOXYRIBONUCLEASE_ACTIVITY æ­¤å¤– GO.dbåŒ…ä¹Ÿæä¾›äº†é™¤ç»„æˆåŸºå› ä»¥å¤–çš„GOæ³¨é‡Šä¿¡æ¯ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 library(GO.db) keytypes(GO.db) # [1] &#34;DEFINITION&#34; &#34;GOID&#34; &#34;ONTOLOGY&#34; &#34;TERM&#34; goids = keys(GO.db, keytype = &#34;GOID&#34;)[1:3] # [1] &#34;GO:0000001&#34; &#34;GO:0000002&#34; &#34;GO:0000003&#34; goids_anno = AnnotationDbi::select(GO.db, keys = goids, columns = c(&#34;TERM&#34;,&#34;ONTOLOGY&#34;,&#34;DEFINITION&#34;), #å…¶ä¸­DEFINITIONä¸ºtermçš„è¯¦ç»†æè¿° keytype=&#34;GOID&#34;) #æ‰€æœ‰çš„BP termçš„GO id goBP = select(GO.db, keys = &#34;BP&#34;, columns = c(&#34;GOID&#34;), keytype=&#34;ONTOLOGY&#34;) 2ã€KEGG 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 hsa_kegg &lt;- clusterProfiler::download_KEGG(&#34;hsa&#34;) names(hsa_kegg) # [1] &#34;KEGGPATHID2EXTID&#34; &#34;KEGGPATHID2NAME&#34; ### KEGG idä¸name head(z) # from to # 1 hsa00010 Glycolysis / Gluconeogenesis # 2 hsa00020 Citrate cycle (TCA cycle) # 3 hsa00030 Pentose phosphate pathway ### KEGG idçš„ç»„æˆåŸºå›  head(hsa_kegg$KEGGPATHID2EXTID) # from to # 1 hsa00010 10327 # 2 hsa00010 124 # 3 hsa00010 125 </p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-16 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-16&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-04-16&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;528&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to ä½¿ç”¨clusterProfilerä¸‹è½½GO&KEGGé€šè·¯åŸºå› " href="https://lishensuo.github.io/en/posts/bioinfo/052%E4%BD%BF%E7%94%A8clusterprofiler%E4%B8%8B%E8%BD%BDgokegg%E9%80%9A%E8%B7%AF%E5%9F%BA%E5%9B%A0/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>äººç±»åŸºå› ç»„åŸºç¡€çŸ¥è¯†ä¸ä¸‹è½½æŸ¥è¯¢
			</h2>
		</header>
		<section class="entry-content">
			<p>ä¸€ã€åŸºå› ç»„å¤§å° ï¼ˆ1ï¼‰äººç±»åŸºå› ç»„ä¸»è¦ç”±ç»†èƒæ ¸çš„23å¯¹æŸ“è‰²ä½“ç»„æˆ(æ ¸åŸºå› ç»„)ï¼Œè¿˜åŒ…æ‹¬çº¿ç²’ä½“ä¸­çš„å°DNAåˆ†å­(çº¿ç²’ä½“åŸºå› ç»„)ã€‚
ï¼ˆ2ï¼‰å•å€ä½“åŸºå› ç»„å¤§æ¦‚æœ‰30äº¿ä¸ªç¢±åŸºå¯¹ç»„æˆï¼Œå…·ä½“åˆ°æ¯ä¸ªæŸ“è‰²ä½“çš„ç¢±åŸºå¯¹é•¿åº¦ä¸åŸºå› æ•°é‡å¦‚ä¸‹æ‰€ç¤º(å‚ç…§UCSCçš„hg38)ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-05-28 00:00:00 &#43;0000 UTC&#39;&gt;2022-05-28&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-05-28&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;3802&amp;nbsp;|&amp;nbsp;8 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to äººç±»åŸºå› ç»„åŸºç¡€çŸ¥è¯†ä¸ä¸‹è½½æŸ¥è¯¢" href="https://lishensuo.github.io/en/posts/bioinfo/053%E4%BA%BA%E7%B1%BB%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%9F%BA%E7%A1%80%E7%9F%A5%E8%AF%86%E4%B8%8E%E4%B8%8B%E8%BD%BD%E6%9F%A5%E8%AF%A2/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>åŸºå› -è›‹ç™½-åŒ–åˆç‰©IDè½¬æ¢
			</h2>
		</header>
		<section class="entry-content">
			<p>1ã€ä¸åŒåŸºå› IDè½¬æ¢ 1.1 org.Hs.eg.dbåŒ… 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 library(dplyr) library(org.Hs.eg.db) keytypes(org.Hs.eg.db) # [1] &#34;ACCNUM&#34; &#34;ALIAS&#34; &#34;ENSEMBL&#34; &#34;ENSEMBLPROT&#34; &#34;ENSEMBLTRANS&#34; &#34;ENTREZID&#34; # [7] &#34;ENZYME&#34; &#34;EVIDENCE&#34; &#34;EVIDENCEALL&#34; &#34;GENENAME&#34; &#34;GENETYPE&#34; &#34;GO&#34; # [13] &#34;GOALL&#34; &#34;IPI&#34; &#34;MAP&#34; &#34;OMIM&#34; &#34;ONTOLOGY&#34; &#34;ONTOLOGYALL&#34; # [19] &#34;PATH&#34; &#34;PFAM&#34; &#34;PMID&#34; &#34;PROSITE&#34; &#34;REFSEQ&#34; &#34;SYMBOL&#34; # [25] &#34;UCSCKG&#34; &#34;UNIPROT&#34; gene_symbol=c(&#34;RHO&#34;,&#34;CALM1&#34;,&#34;MEG3&#34;,&#34;GNGT1&#34;,&#34;SAG&#34;,&#34;RPGRIP1&#34;,&#34;TRPM1&#34;,&#34;PCP2&#34;,&#34;PCP4&#34;,&#34;AP1B1&#34;) gene_ids&lt;-AnnotationDbi::select(org.Hs.eg.db, keys=as.character(gene_symbol), columns=c(&#34;ENSEMBL&#34;,&#34;ENTREZID&#34;), #ç›®æ ‡æ ¼å¼ keytype=&#34;SYMBOL&#34;) #ç›®å‰çš„æ ¼å¼ gene_ids ##å»é‡ gene_ids %&gt;% dplyr::distinct(ENTREZID, .keep_all = T) # SYMBOL ENSEMBL ENTREZID # 1 RHO ENSG00000163914 6010 # 2 CALM1 ENSG00000198668 801 # 3 MEG3 ENSG00000214548 55384 # 4 GNGT1 ENSG00000127928 2792 # 5 SAG ENSG00000130561 6295 # 6 RPGRIP1 ENSG00000092200 57096 # 7 TRPM1 ENSG00000134160 4308 # 8 PCP2 ENSG00000174788 126006 # 9 PCP4 ENSG00000183036 5121 # 10 AP1B1 ENSG00000100280 162 1.2 biomaRtåŒ… 1 2 3 4 5 6 7 8 9 10 11 12 library(&#34;biomaRt&#34;) ensembl = useMart(&#34;ensembl&#34;,dataset=&#34;hsapiens_gene_ensembl&#34;) attributes = listAttributes(ensembl) attributes[1:5,] # library(httr) # httr::set_config(config(ssl_verifypeer = 0L)) gene_symbol=c(&#34;RHO&#34;,&#34;CALM1&#34;,&#34;MEG3&#34;,&#34;GNGT1&#34;,&#34;SAG&#34;,&#34;RPGRIP1&#34;,&#34;TRPM1&#34;,&#34;PCP2&#34;,&#34;PCP4&#34;,&#34;AP1B1&#34;) gene_ids2 &lt;- getBM(filters= &#34;hgnc_symbol&#34;, attributes= c(&#34;hgnc_symbol&#34;,&#34;ensembl_gene_id&#34;,&#34;entrezgene_id&#34;), values = gene_symbol, mart= ensembl) gene_ids2 2ã€é¼ æºåŸºå› è½¬ä¸ºäººç±»åŸºå› ID 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 musGenes &lt;- c(&#34;Hmmr&#34;, &#34;Tlx3&#34;, &#34;Cpeb4&#34;) ## æ–¹å¼1ï¼šç›´æ¥å¤§å°å†™è½¬æ¢ toupper(musGenes) # [1] &#34;HMMR&#34; &#34;TLX3&#34; &#34;CPEB4&#34; ## æ–¹å¼2ï¼šé€šè¿‡biomaRtåŒ…ï¼ˆä¸ç¨³å®šï¼‰ require(&#34;biomaRt&#34;) # library(httr) # httr::set_config(config(ssl_verifypeer = 0L)) human = useMart(&#34;ensembl&#34;, dataset = &#34;hsapiens_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;) mouse = useMart(&#34;ensembl&#34;, dataset = &#34;mmusculus_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;) genes = getLDS(attributes = c(&#34;mgi_symbol&#34;), filters = &#34;mgi_symbol&#34;, values = musGenes, mart = mouse, attributesL = c(&#34;hgnc_symbol&#34;), martL = human, uniqueRows=T) ## æ–¹å¼3ï¼šMGI æ•°æ®åº“ # https://support.bioconductor.org/p/129636/ library(dplyr) mouse_human_genes = read.csv(&#34;http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt&#34;,sep=&#34;\t&#34;) convert_mouse_to_human &lt;- function(gene_list){ output = c() for(gene in gene_list){ class_key = (mouse_human_genes %&gt;% filter(Symbol == gene &amp; Common.Organism.Name==&#34;mouse, laboratory&#34;))[[&#39;DB.Class.Key&#39;]] if(!identical(class_key, integer(0)) ){ human_genes = (mouse_human_genes %&gt;% filter(DB.Class.Key == class_key &amp; Common.Organism.Name==&#34;human&#34;))[,&#34;Symbol&#34;] for(human_gene in human_genes){ output = append(output,human_gene) } } } return (output) } convert_mouse_to_human(musGenes) # 1] &#34;HMMR&#34; &#34;TLX3&#34; &#34;CPEB4&#34; 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 # # https://github.com/lishensuo/utils # # library(&#34;biomaRt&#34;) # # library(httr) # # httr::set_config(config(ssl_verifypeer = 0L)) # human = useMart(&#34;ensembl&#34;, dataset = &#34;hsapiens_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;) # mouse = useMart(&#34;ensembl&#34;, dataset = &#34;mmusculus_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;) # # # https://www.gencodegenes.org/mouse/ # dat = data.table::fread(&#34;gencode.vM33.basic.annotation.gtf.gz&#34;) # dat = subset(dat, V3 == &#34;gene&#34;) # dat_sub = dat[,&#34;V9&#34;] %&gt;% # separate(V9, into = c(&#34;gene_id&#34;,&#34;gene_type&#34;,&#34;gene_name&#34;,&#34;mgi_id&#34;,&#34;havana_gene&#34;), sep = &#34;; &#34;) # dat_sub$gene_name2 = gsub(&#39;gencode.vM33.basic.annotation.gtf.gz &#34;&#39;,&#39;&#39;,dat_sub$gene_name) # dat_sub$gene_name2 = gsub(&#39;&#34;&#39;,&#39;&#39;,dat_sub$gene_name2) # # genes = getLDS(attributes = c(&#34;mgi_symbol&#34;), filters = &#34;mgi_symbol&#34;, # values = dat_sub$gene_name2, # mart = mouse, # attributesL = c(&#34;hgnc_symbol&#34;), # martL = human, uniqueRows=T) # write.csv(genes, file = &#34;mgi2hgnc_biomart.csv&#34;, row.names = F, quote = F) # head(genes) 3ã€è›‹ç™½è´¨ä¸åŸºå› IDè½¬æ¢ https://www.uniprot.org/uploadlists/ ...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-05-28 00:00:00 &#43;0000 UTC&#39;&gt;2022-05-28&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2025-06-25&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2488&amp;nbsp;|&amp;nbsp;5 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to åŸºå› -è›‹ç™½-åŒ–åˆç‰©IDè½¬æ¢" href="https://lishensuo.github.io/en/posts/bioinfo/054%E5%9F%BA%E5%9B%A0-%E8%9B%8B%E7%99%BD-%E5%8C%96%E5%90%88%E7%89%A9id%E8%BD%AC%E6%8D%A2/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>GSEAå¯Œé›†åˆ†æå·¥å…·
			</h2>
		</header>
		<section class="entry-content">
			<p>ä»¥å‰é€šè·¯å¯Œé›†åˆ†æç›´æ¥ä½¿ç”¨clusterprofileråŒ…ï¼Œé˜…è¯»æ–‡çŒ®å‘ç°GSEAåˆ†æåŠå¯è§†åŒ–è¾ƒå¤šä½¿ç”¨Broadå›¢é˜Ÿç ”å‘çš„å·¥å…·ï¼Œç°ç®€è¦å­¦ä¹ å…¶(windowç‰ˆæœ¬)ä½¿ç”¨æ–¹æ³•ã€‚
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-13 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-13&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-13&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1154&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to GSEAå¯Œé›†åˆ†æå·¥å…·" href="https://lishensuo.github.io/en/posts/bioinfo/055gsea%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>clusterProfileråŒ…å¯Œé›†åˆ†æä¸å¯è§†åŒ–
			</h2>
		</header>
		<section class="entry-content">
			<p>1ã€èƒŒæ™¯çŸ¥è¯† ï¼ˆ1ï¼‰ä¸¤ç§å¯Œé›†åˆ†æ åŸºäºè¶…å‡ ä½•æ£€éªŒçš„ORA(over representation analysis)å¯Œé›†åˆ†æ â‘  å‡è®¾å¯¹è½¬å½•ç»„åˆ†ç»„æµ‹åºçš„10000ä¸ªåŸºå› è¡¨è¾¾æ•°æ®è¿›è¡Œå·®å¼‚åˆ†æï¼›
â‘¡ æŒ‰ç…§è§„å®šcutoffé˜ˆå€¼(logFC/Pvalue)ç­›é€‰å¾—åˆ°100ä¸ªå·®å¼‚è¡¨è¾¾åŸºå› ï¼›
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-14 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-14&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-14&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;3645&amp;nbsp;|&amp;nbsp;8 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to clusterProfileråŒ…å¯Œé›†åˆ†æä¸å¯è§†åŒ–" href="https://lishensuo.github.io/en/posts/bioinfo/056clusterprofiler%E5%8C%85%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/"></a>
</article>





<article class="post-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>å¯Œé›†é€šè·¯å¯è§†åŒ–æ–¹å¼
			</h2>
		</header>
		<section class="entry-content">
			<p>0ã€ç¤ºä¾‹æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 library(clusterProfiler) library(org.Hs.eg.db) data(geneList, package=&#34;DOSE&#34;) str(geneList) # Named num [1:12495] 4.57 4.51 4.42 4.14 3.88 ... # - attr(*, &#34;names&#34;)= chr [1:12495] &#34;4312&#34; &#34;8318&#34; &#34;10874&#34; &#34;55143&#34; ... gene_ids&lt;-AnnotationDbi::select(org.Hs.eg.db, keys=as.character(names(gene_list)), columns=&#34;SYMBOL&#34;, #ç›®æ ‡æ ¼å¼ keytype=&#34;ENTREZID&#34;) #ç›®å‰çš„æ ¼å¼ # &#39;select()&#39; returned 1:1 mapping between keys and columns ## ORAåˆ†æ gene_deg = names(geneList)[abs(geneList) &gt; 2] ego &lt;- enrichGO(gene = gene_deg, OrgDb = org.Hs.eg.db, ont = &#34;CC&#34;, # &#34;BP&#34;,&#34;MF&#34;,&#34;ALL&#34; pAdjustMethod = &#34;BH&#34;, pvalueCutoff = 0.01, qvalueCutoff = 0.05, readable = TRUE) ## GSEAåˆ†æ gseago &lt;- gseGO(geneList = geneList, OrgDb = org.Hs.eg.db, ont = &#34;CC&#34;, minGSSize = 100, maxGSSize = 500, pvalueCutoff = 0.05, verbose = FALSE) gseago = setReadable(gseago, OrgDb = org.Hs.eg.db) 1ã€ggplot2 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 ## ORA library(tidyverse) ego_df = ego@result %&gt;% head(10) %&gt;% dplyr::mutate(logp = -log10(p.adjust)) %&gt;% dplyr::arrange(logp) %&gt;% dplyr::mutate(Label=str_wrap(gsub(&#34;_&#34;,&#34; &#34;,Description), width = 50)) %&gt;% dplyr::mutate(Label = factor(Label, levels=Label)) %&gt;% tibble::remove_rownames() %&gt;% dplyr::select(Label, logp) head(ego_df) # Label logp # 1 microtubule associated complex 5.419514 # 2 spindle midzone 5.681085 # 3 spindle microtubule 7.341138 ggplot(ego_df, aes(x=Label, y=logp)) &#43; geom_bar(stat=&#34;identity&#34;, fill=&#34;#1b9e77&#34;) &#43; xlab(&#34;Pathway names&#34;) &#43; ylab(&#34;-log10(P.adj)&#34;) &#43; scale_y_continuous(expand=c(0,0)) &#43; coord_flip() &#43; xlab(&#34;GO Pathway&#34;) &#43; ylab(&#34;Adjusted P value(-log10)&#34;) &#43; geom_text(aes(label = Label, y=0.2), hjust = 0, size=5) &#43; theme_classic() &#43; theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), axis.title.y = element_text(size = 25), axis.title.x = element_text(size = 20), axis.text.x = element_text(size = 16)) 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 ## GSEA library(tidyverse) gseago_df = gseago@result %&gt;% head(20) %&gt;% dplyr::mutate(logp = -log10(p.adjust)) %&gt;% dplyr::arrange(logp) %&gt;% dplyr::mutate(Label=str_wrap(gsub(&#34;_&#34;,&#34; &#34;,Description), width = 50)) %&gt;% dplyr::mutate(Label = factor(Label, levels=Label)) %&gt;% tibble::remove_rownames() %&gt;% dplyr::select(Label, logp, NES) head(gseago_df) # Label logp NES # 1 polymeric cytoskeletal fiber 3.755730 1.487284 # 2 chromosome, telomeric region 4.311816 1.907511 # 3 microtubule associated complex 4.432327 1.944046 ggplot(gseago_df, aes(NES, fct_reorder(Label, NES), fill=qvalue)) &#43; geom_bar(stat=&#39;identity&#39;) &#43; scale_fill_continuous(low=&#39;red&#39;, high=&#39;blue&#39;, guide=guide_colorbar(reverse=TRUE)) &#43; theme_minimal() &#43; ylab(NULL) 2ã€GOplot å®˜æ–¹æ•™ç¨‹ï¼šhttps://wencke.github.io/ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 # install.packages(&#39;GOplot&#39;) library(GOplot) # (1) å‡†å¤‡æ•°æ® ## å¯Œé›†é€šè·¯ç»“æœ res_enrich = ego@result %&gt;% dplyr::select(ID, Description, geneID,p.adjust) %&gt;% dplyr::mutate(Category = &#34;GO&#34;, .before=1) %&gt;% dplyr::mutate(geneID = gsub(&#34;/&#34;,&#34;, &#34;, geneID)) %&gt;% dplyr::rename(&#34;Term&#34;=&#34;Description&#34;, &#34;Genes&#34;=&#34;geneID&#34;, &#34;adj_pval&#34;=&#34;p.adjust&#34;) %&gt;% dplyr::select(Category,ID,Term,Genes,adj_pval) %&gt;% tibble::remove_rownames() t(res_enrich[1,]) # 1 # Category &#34;GO&#34; # ID &#34;GO:0005819&#34; # Term &#34;spindle&#34; # Genes &#34;CDCA8, CDC20, KIF23, CENPE, ASPM, DLGAP5, SKA1, NUSAP1, TPX2, TACC3, NEK2, CDK1, MAD2L1, KIF18A, BIRC5, KIF11, TRAT1, TTK, AURKB, PRC1, KIFC1, KIF18B, KIF20A, AURKA, CCNB1, KIF4A&#34; # adj_pval &#34;6.339976e-11&#34; ## å¦‚ä¸Šï¼ŒIDä¸ºoptionalï¼ŒCategoryä¸ºrequired ## (2) åŸºå› å·®å¼‚å€æ•° res_logfc = data.frame(ID=gene_ids$SYMBOL, logFC=geneList) head(res_logfc) # ID logFC # 1 MMP1 4.572613 # 2 CDC45 4.514594 # 3 NMU 4.418218 ## å¦‚æ— ç›¸å…³æ•°æ®ï¼Œå¯éšä¾¿è®¾ç½®ä¸º0 ## (3) æ„å»ºç»˜å›¾å¯¹è±¡ circ &lt;- circle_dat(res_enrich, res_logfc) head(circ) Circular visualization
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-14 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-14&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-06-17&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1544&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to å¯Œé›†é€šè·¯å¯è§†åŒ–æ–¹å¼" href="https://lishensuo.github.io/en/posts/bioinfo/057%E5%AF%8C%E9%9B%86%E9%80%9A%E8%B7%AF%E5%8F%AF%E8%A7%86%E5%8C%96%E6%96%B9%E5%BC%8F/"></a>
</article>
<footer class="page-footer">
  <nav class="pagination">
    <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/page/4/">Â« Prev Page</a>
    <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/page/6/">Next Page Â»</a>
  </nav>
</footer>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>æ‚¨æ˜¯æœ¬ç«™ç¬¬ <span id="busuanzi_value_site_uv"></span> ä½è®¿é—®è€…ï¼Œæ€»æµè§ˆé‡ä¸º <span id="busuanzi_value_site_pv"></span> æ¬¡
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
