<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--hdWGCNA共表达网络 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
WGCNA是适用于大批量样本的array或者Bulk RNAseq数据的加权基因共表达网络分析。由于单细胞数据的稀疏性，不适用于WGCNA直接分析。hdWGCNA包基于WGCNA包提供了一种针对scRNAseq数据的加权基因共表达网络分析策略。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/016%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--hdwgcna%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/016%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--hdwgcna%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--hdWGCNA共表达网络" />
<meta property="og:description" content="
WGCNA是适用于大批量样本的array或者Bulk RNAseq数据的加权基因共表达网络分析。由于单细胞数据的稀疏性，不适用于WGCNA直接分析。hdWGCNA包基于WGCNA包提供了一种针对scRNAseq数据的加权基因共表达网络分析策略。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/016%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--hdwgcna%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-07-07T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2022-07-07T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--hdWGCNA共表达网络"/>
<meta name="twitter:description" content="
WGCNA是适用于大批量样本的array或者Bulk RNAseq数据的加权基因共表达网络分析。由于单细胞数据的稀疏性，不适用于WGCNA直接分析。hdWGCNA包基于WGCNA包提供了一种针对scRNAseq数据的加权基因共表达网络分析策略。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--hdWGCNA共表达网络",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/016%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--hdwgcna%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--hdWGCNA共表达网络",
  "name": "单细胞分析工具--hdWGCNA共表达网络",
  "description": " WGCNA是适用于大批量样本的array或者Bulk RNAseq数据的加权基因共表达网络分析。由于单细胞数据的稀疏性，不适用于WGCNA直接分析。hdWGCNA包基于WGCNA包提供了一种针对scRNAseq数据的加权基因共表达网络分析策略。\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " WGCNA是适用于大批量样本的array或者Bulk RNAseq数据的加权基因共表达网络分析。由于单细胞数据的稀疏性，不适用于WGCNA直接分析。hdWGCNA包基于WGCNA包提供了一种针对scRNAseq数据的加权基因共表达网络分析策略。\n官方教程：https://smorabit.github.io/hdWGCNA/articles/basic_tutorial.html\n1、加载包并准备示例数据 1 2 3 4 5 6 7 8 9 10 11 12 13 #devtools::install_github('smorabit/hdWGCNA', ref='dev') library(hdWGCNA) library(WGCNA) library(Seurat) library(clustree) library(dplyr) library(patchwork) # 示例数据 # https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz list.files(\"filtered_gene_bc_matrices/hg19/\") # [1] \"barcodes.tsv\" \"genes.tsv\" \"matrix.mtx\" 2、创建Seurat对象并预处理 执行Seurat的常规流程，包括：创建seurat对象—标准化、归一化—高变基因—降维、聚类分群—鉴定细胞类型(optional)\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 sce = Read10X(data.dir = \"filtered_gene_bc_matrices/hg19/\") %\u003e% CreateSeuratObject() # 32738 features across 2700 samples sce = sce %\u003e% NormalizeData() %\u003e% FindVariableFeatures() %\u003e% ScaleData() sce = sce %\u003e% RunPCA() %\u003e% RunUMAP(dims = 1:30) %\u003e% #RunTSNE FindNeighbors(dims = 1:30) %\u003e% FindClusters(resolution = c(0.01, 0.05, 0.1, 0.2, 0.3, 0.5,0.8,1)) clustree(sce@meta.data, prefix = \"RNA_snn_res.\") #colnames(sce@meta.data) table(sce$RNA_snn_res.0.1) # 0 1 2 3 # 1195 698 448 359 Idents(sce) = \"RNA_snn_res.0.1\" ##如上按照分辨率为0.1的分群结果进行后续的分析 3、创建misc slot，挑选基因 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 seurat_obj = SetupForWGCNA( sce, gene_select = \"fraction\", # the gene selection approach fraction = 0.05, # fraction of cells that a gene needs to be expressed in order to be included wgcna_name = \"PBMC\" # the name of the hdWGCNA experiment ) # gene_select参数设置过滤基因的方法，还有是 ## (1) \"variable\" : use the genes stored in the Seurat object’s VariableFeatures. ## (2) \"custom\" : use genes that are specified in a custom list. str(seurat_obj@misc) # List of 2 # $ active_wgcna: chr \"PBMC\" # $ PBMC :List of 2 # ..$ wgcna_group: chr \"all\" # ..$ wgcna_genes: chr [1:3855] \"NOC2L\" \"HES4\" \"ISG15\" \"TNFRSF4\" ... 4、聚合metacell 将多个细胞表达情况(KNN)合并为一个metacell，可避免单细胞数据的稀疏性。\n“多个细胞\"需要是来自同一种细胞类型/分群，可通过group.by参数定义依据。\nk参数：in general a lower number for k can be used for small datasets. We generally use k values between 20 and 75.\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 seurat_obj \u003c- MetacellsByGroups( seurat_obj = seurat_obj, group.by = c(\"RNA_snn_res.0.1\"), # specify the columns in seurat_obj@meta.data to group by k = 25, # nearest-neighbors parameter max_shared = 10, # maximum number of shared cells between two metacells ident.group = 'RNA_snn_res.0.1' # set the Idents of the metacell seurat object ) seurat_obj \u003c- NormalizeMetacells(seurat_obj) seurat_obj@misc$PBMC$wgcna_metacell_obj # 32738 features across 1524 samples within 1 assay seurat_obj@misc$PBMC$wgcna_metacell_obj@assays$RNA@counts[1:4,1:8] # 4 x 8 sparse Matrix of class \"dgCMatrix\" # 0_1 0_2 0_3 0_4 0_5 0_6 0_7 0_8 # MIR1302-10 . . . . . . . . # FAM138A . . . . . . . . # OR4F5 . . . . . . . . # RP11-34P13.7 . . . . . . . . seurat_obj@misc$PBMC$wgcna_metacell_obj@meta.data %\u003e% head # orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.1 # 0_1 0 2618.96 5880 0 # 0_2 0 2486.00 5858 0 # 0_3 0 1340.56 4306 0 # 0_4 0 2112.12 5388 0 # 0_5 0 2839.04 6073 0 # 0_6 0 2198.44 5407 0 5、共表达网络分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 ##(1) 选择一种/多种细胞类型，进行WGCNA分析 seurat_obj \u003c- SetDatExpr( seurat_obj, group_name = c(0,1,2,3), group.by='RNA_snn_res.0.1' # the metadata column containing the cell type info. This same column should have also been used in MetacellsByGroups ) ##(2) 选择合适的软阈值 seurat_obj \u003c- TestSoftPowers( seurat_obj, setDatExpr = FALSE, # set this to FALSE since we did this above ) plot_list \u003c- PlotSoftPowers(seurat_obj) wrap_plots(plot_list, ncol=2) # power_table \u003c- GetPowerTable(seurat_obj) # head(power_table) 1 2 3 4 5 6 ##(3) 根据上图选择合适的软阈值构建网络 seurat_obj \u003c- ConstructNetwork( seurat_obj, soft_power=8, setDatExpr=FALSE ) PlotDendrogram(seurat_obj, main='INH hdWGCNA Dendrogram') 6、模块特征值与hub基因分析 （1）模块组成基因 1 2 3 4 5 6 7 8 9 10 11 12 13 14 seurat_obj@misc$PBMC$wgcna_modules %\u003e% head # gene_name module color # NOC2L NOC2L grey grey # HES4 HES4 turquoise turquoise # ISG15 ISG15 turquoise turquoise # TNFRSF4 TNFRSF4 blue blue # SDF4 SDF4 blue blue # UBE2J2 UBE2J2 blue blue table(seurat_obj@misc$PBMC$wgcna_modules$module) # grey turquoise blue brown yellow # 588 1786 1272 133 76 # modules \u003c- GetModules(seurat_obj) # head(modules) （2）每个细胞对于每个模块的特征值 1 2 3 4 5 6 7 8 9 10 11 12 13 14 seurat_obj \u003c- ModuleEigengenes( seurat_obj ) seurat_obj@misc$PBMC$MEs %\u003e% head # blue grey brown turquoise yellow # AAACATACAACCAC-1 4.755930 0.2424180 2.6565029 -5.832575 -2.0906948 # AAACATTGAGCTAC-1 2.388492 1.3106515 -0.3341197 1.172415 8.8764667 # AAACATTGATCAGC-1 5.807490 2.9008702 0.8873175 -1.132285 -1.8973713 # AAACCGTGCTTCCG-1 -6.662387 -0.3072945 -1.5451723 14.692779 0.1159869 # AAACCGTGTATGCG-1 -16.098322 -1.6676024 8.7666564 -4.131657 -0.9371790 # AAACGCACTGGTAC-1 2.064300 -0.1415680 -1.2313300 -2.824847 -1.1698044 # MEs \u003c- GetMEs(seurat_obj, harmonized=FALSE) # head(MEs) （3）鉴定模块内hub基因– 基因表达与模块特征值的具有高相关性 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 seurat_obj \u003c- ModuleConnectivity( seurat_obj ) seurat_obj@misc$PBMC$wgcna_modules[,c(-1,-2,-3)] %\u003e% head # kME_blue kME_grey kME_brown kME_turquoise kME_yellow # NOC2L 0.02851618 0.08360662 0.02041163 0.003048238 0.038497122 # HES4 -0.24225802 0.06869209 -0.08284058 0.117744840 0.326692215 # ISG15 -0.29705821 0.08027215 -0.07240235 0.351103776 0.314836382 # TNFRSF4 0.16100218 0.08589770 0.13010946 -0.002353201 -0.090182722 # SDF4 0.01120696 0.13914234 0.07918151 0.040070556 0.003869467 # UBE2J2 0.03178884 0.07651548 0.04663831 0.008615164 -0.006942615 # p \u003c- PlotKMEs(seurat_obj, ncol=2) # p hub_df \u003c- GetHubGenes(seurat_obj, n_hubs = 10) hub_df # gene_name module kME # 1 RPS27 blue 0.7207118 # 2 RPS12 blue 0.7258250 # 3 CCL5 brown 0.4934812 # 4 B2M brown 0.5294537 # 5 LGALS1 turquoise 0.6477877 # 6 TYROBP turquoise 0.6542149 # 7 HLA-DRA yellow 0.5817169 # 8 HLA-DPA1 yellow 0.5894472 ## 计算每个细胞对于每个模块hub基因的表达活性(module score),可使用seurat包或者Ucell包 seurat_obj \u003c- ModuleExprScore( seurat_obj, n_genes = 25, method='Seurat' ) seurat_obj@misc$PBMC$module_scores %\u003e% head ## 计算每个细胞对于每个模块hub基因的表达活性(module score),可使用seurat包或者Ucell包 # blue brown turquoise yellow # AAACATACAACCAC-1 3.380281 0.74403040 0.3806462 -0.18699177 # AAACATTGAGCTAC-1 3.035111 0.31949694 0.4019337 1.06061130 # AAACATTGATCAGC-1 2.769994 0.28652286 0.5089602 -0.30358276 # AAACCGTGCTTCCG-1 2.080403 0.07190668 2.6016017 0.45921797 # AAACCGTGTATGCG-1 1.434566 1.07243443 0.9849420 0.14309843 # AAACGCACTGGTAC-1 2.999822 0.25594526 0.6041499 -0.02860217 # library(UCell) # seurat_obj \u003c- ModuleExprScore( # seurat_obj, # n_genes = 25, # method='UCell' # ) # seurat_obj@misc$PBMC$module_scores %\u003e% head 7、结果可视化 每个细胞对于每个模块的特征值 1 2 3 4 5 6 plot_list \u003c- ModuleFeaturePlot( seurat_obj, features='MEs', # plot the hMEs order=TRUE # order so the points with highest hMEs are on top ) wrap_plots(plot_list, ncol=2) 每个细胞对于每个模块hub基因的表达活性 1 2 3 4 5 6 7 8 seurat_obj@misc$PBMC$module_scores %\u003e% head plot_list \u003c- ModuleFeaturePlot( seurat_obj, features='scores', # plot the hub gene scores order='shuffle', # order so cells are shuffled ucell = TRUE # depending on Seurat vs UCell for gene scoring ) wrap_plots(plot_list, ncol=2) 将细胞对于模块的特征值，整合到seurat的meta.data中 1 2 3 4 5 6 7 8 9 10 MEs \u003c- GetMEs(seurat_obj) mods \u003c- colnames(MEs); mods \u003c- mods[mods != 'grey'] seurat_obj@meta.data \u003c- cbind(seurat_obj@meta.data, MEs) p \u003c- DotPlot(seurat_obj, features = mods, group.by = 'RNA_snn_res.0.1') p \u003c- p + coord_flip() + RotatedAxis() + scale_color_gradient2(high='red', mid='grey95', low='blue') p ",
  "wordCount" : "1850",
  "inLanguage": "en",
  "datePublished": "2022-07-07T00:00:00Z",
  "dateModified": "2022-07-07T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/016%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--hdwgcna%E5%85%B1%E8%A1%A8%E8%BE%BE%E7%BD%91%E7%BB%9C/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--hdWGCNA共表达网络
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2022-07-07 00:00:00 &#43;0000 UTC&#39;&gt;2022-07-07&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-07-07&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1850&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e5%8a%a0%e8%bd%bd%e5%8c%85%e5%b9%b6%e5%87%86%e5%a4%87%e7%a4%ba%e4%be%8b%e6%95%b0%e6%8d%ae" aria-label="1、加载包并准备示例数据">1、加载包并准备示例数据</a></li>
                    <li>
                        <a href="#2%e5%88%9b%e5%bb%baseurat%e5%af%b9%e8%b1%a1%e5%b9%b6%e9%a2%84%e5%a4%84%e7%90%86" aria-label="2、创建Seurat对象并预处理">2、创建Seurat对象并预处理</a></li>
                    <li>
                        <a href="#3%e5%88%9b%e5%bb%bamisc-slot%e6%8c%91%e9%80%89%e5%9f%ba%e5%9b%a0" aria-label="3、创建misc slot，挑选基因">3、创建misc slot，挑选基因</a></li>
                    <li>
                        <a href="#4%e8%81%9a%e5%90%88metacell" aria-label="4、聚合metacell">4、聚合metacell</a></li>
                    <li>
                        <a href="#5%e5%85%b1%e8%a1%a8%e8%be%be%e7%bd%91%e7%bb%9c%e5%88%86%e6%9e%90" aria-label="5、共表达网络分析">5、共表达网络分析</a></li>
                    <li>
                        <a href="#6%e6%a8%a1%e5%9d%97%e7%89%b9%e5%be%81%e5%80%bc%e4%b8%8ehub%e5%9f%ba%e5%9b%a0%e5%88%86%e6%9e%90" aria-label="6、模块特征值与hub基因分析">6、模块特征值与hub基因分析</a></li>
                    <li>
                        <a href="#7%e7%bb%93%e6%9e%9c%e5%8f%af%e8%a7%86%e5%8c%96" aria-label="7、结果可视化">7、结果可视化</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>WGCNA是适用于大批量样本的array或者Bulk RNAseq数据的加权基因共表达网络分析。由于单细胞数据的稀疏性，不适用于WGCNA直接分析。hdWGCNA包基于WGCNA包提供了一种针对scRNAseq数据的加权基因共表达网络分析策略。</p>
<p>官方教程：https://smorabit.github.io/hdWGCNA/articles/basic_tutorial.html</p></blockquote>
<h2 id="1加载包并准备示例数据">1、加载包并准备示例数据<a hidden class="anchor" aria-hidden="true" href="#1加载包并准备示例数据">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">#devtools::install_github(&#39;smorabit/hdWGCNA&#39;, ref=&#39;dev&#39;)</span>
</span></span><span style="display:flex;"><span>library(hdWGCNA)
</span></span><span style="display:flex;"><span>library(WGCNA)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>library(clustree)
</span></span><span style="display:flex;"><span>library(dplyr)
</span></span><span style="display:flex;"><span>library(patchwork)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 示例数据</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz</span>
</span></span><span style="display:flex;"><span>list.files(<span style="color:#0ff;font-weight:bold">&#34;filtered_gene_bc_matrices/hg19/&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;barcodes.tsv&#34; &#34;genes.tsv&#34;    &#34;matrix.mtx&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="2创建seurat对象并预处理">2、创建Seurat对象并预处理<a hidden class="anchor" aria-hidden="true" href="#2创建seurat对象并预处理">#</a></h2>
<p>执行Seurat的常规流程，包括：创建seurat对象&mdash;标准化、归一化&mdash;高变基因&mdash;降维、聚类分群&mdash;鉴定细胞类型(optional)</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sce = Read10X(data.dir = <span style="color:#0ff;font-weight:bold">&#34;filtered_gene_bc_matrices/hg19/&#34;</span>) %&gt;% CreateSeuratObject()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 32738 features across 2700 samples</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sce = sce %&gt;%
</span></span><span style="display:flex;"><span>  NormalizeData() %&gt;%
</span></span><span style="display:flex;"><span>  FindVariableFeatures() %&gt;%
</span></span><span style="display:flex;"><span>  ScaleData()
</span></span><span style="display:flex;"><span>sce = sce %&gt;% 
</span></span><span style="display:flex;"><span>  RunPCA() %&gt;% 
</span></span><span style="display:flex;"><span>  RunUMAP(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;%  <span style="color:#007f7f">#RunTSNE</span>
</span></span><span style="display:flex;"><span>  FindNeighbors(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  FindClusters(resolution = c(<span style="color:#ff0;font-weight:bold">0.01</span>, <span style="color:#ff0;font-weight:bold">0.05</span>, <span style="color:#ff0;font-weight:bold">0.1</span>, <span style="color:#ff0;font-weight:bold">0.2</span>, <span style="color:#ff0;font-weight:bold">0.3</span>, <span style="color:#ff0;font-weight:bold">0.5</span>,<span style="color:#ff0;font-weight:bold">0.8</span>,<span style="color:#ff0;font-weight:bold">1</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>clustree(sce@meta.data, prefix = <span style="color:#0ff;font-weight:bold">&#34;RNA_snn_res.&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#colnames(sce@meta.data)</span>
</span></span><span style="display:flex;"><span>table(sce$RNA_snn_res.0.1)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#    0    1    2    3</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1195  698  448  359</span>
</span></span><span style="display:flex;"><span>Idents(sce) = <span style="color:#0ff;font-weight:bold">&#34;RNA_snn_res.0.1&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##如上按照分辨率为0.1的分群结果进行后续的分析</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="3创建misc-slot挑选基因">3、创建misc slot，挑选基因<a hidden class="anchor" aria-hidden="true" href="#3创建misc-slot挑选基因">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>seurat_obj = SetupForWGCNA(
</span></span><span style="display:flex;"><span>  sce,
</span></span><span style="display:flex;"><span>  gene_select = <span style="color:#0ff;font-weight:bold">&#34;fraction&#34;</span>, <span style="color:#007f7f"># the gene selection approach</span>
</span></span><span style="display:flex;"><span>  fraction = <span style="color:#ff0;font-weight:bold">0.05</span>, <span style="color:#007f7f"># fraction of cells that a gene needs to be expressed in order to be included</span>
</span></span><span style="display:flex;"><span>  wgcna_name = <span style="color:#0ff;font-weight:bold">&#34;PBMC&#34;</span> <span style="color:#007f7f"># the name of the hdWGCNA experiment</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gene_select参数设置过滤基因的方法，还有是</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (1) &#34;variable&#34; :  use the genes stored in the Seurat object’s VariableFeatures.</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (2) &#34;custom&#34;   :  use genes that are specified in a custom list.</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>str(seurat_obj@misc)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># List of 2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  $ active_wgcna: chr &#34;PBMC&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  $ PBMC        :List of 2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   ..$ wgcna_group: chr &#34;all&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   ..$ wgcna_genes: chr [1:3855] &#34;NOC2L&#34; &#34;HES4&#34; &#34;ISG15&#34; &#34;TNFRSF4&#34; ...</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="4聚合metacell">4、聚合metacell<a hidden class="anchor" aria-hidden="true" href="#4聚合metacell">#</a></h2>
<p>将多个细胞表达情况(KNN)合并为一个metacell，可避免单细胞数据的稀疏性。</p>
<p>&ldquo;多个细胞&quot;需要是来自同一种细胞类型/分群，可通过<code>group.by</code>参数定义依据。</p>
<blockquote>
<p><code>k</code>参数：in general a lower number for k can be used for small datasets. We generally use k values between 20 and 75.</p></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>seurat_obj &lt;- MetacellsByGroups(
</span></span><span style="display:flex;"><span>  seurat_obj = seurat_obj,
</span></span><span style="display:flex;"><span>  group.by = c(<span style="color:#0ff;font-weight:bold">&#34;RNA_snn_res.0.1&#34;</span>), <span style="color:#007f7f"># specify the columns in seurat_obj@meta.data to group by</span>
</span></span><span style="display:flex;"><span>  k = <span style="color:#ff0;font-weight:bold">25</span>, <span style="color:#007f7f"># nearest-neighbors parameter</span>
</span></span><span style="display:flex;"><span>  max_shared = <span style="color:#ff0;font-weight:bold">10</span>, <span style="color:#007f7f"># maximum number of shared cells between two metacells</span>
</span></span><span style="display:flex;"><span>  ident.group = <span style="color:#0ff;font-weight:bold">&#39;RNA_snn_res.0.1&#39;</span> <span style="color:#007f7f"># set the Idents of the metacell seurat object</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>seurat_obj &lt;- NormalizeMetacells(seurat_obj)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>seurat_obj@misc$PBMC$wgcna_metacell_obj
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 32738 features across 1524 samples within 1 assay</span>
</span></span><span style="display:flex;"><span>seurat_obj@misc$PBMC$wgcna_metacell_obj@assays$RNA@counts[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">8</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 x 8 sparse Matrix of class &#34;dgCMatrix&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#              0_1 0_2 0_3 0_4 0_5 0_6 0_7 0_8</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MIR1302-10     .   .   .   .   .   .   .   .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FAM138A        .   .   .   .   .   .   .   .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># OR4F5          .   .   .   .   .   .   .   .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># RP11-34P13.7   .   .   .   .   .   .   .   .</span>
</span></span><span style="display:flex;"><span>seurat_obj@misc$PBMC$wgcna_metacell_obj@meta.data %&gt;% head
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     orig.ident nCount_RNA nFeature_RNA RNA_snn_res.0.1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 0_1          0    2618.96         5880               0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 0_2          0    2486.00         5858               0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 0_3          0    1340.56         4306               0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 0_4          0    2112.12         5388               0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 0_5          0    2839.04         6073               0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 0_6          0    2198.44         5407               0</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="5共表达网络分析">5、共表达网络分析<a hidden class="anchor" aria-hidden="true" href="#5共表达网络分析">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1) 选择一种/多种细胞类型，进行WGCNA分析</span>
</span></span><span style="display:flex;"><span>seurat_obj &lt;- SetDatExpr(
</span></span><span style="display:flex;"><span>  seurat_obj,
</span></span><span style="display:flex;"><span>  group_name = c(<span style="color:#ff0;font-weight:bold">0</span>,<span style="color:#ff0;font-weight:bold">1</span>,<span style="color:#ff0;font-weight:bold">2</span>,<span style="color:#ff0;font-weight:bold">3</span>),
</span></span><span style="display:flex;"><span>  group.by=<span style="color:#0ff;font-weight:bold">&#39;RNA_snn_res.0.1&#39;</span> <span style="color:#007f7f"># the metadata column containing the cell type info. This same column should have also been used in MetacellsByGroups</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) 选择合适的软阈值</span>
</span></span><span style="display:flex;"><span>seurat_obj &lt;- TestSoftPowers(
</span></span><span style="display:flex;"><span>  seurat_obj,
</span></span><span style="display:flex;"><span>  setDatExpr = <span style="color:#fff;font-weight:bold">FALSE</span>, <span style="color:#007f7f"># set this to FALSE since we did this above</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>plot_list &lt;- PlotSoftPowers(seurat_obj)
</span></span><span style="display:flex;"><span>wrap_plots(plot_list, ncol=<span style="color:#ff0;font-weight:bold">2</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># power_table &lt;- GetPowerTable(seurat_obj)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># head(power_table)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220707142901974.png" alt="image-20220707142901974" style="zoom:50%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(3) 根据上图选择合适的软阈值构建网络</span>
</span></span><span style="display:flex;"><span>seurat_obj &lt;- ConstructNetwork(
</span></span><span style="display:flex;"><span>  seurat_obj, soft_power=<span style="color:#ff0;font-weight:bold">8</span>,
</span></span><span style="display:flex;"><span>  setDatExpr=<span style="color:#fff;font-weight:bold">FALSE</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>PlotDendrogram(seurat_obj, main=<span style="color:#0ff;font-weight:bold">&#39;INH hdWGCNA Dendrogram&#39;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220707143124392.png" alt="image-20220707143124392" style="zoom:50%;" />
<h2 id="6模块特征值与hub基因分析">6、模块特征值与hub基因分析<a hidden class="anchor" aria-hidden="true" href="#6模块特征值与hub基因分析">#</a></h2>
<ul>
<li>（1）模块组成基因</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>seurat_obj@misc$PBMC$wgcna_modules %&gt;% head
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#         gene_name    module     color</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NOC2L       NOC2L      grey      grey</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># HES4         HES4 turquoise turquoise</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ISG15       ISG15 turquoise turquoise</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TNFRSF4   TNFRSF4      blue      blue</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SDF4         SDF4      blue      blue</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># UBE2J2     UBE2J2      blue      blue</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>table(seurat_obj@misc$PBMC$wgcna_modules$module)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># grey turquoise      blue     brown    yellow</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  588      1786      1272       133        76</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># modules &lt;- GetModules(seurat_obj)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># head(modules)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（2）每个细胞对于每个模块的特征值</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>seurat_obj &lt;- ModuleEigengenes(
</span></span><span style="display:flex;"><span> seurat_obj
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>seurat_obj@misc$PBMC$MEs %&gt;% head
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                        blue       grey      brown turquoise     yellow</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATACAACCAC-1   4.755930  0.2424180  2.6565029 -5.832575 -2.0906948</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATTGAGCTAC-1   2.388492  1.3106515 -0.3341197  1.172415  8.8764667</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATTGATCAGC-1   5.807490  2.9008702  0.8873175 -1.132285 -1.8973713</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCGTGCTTCCG-1  -6.662387 -0.3072945 -1.5451723 14.692779  0.1159869</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCGTGTATGCG-1 -16.098322 -1.6676024  8.7666564 -4.131657 -0.9371790</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACGCACTGGTAC-1   2.064300 -0.1415680 -1.2313300 -2.824847 -1.1698044</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MEs &lt;- GetMEs(seurat_obj, harmonized=FALSE)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># head(MEs)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（3）鉴定模块内hub基因&ndash; 基因表达与模块特征值的具有高相关性</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">43
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">44
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">45
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">46
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">47
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">48
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">49
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">50
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>seurat_obj &lt;- ModuleConnectivity(
</span></span><span style="display:flex;"><span>  seurat_obj
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>seurat_obj@misc$PBMC$wgcna_modules[,c(<span style="color:#ff0;font-weight:bold">-1</span>,<span style="color:#ff0;font-weight:bold">-2</span>,<span style="color:#ff0;font-weight:bold">-3</span>)] %&gt;% head
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            kME_blue   kME_grey   kME_brown kME_turquoise   kME_yellow</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NOC2L    0.02851618 0.08360662  0.02041163   0.003048238  0.038497122</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># HES4    -0.24225802 0.06869209 -0.08284058   0.117744840  0.326692215</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ISG15   -0.29705821 0.08027215 -0.07240235   0.351103776  0.314836382</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TNFRSF4  0.16100218 0.08589770  0.13010946  -0.002353201 -0.090182722</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SDF4     0.01120696 0.13914234  0.07918151   0.040070556  0.003869467</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># UBE2J2   0.03178884 0.07651548  0.04663831   0.008615164 -0.006942615</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># p &lt;- PlotKMEs(seurat_obj, ncol=2)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># p</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>hub_df &lt;- GetHubGenes(seurat_obj, n_hubs = <span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span>hub_df
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   gene_name    module       kME</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1     RPS27      blue 0.7207118</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2     RPS12      blue 0.7258250</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3      CCL5     brown 0.4934812</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4       B2M     brown 0.5294537</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5    LGALS1 turquoise 0.6477877</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6    TYROBP turquoise 0.6542149</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 7   HLA-DRA    yellow 0.5817169</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 8  HLA-DPA1    yellow 0.5894472</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 计算每个细胞对于每个模块hub基因的表达活性(module score),可使用seurat包或者Ucell包</span>
</span></span><span style="display:flex;"><span>seurat_obj &lt;- ModuleExprScore(
</span></span><span style="display:flex;"><span>  seurat_obj,
</span></span><span style="display:flex;"><span>  n_genes = <span style="color:#ff0;font-weight:bold">25</span>,
</span></span><span style="display:flex;"><span>  method=<span style="color:#0ff;font-weight:bold">&#39;Seurat&#39;</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>seurat_obj@misc$PBMC$module_scores %&gt;% head
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 计算每个细胞对于每个模块hub基因的表达活性(module score),可使用seurat包或者Ucell包</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                      blue      brown turquoise      yellow</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATACAACCAC-1 3.380281 0.74403040 0.3806462 -0.18699177</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATTGAGCTAC-1 3.035111 0.31949694 0.4019337  1.06061130</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACATTGATCAGC-1 2.769994 0.28652286 0.5089602 -0.30358276</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCGTGCTTCCG-1 2.080403 0.07190668 2.6016017  0.45921797</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACCGTGTATGCG-1 1.434566 1.07243443 0.9849420  0.14309843</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AAACGCACTGGTAC-1 2.999822 0.25594526 0.6041499 -0.02860217</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># library(UCell)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># seurat_obj &lt;- ModuleExprScore(</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   seurat_obj,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   n_genes = 25,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   method=&#39;UCell&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># )</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># seurat_obj@misc$PBMC$module_scores %&gt;% head</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="7结果可视化">7、结果可视化<a hidden class="anchor" aria-hidden="true" href="#7结果可视化">#</a></h2>
<ul>
<li>每个细胞对于每个模块的特征值</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>plot_list &lt;- ModuleFeaturePlot(
</span></span><span style="display:flex;"><span>  seurat_obj,
</span></span><span style="display:flex;"><span>  features=<span style="color:#0ff;font-weight:bold">&#39;MEs&#39;</span>, <span style="color:#007f7f"># plot the hMEs</span>
</span></span><span style="display:flex;"><span>  order=<span style="color:#fff;font-weight:bold">TRUE</span> <span style="color:#007f7f"># order so the points with highest hMEs are on top</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>wrap_plots(plot_list, ncol=<span style="color:#ff0;font-weight:bold">2</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220707144607036.png" alt="image-20220707144607036" style="zoom:50%;" />
<ul>
<li>每个细胞对于每个模块hub基因的表达活性</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>seurat_obj@misc$PBMC$module_scores %&gt;% head
</span></span><span style="display:flex;"><span>plot_list &lt;- ModuleFeaturePlot(
</span></span><span style="display:flex;"><span>  seurat_obj,
</span></span><span style="display:flex;"><span>  features=<span style="color:#0ff;font-weight:bold">&#39;scores&#39;</span>, <span style="color:#007f7f"># plot the hub gene scores</span>
</span></span><span style="display:flex;"><span>  order=<span style="color:#0ff;font-weight:bold">&#39;shuffle&#39;</span>, <span style="color:#007f7f"># order so cells are shuffled</span>
</span></span><span style="display:flex;"><span>  ucell = <span style="color:#fff;font-weight:bold">TRUE</span> <span style="color:#007f7f"># depending on Seurat vs UCell for gene scoring</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>wrap_plots(plot_list, ncol=<span style="color:#ff0;font-weight:bold">2</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220707144654538.png" alt="image-20220707144654538" style="zoom:50%;" />
<ul>
<li>将细胞对于模块的特征值，整合到seurat的meta.data中</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>MEs &lt;- GetMEs(seurat_obj)
</span></span><span style="display:flex;"><span>mods &lt;- colnames(MEs); mods &lt;- mods[mods != <span style="color:#0ff;font-weight:bold">&#39;grey&#39;</span>]
</span></span><span style="display:flex;"><span>seurat_obj@meta.data &lt;- cbind(seurat_obj@meta.data, MEs)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>p &lt;- DotPlot(seurat_obj, features = mods, group.by = <span style="color:#0ff;font-weight:bold">&#39;RNA_snn_res.0.1&#39;</span>)
</span></span><span style="display:flex;"><span>p &lt;- p +
</span></span><span style="display:flex;"><span>  coord_flip() +
</span></span><span style="display:flex;"><span>  RotatedAxis() +
</span></span><span style="display:flex;"><span>  scale_color_gradient2(high=<span style="color:#0ff;font-weight:bold">&#39;red&#39;</span>, mid=<span style="color:#0ff;font-weight:bold">&#39;grey95&#39;</span>, low=<span style="color:#0ff;font-weight:bold">&#39;blue&#39;</span>)
</span></span><span style="display:flex;"><span>p
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220707145405227.png" alt="image-20220707145405227" style="zoom:50%;" />


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/015%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat%E7%A9%BA%E9%97%B4%E8%BD%AC%E5%BD%95%E7%BB%84/">
    <span class="title">« Prev Page</span>
    <br>
    <span>使用Seurat包分析空间转录组数据</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/017%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--ecaugt%E6%8F%90%E5%8F%96heca%E6%95%B0%E6%8D%AE/">
    <span class="title">Next Page »</span>
    <br>
    <span>单细胞分析工具--ECAUGT提取hECA数据</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
