<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--RNAvelocity速率分析 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
RNA velocity(RNA速率)分析是基于单细胞转录组测序数据分析细胞发育状态动力学的方法。简单来说，该模型根据测序read片段属于unspliced pre-mrna以及spliced mature mrna的比例推测单细胞的发育轨迹。 高比例的unspliced pre-mrna的占比越高，velocity速率越大，表明在之后阶段中将产生高表达趋势。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/024%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--rnavelocity%E9%80%9F%E7%8E%87%E5%88%86%E6%9E%90/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/024%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--rnavelocity%E9%80%9F%E7%8E%87%E5%88%86%E6%9E%90/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--RNAvelocity速率分析" />
<meta property="og:description" content="
RNA velocity(RNA速率)分析是基于单细胞转录组测序数据分析细胞发育状态动力学的方法。简单来说，该模型根据测序read片段属于unspliced pre-mrna以及spliced mature mrna的比例推测单细胞的发育轨迹。 高比例的unspliced pre-mrna的占比越高，velocity速率越大，表明在之后阶段中将产生高表达趋势。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/024%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--rnavelocity%E9%80%9F%E7%8E%87%E5%88%86%E6%9E%90/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-01-26T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-01-26T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--RNAvelocity速率分析"/>
<meta name="twitter:description" content="
RNA velocity(RNA速率)分析是基于单细胞转录组测序数据分析细胞发育状态动力学的方法。简单来说，该模型根据测序read片段属于unspliced pre-mrna以及spliced mature mrna的比例推测单细胞的发育轨迹。 高比例的unspliced pre-mrna的占比越高，velocity速率越大，表明在之后阶段中将产生高表达趋势。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--RNAvelocity速率分析",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/024%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--rnavelocity%E9%80%9F%E7%8E%87%E5%88%86%E6%9E%90/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--RNAvelocity速率分析",
  "name": "单细胞分析工具--RNAvelocity速率分析",
  "description": " RNA velocity(RNA速率)分析是基于单细胞转录组测序数据分析细胞发育状态动力学的方法。简单来说，该模型根据测序read片段属于unspliced pre-mrna以及spliced mature mrna的比例推测单细胞的发育轨迹。 高比例的unspliced pre-mrna的占比越高，velocity速率越大，表明在之后阶段中将产生高表达趋势。\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " RNA velocity(RNA速率)分析是基于单细胞转录组测序数据分析细胞发育状态动力学的方法。简单来说，该模型根据测序read片段属于unspliced pre-mrna以及spliced mature mrna的比例推测单细胞的发育轨迹。 高比例的unspliced pre-mrna的占比越高，velocity速率越大，表明在之后阶段中将产生高表达趋势。\n如下将主要学习分析流程，算法原理可参考原始论文：\n2018 | Natue | DOI: 10.1038/s41586-018-0414-6 2020 | Nature Biotechnology |DOI: 10.1038/s41587-020-0591-3 分析流程参考教程：\n上游比对：https://velocyto.org/velocyto.py/index.html 下游分析：https://smorabit.github.io/tutorials/8_velocyto/ 1、上游比对 单细胞测序技术有很多，如下以常用的10X genomics的测序结果为例。 首先根据原始测序结果fastq.gz，经10X官方的cellranger软件进行比对（使用方法参考之前笔记） 根据filtered_feature_bc_matrix 文件夹结果使用Seurat包进行单细胞常规分析； 根据比对的bam文件，使用velocyto软件分别识别出比对到unspliced与spliced mRNA。 示例数据是GSE178911中的GSM5400792样本，为人体急性髓系白血病（AML）骨髓细胞。已经过cellranger比对，结果保存在201008_D0_scRNA_fastq文件夹。\n1 2 3 4 5 6 7 8 9 10 11 12 13 tree -L 1 ./201008_D0_scRNA_fastq/outs/ # ./201008_D0_scRNA_fastq/outs/ # ├── cellsorted_possorted_genome_bam.bam # ├── filtered_feature_bc_matrix # ├── filtered_feature_bc_matrix.h5 # ├── filtered_feature_bc_matrix.tar.gz # ├── metrics_summary.csv # ├── molecule_info.h5 # ├── possorted_genome_bam.bam # ├── possorted_genome_bam.bam.bai # ├── raw_feature_bc_matrix # ├── raw_feature_bc_matrix.h5 # └── web_summary.html 1.1 velocyto分析 1 2 3 4 5 6 7 8 9 10 # (1) 参考下面链接，建立软件环境 # https://velocyto.org/velocyto.py/install/index.html # (2) 准备三份数据 ## 10X比对结果文件夹 cellranger_outDir=/path/to/201008_D8Ven_scRNA_fastq ## 10X参考数据集中的gtf文件 cellranger_gtf=/path/to/refdata-gex-GRCh38-2020-A/genes/genes.gtf ## 重复序列注释文件(optiona;) rmsk_gtf=/path/to/hg38_repeat_rmsk.gtf rmsk_gtf下载链接：https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=611454127_NtvlaW6xBSIRYJEBI0iRDEWisITa\u0026clade=mammal\u0026org=Human\u0026db=0\u0026hgta_group=allTracks\u0026hgta_track=rmsk\u0026hgta_table=rmsk\u0026hgta_regionType=genome\u0026position=\u0026hgta_outputType=gff\u0026hgta_outFileName=mm10_rmsk.gtf\n1 2 3 4 5 6 7 # (3) 开始分析 velocyto run10x -m $rmsk_gtf $cellranger_outDir $cellranger_gtf # (4) 分析结果 tree ./201008_D0_scRNA_fastq/velocyto/ # ./201008_D0_scRNA_fastq/velocyto/ # └── 201008_D0_scRNA_fastq.loom 如上，每个样本都会产生一个loom文件结果。\n1.2 Seurat分析 使用Seurat包完成基本分析，主要包括UMAP降维与分群两步 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 library(Seurat) library(tidyverse) counts = Read10X(\"filtered_feature_bc_matrix/\") sce = CreateSeuratObject(counts) dim(sce) # [1] 36601 6005 sce = sce %\u003e% NormalizeData() %\u003e% FindVariableFeatures(nfeatures = 2000) %\u003e% ScaleData() sce = sce %\u003e% RunPCA() %\u003e% RunUMAP(dims = 1:30) %\u003e% #RunTSNE FindNeighbors(dims = 1:30) %\u003e% FindClusters(resolution = c(0.5)) table(sce$seurat_clusters) # 0 1 2 3 4 5 6 7 8 9 # 1250 1077 827 621 599 541 397 386 215 92 DimPlot(sce) saveRDS(sce, file = \"sce.rds\") 2、下游分析 以下在jupyter notebook中执行\n2.1 读入数据 参考之前scanpy学习笔记，导入Seurat对象为Anndata 1 2 3 4 5 6 7 8 9 ## 准备Python环境 import scanpy as sc import numpy as np import anndata2ri # Activate the anndata2ri conversion between SingleCellExperiment and AnnData anndata2ri.activate() #Loading the rpy2 extension enables cell magic to be used #This runs R code in jupyter notebook cells %load_ext rpy2.ipython 1 2 3 4 %%R library(Seurat) sce = readRDS(\"./201008_D0_scRNA_fastq/sce.rds\") sce 1 2 3 4 %%R -o adata #convert the Seurat object to a SingleCellExperiment object adata = as.SingleCellExperiment(sce) adata 1 2 3 4 5 adata # AnnData object with n_obs × n_vars = 6005 × 36601 # obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.5', 'seurat_clusters', 'ident' # obsm: 'X_pca', 'X_umap' # layers: 'logcounts' 读入velocyto分析的loom文件 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 import scvelo as scv import scanpy as sc import numpy as np import pandas as pd import anndata as ad ldata1 = scv.read('./201008_D0_scRNA_fastq/velocyto/201008_D0_scRNA_fastq.loom') # AnnData object with n_obs × n_vars = 6005 × 36601 # var: 'Accession', 'Chromosome', 'End', 'Start', 'Strand' # layers: 'matrix', 'ambiguous', 'spliced', 'unspliced' ## 修改细胞名barcode，与Seurat保持一致 barcodes = [bc.split(':')[1] for bc in ldata1.obs.index.tolist()] barcodes = [bc[0:len(bc)-1] + '-1' for bc in barcodes] barcodes # ['AACAAGACATGCGGTC-1', # 'AACAGGGCACGACCTG-1', # 'AAAGAACGTTATGTGC-1', # 'AACAACCAGCAATAAC-1', # 'AACAAGAGTCTCTCCA-1'] ldata1.obs.index = barcodes ldata1.var_names_make_unique() 合并上述两个数据 1 2 3 4 5 6 7 8 9 adata = scv.utils.merge(adata, ldata1) adata # AnnData object with n_obs × n_vars = 6005 × 36591 # obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.5', 'seurat_clusters', 'ident', 'initial_size_unspliced', 'initial_size_spliced', 'initial_size' # var: 'Accession', 'Chromosome', 'End', 'Start', 'Strand' # obsm: 'X_pca', 'X_umap' # layers: 'logcounts', 'matrix', 'ambiguous', 'spliced', 'unspliced' sc.pl.umap(adata, color='seurat_clusters', legend_loc='on data') 2.2 scvelo速率分析 （1）统计每个cluster的spliced与unspliced比例 1 scv.pl.proportions(adata, groupby='seurat_clusters') （2）速率分析 1 2 3 4 5 6 7 8 9 10 11 12 13 scv.pp.filter_and_normalize(adata) scv.pp.moments(adata) scv.tl.velocity(adata, mode='stochastic') scv.tl.velocity_graph(adata) adata # AnnData object with n_obs × n_vars = 6005 × 36591 # obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.5', 'seurat_clusters', 'ident', 'initial_size_unspliced', 'initial_size_spliced', 'initial_size', 'n_counts' # var: 'Accession', 'Chromosome', 'End', 'Start', 'Strand', 'gene_count_corr', 'velocity_gamma', 'velocity_qreg_ratio', 'velocity_r2', 'velocity_genes' # uns: 'seurat_clusters_colors', 'neighbors', 'velocity_params' # obsm: 'X_pca', 'X_umap' # layers: 'logcounts', 'matrix', 'ambiguous', 'spliced', 'unspliced', 'Ms', 'Mu', 'velocity', 'variance_velocity' # obsp: 'distances', 'connectivities' （3）速率分析结果可视化 1 2 3 scv.pl.velocity_embedding_grid(adata, basis='umap', color='seurat_clusters', scale=0.25) # 可通过多个参数调整图形(箭头)细节展示 1 2 3 scv.pl.velocity_embedding_stream(adata, basis='umap', color='seurat_clusters') # 可通过多个参数调整图形(箭头)细节展示 1 2 3 4 5 6 scv.tl.rank_velocity_genes(adata, groupby='seurat_clusters', min_corr=.3) df = scv.DataFrame(adata.uns['rank_velocity_genes']['names']) df.head() scv.pl.scatter(adata, df['2'][:3], frameon=True, color='seurat_clusters', size=10, linewidth=1.5) 1 2 scv.tl.velocity_pseudotime(adata) # 由0到1 scv.pl.scatter(adata, color='velocity_pseudotime', cmap='gnuplot') 1 2 3 scv.tl.paga(adata, groups='seurat_clusters') scv.pl.paga(adata, basis='umap', size=50, alpha=.1, min_edge_width=2, node_size_scale=1.5) （4）特定细胞群速率分析 1 2 3 cur_celltypes = [\"0\"] adata_subset = adata[adata.obs['seurat_clusters'].isin(cur_celltypes)] sc.pl.umap(adata_subset) 1 2 3 4 5 6 sc.pp.neighbors(adata_subset, n_neighbors=15, use_rep='X_pca') scv.pp.filter_and_normalize(adata_subset) scv.pp.moments(adata_subset) scv.tl.recover_dynamics(adata_subset) scv.pl.velocity_embedding_stream(adata_subset, basis='umap') 1 2 3 4 df = adata_subset.var df = df[(df['fit_likelihood'] \u003e .1) \u0026 df['velocity_genes'] == True] top_genes = adata_subset.var['fit_likelihood'].sort_values(ascending=False).index[:300] scv.pl.scatter(adata_subset, basis=top_genes[:15], ncols=5, frameon=True) ",
  "wordCount" : "1965",
  "inLanguage": "en",
  "datePublished": "2023-01-26T00:00:00Z",
  "dateModified": "2023-01-26T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/024%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--rnavelocity%E9%80%9F%E7%8E%87%E5%88%86%E6%9E%90/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--RNAvelocity速率分析
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-26 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-26&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-26&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1965&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e4%b8%8a%e6%b8%b8%e6%af%94%e5%af%b9" aria-label="1、上游比对">1、上游比对</a><ul>
                            
                    <li>
                        <a href="#11-velocyto%e5%88%86%e6%9e%90" aria-label="1.1 velocyto分析">1.1 velocyto分析</a></li>
                    <li>
                        <a href="#12-seurat%e5%88%86%e6%9e%90" aria-label="1.2 Seurat分析">1.2 Seurat分析</a></li></ul>
                    </li>
                    <li>
                        <a href="#2%e4%b8%8b%e6%b8%b8%e5%88%86%e6%9e%90" aria-label="2、下游分析">2、下游分析</a><ul>
                            
                    <li>
                        <a href="#21-%e8%af%bb%e5%85%a5%e6%95%b0%e6%8d%ae" aria-label="2.1 读入数据">2.1 读入数据</a></li>
                    <li>
                        <a href="#22-scvelo%e9%80%9f%e7%8e%87%e5%88%86%e6%9e%90" aria-label="2.2 scvelo速率分析">2.2 scvelo速率分析</a>
                    </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>RNA velocity(RNA速率)分析是基于单细胞转录组测序数据分析细胞发育状态动力学的方法。简单来说，该模型根据测序read片段属于<strong>unspliced pre-mrna</strong>以及<strong>spliced mature mrna</strong>的比例推测单细胞的发育轨迹。 高比例的<strong>unspliced pre-mrna</strong>的占比越高，velocity速率越大，表明在之后阶段中将产生高表达趋势。</p>
<p>如下将主要学习分析流程，算法原理可参考原始论文：</p>
<ul>
<li>2018 | Natue | <a href="https://doi.org/10.1038/s41586-018-0414-6">DOI: 10.1038/s41586-018-0414-6</a></li>
<li>2020 | Nature Biotechnology |<a href="https://doi.org/10.1038/s41587-020-0591-3">DOI: 10.1038/s41587-020-0591-3</a></li>
</ul>
<p>分析流程参考教程：</p>
<ul>
<li>上游比对：https://velocyto.org/velocyto.py/index.html</li>
<li>下游分析：https://smorabit.github.io/tutorials/8_velocyto/</li>
</ul>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/MSB-17-e10282-g002.jpg" alt="An external file that holds a picture, illustration, etc. Object name is MSB-17-e10282-g002.jpg" style="zoom: 80%;" /></blockquote>
<h2 id="1上游比对">1、上游比对<a hidden class="anchor" aria-hidden="true" href="#1上游比对">#</a></h2>
<ul>
<li>单细胞测序技术有很多，如下以常用的10X genomics的测序结果为例。</li>
<li>首先根据原始测序结果fastq.gz，经10X官方的cellranger软件进行比对（使用方法参考之前笔记）
<ul>
<li>根据<code>filtered_feature_bc_matrix</code> 文件夹结果使用Seurat包进行单细胞常规分析；</li>
<li>根据比对的bam文件，使用velocyto软件分别识别出比对到unspliced与spliced mRNA。</li>
</ul>
</li>
</ul>
<blockquote>
<p>示例数据是GSE178911中的GSM5400792样本，为人体急性髓系白血病（AML）骨髓细胞。已经过cellranger比对，结果保存在<code>201008_D0_scRNA_fastq</code>文件夹。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>tree -L <span style="color:#ff0;font-weight:bold">1</span> ./201008_D0_scRNA_fastq/outs/
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ./201008_D0_scRNA_fastq/outs/</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── cellsorted_possorted_genome_bam.bam</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── filtered_feature_bc_matrix</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── filtered_feature_bc_matrix.h5</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── filtered_feature_bc_matrix.tar.gz</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── metrics_summary.csv</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── molecule_info.h5</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── possorted_genome_bam.bam</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── possorted_genome_bam.bam.bai</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── raw_feature_bc_matrix</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ├── raw_feature_bc_matrix.h5</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># └── web_summary.html</span>
</span></span></code></pre></td></tr></table>
</div>
</div></blockquote>
<h3 id="11-velocyto分析">1.1 velocyto分析<a hidden class="anchor" aria-hidden="true" href="#11-velocyto分析">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f"># (1) 参考下面链接，建立软件环境</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># https://velocyto.org/velocyto.py/install/index.html</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2) 准备三份数据</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 10X比对结果文件夹 </span>
</span></span><span style="display:flex;"><span>cellranger_outDir=/path/to/201008_D8Ven_scRNA_fastq
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 10X参考数据集中的gtf文件</span>
</span></span><span style="display:flex;"><span>cellranger_gtf=/path/to/refdata-gex-GRCh38-2020-A/genes/genes.gtf
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 重复序列注释文件(optiona;)</span>
</span></span><span style="display:flex;"><span>rmsk_gtf=/path/to/hg38_repeat_rmsk.gtf
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>rmsk_gtf下载链接：https://genome.ucsc.edu/cgi-bin/hgTables?hgsid=611454127_NtvlaW6xBSIRYJEBI0iRDEWisITa&amp;clade=mammal&amp;org=Human&amp;db=0&amp;hgta_group=allTracks&amp;hgta_track=rmsk&amp;hgta_table=rmsk&amp;hgta_regionType=genome&amp;position=&amp;hgta_outputType=gff&amp;hgta_outFileName=mm10_rmsk.gtf</p>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-8a2896878adac63c.png" style="zoom: 50%;" /></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span><span style="color:#007f7f"># (3) 开始分析</span>
</span></span><span style="display:flex;"><span>velocyto run10x -m $rmsk_gtf  $cellranger_outDir $cellranger_gtf
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (4) 分析结果</span>
</span></span><span style="display:flex;"><span>tree ./201008_D0_scRNA_fastq/velocyto/
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ./201008_D0_scRNA_fastq/velocyto/</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># └── 201008_D0_scRNA_fastq.loom</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>如上，每个样本都会产生一个loom文件结果。</p></blockquote>
<h3 id="12-seurat分析">1.2 Seurat分析<a hidden class="anchor" aria-hidden="true" href="#12-seurat分析">#</a></h3>
<ul>
<li>使用Seurat包完成基本分析，主要包括UMAP降维与分群两步</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>counts = Read10X(<span style="color:#0ff;font-weight:bold">&#34;filtered_feature_bc_matrix/&#34;</span>)
</span></span><span style="display:flex;"><span>sce = CreateSeuratObject(counts)
</span></span><span style="display:flex;"><span>dim(sce)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 36601  6005</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>sce = sce %&gt;%
</span></span><span style="display:flex;"><span>  NormalizeData() %&gt;%
</span></span><span style="display:flex;"><span>  FindVariableFeatures(nfeatures = <span style="color:#ff0;font-weight:bold">2000</span>) %&gt;%
</span></span><span style="display:flex;"><span>  ScaleData()
</span></span><span style="display:flex;"><span>sce = sce %&gt;% 
</span></span><span style="display:flex;"><span>  RunPCA() %&gt;% 
</span></span><span style="display:flex;"><span>  RunUMAP(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;%  <span style="color:#007f7f">#RunTSNE</span>
</span></span><span style="display:flex;"><span>  FindNeighbors(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  FindClusters(resolution = c(<span style="color:#ff0;font-weight:bold">0.5</span>))
</span></span><span style="display:flex;"><span>table(sce$seurat_clusters)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#    0    1    2    3    4    5    6    7    8    9 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1250 1077  827  621  599  541  397  386  215   92 </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>DimPlot(sce)
</span></span><span style="display:flex;"><span>saveRDS(sce, file = <span style="color:#0ff;font-weight:bold">&#34;sce.rds&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126113814256.png" alt="image-20230126113814256" style="zoom: 50%;" />
<h2 id="2下游分析">2、下游分析<a hidden class="anchor" aria-hidden="true" href="#2下游分析">#</a></h2>
<blockquote>
<p>以下在jupyter notebook中执行</p></blockquote>
<h3 id="21-读入数据">2.1 读入数据<a hidden class="anchor" aria-hidden="true" href="#21-读入数据">#</a></h3>
<ul>
<li>参考之前scanpy学习笔记，导入Seurat对象为Anndata</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f">## 准备Python环境</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> scanpy <span style="color:#fff;font-weight:bold">as</span> sc
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> numpy <span style="color:#fff;font-weight:bold">as</span> np
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> anndata2ri
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Activate the anndata2ri conversion between SingleCellExperiment and AnnData</span>
</span></span><span style="display:flex;"><span>anndata2ri.activate()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#Loading the rpy2 extension enables cell magic to be used</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#This runs R code in jupyter notebook cells</span>
</span></span><span style="display:flex;"><span>%load_ext rpy2.ipython
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>%%R
</span></span><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>sce = readRDS(<span style="color:#0ff;font-weight:bold">&#34;./201008_D0_scRNA_fastq/sce.rds&#34;</span>)
</span></span><span style="display:flex;"><span>sce
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>%%R -o adata
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#convert the Seurat object to a SingleCellExperiment object</span>
</span></span><span style="display:flex;"><span>adata = <span style="color:#fff;font-weight:bold">as</span>.SingleCellExperiment(sce)
</span></span><span style="display:flex;"><span>adata
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>adata
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AnnData object with n_obs × n_vars = 6005 × 36601</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obs: &#39;orig.ident&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;RNA_snn_res.0.5&#39;, &#39;seurat_clusters&#39;, &#39;ident&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     layers: &#39;logcounts&#39;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>读入velocyto分析的loom文件</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> scvelo <span style="color:#fff;font-weight:bold">as</span> scv
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> scanpy <span style="color:#fff;font-weight:bold">as</span> sc
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> numpy <span style="color:#fff;font-weight:bold">as</span> np
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> pandas <span style="color:#fff;font-weight:bold">as</span> pd
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> anndata <span style="color:#fff;font-weight:bold">as</span> ad
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>ldata1 = scv.read(<span style="color:#0ff;font-weight:bold">&#39;./201008_D0_scRNA_fastq/velocyto/201008_D0_scRNA_fastq.loom&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AnnData object with n_obs × n_vars = 6005 × 36601</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     var: &#39;Accession&#39;, &#39;Chromosome&#39;, &#39;End&#39;, &#39;Start&#39;, &#39;Strand&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     layers: &#39;matrix&#39;, &#39;ambiguous&#39;, &#39;spliced&#39;, &#39;unspliced&#39;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 修改细胞名barcode，与Seurat保持一致</span>
</span></span><span style="display:flex;"><span>barcodes = [bc.split(<span style="color:#0ff;font-weight:bold">&#39;:&#39;</span>)[<span style="color:#ff0;font-weight:bold">1</span>] <span style="color:#fff;font-weight:bold">for</span> bc in ldata1.obs.index.tolist()]
</span></span><span style="display:flex;"><span>barcodes = [bc[<span style="color:#ff0;font-weight:bold">0</span>:<span style="color:#fff;font-weight:bold">len</span>(bc)-<span style="color:#ff0;font-weight:bold">1</span>] + <span style="color:#0ff;font-weight:bold">&#39;-1&#39;</span> <span style="color:#fff;font-weight:bold">for</span> bc in barcodes]
</span></span><span style="display:flex;"><span>barcodes
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;AACAAGACATGCGGTC-1&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;AACAGGGCACGACCTG-1&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;AAAGAACGTTATGTGC-1&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;AACAACCAGCAATAAC-1&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  &#39;AACAAGAGTCTCTCCA-1&#39;]</span>
</span></span><span style="display:flex;"><span>ldata1.obs.index = barcodes
</span></span><span style="display:flex;"><span>ldata1.var_names_make_unique()
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>合并上述两个数据</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>adata = scv.utils.merge(adata, ldata1)
</span></span><span style="display:flex;"><span>adata
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AnnData object with n_obs × n_vars = 6005 × 36591</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obs: &#39;orig.ident&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;RNA_snn_res.0.5&#39;, &#39;seurat_clusters&#39;, &#39;ident&#39;, &#39;initial_size_unspliced&#39;, &#39;initial_size_spliced&#39;, &#39;initial_size&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     var: &#39;Accession&#39;, &#39;Chromosome&#39;, &#39;End&#39;, &#39;Start&#39;, &#39;Strand&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     layers: &#39;logcounts&#39;, &#39;matrix&#39;, &#39;ambiguous&#39;, &#39;spliced&#39;, &#39;unspliced&#39;</span>
</span></span><span style="display:flex;"><span>sc.pl.umap(adata, color=<span style="color:#0ff;font-weight:bold">&#39;seurat_clusters&#39;</span>, 
</span></span><span style="display:flex;"><span>		   legend_loc=<span style="color:#0ff;font-weight:bold">&#39;on data&#39;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126125018274.png" alt="image-20230126125018274" style="zoom:67%;" />
<h3 id="22-scvelo速率分析">2.2 scvelo速率分析<a hidden class="anchor" aria-hidden="true" href="#22-scvelo速率分析">#</a></h3>
<ul>
<li>（1）统计每个cluster的spliced与unspliced比例</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>scv.pl.proportions(adata, groupby=<span style="color:#0ff;font-weight:bold">&#39;seurat_clusters&#39;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126125232129.png" alt="image-20230126125232129"  />
</p>
<ul>
<li>（2）速率分析</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>scv.pp.filter_and_normalize(adata)
</span></span><span style="display:flex;"><span>scv.pp.moments(adata)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>scv.tl.velocity(adata, mode=<span style="color:#0ff;font-weight:bold">&#39;stochastic&#39;</span>)
</span></span><span style="display:flex;"><span>scv.tl.velocity_graph(adata)
</span></span><span style="display:flex;"><span>adata
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AnnData object with n_obs × n_vars = 6005 × 36591</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obs: &#39;orig.ident&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;RNA_snn_res.0.5&#39;, &#39;seurat_clusters&#39;, &#39;ident&#39;, &#39;initial_size_unspliced&#39;, &#39;initial_size_spliced&#39;, &#39;initial_size&#39;, &#39;n_counts&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     var: &#39;Accession&#39;, &#39;Chromosome&#39;, &#39;End&#39;, &#39;Start&#39;, &#39;Strand&#39;, &#39;gene_count_corr&#39;, &#39;velocity_gamma&#39;, &#39;velocity_qreg_ratio&#39;, &#39;velocity_r2&#39;, &#39;velocity_genes&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     uns: &#39;seurat_clusters_colors&#39;, &#39;neighbors&#39;, &#39;velocity_params&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     layers: &#39;logcounts&#39;, &#39;matrix&#39;, &#39;ambiguous&#39;, &#39;spliced&#39;, &#39;unspliced&#39;, &#39;Ms&#39;, &#39;Mu&#39;, &#39;velocity&#39;, &#39;variance_velocity&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obsp: &#39;distances&#39;, &#39;connectivities&#39;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（3）速率分析结果可视化</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>scv.pl.velocity_embedding_grid(adata, basis=<span style="color:#0ff;font-weight:bold">&#39;umap&#39;</span>, color=<span style="color:#0ff;font-weight:bold">&#39;seurat_clusters&#39;</span>, 
</span></span><span style="display:flex;"><span>							   scale=<span style="color:#ff0;font-weight:bold">0.25</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 可通过多个参数调整图形(箭头)细节展示</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126130013583.png" alt="image-20230126130013583" style="zoom:67%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>scv.pl.velocity_embedding_stream(adata, basis=<span style="color:#0ff;font-weight:bold">&#39;umap&#39;</span>, 
</span></span><span style="display:flex;"><span>								 color=<span style="color:#0ff;font-weight:bold">&#39;seurat_clusters&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 可通过多个参数调整图形(箭头)细节展示</span>
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126130132801.png" alt="image-20230126130132801" style="zoom:67%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>scv.tl.rank_velocity_genes(adata, groupby=<span style="color:#0ff;font-weight:bold">&#39;seurat_clusters&#39;</span>, min_corr=<span style="color:#ff0;font-weight:bold">.3</span>)
</span></span><span style="display:flex;"><span>df = scv.DataFrame(adata.uns[<span style="color:#0ff;font-weight:bold">&#39;rank_velocity_genes&#39;</span>][<span style="color:#0ff;font-weight:bold">&#39;names&#39;</span>])
</span></span><span style="display:flex;"><span>df.head()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>scv.pl.scatter(adata, df[<span style="color:#0ff;font-weight:bold">&#39;2&#39;</span>][:<span style="color:#ff0;font-weight:bold">3</span>],  frameon=<span style="color:#fff;font-weight:bold">True</span>, 
</span></span><span style="display:flex;"><span>			   color=<span style="color:#0ff;font-weight:bold">&#39;seurat_clusters&#39;</span>, size=<span style="color:#ff0;font-weight:bold">10</span>, linewidth=<span style="color:#ff0;font-weight:bold">1.5</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126130540798.png" alt="image-20230126130540798"  />
</p>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126130631766.png" alt="image-20230126130631766" style="zoom: 67%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>scv.tl.velocity_pseudotime(adata)  <span style="color:#007f7f"># 由0到1</span>
</span></span><span style="display:flex;"><span>scv.pl.scatter(adata, color=<span style="color:#0ff;font-weight:bold">&#39;velocity_pseudotime&#39;</span>, cmap=<span style="color:#0ff;font-weight:bold">&#39;gnuplot&#39;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126131215163.png" alt="image-20230126131215163" style="zoom:67%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>scv.tl.paga(adata, groups=<span style="color:#0ff;font-weight:bold">&#39;seurat_clusters&#39;</span>)
</span></span><span style="display:flex;"><span>scv.pl.paga(adata, basis=<span style="color:#0ff;font-weight:bold">&#39;umap&#39;</span>, size=<span style="color:#ff0;font-weight:bold">50</span>, alpha=<span style="color:#ff0;font-weight:bold">.1</span>,
</span></span><span style="display:flex;"><span>            min_edge_width=<span style="color:#ff0;font-weight:bold">2</span>, node_size_scale=<span style="color:#ff0;font-weight:bold">1.5</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126131321685.png" alt="image-20230126131321685" style="zoom:67%;" />
<ul>
<li>（4）特定细胞群速率分析</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>cur_celltypes = [<span style="color:#0ff;font-weight:bold">&#34;0&#34;</span>]
</span></span><span style="display:flex;"><span>adata_subset = adata[adata.obs[<span style="color:#0ff;font-weight:bold">&#39;seurat_clusters&#39;</span>].isin(cur_celltypes)]
</span></span><span style="display:flex;"><span>sc.pl.umap(adata_subset)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126131634828.png" alt="image-20230126131634828" style="zoom:67%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>sc.pp.neighbors(adata_subset, n_neighbors=<span style="color:#ff0;font-weight:bold">15</span>, use_rep=<span style="color:#0ff;font-weight:bold">&#39;X_pca&#39;</span>)
</span></span><span style="display:flex;"><span>scv.pp.filter_and_normalize(adata_subset)
</span></span><span style="display:flex;"><span>scv.pp.moments(adata_subset)
</span></span><span style="display:flex;"><span>scv.tl.recover_dynamics(adata_subset)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>scv.pl.velocity_embedding_stream(adata_subset, basis=<span style="color:#0ff;font-weight:bold">&#39;umap&#39;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126132125090.png" alt="image-20230126132125090" style="zoom:67%;" />
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>df = adata_subset.var
</span></span><span style="display:flex;"><span>df = df[(df[<span style="color:#0ff;font-weight:bold">&#39;fit_likelihood&#39;</span>] &gt; <span style="color:#ff0;font-weight:bold">.1</span>) &amp; df[<span style="color:#0ff;font-weight:bold">&#39;velocity_genes&#39;</span>] == <span style="color:#fff;font-weight:bold">True</span>]
</span></span><span style="display:flex;"><span>top_genes = adata_subset.var[<span style="color:#0ff;font-weight:bold">&#39;fit_likelihood&#39;</span>].sort_values(ascending=<span style="color:#fff;font-weight:bold">False</span>).index[:<span style="color:#ff0;font-weight:bold">300</span>]
</span></span><span style="display:flex;"><span>scv.pl.scatter(adata_subset,  basis=top_genes[:<span style="color:#ff0;font-weight:bold">15</span>], ncols=<span style="color:#ff0;font-weight:bold">5</span>, frameon=<span style="color:#fff;font-weight:bold">True</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230126132228007.png" alt="image-20230126132228007"  />
</p>


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/019%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--singler%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B%E6%B3%A8%E9%87%8A/">
    <span class="title">« Prev Page</span>
    <br>
    <span>单细胞分析工具--SingleR细胞类型注释</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/program/020python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6-ipython%E9%AD%94%E6%B3%95%E5%91%BD%E4%BB%A4/">
    <span class="title">Next Page »</span>
    <br>
    <span>Python数据科学-(1)IPython的特点</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
