<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--ShinyCell交互式展示单细胞数据 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
ShinyCell包是由杜克-新加坡国立大学医学院的John F. Ouyang团队开发的单细胞分析工具包，实现基于shiny网页交互式展示单细胞数据；于2021年3月发表于Bioinformatics杂志。如文章中介绍，ShinyCell相比同类工具具有多个优势，例如直观的side-by-side的降维可视化方式，hdf5格式保存表达矩阵从而读取快速，支持pdf/png保存图片，支持多种常见单细胞数据类型等。参考其教程文档，学习记录如下。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/040%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--shinycell%E4%BA%A4%E4%BA%92%E5%BC%8F%E5%B1%95%E7%A4%BA%E5%8D%95%E7%BB%86%E8%83%9E%E6%95%B0%E6%8D%AE/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/040%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--shinycell%E4%BA%A4%E4%BA%92%E5%BC%8F%E5%B1%95%E7%A4%BA%E5%8D%95%E7%BB%86%E8%83%9E%E6%95%B0%E6%8D%AE/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--ShinyCell交互式展示单细胞数据" />
<meta property="og:description" content="
ShinyCell包是由杜克-新加坡国立大学医学院的John F. Ouyang团队开发的单细胞分析工具包，实现基于shiny网页交互式展示单细胞数据；于2021年3月发表于Bioinformatics杂志。如文章中介绍，ShinyCell相比同类工具具有多个优势，例如直观的side-by-side的降维可视化方式，hdf5格式保存表达矩阵从而读取快速，支持pdf/png保存图片，支持多种常见单细胞数据类型等。参考其教程文档，学习记录如下。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/040%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--shinycell%E4%BA%A4%E4%BA%92%E5%BC%8F%E5%B1%95%E7%A4%BA%E5%8D%95%E7%BB%86%E8%83%9E%E6%95%B0%E6%8D%AE/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-08-07T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-08-07T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--ShinyCell交互式展示单细胞数据"/>
<meta name="twitter:description" content="
ShinyCell包是由杜克-新加坡国立大学医学院的John F. Ouyang团队开发的单细胞分析工具包，实现基于shiny网页交互式展示单细胞数据；于2021年3月发表于Bioinformatics杂志。如文章中介绍，ShinyCell相比同类工具具有多个优势，例如直观的side-by-side的降维可视化方式，hdf5格式保存表达矩阵从而读取快速，支持pdf/png保存图片，支持多种常见单细胞数据类型等。参考其教程文档，学习记录如下。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--ShinyCell交互式展示单细胞数据",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/040%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--shinycell%E4%BA%A4%E4%BA%92%E5%BC%8F%E5%B1%95%E7%A4%BA%E5%8D%95%E7%BB%86%E8%83%9E%E6%95%B0%E6%8D%AE/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--ShinyCell交互式展示单细胞数据",
  "name": "单细胞分析工具--ShinyCell交互式展示单细胞数据",
  "description": " ShinyCell包是由杜克-新加坡国立大学医学院的John F. Ouyang团队开发的单细胞分析工具包，实现基于shiny网页交互式展示单细胞数据；于2021年3月发表于Bioinformatics杂志。如文章中介绍，ShinyCell相比同类工具具有多个优势，例如直观的side-by-side的降维可视化方式，hdf5格式保存表达矩阵从而读取快速，支持pdf/png保存图片，支持多种常见单细胞数据类型等。参考其教程文档，学习记录如下。\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " ShinyCell包是由杜克-新加坡国立大学医学院的John F. Ouyang团队开发的单细胞分析工具包，实现基于shiny网页交互式展示单细胞数据；于2021年3月发表于Bioinformatics杂志。如文章中介绍，ShinyCell相比同类工具具有多个优势，例如直观的side-by-side的降维可视化方式，hdf5格式保存表达矩阵从而读取快速，支持pdf/png保存图片，支持多种常见单细胞数据类型等。参考其教程文档，学习记录如下。\nPaper | ShinyCell: simple and sharable visualization of single-cell gene expression data DOI | https://doi.org/10.1093/bioinformatics/btab209 Github | https://github.com/SGDDNB/ShinyCell 1. 安装R包 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 # (1) 安装ShinyCell所需的依赖包 reqPkg = c(\"data.table\", \"Matrix\", \"hdf5r\", \"reticulate\", \"ggplot2\", \"gridExtra\", \"glue\", \"readr\", \"RColorBrewer\", \"R.utils\", \"Seurat\") newPkg = reqPkg[!(reqPkg %in% installed.packages()[,\"Package\"])] if(length(newPkg)){install.packages(newPkg)} # (2) 安装运行shiny所需的R包 reqPkg = c(\"shiny\", \"shinyhelper\", \"data.table\", \"Matrix\", \"DT\", \"hdf5r\", \"reticulate\", \"ggplot2\", \"gridExtra\", \"magrittr\", \"ggdendro\") newPkg = reqPkg[!(reqPkg %in% installed.packages()[,\"Package\"])] if(length(newPkg)){install.packages(newPkg)} # (3) 最后安装ShinyCell包 # github直接安装 devtools::install_github(\"SGDDNB/ShinyCell\") # 源文件本地安装 devtools::install_local(\"ShinyCell-master.zip\") ShinyCell支持多种单细胞数据类型，包括Seurat， SCE(singlecellexperiment), h5ad, loom；并均提供了相应的示例文件； 如其文档所强调，ShinyCell是一个可视化工具，而不是分析工具；所以提供的单细胞数据需要已经完成基础的上游分析； 下面主要以Seurat类型为例进行学习 1 2 3 library(Seurat) library(ShinyCell) library(shiny) 2. 基础用法 2.1 示例数据\n1 2 3 # getExampleData() # ~200MB # http://files.ddnetbio.com/hrpiFiles/readySeu_rset.rds seu = readRDS(\"readySeu_rset.rds\") 单细胞数据里需包括（1）标准化表达矩阵；（2）细胞meta信息；（3）降维信息。\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 ### （1）标准化表达矩阵 dim(seu) # [1] 12590 9852 Assays(seu) # [1] \"RNA\" seu@assays$RNA@data[1:4,1:4] #注意是ENSEMBL基因ID # 4 x 4 sparse Matrix of class \"dgCMatrix\" # FM_AAACCTGAGAAACGCC-1 FM_AAACCTGAGTGGAGTC-1 FM_AAACCTGCATAGAAAC-1 FM_AAACCTGTCATTCACT-1 # ENSG00000000003 . . 0.8413237 . # ENSG00000000419 . . . . # ENSG00000000457 . . . . # ENSG00000000460 . . . . ### （2）细胞meta信息 head(t(seu@meta.data[1,]),10) # FM_AAACCTGAGAAACGCC-1 # orig.ident \"FM\" # RNA_snn_res.0.5 \"5\" # library \"FM\" # nUMI \"13798\" # nGene \"3522\" # pctMT \"1.232063\" # pctHK \"20.77113\" # score_S \"-0.2176265\" # score_G2M \"-0.3646008\" # phase \"G1\" ### （3）降维信息 names(seu@reductions) # [1] \"pca\" \"tsne\" \"umap\" \"diffmap\" \"fdl\" head(Embeddings(seu, reduction = \"umap\")) # UMAP_1 UMAP_2 # FM_AAACCTGAGAAACGCC-1 -7.728941 -3.2370806 # FM_AAACCTGAGTGGAGTC-1 -7.048894 2.8379904 # FM_AAACCTGCATAGAAAC-1 -7.768705 1.6628623 # FM_AAACCTGTCATTCACT-1 -6.059119 3.3121483 # FM_AAACCTGTCTCTGAGA-1 -8.030442 1.5388506 # FM_AAACGGGCAGGCGATA-1 -9.831629 -0.5126803 2.2 生成shiny网页\n（1）生成meta配置文件\n1 2 3 4 5 6 7 8 9 10 11 12 scConf = createConfig(seu) head(scConf) # 每行对应一项meta的title、level name、颜色方案等信息 t(scConf[1,]) # [,1] # ID \"orig.ident\" # UI \"orig.ident\" # fID \"FM|NR|PR|RR\" # fUI \"FM|NR|PR|RR\" # fCL \"#A6CEE3|#B89B74|#ED8F47|#B15928\" # fRow \"1\" # default \"1\" # grp \"TRUE\" ID, fID 对应原始数据；UI, fUI, fCL 分布表示在shiny中展示的meta title，level name，colors\nfROW表示 legend的行数，default 表示是否为默认展示的两种cell info，grp表示是否为分组类型\n（2）生成shiny配置文件\n1 2 3 4 5 6 7 8 9 ?makeShinyApp makeShinyApp(seu, scConf, gene.mapping = TRUE, # default FALSE, only when ENSEMBL gene id gex.assay = \"RNA\", gex.slot = \"data\", shiny.title = \"scRNA-seq shiny app\", shiny.dir = \"shinyApp/\") # 查看结果 list.files(\"shinyApp\") # [1] \"sc1conf.rds\" \"sc1def.rds\" \"sc1gene.rds\" \"sc1gexpr.h5\" \"sc1meta.rds\" \"server.R\" \"ui.R\" 除了其中的gene.mapping参数（需要在单细胞数据的基因名为ENSEMBL格式时设置为TRUE），其余均为默认参数； 此外还有其它参数，包括默认展示的gene，降维类型；shiny.footnotes参数还支持设置网页的脚注用以交代数据集来源。 （3）最后运行shiny网页即可\n1 2 # 在本地浏览器的3344端口打开 runApp(\"shinyApp/\", launch.browser = T, port = 3344) 3. 模块介绍 （1）完成上述步骤后即可在本地生成一个shiny网页，下面结合作者分享的示例http://shinycell1.ddnetbio.com/介绍其内容；\n在每个模块中，均有3个Toggle选项支持筛选细胞，调整部分绘图参数等。\n（2）如上图所示，ShinyCell主要支持7个可视化模块；若从绘图类型来说共是6种图\n模块 图形 含义 e.g. 1 CellInfo vs GeneExpr 降维图 对比可视化细胞的分群信息与表达信息 A+B 2 CellInfo vs CellInfo 降维图 对比可视化细胞的两种不同分群信息 A+A 3 GeneExpr vs GeneExpr 降维图 对比可视化细胞的两种不同表达信息 B+B 4 Gene coexpression 降维图 可视化细胞的两种不同表达信息 C 5 Violinplot / Boxplot 小提琴图/箱图 可视化特定分群下，细胞的表达分布 D 6 Bubbleplot / Heatmap 点图/热图 可视化特定分群下，细胞的多个表达分布 F 7 Proportion plot 柱状图 可视化特定分群下，细胞的另一种分群的分布 E 4. 进阶用法 4.1 meta config调整 如上介绍，生成shiny网页的第一步是根据单细胞数据的meta信息生成相应的config文件，用以设置模块可视化的细节。\n默认情况下会使用全部的meta信息，如需调整一方面可直接修改原来的单细胞数据；另一方面也可以使用ShinyCell包进行部分修改，如下所示。\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 seu \u003c- readRDS(\"readySeu_rset.rds\") seu@meta.data = seu@meta.data[,1:6] scConf = createConfig(seu) # 展示当前的meta config legend showLegend(scConf) # (1)删除若干条细胞注释信息 scConf = delMeta(scConf, c(\"orig.ident\")) # (2)修改在shiny的细胞注释名title name(fID) scConf = modMetaName(scConf, meta.to.mod = c(\"nUMI\", \"nGene\"), new.name = c(\"No. UMIs\", \"No. detected genes\")) # (3)修改在shiny的细胞注释名level name(fUI) scConf = modLabels(scConf, meta.to.mod = \"library\", new.labels = c(\"fm\", \"pr\", \"nr\", \"rr\")) # (4)修改在shiny的细胞注释可视化颜色(fCL) scConf = modColours(scConf, meta.to.mod = \"library\", new.colours= RColorBrewer::brewer.pal(n = 4, name = \"Set1\")) showLegend(scConf) # showOrder(scConf) # scConf = reorderMeta(scConf, rev(scConf$ID)) # showOrder(scConf) 如需修改特定meta的level factor顺序，似乎只能在原始单细胞数据的meta中进行修改。\n4.2 多个数据集 ShinyCell支持在同一个网页内独立展示多个的数据集，可参考示例网页http://shinycell2.ddnetbio.com\n简单生成方法如下\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 ## 数据集1 seu \u003c- readRDS(\"readySeu_rset.rds\") scConf1 = createConfig(seu) makeShinyFiles(seu, scConf1, gene.mapping = TRUE, shiny.prefix = \"sc1\", # 前缀不同 shiny.dir = \"shinyAppMulti/\") ## 数据集2 scConf2 = delMeta(scConf1, c(\"orig.ident\")) makeShinyFiles(seu, scConf2, gene.mapping = TRUE, shiny.prefix = \"sc2\", # 前缀不同 gex.assay = \"RNA\", gex.slot = \"data\", shiny.dir = \"shinyAppMulti/\") ## 生成shiny配置文件 makeShinyCodesMulti( shiny.title = \"Multi-dataset Tutorial\", shiny.prefix = c(\"sc1\", \"sc2\"), shiny.headers = c(\"Dataset1\", \"Dataset2\"), shiny.footnotes = \"\", shiny.dir = \"shinyAppMulti/\") ## 运行网页 runApp(\"shinyAppMulti/\", launch.browser = T, port = 3344) 4.3 linux平台 对于大型单细胞数据集，需要在服务器linux平台分析时，大体与上述方式类似，简述流程如下。 1 2 3 4 5 6 7 8 9 10 11 12 13 14 # (1) 前期需在linux的R语言环境下安装相关R包，并准备相关数据 # (2) 生成shiny网页文件 seu \u003c- readRDS(\"readySeu_rset.rds\") scConf = createConfig(seu) makeShinyApp(seu, scConf, gene.mapping = TRUE, shiny.title = \"ShinyCell Quick Start\") # (3) linux端运行网页，注意需要设置launch.browser参数为FALSE runApp(\"shinyApp/\", launch.browser = F, port = 3344) # (4) 本地window端链接：同时win键+R，输入cmd打开命令窗口，根据下述命令进行链接 ssh -L localhost:3344:localhost:3344 username@192.168.212.xxx -p 22 # (5) 最后本地浏览器打开 http://127.0.0.1:3344/ 即可 ",
  "wordCount" : "2383",
  "inLanguage": "en",
  "datePublished": "2023-08-07T00:00:00Z",
  "dateModified": "2023-08-07T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/040%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--shinycell%E4%BA%A4%E4%BA%92%E5%BC%8F%E5%B1%95%E7%A4%BA%E5%8D%95%E7%BB%86%E8%83%9E%E6%95%B0%E6%8D%AE/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--ShinyCell交互式展示单细胞数据
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-08-07 00:00:00 &#43;0000 UTC&#39;&gt;2023-08-07&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-08-07&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2383&amp;nbsp;|&amp;nbsp;5 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1-%e5%ae%89%e8%a3%85r%e5%8c%85" aria-label="1. 安装R包">1. 安装R包</a></li>
                    <li>
                        <a href="#2-%e5%9f%ba%e7%a1%80%e7%94%a8%e6%b3%95" aria-label="2. 基础用法">2. 基础用法</a></li>
                    <li>
                        <a href="#3-%e6%a8%a1%e5%9d%97%e4%bb%8b%e7%bb%8d" aria-label="3. 模块介绍">3. 模块介绍</a></li>
                    <li>
                        <a href="#4-%e8%bf%9b%e9%98%b6%e7%94%a8%e6%b3%95" aria-label="4. 进阶用法">4. 进阶用法</a><ul>
                            
                    <li>
                        <a href="#41-meta-config%e8%b0%83%e6%95%b4" aria-label="4.1 meta config调整">4.1 meta config调整</a></li>
                    <li>
                        <a href="#42-%e5%a4%9a%e4%b8%aa%e6%95%b0%e6%8d%ae%e9%9b%86" aria-label="4.2 多个数据集">4.2 多个数据集</a></li>
                    <li>
                        <a href="#43-linux%e5%b9%b3%e5%8f%b0" aria-label="4.3 linux平台">4.3 linux平台</a>
                    </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>ShinyCell包是由杜克-新加坡国立大学医学院的John F. Ouyang团队开发的单细胞分析工具包，实现基于shiny网页交互式展示单细胞数据；于2021年3月发表于Bioinformatics杂志。如文章中介绍，ShinyCell相比同类工具具有多个优势，例如直观的side-by-side的降维可视化方式，hdf5格式保存表达矩阵从而读取快速，支持pdf/png保存图片，支持多种常见单细胞数据类型等。参考其教程文档，学习记录如下。</p>
<ul>
<li><strong>Paper |</strong> ShinyCell: simple and sharable visualization of single-cell gene expression data</li>
<li><strong>DOI</strong> <strong>|</strong> <a href="https://doi.org/10.1093/bioinformatics/btab209">https://doi.org/10.1093/bioinformatics/btab209</a></li>
<li><strong>Github |</strong>  <a href="https://github.com/SGDDNB/ShinyCell">https://github.com/SGDDNB/ShinyCell</a></li>
</ul></blockquote>
<h2 id="1-安装r包">1. 安装R包<a hidden class="anchor" aria-hidden="true" href="#1-安装r包">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># (1) 安装ShinyCell所需的依赖包</span>
</span></span><span style="display:flex;"><span>reqPkg = c(<span style="color:#0ff;font-weight:bold">&#34;data.table&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Matrix&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;hdf5r&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;reticulate&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;ggplot2&#34;</span>,
</span></span><span style="display:flex;"><span>           <span style="color:#0ff;font-weight:bold">&#34;gridExtra&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;glue&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;readr&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;RColorBrewer&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;R.utils&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Seurat&#34;</span>)
</span></span><span style="display:flex;"><span>newPkg = reqPkg[!(reqPkg %in% installed.packages()[,<span style="color:#0ff;font-weight:bold">&#34;Package&#34;</span>])]
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">if</span>(length(newPkg)){install.packages(newPkg)}
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2) 安装运行shiny所需的R包</span>
</span></span><span style="display:flex;"><span>reqPkg = c(<span style="color:#0ff;font-weight:bold">&#34;shiny&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;shinyhelper&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;data.table&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Matrix&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;DT&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;hdf5r&#34;</span>,
</span></span><span style="display:flex;"><span>           <span style="color:#0ff;font-weight:bold">&#34;reticulate&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;ggplot2&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;gridExtra&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;magrittr&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;ggdendro&#34;</span>)
</span></span><span style="display:flex;"><span>newPkg = reqPkg[!(reqPkg %in% installed.packages()[,<span style="color:#0ff;font-weight:bold">&#34;Package&#34;</span>])]
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">if</span>(length(newPkg)){install.packages(newPkg)}
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (3) 最后安装ShinyCell包</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># github直接安装</span>
</span></span><span style="display:flex;"><span>devtools::install_github(<span style="color:#0ff;font-weight:bold">&#34;SGDDNB/ShinyCell&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 源文件本地安装</span>
</span></span><span style="display:flex;"><span>devtools::install_local(<span style="color:#0ff;font-weight:bold">&#34;ShinyCell-master.zip&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>ShinyCell支持多种单细胞数据类型，包括Seurat， SCE(singlecellexperiment), h5ad, loom；并均提供了相应的示例文件；</li>
<li>如其文档所强调，ShinyCell是一个可视化工具，而不是分析工具；所以提供的单细胞数据需要已经完成基础的上游分析；</li>
<li>下面主要以Seurat类型为例进行学习</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>library(ShinyCell)
</span></span><span style="display:flex;"><span>library(shiny)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="2-基础用法">2. 基础用法<a hidden class="anchor" aria-hidden="true" href="#2-基础用法">#</a></h2>
<p><strong>2.1 示例数据</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># getExampleData()   # ~200MB</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># http://files.ddnetbio.com/hrpiFiles/readySeu_rset.rds</span>
</span></span><span style="display:flex;"><span>seu = readRDS(<span style="color:#0ff;font-weight:bold">&#34;readySeu_rset.rds&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>单细胞数据里需包括（1）标准化表达矩阵；（2）细胞meta信息；（3）降维信息。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">### （1）标准化表达矩阵</span>
</span></span><span style="display:flex;"><span>dim(seu)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 12590  9852</span>
</span></span><span style="display:flex;"><span>Assays(seu)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;RNA&#34;</span>
</span></span><span style="display:flex;"><span>seu@assays$RNA@data[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]  <span style="color:#007f7f">#注意是ENSEMBL基因ID</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 x 4 sparse Matrix of class &#34;dgCMatrix&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                 FM_AAACCTGAGAAACGCC-1 FM_AAACCTGAGTGGAGTC-1 FM_AAACCTGCATAGAAAC-1 FM_AAACCTGTCATTCACT-1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ENSG00000000003                     .                     .             0.8413237                     .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ENSG00000000419                     .                     .             .                             .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ENSG00000000457                     .                     .             .                             .</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ENSG00000000460                     .                     .             .                             .</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">### （2）细胞meta信息</span>
</span></span><span style="display:flex;"><span>head(t(seu@meta.data[1,]),<span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                 FM_AAACCTGAGAAACGCC-1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># orig.ident      &#34;FM&#34;                 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># RNA_snn_res.0.5 &#34;5&#34;                  </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># library         &#34;FM&#34;                 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># nUMI            &#34;13798&#34;              </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># nGene           &#34;3522&#34;               </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># pctMT           &#34;1.232063&#34;           </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># pctHK           &#34;20.77113&#34;           </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># score_S         &#34;-0.2176265&#34;         </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># score_G2M       &#34;-0.3646008&#34;         </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># phase           &#34;G1&#34;  </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">### （3）降维信息</span>
</span></span><span style="display:flex;"><span>names(seu@reductions)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;pca&#34;     &#34;tsne&#34;    &#34;umap&#34;    &#34;diffmap&#34; &#34;fdl&#34;  </span>
</span></span><span style="display:flex;"><span>head(Embeddings(seu, reduction = <span style="color:#0ff;font-weight:bold">&#34;umap&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                          UMAP_1     UMAP_2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FM_AAACCTGAGAAACGCC-1 -7.728941 -3.2370806</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FM_AAACCTGAGTGGAGTC-1 -7.048894  2.8379904</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FM_AAACCTGCATAGAAAC-1 -7.768705  1.6628623</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FM_AAACCTGTCATTCACT-1 -6.059119  3.3121483</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FM_AAACCTGTCTCTGAGA-1 -8.030442  1.5388506</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FM_AAACGGGCAGGCGATA-1 -9.831629 -0.5126803</span>
</span></span></code></pre></td></tr></table>
</div>
</div></blockquote>
<p><strong>2.2 生成shiny网页</strong></p>
<p>（1）生成meta配置文件</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>scConf = createConfig(seu)
</span></span><span style="display:flex;"><span>head(scConf) <span style="color:#007f7f"># 每行对应一项meta的title、level name、颜色方案等信息</span>
</span></span><span style="display:flex;"><span>t(scConf[1,])   
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#         [,1]                             </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ID      &#34;orig.ident&#34;                     </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># UI      &#34;orig.ident&#34;                     </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># fID     &#34;FM|NR|PR|RR&#34;                    </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># fUI     &#34;FM|NR|PR|RR&#34;                    </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># fCL     &#34;#A6CEE3|#B89B74|#ED8F47|#B15928&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># fRow    &#34;1&#34;                              </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># default &#34;1&#34;                              </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># grp     &#34;TRUE&#34; </span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>ID, fID 对应原始数据；UI, fUI, fCL 分布表示在shiny中展示的meta title，level name，colors</p>
<p>fROW表示 legend的行数，default 表示是否为默认展示的两种cell info，grp表示是否为分组类型</p></blockquote>
<p>（2）生成shiny配置文件</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>?makeShinyApp
</span></span><span style="display:flex;"><span>makeShinyApp(seu, scConf, 
</span></span><span style="display:flex;"><span>			 gene.mapping = <span style="color:#fff;font-weight:bold">TRUE</span>,  <span style="color:#007f7f"># default FALSE, only when ENSEMBL gene id</span>
</span></span><span style="display:flex;"><span>			 gex.assay = <span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>, gex.slot = <span style="color:#0ff;font-weight:bold">&#34;data&#34;</span>,
</span></span><span style="display:flex;"><span>             shiny.title = <span style="color:#0ff;font-weight:bold">&#34;scRNA-seq shiny app&#34;</span>,
</span></span><span style="display:flex;"><span>             shiny.dir = <span style="color:#0ff;font-weight:bold">&#34;shinyApp/&#34;</span>) 
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 查看结果</span>
</span></span><span style="display:flex;"><span>list.files(<span style="color:#0ff;font-weight:bold">&#34;shinyApp&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;sc1conf.rds&#34; &#34;sc1def.rds&#34;  &#34;sc1gene.rds&#34; &#34;sc1gexpr.h5&#34; &#34;sc1meta.rds&#34; &#34;server.R&#34;    &#34;ui.R&#34; </span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<ul>
<li>除了其中的<code>gene.mapping</code>参数（需要在单细胞数据的基因名为ENSEMBL格式时设置为TRUE），其余均为默认参数；</li>
<li>此外还有其它参数，包括默认展示的gene，降维类型；<code>shiny.footnotes</code>参数还支持设置网页的脚注用以交代数据集来源。</li>
</ul></blockquote>
<p>（3）最后运行shiny网页即可</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># 在本地浏览器的3344端口打开</span>
</span></span><span style="display:flex;"><span>runApp(<span style="color:#0ff;font-weight:bold">&#34;shinyApp/&#34;</span>, launch.browser = <span style="color:#fff;font-weight:bold">T</span>, port = <span style="color:#ff0;font-weight:bold">3344</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="3-模块介绍">3. 模块介绍<a hidden class="anchor" aria-hidden="true" href="#3-模块介绍">#</a></h2>
<p>（1）完成上述步骤后即可在本地生成一个shiny网页，下面结合作者分享的示例http://shinycell1.ddnetbio.com/介绍其内容；</p>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230807075943682.png" alt="image-20230807075943682" style="zoom:50%;" />
<blockquote>
<p>在每个模块中，均有3个Toggle选项支持筛选细胞，调整部分绘图参数等。</p></blockquote>
<p>（2）如上图所示，ShinyCell主要支持7个可视化模块；若从绘图类型来说共是6种图</p>
<table>
  <thead>
      <tr>
          <th style="text-align: center"></th>
          <th style="text-align: center">模块</th>
          <th style="text-align: center">图形</th>
          <th style="text-align: center">含义</th>
          <th style="text-align: center">e.g.</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td style="text-align: center">1</td>
          <td style="text-align: center">CellInfo vs GeneExpr</td>
          <td style="text-align: center">降维图</td>
          <td style="text-align: center"><strong>对比</strong>可视化细胞的分群信息与表达信息</td>
          <td style="text-align: center">A+B</td>
      </tr>
      <tr>
          <td style="text-align: center">2</td>
          <td style="text-align: center">CellInfo vs CellInfo</td>
          <td style="text-align: center">降维图</td>
          <td style="text-align: center"><strong>对比</strong>可视化细胞的两种不同分群信息</td>
          <td style="text-align: center">A+A</td>
      </tr>
      <tr>
          <td style="text-align: center">3</td>
          <td style="text-align: center">GeneExpr vs GeneExpr</td>
          <td style="text-align: center">降维图</td>
          <td style="text-align: center"><strong>对比</strong>可视化细胞的两种不同表达信息</td>
          <td style="text-align: center">B+B</td>
      </tr>
      <tr>
          <td style="text-align: center">4</td>
          <td style="text-align: center">Gene coexpression</td>
          <td style="text-align: center">降维图</td>
          <td style="text-align: center">可视化细胞的两种不同表达信息</td>
          <td style="text-align: center">C</td>
      </tr>
      <tr>
          <td style="text-align: center">5</td>
          <td style="text-align: center">Violinplot / Boxplot</td>
          <td style="text-align: center">小提琴图/箱图</td>
          <td style="text-align: center">可视化特定分群下，细胞的表达分布</td>
          <td style="text-align: center">D</td>
      </tr>
      <tr>
          <td style="text-align: center">6</td>
          <td style="text-align: center">Bubbleplot / Heatmap</td>
          <td style="text-align: center">点图/热图</td>
          <td style="text-align: center">可视化特定分群下，细胞的多个表达分布</td>
          <td style="text-align: center">F</td>
      </tr>
      <tr>
          <td style="text-align: center">7</td>
          <td style="text-align: center">Proportion plot</td>
          <td style="text-align: center">柱状图</td>
          <td style="text-align: center">可视化特定分群下，细胞的另一种分群的分布</td>
          <td style="text-align: center">E</td>
      </tr>
  </tbody>
</table>
<img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230807084047875.png" alt="image-20230807084047875" style="zoom: 80%;" />
<h2 id="4-进阶用法">4. 进阶用法<a hidden class="anchor" aria-hidden="true" href="#4-进阶用法">#</a></h2>
<h3 id="41-meta-config调整">4.1 meta config调整<a hidden class="anchor" aria-hidden="true" href="#41-meta-config调整">#</a></h3>
<ul>
<li>
<p>如上介绍，生成shiny网页的第一步是根据单细胞数据的meta信息生成相应的config文件，用以设置模块可视化的细节。</p>
</li>
<li>
<p>默认情况下会使用全部的meta信息，如需调整一方面可直接修改原来的单细胞数据；另一方面也可以使用ShinyCell包进行部分修改，如下所示。</p>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>seu &lt;- readRDS(<span style="color:#0ff;font-weight:bold">&#34;readySeu_rset.rds&#34;</span>)
</span></span><span style="display:flex;"><span>seu@meta.data = seu@meta.data[,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">6</span>]
</span></span><span style="display:flex;"><span>scConf = createConfig(seu)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 展示当前的meta config legend</span>
</span></span><span style="display:flex;"><span>showLegend(scConf) 
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (1)删除若干条细胞注释信息</span>
</span></span><span style="display:flex;"><span>scConf = delMeta(scConf, c(<span style="color:#0ff;font-weight:bold">&#34;orig.ident&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2)修改在shiny的细胞注释名title name(fID)</span>
</span></span><span style="display:flex;"><span>scConf = modMetaName(scConf,
</span></span><span style="display:flex;"><span>                     meta.to.mod = c(<span style="color:#0ff;font-weight:bold">&#34;nUMI&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;nGene&#34;</span>),
</span></span><span style="display:flex;"><span>                     new.name = c(<span style="color:#0ff;font-weight:bold">&#34;No. UMIs&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;No. detected genes&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (3)修改在shiny的细胞注释名level name(fUI)</span>
</span></span><span style="display:flex;"><span>scConf = modLabels(scConf, meta.to.mod = <span style="color:#0ff;font-weight:bold">&#34;library&#34;</span>,
</span></span><span style="display:flex;"><span>                   new.labels = c(<span style="color:#0ff;font-weight:bold">&#34;fm&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;pr&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;nr&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;rr&#34;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (4)修改在shiny的细胞注释可视化颜色(fCL)</span>
</span></span><span style="display:flex;"><span>scConf = modColours(scConf, meta.to.mod = <span style="color:#0ff;font-weight:bold">&#34;library&#34;</span>,
</span></span><span style="display:flex;"><span>                    new.colours= RColorBrewer::brewer.pal(n = <span style="color:#ff0;font-weight:bold">4</span>, name = <span style="color:#0ff;font-weight:bold">&#34;Set1&#34;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>showLegend(scConf)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># showOrder(scConf)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># scConf = reorderMeta(scConf, rev(scConf$ID))</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># showOrder(scConf)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>如需修改特定meta的level factor顺序，似乎只能在原始单细胞数据的meta中进行修改。</p></blockquote>
<h3 id="42-多个数据集">4.2 多个数据集<a hidden class="anchor" aria-hidden="true" href="#42-多个数据集">#</a></h3>
<ul>
<li>
<p>ShinyCell支持在同一个网页内独立展示多个的数据集，可参考示例网页http://shinycell2.ddnetbio.com</p>
</li>
<li>
<p>简单生成方法如下</p>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## 数据集1</span>
</span></span><span style="display:flex;"><span>seu &lt;- readRDS(<span style="color:#0ff;font-weight:bold">&#34;readySeu_rset.rds&#34;</span>)
</span></span><span style="display:flex;"><span>scConf1 = createConfig(seu)
</span></span><span style="display:flex;"><span>makeShinyFiles(seu, scConf1, 
</span></span><span style="display:flex;"><span>			 gene.mapping = <span style="color:#fff;font-weight:bold">TRUE</span>,  
</span></span><span style="display:flex;"><span>			 shiny.prefix = <span style="color:#0ff;font-weight:bold">&#34;sc1&#34;</span>, <span style="color:#007f7f"># 前缀不同</span>
</span></span><span style="display:flex;"><span>             shiny.dir = <span style="color:#0ff;font-weight:bold">&#34;shinyAppMulti/&#34;</span>) 
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 数据集2</span>
</span></span><span style="display:flex;"><span>scConf2 = delMeta(scConf1, c(<span style="color:#0ff;font-weight:bold">&#34;orig.ident&#34;</span>))
</span></span><span style="display:flex;"><span>makeShinyFiles(seu, scConf2, 
</span></span><span style="display:flex;"><span>			 gene.mapping = <span style="color:#fff;font-weight:bold">TRUE</span>,  
</span></span><span style="display:flex;"><span>			 shiny.prefix = <span style="color:#0ff;font-weight:bold">&#34;sc2&#34;</span>, <span style="color:#007f7f"># 前缀不同</span>
</span></span><span style="display:flex;"><span>			 gex.assay = <span style="color:#0ff;font-weight:bold">&#34;RNA&#34;</span>, gex.slot = <span style="color:#0ff;font-weight:bold">&#34;data&#34;</span>,
</span></span><span style="display:flex;"><span>             shiny.dir = <span style="color:#0ff;font-weight:bold">&#34;shinyAppMulti/&#34;</span>) 
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 生成shiny配置文件</span>
</span></span><span style="display:flex;"><span>makeShinyCodesMulti(
</span></span><span style="display:flex;"><span>  shiny.title = <span style="color:#0ff;font-weight:bold">&#34;Multi-dataset Tutorial&#34;</span>, 
</span></span><span style="display:flex;"><span>  shiny.prefix = c(<span style="color:#0ff;font-weight:bold">&#34;sc1&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;sc2&#34;</span>),
</span></span><span style="display:flex;"><span>  shiny.headers = c(<span style="color:#0ff;font-weight:bold">&#34;Dataset1&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Dataset2&#34;</span>),
</span></span><span style="display:flex;"><span>  shiny.footnotes = <span style="color:#0ff;font-weight:bold">&#34;&#34;</span>,
</span></span><span style="display:flex;"><span>  shiny.dir = <span style="color:#0ff;font-weight:bold">&#34;shinyAppMulti/&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 运行网页</span>
</span></span><span style="display:flex;"><span>runApp(<span style="color:#0ff;font-weight:bold">&#34;shinyAppMulti/&#34;</span>, launch.browser = <span style="color:#fff;font-weight:bold">T</span>, port = <span style="color:#ff0;font-weight:bold">3344</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230807091629863.png" alt="image-20230807091629863"  />
</p>
<h3 id="43-linux平台">4.3 linux平台<a hidden class="anchor" aria-hidden="true" href="#43-linux平台">#</a></h3>
<ul>
<li>对于大型单细胞数据集，需要在服务器linux平台分析时，大体与上述方式类似，简述流程如下。</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># (1) 前期需在linux的R语言环境下安装相关R包，并准备相关数据</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2) 生成shiny网页文件 </span>
</span></span><span style="display:flex;"><span>seu &lt;- readRDS(<span style="color:#0ff;font-weight:bold">&#34;readySeu_rset.rds&#34;</span>)
</span></span><span style="display:flex;"><span>scConf = createConfig(seu)
</span></span><span style="display:flex;"><span>makeShinyApp(seu, scConf, gene.mapping = <span style="color:#fff;font-weight:bold">TRUE</span>,
</span></span><span style="display:flex;"><span>             shiny.title = <span style="color:#0ff;font-weight:bold">&#34;ShinyCell Quick Start&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (3) linux端运行网页，注意需要设置launch.browser参数为FALSE</span>
</span></span><span style="display:flex;"><span>runApp(<span style="color:#0ff;font-weight:bold">&#34;shinyApp/&#34;</span>, launch.browser = <span style="color:#fff;font-weight:bold">F</span>, port = <span style="color:#ff0;font-weight:bold">3344</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (4) 本地window端链接：同时win键+R，输入cmd打开命令窗口，根据下述命令进行链接</span>
</span></span><span style="display:flex;"><span>ssh -L localhost:<span style="color:#ff0;font-weight:bold">3344</span>:localhost:<span style="color:#ff0;font-weight:bold">3344</span> username@<span style="color:#ff0;font-weight:bold">192.168.212</span>.xxx -p <span style="color:#ff0;font-weight:bold">22</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (5) 最后本地浏览器打开 http://127.0.0.1:3344/ 即可</span>
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/037%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--sccode%E7%BB%BC%E5%90%88%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/">
    <span class="title">« Prev Page</span>
    <br>
    <span>单细胞分析工具--sccode综合差异分析</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/041%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--seurat5%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/">
    <span class="title">Next Page »</span>
    <br>
    <span>单细胞分析工具--Seurat5基础用法</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
