<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--MAST差异基因分析 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
MAST，全称Model-based Analysis of Single-cell Transcriptomics，是2015年于Genome Biology发表的R包工具，主要用于单细胞数据的差异表达分析。参看官方教程，简单记录用法如下。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/030%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--mast%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E5%88%86%E6%9E%90/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/030%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--mast%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E5%88%86%E6%9E%90/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--MAST差异基因分析" />
<meta property="og:description" content="
MAST，全称Model-based Analysis of Single-cell Transcriptomics，是2015年于Genome Biology发表的R包工具，主要用于单细胞数据的差异表达分析。参看官方教程，简单记录用法如下。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/030%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--mast%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E5%88%86%E6%9E%90/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-02-10T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-02-10T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--MAST差异基因分析"/>
<meta name="twitter:description" content="
MAST，全称Model-based Analysis of Single-cell Transcriptomics，是2015年于Genome Biology发表的R包工具，主要用于单细胞数据的差异表达分析。参看官方教程，简单记录用法如下。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--MAST差异基因分析",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/030%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--mast%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E5%88%86%E6%9E%90/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--MAST差异基因分析",
  "name": "单细胞分析工具--MAST差异基因分析",
  "description": " MAST，全称Model-based Analysis of Single-cell Transcriptomics，是2015年于Genome Biology发表的R包工具，主要用于单细胞数据的差异表达分析。参看官方教程，简单记录用法如下。\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " MAST，全称Model-based Analysis of Single-cell Transcriptomics，是2015年于Genome Biology发表的R包工具，主要用于单细胞数据的差异表达分析。参看官方教程，简单记录用法如下。\n原始论文：https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5\nGithub：https://github.com/RGLab/MAST\n官方手册：http://bioconductor.org/packages/release/bioc/vignettes/MAST/inst/doc/MAITAnalysis.html\n1 BiocManager::install(\"MAST\") 1 2 library(MAST) library(tidyverse) 1、示例数据 经细胞因子刺激前后的MAITs细胞的单细胞表达数据，使用MAST包分析其差异表达基因。 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 data(maits, package='MAST') dim(maits$expressionmat) #共96个细胞 # [1] 96 16302 maits$expressionmat[1:4,1:4] head(maits$cdat) #细胞注释数据 head(maits$fdat) #基因注释数据 ##整理表达矩阵，使用symbol作为基因名 idx = -which(duplicated(maits$fdat$symbolid)) maits$fdat = maits$fdat[idx,] maits$fdat = maits$fdat[,-1] rownames(maits$fdat) = maits$fdat$symbolid maits$expressionmat = maits$expressionmat[,idx] colnames(maits$expressionmat) = rownames(maits$fdat) dim(maits$expressionmat) # [1] 96 15077 maits$expressionmat[1:4,1:4] # WASH7P FAM138F LOC100132062 LOC100133331 # 1-MAIT-Stim-C05_S91 0 0.08406426 3.7158934 6.829215 # 1-MAIT-Stim-C07_S70 0 0.12432814 0.0000000 0.000000 # 1-MAIT-Stim-C08_S68 0 1.79077204 0.0000000 0.000000 # 1-MAIT-Stim-C10_S95 0 1.71369581 0.2141248 0.000000 head(maits$fdat) # entrez symbolid # WASH7P 653635 WASH7P # FAM138F 641702 FAM138F # LOC100132062 100132062 LOC100132062 2、构建SingleCellAssay对象 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 sca = FromMatrix(t(maits$expressionmat), maits$cdat, maits$fdat) ## 方便演示，随机筛选出2000个基因 sca = sca[unique(sample(rownames(sca),2000)),] dim(sca) # [1] 16302 96 ## (1) 目标分组 table(colData(sca)$condition) # Stim Unstim # 47 49 ## 将Unstim设置为参考组 cond\u003c-factor(colData(sca)$condition) cond\u003c-relevel(cond,\"Unstim\") colData(sca)$condition\u003c-cond ## (2) 考虑一个协变量因素 summary(colData(sca)$cngeneson) # Min. 1st Qu. Median Mean 3rd Qu. Max. # -0.17159 -0.02578 0.01408 0.00000 0.05943 0.15415 ## (3) 模拟一个分类变量分组因素 set.seed(123) colData(sca)$region = sample(letters[1:3], ncol(sca), replace=T) table(colData(sca)$region) # a b c #32 31 33 3、回归分析 MAST做差异分析的主要优势在于可以校正多种协变量因素，如下演示几种常用的用法。 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 ## (1) 校正cngeneson协变量:默认参数 zlmCond \u003c- zlm(~condition + cngeneson, sca, method=\"bayesglm\", ebayes=TRUE) summaryCond \u003c- summary(zlmCond, doLRT='conditionStim') summaryDt \u003c- summaryCond$datatable levels(summaryDt$contrast) # [1] \"conditionStim\" \"(Intercept)\" \"cngeneson\" ## (2) 校正cngeneson与region协变量:默认参数 zlmCond \u003c- zlm(~condition + cngeneson + region, sca, method = \"bayesglm\", ebayes=TRUE) summaryCond \u003c- summary(zlmCond, doLRT='conditionStim') summaryDt \u003c- summaryCond$datatable levels(summaryDt$contrast) # [1] \"conditionStim\" \"(Intercept)\" \"cngeneson\" \"regionb\" \"regionc\" ## (3) 校正cngeneson协变量，将region认为是随机因素 zlmCond \u003c- zlm(~condition + cngeneson + (1|region), sca, method = \"glmer\", ebayes=FALSE) summaryCond \u003c- summary(zlmCond, doLRT='conditionStim') summaryDt \u003c- summaryCond$datatable levels(summaryDt$contrast) # [1] \"conditionStim\" \"(Intercept)\" \"cngeneson\" 汇总差异基因结果 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 df_pval = summaryDt %\u003e% dplyr::filter(contrast=='conditionStim') %\u003e% dplyr::filter(component=='H') %\u003e% dplyr::select(primerid, `Pr(\u003eChisq)`) df_logfc = summaryDt %\u003e% dplyr::filter(contrast=='conditionStim') %\u003e% dplyr::filter(component=='logFC') %\u003e% dplyr::select(primerid, coef, ci.hi, ci.lo) df_stat = dplyr::inner_join(df_logfc, df_pval) %\u003e% dplyr::rename(\"symbol\"=\"primerid\") %\u003e% dplyr::rename(\"pval\"=\"Pr(\u003eChisq)\",\"logFC\"=\"coef\") %\u003e% dplyr::mutate(\"fdr\" = p.adjust(pval)) %\u003e% dplyr::arrange(fdr) head(df_stat) # symbol logFC ci.hi ci.lo pval fdr # 1: NDUFV2 7.522343 8.8016689 6.2430166 2.586874e-18 5.109075e-15 # 2: KIF9 -0.225052 0.1273536 -0.5774576 1.760028e-16 3.474294e-13 # 3: DUSP1 -6.513010 -5.2606967 -7.7653235 8.668938e-15 1.710382e-11 # 4: VCP 5.155219 6.2345839 4.0758540 1.530723e-13 3.018586e-10 # 5: SARS 5.577095 6.8474645 4.3067258 1.216538e-12 2.397796e-09 # 6: HSPD1 6.261077 7.6919267 4.8302277 1.969255e-12 3.879433e-09 上述第3种差异分析方式在一些文献中经常提及，主要用于避免不同样本来源对差异分析造成的影响，例如下面两篇文章。\n（1）The landscape of immune dysregulation in Crohn’s disease revealed through single-cell transcriptomic profiling in the ileum and colon\n（2）Diverse human astrocyte and microglial transcriptional responses to Alzheimer’s pathology\n",
  "wordCount" : "1131",
  "inLanguage": "en",
  "datePublished": "2023-02-10T00:00:00Z",
  "dateModified": "2023-02-10T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/030%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--mast%E5%B7%AE%E5%BC%82%E5%9F%BA%E5%9B%A0%E5%88%86%E6%9E%90/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--MAST差异基因分析
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-02-10 00:00:00 &#43;0000 UTC&#39;&gt;2023-02-10&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-02-10&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1131&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e7%a4%ba%e4%be%8b%e6%95%b0%e6%8d%ae" aria-label="1、示例数据">1、示例数据</a><ul>
                            
                    <li>
                        <a href="#2%e6%9e%84%e5%bb%basinglecellassay%e5%af%b9%e8%b1%a1" aria-label="2、构建SingleCellAssay对象">2、构建SingleCellAssay对象</a></li></ul>
                    </li>
                    <li>
                        <a href="#3%e5%9b%9e%e5%bd%92%e5%88%86%e6%9e%90" aria-label="3、回归分析">3、回归分析</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>MAST，全称Model-based Analysis of Single-cell Transcriptomics，是2015年于Genome Biology发表的R包工具，主要用于单细胞数据的差异表达分析。参看官方教程，简单记录用法如下。</p>
<p>原始论文：https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5</p>
<p>Github：https://github.com/RGLab/MAST</p>
<p>官方手册：http://bioconductor.org/packages/release/bioc/vignettes/MAST/inst/doc/MAITAnalysis.html</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230209215214802.png" alt="image-20230209215214802"  />
</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>BiocManager::install(<span style="color:#0ff;font-weight:bold">&#34;MAST&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(MAST)
</span></span><span style="display:flex;"><span>library(tidyverse)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="1示例数据">1、示例数据<a hidden class="anchor" aria-hidden="true" href="#1示例数据">#</a></h2>
<ul>
<li>经细胞因子刺激前后的MAITs细胞的单细胞表达数据，使用MAST包分析其差异表达基因。</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>data(maits, package=<span style="color:#0ff;font-weight:bold">&#39;MAST&#39;</span>)
</span></span><span style="display:flex;"><span>dim(maits$expressionmat) <span style="color:#007f7f">#共96个细胞</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1]    96 16302</span>
</span></span><span style="display:flex;"><span>maits$expressionmat[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span>head(maits$cdat) <span style="color:#007f7f">#细胞注释数据</span>
</span></span><span style="display:flex;"><span>head(maits$fdat) <span style="color:#007f7f">#基因注释数据</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##整理表达矩阵，使用symbol作为基因名</span>
</span></span><span style="display:flex;"><span>idx = -which(duplicated(maits$fdat$symbolid))
</span></span><span style="display:flex;"><span>maits$fdat = maits$fdat[idx,]
</span></span><span style="display:flex;"><span>maits$fdat = maits$fdat[,<span style="color:#ff0;font-weight:bold">-1</span>]
</span></span><span style="display:flex;"><span>rownames(maits$fdat) = maits$fdat$symbolid
</span></span><span style="display:flex;"><span>maits$expressionmat = maits$expressionmat[,idx]
</span></span><span style="display:flex;"><span>colnames(maits$expressionmat) = rownames(maits$fdat)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>dim(maits$expressionmat)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1]    96 15077</span>
</span></span><span style="display:flex;"><span>maits$expressionmat[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                     WASH7P    FAM138F LOC100132062 LOC100133331</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1-MAIT-Stim-C05_S91      0 0.08406426    3.7158934     6.829215</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1-MAIT-Stim-C07_S70      0 0.12432814    0.0000000     0.000000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1-MAIT-Stim-C08_S68      0 1.79077204    0.0000000     0.000000</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1-MAIT-Stim-C10_S95      0 1.71369581    0.2141248     0.000000</span>
</span></span><span style="display:flex;"><span>head(maits$fdat)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                 entrez     symbolid</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># WASH7P          653635       WASH7P</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FAM138F         641702      FAM138F</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># LOC100132062 100132062 LOC100132062</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="2构建singlecellassay对象">2、构建SingleCellAssay对象<a hidden class="anchor" aria-hidden="true" href="#2构建singlecellassay对象">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>sca = FromMatrix(t(maits$expressionmat), maits$cdat, maits$fdat)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 方便演示，随机筛选出2000个基因</span>
</span></span><span style="display:flex;"><span>sca = sca[unique(sample(rownames(sca),<span style="color:#ff0;font-weight:bold">2000</span>)),]
</span></span><span style="display:flex;"><span>dim(sca)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 16302    96</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (1) 目标分组</span>
</span></span><span style="display:flex;"><span>table(colData(sca)$condition)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Stim Unstim </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 47     49 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 将Unstim设置为参考组</span>
</span></span><span style="display:flex;"><span>cond&lt;-factor(colData(sca)$condition)
</span></span><span style="display:flex;"><span>cond&lt;-relevel(cond,<span style="color:#0ff;font-weight:bold">&#34;Unstim&#34;</span>)
</span></span><span style="display:flex;"><span>colData(sca)$condition&lt;-cond
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (2) 考虑一个协变量因素</span>
</span></span><span style="display:flex;"><span>summary(colData(sca)$cngeneson)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># -0.17159 -0.02578  0.01408  0.00000  0.05943  0.15415 </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (3) 模拟一个分类变量分组因素</span>
</span></span><span style="display:flex;"><span>set.seed(<span style="color:#ff0;font-weight:bold">123</span>)
</span></span><span style="display:flex;"><span>colData(sca)$region = sample(<span style="color:#fff;font-weight:bold">letters</span>[1:<span style="color:#ff0;font-weight:bold">3</span>], ncol(sca), replace=<span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>table(colData(sca)$region)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># a  b  c </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#32 31 33 </span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="3回归分析">3、回归分析<a hidden class="anchor" aria-hidden="true" href="#3回归分析">#</a></h2>
<ul>
<li>MAST做差异分析的主要优势在于可以校正多种协变量因素，如下演示几种常用的用法。</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## (1) 校正cngeneson协变量:默认参数</span>
</span></span><span style="display:flex;"><span>zlmCond &lt;- zlm(~condition + cngeneson, sca,
</span></span><span style="display:flex;"><span>    method=<span style="color:#0ff;font-weight:bold">&#34;bayesglm&#34;</span>, ebayes=<span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>summaryCond &lt;- summary(zlmCond, doLRT=<span style="color:#0ff;font-weight:bold">&#39;conditionStim&#39;</span>)
</span></span><span style="display:flex;"><span>summaryDt &lt;- summaryCond$datatable
</span></span><span style="display:flex;"><span>levels(summaryDt$contrast)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;conditionStim&#34; &#34;(Intercept)&#34;   &#34;cngeneson&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (2) 校正cngeneson与region协变量:默认参数</span>
</span></span><span style="display:flex;"><span>zlmCond &lt;- zlm(~condition + cngeneson + region, sca, 
</span></span><span style="display:flex;"><span>    method = <span style="color:#0ff;font-weight:bold">&#34;bayesglm&#34;</span>, ebayes=<span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>summaryCond &lt;- summary(zlmCond, doLRT=<span style="color:#0ff;font-weight:bold">&#39;conditionStim&#39;</span>)
</span></span><span style="display:flex;"><span>summaryDt &lt;- summaryCond$datatable
</span></span><span style="display:flex;"><span>levels(summaryDt$contrast)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;conditionStim&#34; &#34;(Intercept)&#34;   &#34;cngeneson&#34;     &#34;regionb&#34;       &#34;regionc&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## (3) 校正cngeneson协变量，将region认为是随机因素</span>
</span></span><span style="display:flex;"><span>zlmCond &lt;- zlm(~condition + cngeneson + (<span style="color:#ff0;font-weight:bold">1</span>|region), sca, 
</span></span><span style="display:flex;"><span>    method = <span style="color:#0ff;font-weight:bold">&#34;glmer&#34;</span>, ebayes=<span style="color:#fff;font-weight:bold">FALSE</span>)
</span></span><span style="display:flex;"><span>summaryCond &lt;- summary(zlmCond, doLRT=<span style="color:#0ff;font-weight:bold">&#39;conditionStim&#39;</span>)
</span></span><span style="display:flex;"><span>summaryDt &lt;- summaryCond$datatable
</span></span><span style="display:flex;"><span>levels(summaryDt$contrast)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;conditionStim&#34; &#34;(Intercept)&#34;   &#34;cngeneson&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>汇总差异基因结果</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>df_pval = summaryDt %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::filter(contrast==<span style="color:#0ff;font-weight:bold">&#39;conditionStim&#39;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::filter(component==<span style="color:#0ff;font-weight:bold">&#39;H&#39;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::select(primerid, `Pr(&gt;Chisq)`)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>df_logfc = summaryDt %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::filter(contrast==<span style="color:#0ff;font-weight:bold">&#39;conditionStim&#39;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::filter(component==<span style="color:#0ff;font-weight:bold">&#39;logFC&#39;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::select(primerid, coef, ci.hi, ci.lo)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>df_stat = dplyr::inner_join(df_logfc, df_pval) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::rename(<span style="color:#0ff;font-weight:bold">&#34;symbol&#34;</span>=<span style="color:#0ff;font-weight:bold">&#34;primerid&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::rename(<span style="color:#0ff;font-weight:bold">&#34;pval&#34;</span>=<span style="color:#0ff;font-weight:bold">&#34;Pr(&gt;Chisq)&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;logFC&#34;</span>=<span style="color:#0ff;font-weight:bold">&#34;coef&#34;</span>) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::mutate(<span style="color:#0ff;font-weight:bold">&#34;fdr&#34;</span> = p.adjust(pval)) %&gt;% 
</span></span><span style="display:flex;"><span>    dplyr::arrange(fdr)
</span></span><span style="display:flex;"><span>head(df_stat)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#    symbol     logFC      ci.hi      ci.lo         pval          fdr</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1: NDUFV2  7.522343  8.8016689  6.2430166 2.586874e-18 5.109075e-15</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2:   KIF9 -0.225052  0.1273536 -0.5774576 1.760028e-16 3.474294e-13</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3:  DUSP1 -6.513010 -5.2606967 -7.7653235 8.668938e-15 1.710382e-11</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4:    VCP  5.155219  6.2345839  4.0758540 1.530723e-13 3.018586e-10</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5:   SARS  5.577095  6.8474645  4.3067258 1.216538e-12 2.397796e-09</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6:  HSPD1  6.261077  7.6919267  4.8302277 1.969255e-12 3.879433e-09</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>上述第3种差异分析方式在一些文献中经常提及，主要用于避免不同样本来源对差异分析造成的影响，例如下面两篇文章。</p>
<p>（1）The landscape of immune dysregulation in Crohn’s disease revealed through single-cell transcriptomic profiling in the ileum and colon</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230210084713190.png" alt="image-20230210084713190"  />
</p>
<p>（2）Diverse human astrocyte and microglial transcriptional responses to Alzheimer’s pathology</p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20230210084841322.png" alt="image-20230210084841322"  />
</p></blockquote>


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/program/023python%E6%95%B0%E6%8D%AE%E7%A7%91%E5%AD%A6-matplotlib%E5%8F%AF%E8%A7%86%E5%8C%96/">
    <span class="title">« Prev Page</span>
    <br>
    <span>Python数据科学-(4)Matplotlib可视化</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/031%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scissor%E8%A1%A8%E5%9E%8B%E7%9B%B8%E5%85%B3%E4%BA%9A%E7%BE%A4%E9%89%B4%E5%AE%9A/">
    <span class="title">Next Page »</span>
    <br>
    <span>单细胞分析工具--Scissor表型相关亚群鉴定</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
