<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>差异分析R包-DESeq2&#43;edgeR&#43;limma | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="转录组">
<meta name="description" content="在对转录组数据分析时，分组比较的差异分析前提时获得测序表达矩阵。
根据测序技术分为两种，对应的分析R包如下所示。

RNAseq的原始count表达矩阵

DESeq2
edgeR
limma


芯片array标准化后的表达矩阵

limma



1、示例count表达矩阵
如下使用TCGAbiolinks包下载共9个癌症样本、35个正常样本的表达矩阵；其中涉及到8对配对样本（tumor与normal来自同一病人）">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/045%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90r%E5%8C%85-deseq2&#43;edger&#43;limma/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/045%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90r%E5%8C%85-deseq2&#43;edger&#43;limma/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="差异分析R包-DESeq2&#43;edgeR&#43;limma" />
<meta property="og:description" content="在对转录组数据分析时，分组比较的差异分析前提时获得测序表达矩阵。
根据测序技术分为两种，对应的分析R包如下所示。

RNAseq的原始count表达矩阵

DESeq2
edgeR
limma


芯片array标准化后的表达矩阵

limma



1、示例count表达矩阵
如下使用TCGAbiolinks包下载共9个癌症样本、35个正常样本的表达矩阵；其中涉及到8对配对样本（tumor与normal来自同一病人）" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/045%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90r%E5%8C%85-deseq2&#43;edger&#43;limma/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-06-11T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2022-06-11T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="差异分析R包-DESeq2&#43;edgeR&#43;limma"/>
<meta name="twitter:description" content="在对转录组数据分析时，分组比较的差异分析前提时获得测序表达矩阵。
根据测序技术分为两种，对应的分析R包如下所示。

RNAseq的原始count表达矩阵

DESeq2
edgeR
limma


芯片array标准化后的表达矩阵

limma



1、示例count表达矩阵
如下使用TCGAbiolinks包下载共9个癌症样本、35个正常样本的表达矩阵；其中涉及到8对配对样本（tumor与normal来自同一病人）"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "差异分析R包-DESeq2+edgeR+limma",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/045%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90r%E5%8C%85-deseq2+edger+limma/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "差异分析R包-DESeq2+edgeR+limma",
  "name": "差异分析R包-DESeq2\u002bedgeR\u002blimma",
  "description": "在对转录组数据分析时，分组比较的差异分析前提时获得测序表达矩阵。\n根据测序技术分为两种，对应的分析R包如下所示。\nRNAseq的原始count表达矩阵 DESeq2 edgeR limma 芯片array标准化后的表达矩阵 limma 1、示例count表达矩阵 如下使用TCGAbiolinks包下载共9个癌症样本、35个正常样本的表达矩阵；其中涉及到8对配对样本（tumor与normal来自同一病人）\n",
  "keywords": [
    "转录组"
  ],
  "articleBody": "在对转录组数据分析时，分组比较的差异分析前提时获得测序表达矩阵。\n根据测序技术分为两种，对应的分析R包如下所示。\nRNAseq的原始count表达矩阵 DESeq2 edgeR limma 芯片array标准化后的表达矩阵 limma 1、示例count表达矩阵 如下使用TCGAbiolinks包下载共9个癌症样本、35个正常样本的表达矩阵；其中涉及到8对配对样本（tumor与normal来自同一病人）\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 library(TCGAbiolinks) library(SummarizedExperiment) library(tidyverse) query \u003c- GDCquery(project = \"TCGA-CHOL\", legacy = FALSE, data.category = \"Transcriptome Profiling\", data.type = \"Gene Expression Quantification\") GDCdownload(query, files.per.chunk = 10) data = GDCprepare(query, save = F) rownames(data) = rowData(data)$gene_name data = data[rowData(data)$gene_type==\"protein_coding\",] count = assay(data, \"unstranded\") count = count[!duplicated(rownames(count)),] dim(count) # [1] 19938 44 count[1:4,1:4] # TCGA-W5-AA2X-01A-11R-A41I-07 TCGA-W5-AA2X-11A-11R-A41I-07 # TSPAN6 3310 2322 # TNMD 0 0 # DPM1 1881 716 # SCYL3 966 315 # TCGA-W5-AA33-01A-11R-A41I-07 TCGA-ZH-A8Y2-01A-11R-A41I-07 # TSPAN6 11312 6768 # TNMD 1 0 # DPM1 968 1273 # SCYL3 1296 774 meta = colData(data)[,c(\"barcode\",\"sample\", \"patient\",\"shortLetterCode\")] %\u003e% as.data.frame() meta$Group = meta$shortLetterCode # NT: Solid Tissue Normal; TP: Primary Tumor head(meta[,\"Group\",drop=F]) # Group # TCGA-W5-AA2X-01A-11R-A41I-07 TP # TCGA-W5-AA2X-11A-11R-A41I-07 NT # TCGA-W5-AA33-01A-11R-A41I-07 TP # TCGA-ZH-A8Y2-01A-11R-A41I-07 TP # TCGA-ZH-A8Y6-01A-11R-A41I-07 TP # TCGA-W5-AA2Z-01A-11R-A41I-07 TP table(meta$Group) # NT TP # 9 35 identical(colnames(count), rownames(meta)) # [1] TRUE ## 8对配对样本 meta_pair = meta[TCGAquery_MatchedCoupledSampleTypes(meta$barcode, c(\"NT\",\"TP\")), ] # TCGA-W5-AA2I TCGA-W5-AA2Q TCGA-W5-AA2U TCGA-W5-AA2X TCGA-W5-AA30 TCGA-W5-AA31 TCGA-W5-AA34 TCGA-ZU-A8S4 # 2 2 2 2 2 2 2 2 count_pair = count[, meta_pair$barcode] identical(meta_pair$barcode, colnames(count_pair)) #[1] TRUE 2、DESeq2 常规分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 library(DESeq2) # 分组因子化：对照组在前，实验组在后 meta$Group = factor(meta$Group, levels = c(\"NT\", \"TP\")) dds = DESeqDataSetFromMatrix(countData = count, colData = meta, design= ~ Group) # keep \u003c- rowSums(counts(dds)) \u003e= 10 # dds \u003c- dds[keep,] dds = DESeq(dds) resultsNames(dds) # lists the coefficients # [1] \"Intercept\" \"Group_TP_vs_NT\" res = results(dds, name=\"Group_TP_vs_NT\") deg = as.data.frame(res) deg = deg[order(deg$padj), ] head(deg) # baseMean log2FoldChange lfcSE stat pvalue padj # MSMO1 7585.3612 -3.748488 0.2109579 -17.76889 1.231028e-70 1.105094e-66 # RMDN2 396.7567 -2.929035 0.1648278 -17.77027 1.201019e-70 1.105094e-66 # GCDH 2272.9263 -3.302603 0.1863973 -17.71809 3.040673e-70 1.819742e-66 # TCAIM 1239.0505 -2.235421 0.1303740 -17.14621 6.708695e-66 3.011198e-62 # KCNN2 360.4449 -5.504765 0.3234836 -17.01714 6.129148e-65 2.200854e-61 # USH2A 550.8569 -6.469569 0.3914803 -16.52591 2.388048e-61 6.125002e-58 count_log = log2(counts(dds, normalized=TRUE) + 1) gene_exp = count_log[\"MSMO1\", ,drop=F] %\u003e% t() %\u003e% as.data.frame() %\u003e% dplyr::mutate(Group=meta$Group) ggplot(gene_exp, aes(x=Group, y=MSMO1)) + geom_boxplot() 配对分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 library(DESeq2) dim(count_pair) # [1] 19938 16 head(meta_pair[,c(\"Group\",\"patient\")]) # Group patient # TCGA-W5-AA2X-11A-11R-A41I-07 NT TCGA-W5-AA2X # TCGA-W5-AA2Q-11A-11R-A41I-07 NT TCGA-W5-AA2Q # TCGA-W5-AA30-11A-11R-A41I-07 NT TCGA-W5-AA30 # TCGA-ZU-A8S4-11A-11R-A41I-07 NT TCGA-ZU-A8S4 # TCGA-W5-AA31-11A-11R-A41I-07 NT TCGA-W5-AA31 # TCGA-W5-AA2I-11A-11R-A41I-07 NT TCGA-W5-AA2I dds2 = DESeqDataSetFromMatrix(countData = count_pair, colData = meta_pair, design= ~ Group + patient) dds2 = DESeq(dds2) resultsNames(dds2) # lists the coefficients # [1] \"Intercept\" \"Group_TP_vs_NT\" # [3] \"patient_TCGA.W5.AA2Q_vs_TCGA.W5.AA2I\" \"patient_TCGA.W5.AA2U_vs_TCGA.W5.AA2I\" # [5] \"patient_TCGA.W5.AA2X_vs_TCGA.W5.AA2I\" \"patient_TCGA.W5.AA30_vs_TCGA.W5.AA2I\" # [7] \"patient_TCGA.W5.AA31_vs_TCGA.W5.AA2I\" \"patient_TCGA.W5.AA34_vs_TCGA.W5.AA2I\" # [9] \"patient_TCGA.ZU.A8S4_vs_TCGA.W5.AA2I\" res2 = results(dds2, name=\"Group_TP_vs_NT\") deg2 = as.data.frame(res2) deg2 = deg2[order(deg2$padj), ] head(deg2) # baseMean log2FoldChange lfcSE stat pvalue padj # PCK2 22699.219 -5.653594 0.1854230 -30.49025 3.508987e-204 6.059670e-200 # GRHPR 17379.340 -3.574706 0.1551980 -23.03319 2.168223e-117 1.872152e-113 # KDM8 1614.762 -4.555835 0.1989067 -22.90438 4.201674e-116 2.418624e-112 # CTH 2367.082 -5.254396 0.2324677 -22.60270 4.077093e-113 1.760183e-109 # ADI1 18162.701 -3.714171 0.1659464 -22.38174 5.927946e-111 2.047394e-107 # PBLD 8315.518 -5.683135 0.2594054 -21.90831 2.164593e-106 6.230059e-103 library(\"BiocParallel\") # Parallelizing DESeq, results, and lfcShrink # not supported on Windows register(MulticoreParam(4)) dds = DESeq(dds, parallel = TRUE) 3、edgeR 常规分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 library(edgeR) ##(1) DGEList对象 # 分组因子字符串：对照组在前；实验组在后 Group = factor(meta$Group, levels = c(\"NT\",\"TP\")) table(Group) # Group # NT TP # 9 35 dge.list = DGEList(counts=count, group=Group) dge.list$counts[1:4,1:4] dge.list$samples %\u003e% head() # group lib.size norm.factors # TCGA-W5-AA2X-01A-11R-A41I-07 TP 41184464 1 # TCGA-W5-AA2X-11A-11R-A41I-07 NT 36468189 1 # TCGA-W5-AA33-01A-11R-A41I-07 TP 43378374 1 ##(2) 计算文库因子 dge.list = calcNormFactors(dge.list) # Default method = 'TMM' dge.list$samples %\u003e% head() # group lib.size norm.factors # TCGA-W5-AA2X-01A-11R-A41I-07 TP 41184464 1.2238691 # TCGA-W5-AA2X-11A-11R-A41I-07 NT 36468189 0.6258154 # TCGA-W5-AA33-01A-11R-A41I-07 TP 43378374 1.3504490 ##(3) 根据分组设计，进行差异分析 design = model.matrix(~ Group) #### 如有其它协变量因素，放在主要变量之前即可，如下示例 # design = model.matrix(~ Gender + Age + Group) head(design) # (Intercept) GroupTP # 1 1 1 # 2 1 0 # 3 1 1 dge.list = estimateDisp(dge.list, design) fit = glmFit(dge.list, design) # Defaults to the last coefficient. fit = glmLRT(fit) deg = topTags(fit, n = Inf) deg = as.data.frame(deg) head(deg) # logFC logCPM LR PValue FDR # MSMO1 -3.636116 7.467688 371.8917 7.249780e-83 1.445461e-78 # GCDH -3.189232 5.737307 359.9146 2.938892e-80 2.929781e-76 # KCNN2 -5.382560 3.055588 344.2997 7.387479e-77 4.909718e-73 # RCL1 -3.799208 5.786302 338.9246 1.094155e-75 5.453816e-72 # SC5D -3.852832 6.657183 338.3414 1.465859e-75 5.845260e-72 # USH2A -6.352713 3.665099 329.5388 1.211265e-73 4.025034e-70 count_log = log2(cpm(dge.list) + 1) # TMM normalized count_log[1:4,1:4] gene_exp = count_log[\"MSMO1\", ,drop=F] %\u003e% t() %\u003e% as.data.frame() %\u003e% dplyr::mutate(Group=meta$Group) ggplot(gene_exp, aes(x=Group, y=MSMO1)) + geom_boxplot() 配对分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 library(edgeR) Group = factor(meta_pair$Group, levels = c(\"NT\",\"TP\")) # 对照组在前；实验组在后 Patient = meta_pair$patient dge.list = DGEList(counts=count_pair, group=Group) dge.list = calcNormFactors(dge.list) # Default method = 'TMM' design = model.matrix(~ Patient + Group) head(design) # (Intercept) PatientTCGA-W5-AA2Q PatientTCGA-W5-AA2U PatientTCGA-W5-AA2X PatientTCGA-W5-AA30 # 1 1 0 0 1 0 # 2 1 1 0 0 0 # 3 1 0 0 0 1 # 4 1 0 0 0 0 # 5 1 0 0 0 0 # 6 1 0 0 0 0 # PatientTCGA-W5-AA31 PatientTCGA-W5-AA34 PatientTCGA-ZU-A8S4 GroupTP # 1 0 0 0 0 # 2 0 0 0 0 # 3 0 0 0 0 # 4 0 0 1 0 # 5 1 0 0 0 # 6 0 0 0 0 dge.list = estimateDisp(dge.list, design) fit = glmQLFit(dge.list, design, robust = TRUE) fit = glmQLFTest(fit) # By default, the test is for the last coefficient in the design matrix, deg = topTags(fit, n = Inf) deg = as.data.frame(deg) head(deg) # logFC logCPM F PValue FDR # CTH -5.168010 5.735511 451.8950 1.891581e-10 2.184341e-06 # KDM8 -4.469608 5.191592 439.9909 2.191134e-10 2.184341e-06 # PCK2 -5.566911 8.997974 385.1885 4.551845e-10 2.240078e-06 # IQGAP3 5.382710 4.043167 378.8504 4.985585e-10 2.240078e-06 # EPB41L4B -3.747267 6.347686 360.6129 6.533619e-10 2.240078e-06 # GCH1 -3.670607 5.705010 358.5600 6.741130e-10 2.240078e-06 4、limma RNAseq差异分析 （1）常规分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 Group = factor(meta$Group, levels = c(\"NT\",\"TP\")) table(Group) # Group # NT TP # 9 35 design = model.matrix(~ Group) head(design) # (Intercept) GroupTP # 1 1 1 # 2 1 0 # 3 1 1 # 4 1 1 # 5 1 1 # 6 1 1 v = voom(count,design,normalize=\"quantile\") fit = lmFit(v,design) fit = eBayes(fit) colnames(coef(fit)) #[1] \"(Intercept)\" \"GroupTP\" tT = topTable(fit, adjust=\"fdr\", sort.by=\"p\", number=Inf) head(tT) # logFC AveExpr t P.Value adj.P.Val B # RMDN2 -2.900127 2.5323698 -28.67963 1.336773e-29 2.665258e-25 56.90314 # USH2A -6.535628 0.3847154 -23.73211 2.788056e-26 2.005247e-22 49.38508 # KCNN2 -5.460617 0.6268550 -23.68516 3.017223e-26 2.005247e-22 49.31589 # ESR1 -5.701844 1.4121136 -21.54736 1.278570e-24 6.373031e-21 45.80920 # AC132217.2 -5.094348 -5.1967122 -20.17979 1.659743e-23 6.430672e-20 42.06491 # SOX5 -4.944365 0.7859657 -20.10020 1.935201e-23 6.430672e-20 43.12095 （2）配对分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 Group = factor(meta_pair$Group, levels = c(\"NT\",\"TP\")) table(Group) # Group # NT TP # 8 8 Patient = factor(meta_pair$patient) table(Patient) Patient # TCGA-W5-AA2I TCGA-W5-AA2Q TCGA-W5-AA2U TCGA-W5-AA2X TCGA-W5-AA30 TCGA-W5-AA31 TCGA-W5-AA34 TCGA-ZU-A8S4 # 2 2 2 2 2 2 2 2 design = model.matrix(~ Patient + Group) head(design) # (Intercept) PatientTCGA-W5-AA2Q PatientTCGA-W5-AA2U PatientTCGA-W5-AA2X PatientTCGA-W5-AA30 # 1 1 0 0 1 0 # 2 1 1 0 0 0 # 3 1 0 0 0 1 # 4 1 0 0 0 0 # 5 1 0 0 0 0 # 6 1 0 0 0 0 # PatientTCGA-W5-AA31 PatientTCGA-W5-AA34 PatientTCGA-ZU-A8S4 GroupTP # 1 0 0 0 0 # 2 0 0 0 0 # 3 0 0 0 0 # 4 0 0 1 0 # 5 1 0 0 0 # 6 0 0 0 0 v = voom(count_pair, design, normalize=\"quantile\") fit = lmFit(v,design) fit = eBayes(fit) colnames(coef(fit)) # [1] \"(Intercept)\" \"PatientTCGA-W5-AA2Q\" \"PatientTCGA-W5-AA2U\" \"PatientTCGA-W5-AA2X\" # [5] \"PatientTCGA-W5-AA30\" \"PatientTCGA-W5-AA31\" \"PatientTCGA-W5-AA34\" \"PatientTCGA-ZU-A8S4\" # [9] \"GroupTP\" tT = topTable(fit, coef = \"GroupTP\", adjust=\"fdr\", sort.by=\"p\", number=Inf) head(tT) # logFC AveExpr t P.Value adj.P.Val B # BTD -2.500130 5.412028 -37.40476 5.067356e-10 8.241491e-06 13.61382 # MCM3 2.535346 5.765008 35.09944 8.267119e-10 8.241491e-06 13.22475 # ABCC10 1.768879 3.760942 31.34388 1.973016e-09 9.926863e-06 12.32708 # FGGY -2.559748 4.447616 -29.91541 2.822782e-09 9.926863e-06 12.07469 # ZCCHC2 -1.924153 4.452133 -29.51382 3.131304e-09 9.926863e-06 12.01657 # ACOT1 -4.146553 1.943388 -28.79614 3.782277e-09 9.926863e-06 11.27019 芯片数据差异分析 （1）GEO芯片数据 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 library(GEOquery) Sys.setenv(\"VROOM_CONNECTION_SIZE\" = 131072 * 10) gse = getGEO('GSE109169', getGPL = F) gse = gse[[1]] exp_dat = exprs(gse) %\u003e% as.data.frame() dim(exp_dat) # [1] 19076 50 exp_dat[1:4,1:4] # GSM2934481 GSM2934482 GSM2934483 GSM2934484 # 2315554 7.31 7.27 7.63 7.38 # 2315633 7.24 7.52 7.42 7.37 # 2315674 7.36 7.42 7.61 7.31 # 2315739 7.24 7.56 7.76 7.46 meta = pData(gse) %\u003e% dplyr::select(title, dplyr::ends_with(\"ch1\")) colnames(meta)=gsub(\"[:_]ch1\", \"\", colnames(meta)) meta$Group = ifelse(meta$source_name==\"breast cancer tissue\", \"tumor\", \"normal\") # normal tumor # 25 25 meta$Patient = stringr::str_split(meta$title, \"-\", simplify=T)[,1] table(meta$Patient) # 2102 2110 2146 2181 2190 2280 2297 2353 2365 2397 2451 2454 2706 2780 2910 2917 2943 2972 2985 3156 3159 # 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 # 3197 3232 3297 3355 # 2 2 2 2 head(meta[,c(\"Group\",\"Patient\")]) # Group Patient # GSM2934481 normal 2102 # GSM2934482 tumor 2102 # GSM2934483 normal 2110 # GSM2934484 tumor 2110 # GSM2934485 normal 2146 # GSM2934486 tumor 2146 identical(rownames(meta), colnames(exp_dat)) # [1] TRUE （2）常规分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 Group = factor(meta$Group, levels = c(\"normal\",\"tumor\")) table(Group) # Group # normal tumor # 25 25 design=model.matrix(~ Group) head(design) # (Intercept) Grouptumor # 1 1 0 # 2 1 1 # 3 1 0 # 4 1 1 # 5 1 0 # 6 1 1 fit = lmFit(exp_dat, design) fit = eBayes(fit) colnames(coef(fit)) # [1] \"(Intercept)\" \"Grouptumor\" tT = topTable(fit, adjust=\"fdr\", sort.by=\"p\", number=Inf) head(tT) # logFC AveExpr t P.Value adj.P.Val B # 4004044 -2.6860 6.8706 -21.16464 6.424093e-27 1.225460e-22 50.47849 # 4000538 -4.1400 6.6420 -19.36848 3.622286e-25 3.454937e-21 46.66686 # 2591421 -2.2128 7.1468 -16.18728 9.688784e-22 6.160775e-18 39.10226 # 3830189 -1.8416 8.0504 -16.03976 1.433530e-21 6.836502e-18 38.72368 # 3519309 -2.1232 7.7964 -15.30913 1.034876e-20 3.948259e-17 36.80964 # 3745525 -1.0784 6.1008 -15.21411 1.344319e-20 4.274038e-17 36.55587 （3）配对分析 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 Group = factor(meta$Group, levels = c(\"normal\",\"tumor\")) table(Group) # Group # normal tumor # 25 25 Patient = factor(meta$Patient) table(Patient) # Patient # 2102 2110 2146 2181 2190 2280 2297 2353 2365 2397 2451 2454 2706 2780 2910 2917 2943 2972 2985 3156 3159 # 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 # 3197 3232 3297 3355 # 2 2 2 2 design=model.matrix(~ Patient + Group) head(design) # Groupnormal Grouptumor # 1 1 0 # 2 0 1 # 3 1 0 fit = lmFit(exp_dat, design) fit \u003c- eBayes(fit) colnames(coef(fit)) # [1] \"(Intercept)\" \"Patient2110\" \"Patient2146\" \"Patient2181\" \"Patient2190\" \"Patient2280\" \"Patient2297\" # [8] \"Patient2353\" \"Patient2365\" \"Patient2397\" \"Patient2451\" \"Patient2454\" \"Patient2706\" \"Patient2780\" # [15] \"Patient2910\" \"Patient2917\" \"Patient2943\" \"Patient2972\" \"Patient2985\" \"Patient3156\" \"Patient3159\" # [22] \"Patient3197\" \"Patient3232\" \"Patient3297\" \"Patient3355\" \"Grouptumor\" tT \u003c- topTable(fit, coef = \"Grouptumor\", adjust=\"fdr\", sort.by=\"p\", number=Inf) head(tT) # logFC AveExpr t P.Value adj.P.Val B # 4004044 -2.6860 6.8706 -23.34059 1.764542e-19 3.366041e-15 33.97414 # 4000538 -4.1400 6.6420 -21.21050 2.072006e-18 1.976279e-14 31.71755 # 3830189 -1.8416 8.0504 -19.34286 2.168752e-17 1.379037e-13 29.52627 # 2701109 -2.4704 7.0796 -17.61105 2.298621e-16 1.096212e-12 27.28943 # 3062082 -2.9824 9.0236 -17.27304 3.728465e-16 1.422484e-12 26.82740 # 2591421 -2.2128 7.1468 -17.03824 5.241945e-16 1.564468e-12 26.50125 ",
  "wordCount" : "2476",
  "inLanguage": "en",
  "datePublished": "2022-06-11T00:00:00Z",
  "dateModified": "2022-06-11T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/045%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90r%E5%8C%85-deseq2+edger+limma/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      差异分析R包-DESeq2&#43;edgeR&#43;limma
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2022-06-11 00:00:00 &#43;0000 UTC&#39;&gt;2022-06-11&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-06-11&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2476&amp;nbsp;|&amp;nbsp;5 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e7%a4%ba%e4%be%8bcount%e8%a1%a8%e8%be%be%e7%9f%a9%e9%98%b5" aria-label="1、示例count表达矩阵">1、示例count表达矩阵</a></li>
                    <li>
                        <a href="#2deseq2" aria-label="2、DESeq2">2、DESeq2</a><ul>
                            
                    <li>
                        <a href="#%e5%b8%b8%e8%a7%84%e5%88%86%e6%9e%90" aria-label="常规分析">常规分析</a></li>
                    <li>
                        <a href="#%e9%85%8d%e5%af%b9%e5%88%86%e6%9e%90" aria-label="配对分析">配对分析</a></li></ul>
                    </li>
                    <li>
                        <a href="#3edger" aria-label="3、edgeR">3、edgeR</a><ul>
                            
                    <li>
                        <a href="#%e5%b8%b8%e8%a7%84%e5%88%86%e6%9e%90-1" aria-label="常规分析">常规分析</a></li>
                    <li>
                        <a href="#%e9%85%8d%e5%af%b9%e5%88%86%e6%9e%90-1" aria-label="配对分析">配对分析</a></li></ul>
                    </li>
                    <li>
                        <a href="#4limma" aria-label="4、limma">4、limma</a><ul>
                            
                    <li>
                        <a href="#rnaseq%e5%b7%ae%e5%bc%82%e5%88%86%e6%9e%90" aria-label="RNAseq差异分析">RNAseq差异分析</a><ul>
                            
                    <li>
                        <a href="#1%e5%b8%b8%e8%a7%84%e5%88%86%e6%9e%90" aria-label="（1）常规分析">（1）常规分析</a></li>
                    <li>
                        <a href="#2%e9%85%8d%e5%af%b9%e5%88%86%e6%9e%90" aria-label="（2）配对分析">（2）配对分析</a></li></ul>
                    </li>
                    <li>
                        <a href="#%e8%8a%af%e7%89%87%e6%95%b0%e6%8d%ae%e5%b7%ae%e5%bc%82%e5%88%86%e6%9e%90" aria-label="芯片数据差异分析">芯片数据差异分析</a><ul>
                            
                    <li>
                        <a href="#1geo%e8%8a%af%e7%89%87%e6%95%b0%e6%8d%ae" aria-label="（1）GEO芯片数据">（1）GEO芯片数据</a></li>
                    <li>
                        <a href="#2%e5%b8%b8%e8%a7%84%e5%88%86%e6%9e%90" aria-label="（2）常规分析">（2）常规分析</a></li>
                    <li>
                        <a href="#3%e9%85%8d%e5%af%b9%e5%88%86%e6%9e%90" aria-label="（3）配对分析">（3）配对分析</a>
                    </li>
                </ul>
                </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><p>在对转录组数据分析时，分组比较的差异分析前提时获得测序表达矩阵。</p>
<p>根据测序技术分为两种，对应的分析R包如下所示。</p>
<ul>
<li>RNAseq的原始count表达矩阵
<ul>
<li>DESeq2</li>
<li>edgeR</li>
<li>limma</li>
</ul>
</li>
<li>芯片array标准化后的表达矩阵
<ul>
<li>limma</li>
</ul>
</li>
</ul>
<h1 id="1示例count表达矩阵">1、示例count表达矩阵<a hidden class="anchor" aria-hidden="true" href="#1示例count表达矩阵">#</a></h1>
<p>如下使用TCGAbiolinks包下载共9个癌症样本、35个正常样本的表达矩阵；其中涉及到8对配对样本（tumor与normal来自同一病人）</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">43
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">44
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">45
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">46
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">47
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">48
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">49
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">50
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">51
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">52
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">53
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(TCGAbiolinks)
</span></span><span style="display:flex;"><span>library(SummarizedExperiment)
</span></span><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>query &lt;- GDCquery(project = <span style="color:#0ff;font-weight:bold">&#34;TCGA-CHOL&#34;</span>,
</span></span><span style="display:flex;"><span>                  legacy = <span style="color:#fff;font-weight:bold">FALSE</span>,
</span></span><span style="display:flex;"><span>                  data.category = <span style="color:#0ff;font-weight:bold">&#34;Transcriptome Profiling&#34;</span>,
</span></span><span style="display:flex;"><span>                  data.type = <span style="color:#0ff;font-weight:bold">&#34;Gene Expression Quantification&#34;</span>)
</span></span><span style="display:flex;"><span>GDCdownload(query, files.per.chunk = <span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span>data = GDCprepare(query, save = <span style="color:#fff;font-weight:bold">F</span>)
</span></span><span style="display:flex;"><span>rownames(data) = rowData(data)$gene_name
</span></span><span style="display:flex;"><span>data = data[rowData(data)$gene_type==<span style="color:#0ff;font-weight:bold">&#34;protein_coding&#34;</span>,]
</span></span><span style="display:flex;"><span>count = assay(data, <span style="color:#0ff;font-weight:bold">&#34;unstranded&#34;</span>)
</span></span><span style="display:flex;"><span>count = count[!duplicated(rownames(count)),]
</span></span><span style="display:flex;"><span>dim(count)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 19938    44</span>
</span></span><span style="display:flex;"><span>count[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        TCGA-W5-AA2X-01A-11R-A41I-07 TCGA-W5-AA2X-11A-11R-A41I-07</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TSPAN6                         3310                         2322</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TNMD                              0                            0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DPM1                           1881                          716</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SCYL3                           966                          315</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        TCGA-W5-AA33-01A-11R-A41I-07 TCGA-ZH-A8Y2-01A-11R-A41I-07</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TSPAN6                        11312                         6768</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TNMD                              1                            0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DPM1                            968                         1273</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SCYL3                          1296                          774</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>meta = colData(data)[,c(<span style="color:#0ff;font-weight:bold">&#34;barcode&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;sample&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;patient&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;shortLetterCode&#34;</span>)] %&gt;% 
</span></span><span style="display:flex;"><span>  as.data.frame()
</span></span><span style="display:flex;"><span>meta$Group = meta$shortLetterCode <span style="color:#007f7f"># NT: Solid Tissue Normal; TP: Primary Tumor </span>
</span></span><span style="display:flex;"><span>head(meta[,<span style="color:#0ff;font-weight:bold">&#34;Group&#34;</span>,drop=<span style="color:#fff;font-weight:bold">F</span>])
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                              Group</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2X-01A-11R-A41I-07    TP</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2X-11A-11R-A41I-07    NT</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA33-01A-11R-A41I-07    TP</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-ZH-A8Y2-01A-11R-A41I-07    TP</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-ZH-A8Y6-01A-11R-A41I-07    TP</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2Z-01A-11R-A41I-07    TP</span>
</span></span><span style="display:flex;"><span>table(meta$Group)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NT TP </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 9 35</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>identical(colnames(count), rownames(meta))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] TRUE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 8对配对样本</span>
</span></span><span style="display:flex;"><span>meta_pair = meta[TCGAquery_MatchedCoupledSampleTypes(meta$barcode, c(<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>)), ]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2I TCGA-W5-AA2Q TCGA-W5-AA2U TCGA-W5-AA2X TCGA-W5-AA30 TCGA-W5-AA31 TCGA-W5-AA34 TCGA-ZU-A8S4 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2            2            2            2            2            2            2            2 </span>
</span></span><span style="display:flex;"><span>count_pair = count[, meta_pair$barcode]
</span></span><span style="display:flex;"><span>identical(meta_pair$barcode, colnames(count_pair))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1] TRUE</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="2deseq2">2、DESeq2<a hidden class="anchor" aria-hidden="true" href="#2deseq2">#</a></h1>
<h2 id="常规分析">常规分析<a hidden class="anchor" aria-hidden="true" href="#常规分析">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(DESeq2)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 分组因子化：对照组在前，实验组在后</span>
</span></span><span style="display:flex;"><span>meta$Group = factor(meta$Group, levels = c(<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>))
</span></span><span style="display:flex;"><span>dds = DESeqDataSetFromMatrix(countData = count,
</span></span><span style="display:flex;"><span>                              colData = meta,
</span></span><span style="display:flex;"><span>                              design= ~ Group)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># keep &lt;- rowSums(counts(dds)) &gt;= 10</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># dds &lt;- dds[keep,]</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>dds = DESeq(dds)
</span></span><span style="display:flex;"><span>resultsNames(dds) <span style="color:#007f7f"># lists the coefficients</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;Intercept&#34;      &#34;Group_TP_vs_NT&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>res = results(dds, name=<span style="color:#0ff;font-weight:bold">&#34;Group_TP_vs_NT&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>deg = as.data.frame(res)
</span></span><span style="display:flex;"><span>deg = deg[order(deg$padj), ]
</span></span><span style="display:flex;"><span>head(deg)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        baseMean log2FoldChange     lfcSE      stat       pvalue         padj</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MSMO1 7585.3612      -3.748488 0.2109579 -17.76889 1.231028e-70 1.105094e-66</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># RMDN2  396.7567      -2.929035 0.1648278 -17.77027 1.201019e-70 1.105094e-66</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GCDH  2272.9263      -3.302603 0.1863973 -17.71809 3.040673e-70 1.819742e-66</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCAIM 1239.0505      -2.235421 0.1303740 -17.14621 6.708695e-66 3.011198e-62</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># KCNN2  360.4449      -5.504765 0.3234836 -17.01714 6.129148e-65 2.200854e-61</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># USH2A  550.8569      -6.469569 0.3914803 -16.52591 2.388048e-61 6.125002e-58</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>count_log = log2(counts(dds, normalized=<span style="color:#fff;font-weight:bold">TRUE</span>) + <span style="color:#ff0;font-weight:bold">1</span>)
</span></span><span style="display:flex;"><span>gene_exp = count_log[<span style="color:#0ff;font-weight:bold">&#34;MSMO1&#34;</span>, ,drop=<span style="color:#fff;font-weight:bold">F</span>] %&gt;% 
</span></span><span style="display:flex;"><span>  t() %&gt;% as.data.frame() %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::mutate(Group=meta$Group) 
</span></span><span style="display:flex;"><span>ggplot(gene_exp, aes(x=Group, y=MSMO1)) + 
</span></span><span style="display:flex;"><span>  geom_boxplot()
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220611150552173.png" alt="image-20220611150552173" style="zoom:50%;" />
<h2 id="配对分析">配对分析<a hidden class="anchor" aria-hidden="true" href="#配对分析">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(DESeq2)
</span></span><span style="display:flex;"><span>dim(count_pair)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 19938    16</span>
</span></span><span style="display:flex;"><span>head(meta_pair[,c(<span style="color:#0ff;font-weight:bold">&#34;Group&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;patient&#34;</span>)])
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                              Group      patient</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2X-11A-11R-A41I-07    NT TCGA-W5-AA2X</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2Q-11A-11R-A41I-07    NT TCGA-W5-AA2Q</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA30-11A-11R-A41I-07    NT TCGA-W5-AA30</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-ZU-A8S4-11A-11R-A41I-07    NT TCGA-ZU-A8S4</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA31-11A-11R-A41I-07    NT TCGA-W5-AA31</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2I-11A-11R-A41I-07    NT TCGA-W5-AA2I</span>
</span></span><span style="display:flex;"><span>dds2 = DESeqDataSetFromMatrix(countData = count_pair,
</span></span><span style="display:flex;"><span>                             colData = meta_pair,
</span></span><span style="display:flex;"><span>                             design= ~ Group + patient)
</span></span><span style="display:flex;"><span>dds2 = DESeq(dds2)
</span></span><span style="display:flex;"><span>resultsNames(dds2) <span style="color:#007f7f"># lists the coefficients</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;Intercept&#34;                            &#34;Group_TP_vs_NT&#34;                      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [3] &#34;patient_TCGA.W5.AA2Q_vs_TCGA.W5.AA2I&#34; &#34;patient_TCGA.W5.AA2U_vs_TCGA.W5.AA2I&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;patient_TCGA.W5.AA2X_vs_TCGA.W5.AA2I&#34; &#34;patient_TCGA.W5.AA30_vs_TCGA.W5.AA2I&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [7] &#34;patient_TCGA.W5.AA31_vs_TCGA.W5.AA2I&#34; &#34;patient_TCGA.W5.AA34_vs_TCGA.W5.AA2I&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [9] &#34;patient_TCGA.ZU.A8S4_vs_TCGA.W5.AA2I&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>res2 = results(dds2, name=<span style="color:#0ff;font-weight:bold">&#34;Group_TP_vs_NT&#34;</span>)
</span></span><span style="display:flex;"><span>deg2 = as.data.frame(res2)
</span></span><span style="display:flex;"><span>deg2 = deg2[order(deg2$padj), ]
</span></span><span style="display:flex;"><span>head(deg2)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        baseMean log2FoldChange     lfcSE      stat        pvalue          padj</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># PCK2  22699.219      -5.653594 0.1854230 -30.49025 3.508987e-204 6.059670e-200</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GRHPR 17379.340      -3.574706 0.1551980 -23.03319 2.168223e-117 1.872152e-113</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># KDM8   1614.762      -4.555835 0.1989067 -22.90438 4.201674e-116 2.418624e-112</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CTH    2367.082      -5.254396 0.2324677 -22.60270 4.077093e-113 1.760183e-109</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ADI1  18162.701      -3.714171 0.1659464 -22.38174 5.927946e-111 2.047394e-107</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># PBLD   8315.518      -5.683135 0.2594054 -21.90831 2.164593e-106 6.230059e-103</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>library(<span style="color:#0ff;font-weight:bold">&#34;BiocParallel&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Parallelizing DESeq, results, and lfcShrink</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># not supported on Windows</span>
</span></span><span style="display:flex;"><span>register(MulticoreParam(<span style="color:#ff0;font-weight:bold">4</span>))
</span></span><span style="display:flex;"><span>dds = DESeq(dds, parallel = <span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="3edger">3、edgeR<a hidden class="anchor" aria-hidden="true" href="#3edger">#</a></h1>
<h2 id="常规分析-1">常规分析<a hidden class="anchor" aria-hidden="true" href="#常规分析-1">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">43
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">44
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">45
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">46
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">47
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">48
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">49
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">50
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">51
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">52
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">53
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">54
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">55
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">56
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">57
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(edgeR)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(1) DGEList对象</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 分组因子字符串：对照组在前；实验组在后</span>
</span></span><span style="display:flex;"><span>Group = factor(meta$Group, levels = c(<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>))
</span></span><span style="display:flex;"><span>table(Group)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Group</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NT TP </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 9 35</span>
</span></span><span style="display:flex;"><span>dge.list = DGEList(counts=count,
</span></span><span style="display:flex;"><span>                    group=Group)
</span></span><span style="display:flex;"><span>dge.list$counts[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span>dge.list$samples %&gt;% head()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                              group lib.size norm.factors</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2X-01A-11R-A41I-07    TP 41184464            1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2X-11A-11R-A41I-07    NT 36468189            1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA33-01A-11R-A41I-07    TP 43378374            1</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) 计算文库因子</span>
</span></span><span style="display:flex;"><span>dge.list = calcNormFactors(dge.list)  <span style="color:#007f7f"># Default  method = &#39;TMM&#39;</span>
</span></span><span style="display:flex;"><span>dge.list$samples %&gt;% head()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                              group lib.size norm.factors</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2X-01A-11R-A41I-07    TP 41184464    1.2238691</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2X-11A-11R-A41I-07    NT 36468189    0.6258154</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA33-01A-11R-A41I-07    TP 43378374    1.3504490 </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) 根据分组设计，进行差异分析</span>
</span></span><span style="display:flex;"><span>design = model.matrix(~ Group)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#### 如有其它协变量因素，放在主要变量之前即可，如下示例</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># design = model.matrix(~ Gender + Age + Group)</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>head(design)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   (Intercept) GroupTP</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1           1       1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2           1       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3           1       1</span>
</span></span><span style="display:flex;"><span>dge.list = estimateDisp(dge.list, design)
</span></span><span style="display:flex;"><span>fit = glmFit(dge.list, design)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Defaults to the last coefficient.</span>
</span></span><span style="display:flex;"><span>fit = glmLRT(fit) 
</span></span><span style="display:flex;"><span>deg = topTags(fit, n = <span style="color:#fff;font-weight:bold">Inf</span>)
</span></span><span style="display:flex;"><span>deg = as.data.frame(deg)
</span></span><span style="display:flex;"><span>head(deg)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#          logFC   logCPM       LR       PValue          FDR</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MSMO1 -3.636116 7.467688 371.8917 7.249780e-83 1.445461e-78</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GCDH  -3.189232 5.737307 359.9146 2.938892e-80 2.929781e-76</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># KCNN2 -5.382560 3.055588 344.2997 7.387479e-77 4.909718e-73</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># RCL1  -3.799208 5.786302 338.9246 1.094155e-75 5.453816e-72</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SC5D  -3.852832 6.657183 338.3414 1.465859e-75 5.845260e-72</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># USH2A -6.352713 3.665099 329.5388 1.211265e-73 4.025034e-70</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>count_log = log2(cpm(dge.list) + <span style="color:#ff0;font-weight:bold">1</span>)   <span style="color:#007f7f"># TMM normalized </span>
</span></span><span style="display:flex;"><span>count_log[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span>gene_exp = count_log[<span style="color:#0ff;font-weight:bold">&#34;MSMO1&#34;</span>, ,drop=<span style="color:#fff;font-weight:bold">F</span>] %&gt;% 
</span></span><span style="display:flex;"><span>  t() %&gt;% as.data.frame() %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::mutate(Group=meta$Group) 
</span></span><span style="display:flex;"><span>ggplot(gene_exp, aes(x=Group, y=MSMO1)) + 
</span></span><span style="display:flex;"><span>  geom_boxplot()
</span></span></code></pre></td></tr></table>
</div>
</div><img src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220611153117041.png" alt="image-20220611153117041" style="zoom:50%;" />
<h2 id="配对分析-1">配对分析<a hidden class="anchor" aria-hidden="true" href="#配对分析-1">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(edgeR)
</span></span><span style="display:flex;"><span>Group = factor(meta_pair$Group, levels = c(<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 对照组在前；实验组在后</span>
</span></span><span style="display:flex;"><span>Patient = meta_pair$patient
</span></span><span style="display:flex;"><span>dge.list = DGEList(counts=count_pair,
</span></span><span style="display:flex;"><span>                   group=Group)
</span></span><span style="display:flex;"><span>dge.list = calcNormFactors(dge.list)  <span style="color:#007f7f"># Default  method = &#39;TMM&#39;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>design = model.matrix(~ Patient + Group)
</span></span><span style="display:flex;"><span>head(design)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   (Intercept) PatientTCGA-W5-AA2Q PatientTCGA-W5-AA2U PatientTCGA-W5-AA2X PatientTCGA-W5-AA30</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1           1                   0                   0                   1                   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2           1                   1                   0                   0                   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3           1                   0                   0                   0                   1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4           1                   0                   0                   0                   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5           1                   0                   0                   0                   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6           1                   0                   0                   0                   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   PatientTCGA-W5-AA31 PatientTCGA-W5-AA34 PatientTCGA-ZU-A8S4 GroupTP</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1                   0                   0                   0       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2                   0                   0                   0       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3                   0                   0                   0       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4                   0                   0                   1       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5                   1                   0                   0       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6                   0                   0                   0       0</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>dge.list = estimateDisp(dge.list, design)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>fit = glmQLFit(dge.list, design, robust = <span style="color:#fff;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>fit = glmQLFTest(fit) 
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># By default, the test is for the last coefficient in the design matrix,</span>
</span></span><span style="display:flex;"><span>deg = topTags(fit, n = <span style="color:#fff;font-weight:bold">Inf</span>)
</span></span><span style="display:flex;"><span>deg = as.data.frame(deg)
</span></span><span style="display:flex;"><span>head(deg)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#              logFC   logCPM        F       PValue          FDR</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CTH      -5.168010 5.735511 451.8950 1.891581e-10 2.184341e-06</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># KDM8     -4.469608 5.191592 439.9909 2.191134e-10 2.184341e-06</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># PCK2     -5.566911 8.997974 385.1885 4.551845e-10 2.240078e-06</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># IQGAP3    5.382710 4.043167 378.8504 4.985585e-10 2.240078e-06</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># EPB41L4B -3.747267 6.347686 360.6129 6.533619e-10 2.240078e-06</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GCH1     -3.670607 5.705010 358.5600 6.741130e-10 2.240078e-06</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="4limma">4、limma<a hidden class="anchor" aria-hidden="true" href="#4limma">#</a></h1>
<h2 id="rnaseq差异分析">RNAseq差异分析<a hidden class="anchor" aria-hidden="true" href="#rnaseq差异分析">#</a></h2>
<h3 id="1常规分析">（1）常规分析<a hidden class="anchor" aria-hidden="true" href="#1常规分析">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>Group = factor(meta$Group, levels = c(<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>))
</span></span><span style="display:flex;"><span>table(Group)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Group</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NT TP </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 9 35</span>
</span></span><span style="display:flex;"><span>design = model.matrix(~ Group)
</span></span><span style="display:flex;"><span>head(design)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   (Intercept) GroupTP</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1           1       1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2           1       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3           1       1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4           1       1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5           1       1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6           1       1</span>
</span></span><span style="display:flex;"><span>v = voom(count,design,normalize=<span style="color:#0ff;font-weight:bold">&#34;quantile&#34;</span>)
</span></span><span style="display:flex;"><span>fit = lmFit(v,design)
</span></span><span style="display:flex;"><span>fit = eBayes(fit)
</span></span><span style="display:flex;"><span>colnames(coef(fit))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1] &#34;(Intercept)&#34; &#34;GroupTP&#34;</span>
</span></span><span style="display:flex;"><span>tT = topTable(fit, adjust=<span style="color:#0ff;font-weight:bold">&#34;fdr&#34;</span>, sort.by=<span style="color:#0ff;font-weight:bold">&#34;p&#34;</span>, number=<span style="color:#fff;font-weight:bold">Inf</span>)
</span></span><span style="display:flex;"><span>head(tT)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                logFC    AveExpr         t      P.Value    adj.P.Val        B</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># RMDN2      -2.900127  2.5323698 -28.67963 1.336773e-29 2.665258e-25 56.90314</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># USH2A      -6.535628  0.3847154 -23.73211 2.788056e-26 2.005247e-22 49.38508</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># KCNN2      -5.460617  0.6268550 -23.68516 3.017223e-26 2.005247e-22 49.31589</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ESR1       -5.701844  1.4121136 -21.54736 1.278570e-24 6.373031e-21 45.80920</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AC132217.2 -5.094348 -5.1967122 -20.17979 1.659743e-23 6.430672e-20 42.06491</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SOX5       -4.944365  0.7859657 -20.10020 1.935201e-23 6.430672e-20 43.12095</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="2配对分析">（2）配对分析<a hidden class="anchor" aria-hidden="true" href="#2配对分析">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>Group = factor(meta_pair$Group, levels = c(<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>))
</span></span><span style="display:flex;"><span>table(Group)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Group</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NT TP </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 8  8 </span>
</span></span><span style="display:flex;"><span>Patient = factor(meta_pair$patient)
</span></span><span style="display:flex;"><span>table(Patient)
</span></span><span style="display:flex;"><span>Patient
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2I TCGA-W5-AA2Q TCGA-W5-AA2U TCGA-W5-AA2X TCGA-W5-AA30 TCGA-W5-AA31 TCGA-W5-AA34 TCGA-ZU-A8S4 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2            2            2            2            2            2            2            2 </span>
</span></span><span style="display:flex;"><span>design = model.matrix(~ Patient + Group)
</span></span><span style="display:flex;"><span>head(design)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   (Intercept) PatientTCGA-W5-AA2Q PatientTCGA-W5-AA2U PatientTCGA-W5-AA2X PatientTCGA-W5-AA30</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1           1                   0                   0                   1                   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2           1                   1                   0                   0                   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3           1                   0                   0                   0                   1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4           1                   0                   0                   0                   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5           1                   0                   0                   0                   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6           1                   0                   0                   0                   0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   PatientTCGA-W5-AA31 PatientTCGA-W5-AA34 PatientTCGA-ZU-A8S4 GroupTP</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1                   0                   0                   0       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2                   0                   0                   0       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3                   0                   0                   0       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4                   0                   0                   1       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5                   1                   0                   0       0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6                   0                   0                   0       0</span>
</span></span><span style="display:flex;"><span>v = voom(count_pair, design, normalize=<span style="color:#0ff;font-weight:bold">&#34;quantile&#34;</span>)
</span></span><span style="display:flex;"><span>fit = lmFit(v,design)
</span></span><span style="display:flex;"><span>fit = eBayes(fit)
</span></span><span style="display:flex;"><span>colnames(coef(fit))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;(Intercept)&#34;         &#34;PatientTCGA-W5-AA2Q&#34; &#34;PatientTCGA-W5-AA2U&#34; &#34;PatientTCGA-W5-AA2X&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;PatientTCGA-W5-AA30&#34; &#34;PatientTCGA-W5-AA31&#34; &#34;PatientTCGA-W5-AA34&#34; &#34;PatientTCGA-ZU-A8S4&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [9] &#34;GroupTP&#34; </span>
</span></span><span style="display:flex;"><span>tT = topTable(fit, coef = <span style="color:#0ff;font-weight:bold">&#34;GroupTP&#34;</span>, adjust=<span style="color:#0ff;font-weight:bold">&#34;fdr&#34;</span>, sort.by=<span style="color:#0ff;font-weight:bold">&#34;p&#34;</span>, number=<span style="color:#fff;font-weight:bold">Inf</span>)
</span></span><span style="display:flex;"><span>head(tT)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            logFC  AveExpr         t      P.Value    adj.P.Val        B</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># BTD    -2.500130 5.412028 -37.40476 5.067356e-10 8.241491e-06 13.61382</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MCM3    2.535346 5.765008  35.09944 8.267119e-10 8.241491e-06 13.22475</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ABCC10  1.768879 3.760942  31.34388 1.973016e-09 9.926863e-06 12.32708</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># FGGY   -2.559748 4.447616 -29.91541 2.822782e-09 9.926863e-06 12.07469</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ZCCHC2 -1.924153 4.452133 -29.51382 3.131304e-09 9.926863e-06 12.01657</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ACOT1  -4.146553 1.943388 -28.79614 3.782277e-09 9.926863e-06 11.27019</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="芯片数据差异分析">芯片数据差异分析<a hidden class="anchor" aria-hidden="true" href="#芯片数据差异分析">#</a></h2>
<h3 id="1geo芯片数据">（1）GEO芯片数据<a hidden class="anchor" aria-hidden="true" href="#1geo芯片数据">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(GEOquery)
</span></span><span style="display:flex;"><span>Sys.setenv(<span style="color:#0ff;font-weight:bold">&#34;VROOM_CONNECTION_SIZE&#34;</span> = <span style="color:#ff0;font-weight:bold">131072</span> * <span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span>gse = getGEO(<span style="color:#0ff;font-weight:bold">&#39;GSE109169&#39;</span>, getGPL = <span style="color:#fff;font-weight:bold">F</span>) 
</span></span><span style="display:flex;"><span>gse = gse[[1]]
</span></span><span style="display:flex;"><span>exp_dat = exprs(gse) %&gt;% as.data.frame()
</span></span><span style="display:flex;"><span>dim(exp_dat)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 19076    50</span>
</span></span><span style="display:flex;"><span>exp_dat[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#         GSM2934481 GSM2934482 GSM2934483 GSM2934484</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2315554       7.31       7.27       7.63       7.38</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2315633       7.24       7.52       7.42       7.37</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2315674       7.36       7.42       7.61       7.31</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2315739       7.24       7.56       7.76       7.46</span>
</span></span><span style="display:flex;"><span>meta = pData(gse) %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::select(title, dplyr::ends_with(<span style="color:#0ff;font-weight:bold">&#34;ch1&#34;</span>))
</span></span><span style="display:flex;"><span>colnames(meta)=gsub(<span style="color:#0ff;font-weight:bold">&#34;[:_]ch1&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;&#34;</span>, colnames(meta))
</span></span><span style="display:flex;"><span>meta$Group = ifelse(meta$source_name==<span style="color:#0ff;font-weight:bold">&#34;breast cancer tissue&#34;</span>,
</span></span><span style="display:flex;"><span>                    <span style="color:#0ff;font-weight:bold">&#34;tumor&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;normal&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># normal  tumor </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 25     25 </span>
</span></span><span style="display:flex;"><span>meta$Patient = stringr::str_split(meta$title, <span style="color:#0ff;font-weight:bold">&#34;-&#34;</span>, simplify=<span style="color:#fff;font-weight:bold">T</span>)[,<span style="color:#ff0;font-weight:bold">1</span>]
</span></span><span style="display:flex;"><span>table(meta$Patient)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2102 2110 2146 2181 2190 2280 2297 2353 2365 2397 2451 2454 2706 2780 2910 2917 2943 2972 2985 3156 3159 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3197 3232 3297 3355 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2    2    2    2 </span>
</span></span><span style="display:flex;"><span>head(meta[,c(<span style="color:#0ff;font-weight:bold">&#34;Group&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;Patient&#34;</span>)])
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#             Group Patient</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GSM2934481 normal    2102</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GSM2934482  tumor    2102</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GSM2934483 normal    2110</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GSM2934484  tumor    2110</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GSM2934485 normal    2146</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GSM2934486  tumor    2146</span>
</span></span><span style="display:flex;"><span>identical(rownames(meta), colnames(exp_dat))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] TRUE</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="2常规分析">（2）常规分析<a hidden class="anchor" aria-hidden="true" href="#2常规分析">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>Group = factor(meta$Group, levels = c(<span style="color:#0ff;font-weight:bold">&#34;normal&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;tumor&#34;</span>))
</span></span><span style="display:flex;"><span>table(Group)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Group</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># normal  tumor </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 25     25 </span>
</span></span><span style="display:flex;"><span>design=model.matrix(~  Group) 
</span></span><span style="display:flex;"><span>head(design)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   (Intercept) Grouptumor</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1           1          0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2           1          1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3           1          0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4           1          1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5           1          0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6           1          1</span>
</span></span><span style="display:flex;"><span>fit = lmFit(exp_dat, design)
</span></span><span style="display:flex;"><span>fit = eBayes(fit)
</span></span><span style="display:flex;"><span>colnames(coef(fit))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;(Intercept)&#34; &#34;Grouptumor&#34;</span>
</span></span><span style="display:flex;"><span>tT = topTable(fit, adjust=<span style="color:#0ff;font-weight:bold">&#34;fdr&#34;</span>, sort.by=<span style="color:#0ff;font-weight:bold">&#34;p&#34;</span>, number=<span style="color:#fff;font-weight:bold">Inf</span>)
</span></span><span style="display:flex;"><span>head(tT)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#           logFC AveExpr         t      P.Value    adj.P.Val        B</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4004044 -2.6860  6.8706 -21.16464 6.424093e-27 1.225460e-22 50.47849</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4000538 -4.1400  6.6420 -19.36848 3.622286e-25 3.454937e-21 46.66686</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2591421 -2.2128  7.1468 -16.18728 9.688784e-22 6.160775e-18 39.10226</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3830189 -1.8416  8.0504 -16.03976 1.433530e-21 6.836502e-18 38.72368</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3519309 -2.1232  7.7964 -15.30913 1.034876e-20 3.948259e-17 36.80964</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3745525 -1.0784  6.1008 -15.21411 1.344319e-20 4.274038e-17 36.55587</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="3配对分析">（3）配对分析<a hidden class="anchor" aria-hidden="true" href="#3配对分析">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>Group = factor(meta$Group, levels = c(<span style="color:#0ff;font-weight:bold">&#34;normal&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;tumor&#34;</span>))
</span></span><span style="display:flex;"><span>table(Group)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Group</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># normal  tumor </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 25     25 </span>
</span></span><span style="display:flex;"><span>Patient = factor(meta$Patient)
</span></span><span style="display:flex;"><span>table(Patient)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Patient</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2102 2110 2146 2181 2190 2280 2297 2353 2365 2397 2451 2454 2706 2780 2910 2917 2943 2972 2985 3156 3159 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2    2 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3197 3232 3297 3355 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2    2    2    2</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>design=model.matrix(~  Patient + Group) 
</span></span><span style="display:flex;"><span>head(design)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   Groupnormal Grouptumor</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1           1          0</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2           0          1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3           1          0</span>
</span></span><span style="display:flex;"><span>fit = lmFit(exp_dat, design)
</span></span><span style="display:flex;"><span>fit &lt;- eBayes(fit)
</span></span><span style="display:flex;"><span>colnames(coef(fit))
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;(Intercept)&#34; &#34;Patient2110&#34; &#34;Patient2146&#34; &#34;Patient2181&#34; &#34;Patient2190&#34; &#34;Patient2280&#34; &#34;Patient2297&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [8] &#34;Patient2353&#34; &#34;Patient2365&#34; &#34;Patient2397&#34; &#34;Patient2451&#34; &#34;Patient2454&#34; &#34;Patient2706&#34; &#34;Patient2780&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [15] &#34;Patient2910&#34; &#34;Patient2917&#34; &#34;Patient2943&#34; &#34;Patient2972&#34; &#34;Patient2985&#34; &#34;Patient3156&#34; &#34;Patient3159&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [22] &#34;Patient3197&#34; &#34;Patient3232&#34; &#34;Patient3297&#34; &#34;Patient3355&#34; &#34;Grouptumor&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>tT &lt;- topTable(fit, coef = <span style="color:#0ff;font-weight:bold">&#34;Grouptumor&#34;</span>, adjust=<span style="color:#0ff;font-weight:bold">&#34;fdr&#34;</span>, sort.by=<span style="color:#0ff;font-weight:bold">&#34;p&#34;</span>, number=<span style="color:#fff;font-weight:bold">Inf</span>)
</span></span><span style="display:flex;"><span>head(tT)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#           logFC AveExpr         t      P.Value    adj.P.Val        B</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4004044 -2.6860  6.8706 -23.34059 1.764542e-19 3.366041e-15 33.97414</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4000538 -4.1400  6.6420 -21.21050 2.072006e-18 1.976279e-14 31.71755</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3830189 -1.8416  8.0504 -19.34286 2.168752e-17 1.379037e-13 29.52627</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2701109 -2.4704  7.0796 -17.61105 2.298621e-16 1.096212e-12 27.28943</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3062082 -2.9824  9.0236 -17.27304 3.728465e-16 1.422484e-12 26.82740</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2591421 -2.2128  7.1468 -17.03824 5.241945e-16 1.564468e-12 26.50125</span>
</span></span></code></pre></td></tr></table>
</div>
</div>

  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">转录组</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/039%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--cosg%E9%89%B4%E5%AE%9Amarker%E5%9F%BA%E5%9B%A0/">
    <span class="title">« Prev Page</span>
    <br>
    <span>单细胞分析工具--COSG鉴定marker基因</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/046tcgabiolinks%E5%8C%85-tcga%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD%E4%B8%8E%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/">
    <span class="title">Next Page »</span>
    <br>
    <span>TCGAbiolinks包-TCGA数据下载与差异分析</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
