<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>基因-蛋白-化合物ID转换 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="生信基础, 化合物">
<meta name="description" content="1、不同基因ID转换
1.1 org.Hs.eg.db包


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30


library(dplyr)
library(org.Hs.eg.db)
keytypes(org.Hs.eg.db) 
#  [1] &#34;ACCNUM&#34;       &#34;ALIAS&#34;        &#34;ENSEMBL&#34;      &#34;ENSEMBLPROT&#34;  &#34;ENSEMBLTRANS&#34; &#34;ENTREZID&#34;    
#  [7] &#34;ENZYME&#34;       &#34;EVIDENCE&#34;     &#34;EVIDENCEALL&#34;  &#34;GENENAME&#34;     &#34;GENETYPE&#34;     &#34;GO&#34;          
# [13] &#34;GOALL&#34;        &#34;IPI&#34;          &#34;MAP&#34;          &#34;OMIM&#34;         &#34;ONTOLOGY&#34;     &#34;ONTOLOGYALL&#34; 
# [19] &#34;PATH&#34;         &#34;PFAM&#34;         &#34;PMID&#34;         &#34;PROSITE&#34;      &#34;REFSEQ&#34;       &#34;SYMBOL&#34;      
# [25] &#34;UCSCKG&#34;       &#34;UNIPROT&#34;

gene_symbol=c(&#34;RHO&#34;,&#34;CALM1&#34;,&#34;MEG3&#34;,&#34;GNGT1&#34;,&#34;SAG&#34;,&#34;RPGRIP1&#34;,&#34;TRPM1&#34;,&#34;PCP2&#34;,&#34;PCP4&#34;,&#34;AP1B1&#34;)

gene_ids&lt;-AnnotationDbi::select(org.Hs.eg.db, keys=as.character(gene_symbol), 
                                columns=c(&#34;ENSEMBL&#34;,&#34;ENTREZID&#34;), #目标格式
                                keytype=&#34;SYMBOL&#34;) #目前的格式
gene_ids

##去重
gene_ids %&gt;% 
  dplyr::distinct(ENTREZID, .keep_all = T)
#     SYMBOL         ENSEMBL ENTREZID
# 1      RHO ENSG00000163914     6010
# 2    CALM1 ENSG00000198668      801
# 3     MEG3 ENSG00000214548    55384
# 4    GNGT1 ENSG00000127928     2792
# 5      SAG ENSG00000130561     6295
# 6  RPGRIP1 ENSG00000092200    57096
# 7    TRPM1 ENSG00000134160     4308
# 8     PCP2 ENSG00000174788   126006
# 9     PCP4 ENSG00000183036     5121
# 10   AP1B1 ENSG00000100280      162


1.2 biomaRt包


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12


library(&#34;biomaRt&#34;)
ensembl = useMart(&#34;ensembl&#34;,dataset=&#34;hsapiens_gene_ensembl&#34;) 
attributes = listAttributes(ensembl)
attributes[1:5,]
# library(httr)
# httr::set_config(config(ssl_verifypeer = 0L))

gene_symbol=c(&#34;RHO&#34;,&#34;CALM1&#34;,&#34;MEG3&#34;,&#34;GNGT1&#34;,&#34;SAG&#34;,&#34;RPGRIP1&#34;,&#34;TRPM1&#34;,&#34;PCP2&#34;,&#34;PCP4&#34;,&#34;AP1B1&#34;)
gene_ids2 &lt;- getBM(filters= &#34;hgnc_symbol&#34;, 
                    attributes= c(&#34;hgnc_symbol&#34;,&#34;ensembl_gene_id&#34;,&#34;entrezgene_id&#34;),
                    values = gene_symbol, mart= ensembl)
gene_ids2


2、鼠源基因转为人类基因ID


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42


musGenes &lt;- c(&#34;Hmmr&#34;, &#34;Tlx3&#34;, &#34;Cpeb4&#34;)

## 方式1：直接大小写转换
toupper(musGenes)
# [1] &#34;HMMR&#34;  &#34;TLX3&#34;  &#34;CPEB4&#34;

## 方式2：通过biomaRt包（不稳定）
require(&#34;biomaRt&#34;)
# library(httr)
# httr::set_config(config(ssl_verifypeer = 0L))
human = useMart(&#34;ensembl&#34;, dataset = &#34;hsapiens_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
mouse = useMart(&#34;ensembl&#34;, dataset = &#34;mmusculus_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
genes = getLDS(attributes = c(&#34;mgi_symbol&#34;), filters = &#34;mgi_symbol&#34;, 
               values = musGenes, 
               mart = mouse, 
               attributesL = c(&#34;hgnc_symbol&#34;), 
               martL = human, uniqueRows=T)

## 方式3：MGI 数据库
# https://support.bioconductor.org/p/129636/
library(dplyr)
mouse_human_genes = read.csv(&#34;http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt&#34;,sep=&#34;\t&#34;)

convert_mouse_to_human &lt;- function(gene_list){
  
  output = c()
  
  for(gene in gene_list){
    class_key = (mouse_human_genes %&gt;% filter(Symbol == gene &amp; Common.Organism.Name==&#34;mouse, laboratory&#34;))[[&#39;DB.Class.Key&#39;]]
    if(!identical(class_key, integer(0)) ){
      human_genes = (mouse_human_genes %&gt;% filter(DB.Class.Key == class_key &amp; Common.Organism.Name==&#34;human&#34;))[,&#34;Symbol&#34;]
      for(human_gene in human_genes){
        output = append(output,human_gene)
      }
    }
  }
  
  return (output)
}

convert_mouse_to_human(musGenes)
# 1] &#34;HMMR&#34;  &#34;TLX3&#34;  &#34;CPEB4&#34;




 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23


# # https://github.com/lishensuo/utils
# 
# library(&#34;biomaRt&#34;)
# # library(httr)
# # httr::set_config(config(ssl_verifypeer = 0L))
# human = useMart(&#34;ensembl&#34;, dataset = &#34;hsapiens_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
# mouse = useMart(&#34;ensembl&#34;, dataset = &#34;mmusculus_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
# 
# # https://www.gencodegenes.org/mouse/
# dat = data.table::fread(&#34;gencode.vM33.basic.annotation.gtf.gz&#34;)
# dat = subset(dat, V3 == &#34;gene&#34;)
# dat_sub = dat[,&#34;V9&#34;] %&gt;% 
#   separate(V9, into = c(&#34;gene_id&#34;,&#34;gene_type&#34;,&#34;gene_name&#34;,&#34;mgi_id&#34;,&#34;havana_gene&#34;), sep = &#34;; &#34;)
# dat_sub$gene_name2 = gsub(&#39;gencode.vM33.basic.annotation.gtf.gz &#34;&#39;,&#39;&#39;,dat_sub$gene_name)
# dat_sub$gene_name2 = gsub(&#39;&#34;&#39;,&#39;&#39;,dat_sub$gene_name2)
# 
# genes = getLDS(attributes = c(&#34;mgi_symbol&#34;), filters = &#34;mgi_symbol&#34;, 
#                values = dat_sub$gene_name2, 
#                mart = mouse, 
#                attributesL = c(&#34;hgnc_symbol&#34;), 
#                martL = human, uniqueRows=T)
# write.csv(genes, file = &#34;mgi2hgnc_biomart.csv&#34;, row.names = F, quote = F)
# head(genes)


3、蛋白质与基因ID转换

https://www.uniprot.org/uploadlists/


">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/054%E5%9F%BA%E5%9B%A0-%E8%9B%8B%E7%99%BD-%E5%8C%96%E5%90%88%E7%89%A9id%E8%BD%AC%E6%8D%A2/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/054%E5%9F%BA%E5%9B%A0-%E8%9B%8B%E7%99%BD-%E5%8C%96%E5%90%88%E7%89%A9id%E8%BD%AC%E6%8D%A2/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="基因-蛋白-化合物ID转换" />
<meta property="og:description" content="1、不同基因ID转换
1.1 org.Hs.eg.db包


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30


library(dplyr)
library(org.Hs.eg.db)
keytypes(org.Hs.eg.db) 
#  [1] &#34;ACCNUM&#34;       &#34;ALIAS&#34;        &#34;ENSEMBL&#34;      &#34;ENSEMBLPROT&#34;  &#34;ENSEMBLTRANS&#34; &#34;ENTREZID&#34;    
#  [7] &#34;ENZYME&#34;       &#34;EVIDENCE&#34;     &#34;EVIDENCEALL&#34;  &#34;GENENAME&#34;     &#34;GENETYPE&#34;     &#34;GO&#34;          
# [13] &#34;GOALL&#34;        &#34;IPI&#34;          &#34;MAP&#34;          &#34;OMIM&#34;         &#34;ONTOLOGY&#34;     &#34;ONTOLOGYALL&#34; 
# [19] &#34;PATH&#34;         &#34;PFAM&#34;         &#34;PMID&#34;         &#34;PROSITE&#34;      &#34;REFSEQ&#34;       &#34;SYMBOL&#34;      
# [25] &#34;UCSCKG&#34;       &#34;UNIPROT&#34;

gene_symbol=c(&#34;RHO&#34;,&#34;CALM1&#34;,&#34;MEG3&#34;,&#34;GNGT1&#34;,&#34;SAG&#34;,&#34;RPGRIP1&#34;,&#34;TRPM1&#34;,&#34;PCP2&#34;,&#34;PCP4&#34;,&#34;AP1B1&#34;)

gene_ids&lt;-AnnotationDbi::select(org.Hs.eg.db, keys=as.character(gene_symbol), 
                                columns=c(&#34;ENSEMBL&#34;,&#34;ENTREZID&#34;), #目标格式
                                keytype=&#34;SYMBOL&#34;) #目前的格式
gene_ids

##去重
gene_ids %&gt;% 
  dplyr::distinct(ENTREZID, .keep_all = T)
#     SYMBOL         ENSEMBL ENTREZID
# 1      RHO ENSG00000163914     6010
# 2    CALM1 ENSG00000198668      801
# 3     MEG3 ENSG00000214548    55384
# 4    GNGT1 ENSG00000127928     2792
# 5      SAG ENSG00000130561     6295
# 6  RPGRIP1 ENSG00000092200    57096
# 7    TRPM1 ENSG00000134160     4308
# 8     PCP2 ENSG00000174788   126006
# 9     PCP4 ENSG00000183036     5121
# 10   AP1B1 ENSG00000100280      162


1.2 biomaRt包


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12


library(&#34;biomaRt&#34;)
ensembl = useMart(&#34;ensembl&#34;,dataset=&#34;hsapiens_gene_ensembl&#34;) 
attributes = listAttributes(ensembl)
attributes[1:5,]
# library(httr)
# httr::set_config(config(ssl_verifypeer = 0L))

gene_symbol=c(&#34;RHO&#34;,&#34;CALM1&#34;,&#34;MEG3&#34;,&#34;GNGT1&#34;,&#34;SAG&#34;,&#34;RPGRIP1&#34;,&#34;TRPM1&#34;,&#34;PCP2&#34;,&#34;PCP4&#34;,&#34;AP1B1&#34;)
gene_ids2 &lt;- getBM(filters= &#34;hgnc_symbol&#34;, 
                    attributes= c(&#34;hgnc_symbol&#34;,&#34;ensembl_gene_id&#34;,&#34;entrezgene_id&#34;),
                    values = gene_symbol, mart= ensembl)
gene_ids2


2、鼠源基因转为人类基因ID


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42


musGenes &lt;- c(&#34;Hmmr&#34;, &#34;Tlx3&#34;, &#34;Cpeb4&#34;)

## 方式1：直接大小写转换
toupper(musGenes)
# [1] &#34;HMMR&#34;  &#34;TLX3&#34;  &#34;CPEB4&#34;

## 方式2：通过biomaRt包（不稳定）
require(&#34;biomaRt&#34;)
# library(httr)
# httr::set_config(config(ssl_verifypeer = 0L))
human = useMart(&#34;ensembl&#34;, dataset = &#34;hsapiens_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
mouse = useMart(&#34;ensembl&#34;, dataset = &#34;mmusculus_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
genes = getLDS(attributes = c(&#34;mgi_symbol&#34;), filters = &#34;mgi_symbol&#34;, 
               values = musGenes, 
               mart = mouse, 
               attributesL = c(&#34;hgnc_symbol&#34;), 
               martL = human, uniqueRows=T)

## 方式3：MGI 数据库
# https://support.bioconductor.org/p/129636/
library(dplyr)
mouse_human_genes = read.csv(&#34;http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt&#34;,sep=&#34;\t&#34;)

convert_mouse_to_human &lt;- function(gene_list){
  
  output = c()
  
  for(gene in gene_list){
    class_key = (mouse_human_genes %&gt;% filter(Symbol == gene &amp; Common.Organism.Name==&#34;mouse, laboratory&#34;))[[&#39;DB.Class.Key&#39;]]
    if(!identical(class_key, integer(0)) ){
      human_genes = (mouse_human_genes %&gt;% filter(DB.Class.Key == class_key &amp; Common.Organism.Name==&#34;human&#34;))[,&#34;Symbol&#34;]
      for(human_gene in human_genes){
        output = append(output,human_gene)
      }
    }
  }
  
  return (output)
}

convert_mouse_to_human(musGenes)
# 1] &#34;HMMR&#34;  &#34;TLX3&#34;  &#34;CPEB4&#34;




 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23


# # https://github.com/lishensuo/utils
# 
# library(&#34;biomaRt&#34;)
# # library(httr)
# # httr::set_config(config(ssl_verifypeer = 0L))
# human = useMart(&#34;ensembl&#34;, dataset = &#34;hsapiens_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
# mouse = useMart(&#34;ensembl&#34;, dataset = &#34;mmusculus_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
# 
# # https://www.gencodegenes.org/mouse/
# dat = data.table::fread(&#34;gencode.vM33.basic.annotation.gtf.gz&#34;)
# dat = subset(dat, V3 == &#34;gene&#34;)
# dat_sub = dat[,&#34;V9&#34;] %&gt;% 
#   separate(V9, into = c(&#34;gene_id&#34;,&#34;gene_type&#34;,&#34;gene_name&#34;,&#34;mgi_id&#34;,&#34;havana_gene&#34;), sep = &#34;; &#34;)
# dat_sub$gene_name2 = gsub(&#39;gencode.vM33.basic.annotation.gtf.gz &#34;&#39;,&#39;&#39;,dat_sub$gene_name)
# dat_sub$gene_name2 = gsub(&#39;&#34;&#39;,&#39;&#39;,dat_sub$gene_name2)
# 
# genes = getLDS(attributes = c(&#34;mgi_symbol&#34;), filters = &#34;mgi_symbol&#34;, 
#                values = dat_sub$gene_name2, 
#                mart = mouse, 
#                attributesL = c(&#34;hgnc_symbol&#34;), 
#                martL = human, uniqueRows=T)
# write.csv(genes, file = &#34;mgi2hgnc_biomart.csv&#34;, row.names = F, quote = F)
# head(genes)


3、蛋白质与基因ID转换

https://www.uniprot.org/uploadlists/


" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/054%E5%9F%BA%E5%9B%A0-%E8%9B%8B%E7%99%BD-%E5%8C%96%E5%90%88%E7%89%A9id%E8%BD%AC%E6%8D%A2/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-05-28T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2025-06-25T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="基因-蛋白-化合物ID转换"/>
<meta name="twitter:description" content="1、不同基因ID转换
1.1 org.Hs.eg.db包


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30


library(dplyr)
library(org.Hs.eg.db)
keytypes(org.Hs.eg.db) 
#  [1] &#34;ACCNUM&#34;       &#34;ALIAS&#34;        &#34;ENSEMBL&#34;      &#34;ENSEMBLPROT&#34;  &#34;ENSEMBLTRANS&#34; &#34;ENTREZID&#34;    
#  [7] &#34;ENZYME&#34;       &#34;EVIDENCE&#34;     &#34;EVIDENCEALL&#34;  &#34;GENENAME&#34;     &#34;GENETYPE&#34;     &#34;GO&#34;          
# [13] &#34;GOALL&#34;        &#34;IPI&#34;          &#34;MAP&#34;          &#34;OMIM&#34;         &#34;ONTOLOGY&#34;     &#34;ONTOLOGYALL&#34; 
# [19] &#34;PATH&#34;         &#34;PFAM&#34;         &#34;PMID&#34;         &#34;PROSITE&#34;      &#34;REFSEQ&#34;       &#34;SYMBOL&#34;      
# [25] &#34;UCSCKG&#34;       &#34;UNIPROT&#34;

gene_symbol=c(&#34;RHO&#34;,&#34;CALM1&#34;,&#34;MEG3&#34;,&#34;GNGT1&#34;,&#34;SAG&#34;,&#34;RPGRIP1&#34;,&#34;TRPM1&#34;,&#34;PCP2&#34;,&#34;PCP4&#34;,&#34;AP1B1&#34;)

gene_ids&lt;-AnnotationDbi::select(org.Hs.eg.db, keys=as.character(gene_symbol), 
                                columns=c(&#34;ENSEMBL&#34;,&#34;ENTREZID&#34;), #目标格式
                                keytype=&#34;SYMBOL&#34;) #目前的格式
gene_ids

##去重
gene_ids %&gt;% 
  dplyr::distinct(ENTREZID, .keep_all = T)
#     SYMBOL         ENSEMBL ENTREZID
# 1      RHO ENSG00000163914     6010
# 2    CALM1 ENSG00000198668      801
# 3     MEG3 ENSG00000214548    55384
# 4    GNGT1 ENSG00000127928     2792
# 5      SAG ENSG00000130561     6295
# 6  RPGRIP1 ENSG00000092200    57096
# 7    TRPM1 ENSG00000134160     4308
# 8     PCP2 ENSG00000174788   126006
# 9     PCP4 ENSG00000183036     5121
# 10   AP1B1 ENSG00000100280      162


1.2 biomaRt包


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12


library(&#34;biomaRt&#34;)
ensembl = useMart(&#34;ensembl&#34;,dataset=&#34;hsapiens_gene_ensembl&#34;) 
attributes = listAttributes(ensembl)
attributes[1:5,]
# library(httr)
# httr::set_config(config(ssl_verifypeer = 0L))

gene_symbol=c(&#34;RHO&#34;,&#34;CALM1&#34;,&#34;MEG3&#34;,&#34;GNGT1&#34;,&#34;SAG&#34;,&#34;RPGRIP1&#34;,&#34;TRPM1&#34;,&#34;PCP2&#34;,&#34;PCP4&#34;,&#34;AP1B1&#34;)
gene_ids2 &lt;- getBM(filters= &#34;hgnc_symbol&#34;, 
                    attributes= c(&#34;hgnc_symbol&#34;,&#34;ensembl_gene_id&#34;,&#34;entrezgene_id&#34;),
                    values = gene_symbol, mart= ensembl)
gene_ids2


2、鼠源基因转为人类基因ID


 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42


musGenes &lt;- c(&#34;Hmmr&#34;, &#34;Tlx3&#34;, &#34;Cpeb4&#34;)

## 方式1：直接大小写转换
toupper(musGenes)
# [1] &#34;HMMR&#34;  &#34;TLX3&#34;  &#34;CPEB4&#34;

## 方式2：通过biomaRt包（不稳定）
require(&#34;biomaRt&#34;)
# library(httr)
# httr::set_config(config(ssl_verifypeer = 0L))
human = useMart(&#34;ensembl&#34;, dataset = &#34;hsapiens_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
mouse = useMart(&#34;ensembl&#34;, dataset = &#34;mmusculus_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
genes = getLDS(attributes = c(&#34;mgi_symbol&#34;), filters = &#34;mgi_symbol&#34;, 
               values = musGenes, 
               mart = mouse, 
               attributesL = c(&#34;hgnc_symbol&#34;), 
               martL = human, uniqueRows=T)

## 方式3：MGI 数据库
# https://support.bioconductor.org/p/129636/
library(dplyr)
mouse_human_genes = read.csv(&#34;http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt&#34;,sep=&#34;\t&#34;)

convert_mouse_to_human &lt;- function(gene_list){
  
  output = c()
  
  for(gene in gene_list){
    class_key = (mouse_human_genes %&gt;% filter(Symbol == gene &amp; Common.Organism.Name==&#34;mouse, laboratory&#34;))[[&#39;DB.Class.Key&#39;]]
    if(!identical(class_key, integer(0)) ){
      human_genes = (mouse_human_genes %&gt;% filter(DB.Class.Key == class_key &amp; Common.Organism.Name==&#34;human&#34;))[,&#34;Symbol&#34;]
      for(human_gene in human_genes){
        output = append(output,human_gene)
      }
    }
  }
  
  return (output)
}

convert_mouse_to_human(musGenes)
# 1] &#34;HMMR&#34;  &#34;TLX3&#34;  &#34;CPEB4&#34;




 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23


# # https://github.com/lishensuo/utils
# 
# library(&#34;biomaRt&#34;)
# # library(httr)
# # httr::set_config(config(ssl_verifypeer = 0L))
# human = useMart(&#34;ensembl&#34;, dataset = &#34;hsapiens_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
# mouse = useMart(&#34;ensembl&#34;, dataset = &#34;mmusculus_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)
# 
# # https://www.gencodegenes.org/mouse/
# dat = data.table::fread(&#34;gencode.vM33.basic.annotation.gtf.gz&#34;)
# dat = subset(dat, V3 == &#34;gene&#34;)
# dat_sub = dat[,&#34;V9&#34;] %&gt;% 
#   separate(V9, into = c(&#34;gene_id&#34;,&#34;gene_type&#34;,&#34;gene_name&#34;,&#34;mgi_id&#34;,&#34;havana_gene&#34;), sep = &#34;; &#34;)
# dat_sub$gene_name2 = gsub(&#39;gencode.vM33.basic.annotation.gtf.gz &#34;&#39;,&#39;&#39;,dat_sub$gene_name)
# dat_sub$gene_name2 = gsub(&#39;&#34;&#39;,&#39;&#39;,dat_sub$gene_name2)
# 
# genes = getLDS(attributes = c(&#34;mgi_symbol&#34;), filters = &#34;mgi_symbol&#34;, 
#                values = dat_sub$gene_name2, 
#                mart = mouse, 
#                attributesL = c(&#34;hgnc_symbol&#34;), 
#                martL = human, uniqueRows=T)
# write.csv(genes, file = &#34;mgi2hgnc_biomart.csv&#34;, row.names = F, quote = F)
# head(genes)


3、蛋白质与基因ID转换

https://www.uniprot.org/uploadlists/


"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "基因-蛋白-化合物ID转换",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/054%E5%9F%BA%E5%9B%A0-%E8%9B%8B%E7%99%BD-%E5%8C%96%E5%90%88%E7%89%A9id%E8%BD%AC%E6%8D%A2/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "基因-蛋白-化合物ID转换",
  "name": "基因-蛋白-化合物ID转换",
  "description": "1、不同基因ID转换 1.1 org.Hs.eg.db包 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 library(dplyr) library(org.Hs.eg.db) keytypes(org.Hs.eg.db) # [1] \u0026#34;ACCNUM\u0026#34; \u0026#34;ALIAS\u0026#34; \u0026#34;ENSEMBL\u0026#34; \u0026#34;ENSEMBLPROT\u0026#34; \u0026#34;ENSEMBLTRANS\u0026#34; \u0026#34;ENTREZID\u0026#34; # [7] \u0026#34;ENZYME\u0026#34; \u0026#34;EVIDENCE\u0026#34; \u0026#34;EVIDENCEALL\u0026#34; \u0026#34;GENENAME\u0026#34; \u0026#34;GENETYPE\u0026#34; \u0026#34;GO\u0026#34; # [13] \u0026#34;GOALL\u0026#34; \u0026#34;IPI\u0026#34; \u0026#34;MAP\u0026#34; \u0026#34;OMIM\u0026#34; \u0026#34;ONTOLOGY\u0026#34; \u0026#34;ONTOLOGYALL\u0026#34; # [19] \u0026#34;PATH\u0026#34; \u0026#34;PFAM\u0026#34; \u0026#34;PMID\u0026#34; \u0026#34;PROSITE\u0026#34; \u0026#34;REFSEQ\u0026#34; \u0026#34;SYMBOL\u0026#34; # [25] \u0026#34;UCSCKG\u0026#34; \u0026#34;UNIPROT\u0026#34; gene_symbol=c(\u0026#34;RHO\u0026#34;,\u0026#34;CALM1\u0026#34;,\u0026#34;MEG3\u0026#34;,\u0026#34;GNGT1\u0026#34;,\u0026#34;SAG\u0026#34;,\u0026#34;RPGRIP1\u0026#34;,\u0026#34;TRPM1\u0026#34;,\u0026#34;PCP2\u0026#34;,\u0026#34;PCP4\u0026#34;,\u0026#34;AP1B1\u0026#34;) gene_ids\u0026lt;-AnnotationDbi::select(org.Hs.eg.db, keys=as.character(gene_symbol), columns=c(\u0026#34;ENSEMBL\u0026#34;,\u0026#34;ENTREZID\u0026#34;), #目标格式 keytype=\u0026#34;SYMBOL\u0026#34;) #目前的格式 gene_ids ##去重 gene_ids %\u0026gt;% dplyr::distinct(ENTREZID, .keep_all = T) # SYMBOL ENSEMBL ENTREZID # 1 RHO ENSG00000163914 6010 # 2 CALM1 ENSG00000198668 801 # 3 MEG3 ENSG00000214548 55384 # 4 GNGT1 ENSG00000127928 2792 # 5 SAG ENSG00000130561 6295 # 6 RPGRIP1 ENSG00000092200 57096 # 7 TRPM1 ENSG00000134160 4308 # 8 PCP2 ENSG00000174788 126006 # 9 PCP4 ENSG00000183036 5121 # 10 AP1B1 ENSG00000100280 162 1.2 biomaRt包 1 2 3 4 5 6 7 8 9 10 11 12 library(\u0026#34;biomaRt\u0026#34;) ensembl = useMart(\u0026#34;ensembl\u0026#34;,dataset=\u0026#34;hsapiens_gene_ensembl\u0026#34;) attributes = listAttributes(ensembl) attributes[1:5,] # library(httr) # httr::set_config(config(ssl_verifypeer = 0L)) gene_symbol=c(\u0026#34;RHO\u0026#34;,\u0026#34;CALM1\u0026#34;,\u0026#34;MEG3\u0026#34;,\u0026#34;GNGT1\u0026#34;,\u0026#34;SAG\u0026#34;,\u0026#34;RPGRIP1\u0026#34;,\u0026#34;TRPM1\u0026#34;,\u0026#34;PCP2\u0026#34;,\u0026#34;PCP4\u0026#34;,\u0026#34;AP1B1\u0026#34;) gene_ids2 \u0026lt;- getBM(filters= \u0026#34;hgnc_symbol\u0026#34;, attributes= c(\u0026#34;hgnc_symbol\u0026#34;,\u0026#34;ensembl_gene_id\u0026#34;,\u0026#34;entrezgene_id\u0026#34;), values = gene_symbol, mart= ensembl) gene_ids2 2、鼠源基因转为人类基因ID 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 musGenes \u0026lt;- c(\u0026#34;Hmmr\u0026#34;, \u0026#34;Tlx3\u0026#34;, \u0026#34;Cpeb4\u0026#34;) ## 方式1：直接大小写转换 toupper(musGenes) # [1] \u0026#34;HMMR\u0026#34; \u0026#34;TLX3\u0026#34; \u0026#34;CPEB4\u0026#34; ## 方式2：通过biomaRt包（不稳定） require(\u0026#34;biomaRt\u0026#34;) # library(httr) # httr::set_config(config(ssl_verifypeer = 0L)) human = useMart(\u0026#34;ensembl\u0026#34;, dataset = \u0026#34;hsapiens_gene_ensembl\u0026#34;,host = \u0026#34;dec2021.archive.ensembl.org\u0026#34;) mouse = useMart(\u0026#34;ensembl\u0026#34;, dataset = \u0026#34;mmusculus_gene_ensembl\u0026#34;,host = \u0026#34;dec2021.archive.ensembl.org\u0026#34;) genes = getLDS(attributes = c(\u0026#34;mgi_symbol\u0026#34;), filters = \u0026#34;mgi_symbol\u0026#34;, values = musGenes, mart = mouse, attributesL = c(\u0026#34;hgnc_symbol\u0026#34;), martL = human, uniqueRows=T) ## 方式3：MGI 数据库 # https://support.bioconductor.org/p/129636/ library(dplyr) mouse_human_genes = read.csv(\u0026#34;http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt\u0026#34;,sep=\u0026#34;\\t\u0026#34;) convert_mouse_to_human \u0026lt;- function(gene_list){ output = c() for(gene in gene_list){ class_key = (mouse_human_genes %\u0026gt;% filter(Symbol == gene \u0026amp; Common.Organism.Name==\u0026#34;mouse, laboratory\u0026#34;))[[\u0026#39;DB.Class.Key\u0026#39;]] if(!identical(class_key, integer(0)) ){ human_genes = (mouse_human_genes %\u0026gt;% filter(DB.Class.Key == class_key \u0026amp; Common.Organism.Name==\u0026#34;human\u0026#34;))[,\u0026#34;Symbol\u0026#34;] for(human_gene in human_genes){ output = append(output,human_gene) } } } return (output) } convert_mouse_to_human(musGenes) # 1] \u0026#34;HMMR\u0026#34; \u0026#34;TLX3\u0026#34; \u0026#34;CPEB4\u0026#34; 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 # # https://github.com/lishensuo/utils # # library(\u0026#34;biomaRt\u0026#34;) # # library(httr) # # httr::set_config(config(ssl_verifypeer = 0L)) # human = useMart(\u0026#34;ensembl\u0026#34;, dataset = \u0026#34;hsapiens_gene_ensembl\u0026#34;,host = \u0026#34;dec2021.archive.ensembl.org\u0026#34;) # mouse = useMart(\u0026#34;ensembl\u0026#34;, dataset = \u0026#34;mmusculus_gene_ensembl\u0026#34;,host = \u0026#34;dec2021.archive.ensembl.org\u0026#34;) # # # https://www.gencodegenes.org/mouse/ # dat = data.table::fread(\u0026#34;gencode.vM33.basic.annotation.gtf.gz\u0026#34;) # dat = subset(dat, V3 == \u0026#34;gene\u0026#34;) # dat_sub = dat[,\u0026#34;V9\u0026#34;] %\u0026gt;% # separate(V9, into = c(\u0026#34;gene_id\u0026#34;,\u0026#34;gene_type\u0026#34;,\u0026#34;gene_name\u0026#34;,\u0026#34;mgi_id\u0026#34;,\u0026#34;havana_gene\u0026#34;), sep = \u0026#34;; \u0026#34;) # dat_sub$gene_name2 = gsub(\u0026#39;gencode.vM33.basic.annotation.gtf.gz \u0026#34;\u0026#39;,\u0026#39;\u0026#39;,dat_sub$gene_name) # dat_sub$gene_name2 = gsub(\u0026#39;\u0026#34;\u0026#39;,\u0026#39;\u0026#39;,dat_sub$gene_name2) # # genes = getLDS(attributes = c(\u0026#34;mgi_symbol\u0026#34;), filters = \u0026#34;mgi_symbol\u0026#34;, # values = dat_sub$gene_name2, # mart = mouse, # attributesL = c(\u0026#34;hgnc_symbol\u0026#34;), # martL = human, uniqueRows=T) # write.csv(genes, file = \u0026#34;mgi2hgnc_biomart.csv\u0026#34;, row.names = F, quote = F) # head(genes) 3、蛋白质与基因ID转换 https://www.uniprot.org/uploadlists/ ",
  "keywords": [
    "生信基础", "化合物"
  ],
  "articleBody": "1、不同基因ID转换 1.1 org.Hs.eg.db包 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 library(dplyr) library(org.Hs.eg.db) keytypes(org.Hs.eg.db) # [1] \"ACCNUM\" \"ALIAS\" \"ENSEMBL\" \"ENSEMBLPROT\" \"ENSEMBLTRANS\" \"ENTREZID\" # [7] \"ENZYME\" \"EVIDENCE\" \"EVIDENCEALL\" \"GENENAME\" \"GENETYPE\" \"GO\" # [13] \"GOALL\" \"IPI\" \"MAP\" \"OMIM\" \"ONTOLOGY\" \"ONTOLOGYALL\" # [19] \"PATH\" \"PFAM\" \"PMID\" \"PROSITE\" \"REFSEQ\" \"SYMBOL\" # [25] \"UCSCKG\" \"UNIPROT\" gene_symbol=c(\"RHO\",\"CALM1\",\"MEG3\",\"GNGT1\",\"SAG\",\"RPGRIP1\",\"TRPM1\",\"PCP2\",\"PCP4\",\"AP1B1\") gene_ids\u003c-AnnotationDbi::select(org.Hs.eg.db, keys=as.character(gene_symbol), columns=c(\"ENSEMBL\",\"ENTREZID\"), #目标格式 keytype=\"SYMBOL\") #目前的格式 gene_ids ##去重 gene_ids %\u003e% dplyr::distinct(ENTREZID, .keep_all = T) # SYMBOL ENSEMBL ENTREZID # 1 RHO ENSG00000163914 6010 # 2 CALM1 ENSG00000198668 801 # 3 MEG3 ENSG00000214548 55384 # 4 GNGT1 ENSG00000127928 2792 # 5 SAG ENSG00000130561 6295 # 6 RPGRIP1 ENSG00000092200 57096 # 7 TRPM1 ENSG00000134160 4308 # 8 PCP2 ENSG00000174788 126006 # 9 PCP4 ENSG00000183036 5121 # 10 AP1B1 ENSG00000100280 162 1.2 biomaRt包 1 2 3 4 5 6 7 8 9 10 11 12 library(\"biomaRt\") ensembl = useMart(\"ensembl\",dataset=\"hsapiens_gene_ensembl\") attributes = listAttributes(ensembl) attributes[1:5,] # library(httr) # httr::set_config(config(ssl_verifypeer = 0L)) gene_symbol=c(\"RHO\",\"CALM1\",\"MEG3\",\"GNGT1\",\"SAG\",\"RPGRIP1\",\"TRPM1\",\"PCP2\",\"PCP4\",\"AP1B1\") gene_ids2 \u003c- getBM(filters= \"hgnc_symbol\", attributes= c(\"hgnc_symbol\",\"ensembl_gene_id\",\"entrezgene_id\"), values = gene_symbol, mart= ensembl) gene_ids2 2、鼠源基因转为人类基因ID 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 musGenes \u003c- c(\"Hmmr\", \"Tlx3\", \"Cpeb4\") ## 方式1：直接大小写转换 toupper(musGenes) # [1] \"HMMR\" \"TLX3\" \"CPEB4\" ## 方式2：通过biomaRt包（不稳定） require(\"biomaRt\") # library(httr) # httr::set_config(config(ssl_verifypeer = 0L)) human = useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\",host = \"dec2021.archive.ensembl.org\") mouse = useMart(\"ensembl\", dataset = \"mmusculus_gene_ensembl\",host = \"dec2021.archive.ensembl.org\") genes = getLDS(attributes = c(\"mgi_symbol\"), filters = \"mgi_symbol\", values = musGenes, mart = mouse, attributesL = c(\"hgnc_symbol\"), martL = human, uniqueRows=T) ## 方式3：MGI 数据库 # https://support.bioconductor.org/p/129636/ library(dplyr) mouse_human_genes = read.csv(\"http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt\",sep=\"\\t\") convert_mouse_to_human \u003c- function(gene_list){ output = c() for(gene in gene_list){ class_key = (mouse_human_genes %\u003e% filter(Symbol == gene \u0026 Common.Organism.Name==\"mouse, laboratory\"))[['DB.Class.Key']] if(!identical(class_key, integer(0)) ){ human_genes = (mouse_human_genes %\u003e% filter(DB.Class.Key == class_key \u0026 Common.Organism.Name==\"human\"))[,\"Symbol\"] for(human_gene in human_genes){ output = append(output,human_gene) } } } return (output) } convert_mouse_to_human(musGenes) # 1] \"HMMR\" \"TLX3\" \"CPEB4\" 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 # # https://github.com/lishensuo/utils # # library(\"biomaRt\") # # library(httr) # # httr::set_config(config(ssl_verifypeer = 0L)) # human = useMart(\"ensembl\", dataset = \"hsapiens_gene_ensembl\",host = \"dec2021.archive.ensembl.org\") # mouse = useMart(\"ensembl\", dataset = \"mmusculus_gene_ensembl\",host = \"dec2021.archive.ensembl.org\") # # # https://www.gencodegenes.org/mouse/ # dat = data.table::fread(\"gencode.vM33.basic.annotation.gtf.gz\") # dat = subset(dat, V3 == \"gene\") # dat_sub = dat[,\"V9\"] %\u003e% # separate(V9, into = c(\"gene_id\",\"gene_type\",\"gene_name\",\"mgi_id\",\"havana_gene\"), sep = \"; \") # dat_sub$gene_name2 = gsub('gencode.vM33.basic.annotation.gtf.gz \"','',dat_sub$gene_name) # dat_sub$gene_name2 = gsub('\"','',dat_sub$gene_name2) # # genes = getLDS(attributes = c(\"mgi_symbol\"), filters = \"mgi_symbol\", # values = dat_sub$gene_name2, # mart = mouse, # attributesL = c(\"hgnc_symbol\"), # martL = human, uniqueRows=T) # write.csv(genes, file = \"mgi2hgnc_biomart.csv\", row.names = F, quote = F) # head(genes) 3、蛋白质与基因ID转换 https://www.uniprot.org/uploadlists/ 4、化合物ID转换 4.1 网页转换 https://pubchem.ncbi.nlm.nih.gov/idexchange/idexchange.cgi\n4.2 python工具转换 （1）https://pubchempy.readthedocs.io/en/v1.0.4/index.html\n1 pip install pubchempy （2）简单用法如下\n1 import pubchempy as pcp ①可直接根据化合物CID，构建出pubchempy.Compound对象 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 c = pcp.Compound.from_cid(5090) type(c) # pubchempy.Compound ##举例3个属性值 c.molecular_formula # 'C17H14O4S' c.canonical_smiles # 'CS(=O)(=O)C1=CC=C(C=C1)C2=C(C(=O)OC2)C3=CC=CC=C3' c.synonyms[:3] # ['rofecoxib', '162011-90-7', 'Vioxx'] ## 转换成表格，显示全部属性 c_df = pcp.compounds_to_frame(c) c_df.columns # Index(['atom_stereo_count', 'atoms', 'bond_stereo_count', 'bonds', # 'cactvs_fingerprint', 'canonical_smiles', 'charge', 'complexity', # 'conformer_id_3d', 'conformer_rmsd_3d', 'coordinate_type', # 'covalent_unit_count', 'defined_atom_stereo_count', # 'defined_bond_stereo_count', 'effective_rotor_count_3d', 'elements', # 'exact_mass', 'feature_selfoverlap_3d', 'fingerprint', # 'h_bond_acceptor_count', 'h_bond_donor_count', 'heavy_atom_count', # 'inchi', 'inchikey', 'isomeric_smiles', 'isotope_atom_count', # 'iupac_name', 'mmff94_energy_3d', 'mmff94_partial_charges_3d', # 'molecular_formula', 'molecular_weight', 'monoisotopic_mass', # 'multipoles_3d', 'pharmacophore_features_3d', 'record', # 'rotatable_bond_count', 'shape_fingerprint_3d', 'shape_selfoverlap_3d', # 'tpsa', 'undefined_atom_stereo_count', 'undefined_bond_stereo_count', # 'volume_3d', 'xlogp'], # dtype='object') ②根据小分子属性值，搜索符合条件的小分子对象\n仅支持6种属性：[“name”,“smiles”,“sdf”,“inchi”,“inchikey”,“formula”]\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 ## 第一个参数--属性值，第二个参数--属性类型 ## 返回结果是包含若干个小分子对象的list results = pcp.get_compounds('Glucose', 'name') results = pcp.get_compounds('C1=CC2=C(C3=C(C=CC=N3)C=C2)N=C1', 'smiles') multi_results = pcp.get_compounds('C6H12O6', 'formula') results = multi_results[:3] for cp in results: print(cp.isomeric_smiles) # C1[C@H]([C@H]([C@@H](C(O1)(CO)O)O)O)O # C([C@H]([C@H]([C@@H]([C@H](C=O)O)O)O)O)O # C(C(C(C(C(C=O)O)O)O)O)O cs = pcp.get_compounds('C20H41Br', 'formula') cs[:3] # [Compound(20271), Compound(23148745), Compound(10808570)] cs_df = pcp.compounds_to_frame(cs, properties=['isomeric_smiles', 'xlogp', 'rotatable_bond_count']) cs_df.shape # (43, 3) 补充：MyGene MyGene.info 是一个为生物信息学提供高效、统一、RESTful API 的基因注释服务平台，支持快速查询基因的ID转换、结构、功能、通路、蛋白、疾病、药物靶点等多种注释信息；数据来源整合了多个主流数据库（如 NCBI, Ensembl, UniProt, GO, KEGG, ChEMBL 等）。\nhttps://docs.mygene.info/en/latest/index.html https://github.com/biothings/mygene.info https://pypi.org/project/mygene/ 使用方式也非常简单，可以直接通过API接口，也提供了封装好的Python、R包进行调用。下面简单记录其Python包的用法，方便以后使用。\n1 2 3 4 5 # pip install mygene import mygene mg = mygene.MyGeneInfo() 方式1：根据指定的entrezgene/ensembl id查询\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 res = mg.getgene(1017, fields='all') res.keys() # dict_keys(['AllianceGenome', 'HGNC', 'MIM', '_id', '_version', 'accession', 'agr', # 'alias', 'chembl', 'ec', 'ensembl', 'entrezgene', 'exac', 'exons', 'exons_hg19', # 'generif', 'genomic_pos', 'genomic_pos_hg19', 'go', 'homologene', 'interpro', 'ipi', # 'map_location', 'name', 'other_names', 'pantherdb', 'pathway', 'pdb', 'pfam', 'pharmgkb', # 'pharos', 'pir', 'prosite', 'reagent', 'refseq', 'reporter', 'summary', 'symbol', 'taxid', # 'type_of_gene', 'umls', 'unigene', 'unii', 'uniprot', 'wikipedia']) ## 笔记最后列出了每个key/field的含义说明 # 使用fields参数限定返回的注释条目 mg.getgene(1017, fields='name,symbol,refseq') mg.getgene(1017, fields=['name', 'symbol', 'refseq.rna']) # 同时查询多个基因 mg.getgenes([1017,1018,'ENSG00000148795'], fields='name,symbol,entrezgene,taxid,type_of_gene') mg.getgenes([1017,1018,'ENSG00000148795'], fields='name,symbol,entrezgene,taxid,ensembl', as_dataframe=True) 方式2：根据关键字匹配搜索\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 mg.query('cdk2') # 默认返回最匹配的10个结果，且返回条目进包括entrezgene/name/symbol/taxid # 针对特定条目进行搜索，且筛选物种 mg.query('symbol:cdk2', species='human', fields='name,symbol,entrezgene,taxid,ensembl') # 批量搜索，scopes参数设置针对某一条目进行搜索 mg.querymany([1017, 695], scopes='entrezgene', species='human') mg.querymany([\"TP53\", \"EGFR\", 'NA_TEST'], scopes='symbol', species='human', as_dataframe=True) # 当fetch_all=True，生成一个可迭代对象 kinases = mg.query('name:kinase', species='human', fetch_all=True) for gene in kinases: print(gene['_id'], gene['symbol']) 基本信息 字段名 含义 _id / entrezgene 主键，通常等于 Entrez Gene ID，如 1017 _version 数据版本号 symbol 基因符号，例如 CDK2 name 基因全名，例如 “cyclin-dependent kinase 2” taxid NCBI 物种 ID，9606 表示人类 type_of_gene 基因类型，如 protein-coding、ncRNA、pseudo 等 summary 基因功能简要描述 alias 其他名称或别名（例如缩写、旧称） other_names 附加名称（全称、别称等） 序列信息 字段名 含义 accession 基因相关的 GenBank/RefSeq accession 列表 refseq RefSeq 序列 ID（mRNA、protein、genomic 等） genomic_pos 基因在最新人类基因组版本（如 hg38）上的位置 genomic_pos_hg19 在旧版本（hg19）上的位置信息 exons 外显子位置（hg38） exons_hg19 外显子位置（hg19） 生物通路与功能注释\n字段名 含义 go Gene Ontology 注释，包括 BP（生物过程）、MF（分子功能）、CC（细胞成分） pathway 涉及的信号通路信息（来自 Reactome、WikiPathways、KEGG 等） pantherdb Panther 通路数据库注释 pharmgkb 药物-基因相互作用信息 pharos 药物开发相关注释（Drug Target 分类） 蛋白结构和功能域\n字段名 含义 uniprot UniProt 蛋白数据库的 ID 和注释 interpro InterPro 蛋白家族或功能域 pfam Pfam 蛋白结构域 prosite Prosite 模式/结构域数据库 pdb 相关蛋白质的三维结构（来自 Protein Data Bank） ec 酶编号（Enzyme Commission Number），如 kinase 常有 ipi International Protein Index（已弃用） pir PIR 蛋白数据库 ID 同源与进化 字段名 含义 homologene NCBI HomoloGene 同源基因信息 ([tax_id, entrezgene_id]) ensembl Ensembl 基因数据库的 ID 和注释 MIM OMIM（人类遗传疾病数据库）相关信息 unigene NCBI UniGene 聚类（旧） umls Unified Medical Language System 编号（跨数据库映射） exac ExAC 等人群变异数据汇总（Exome Aggregation Consortium） HGNC HGNC（人类基因命名委员会）ID 和详细命名信息 实验与试剂相关\n字段名 含义 reporter 基因在不同表达芯片平台上的探针 ID（如 Affymetrix） reagent BioGPS 等平台上的实验试剂资源 agr Alliance of Genome Resources ID（联合资源数据库） AllianceGenome Alliance Genome 注释（也可能重复 agr） 药物相关\n字段名 含义 chembl ChEMBL 中的药物靶点信息 unii FDA UNII（药物成分唯一标识符） 网络资源 字段名 含义 wikipedia Wikipedia 页面链接（如果有） map_location 染色体图谱位置，如 12q13.2（12号染色体长臂） generif GeneRIFs：简短文献注释，来自 PubMed 摘要（Gene Reference Into Function） ",
  "wordCount" : "2488",
  "inLanguage": "en",
  "datePublished": "2022-05-28T00:00:00Z",
  "dateModified": "2025-06-25T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/054%E5%9F%BA%E5%9B%A0-%E8%9B%8B%E7%99%BD-%E5%8C%96%E5%90%88%E7%89%A9id%E8%BD%AC%E6%8D%A2/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      基因-蛋白-化合物ID转换
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2022-05-28 00:00:00 &#43;0000 UTC&#39;&gt;2022-05-28&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2025-06-25&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2488&amp;nbsp;|&amp;nbsp;5 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e4%b8%8d%e5%90%8c%e5%9f%ba%e5%9b%a0id%e8%bd%ac%e6%8d%a2" aria-label="1、不同基因ID转换">1、不同基因ID转换</a><ul>
                            
                    <li>
                        <a href="#11-orghsegdb%e5%8c%85" aria-label="1.1 org.Hs.eg.db包">1.1 org.Hs.eg.db包</a></li>
                    <li>
                        <a href="#12-biomart%e5%8c%85" aria-label="1.2 biomaRt包">1.2 biomaRt包</a></li></ul>
                    </li>
                    <li>
                        <a href="#2%e9%bc%a0%e6%ba%90%e5%9f%ba%e5%9b%a0%e8%bd%ac%e4%b8%ba%e4%ba%ba%e7%b1%bb%e5%9f%ba%e5%9b%a0id" aria-label="2、鼠源基因转为人类基因ID">2、鼠源基因转为人类基因ID</a></li>
                    <li>
                        <a href="#3%e8%9b%8b%e7%99%bd%e8%b4%a8%e4%b8%8e%e5%9f%ba%e5%9b%a0id%e8%bd%ac%e6%8d%a2" aria-label="3、蛋白质与基因ID转换">3、蛋白质与基因ID转换</a></li>
                    <li>
                        <a href="#4%e5%8c%96%e5%90%88%e7%89%a9id%e8%bd%ac%e6%8d%a2" aria-label="4、化合物ID转换">4、化合物ID转换</a><ul>
                            
                    <li>
                        <a href="#41-%e7%bd%91%e9%a1%b5%e8%bd%ac%e6%8d%a2" aria-label="4.1 网页转换">4.1 网页转换</a></li>
                    <li>
                        <a href="#42-python%e5%b7%a5%e5%85%b7%e8%bd%ac%e6%8d%a2" aria-label="4.2 python工具转换">4.2 python工具转换</a></li></ul>
                    </li>
                    <li>
                        <a href="#%e8%a1%a5%e5%85%85mygene" aria-label="补充：MyGene">补充：MyGene</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><h1 id="1不同基因id转换">1、不同基因ID转换<a hidden class="anchor" aria-hidden="true" href="#1不同基因id转换">#</a></h1>
<h2 id="11-orghsegdb包">1.1 org.Hs.eg.db包<a hidden class="anchor" aria-hidden="true" href="#11-orghsegdb包">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(dplyr)
</span></span><span style="display:flex;"><span>library(org.Hs.eg.db)
</span></span><span style="display:flex;"><span>keytypes(org.Hs.eg.db) 
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  [1] &#34;ACCNUM&#34;       &#34;ALIAS&#34;        &#34;ENSEMBL&#34;      &#34;ENSEMBLPROT&#34;  &#34;ENSEMBLTRANS&#34; &#34;ENTREZID&#34;    </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#  [7] &#34;ENZYME&#34;       &#34;EVIDENCE&#34;     &#34;EVIDENCEALL&#34;  &#34;GENENAME&#34;     &#34;GENETYPE&#34;     &#34;GO&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [13] &#34;GOALL&#34;        &#34;IPI&#34;          &#34;MAP&#34;          &#34;OMIM&#34;         &#34;ONTOLOGY&#34;     &#34;ONTOLOGYALL&#34; </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [19] &#34;PATH&#34;         &#34;PFAM&#34;         &#34;PMID&#34;         &#34;PROSITE&#34;      &#34;REFSEQ&#34;       &#34;SYMBOL&#34;      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [25] &#34;UCSCKG&#34;       &#34;UNIPROT&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gene_symbol=c(<span style="color:#0ff;font-weight:bold">&#34;RHO&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;CALM1&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;MEG3&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;GNGT1&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;SAG&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;RPGRIP1&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TRPM1&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;PCP2&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;PCP4&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;AP1B1&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gene_ids&lt;-AnnotationDbi::select(org.Hs.eg.db, keys=as.character(gene_symbol), 
</span></span><span style="display:flex;"><span>                                columns=c(<span style="color:#0ff;font-weight:bold">&#34;ENSEMBL&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;ENTREZID&#34;</span>), <span style="color:#007f7f">#目标格式</span>
</span></span><span style="display:flex;"><span>                                keytype=<span style="color:#0ff;font-weight:bold">&#34;SYMBOL&#34;</span>) <span style="color:#007f7f">#目前的格式</span>
</span></span><span style="display:flex;"><span>gene_ids
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##去重</span>
</span></span><span style="display:flex;"><span>gene_ids %&gt;% 
</span></span><span style="display:flex;"><span>  dplyr::distinct(ENTREZID, .keep_all = <span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     SYMBOL         ENSEMBL ENTREZID</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1      RHO ENSG00000163914     6010</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2    CALM1 ENSG00000198668      801</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3     MEG3 ENSG00000214548    55384</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4    GNGT1 ENSG00000127928     2792</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5      SAG ENSG00000130561     6295</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6  RPGRIP1 ENSG00000092200    57096</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 7    TRPM1 ENSG00000134160     4308</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 8     PCP2 ENSG00000174788   126006</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 9     PCP4 ENSG00000183036     5121</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 10   AP1B1 ENSG00000100280      162</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="12-biomart包">1.2 biomaRt包<a hidden class="anchor" aria-hidden="true" href="#12-biomart包">#</a></h2>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(<span style="color:#0ff;font-weight:bold">&#34;biomaRt&#34;</span>)
</span></span><span style="display:flex;"><span>ensembl = useMart(<span style="color:#0ff;font-weight:bold">&#34;ensembl&#34;</span>,dataset=<span style="color:#0ff;font-weight:bold">&#34;hsapiens_gene_ensembl&#34;</span>) 
</span></span><span style="display:flex;"><span>attributes = listAttributes(ensembl)
</span></span><span style="display:flex;"><span>attributes[1:<span style="color:#ff0;font-weight:bold">5</span>,]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># library(httr)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># httr::set_config(config(ssl_verifypeer = 0L))</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gene_symbol=c(<span style="color:#0ff;font-weight:bold">&#34;RHO&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;CALM1&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;MEG3&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;GNGT1&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;SAG&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;RPGRIP1&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TRPM1&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;PCP2&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;PCP4&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;AP1B1&#34;</span>)
</span></span><span style="display:flex;"><span>gene_ids2 &lt;- getBM(filters= <span style="color:#0ff;font-weight:bold">&#34;hgnc_symbol&#34;</span>, 
</span></span><span style="display:flex;"><span>                    attributes= c(<span style="color:#0ff;font-weight:bold">&#34;hgnc_symbol&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;ensembl_gene_id&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;entrezgene_id&#34;</span>),
</span></span><span style="display:flex;"><span>                    values = gene_symbol, mart= ensembl)
</span></span><span style="display:flex;"><span>gene_ids2
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="2鼠源基因转为人类基因id">2、鼠源基因转为人类基因ID<a hidden class="anchor" aria-hidden="true" href="#2鼠源基因转为人类基因id">#</a></h1>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>musGenes &lt;- c(<span style="color:#0ff;font-weight:bold">&#34;Hmmr&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Tlx3&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Cpeb4&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 方式1：直接大小写转换</span>
</span></span><span style="display:flex;"><span>toupper(musGenes)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;HMMR&#34;  &#34;TLX3&#34;  &#34;CPEB4&#34;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 方式2：通过biomaRt包（不稳定）</span>
</span></span><span style="display:flex;"><span>require(<span style="color:#0ff;font-weight:bold">&#34;biomaRt&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># library(httr)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># httr::set_config(config(ssl_verifypeer = 0L))</span>
</span></span><span style="display:flex;"><span>human = useMart(<span style="color:#0ff;font-weight:bold">&#34;ensembl&#34;</span>, dataset = <span style="color:#0ff;font-weight:bold">&#34;hsapiens_gene_ensembl&#34;</span>,host = <span style="color:#0ff;font-weight:bold">&#34;dec2021.archive.ensembl.org&#34;</span>)
</span></span><span style="display:flex;"><span>mouse = useMart(<span style="color:#0ff;font-weight:bold">&#34;ensembl&#34;</span>, dataset = <span style="color:#0ff;font-weight:bold">&#34;mmusculus_gene_ensembl&#34;</span>,host = <span style="color:#0ff;font-weight:bold">&#34;dec2021.archive.ensembl.org&#34;</span>)
</span></span><span style="display:flex;"><span>genes = getLDS(attributes = c(<span style="color:#0ff;font-weight:bold">&#34;mgi_symbol&#34;</span>), filters = <span style="color:#0ff;font-weight:bold">&#34;mgi_symbol&#34;</span>, 
</span></span><span style="display:flex;"><span>               values = musGenes, 
</span></span><span style="display:flex;"><span>               mart = mouse, 
</span></span><span style="display:flex;"><span>               attributesL = c(<span style="color:#0ff;font-weight:bold">&#34;hgnc_symbol&#34;</span>), 
</span></span><span style="display:flex;"><span>               martL = human, uniqueRows=<span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 方式3：MGI 数据库</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># https://support.bioconductor.org/p/129636/</span>
</span></span><span style="display:flex;"><span>library(dplyr)
</span></span><span style="display:flex;"><span>mouse_human_genes = read.csv(<span style="color:#0ff;font-weight:bold">&#34;http://www.informatics.jax.org/downloads/reports/HOM_MouseHumanSequence.rpt&#34;</span>,sep=<span style="color:#0ff;font-weight:bold">&#34;\t&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>convert_mouse_to_human &lt;- <span style="color:#fff;font-weight:bold">function</span>(gene_list){
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  output = c()
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  <span style="color:#fff;font-weight:bold">for</span>(gene <span style="color:#fff;font-weight:bold">in</span> gene_list){
</span></span><span style="display:flex;"><span>    class_key = (mouse_human_genes %&gt;% filter(Symbol == gene &amp; Common.Organism.Name==<span style="color:#0ff;font-weight:bold">&#34;mouse, laboratory&#34;</span>))[[<span style="color:#0ff;font-weight:bold">&#39;DB.Class.Key&#39;</span>]]
</span></span><span style="display:flex;"><span>    <span style="color:#fff;font-weight:bold">if</span>(!identical(class_key, integer(<span style="color:#ff0;font-weight:bold">0</span>)) ){
</span></span><span style="display:flex;"><span>      human_genes = (mouse_human_genes %&gt;% filter(DB.Class.Key == class_key &amp; Common.Organism.Name==<span style="color:#0ff;font-weight:bold">&#34;human&#34;</span>))[,<span style="color:#0ff;font-weight:bold">&#34;Symbol&#34;</span>]
</span></span><span style="display:flex;"><span>      <span style="color:#fff;font-weight:bold">for</span>(human_gene <span style="color:#fff;font-weight:bold">in</span> human_genes){
</span></span><span style="display:flex;"><span>        output = append(output,human_gene)
</span></span><span style="display:flex;"><span>      }
</span></span><span style="display:flex;"><span>    }
</span></span><span style="display:flex;"><span>  }
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span>  <span style="color:#fff;font-weight:bold">return</span> (output)
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>convert_mouse_to_human(musGenes)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1] &#34;HMMR&#34;  &#34;TLX3&#34;  &#34;CPEB4&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f"># # https://github.com/lishensuo/utils</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># library(&#34;biomaRt&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># # library(httr)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># # httr::set_config(config(ssl_verifypeer = 0L))</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># human = useMart(&#34;ensembl&#34;, dataset = &#34;hsapiens_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># mouse = useMart(&#34;ensembl&#34;, dataset = &#34;mmusculus_gene_ensembl&#34;,host = &#34;dec2021.archive.ensembl.org&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># # https://www.gencodegenes.org/mouse/</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># dat = data.table::fread(&#34;gencode.vM33.basic.annotation.gtf.gz&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># dat = subset(dat, V3 == &#34;gene&#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># dat_sub = dat[,&#34;V9&#34;] %&gt;% </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   separate(V9, into = c(&#34;gene_id&#34;,&#34;gene_type&#34;,&#34;gene_name&#34;,&#34;mgi_id&#34;,&#34;havana_gene&#34;), sep = &#34;; &#34;)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># dat_sub$gene_name2 = gsub(&#39;gencode.vM33.basic.annotation.gtf.gz &#34;&#39;,&#39;&#39;,dat_sub$gene_name)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># dat_sub$gene_name2 = gsub(&#39;&#34;&#39;,&#39;&#39;,dat_sub$gene_name2)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># genes = getLDS(attributes = c(&#34;mgi_symbol&#34;), filters = &#34;mgi_symbol&#34;, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                values = dat_sub$gene_name2, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                mart = mouse, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                attributesL = c(&#34;hgnc_symbol&#34;), </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                martL = human, uniqueRows=T)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># write.csv(genes, file = &#34;mgi2hgnc_biomart.csv&#34;, row.names = F, quote = F)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># head(genes)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="3蛋白质与基因id转换">3、蛋白质与基因ID转换<a hidden class="anchor" aria-hidden="true" href="#3蛋白质与基因id转换">#</a></h1>
<ul>
<li><a href="https://www.uniprot.org/uploadlists/">https://www.uniprot.org/uploadlists/</a></li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220528214617188.png" alt="image-20220528214617188"  />
</p>
<h1 id="4化合物id转换">4、化合物ID转换<a hidden class="anchor" aria-hidden="true" href="#4化合物id转换">#</a></h1>
<h2 id="41-网页转换">4.1 网页转换<a hidden class="anchor" aria-hidden="true" href="#41-网页转换">#</a></h2>
<p><a href="https://pubchem.ncbi.nlm.nih.gov/idexchange/idexchange.cgi">https://pubchem.ncbi.nlm.nih.gov/idexchange/idexchange.cgi</a></p>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220528214848722.png" alt="image-20220528214848722"  />
</p>
<h2 id="42-python工具转换">4.2 python工具转换<a hidden class="anchor" aria-hidden="true" href="#42-python工具转换">#</a></h2>
<p>（1）https://pubchempy.readthedocs.io/en/v1.0.4/index.html</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-shell" data-lang="shell"><span style="display:flex;"><span>pip install pubchempy
</span></span></code></pre></td></tr></table>
</div>
</div><p>（2）简单用法如下</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> pubchempy <span style="color:#fff;font-weight:bold">as</span> pcp
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>①可直接根据化合物CID，构建出pubchempy.Compound对象</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>c = pcp.Compound.from_cid(<span style="color:#ff0;font-weight:bold">5090</span>)
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">type</span>(c)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># pubchempy.Compound</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##举例3个属性值</span>
</span></span><span style="display:flex;"><span>c.molecular_formula
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &#39;C17H14O4S&#39;</span>
</span></span><span style="display:flex;"><span>c.canonical_smiles
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># &#39;CS(=O)(=O)C1=CC=C(C=C1)C2=C(C(=O)OC2)C3=CC=CC=C3&#39;</span>
</span></span><span style="display:flex;"><span>c.synonyms[:<span style="color:#ff0;font-weight:bold">3</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;rofecoxib&#39;, &#39;162011-90-7&#39;, &#39;Vioxx&#39;]</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 转换成表格，显示全部属性</span>
</span></span><span style="display:flex;"><span>c_df = pcp.compounds_to_frame(c)
</span></span><span style="display:flex;"><span>c_df.columns
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Index([&#39;atom_stereo_count&#39;, &#39;atoms&#39;, &#39;bond_stereo_count&#39;, &#39;bonds&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;cactvs_fingerprint&#39;, &#39;canonical_smiles&#39;, &#39;charge&#39;, &#39;complexity&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;conformer_id_3d&#39;, &#39;conformer_rmsd_3d&#39;, &#39;coordinate_type&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;covalent_unit_count&#39;, &#39;defined_atom_stereo_count&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;defined_bond_stereo_count&#39;, &#39;effective_rotor_count_3d&#39;, &#39;elements&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;exact_mass&#39;, &#39;feature_selfoverlap_3d&#39;, &#39;fingerprint&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;h_bond_acceptor_count&#39;, &#39;h_bond_donor_count&#39;, &#39;heavy_atom_count&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;inchi&#39;, &#39;inchikey&#39;, &#39;isomeric_smiles&#39;, &#39;isotope_atom_count&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;iupac_name&#39;, &#39;mmff94_energy_3d&#39;, &#39;mmff94_partial_charges_3d&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;molecular_formula&#39;, &#39;molecular_weight&#39;, &#39;monoisotopic_mass&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;multipoles_3d&#39;, &#39;pharmacophore_features_3d&#39;, &#39;record&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;rotatable_bond_count&#39;, &#39;shape_fingerprint_3d&#39;, &#39;shape_selfoverlap_3d&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;tpsa&#39;, &#39;undefined_atom_stereo_count&#39;, &#39;undefined_bond_stereo_count&#39;,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        &#39;volume_3d&#39;, &#39;xlogp&#39;],</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#       dtype=&#39;object&#39;)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>
<p>②根据小分子属性值，搜索符合条件的小分子对象</p>
<p>仅支持6种属性：[&ldquo;name&rdquo;,&ldquo;smiles&rdquo;,&ldquo;sdf&rdquo;,&ldquo;inchi&rdquo;,&ldquo;inchikey&rdquo;,&ldquo;formula&rdquo;]</p>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f">## 第一个参数--属性值，第二个参数--属性类型</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 返回结果是包含若干个小分子对象的list</span>
</span></span><span style="display:flex;"><span>results = pcp.get_compounds(<span style="color:#0ff;font-weight:bold">&#39;Glucose&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;name&#39;</span>)
</span></span><span style="display:flex;"><span>results = pcp.get_compounds(<span style="color:#0ff;font-weight:bold">&#39;C1=CC2=C(C3=C(C=CC=N3)C=C2)N=C1&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;smiles&#39;</span>)
</span></span><span style="display:flex;"><span>multi_results = pcp.get_compounds(<span style="color:#0ff;font-weight:bold">&#39;C6H12O6&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;formula&#39;</span>)
</span></span><span style="display:flex;"><span>results = multi_results[:<span style="color:#ff0;font-weight:bold">3</span>]
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span> cp in results:
</span></span><span style="display:flex;"><span>    <span style="color:#fff;font-weight:bold">print</span>(cp.isomeric_smiles)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># C1[C@H]([C@H]([C@@H](C(O1)(CO)O)O)O)O</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># C([C@H]([C@H]([C@@H]([C@H](C=O)O)O)O)O)O</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># C(C(C(C(C(C=O)O)O)O)O)O</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>cs = pcp.get_compounds(<span style="color:#0ff;font-weight:bold">&#39;C20H41Br&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;formula&#39;</span>)
</span></span><span style="display:flex;"><span>cs[:<span style="color:#ff0;font-weight:bold">3</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [Compound(20271), Compound(23148745), Compound(10808570)]</span>
</span></span><span style="display:flex;"><span>cs_df = pcp.compounds_to_frame(cs, properties=[<span style="color:#0ff;font-weight:bold">&#39;isomeric_smiles&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;xlogp&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;rotatable_bond_count&#39;</span>])
</span></span><span style="display:flex;"><span>cs_df.shape
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (43, 3)</span>
</span></span></code></pre></td></tr></table>
</div>
</div><hr>
<h1 id="补充mygene">补充：MyGene<a hidden class="anchor" aria-hidden="true" href="#补充mygene">#</a></h1>
<p>MyGene.info 是一个为生物信息学提供高效、统一、RESTful API 的基因注释服务平台，支持快速查询基因的<strong>ID转换</strong>、结构、功能、通路、蛋白、疾病、药物靶点等多种注释信息；数据来源整合了多个主流数据库（如 NCBI, Ensembl, UniProt, GO, KEGG, ChEMBL 等）。</p>
<ul>
<li><a href="https://docs.mygene.info/en/latest/index.html">https://docs.mygene.info/en/latest/index.html</a></li>
<li><a href="https://github.com/biothings/mygene.info">https://github.com/biothings/mygene.info</a></li>
<li><a href="https://pypi.org/project/mygene/">https://pypi.org/project/mygene/</a></li>
</ul>
<p>使用方式也非常简单，可以直接通过API接口，也提供了封装好的Python、R包进行调用。下面简单记录其Python包的用法，方便以后使用。</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f"># pip install mygene</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> mygene
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>mg = mygene.MyGeneInfo()
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>方式1：根据指定的entrezgene/ensembl id查询</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>res = mg.getgene(<span style="color:#ff0;font-weight:bold">1017</span>, fields=<span style="color:#0ff;font-weight:bold">&#39;all&#39;</span>)
</span></span><span style="display:flex;"><span>res.keys()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># dict_keys([&#39;AllianceGenome&#39;, &#39;HGNC&#39;, &#39;MIM&#39;, &#39;_id&#39;, &#39;_version&#39;, &#39;accession&#39;, &#39;agr&#39;, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            &#39;alias&#39;, &#39;chembl&#39;, &#39;ec&#39;, &#39;ensembl&#39;, &#39;entrezgene&#39;, &#39;exac&#39;, &#39;exons&#39;, &#39;exons_hg19&#39;, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            &#39;generif&#39;, &#39;genomic_pos&#39;, &#39;genomic_pos_hg19&#39;, &#39;go&#39;, &#39;homologene&#39;, &#39;interpro&#39;, &#39;ipi&#39;, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            &#39;map_location&#39;, &#39;name&#39;, &#39;other_names&#39;, &#39;pantherdb&#39;, &#39;pathway&#39;, &#39;pdb&#39;, &#39;pfam&#39;, &#39;pharmgkb&#39;, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            &#39;pharos&#39;, &#39;pir&#39;, &#39;prosite&#39;, &#39;reagent&#39;, &#39;refseq&#39;, &#39;reporter&#39;,  &#39;summary&#39;, &#39;symbol&#39;, &#39;taxid&#39;, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            &#39;type_of_gene&#39;, &#39;umls&#39;, &#39;unigene&#39;, &#39;unii&#39;, &#39;uniprot&#39;, &#39;wikipedia&#39;])</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 笔记最后列出了每个key/field的含义说明</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 使用fields参数限定返回的注释条目</span>
</span></span><span style="display:flex;"><span>mg.getgene(<span style="color:#ff0;font-weight:bold">1017</span>, fields=<span style="color:#0ff;font-weight:bold">&#39;name,symbol,refseq&#39;</span>)
</span></span><span style="display:flex;"><span>mg.getgene(<span style="color:#ff0;font-weight:bold">1017</span>, fields=[<span style="color:#0ff;font-weight:bold">&#39;name&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;symbol&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;refseq.rna&#39;</span>])
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 同时查询多个基因</span>
</span></span><span style="display:flex;"><span>mg.getgenes([<span style="color:#ff0;font-weight:bold">1017</span>,<span style="color:#ff0;font-weight:bold">1018</span>,<span style="color:#0ff;font-weight:bold">&#39;ENSG00000148795&#39;</span>], fields=<span style="color:#0ff;font-weight:bold">&#39;name,symbol,entrezgene,taxid,type_of_gene&#39;</span>)
</span></span><span style="display:flex;"><span>mg.getgenes([<span style="color:#ff0;font-weight:bold">1017</span>,<span style="color:#ff0;font-weight:bold">1018</span>,<span style="color:#0ff;font-weight:bold">&#39;ENSG00000148795&#39;</span>], fields=<span style="color:#0ff;font-weight:bold">&#39;name,symbol,entrezgene,taxid,ensembl&#39;</span>, as_dataframe=<span style="color:#fff;font-weight:bold">True</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><p><strong>方式2：根据关键字匹配搜索</strong></p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>mg.query(<span style="color:#0ff;font-weight:bold">&#39;cdk2&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 默认返回最匹配的10个结果，且返回条目进包括entrezgene/name/symbol/taxid</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 针对特定条目进行搜索，且筛选物种</span>
</span></span><span style="display:flex;"><span>mg.query(<span style="color:#0ff;font-weight:bold">&#39;symbol:cdk2&#39;</span>, species=<span style="color:#0ff;font-weight:bold">&#39;human&#39;</span>, fields=<span style="color:#0ff;font-weight:bold">&#39;name,symbol,entrezgene,taxid,ensembl&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 批量搜索，scopes参数设置针对某一条目进行搜索</span>
</span></span><span style="display:flex;"><span>mg.querymany([<span style="color:#ff0;font-weight:bold">1017</span>, <span style="color:#ff0;font-weight:bold">695</span>], scopes=<span style="color:#0ff;font-weight:bold">&#39;entrezgene&#39;</span>, species=<span style="color:#0ff;font-weight:bold">&#39;human&#39;</span>)
</span></span><span style="display:flex;"><span>mg.querymany([<span style="color:#0ff;font-weight:bold">&#34;TP53&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;EGFR&#34;</span>, <span style="color:#0ff;font-weight:bold">&#39;NA_TEST&#39;</span>], scopes=<span style="color:#0ff;font-weight:bold">&#39;symbol&#39;</span>, species=<span style="color:#0ff;font-weight:bold">&#39;human&#39;</span>, as_dataframe=<span style="color:#fff;font-weight:bold">True</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 当fetch_all=True，生成一个可迭代对象</span>
</span></span><span style="display:flex;"><span>kinases = mg.query(<span style="color:#0ff;font-weight:bold">&#39;name:kinase&#39;</span>, species=<span style="color:#0ff;font-weight:bold">&#39;human&#39;</span>, fetch_all=<span style="color:#fff;font-weight:bold">True</span>)
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span> gene in kinases:
</span></span><span style="display:flex;"><span>    <span style="color:#fff;font-weight:bold">print</span>(gene[<span style="color:#0ff;font-weight:bold">&#39;_id&#39;</span>], gene[<span style="color:#0ff;font-weight:bold">&#39;symbol&#39;</span>])
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>基本信息</li>
</ul>
<table>
  <thead>
      <tr>
          <th>字段名</th>
          <th>含义</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td><code>_id</code> / <code>entrezgene</code></td>
          <td>主键，通常等于 Entrez Gene ID，如 <code>1017</code></td>
      </tr>
      <tr>
          <td><code>_version</code></td>
          <td>数据版本号</td>
      </tr>
      <tr>
          <td><code>symbol</code></td>
          <td>基因符号，例如 <code>CDK2</code></td>
      </tr>
      <tr>
          <td><code>name</code></td>
          <td>基因全名，例如 &ldquo;cyclin-dependent kinase 2&rdquo;</td>
      </tr>
      <tr>
          <td><code>taxid</code></td>
          <td>NCBI 物种 ID，<code>9606</code> 表示人类</td>
      </tr>
      <tr>
          <td><code>type_of_gene</code></td>
          <td>基因类型，如 <code>protein-coding</code>、<code>ncRNA</code>、<code>pseudo</code> 等</td>
      </tr>
      <tr>
          <td><code>summary</code></td>
          <td>基因功能简要描述</td>
      </tr>
      <tr>
          <td><code>alias</code></td>
          <td>其他名称或别名（例如缩写、旧称）</td>
      </tr>
      <tr>
          <td><code>other_names</code></td>
          <td>附加名称（全称、别称等）</td>
      </tr>
  </tbody>
</table>
<ul>
<li>序列信息</li>
</ul>
<table>
  <thead>
      <tr>
          <th>字段名</th>
          <th>含义</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td><code>accession</code></td>
          <td>基因相关的 GenBank/RefSeq accession 列表</td>
      </tr>
      <tr>
          <td><code>refseq</code></td>
          <td>RefSeq 序列 ID（mRNA、protein、genomic 等）</td>
      </tr>
      <tr>
          <td><code>genomic_pos</code></td>
          <td>基因在最新人类基因组版本（如 hg38）上的位置</td>
      </tr>
      <tr>
          <td><code>genomic_pos_hg19</code></td>
          <td>在旧版本（hg19）上的位置信息</td>
      </tr>
      <tr>
          <td><code>exons</code></td>
          <td>外显子位置（hg38）</td>
      </tr>
      <tr>
          <td><code>exons_hg19</code></td>
          <td>外显子位置（hg19）</td>
      </tr>
  </tbody>
</table>
<ul>
<li>
<p>生物通路与功能注释</p>
<table>
  <thead>
      <tr>
          <th>字段名</th>
          <th>含义</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td><code>go</code></td>
          <td>Gene Ontology 注释，包括 BP（生物过程）、MF（分子功能）、CC（细胞成分）</td>
      </tr>
      <tr>
          <td><code>pathway</code></td>
          <td>涉及的信号通路信息（来自 Reactome、WikiPathways、KEGG 等）</td>
      </tr>
      <tr>
          <td><code>pantherdb</code></td>
          <td>Panther 通路数据库注释</td>
      </tr>
      <tr>
          <td><code>pharmgkb</code></td>
          <td>药物-基因相互作用信息</td>
      </tr>
      <tr>
          <td><code>pharos</code></td>
          <td>药物开发相关注释（Drug Target 分类）</td>
      </tr>
  </tbody>
</table>
</li>
<li>
<p>蛋白结构和功能域</p>
</li>
</ul>
<table>
  <thead>
      <tr>
          <th>字段名</th>
          <th>含义</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td><code>uniprot</code></td>
          <td>UniProt 蛋白数据库的 ID 和注释</td>
      </tr>
      <tr>
          <td><code>interpro</code></td>
          <td>InterPro 蛋白家族或功能域</td>
      </tr>
      <tr>
          <td><code>pfam</code></td>
          <td>Pfam 蛋白结构域</td>
      </tr>
      <tr>
          <td><code>prosite</code></td>
          <td>Prosite 模式/结构域数据库</td>
      </tr>
      <tr>
          <td><code>pdb</code></td>
          <td>相关蛋白质的三维结构（来自 Protein Data Bank）</td>
      </tr>
      <tr>
          <td><code>ec</code></td>
          <td>酶编号（Enzyme Commission Number），如 kinase 常有</td>
      </tr>
      <tr>
          <td><code>ipi</code></td>
          <td>International Protein Index（已弃用）</td>
      </tr>
      <tr>
          <td><code>pir</code></td>
          <td>PIR 蛋白数据库 ID</td>
      </tr>
  </tbody>
</table>
<ul>
<li>同源与进化</li>
</ul>
<table>
  <thead>
      <tr>
          <th>字段名</th>
          <th>含义</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td><code>homologene</code></td>
          <td>NCBI HomoloGene 同源基因信息 ([tax_id, entrezgene_id])</td>
      </tr>
      <tr>
          <td><code>ensembl</code></td>
          <td>Ensembl 基因数据库的 ID 和注释</td>
      </tr>
      <tr>
          <td><code>MIM</code></td>
          <td>OMIM（人类遗传疾病数据库）相关信息</td>
      </tr>
      <tr>
          <td><code>unigene</code></td>
          <td>NCBI UniGene 聚类（旧）</td>
      </tr>
      <tr>
          <td><code>umls</code></td>
          <td>Unified Medical Language System 编号（跨数据库映射）</td>
      </tr>
      <tr>
          <td><code>exac</code></td>
          <td>ExAC 等人群变异数据汇总（Exome Aggregation Consortium）</td>
      </tr>
      <tr>
          <td><code>HGNC</code></td>
          <td>HGNC（人类基因命名委员会）ID 和详细命名信息</td>
      </tr>
  </tbody>
</table>
<ul>
<li>
<p>实验与试剂相关</p>
<table>
  <thead>
      <tr>
          <th>字段名</th>
          <th>含义</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td><code>reporter</code></td>
          <td>基因在不同表达芯片平台上的探针 ID（如 Affymetrix）</td>
      </tr>
      <tr>
          <td><code>reagent</code></td>
          <td>BioGPS 等平台上的实验试剂资源</td>
      </tr>
      <tr>
          <td><code>agr</code></td>
          <td>Alliance of Genome Resources ID（联合资源数据库）</td>
      </tr>
      <tr>
          <td><code>AllianceGenome</code></td>
          <td>Alliance Genome 注释（也可能重复 agr）</td>
      </tr>
  </tbody>
</table>
</li>
<li>
<p>药物相关</p>
</li>
</ul>
<table>
  <thead>
      <tr>
          <th>字段名</th>
          <th>含义</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td><code>chembl</code></td>
          <td>ChEMBL 中的药物靶点信息</td>
      </tr>
      <tr>
          <td><code>unii</code></td>
          <td>FDA UNII（药物成分唯一标识符）</td>
      </tr>
  </tbody>
</table>
<ul>
<li>网络资源</li>
</ul>
<table>
  <thead>
      <tr>
          <th>字段名</th>
          <th>含义</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td><code>wikipedia</code></td>
          <td>Wikipedia 页面链接（如果有）</td>
      </tr>
      <tr>
          <td><code>map_location</code></td>
          <td>染色体图谱位置，如 <code>12q13.2</code>（12号染色体长臂）</td>
      </tr>
      <tr>
          <td><code>generif</code></td>
          <td>GeneRIFs：简短文献注释，来自 PubMed 摘要（Gene Reference Into Function）</td>
      </tr>
  </tbody>
</table>


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E7%94%9F%E4%BF%A1%E5%9F%BA%E7%A1%80/">生信基础</a></li>
      <li><a href="https://lishensuo.github.io/en/tags/%E5%8C%96%E5%90%88%E7%89%A9/">化合物</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/053%E4%BA%BA%E7%B1%BB%E5%9F%BA%E5%9B%A0%E7%BB%84%E5%9F%BA%E7%A1%80%E7%9F%A5%E8%AF%86%E4%B8%8E%E4%B8%8B%E8%BD%BD%E6%9F%A5%E8%AF%A2/">
    <span class="title">« Prev Page</span>
    <br>
    <span>人类基因组基础知识与下载查询</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/055gsea%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7/">
    <span class="title">Next Page »</span>
    <br>
    <span>GSEA富集分析工具</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
