<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>TCGAbiolinksåŒ…-TCGAæ•°æ®ä¸‹è½½ä¸å·®å¼‚åˆ†æ | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="è½¬å½•ç»„">
<meta name="description" content="TCGAbiolinksåŒ…æ˜¯ä¸€ç«™å¼åˆ†æTCGAæ•°æ®çš„RåŒ…å·¥å…·ï¼Œå®ƒé›†æˆäº†TCGAæ•°æ®ä¸‹è½½ã€åˆ†æã€å¯è§†åŒ–çš„å…¨éƒ¨æµç¨‹ã€‚æ­¤æ¬¡ç³»åˆ—ç¬”è®°ä¸»è¦è·Ÿç€ TCGAbiolinkså¸®åŠ©æ–‡æ¡£é‡æ–°å­¦ä¹ ä¸‹TCGAæ•°æ®æŒ–æ˜æµç¨‹ã€‚">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/046tcgabiolinks%E5%8C%85-tcga%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD%E4%B8%8E%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/046tcgabiolinks%E5%8C%85-tcga%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD%E4%B8%8E%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="TCGAbiolinksåŒ…-TCGAæ•°æ®ä¸‹è½½ä¸å·®å¼‚åˆ†æ" />
<meta property="og:description" content="TCGAbiolinksåŒ…æ˜¯ä¸€ç«™å¼åˆ†æTCGAæ•°æ®çš„RåŒ…å·¥å…·ï¼Œå®ƒé›†æˆäº†TCGAæ•°æ®ä¸‹è½½ã€åˆ†æã€å¯è§†åŒ–çš„å…¨éƒ¨æµç¨‹ã€‚æ­¤æ¬¡ç³»åˆ—ç¬”è®°ä¸»è¦è·Ÿç€ TCGAbiolinkså¸®åŠ©æ–‡æ¡£é‡æ–°å­¦ä¹ ä¸‹TCGAæ•°æ®æŒ–æ˜æµç¨‹ã€‚" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/046tcgabiolinks%E5%8C%85-tcga%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD%E4%B8%8E%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2022-04-30T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2022-05-06T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="TCGAbiolinksåŒ…-TCGAæ•°æ®ä¸‹è½½ä¸å·®å¼‚åˆ†æ"/>
<meta name="twitter:description" content="TCGAbiolinksåŒ…æ˜¯ä¸€ç«™å¼åˆ†æTCGAæ•°æ®çš„RåŒ…å·¥å…·ï¼Œå®ƒé›†æˆäº†TCGAæ•°æ®ä¸‹è½½ã€åˆ†æã€å¯è§†åŒ–çš„å…¨éƒ¨æµç¨‹ã€‚æ­¤æ¬¡ç³»åˆ—ç¬”è®°ä¸»è¦è·Ÿç€ TCGAbiolinkså¸®åŠ©æ–‡æ¡£é‡æ–°å­¦ä¹ ä¸‹TCGAæ•°æ®æŒ–æ˜æµç¨‹ã€‚"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "åˆ†ç±»",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "TCGAbiolinksåŒ…-TCGAæ•°æ®ä¸‹è½½ä¸å·®å¼‚åˆ†æ",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/046tcgabiolinks%E5%8C%85-tcga%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD%E4%B8%8E%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "TCGAbiolinksåŒ…-TCGAæ•°æ®ä¸‹è½½ä¸å·®å¼‚åˆ†æ",
  "name": "TCGAbiolinksåŒ…-TCGAæ•°æ®ä¸‹è½½ä¸å·®å¼‚åˆ†æ",
  "description": "TCGAbiolinksåŒ…æ˜¯ä¸€ç«™å¼åˆ†æTCGAæ•°æ®çš„RåŒ…å·¥å…·ï¼Œå®ƒé›†æˆäº†TCGAæ•°æ®ä¸‹è½½ã€åˆ†æã€å¯è§†åŒ–çš„å…¨éƒ¨æµç¨‹ã€‚æ­¤æ¬¡ç³»åˆ—ç¬”è®°ä¸»è¦è·Ÿç€ TCGAbiolinkså¸®åŠ©æ–‡æ¡£é‡æ–°å­¦ä¹ ä¸‹TCGAæ•°æ®æŒ–æ˜æµç¨‹ã€‚\n",
  "keywords": [
    "è½¬å½•ç»„"
  ],
  "articleBody": "TCGAbiolinksåŒ…æ˜¯ä¸€ç«™å¼åˆ†æTCGAæ•°æ®çš„RåŒ…å·¥å…·ï¼Œå®ƒé›†æˆäº†TCGAæ•°æ®ä¸‹è½½ã€åˆ†æã€å¯è§†åŒ–çš„å…¨éƒ¨æµç¨‹ã€‚æ­¤æ¬¡ç³»åˆ—ç¬”è®°ä¸»è¦è·Ÿç€ TCGAbiolinkså¸®åŠ©æ–‡æ¡£é‡æ–°å­¦ä¹ ä¸‹TCGAæ•°æ®æŒ–æ˜æµç¨‹ã€‚\nå®˜æ–¹æ–‡æ¡£ï¼šhttps://bioconductor.org/packages/release/bioc/vignettes/TCGAbiolinks/inst/doc/index.html\næ–‡çŒ®ï¼šTCGAbiolinks: an R/Bioconductor package for integrative analysis of TCGA data https://pubmed.ncbi.nlm.nih.gov/26704973/\nå®‰è£…åŒ…\n1 2 3 4 5 6 7 if (!requireNamespace(\"BiocManager\", quietly = TRUE)) install.packages(\"BiocManager\") #From github BiocManager::install(\"BioinformaticsFMRP/TCGAbiolinksGUI.data\") BiocManager::install(\"BioinformaticsFMRP/TCGAbiolinks\") #From Bioconductor BiocManager::install(\"TCGAbiolinks\") ä¸€ã€æŸ¥æ‰¾TCGAæ•°æ® ä¸»è¦é€šè¿‡GDCquery()å‡½æ•°æŸ¥æ‰¾æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 library(TCGAbiolinks) GDCquery( project, data.category, data.type, workflow.type, legacy = FALSE, access, platform, file.type, barcode, data.format, experimental.strategy, sample.type ) 1 é€‰æ‹©è‚¿ç˜¤ç±»å‹ projectå‚æ•°ï¼šæŒ‡å®šä¸€ä¸ªæˆ–å¤šä¸ªæ„Ÿå…´è¶£çš„TCGAé¡¹ç›®å å¦‚ä¸‹ä»£ç æ‰€ç¤ºï¼Œä¾›åŒ…æ‹¬33ç§TCGAç™Œç—‡ç±»å‹ã€‚ç™Œç—‡ç®€ç§°å¯¹åº”å…¨ç§°åŠä¸­æ–‡åè§æ–‡æœ«ã€‚ 1 2 3 4 5 6 7 8 9 projects = TCGAbiolinks:::getGDCprojects()$project_id TCGAs = grep(\"TCGA\", projects, value = T) sort(TCGAs) # [1] \"TCGA-ACC\" \"TCGA-BLCA\" \"TCGA-BRCA\" \"TCGA-CESC\" \"TCGA-CHOL\" \"TCGA-COAD\" # [7] \"TCGA-DLBC\" \"TCGA-ESCA\" \"TCGA-GBM\" \"TCGA-HNSC\" \"TCGA-KICH\" \"TCGA-KIRC\" # [13] \"TCGA-KIRP\" \"TCGA-LAML\" \"TCGA-LGG\" \"TCGA-LIHC\" \"TCGA-LUAD\" \"TCGA-LUSC\" # [19] \"TCGA-MESO\" \"TCGA-OV\" \"TCGA-PAAD\" \"TCGA-PCPG\" \"TCGA-PRAD\" \"TCGA-READ\" # [25] \"TCGA-SARC\" \"TCGA-SKCM\" \"TCGA-STAD\" \"TCGA-TGCT\" \"TCGA-THCA\" \"TCGA-THYM\" # [31] \"TCGA-UCEC\" \"TCGA-UCS\" \"TCGA-UVM\" 2 hg19/hg38 ä¸»è¦æ ¹æ®å‚è€ƒåŸºå› ç»„çš„ä¸åŒï¼ŒåŒ…å«ä¸¤å¥—æ•°æ®ï¼šGDC Legacy Archiveã€ä¸»è¦GRCh37 (hg19)ã€‘ï¼ŒGDC harmonized databaseã€GRCh38 (hg38)ã€‘ é€šè¿‡è®¾ç½®å‚æ•°legacy ï¼Œé»˜è®¤ä¸ºFALSE(hg38)ï¼›TRUEåˆ™è¡¨ç¤ºä½¿ç”¨hg19å‚è€ƒåŸºå› ç»„çš„æµ‹åºæ•°æ®ã€‚ 3 ä¸‹è½½æ•°æ®ç±»å‹ åŸºäºä¸Šè¿°çš„å‚æ•°ï¼Œæˆ‘ä»¬å¯ä»¥è®¾ç½®å¦‚ä¸‹å‚æ•°ï¼Œäº¤ä»£æˆ‘ä»¬çš„ç›®æ ‡æ•°æ®ç±»å‹\ndata.category = æŒ‡å®šä¸‹è½½ä»€ä¹ˆç±»å‹çš„æ•°æ®ï¼šå¦‚è½¬å½•ç»„æ•°æ®ã€ä¸´åºŠæ•°æ®â€¦. 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 ##(1) GDC harmonized databaseã€GRCh38 (hg38)ã€‘ TCGAbiolinks:::getProjectSummary(\"TCGA-BRCA\", legacy = FALSE)$data_categories$data_category # [1] \"Structural Variation\" \"Simple Nucleotide Variation\" # [3] \"Copy Number Variation\" \"Transcriptome Profiling\" # [5] \"DNA Methylation\" \"Sequencing Reads\" # [7] \"Biospecimen\" \"Clinical\" # [9] \"Proteome Profiling\" ##(2) GDC Legacy Archiveã€ä¸»è¦GRCh37 (hg19)ã€‘ TCGAbiolinks:::getProjectSummary(\"TCGA-BRCA\", legacy = TRUE)$data_categories$data_category # [1] \"Protein expression\" \"Copy number variation\" # [3] \"Biospecimen\" \"Simple nucleotide variation\" # [5] \"Gene expression\" \"Raw microarray data\" # [7] \"DNA methylation\" \"Clinical\" # [9] \"Raw sequencing data\" data.type = æ›´åŠ ç»†èŠ‚çš„æ•°æ®ç±»å‹é€‰æ‹©ï¼Œä¾‹å¦‚harmonizedçš„Transcriptome Profilingæœ‰ Gene Expression Quantification, Isoform Expression Quantification, miRNA Expression Quantification, Splice Junction Quantification4ç§æ•°æ®ã€‚è€Œæˆ‘ä»¬ä¸€èˆ¬éœ€è¦çš„æ˜¯Gene Expression Quantificationã€‚ workflow.type = åŒä¸€ä¸ªæµ‹åºæ•°æ®å¯èƒ½æœ‰ä¸åŒçš„pipelineå¤„ç†æµç¨‹ï¼Œå¦‚æœæœ‰ä¸åŒçš„å¤„ç†æµç¨‹ï¼Œéœ€è¦é€‰æ‹©ã€‚æ®æœ€æ–°çš„TCGAç‰ˆæœ¬ï¼Œä»…æä¾›STAR - Countsã€‚ platform = æµ‹åºå¹³å°ï¼ˆoptionalï¼‰ file.type = å…·ä½“çš„æ•°æ®æ–‡ä»¶ï¼ˆoptional, for legacyï¼‰ å¦‚æœä¸çŸ¥é“ç›®æ ‡æ•°æ®çš„ä¸Šè¿°ä¿¡æ¯ï¼Œå¯ä»¥å‚è€ƒä¸‹é¢çš„æ¦‚è¿° å…·ä½“å‚è€ƒæ–‡æœ«ç¬¬äºŒç‚¹ã€‚\n4 æ ·æœ¬æ ‡ç­¾Barcode å®Œæ•´çš„barcodeï¼šå½¢å¦‚ TCGA-G4-6317-02A-11D-2064-05ï¼Œè¿™ä¸ªæ ‡ç­¾åŒ…å«äº†ä»ç—…äººæ¥æºåˆ°æµ‹åºè¿‡ç¨‹ã€åˆ†æçš„æ‰€æœ‰ä¿¡æ¯ï¼Œå¦‚ä¸‹å›¾æ‰€ç¤ºæ¯”è¾ƒé‡è¦çš„æ˜¯Participantã€Sample ã€Portionä¸‰ä¸ªéƒ¨åˆ†ï¼Œåˆ†åˆ«äº¤ä»£äº†ç—…äººç¼–å·ã€æ ·æœ¬ç±»å‹ã€æµ‹åºç±»å‹ ç—…äººçš„idï¼šå½¢å¦‚ TCGA-G4-6317 æ ·æœ¬æ¥æºçš„idï¼šå½¢å¦‚ TCGA-G4-6317-02 å…¶ä¸­æ¯”è¾ƒé‡è¦çš„æ˜¯äº¤ä»£æ ·æœ¬ç±»å‹çš„Sampleçš„ä¸¤ä½æ•°ä¿¡æ¯ï¼Œæ˜¯åé¢è¿›è¡Œå·®å¼‚åˆ†æçš„åˆ†ç»„ä¾æ®ã€‚å…·ä½“å¯¹åº”çš„å«ä¹‰å¦‚ä¸‹ã€‚ä¾‹å¦‚01è¡¨ç¤ºç—…äººçš„åŸä½ç˜¤ç»„ç»‡ï¼›11è¡¨ç¤ºæ¥è‡ªç—…äººçš„æ­£å¸¸ç»„ç»‡â€¦. åŸºäºä¸Šè¿°ç†è§£ï¼Œæˆ‘ä»¬ä¹Ÿå¯ä»¥è®¾ç½®sample.type = å‚æ•°æŒ‡å®šä¸‹è½½æ„Ÿå…´è¶£çš„æ ·æœ¬ç±»å‹æ•°æ®ï¼Œä¾‹å¦‚sample.type = \"Primary Tumor\"\nå¯¹äºç»™å®šçš„TCGA barcodeï¼Œå¯ä»¥åˆ©ç”¨TCGAquery_SampleTypes()æå–å‡ºç›®æ ‡åˆ†ç»„çš„æ ·æœ¬ï¼›TCGAquery_MatchedCoupledSampleTypes()å‡½æ•°å¯ä»¥æå–æ¥è‡ªåŒä¸€ç—…äººçš„é…å¯¹æ ·æœ¬æ•°æ®ã€‚\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 query \u003c- GDCquery(project = c(\"TCGA-BRCA\"), legacy = FALSE, data.category = \"Transcriptome Profiling\", data.type = \"Gene Expression Quantification\", workflow.type = \"HTSeq - Counts\") dim(getResults(query)) #[1] 1222 29 query_info = getResults(query) TP = TCGAquery_SampleTypes(query_info$sample.submitter_id,\"TP\") NT = TCGAquery_SampleTypes(query_info$sample.submitter_id,\"NT\") query \u003c- GDCquery(project = c(\"TCGA-BRCA\"), legacy = FALSE, data.category = \"Transcriptome Profiling\", data.type = \"Gene Expression Quantification\", workflow.type = \"HTSeq - Counts\", barcode = c(TP, NT)) dim(getResults(query)) #[1] 1215 29 Pair_sample = TCGAquery_MatchedCoupledSampleTypes(query_info$sample.submitter_id,c(\"NT\",\"TP\")) query \u003c- GDCquery(project = c(\"TCGA-BRCA\"), legacy = FALSE, data.category = \"Transcriptome Profiling\", data.type = \"Gene Expression Quantification\", workflow.type = \"HTSeq - Counts\", barcode = Pair_sample) dim(getResults(query)) #[1] 229 29 å¦‚ä¸Šæ˜¯æŸ¥è¯¢TCGAç›®æ ‡æ•°æ®çš„å‡ ç§å¸¸è§æ ‡å‡†ï¼Œè¿˜æœ‰å‡ ä¸ªå‚æ•°æ²¡æœ‰ä»‹ç»ï¼Œå¯å‚çœ‹å‡½æ•°å¸®åŠ©æ–‡æ¡£ã€‚å¯æ ¹æ®è‡ªå·±çš„ç›®çš„çµæ´»è®¾ç½®ä¸Šè¿°å‚æ•°ã€‚\näºŒã€ä»æ•°æ®ä¸‹è½½åˆ°å·®å¼‚åˆ†æ ä»¥èƒ†ç®¡ç™Œä¸ºä¾‹ï¼Œæ¼”ç¤ºä»æ•°æ®ä¸‹è½½åˆ°å·®å¼‚åˆ†æçš„å…¨æµç¨‹ã€‚\n1 Legacy(hg19) 1.1 ä¸‹è½½æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 library(TCGAbiolinks) ##(1) è·å–ä¸‹è½½ç›®æ ‡æ–‡ä»¶ query \u003c- GDCquery(project = \"TCGA-CHOL\", legacy = TRUE, data.category = \"Gene expression\", data.type = \"Gene expression quantification\", platform = \"Illumina HiSeq\", file.type = \"results\") ##(2) ä¸‹è½½æ•°æ®åˆ°æœ¬åœ° # files.per.chunk = 10 è¡¨ç¤ºåˆ†æ‰¹ä¸‹è½½ï¼Œæ¯æ‰¹ä¸‹è½½10ä¸ªç—…äººçš„æ•°æ®ï¼Œå¯é¿å…ä¸­é€”æŠ¥é”™ï¼Œè€Œå‰åŠŸå°½å¼ƒã€‚ GDCdownload(query, files.per.chunk = 10) ##(3) åŠ è½½æ•°æ®è‡³å½“å‰ç¯å¢ƒ data \u003c- GDCprepare(query, save = T, save.filename = \"CHOL_RNAseq_hg19.rda\") 1.2 æ¢ç´¢æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 library(SummarizedExperiment) dim(assay(data)) #[1] 19947 45 ##(1) æŸ¥çœ‹è¡¨è¾¾çŸ©é˜µ assayNames(data) # [1] \"raw_count\" \"scaled_estimate\" assay(data, \"raw_count\")[1:3,1:3] # TCGA-W5-AA2I-11A-11R-A41I-07 TCGA-W5-AA2U-11A-11R-A41I-07 TCGA-W5-AA2I-01A-32R-A41I-07 # A1BG|1 80062.65 79948.1 2519.20 # A2M|2 79697.45 126813.9 4107.88 # NAT1|9 364.00 397.0 126.00 ##(2) æ ·æœ¬(ä¸´åºŠ)ä¿¡æ¯ dim(colData(data)) #[1] 45 201 colData(data)[1:4,1:4] ##(3) åŸºå› ä¿¡æ¯ rowData(data)[1:6,1:5] # DataFrame with 6 rows and 3 columns # gene_id entrezgene ensembl_gene_id # # A1BG A1BG 1 ENSG00000121410 # A2M A2M 2 ENSG00000175899 # NAT1 NAT1 9 ENSG00000171428 rowRanges(data) 1.3 æ•´ç†æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 ##(1)æ›´æ”¹è¡¨è¾¾çŸ©é˜µè¡Œå/åŸºå› åæ ¼å¼ rownames(data) = stringr::str_split(rownames(data), \"[|]\", simplify = T)[,1] data = data[!duplicated(rownames(data)),] dim(data) # [1] 19938 45 assay(data, \"raw_count\")[1:3,1:3] # TCGA-W5-AA2I-11A-11R-A41I-07 TCGA-W5-AA2U-11A-11R-A41I-07 TCGA-W5-AA2I-01A-32R-A41I-07 # A1BG 80062.65 79948.1 2519.20 # A2M 79697.45 126813.9 4107.88 # NAT1 364.00 397.0 126.00 ##(2) åˆ†ææ ·æœ¬é—´ç›¸å…³æ€§ï¼Œé‰´å®šç¦»ç¾¤ç‚¹æ ·æœ¬ CorOutliers \u003c- TCGAanalyze_Preprocessing(data, cor.cut = 0, filename=\"CHOL45_hg19_cor.png\") ##(3) æ ‡å‡†åŒ– dataNorm \u003c- TCGAanalyze_Normalization(tabDF = data, geneInfo = geneInfo) #æ²¡æœ‰logå¤„ç† dim(dataNorm) # [1] 17618 45 class(dataNorm) # [1] \"matrix\" \"array\" dataNorm[1:3,1:3] # TCGA-W5-AA2I-11A-11R-A41I-07 TCGA-W5-AA2U-11A-11R-A41I-07 TCGA-W5-AA2I-01A-32R-A41I-07 # A1BG 80063 79948 2519 # A2M 79697 126814 4108 # NAT1 364 397 126 ##(4) è¿‡æ»¤ä½è¡¨è¾¾åŸºå›  dataFilt \u003c- TCGAanalyze_Filtering(tabDF = dataNorm, method = \"quantile\", qnt.cut = 0.05) dim(dataFilt) exp_data = dataFilt 1.4 å·®å¼‚åˆ†æ 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 ##(1) ç¡®å®šåˆ†ç»„æ ·æœ¬ samplesNT \u003c- TCGAquery_SampleTypes(barcode = colnames(exp_data), typesample = c(\"NT\")) samplesTP \u003c- TCGAquery_SampleTypes(barcode = colnames(exp_data), typesample = c(\"TP\")) ##(2) å·®å¼‚åˆ†æ dataDEGs \u003c- TCGAanalyze_DEA(mat1 = exp_data[,samplesNT], mat2 = exp_data[,samplesTP], Cond1type = \"Normal\", Cond2type = \"Tumor\", pipeline = \"edgeR\", method = \"glmLRT\") ##(3) æ•´ç†ç»“æœ DEGs \u003c- TCGAanalyze_LevelTab(dataDEGs,\"Tumor\",\"Normal\", exp_data[,samplesTP],exp_data[,samplesNT]) DEGs = DEGs[order(DEGs$FDR),] DEGs$Direct = ifelse(DEGs$logFC\u003e0, \"Up\", \"Down\") head(DEGs) # mRNA logFC FDR Delta Tumor Normal Direct # USH2A USH2A -5.261122 1.805847e-46 -150.9650 28.69444 1207.88889 Down # KCNN2 KCNN2 -4.344681 1.045519e-36 -188.2695 43.33333 893.77778 Down # LCAT LCAT -3.683368 2.344098e-31 -1875.2437 509.11111 6640.22222 Down # IQCE IQCE 4.229960 6.301169e-30 12462.2846 2946.19444 153.33333 Up # DCXR DCXR -4.070648 9.023484e-30 -7408.0132 1819.86111 32284.22222 Down # PRSS16 PRSS16 6.947385 3.517876e-28 9163.9866 1319.05556 10.22222 Up DEGs = DEGs[order(DEGs$FDR),] DEGs$Direct = ifelse(DEGs$logFC\u003e0, \"Up\", \"Down\") table(DEGs$FDR\u003c0.01,DEGs$Direct) DEGs_sig = subset(DEGs, FDR\u003c0.01) table(DEGs_sig$Direct, abs(DEGs_sig$logFC)\u003e1) 2 Harmonized(hg38) 2.1 ä¸‹è½½æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 library(TCGAbiolinks) ##(1) è·å–ä¸‹è½½ç›®æ ‡æ–‡ä»¶ query \u003c- GDCquery(project = \"TCGA-CHOL\", legacy = FALSE, data.category = \"Transcriptome Profiling\", data.type = \"Gene Expression Quantification\") ##(2) ä¸‹è½½æ•°æ®åˆ°æœ¬åœ° # files.per.chunk = 10 è¡¨ç¤ºåˆ†æ‰¹ä¸‹è½½ï¼Œæ¯æ‰¹ä¸‹è½½10ä¸ªç—…äººçš„æ•°æ®ï¼Œå¯é¿å…ä¸­é€”æŠ¥é”™ï¼Œè€Œå‰åŠŸå°½å¼ƒã€‚ GDCdownload(query, files.per.chunk = 10) ##(3) åŠ è½½æ•°æ®è‡³å½“å‰ç¯å¢ƒ data \u003c- GDCprepare(query, save = T, save.filename = \"CHOL_RNAseq_hg38.rda\") 2.2 æ¢ç´¢æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 library(SummarizedExperiment) dim(assay(data)) #[1] 60660 44 ##(1) æŸ¥çœ‹è¡¨è¾¾çŸ©é˜µ assayNames(data) #[1] \"unstranded\" \"stranded_first\" \"stranded_second\" \"tpm_unstrand\" #[5] \"fpkm_unstrand\" \"fpkm_uq_unstrand\" assay(data, \"unstranded\")[1:3,1:3] # TCGA-W5-AA2X-01A-11R-A41I-07 TCGA-W5-AA2X-11A-11R-A41I-07 TCGA-W5-AA33-01A-11R-A41I-07 # ENSG00000000003.15 3310 2322 11312 # ENSG00000000005.6 0 0 1 # ENSG00000000419.13 1881 716 968 ##(2) æ ·æœ¬(ä¸´åºŠ)ä¿¡æ¯ dim(colData(data)) #[1] 44 201 colData(data)[1:4,1:4] ##(3) åŸºå› ä¿¡æ¯ t(rowData(data)[1,]) # TransposedDataFrame with 10 rows and 1 column # ENSG00000000003.15 # source HAVANA # type gene # score NA # phase NA # gene_id ENSG00000000003.15 # gene_type protein_coding # gene_name TSPAN6 # level 2 # hgnc_id HGNC:11858 # havana_gene OTTHUMG00000022002.2 table(rowData(data)$gene_type) %\u003e% sort(decreasing = T) %\u003e% head() # protein_coding lncRNA processed_pseudogene unprocessed_pseudogene misc_RNA snRNA # 19962 16901 10167 2614 2212 1901 rowRanges(data) 2.3 æ•´ç†æ•°æ® 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 ##(1)æ›´æ”¹è¡¨è¾¾çŸ©é˜µè¡Œå/åŸºå› åæ ¼å¼ï¼Œç­›é€‰åŸºå›  rownames(data) = rowData(data)$gene_name data = data[!duplicated(rownames(data)),] data = data[rowData(data)$gene_type==\"protein_coding\",] dim(data) # [1] 19934 44 assay(data, \"unstranded\")[1:3,1:3] # TCGA-W5-AA2X-01A-11R-A41I-07 TCGA-W5-AA2X-11A-11R-A41I-07 TCGA-W5-AA33-01A-11R-A41I-07 # TSPAN6 3310 2322 11312 # TNMD 0 0 1 # DPM1 1881 716 968 ##(2) åˆ†ææ ·æœ¬é—´ç›¸å…³æ€§ï¼Œé‰´å®šç¦»ç¾¤ç‚¹æ ·æœ¬ CorOutliers \u003c- TCGAanalyze_Preprocessing(data, cor.cut = 0, filename=\"CHOL45_hg38_cor.png\") ##(3) æ ‡å‡†åŒ– dataNorm \u003c- TCGAanalyze_Normalization(tabDF = data, geneInfo = geneInfo) #æ²¡æœ‰logå¤„ç† dim(dataNorm) # [1] 15196 44 class(dataNorm) # [1] \"matrix\" \"array\" dataNorm[1:3,1:3] # TCGA-W5-AA2X-01A-11R-A41I-07 TCGA-W5-AA2X-11A-11R-A41I-07 TCGA-W5-AA33-01A-11R-A41I-07 # TSPAN6 3310 2322 11312 # DPM1 1881 716 968 # SCYL3 966 315 1296 ##(4) è¿‡æ»¤ä½è¡¨è¾¾åŸºå›  # dataFilt \u003c- TCGAanalyze_Filtering(tabDF = dataNorm, # method = \"quantile\", # qnt.cut = 0.05) # dim(dataFilt) exp_data = dataNorm 2.4 å·®å¼‚åˆ†æ (1) edgeR 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 ##(1) ç¡®å®šåˆ†ç»„æ ·æœ¬ samplesNT \u003c- TCGAquery_SampleTypes(barcode = colnames(exp_data), typesample = c(\"NT\")) samplesTP \u003c- TCGAquery_SampleTypes(barcode = colnames(exp_data), typesample = c(\"TP\")) ##(2) å·®å¼‚åˆ†æ dataDEGs \u003c- TCGAanalyze_DEA(mat1 = exp_data[,samplesNT], mat2 = exp_data[,samplesTP], Cond1type = \"Normal\", Cond2type = \"Tumor\", pipeline = \"edgeR\", method = \"glmLRT\") ##(3) æ•´ç†ç»“æœ DEGs \u003c- TCGAanalyze_LevelTab(dataDEGs,\"Tumor\",\"Normal\", exp_data[,samplesTP],exp_data[,samplesNT]) DEGs = DEGs[order(DEGs$FDR),] DEGs$Direct = ifelse(DEGs$logFC\u003e0, \"Up\", \"Down\") head(DEGs) # mRNA logFC FDR Delta Tumor Normal Direct # USH2A USH2A -5.261122 1.805847e-46 -150.9650 28.69444 1207.88889 Down # KCNN2 KCNN2 -4.344681 1.045519e-36 -188.2695 43.33333 893.77778 Down # LCAT LCAT -3.683368 2.344098e-31 -1875.2437 509.11111 6640.22222 Down # IQCE IQCE 4.229960 6.301169e-30 12462.2846 2946.19444 153.33333 Up # DCXR DCXR -4.070648 9.023484e-30 -7408.0132 1819.86111 32284.22222 Down # PRSS16 PRSS16 6.947385 3.517876e-28 9163.9866 1319.05556 10.22222 Up DEGs = DEGs[order(DEGs$FDR),] DEGs$Direct = ifelse(DEGs$logFC\u003e0, \"Up\", \"Down\") table(DEGs$FDR\u003c0.01,DEGs$Direct) DEGs_sig = subset(DEGs, FDR\u003c0.01) table(DEGs_sig$Direct, abs(DEGs_sig$logFC)\u003e1) (2) Deseq2 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 ##(1)Deseq2å·®å¼‚åˆ†æ library(TCGAbiolinks) library(DESeq2) library(tidyverse) data$sample_type = factor(data$sample_type, levels = c(\"Solid Tissue Normal\",\"Primary Tumor\")) ddsSE \u003c- DESeqDataSet(data, design = ~ sample_type) keep \u003c- rowSums(counts(ddsSE)) \u003e= 10 ddsSE \u003c- ddsSE[keep,] ddsSE \u003c- DESeq(ddsSE) resultsNames(ddsSE) res \u003c- results(ddsSE, name = \"sample_type_Primary.Tumor_vs_Solid.Tissue.Normal\") dea \u003c- as.data.frame(res) %\u003e% arrange(padj) head(dea) # baseMean log2FoldChange lfcSE stat pvalue padj # MSMO1 7585.3117 -3.748458 0.2108963 -17.77394 1.125105e-70 1.012094e-66 # RMDN2 396.7564 -2.929007 0.1648043 -17.77264 1.151350e-70 1.012094e-66 # GCDH 2272.8992 -3.302565 0.1863391 -17.72342 2.765888e-70 1.620903e-66 # TCAIM 1239.0498 -2.235399 0.1303361 -17.15104 6.174288e-66 2.713754e-62 # KCNN2 360.4370 -5.504733 0.3234314 -17.01979 5.858475e-65 2.059957e-61 # RCL1 2371.8288 -3.917907 0.2368974 -16.53842 1.940540e-61 5.686107e-58 ##(2)é…å¯¹å·®å¼‚åˆ†æ data_pair = data[,data$barcode %in% TCGAquery_MatchedCoupledSampleTypes(data$barcode, c(\"NT\",\"TP\"))] dim(data_pair) data_pair$pair = substr(data_pair$barcode, 1, 12) table(data_pair$pair) # TCGA-W5-AA2I TCGA-W5-AA2Q TCGA-W5-AA2U TCGA-W5-AA2X TCGA-W5-AA30 TCGA-W5-AA31 TCGA-W5-AA34 TCGA-ZU-A8S4 # 2 2 2 2 2 2 2 2 data_pair$sample_type = factor(data_pair$sample_type, levels = c(\"Solid Tissue Normal\",\"Primary Tumor\")) ddsSE \u003c- DESeqDataSet(data_pair, design = ~ sample_type + pair) keep \u003c- rowSums(counts(ddsSE)) \u003e= 10 ddsSE \u003c- ddsSE[keep,] ddsSE \u003c- DESeq(ddsSE) resultsNames(ddsSE) res \u003c- results(ddsSE, name = \"sample_type_Primary.Tumor_vs_Solid.Tissue.Normal\") dea_p \u003c- as.data.frame(res) %\u003e% arrange(padj) head(dea_p) # baseMean log2FoldChange lfcSE stat pvalue padj # PCK2 22698.702 -5.653596 0.1854224 -30.49037 3.496861e-204 5.903052e-200 # GRHPR 17379.001 -3.574709 0.1552370 -23.02743 2.476640e-117 2.090408e-113 # KDM8 1614.739 -4.555841 0.1989383 -22.90077 4.564515e-116 2.568453e-112 # CTH 2367.036 -5.254399 0.2324287 -22.60650 3.740448e-113 1.578563e-109 # ADI1 18162.322 -3.714172 0.1659180 -22.38559 5.438272e-111 1.836069e-107 # PBLD 8315.323 -5.683141 0.2593493 -21.91308 1.949637e-106 5.485303e-103 å››ã€å…³äºç—…äººçš„ä¸´åºŠæ•°æ®ä¸è‚¿ç˜¤åˆ†å‹ 1ã€è·å–ç—…äººçš„ä¸´åºŠæ•°æ® å¦‚ä¸Šåœ¨GDCprepare()è¿‡ç¨‹ä¸­ï¼Œä¼šè‡ªåŠ¨æ³¨é‡Šç—…äººæ ·æœ¬çš„ä¸´åºŠä¿¡æ¯ã€‚ æˆ‘ä»¬ä¹Ÿå¯ä»¥é¢„å…ˆå•ç‹¬ä¸‹è½½æ¯ä¸ªç—…äººçš„ä¸´åºŠæ•°æ®ï¼Œä»¥ä¾›å‚è€ƒã€‚ æ–¹æ³•ä¸€ï¼šGDCquery() pipeline 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 query \u003c- GDCquery(project = \"TCGA-ACC\", data.category = \"Clinical\", data.type = \"Clinical Supplement\", data.format = \"BCR Biotab\") GDCdownload(query, files.per.chunk = 20) clinical.BCRtab.all \u003c- GDCprepare(query) grep(\"clinical_\", names(clinical.BCRtab.all), value = T) # [1] \"clinical_drug_brca\" \"clinical_omf_v4.0_brca\" # [3] \"clinical_follow_up_v4.0_brca\" \"clinical_follow_up_v1.5_brca\" # [5] \"clinical_follow_up_v4.0_nte_brca\" \"clinical_patient_brca\" # [7] \"clinical_radiation_brca\" \"clinical_nte_brca\" # [9] \"clinical_follow_up_v2.1_brca\" clinical_patient_brca = as.data.frame(clinical.BCRtab.all$clinical_patient_brca) clinical_patient_brca[1:4,1:4] # bcr_patient_uuid bcr_patient_barcode form_completion_date prospective_collection # 1 bcr_patient_uuid bcr_patient_barcode form_completion_date tissue_prospective_collection_indicator # 2 CDE_ID: CDE_ID:2003301 CDE_ID: CDE_ID:3088492 # 3 6E7D5EC6-A469-467C-B748-237353C23416 TCGA-3C-AAAU 2014-1-13 NO # 4 55262FCB-1B01-4480-B322-36570430C917 TCGA-3C-AALI 2014-7-28 NO æ–¹æ³•äºŒï¼šGDCquery_clinic() æ ¹æ®å®˜æ–¹ä»‹ç»ï¼Œè¿™ä¸ªå‡½æ•°ä¸‹è½½çš„æ˜¯indexed clinical: a refined clinical data that is created using the XML files(æ–¹æ³•ä¸€). è¿™ç§æ–¹æ³•ä¸‹è½½é€Ÿåº¦è¾ƒå¿«ï¼Œå»ºè®®ä¼˜å…ˆä½¿ç”¨ã€‚å¦‚æœæ²¡æœ‰æƒ³è¦çš„ä¿¡æ¯ï¼Œå†ä½¿ç”¨æ–¹æ³•ä¸€ã€‚ 1 2 3 4 5 6 7 8 clinical \u003c- GDCquery_clinic(project = \"TCGA-BRCA\", type = \"clinical\") clinical \u003c- GDCquery_clinic(project = \"TCGA-BRCA\", type = \"clinical\") clinical[1:4,1:4] # submitter_id synchronous_malignancy ajcc_pathologic_stage tumor_stage # 1 TCGA-E2-A14U No Stage I stage i # 2 TCGA-E9-A1RC No Stage IIIC stage iiic # 3 TCGA-D8-A1J9 No Stage IA stage ia # 4 TCGA-E2-A14P No Stage IIIC stage iiic 2ã€è·å–ç—…äººçš„è‚¿ç˜¤åˆ†å‹ PanCancerAtlas_subtypes() The columns â€œSubtype_Selectedâ€ was selected as most prominent subtype classification (from the other columns) 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 subtypes \u003c- PanCancerAtlas_subtypes() dim(subtypes) #[1] 7734 10 table(subtypes$cancer.type) # ACC AML BLCA BRCA COAD ESCA GBM HNSC KICH KIRC KIRP LGG LIHC LUAD LUSC OVCA PCPG # 91 187 129 1218 341 169 606 279 66 442 161 516 196 230 178 489 178 # PRAD READ SKCM STAD THCA UCEC UCS # 333 118 333 383 496 538 57 head(as.data.frame(subtypes)) # pan.samplesID cancer.type Subtype_mRNA Subtype_DNAmeth Subtype_protein Subtype_miRNA Subtype_CNA Subtype_Integrative Subtype_other Subtype_Selected # 1 TCGA-OR-A5J1 ACC steroid-phenotype-high+proliferation CIMP-high NA miRNA_1 Quiet COC3 C1A ACC.CIMP-high # 2 TCGA-OR-A5J2 ACC steroid-phenotype-high+proliferation CIMP-low 1 miRNA_1 Noisy COC3 C1A ACC.CIMP-low # 3 TCGA-OR-A5J3 ACC steroid-phenotype-high CIMP-intermediate 3 miRNA_6 Chromosomal COC2 C1A ACC.CIMP-intermediate # 4 TCGA-OR-A5J4 ACC CIMP-high NA miRNA_6 Chromosomal ACC.CIMP-high # 5 TCGA-OR-A5J5 ACC steroid-phenotype-high CIMP-intermediate NA miRNA_2 Chromosomal COC2 C1A ACC.CIMP-intermediate # 6 TCGA-OR-A5J6 ACC steroid-phenotype-low CIMP-low 2 miRNA_1 Noisy COC1 C1B ACC.CIMP-low TCGAquery_subtype() These subtypes will be automatically added in the summarizedExperiment object through GDCprepare. But you can also use the TCGAquery_subtype function to retrieve this information. 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 brca.subtype \u003c- TCGAquery_subtype(tumor = \"brca\") t(brca.subtype[1,]) # [,1] # patient \"TCGA-3C-AAAU\" # Tumor.Type \"BRCA\" # Included_in_previous_marker_papers \"NO\" # vital_status \"Alive\" # days_to_birth \"-20211\" # days_to_death \"NA\" # days_to_last_followup \"4047\" # age_at_initial_pathologic_diagnosis \"55\" # pathologic_stage \"NA\" # Tumor_Grade \"NA\" # BRCA_Pathology \"NA\" # BRCA_Subtype_PAM50 \"LumA\" # MSI_status \"NA\" # HPV_Status \"NA\" # tobacco_smoking_history \"NA\" # CNV Clusters \"C6\" # Mutation Clusters \"C7\" # DNA.Methylation Clusters \"C1\" # mRNA Clusters \"C1\" # miRNA Clusters \"C3\" # lncRNA Clusters \"NA\" # Protein Clusters \"NA\" # PARADIGM Clusters \"C5\" # Pan-Gyn Clusters \"NA\" GDCquery_Maf()å‡½æ•°å¯ä»¥æ”¯æŒä¸‹è½½çªå˜æ•°æ®ï¼Œè¿™é‡Œå°±æš‚æ—¶ä¸å­¦ä¹ äº†ã€‚ä¹‹åæœ‰æœºä¼šå†äº†è§£ä¸€ä¸‹ã€‚\nè¡¥å…… 1 ã€TCGAç®€ç§°å¯¹åº”å…¨ç§°ä»¥åŠä¸­æ–‡å Study Abbreviation Study Name ä¸­æ–‡å ACC Adrenocortical carcinoma è‚¾ä¸Šè…ºçš®è´¨ç™Œ BLCA Bladder Urothelial Carcinoma è†€èƒ±å°¿è·¯ä¸Šçš®ç™Œ BRCA Breast invasive carcinoma æµ¸æ¶¦æ€§ä¹³è…ºç™Œ CESC Cervical squamous cell carcinoma and endocervical adenocarcinoma å®«é¢ˆé³çŠ¶ç»†èƒç™Œå’Œå®«é¢ˆå†…è…ºç™Œ CHOL Cholangiocarcinoma èƒ†ç®¡ç™Œ COAD Colon adenocarcinoma ç»“è‚ è…ºç™Œ DLBC Lymphoid Neoplasm Diffuse Large B-cell Lymphoma æ·‹å·´æ ·è‚¿ç˜¤å¼¥æ¼«å¤§bç»†èƒæ·‹å·´ç˜¤ ESCA Esophageal carcinoma é£Ÿç®¡ç™Œç™Œ GBM Glioblastoma multiforme å¤šå½¢æ€§æˆèƒ¶è´¨ç»†èƒç˜¤ HNSC Head and Neck squamous cell carcinoma å¤´é¢ˆéƒ¨é³çŠ¶ç»†èƒç™Œ KICH Kidney Chromophobe è‚¾å«Œè‰²ç»†èƒç™Œ KIRC Kidney renal clear cell carcinoma è‚¾é€æ˜ç»†èƒç™Œ KIRP Kidney renal papillary cell carcinoma è‚¾ä¹³å¤´çŠ¶ç»†èƒç™Œ LAML Acute Myeloid Leukemia æ€¥æ€§é«“ç³»ç™½è¡€ç—… LGG Brain Lower Grade Glioma è„‘ä½çº§åˆ«èƒ¶è´¨ç˜¤ LIHC Liver hepatocellular carcinoma è‚è„è‚ç»†èƒç™Œ LUAD Lung adenocarcinoma è‚ºè…ºç™Œ LUSC Lung squamous cell carcinoma è‚ºé³ç™Œ MESO Mesothelioma é—´çš®ç˜¤ OV Ovarian serous cystadenocarcinoma åµå·¢æµ†æ¶²æ€§å›Šè…ºç™Œ PAAD Pancreatic adenocarcinoma èƒ°è…ºè…ºç™Œ PCPG Pheochromocytoma and Paraganglioma å—œé“¬ç»†èƒç˜¤å’Œå‰¯ç¥ç»èŠ‚ç˜¤ PRAD Prostate adenocarcinoma å‰åˆ—è…ºè…ºç™Œ READ Rectum adenocarcinoma ç›´è‚ è…ºç™Œ SARC Sarcoma è‚‰ç˜¤ SKCM Skin Cutaneous Melanoma çš®è‚¤çš®è‚¤é»‘è‰²ç´ ç˜¤ STAD Stomach adenocarcinoma èƒƒè…ºç™Œ TGCT Testicular Germ Cell Tumors ç¾ä¸¸ç”Ÿæ®–ç»†èƒè‚¿ç˜¤ THCA Thyroid carcinoma ç”²çŠ¶è…ºç™Œ THYM Thymoma èƒ¸è…ºç˜¤ UCEC Uterine Corpus Endometrial Carcinoma å­å®«å†…è†œç™Œ UCS Uterine Carcinosarcoma å­å®«ç™Œè‚‰ç˜¤ UVM Uveal Melanoma è‘¡è„è†œé»‘è‰²ç´ ç˜¤ 2 ã€GDCæ‰€æœ‰æ•°æ®ç±»å‹ 2.1 GDC harmonized database Data.category Data.type Workflow.Type Platform Transcriptome Profiling Gene Expression Quantification HTSeq - Counts Transcriptome Profiling Gene Expression Quantification HTSeq - FPKM Transcriptome Profiling Gene Expression Quantification HTSeq - FPKM-UQ Transcriptome Profiling Gene Expression Quantification STAR - Counts Transcriptome Profiling Isoform Expression Quantification - Transcriptome Profiling miRNA Expression Quantification - Transcriptome Profiling Splice Junction Quantification Copy number variation Copy Number Segment Copy number variation Masked Copy Number Segment Copy number variation Gene Level Copy Number Scores Simple Nucleotide Variation Masked Somatic Mutation MuSE Variant Aggregation and Masking Simple Nucleotide Variation Masked Somatic Mutation MuTect2 Variant Aggregation and Masking Simple Nucleotide Variation Masked Somatic Mutation SomaticSniper Variant Aggregation and Masking Simple Nucleotide Variation Masked Somatic Mutation VarScan2 Variant Aggregation and Masking Raw Sequencing Data - Biospecimen Slide Image Biospecimen Biospecimen Supplement Clinical - DNA Methylation Methylation Beta Value Illumina Human Methylation 450 DNA Methylation Methylation Beta Value Illumina Human Methylation 27 2.2 GDC Legacy Archive Data.category Data.type Platform file.type Copy number variation - Affymetrix SNP Array 6.0 nocnv_hg18.seg Copy number variation - Affymetrix SNP Array 6.0 hg18.seg Copy number variation - Affymetrix SNP Array 6.0 nocnv_hg19.seg Copy number variation - Affymetrix SNP Array 6.0 hg19.seg Copy number variation - Illumina HiSeq - Simple nucleotide variation Simple somatic mutation Raw sequencing data Biospecimen Clinical Protein expression MDA RPPA Core - Gene expression Gene expression quantification Illumina HiSeq normalized_results Gene expression Gene expression quantification Illumina HiSeq results Gene expression Gene expression quantification HT_HG-U133A - Gene expression Gene expression quantification AgilentG4502A_07_2 - Gene expression Gene expression quantification AgilentG4502A_07_1 - Gene expression Gene expression quantification HuEx-1_0-st-v2 FIRMA.txt Gene expression Gene expression quantification gene.txt Gene expression Isoform expression quantification - - Gene expression miRNA gene quantification - hg19.mirna Gene expression miRNA gene quantification hg19.mirbase20 Gene expression miRNA gene quantification mirna Gene expression Exon junction quantification - - Gene expression Exon quantification - - Gene expression miRNA isoform quantification - hg19.isoform Gene expression miRNA isoform quantification - isoform DNA methylation Illumina Human Methylation 450 Not used DNA methylation Illumina Human Methylation 27 Not used DNA methylation Illumina DNA Methylation OMA003 CPI Not used DNA methylation Illumina DNA Methylation OMA002 CPI Not used DNA methylation Illumina Hi Seq DNA methylation Bisulfite sequence alignment DNA methylation Methylation percentage DNA methylation Aligned reads Raw microarray data Raw intensities Illumina Human Methylation 450 idat Raw Microarray Data Raw intensities Illumina Human Methylation 27 idat Structural Rearrangement Other ",
  "wordCount" : "4785",
  "inLanguage": "en",
  "datePublished": "2022-04-30T00:00:00Z",
  "dateModified": "2022-05-06T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/046tcgabiolinks%E5%8C%85-tcga%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD%E4%B8%8E%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="ä¸»é¡µ">
                    <span>ä¸»é¡µ</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="åˆ†ç±»">
                    <span>åˆ†ç±»</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="æ ‡ç­¾">
                    <span>æ ‡ç­¾</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="å½’æ¡£">
                    <span>å½’æ¡£</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="å…³äº">
                    <span>å…³äº</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="æœç´¢ (Alt &#43; /)" accesskey=/>
                    <span>æœç´¢</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/">åˆ†ç±»</a>&nbsp;Â»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">ğŸ“– ç”Ÿä¿¡æ•°æ®åˆ†æ--åˆ†ææµç¨‹ï¼Œå·¥å…·åŒ…ç­‰</a></div>
    <h1 class="post-title">
      TCGAbiolinksåŒ…-TCGAæ•°æ®ä¸‹è½½ä¸å·®å¼‚åˆ†æ
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-30 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-30&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-05-06&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;4785&amp;nbsp;|&amp;nbsp;10 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#%e4%b8%80%e6%9f%a5%e6%89%betcga%e6%95%b0%e6%8d%ae" aria-label="ä¸€ã€æŸ¥æ‰¾TCGAæ•°æ®">ä¸€ã€æŸ¥æ‰¾TCGAæ•°æ®</a><ul>
                            
                    <li>
                        <a href="#1-%e9%80%89%e6%8b%a9%e8%82%bf%e7%98%a4%e7%b1%bb%e5%9e%8b" aria-label="1 é€‰æ‹©è‚¿ç˜¤ç±»å‹">1 é€‰æ‹©è‚¿ç˜¤ç±»å‹</a></li>
                    <li>
                        <a href="#2-hg19hg38" aria-label="2 hg19/hg38">2 hg19/hg38</a></li>
                    <li>
                        <a href="#3-%e4%b8%8b%e8%bd%bd%e6%95%b0%e6%8d%ae%e7%b1%bb%e5%9e%8b" aria-label="3 ä¸‹è½½æ•°æ®ç±»å‹">3 ä¸‹è½½æ•°æ®ç±»å‹</a></li>
                    <li>
                        <a href="#4-%e6%a0%b7%e6%9c%ac%e6%a0%87%e7%ad%bebarcode" aria-label="4 æ ·æœ¬æ ‡ç­¾Barcode">4 æ ·æœ¬æ ‡ç­¾Barcode</a></li></ul>
                    </li>
                    <li>
                        <a href="#%e4%ba%8c%e4%bb%8e%e6%95%b0%e6%8d%ae%e4%b8%8b%e8%bd%bd%e5%88%b0%e5%b7%ae%e5%bc%82%e5%88%86%e6%9e%90" aria-label="äºŒã€ä»æ•°æ®ä¸‹è½½åˆ°å·®å¼‚åˆ†æ">äºŒã€ä»æ•°æ®ä¸‹è½½åˆ°å·®å¼‚åˆ†æ</a><ul>
                            
                    <li>
                        <a href="#1-legacyhg19" aria-label="1 Legacy(hg19)">1 Legacy(hg19)</a><ul>
                            
                    <li>
                        <a href="#11-%e4%b8%8b%e8%bd%bd%e6%95%b0%e6%8d%ae" aria-label="1.1 ä¸‹è½½æ•°æ®">1.1 ä¸‹è½½æ•°æ®</a></li>
                    <li>
                        <a href="#12-%e6%8e%a2%e7%b4%a2%e6%95%b0%e6%8d%ae" aria-label="1.2 æ¢ç´¢æ•°æ®">1.2 æ¢ç´¢æ•°æ®</a></li>
                    <li>
                        <a href="#13-%e6%95%b4%e7%90%86%e6%95%b0%e6%8d%ae" aria-label="1.3 æ•´ç†æ•°æ®">1.3 æ•´ç†æ•°æ®</a></li>
                    <li>
                        <a href="#14-%e5%b7%ae%e5%bc%82%e5%88%86%e6%9e%90" aria-label="1.4 å·®å¼‚åˆ†æ">1.4 å·®å¼‚åˆ†æ</a></li></ul>
                    </li>
                    <li>
                        <a href="#2-harmonizedhg38" aria-label="2 Harmonized(hg38)">2 Harmonized(hg38)</a><ul>
                            
                    <li>
                        <a href="#21-%e4%b8%8b%e8%bd%bd%e6%95%b0%e6%8d%ae" aria-label="2.1 ä¸‹è½½æ•°æ®">2.1 ä¸‹è½½æ•°æ®</a></li>
                    <li>
                        <a href="#22-%e6%8e%a2%e7%b4%a2%e6%95%b0%e6%8d%ae" aria-label="2.2 æ¢ç´¢æ•°æ®">2.2 æ¢ç´¢æ•°æ®</a></li>
                    <li>
                        <a href="#23-%e6%95%b4%e7%90%86%e6%95%b0%e6%8d%ae" aria-label="2.3 æ•´ç†æ•°æ®">2.3 æ•´ç†æ•°æ®</a></li>
                    <li>
                        <a href="#24-%e5%b7%ae%e5%bc%82%e5%88%86%e6%9e%90" aria-label="2.4 å·®å¼‚åˆ†æ">2.4 å·®å¼‚åˆ†æ</a><ul>
                            
                    <li>
                        <a href="#1-edger" aria-label="(1) edgeR">(1) edgeR</a></li>
                    <li>
                        <a href="#2-deseq2" aria-label="(2) Deseq2">(2) Deseq2</a></li></ul>
                    </li></ul>
                    </li></ul>
                    </li>
                    <li>
                        <a href="#%e5%9b%9b%e5%85%b3%e4%ba%8e%e7%97%85%e4%ba%ba%e7%9a%84%e4%b8%b4%e5%ba%8a%e6%95%b0%e6%8d%ae%e4%b8%8e%e8%82%bf%e7%98%a4%e5%88%86%e5%9e%8b" aria-label="å››ã€å…³äºç—…äººçš„ä¸´åºŠæ•°æ®ä¸è‚¿ç˜¤åˆ†å‹">å››ã€å…³äºç—…äººçš„ä¸´åºŠæ•°æ®ä¸è‚¿ç˜¤åˆ†å‹</a><ul>
                            
                    <li>
                        <a href="#1%e8%8e%b7%e5%8f%96%e7%97%85%e4%ba%ba%e7%9a%84%e4%b8%b4%e5%ba%8a%e6%95%b0%e6%8d%ae" aria-label="1ã€è·å–ç—…äººçš„ä¸´åºŠæ•°æ®">1ã€è·å–ç—…äººçš„ä¸´åºŠæ•°æ®</a><ul>
                            <ul>
                            <ul>
                            
                    <li>
                        <a href="#%e6%96%b9%e6%b3%95%e4%b8%80gdcquery-pipeline" aria-label="æ–¹æ³•ä¸€ï¼šGDCquery() pipeline">æ–¹æ³•ä¸€ï¼šGDCquery() pipeline</a></li>
                    <li>
                        <a href="#%e6%96%b9%e6%b3%95%e4%ba%8cgdcquery_clinic" aria-label="æ–¹æ³•äºŒï¼šGDCquery_clinic()">æ–¹æ³•äºŒï¼šGDCquery_clinic()</a></li></ul>
                        </ul>
                        </ul>
                    </li>
                    <li>
                        <a href="#2%e8%8e%b7%e5%8f%96%e7%97%85%e4%ba%ba%e7%9a%84%e8%82%bf%e7%98%a4%e5%88%86%e5%9e%8b" aria-label="2ã€è·å–ç—…äººçš„è‚¿ç˜¤åˆ†å‹">2ã€è·å–ç—…äººçš„è‚¿ç˜¤åˆ†å‹</a></li></ul>
                    </li>
                    <li>
                        <a href="#%e8%a1%a5%e5%85%85" aria-label="è¡¥å……">è¡¥å……</a><ul>
                            
                    <li>
                        <a href="#1-tcga%e7%ae%80%e7%a7%b0%e5%af%b9%e5%ba%94%e5%85%a8%e7%a7%b0%e4%bb%a5%e5%8f%8a%e4%b8%ad%e6%96%87%e5%90%8d" aria-label="1 ã€TCGAç®€ç§°å¯¹åº”å…¨ç§°ä»¥åŠä¸­æ–‡å">1 ã€TCGAç®€ç§°å¯¹åº”å…¨ç§°ä»¥åŠä¸­æ–‡å</a></li>
                    <li>
                        <a href="#2-gdc%e6%89%80%e6%9c%89%e6%95%b0%e6%8d%ae%e7%b1%bb%e5%9e%8b" aria-label="2 ã€GDCæ‰€æœ‰æ•°æ®ç±»å‹">2 ã€GDCæ‰€æœ‰æ•°æ®ç±»å‹</a><ul>
                            
                    <li>
                        <a href="#21--gdc-harmonized-database" aria-label="2.1  GDC harmonized database">2.1  GDC harmonized database</a></li>
                    <li>
                        <a href="#22--gdc-legacy-archive" aria-label="2.2  GDC Legacy Archive">2.2  GDC Legacy Archive</a>
                    </li>
                </ul>
                </li>
                </ul>
                </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><p><code>TCGAbiolinks</code>åŒ…æ˜¯ä¸€ç«™å¼åˆ†æTCGAæ•°æ®çš„RåŒ…å·¥å…·ï¼Œå®ƒé›†æˆäº†TCGAæ•°æ®ä¸‹è½½ã€åˆ†æã€å¯è§†åŒ–çš„å…¨éƒ¨æµç¨‹ã€‚æ­¤æ¬¡ç³»åˆ—ç¬”è®°ä¸»è¦è·Ÿç€ TCGAbiolinkså¸®åŠ©æ–‡æ¡£é‡æ–°å­¦ä¹ ä¸‹TCGAæ•°æ®æŒ–æ˜æµç¨‹ã€‚</p>
<blockquote>
<ul>
<li>
<p>å®˜æ–¹æ–‡æ¡£ï¼šhttps://bioconductor.org/packages/release/bioc/vignettes/TCGAbiolinks/inst/doc/index.html</p>
</li>
<li>
<p>æ–‡çŒ®ï¼šTCGAbiolinks: an R/Bioconductor package for integrative analysis of TCGA data <a href="https://pubmed.ncbi.nlm.nih.gov/26704973/">https://pubmed.ncbi.nlm.nih.gov/26704973/</a></p>
</li>
<li>
<p>å®‰è£…åŒ…</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">if</span> (!requireNamespace(<span style="color:#0ff;font-weight:bold">&#34;BiocManager&#34;</span>, quietly = <span style="color:#fff;font-weight:bold">TRUE</span>))
</span></span><span style="display:flex;"><span>    install.packages(<span style="color:#0ff;font-weight:bold">&#34;BiocManager&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#From github</span>
</span></span><span style="display:flex;"><span>BiocManager::install(<span style="color:#0ff;font-weight:bold">&#34;BioinformaticsFMRP/TCGAbiolinksGUI.data&#34;</span>)
</span></span><span style="display:flex;"><span>BiocManager::install(<span style="color:#0ff;font-weight:bold">&#34;BioinformaticsFMRP/TCGAbiolinks&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#From Bioconductor</span>
</span></span><span style="display:flex;"><span>BiocManager::install(<span style="color:#0ff;font-weight:bold">&#34;TCGAbiolinks&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div></li>
</ul></blockquote>
<h1 id="ä¸€æŸ¥æ‰¾tcgaæ•°æ®">ä¸€ã€æŸ¥æ‰¾TCGAæ•°æ®<a hidden class="anchor" aria-hidden="true" href="#ä¸€æŸ¥æ‰¾tcgaæ•°æ®">#</a></h1>
<ul>
<li>ä¸»è¦é€šè¿‡<code>GDCquery()</code>å‡½æ•°æŸ¥æ‰¾æ•°æ®</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(TCGAbiolinks)
</span></span><span style="display:flex;"><span>GDCquery(
</span></span><span style="display:flex;"><span>  project,
</span></span><span style="display:flex;"><span>  data.category,
</span></span><span style="display:flex;"><span>  data.type,
</span></span><span style="display:flex;"><span>  workflow.type,
</span></span><span style="display:flex;"><span>  legacy = <span style="color:#fff;font-weight:bold">FALSE</span>,
</span></span><span style="display:flex;"><span>  access,
</span></span><span style="display:flex;"><span>  platform,
</span></span><span style="display:flex;"><span>  file.type,
</span></span><span style="display:flex;"><span>  barcode,
</span></span><span style="display:flex;"><span>  data.format,
</span></span><span style="display:flex;"><span>  experimental.strategy,
</span></span><span style="display:flex;"><span>  sample.type
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="1-é€‰æ‹©è‚¿ç˜¤ç±»å‹">1 é€‰æ‹©è‚¿ç˜¤ç±»å‹<a hidden class="anchor" aria-hidden="true" href="#1-é€‰æ‹©è‚¿ç˜¤ç±»å‹">#</a></h2>
<ul>
<li><code>project</code>å‚æ•°ï¼šæŒ‡å®šä¸€ä¸ªæˆ–å¤šä¸ªæ„Ÿå…´è¶£çš„TCGAé¡¹ç›®å</li>
<li>å¦‚ä¸‹ä»£ç æ‰€ç¤ºï¼Œä¾›åŒ…æ‹¬33ç§TCGAç™Œç—‡ç±»å‹ã€‚ç™Œç—‡ç®€ç§°å¯¹åº”å…¨ç§°åŠä¸­æ–‡åè§æ–‡æœ«ã€‚</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>projects = TCGAbiolinks:::getGDCprojects()$project_id
</span></span><span style="display:flex;"><span>TCGAs = grep(<span style="color:#0ff;font-weight:bold">&#34;TCGA&#34;</span>, projects, value = <span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span>sort(TCGAs)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;TCGA-ACC&#34;  &#34;TCGA-BLCA&#34; &#34;TCGA-BRCA&#34; &#34;TCGA-CESC&#34; &#34;TCGA-CHOL&#34; &#34;TCGA-COAD&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [7] &#34;TCGA-DLBC&#34; &#34;TCGA-ESCA&#34; &#34;TCGA-GBM&#34;  &#34;TCGA-HNSC&#34; &#34;TCGA-KICH&#34; &#34;TCGA-KIRC&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [13] &#34;TCGA-KIRP&#34; &#34;TCGA-LAML&#34; &#34;TCGA-LGG&#34;  &#34;TCGA-LIHC&#34; &#34;TCGA-LUAD&#34; &#34;TCGA-LUSC&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [19] &#34;TCGA-MESO&#34; &#34;TCGA-OV&#34;   &#34;TCGA-PAAD&#34; &#34;TCGA-PCPG&#34; &#34;TCGA-PRAD&#34; &#34;TCGA-READ&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [25] &#34;TCGA-SARC&#34; &#34;TCGA-SKCM&#34; &#34;TCGA-STAD&#34; &#34;TCGA-TGCT&#34; &#34;TCGA-THCA&#34; &#34;TCGA-THYM&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [31] &#34;TCGA-UCEC&#34; &#34;TCGA-UCS&#34;  &#34;TCGA-UVM&#34; </span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="2-hg19hg38">2 hg19/hg38<a hidden class="anchor" aria-hidden="true" href="#2-hg19hg38">#</a></h2>
<ul>
<li>ä¸»è¦æ ¹æ®å‚è€ƒåŸºå› ç»„çš„ä¸åŒï¼ŒåŒ…å«ä¸¤å¥—æ•°æ®ï¼šGDC Legacy Archiveã€ä¸»è¦GRCh37 (hg19)ã€‘ï¼ŒGDC harmonized databaseã€GRCh38 (hg38)ã€‘</li>
<li>é€šè¿‡è®¾ç½®å‚æ•°<code>legacy</code> ï¼Œé»˜è®¤ä¸ºFALSE(hg38)ï¼›TRUEåˆ™è¡¨ç¤ºä½¿ç”¨hg19å‚è€ƒåŸºå› ç»„çš„æµ‹åºæ•°æ®ã€‚</li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/image-20220427092526157.png" alt="image-20220427092526157"  />
</p>
<h2 id="3-ä¸‹è½½æ•°æ®ç±»å‹">3 ä¸‹è½½æ•°æ®ç±»å‹<a hidden class="anchor" aria-hidden="true" href="#3-ä¸‹è½½æ•°æ®ç±»å‹">#</a></h2>
<p>åŸºäºä¸Šè¿°çš„å‚æ•°ï¼Œæˆ‘ä»¬å¯ä»¥è®¾ç½®å¦‚ä¸‹å‚æ•°ï¼Œäº¤ä»£æˆ‘ä»¬çš„ç›®æ ‡æ•°æ®ç±»å‹</p>
<ul>
<li><code>data.category = </code> æŒ‡å®šä¸‹è½½ä»€ä¹ˆç±»å‹çš„æ•°æ®ï¼šå¦‚è½¬å½•ç»„æ•°æ®ã€ä¸´åºŠæ•°æ®&hellip;.</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1) GDC harmonized databaseã€GRCh38 (hg38)ã€‘</span>
</span></span><span style="display:flex;"><span>TCGAbiolinks:::getProjectSummary(<span style="color:#0ff;font-weight:bold">&#34;TCGA-BRCA&#34;</span>, legacy = <span style="color:#fff;font-weight:bold">FALSE</span>)$data_categories$data_category
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;Structural Variation&#34;        &#34;Simple Nucleotide Variation&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [3] &#34;Copy Number Variation&#34;       &#34;Transcriptome Profiling&#34;    </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;DNA Methylation&#34;             &#34;Sequencing Reads&#34;           </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [7] &#34;Biospecimen&#34;                 &#34;Clinical&#34;                   </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [9] &#34;Proteome Profiling&#34;  </span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) GDC Legacy Archiveã€ä¸»è¦GRCh37 (hg19)ã€‘</span>
</span></span><span style="display:flex;"><span>TCGAbiolinks:::getProjectSummary(<span style="color:#0ff;font-weight:bold">&#34;TCGA-BRCA&#34;</span>, legacy = <span style="color:#fff;font-weight:bold">TRUE</span>)$data_categories$data_category
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;Protein expression&#34;          &#34;Copy number variation&#34;      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [3] &#34;Biospecimen&#34;                 &#34;Simple nucleotide variation&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;Gene expression&#34;             &#34;Raw microarray data&#34;        </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [7] &#34;DNA methylation&#34;             &#34;Clinical&#34;                   </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [9] &#34;Raw sequencing data&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li><code>data.type = </code> æ›´åŠ ç»†èŠ‚çš„æ•°æ®ç±»å‹é€‰æ‹©ï¼Œä¾‹å¦‚<strong>harmonized</strong>çš„<code>Transcriptome Profiling</code>æœ‰<code>  Gene Expression Quantification</code>, <code>Isoform Expression Quantification</code>, <code>miRNA Expression Quantification</code>, <code>  Splice Junction Quantification</code>4ç§æ•°æ®ã€‚è€Œæˆ‘ä»¬ä¸€èˆ¬éœ€è¦çš„æ˜¯<code>Gene Expression Quantification</code>ã€‚</li>
<li><code>workflow.type = </code> åŒä¸€ä¸ªæµ‹åºæ•°æ®å¯èƒ½æœ‰ä¸åŒçš„pipelineå¤„ç†æµç¨‹ï¼Œå¦‚æœæœ‰ä¸åŒçš„å¤„ç†æµç¨‹ï¼Œéœ€è¦é€‰æ‹©ã€‚æ®æœ€æ–°çš„TCGAç‰ˆæœ¬ï¼Œä»…æä¾›STAR - Countsã€‚</li>
<li><code>platform = </code> æµ‹åºå¹³å°ï¼ˆoptionalï¼‰</li>
<li><code>file.type = </code> å…·ä½“çš„æ•°æ®æ–‡ä»¶ï¼ˆoptional, for legacyï¼‰
å¦‚æœä¸çŸ¥é“ç›®æ ‡æ•°æ®çš„ä¸Šè¿°ä¿¡æ¯ï¼Œå¯ä»¥å‚è€ƒä¸‹é¢çš„æ¦‚è¿°</li>
</ul>
<blockquote>
<p>å…·ä½“å‚è€ƒæ–‡æœ«ç¬¬äºŒç‚¹ã€‚</p></blockquote>
<h2 id="4-æ ·æœ¬æ ‡ç­¾barcode">4 æ ·æœ¬æ ‡ç­¾Barcode<a hidden class="anchor" aria-hidden="true" href="#4-æ ·æœ¬æ ‡ç­¾barcode">#</a></h2>
<p>å®Œæ•´çš„barcodeï¼šå½¢å¦‚ TCGA-G4-6317-02A-11D-2064-05ï¼Œè¿™ä¸ªæ ‡ç­¾åŒ…å«äº†ä»ç—…äººæ¥æºåˆ°æµ‹åºè¿‡ç¨‹ã€åˆ†æçš„æ‰€æœ‰ä¿¡æ¯ï¼Œå¦‚ä¸‹å›¾æ‰€ç¤ºæ¯”è¾ƒé‡è¦çš„æ˜¯<code>Participant</code>ã€<code>Sample</code>  ã€<code>Portion</code>ä¸‰ä¸ªéƒ¨åˆ†ï¼Œåˆ†åˆ«äº¤ä»£äº†ç—…äººç¼–å·ã€æ ·æœ¬ç±»å‹ã€æµ‹åºç±»å‹
ç—…äººçš„idï¼šå½¢å¦‚ TCGA-G4-6317
æ ·æœ¬æ¥æºçš„idï¼šå½¢å¦‚ TCGA-G4-6317-02
<img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-04e184c062c99649.png" alt=""  />
</p>
<ul>
<li>
<p>å…¶ä¸­æ¯”è¾ƒé‡è¦çš„æ˜¯äº¤ä»£æ ·æœ¬ç±»å‹çš„<code>Sample</code>çš„ä¸¤ä½æ•°ä¿¡æ¯ï¼Œæ˜¯åé¢è¿›è¡Œå·®å¼‚åˆ†æçš„åˆ†ç»„ä¾æ®ã€‚å…·ä½“å¯¹åº”çš„å«ä¹‰å¦‚ä¸‹ã€‚ä¾‹å¦‚<code>01</code>è¡¨ç¤ºç—…äººçš„åŸä½ç˜¤ç»„ç»‡ï¼›<code>11</code>è¡¨ç¤ºæ¥è‡ªç—…äººçš„æ­£å¸¸ç»„ç»‡&hellip;.
<img src="https://raw.githubusercontent.com/lishensuo/images/main/20354525-f58d8cb810877a3e.png" style="zoom: 67%;" /></p>
</li>
<li>
<p>åŸºäºä¸Šè¿°ç†è§£ï¼Œæˆ‘ä»¬ä¹Ÿå¯ä»¥è®¾ç½®<code>sample.type = </code>å‚æ•°æŒ‡å®šä¸‹è½½æ„Ÿå…´è¶£çš„æ ·æœ¬ç±»å‹æ•°æ®ï¼Œä¾‹å¦‚<code>sample.type = &quot;Primary Tumor&quot;</code></p>
</li>
<li>
<p>å¯¹äºç»™å®šçš„TCGA barcodeï¼Œå¯ä»¥åˆ©ç”¨<code>TCGAquery_SampleTypes()</code>æå–å‡ºç›®æ ‡åˆ†ç»„çš„æ ·æœ¬ï¼›<code>TCGAquery_MatchedCoupledSampleTypes()</code>å‡½æ•°å¯ä»¥æå–æ¥è‡ªåŒä¸€ç—…äººçš„é…å¯¹æ ·æœ¬æ•°æ®ã€‚</p>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>query &lt;- GDCquery(project = c(<span style="color:#0ff;font-weight:bold">&#34;TCGA-BRCA&#34;</span>),
</span></span><span style="display:flex;"><span>                  legacy = <span style="color:#fff;font-weight:bold">FALSE</span>, 
</span></span><span style="display:flex;"><span>                  data.category = <span style="color:#0ff;font-weight:bold">&#34;Transcriptome Profiling&#34;</span>,
</span></span><span style="display:flex;"><span>                  data.type = <span style="color:#0ff;font-weight:bold">&#34;Gene Expression Quantification&#34;</span>,
</span></span><span style="display:flex;"><span>                  workflow.type = <span style="color:#0ff;font-weight:bold">&#34;HTSeq - Counts&#34;</span>)
</span></span><span style="display:flex;"><span>dim(getResults(query))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1] 1222   29</span>
</span></span><span style="display:flex;"><span>query_info = getResults(query)
</span></span><span style="display:flex;"><span>TP = TCGAquery_SampleTypes(query_info$sample.submitter_id,<span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>)
</span></span><span style="display:flex;"><span>NT = TCGAquery_SampleTypes(query_info$sample.submitter_id,<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>)
</span></span><span style="display:flex;"><span>query &lt;- GDCquery(project = c(<span style="color:#0ff;font-weight:bold">&#34;TCGA-BRCA&#34;</span>),
</span></span><span style="display:flex;"><span>                  legacy = <span style="color:#fff;font-weight:bold">FALSE</span>, 
</span></span><span style="display:flex;"><span>                  data.category = <span style="color:#0ff;font-weight:bold">&#34;Transcriptome Profiling&#34;</span>,
</span></span><span style="display:flex;"><span>                  data.type = <span style="color:#0ff;font-weight:bold">&#34;Gene Expression Quantification&#34;</span>,
</span></span><span style="display:flex;"><span>                  workflow.type = <span style="color:#0ff;font-weight:bold">&#34;HTSeq - Counts&#34;</span>,
</span></span><span style="display:flex;"><span>                  barcode = c(TP, NT))
</span></span><span style="display:flex;"><span>dim(getResults(query))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1] 1215   29</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>Pair_sample = TCGAquery_MatchedCoupledSampleTypes(query_info$sample.submitter_id,c(<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>))
</span></span><span style="display:flex;"><span>query &lt;- GDCquery(project = c(<span style="color:#0ff;font-weight:bold">&#34;TCGA-BRCA&#34;</span>),
</span></span><span style="display:flex;"><span>                  legacy = <span style="color:#fff;font-weight:bold">FALSE</span>, 
</span></span><span style="display:flex;"><span>                  data.category = <span style="color:#0ff;font-weight:bold">&#34;Transcriptome Profiling&#34;</span>,
</span></span><span style="display:flex;"><span>                  data.type = <span style="color:#0ff;font-weight:bold">&#34;Gene Expression Quantification&#34;</span>,
</span></span><span style="display:flex;"><span>                  workflow.type = <span style="color:#0ff;font-weight:bold">&#34;HTSeq - Counts&#34;</span>,
</span></span><span style="display:flex;"><span>                  barcode = Pair_sample)
</span></span><span style="display:flex;"><span>dim(getResults(query))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1] 229  29</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>å¦‚ä¸Šæ˜¯æŸ¥è¯¢TCGAç›®æ ‡æ•°æ®çš„å‡ ç§å¸¸è§æ ‡å‡†ï¼Œè¿˜æœ‰å‡ ä¸ªå‚æ•°æ²¡æœ‰ä»‹ç»ï¼Œå¯å‚çœ‹å‡½æ•°å¸®åŠ©æ–‡æ¡£ã€‚å¯æ ¹æ®è‡ªå·±çš„ç›®çš„çµæ´»è®¾ç½®ä¸Šè¿°å‚æ•°ã€‚</p></blockquote>
<h1 id="äºŒä»æ•°æ®ä¸‹è½½åˆ°å·®å¼‚åˆ†æ">äºŒã€ä»æ•°æ®ä¸‹è½½åˆ°å·®å¼‚åˆ†æ<a hidden class="anchor" aria-hidden="true" href="#äºŒä»æ•°æ®ä¸‹è½½åˆ°å·®å¼‚åˆ†æ">#</a></h1>
<p>ä»¥èƒ†ç®¡ç™Œä¸ºä¾‹ï¼Œæ¼”ç¤ºä»æ•°æ®ä¸‹è½½åˆ°å·®å¼‚åˆ†æçš„å…¨æµç¨‹ã€‚</p>
<h2 id="1-legacyhg19">1 Legacy(hg19)<a hidden class="anchor" aria-hidden="true" href="#1-legacyhg19">#</a></h2>
<h3 id="11-ä¸‹è½½æ•°æ®">1.1 ä¸‹è½½æ•°æ®<a hidden class="anchor" aria-hidden="true" href="#11-ä¸‹è½½æ•°æ®">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(TCGAbiolinks)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(1) è·å–ä¸‹è½½ç›®æ ‡æ–‡ä»¶</span>
</span></span><span style="display:flex;"><span>query &lt;- GDCquery(project = <span style="color:#0ff;font-weight:bold">&#34;TCGA-CHOL&#34;</span>,
</span></span><span style="display:flex;"><span>                  legacy = <span style="color:#fff;font-weight:bold">TRUE</span>,
</span></span><span style="display:flex;"><span>                  data.category = <span style="color:#0ff;font-weight:bold">&#34;Gene expression&#34;</span>,
</span></span><span style="display:flex;"><span>                  data.type = <span style="color:#0ff;font-weight:bold">&#34;Gene expression quantification&#34;</span>,
</span></span><span style="display:flex;"><span>                  platform = <span style="color:#0ff;font-weight:bold">&#34;Illumina HiSeq&#34;</span>, 
</span></span><span style="display:flex;"><span>                  file.type  = <span style="color:#0ff;font-weight:bold">&#34;results&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) ä¸‹è½½æ•°æ®åˆ°æœ¬åœ°</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># files.per.chunk = 10 è¡¨ç¤ºåˆ†æ‰¹ä¸‹è½½ï¼Œæ¯æ‰¹ä¸‹è½½10ä¸ªç—…äººçš„æ•°æ®ï¼Œå¯é¿å…ä¸­é€”æŠ¥é”™ï¼Œè€Œå‰åŠŸå°½å¼ƒã€‚</span>
</span></span><span style="display:flex;"><span>GDCdownload(query, files.per.chunk = <span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) åŠ è½½æ•°æ®è‡³å½“å‰ç¯å¢ƒ</span>
</span></span><span style="display:flex;"><span>data &lt;- GDCprepare(query, save = <span style="color:#fff;font-weight:bold">T</span>, save.filename = <span style="color:#0ff;font-weight:bold">&#34;CHOL_RNAseq_hg19.rda&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="12-æ¢ç´¢æ•°æ®">1.2 æ¢ç´¢æ•°æ®<a hidden class="anchor" aria-hidden="true" href="#12-æ¢ç´¢æ•°æ®">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(SummarizedExperiment)
</span></span><span style="display:flex;"><span>dim(assay(data))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1] 19947    45</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(1) æŸ¥çœ‹è¡¨è¾¾çŸ©é˜µ</span>
</span></span><span style="display:flex;"><span>assayNames(data)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;raw_count&#34;       &#34;scaled_estimate&#34;</span>
</span></span><span style="display:flex;"><span>assay(data, <span style="color:#0ff;font-weight:bold">&#34;raw_count&#34;</span>)[1:<span style="color:#ff0;font-weight:bold">3</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">3</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        TCGA-W5-AA2I-11A-11R-A41I-07 TCGA-W5-AA2U-11A-11R-A41I-07 TCGA-W5-AA2I-01A-32R-A41I-07</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># A1BG|1                     80062.65                      79948.1                      2519.20</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># A2M|2                      79697.45                     126813.9                      4107.88</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NAT1|9                       364.00                        397.0                       126.00</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) æ ·æœ¬(ä¸´åºŠ)ä¿¡æ¯</span>
</span></span><span style="display:flex;"><span>dim(colData(data))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1]  45 201</span>
</span></span><span style="display:flex;"><span>colData(data)[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) åŸºå› ä¿¡æ¯</span>
</span></span><span style="display:flex;"><span>rowData(data)[1:<span style="color:#ff0;font-weight:bold">6</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">5</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DataFrame with 6 rows and 3 columns</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                   gene_id entrezgene ensembl_gene_id</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                   &lt;character&gt;  &lt;integer&gt;     &lt;character&gt;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># A1BG                 A1BG          1 ENSG00000121410</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># A2M                   A2M          2 ENSG00000175899</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NAT1                 NAT1          9 ENSG00000171428</span>
</span></span><span style="display:flex;"><span>rowRanges(data)
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="13-æ•´ç†æ•°æ®">1.3 æ•´ç†æ•°æ®<a hidden class="anchor" aria-hidden="true" href="#13-æ•´ç†æ•°æ®">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1)æ›´æ”¹è¡¨è¾¾çŸ©é˜µè¡Œå/åŸºå› åæ ¼å¼</span>
</span></span><span style="display:flex;"><span>rownames(data) = stringr::str_split(rownames(data),
</span></span><span style="display:flex;"><span>                                    <span style="color:#0ff;font-weight:bold">&#34;[|]&#34;</span>, simplify = <span style="color:#fff;font-weight:bold">T</span>)[,<span style="color:#ff0;font-weight:bold">1</span>]
</span></span><span style="display:flex;"><span>data = data[!duplicated(rownames(data)),]
</span></span><span style="display:flex;"><span>dim(data)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 19938    45</span>
</span></span><span style="display:flex;"><span>assay(data, <span style="color:#0ff;font-weight:bold">&#34;raw_count&#34;</span>)[1:<span style="color:#ff0;font-weight:bold">3</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">3</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        TCGA-W5-AA2I-11A-11R-A41I-07 TCGA-W5-AA2U-11A-11R-A41I-07 TCGA-W5-AA2I-01A-32R-A41I-07</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># A1BG                     80062.65                      79948.1                      2519.20</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># A2M                      79697.45                     126813.9                      4107.88</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NAT1                       364.00                        397.0                       126.00</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) åˆ†ææ ·æœ¬é—´ç›¸å…³æ€§ï¼Œé‰´å®šç¦»ç¾¤ç‚¹æ ·æœ¬</span>
</span></span><span style="display:flex;"><span>CorOutliers &lt;- TCGAanalyze_Preprocessing(data, cor.cut = <span style="color:#ff0;font-weight:bold">0</span>, 
</span></span><span style="display:flex;"><span>                                         filename=<span style="color:#0ff;font-weight:bold">&#34;CHOL45_hg19_cor.png&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) æ ‡å‡†åŒ–</span>
</span></span><span style="display:flex;"><span>dataNorm &lt;- TCGAanalyze_Normalization(tabDF = data, geneInfo =  geneInfo) <span style="color:#007f7f">#æ²¡æœ‰logå¤„ç†</span>
</span></span><span style="display:flex;"><span>dim(dataNorm)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 17618    45</span>
</span></span><span style="display:flex;"><span>class(dataNorm) 
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;matrix&#34; &#34;array&#34;</span>
</span></span><span style="display:flex;"><span>dataNorm[1:<span style="color:#ff0;font-weight:bold">3</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">3</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#      TCGA-W5-AA2I-11A-11R-A41I-07 TCGA-W5-AA2U-11A-11R-A41I-07 TCGA-W5-AA2I-01A-32R-A41I-07</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># A1BG                        80063                        79948                         2519</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># A2M                         79697                       126814                         4108</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># NAT1                          364                          397                          126</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(4) è¿‡æ»¤ä½è¡¨è¾¾åŸºå› </span>
</span></span><span style="display:flex;"><span>dataFilt &lt;- TCGAanalyze_Filtering(tabDF = dataNorm,
</span></span><span style="display:flex;"><span>                                  method = <span style="color:#0ff;font-weight:bold">&#34;quantile&#34;</span>, 
</span></span><span style="display:flex;"><span>                                  qnt.cut =  <span style="color:#ff0;font-weight:bold">0.05</span>)
</span></span><span style="display:flex;"><span>dim(dataFilt)
</span></span><span style="display:flex;"><span>exp_data = dataFilt
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="14-å·®å¼‚åˆ†æ">1.4 å·®å¼‚åˆ†æ<a hidden class="anchor" aria-hidden="true" href="#14-å·®å¼‚åˆ†æ">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1) ç¡®å®šåˆ†ç»„æ ·æœ¬</span>
</span></span><span style="display:flex;"><span>samplesNT &lt;- TCGAquery_SampleTypes(barcode = colnames(exp_data),
</span></span><span style="display:flex;"><span>                                   typesample = c(<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>))
</span></span><span style="display:flex;"><span>samplesTP &lt;- TCGAquery_SampleTypes(barcode = colnames(exp_data), 
</span></span><span style="display:flex;"><span>                                   typesample = c(<span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) å·®å¼‚åˆ†æ</span>
</span></span><span style="display:flex;"><span>dataDEGs &lt;- TCGAanalyze_DEA(mat1 = exp_data[,samplesNT],
</span></span><span style="display:flex;"><span>                            mat2 = exp_data[,samplesTP],
</span></span><span style="display:flex;"><span>                            Cond1type = <span style="color:#0ff;font-weight:bold">&#34;Normal&#34;</span>,
</span></span><span style="display:flex;"><span>                            Cond2type = <span style="color:#0ff;font-weight:bold">&#34;Tumor&#34;</span>,
</span></span><span style="display:flex;"><span>                            pipeline = <span style="color:#0ff;font-weight:bold">&#34;edgeR&#34;</span>,
</span></span><span style="display:flex;"><span>                            method = <span style="color:#0ff;font-weight:bold">&#34;glmLRT&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) æ•´ç†ç»“æœ</span>
</span></span><span style="display:flex;"><span>DEGs &lt;- TCGAanalyze_LevelTab(dataDEGs,<span style="color:#0ff;font-weight:bold">&#34;Tumor&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;Normal&#34;</span>,
</span></span><span style="display:flex;"><span>                             exp_data[,samplesTP],exp_data[,samplesNT])
</span></span><span style="display:flex;"><span>DEGs = DEGs[order(DEGs$FDR),]
</span></span><span style="display:flex;"><span>DEGs$Direct = ifelse(DEGs$logFC&gt;<span style="color:#ff0;font-weight:bold">0</span>, <span style="color:#0ff;font-weight:bold">&#34;Up&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Down&#34;</span>)
</span></span><span style="display:flex;"><span>head(DEGs)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#          mRNA     logFC          FDR      Delta      Tumor      Normal Direct</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># USH2A   USH2A -5.261122 1.805847e-46  -150.9650   28.69444  1207.88889   Down</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># KCNN2   KCNN2 -4.344681 1.045519e-36  -188.2695   43.33333   893.77778   Down</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># LCAT     LCAT -3.683368 2.344098e-31 -1875.2437  509.11111  6640.22222   Down</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># IQCE     IQCE  4.229960 6.301169e-30 12462.2846 2946.19444   153.33333     Up</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DCXR     DCXR -4.070648 9.023484e-30 -7408.0132 1819.86111 32284.22222   Down</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># PRSS16 PRSS16  6.947385 3.517876e-28  9163.9866 1319.05556    10.22222     Up</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>DEGs = DEGs[order(DEGs$FDR),]
</span></span><span style="display:flex;"><span>DEGs$Direct = ifelse(DEGs$logFC&gt;<span style="color:#ff0;font-weight:bold">0</span>, <span style="color:#0ff;font-weight:bold">&#34;Up&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Down&#34;</span>)
</span></span><span style="display:flex;"><span>table(DEGs$FDR&lt;<span style="color:#ff0;font-weight:bold">0.01</span>,DEGs$Direct)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>DEGs_sig = subset(DEGs, FDR&lt;<span style="color:#ff0;font-weight:bold">0.01</span>)
</span></span><span style="display:flex;"><span>table(DEGs_sig$Direct, abs(DEGs_sig$logFC)&gt;<span style="color:#ff0;font-weight:bold">1</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="2-harmonizedhg38">2 Harmonized(hg38)<a hidden class="anchor" aria-hidden="true" href="#2-harmonizedhg38">#</a></h2>
<h3 id="21-ä¸‹è½½æ•°æ®">2.1 ä¸‹è½½æ•°æ®<a hidden class="anchor" aria-hidden="true" href="#21-ä¸‹è½½æ•°æ®">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(TCGAbiolinks)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(1) è·å–ä¸‹è½½ç›®æ ‡æ–‡ä»¶</span>
</span></span><span style="display:flex;"><span>query &lt;- GDCquery(project = <span style="color:#0ff;font-weight:bold">&#34;TCGA-CHOL&#34;</span>,
</span></span><span style="display:flex;"><span>                  legacy = <span style="color:#fff;font-weight:bold">FALSE</span>,
</span></span><span style="display:flex;"><span>                  data.category = <span style="color:#0ff;font-weight:bold">&#34;Transcriptome Profiling&#34;</span>,
</span></span><span style="display:flex;"><span>                  data.type = <span style="color:#0ff;font-weight:bold">&#34;Gene Expression Quantification&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) ä¸‹è½½æ•°æ®åˆ°æœ¬åœ°</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># files.per.chunk = 10 è¡¨ç¤ºåˆ†æ‰¹ä¸‹è½½ï¼Œæ¯æ‰¹ä¸‹è½½10ä¸ªç—…äººçš„æ•°æ®ï¼Œå¯é¿å…ä¸­é€”æŠ¥é”™ï¼Œè€Œå‰åŠŸå°½å¼ƒã€‚</span>
</span></span><span style="display:flex;"><span>GDCdownload(query, files.per.chunk = <span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) åŠ è½½æ•°æ®è‡³å½“å‰ç¯å¢ƒ</span>
</span></span><span style="display:flex;"><span>data &lt;- GDCprepare(query, save = <span style="color:#fff;font-weight:bold">T</span>, save.filename = <span style="color:#0ff;font-weight:bold">&#34;CHOL_RNAseq_hg38.rda&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="22-æ¢ç´¢æ•°æ®">2.2 æ¢ç´¢æ•°æ®<a hidden class="anchor" aria-hidden="true" href="#22-æ¢ç´¢æ•°æ®">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>library(SummarizedExperiment)
</span></span><span style="display:flex;"><span>dim(assay(data))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1] 60660    44</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(1) æŸ¥çœ‹è¡¨è¾¾çŸ©é˜µ</span>
</span></span><span style="display:flex;"><span>assayNames(data)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1] &#34;unstranded&#34;       &#34;stranded_first&#34;   &#34;stranded_second&#34;  &#34;tpm_unstrand&#34;    </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[5] &#34;fpkm_unstrand&#34;    &#34;fpkm_uq_unstrand&#34;</span>
</span></span><span style="display:flex;"><span>assay(data, <span style="color:#0ff;font-weight:bold">&#34;unstranded&#34;</span>)[1:<span style="color:#ff0;font-weight:bold">3</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">3</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                    TCGA-W5-AA2X-01A-11R-A41I-07 TCGA-W5-AA2X-11A-11R-A41I-07 TCGA-W5-AA33-01A-11R-A41I-07</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ENSG00000000003.15                         3310                         2322                        11312</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ENSG00000000005.6                             0                            0                            1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ENSG00000000419.13                         1881                          716                          968</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) æ ·æœ¬(ä¸´åºŠ)ä¿¡æ¯</span>
</span></span><span style="display:flex;"><span>dim(colData(data))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1]  44 201</span>
</span></span><span style="display:flex;"><span>colData(data)[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) åŸºå› ä¿¡æ¯</span>
</span></span><span style="display:flex;"><span>t(rowData(data)[1,])
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TransposedDataFrame with 10 rows and 1 column</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                           ENSG00000000003.15</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># source      &lt;factor&gt;                  HAVANA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># type        &lt;factor&gt;                    gene</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># score       &lt;numeric&gt;                     NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># phase       &lt;integer&gt;                     NA</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gene_id     &lt;character&gt;   ENSG00000000003.15</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gene_type   &lt;character&gt;       protein_coding</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># gene_name   &lt;character&gt;               TSPAN6</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># level       &lt;character&gt;                    2</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># hgnc_id     &lt;character&gt;           HGNC:11858</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># havana_gene &lt;character&gt; OTTHUMG00000022002.2</span>
</span></span><span style="display:flex;"><span>table(rowData(data)$gene_type) %&gt;% sort(decreasing = <span style="color:#fff;font-weight:bold">T</span>)  %&gt;% head()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># protein_coding                 lncRNA   processed_pseudogene unprocessed_pseudogene               misc_RNA                  snRNA </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#          19962                  16901                  10167                   2614                   2212                   1901 </span>
</span></span><span style="display:flex;"><span>rowRanges(data)
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="23-æ•´ç†æ•°æ®">2.3 æ•´ç†æ•°æ®<a hidden class="anchor" aria-hidden="true" href="#23-æ•´ç†æ•°æ®">#</a></h3>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1)æ›´æ”¹è¡¨è¾¾çŸ©é˜µè¡Œå/åŸºå› åæ ¼å¼ï¼Œç­›é€‰åŸºå› </span>
</span></span><span style="display:flex;"><span>rownames(data) = rowData(data)$gene_name
</span></span><span style="display:flex;"><span>data = data[!duplicated(rownames(data)),]
</span></span><span style="display:flex;"><span>data = data[rowData(data)$gene_type==<span style="color:#0ff;font-weight:bold">&#34;protein_coding&#34;</span>,]
</span></span><span style="display:flex;"><span>dim(data)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 19934    44</span>
</span></span><span style="display:flex;"><span>assay(data, <span style="color:#0ff;font-weight:bold">&#34;unstranded&#34;</span>)[1:<span style="color:#ff0;font-weight:bold">3</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">3</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        TCGA-W5-AA2X-01A-11R-A41I-07 TCGA-W5-AA2X-11A-11R-A41I-07 TCGA-W5-AA33-01A-11R-A41I-07</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TSPAN6                         3310                         2322                        11312</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TNMD                              0                            0                            1</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DPM1                           1881                          716                          968</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) åˆ†ææ ·æœ¬é—´ç›¸å…³æ€§ï¼Œé‰´å®šç¦»ç¾¤ç‚¹æ ·æœ¬</span>
</span></span><span style="display:flex;"><span>CorOutliers &lt;- TCGAanalyze_Preprocessing(data, cor.cut = <span style="color:#ff0;font-weight:bold">0</span>, 
</span></span><span style="display:flex;"><span>                                         filename=<span style="color:#0ff;font-weight:bold">&#34;CHOL45_hg38_cor.png&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) æ ‡å‡†åŒ–</span>
</span></span><span style="display:flex;"><span>dataNorm &lt;- TCGAanalyze_Normalization(tabDF = data, geneInfo =  geneInfo) <span style="color:#007f7f">#æ²¡æœ‰logå¤„ç†</span>
</span></span><span style="display:flex;"><span>dim(dataNorm)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] 15196    44</span>
</span></span><span style="display:flex;"><span>class(dataNorm) 
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;matrix&#34; &#34;array&#34;</span>
</span></span><span style="display:flex;"><span>dataNorm[1:<span style="color:#ff0;font-weight:bold">3</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">3</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        TCGA-W5-AA2X-01A-11R-A41I-07 TCGA-W5-AA2X-11A-11R-A41I-07 TCGA-W5-AA33-01A-11R-A41I-07</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TSPAN6                         3310                         2322                        11312</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DPM1                           1881                          716                          968</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># SCYL3                           966                          315                         1296</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(4) è¿‡æ»¤ä½è¡¨è¾¾åŸºå› </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># dataFilt &lt;- TCGAanalyze_Filtering(tabDF = dataNorm,</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                                   method = &#34;quantile&#34;, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                                   qnt.cut =  0.05)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># dim(dataFilt)</span>
</span></span><span style="display:flex;"><span>exp_data = dataNorm
</span></span></code></pre></td></tr></table>
</div>
</div><h3 id="24-å·®å¼‚åˆ†æ">2.4 å·®å¼‚åˆ†æ<a hidden class="anchor" aria-hidden="true" href="#24-å·®å¼‚åˆ†æ">#</a></h3>
<h4 id="1-edger">(1) edgeR<a hidden class="anchor" aria-hidden="true" href="#1-edger">#</a></h4>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1) ç¡®å®šåˆ†ç»„æ ·æœ¬</span>
</span></span><span style="display:flex;"><span>samplesNT &lt;- TCGAquery_SampleTypes(barcode = colnames(exp_data),
</span></span><span style="display:flex;"><span>                                   typesample = c(<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>))
</span></span><span style="display:flex;"><span>samplesTP &lt;- TCGAquery_SampleTypes(barcode = colnames(exp_data), 
</span></span><span style="display:flex;"><span>                                   typesample = c(<span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2) å·®å¼‚åˆ†æ</span>
</span></span><span style="display:flex;"><span>dataDEGs &lt;- TCGAanalyze_DEA(mat1 = exp_data[,samplesNT],
</span></span><span style="display:flex;"><span>                            mat2 = exp_data[,samplesTP],
</span></span><span style="display:flex;"><span>                            Cond1type = <span style="color:#0ff;font-weight:bold">&#34;Normal&#34;</span>,
</span></span><span style="display:flex;"><span>                            Cond2type = <span style="color:#0ff;font-weight:bold">&#34;Tumor&#34;</span>,
</span></span><span style="display:flex;"><span>                            pipeline = <span style="color:#0ff;font-weight:bold">&#34;edgeR&#34;</span>,
</span></span><span style="display:flex;"><span>                            method = <span style="color:#0ff;font-weight:bold">&#34;glmLRT&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(3) æ•´ç†ç»“æœ</span>
</span></span><span style="display:flex;"><span>DEGs &lt;- TCGAanalyze_LevelTab(dataDEGs,<span style="color:#0ff;font-weight:bold">&#34;Tumor&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;Normal&#34;</span>,
</span></span><span style="display:flex;"><span>                             exp_data[,samplesTP],exp_data[,samplesNT])
</span></span><span style="display:flex;"><span>DEGs = DEGs[order(DEGs$FDR),]
</span></span><span style="display:flex;"><span>DEGs$Direct = ifelse(DEGs$logFC&gt;<span style="color:#ff0;font-weight:bold">0</span>, <span style="color:#0ff;font-weight:bold">&#34;Up&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Down&#34;</span>)
</span></span><span style="display:flex;"><span>head(DEGs)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#          mRNA     logFC          FDR      Delta      Tumor      Normal Direct</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># USH2A   USH2A -5.261122 1.805847e-46  -150.9650   28.69444  1207.88889   Down</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># KCNN2   KCNN2 -4.344681 1.045519e-36  -188.2695   43.33333   893.77778   Down</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># LCAT     LCAT -3.683368 2.344098e-31 -1875.2437  509.11111  6640.22222   Down</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># IQCE     IQCE  4.229960 6.301169e-30 12462.2846 2946.19444   153.33333     Up</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DCXR     DCXR -4.070648 9.023484e-30 -7408.0132 1819.86111 32284.22222   Down</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># PRSS16 PRSS16  6.947385 3.517876e-28  9163.9866 1319.05556    10.22222     Up</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>DEGs = DEGs[order(DEGs$FDR),]
</span></span><span style="display:flex;"><span>DEGs$Direct = ifelse(DEGs$logFC&gt;<span style="color:#ff0;font-weight:bold">0</span>, <span style="color:#0ff;font-weight:bold">&#34;Up&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Down&#34;</span>)
</span></span><span style="display:flex;"><span>table(DEGs$FDR&lt;<span style="color:#ff0;font-weight:bold">0.01</span>,DEGs$Direct)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>DEGs_sig = subset(DEGs, FDR&lt;<span style="color:#ff0;font-weight:bold">0.01</span>)
</span></span><span style="display:flex;"><span>table(DEGs_sig$Direct, abs(DEGs_sig$logFC)&gt;<span style="color:#ff0;font-weight:bold">1</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><h4 id="2-deseq2">(2) Deseq2<a hidden class="anchor" aria-hidden="true" href="#2-deseq2">#</a></h4>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">31
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">32
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">33
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">34
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">35
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">36
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">37
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">38
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">39
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">40
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">41
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">42
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">43
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">44
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">45
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">46
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">##(1)Deseq2å·®å¼‚åˆ†æ</span>
</span></span><span style="display:flex;"><span>library(TCGAbiolinks)
</span></span><span style="display:flex;"><span>library(DESeq2)
</span></span><span style="display:flex;"><span>library(tidyverse)
</span></span><span style="display:flex;"><span>data$sample_type = factor(data$sample_type, 
</span></span><span style="display:flex;"><span>                          levels = c(<span style="color:#0ff;font-weight:bold">&#34;Solid Tissue Normal&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;Primary Tumor&#34;</span>))
</span></span><span style="display:flex;"><span>ddsSE &lt;- DESeqDataSet(data, design = ~ sample_type)
</span></span><span style="display:flex;"><span>keep &lt;- rowSums(counts(ddsSE)) &gt;= <span style="color:#ff0;font-weight:bold">10</span>
</span></span><span style="display:flex;"><span>ddsSE &lt;- ddsSE[keep,]
</span></span><span style="display:flex;"><span>ddsSE &lt;- DESeq(ddsSE)
</span></span><span style="display:flex;"><span>resultsNames(ddsSE)
</span></span><span style="display:flex;"><span>res &lt;- results(ddsSE, name = <span style="color:#0ff;font-weight:bold">&#34;sample_type_Primary.Tumor_vs_Solid.Tissue.Normal&#34;</span>)
</span></span><span style="display:flex;"><span>dea &lt;- as.data.frame(res) %&gt;% arrange(padj)
</span></span><span style="display:flex;"><span>head(dea)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        baseMean log2FoldChange     lfcSE      stat       pvalue         padj</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MSMO1 7585.3117      -3.748458 0.2108963 -17.77394 1.125105e-70 1.012094e-66</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># RMDN2  396.7564      -2.929007 0.1648043 -17.77264 1.151350e-70 1.012094e-66</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GCDH  2272.8992      -3.302565 0.1863391 -17.72342 2.765888e-70 1.620903e-66</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCAIM 1239.0498      -2.235399 0.1303361 -17.15104 6.174288e-66 2.713754e-62</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># KCNN2  360.4370      -5.504733 0.3234314 -17.01979 5.858475e-65 2.059957e-61</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># RCL1  2371.8288      -3.917907 0.2368974 -16.53842 1.940540e-61 5.686107e-58</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">##(2)é…å¯¹å·®å¼‚åˆ†æ</span>
</span></span><span style="display:flex;"><span>data_pair = data[,data$barcode  %in% TCGAquery_MatchedCoupledSampleTypes(data$barcode, c(<span style="color:#0ff;font-weight:bold">&#34;NT&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;TP&#34;</span>))]
</span></span><span style="display:flex;"><span>dim(data_pair)
</span></span><span style="display:flex;"><span>data_pair$pair = substr(data_pair$barcode, <span style="color:#ff0;font-weight:bold">1</span>, <span style="color:#ff0;font-weight:bold">12</span>)
</span></span><span style="display:flex;"><span>table(data_pair$pair)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># TCGA-W5-AA2I TCGA-W5-AA2Q TCGA-W5-AA2U TCGA-W5-AA2X TCGA-W5-AA30 TCGA-W5-AA31 TCGA-W5-AA34 TCGA-ZU-A8S4 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#            2            2            2            2            2            2            2            2</span>
</span></span><span style="display:flex;"><span>data_pair$sample_type = factor(data_pair$sample_type, 
</span></span><span style="display:flex;"><span>                          levels = c(<span style="color:#0ff;font-weight:bold">&#34;Solid Tissue Normal&#34;</span>,<span style="color:#0ff;font-weight:bold">&#34;Primary Tumor&#34;</span>))
</span></span><span style="display:flex;"><span>ddsSE &lt;- DESeqDataSet(data_pair, design = ~ sample_type + pair)
</span></span><span style="display:flex;"><span>keep &lt;- rowSums(counts(ddsSE)) &gt;= <span style="color:#ff0;font-weight:bold">10</span>
</span></span><span style="display:flex;"><span>ddsSE &lt;- ddsSE[keep,]
</span></span><span style="display:flex;"><span>ddsSE &lt;- DESeq(ddsSE)
</span></span><span style="display:flex;"><span>resultsNames(ddsSE)
</span></span><span style="display:flex;"><span>res &lt;- results(ddsSE, name = <span style="color:#0ff;font-weight:bold">&#34;sample_type_Primary.Tumor_vs_Solid.Tissue.Normal&#34;</span>)
</span></span><span style="display:flex;"><span>dea_p &lt;- as.data.frame(res) %&gt;% arrange(padj)
</span></span><span style="display:flex;"><span>head(dea_p)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#        baseMean log2FoldChange     lfcSE      stat        pvalue          padj</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># PCK2  22698.702      -5.653596 0.1854224 -30.49037 3.496861e-204 5.903052e-200</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># GRHPR 17379.001      -3.574709 0.1552370 -23.02743 2.476640e-117 2.090408e-113</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># KDM8   1614.739      -4.555841 0.1989383 -22.90077 4.564515e-116 2.568453e-112</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CTH    2367.036      -5.254399 0.2324287 -22.60650 3.740448e-113 1.578563e-109</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ADI1  18162.322      -3.714172 0.1659180 -22.38559 5.438272e-111 1.836069e-107</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># PBLD   8315.323      -5.683141 0.2593493 -21.91308 1.949637e-106 5.485303e-103</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h1 id="å››å…³äºç—…äººçš„ä¸´åºŠæ•°æ®ä¸è‚¿ç˜¤åˆ†å‹">å››ã€å…³äºç—…äººçš„ä¸´åºŠæ•°æ®ä¸è‚¿ç˜¤åˆ†å‹<a hidden class="anchor" aria-hidden="true" href="#å››å…³äºç—…äººçš„ä¸´åºŠæ•°æ®ä¸è‚¿ç˜¤åˆ†å‹">#</a></h1>
<h2 id="1è·å–ç—…äººçš„ä¸´åºŠæ•°æ®">1ã€è·å–ç—…äººçš„ä¸´åºŠæ•°æ®<a hidden class="anchor" aria-hidden="true" href="#1è·å–ç—…äººçš„ä¸´åºŠæ•°æ®">#</a></h2>
<ul>
<li>å¦‚ä¸Šåœ¨<code>GDCprepare()</code>è¿‡ç¨‹ä¸­ï¼Œä¼šè‡ªåŠ¨æ³¨é‡Šç—…äººæ ·æœ¬çš„ä¸´åºŠä¿¡æ¯ã€‚</li>
<li>æˆ‘ä»¬ä¹Ÿå¯ä»¥é¢„å…ˆå•ç‹¬ä¸‹è½½æ¯ä¸ªç—…äººçš„ä¸´åºŠæ•°æ®ï¼Œä»¥ä¾›å‚è€ƒã€‚</li>
</ul>
<h5 id="æ–¹æ³•ä¸€gdcquery-pipeline">æ–¹æ³•ä¸€ï¼šGDCquery() pipeline<a hidden class="anchor" aria-hidden="true" href="#æ–¹æ³•ä¸€gdcquery-pipeline">#</a></h5>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>query &lt;- GDCquery(project = <span style="color:#0ff;font-weight:bold">&#34;TCGA-ACC&#34;</span>, 
</span></span><span style="display:flex;"><span>                  data.category = <span style="color:#0ff;font-weight:bold">&#34;Clinical&#34;</span>,
</span></span><span style="display:flex;"><span>                  data.type = <span style="color:#0ff;font-weight:bold">&#34;Clinical Supplement&#34;</span>, 
</span></span><span style="display:flex;"><span>                  data.format = <span style="color:#0ff;font-weight:bold">&#34;BCR Biotab&#34;</span>)
</span></span><span style="display:flex;"><span>GDCdownload(query, files.per.chunk = <span style="color:#ff0;font-weight:bold">20</span>)
</span></span><span style="display:flex;"><span>clinical.BCRtab.all &lt;- GDCprepare(query)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>grep(<span style="color:#0ff;font-weight:bold">&#34;clinical_&#34;</span>, names(clinical.BCRtab.all), value = <span style="color:#fff;font-weight:bold">T</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [1] &#34;clinical_drug_brca&#34;               &#34;clinical_omf_v4.0_brca&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [3] &#34;clinical_follow_up_v4.0_brca&#34;     &#34;clinical_follow_up_v1.5_brca&#34;    </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [5] &#34;clinical_follow_up_v4.0_nte_brca&#34; &#34;clinical_patient_brca&#34;           </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [7] &#34;clinical_radiation_brca&#34;          &#34;clinical_nte_brca&#34;               </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [9] &#34;clinical_follow_up_v2.1_brca&#34; </span>
</span></span><span style="display:flex;"><span>clinical_patient_brca = as.data.frame(clinical.BCRtab.all$clinical_patient_brca)
</span></span><span style="display:flex;"><span>clinical_patient_brca[1:<span style="color:#ff0;font-weight:bold">4</span>,<span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                       bcr_patient_uuid bcr_patient_barcode form_completion_date                  prospective_collection</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1                     bcr_patient_uuid bcr_patient_barcode form_completion_date tissue_prospective_collection_indicator</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2                              CDE_ID:      CDE_ID:2003301              CDE_ID:                          CDE_ID:3088492</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3 6E7D5EC6-A469-467C-B748-237353C23416        TCGA-3C-AAAU            2014-1-13                                      NO</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4 55262FCB-1B01-4480-B322-36570430C917        TCGA-3C-AALI            2014-7-28                                      NO</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h5 id="æ–¹æ³•äºŒgdcquery_clinic">æ–¹æ³•äºŒï¼šGDCquery_clinic()<a hidden class="anchor" aria-hidden="true" href="#æ–¹æ³•äºŒgdcquery_clinic">#</a></h5>
<ul>
<li>æ ¹æ®å®˜æ–¹ä»‹ç»ï¼Œè¿™ä¸ªå‡½æ•°ä¸‹è½½çš„æ˜¯indexed clinical: a refined clinical data that is created using the XML files(æ–¹æ³•ä¸€).</li>
<li>è¿™ç§æ–¹æ³•ä¸‹è½½é€Ÿåº¦è¾ƒå¿«ï¼Œå»ºè®®ä¼˜å…ˆä½¿ç”¨ã€‚å¦‚æœæ²¡æœ‰æƒ³è¦çš„ä¿¡æ¯ï¼Œå†ä½¿ç”¨æ–¹æ³•ä¸€ã€‚</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-fallback" data-lang="fallback"><span style="display:flex;"><span>clinical &lt;- GDCquery_clinic(project = &#34;TCGA-BRCA&#34;, type = &#34;clinical&#34;)
</span></span><span style="display:flex;"><span>clinical &lt;- GDCquery_clinic(project = &#34;TCGA-BRCA&#34;, type = &#34;clinical&#34;)
</span></span><span style="display:flex;"><span>clinical[1:4,1:4]
</span></span><span style="display:flex;"><span>#   submitter_id synchronous_malignancy ajcc_pathologic_stage tumor_stage
</span></span><span style="display:flex;"><span># 1 TCGA-E2-A14U                     No               Stage I     stage i
</span></span><span style="display:flex;"><span># 2 TCGA-E9-A1RC                     No            Stage IIIC  stage iiic
</span></span><span style="display:flex;"><span># 3 TCGA-D8-A1J9                     No              Stage IA    stage ia
</span></span><span style="display:flex;"><span># 4 TCGA-E2-A14P                     No            Stage IIIC  stage iiic
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="2è·å–ç—…äººçš„è‚¿ç˜¤åˆ†å‹">2ã€è·å–ç—…äººçš„è‚¿ç˜¤åˆ†å‹<a hidden class="anchor" aria-hidden="true" href="#2è·å–ç—…äººçš„è‚¿ç˜¤åˆ†å‹">#</a></h2>
<ul>
<li><code>PanCancerAtlas_subtypes()</code>
The columns â€œSubtype_Selectedâ€ was selected as most prominent subtype classification (from the other columns)</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>subtypes &lt;- PanCancerAtlas_subtypes()
</span></span><span style="display:flex;"><span>dim(subtypes)
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#[1] 7734   10</span>
</span></span><span style="display:flex;"><span>table(subtypes$cancer.type)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ACC  AML BLCA BRCA COAD ESCA  GBM HNSC KICH KIRC KIRP  LGG LIHC LUAD LUSC OVCA PCPG </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 91  187  129 1218  341  169  606  279   66  442  161  516  196  230  178  489  178 </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># PRAD READ SKCM STAD THCA UCEC  UCS </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 333  118  333  383  496  538   57</span>
</span></span><span style="display:flex;"><span>head(as.data.frame(subtypes))
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#   pan.samplesID cancer.type                         Subtype_mRNA   Subtype_DNAmeth Subtype_protein Subtype_miRNA Subtype_CNA Subtype_Integrative Subtype_other      Subtype_Selected</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 1  TCGA-OR-A5J1         ACC steroid-phenotype-high+proliferation         CIMP-high              NA       miRNA_1       Quiet                COC3           C1A         ACC.CIMP-high</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 2  TCGA-OR-A5J2         ACC steroid-phenotype-high+proliferation          CIMP-low               1       miRNA_1       Noisy                COC3           C1A          ACC.CIMP-low</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 3  TCGA-OR-A5J3         ACC               steroid-phenotype-high CIMP-intermediate               3       miRNA_6 Chromosomal                COC2           C1A ACC.CIMP-intermediate</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 4  TCGA-OR-A5J4         ACC                                 &lt;NA&gt;         CIMP-high              NA       miRNA_6 Chromosomal                &lt;NA&gt;          &lt;NA&gt;         ACC.CIMP-high</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 5  TCGA-OR-A5J5         ACC               steroid-phenotype-high CIMP-intermediate              NA       miRNA_2 Chromosomal                COC2           C1A ACC.CIMP-intermediate</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 6  TCGA-OR-A5J6         ACC                steroid-phenotype-low          CIMP-low               2       miRNA_1       Noisy                COC1           C1B          ACC.CIMP-low</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li><code>TCGAquery_subtype()</code>
These subtypes will be automatically added in the summarizedExperiment object through GDCprepare. But you can also use the TCGAquery_subtype function to retrieve this information.</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span>brca.subtype &lt;- TCGAquery_subtype(tumor = <span style="color:#0ff;font-weight:bold">&#34;brca&#34;</span>)
</span></span><span style="display:flex;"><span>t(brca.subtype[1,])
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#                                     [,1]          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># patient                             &#34;TCGA-3C-AAAU&#34;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Tumor.Type                          &#34;BRCA&#34;        </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Included_in_previous_marker_papers  &#34;NO&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># vital_status                        &#34;Alive&#34;       </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># days_to_birth                       &#34;-20211&#34;      </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># days_to_death                       &#34;NA&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># days_to_last_followup               &#34;4047&#34;        </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># age_at_initial_pathologic_diagnosis &#34;55&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># pathologic_stage                    &#34;NA&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Tumor_Grade                         &#34;NA&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># BRCA_Pathology                      &#34;NA&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># BRCA_Subtype_PAM50                  &#34;LumA&#34;        </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># MSI_status                          &#34;NA&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># HPV_Status                          &#34;NA&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># tobacco_smoking_history             &#34;NA&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># CNV Clusters                        &#34;C6&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Mutation Clusters                   &#34;C7&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># DNA.Methylation Clusters            &#34;C1&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># mRNA Clusters                       &#34;C1&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># miRNA Clusters                      &#34;C3&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># lncRNA Clusters                     &#34;NA&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Protein Clusters                    &#34;NA&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># PARADIGM Clusters                   &#34;C5&#34;          </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Pan-Gyn Clusters                    &#34;NA&#34;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p><code>GDCquery_Maf()</code>å‡½æ•°å¯ä»¥æ”¯æŒä¸‹è½½çªå˜æ•°æ®ï¼Œè¿™é‡Œå°±æš‚æ—¶ä¸å­¦ä¹ äº†ã€‚ä¹‹åæœ‰æœºä¼šå†äº†è§£ä¸€ä¸‹ã€‚</p></blockquote>
<h1 id="è¡¥å……">è¡¥å……<a hidden class="anchor" aria-hidden="true" href="#è¡¥å……">#</a></h1>
<h2 id="1-tcgaç®€ç§°å¯¹åº”å…¨ç§°ä»¥åŠä¸­æ–‡å">1 ã€TCGAç®€ç§°å¯¹åº”å…¨ç§°ä»¥åŠä¸­æ–‡å<a hidden class="anchor" aria-hidden="true" href="#1-tcgaç®€ç§°å¯¹åº”å…¨ç§°ä»¥åŠä¸­æ–‡å">#</a></h2>
<table>
  <thead>
      <tr>
          <th>Study Abbreviation</th>
          <th>Study Name</th>
          <th>ä¸­æ–‡å</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td>ACC</td>
          <td>Adrenocortical carcinoma</td>
          <td>è‚¾ä¸Šè…ºçš®è´¨ç™Œ</td>
      </tr>
      <tr>
          <td>BLCA</td>
          <td>Bladder Urothelial Carcinoma</td>
          <td>è†€èƒ±å°¿è·¯ä¸Šçš®ç™Œ</td>
      </tr>
      <tr>
          <td>BRCA</td>
          <td>Breast invasive carcinoma</td>
          <td>æµ¸æ¶¦æ€§ä¹³è…ºç™Œ</td>
      </tr>
      <tr>
          <td>CESC</td>
          <td>Cervical squamous cell carcinoma and endocervical adenocarcinoma</td>
          <td>å®«é¢ˆé³çŠ¶ç»†èƒç™Œå’Œå®«é¢ˆå†…è…ºç™Œ</td>
      </tr>
      <tr>
          <td>CHOL</td>
          <td>Cholangiocarcinoma</td>
          <td>èƒ†ç®¡ç™Œ</td>
      </tr>
      <tr>
          <td>COAD</td>
          <td>Colon adenocarcinoma</td>
          <td>ç»“è‚ è…ºç™Œ</td>
      </tr>
      <tr>
          <td>DLBC</td>
          <td>Lymphoid Neoplasm Diffuse Large B-cell Lymphoma</td>
          <td>æ·‹å·´æ ·è‚¿ç˜¤å¼¥æ¼«å¤§bç»†èƒæ·‹å·´ç˜¤</td>
      </tr>
      <tr>
          <td>ESCA</td>
          <td>Esophageal carcinoma</td>
          <td>é£Ÿç®¡ç™Œç™Œ</td>
      </tr>
      <tr>
          <td>GBM</td>
          <td>Glioblastoma multiforme</td>
          <td>å¤šå½¢æ€§æˆèƒ¶è´¨ç»†èƒç˜¤</td>
      </tr>
      <tr>
          <td>HNSC</td>
          <td>Head and Neck squamous cell carcinoma</td>
          <td>å¤´é¢ˆéƒ¨é³çŠ¶ç»†èƒç™Œ</td>
      </tr>
      <tr>
          <td>KICH</td>
          <td>Kidney Chromophobe</td>
          <td>è‚¾å«Œè‰²ç»†èƒç™Œ</td>
      </tr>
      <tr>
          <td>KIRC</td>
          <td>Kidney renal clear cell carcinoma</td>
          <td>è‚¾é€æ˜ç»†èƒç™Œ</td>
      </tr>
      <tr>
          <td>KIRP</td>
          <td>Kidney renal papillary cell carcinoma</td>
          <td>è‚¾ä¹³å¤´çŠ¶ç»†èƒç™Œ</td>
      </tr>
      <tr>
          <td>LAML</td>
          <td>Acute Myeloid Leukemia</td>
          <td>æ€¥æ€§é«“ç³»ç™½è¡€ç—…</td>
      </tr>
      <tr>
          <td>LGG</td>
          <td>Brain Lower Grade Glioma</td>
          <td>è„‘ä½çº§åˆ«èƒ¶è´¨ç˜¤</td>
      </tr>
      <tr>
          <td>LIHC</td>
          <td>Liver hepatocellular carcinoma</td>
          <td>è‚è„è‚ç»†èƒç™Œ</td>
      </tr>
      <tr>
          <td>LUAD</td>
          <td>Lung adenocarcinoma</td>
          <td>è‚ºè…ºç™Œ</td>
      </tr>
      <tr>
          <td>LUSC</td>
          <td>Lung squamous cell carcinoma</td>
          <td>è‚ºé³ç™Œ</td>
      </tr>
      <tr>
          <td>MESO</td>
          <td>Mesothelioma</td>
          <td>é—´çš®ç˜¤</td>
      </tr>
      <tr>
          <td>OV</td>
          <td>Ovarian serous cystadenocarcinoma</td>
          <td>åµå·¢æµ†æ¶²æ€§å›Šè…ºç™Œ</td>
      </tr>
      <tr>
          <td>PAAD</td>
          <td>Pancreatic adenocarcinoma</td>
          <td>èƒ°è…ºè…ºç™Œ</td>
      </tr>
      <tr>
          <td>PCPG</td>
          <td>Pheochromocytoma and Paraganglioma</td>
          <td>å—œé“¬ç»†èƒç˜¤å’Œå‰¯ç¥ç»èŠ‚ç˜¤</td>
      </tr>
      <tr>
          <td>PRAD</td>
          <td>Prostate adenocarcinoma</td>
          <td>å‰åˆ—è…ºè…ºç™Œ</td>
      </tr>
      <tr>
          <td>READ</td>
          <td>Rectum adenocarcinoma</td>
          <td>ç›´è‚ è…ºç™Œ</td>
      </tr>
      <tr>
          <td>SARC</td>
          <td>Sarcoma</td>
          <td>è‚‰ç˜¤</td>
      </tr>
      <tr>
          <td>SKCM</td>
          <td>Skin Cutaneous Melanoma</td>
          <td>çš®è‚¤çš®è‚¤é»‘è‰²ç´ ç˜¤</td>
      </tr>
      <tr>
          <td>STAD</td>
          <td>Stomach adenocarcinoma</td>
          <td>èƒƒè…ºç™Œ</td>
      </tr>
      <tr>
          <td>TGCT</td>
          <td>Testicular Germ Cell Tumors</td>
          <td>ç¾ä¸¸ç”Ÿæ®–ç»†èƒè‚¿ç˜¤</td>
      </tr>
      <tr>
          <td>THCA</td>
          <td>Thyroid carcinoma</td>
          <td>ç”²çŠ¶è…ºç™Œ</td>
      </tr>
      <tr>
          <td>THYM</td>
          <td>Thymoma</td>
          <td>èƒ¸è…ºç˜¤</td>
      </tr>
      <tr>
          <td>UCEC</td>
          <td>Uterine Corpus Endometrial Carcinoma</td>
          <td>å­å®«å†…è†œç™Œ</td>
      </tr>
      <tr>
          <td>UCS</td>
          <td>Uterine Carcinosarcoma</td>
          <td>å­å®«ç™Œè‚‰ç˜¤</td>
      </tr>
      <tr>
          <td>UVM</td>
          <td>Uveal Melanoma</td>
          <td>è‘¡è„è†œé»‘è‰²ç´ ç˜¤</td>
      </tr>
  </tbody>
</table>
<h2 id="2-gdcæ‰€æœ‰æ•°æ®ç±»å‹">2 ã€GDCæ‰€æœ‰æ•°æ®ç±»å‹<a hidden class="anchor" aria-hidden="true" href="#2-gdcæ‰€æœ‰æ•°æ®ç±»å‹">#</a></h2>
<h3 id="21--gdc-harmonized-database">2.1  GDC harmonized database<a hidden class="anchor" aria-hidden="true" href="#21--gdc-harmonized-database">#</a></h3>
<table>
  <thead>
      <tr>
          <th>Data.category</th>
          <th>Data.type</th>
          <th>Workflow.Type</th>
          <th>Platform</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td>Transcriptome Profiling</td>
          <td>Gene Expression Quantification</td>
          <td>HTSeq - Counts</td>
          <td></td>
      </tr>
      <tr>
          <td>Transcriptome Profiling</td>
          <td>Gene Expression Quantification</td>
          <td>HTSeq - FPKM</td>
          <td></td>
      </tr>
      <tr>
          <td>Transcriptome Profiling</td>
          <td>Gene Expression Quantification</td>
          <td>HTSeq - FPKM-UQ</td>
          <td></td>
      </tr>
      <tr>
          <td>Transcriptome Profiling</td>
          <td>Gene Expression Quantification</td>
          <td>STAR - Counts</td>
          <td></td>
      </tr>
      <tr>
          <td>Transcriptome Profiling</td>
          <td>Isoform Expression Quantification</td>
          <td>-</td>
          <td></td>
      </tr>
      <tr>
          <td>Transcriptome Profiling</td>
          <td>miRNA Expression Quantification</td>
          <td>-</td>
          <td></td>
      </tr>
      <tr>
          <td>Transcriptome Profiling</td>
          <td>Splice Junction Quantification</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Copy number variation</td>
          <td>Copy Number Segment</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Copy number variation</td>
          <td>Masked Copy Number Segment</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Copy number variation</td>
          <td>Gene Level Copy Number Scores</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Simple Nucleotide Variation</td>
          <td>Masked Somatic Mutation</td>
          <td>MuSE Variant Aggregation and Masking</td>
          <td></td>
      </tr>
      <tr>
          <td>Simple Nucleotide Variation</td>
          <td>Masked Somatic Mutation</td>
          <td>MuTect2 Variant Aggregation and Masking</td>
          <td></td>
      </tr>
      <tr>
          <td>Simple Nucleotide Variation</td>
          <td>Masked Somatic Mutation</td>
          <td>SomaticSniper Variant Aggregation and Masking</td>
          <td></td>
      </tr>
      <tr>
          <td>Simple Nucleotide Variation</td>
          <td>Masked Somatic Mutation</td>
          <td>VarScan2 Variant Aggregation and Masking</td>
          <td></td>
      </tr>
      <tr>
          <td>Raw Sequencing Data</td>
          <td>-</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Biospecimen</td>
          <td>Slide Image</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Biospecimen</td>
          <td>Biospecimen Supplement</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Clinical</td>
          <td>-</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>DNA Methylation</td>
          <td>Methylation Beta Value</td>
          <td></td>
          <td>Illumina Human Methylation 450</td>
      </tr>
      <tr>
          <td>DNA Methylation</td>
          <td>Methylation Beta Value</td>
          <td></td>
          <td>Illumina Human Methylation 27</td>
      </tr>
  </tbody>
</table>
<h3 id="22--gdc-legacy-archive">2.2  GDC Legacy Archive<a hidden class="anchor" aria-hidden="true" href="#22--gdc-legacy-archive">#</a></h3>
<table>
  <thead>
      <tr>
          <th>Data.category</th>
          <th>Data.type</th>
          <th>Platform</th>
          <th>file.type</th>
      </tr>
  </thead>
  <tbody>
      <tr>
          <td>Copy number variation</td>
          <td>-</td>
          <td>Affymetrix SNP Array 6.0</td>
          <td>nocnv_hg18.seg</td>
      </tr>
      <tr>
          <td>Copy number variation</td>
          <td>-</td>
          <td>Affymetrix SNP Array 6.0</td>
          <td>hg18.seg</td>
      </tr>
      <tr>
          <td>Copy number variation</td>
          <td>-</td>
          <td>Affymetrix SNP Array 6.0</td>
          <td>nocnv_hg19.seg</td>
      </tr>
      <tr>
          <td>Copy number variation</td>
          <td>-</td>
          <td>Affymetrix SNP Array 6.0</td>
          <td>hg19.seg</td>
      </tr>
      <tr>
          <td>Copy number variation</td>
          <td>-</td>
          <td>Illumina HiSeq</td>
          <td>-</td>
      </tr>
      <tr>
          <td>Simple nucleotide variation</td>
          <td>Simple somatic mutation</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Raw sequencing data</td>
          <td></td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Biospecimen</td>
          <td></td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Clinical</td>
          <td></td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Protein expression</td>
          <td></td>
          <td>MDA RPPA Core</td>
          <td>-</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>Gene expression quantification</td>
          <td>Illumina HiSeq</td>
          <td>normalized_results</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>Gene expression quantification</td>
          <td>Illumina HiSeq</td>
          <td>results</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>Gene expression quantification</td>
          <td>HT_HG-U133A</td>
          <td>-</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>Gene expression quantification</td>
          <td>AgilentG4502A_07_2</td>
          <td>-</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>Gene expression quantification</td>
          <td>AgilentG4502A_07_1</td>
          <td>-</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>Gene expression quantification</td>
          <td>HuEx-1_0-st-v2</td>
          <td>FIRMA.txt</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>Gene expression quantification</td>
          <td></td>
          <td>gene.txt</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>Isoform expression quantification</td>
          <td>-</td>
          <td>-</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>miRNA gene quantification</td>
          <td>-</td>
          <td>hg19.mirna</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>miRNA gene quantification</td>
          <td></td>
          <td>hg19.mirbase20</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>miRNA gene quantification</td>
          <td></td>
          <td>mirna</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>Exon junction quantification</td>
          <td>-</td>
          <td>-</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>Exon quantification</td>
          <td>-</td>
          <td>-</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>miRNA isoform quantification</td>
          <td>-</td>
          <td>hg19.isoform</td>
      </tr>
      <tr>
          <td>Gene expression</td>
          <td>miRNA isoform quantification</td>
          <td>-</td>
          <td>isoform</td>
      </tr>
      <tr>
          <td>DNA methylation</td>
          <td></td>
          <td>Illumina Human Methylation 450</td>
          <td>Not used</td>
      </tr>
      <tr>
          <td>DNA methylation</td>
          <td></td>
          <td>Illumina Human Methylation 27</td>
          <td>Not used</td>
      </tr>
      <tr>
          <td>DNA methylation</td>
          <td></td>
          <td>Illumina DNA Methylation OMA003 CPI</td>
          <td>Not used</td>
      </tr>
      <tr>
          <td>DNA methylation</td>
          <td></td>
          <td>Illumina DNA Methylation OMA002 CPI</td>
          <td>Not used</td>
      </tr>
      <tr>
          <td>DNA methylation</td>
          <td></td>
          <td>Illumina Hi Seq</td>
          <td></td>
      </tr>
      <tr>
          <td>DNA methylation</td>
          <td>Bisulfite sequence alignment</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>DNA methylation</td>
          <td>Methylation percentage</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>DNA methylation</td>
          <td>Aligned reads</td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Raw microarray data</td>
          <td>Raw intensities</td>
          <td>Illumina Human Methylation 450</td>
          <td>idat</td>
      </tr>
      <tr>
          <td>Raw Microarray Data</td>
          <td>Raw intensities</td>
          <td>Illumina Human Methylation 27</td>
          <td>idat</td>
      </tr>
      <tr>
          <td>Structural Rearrangement</td>
          <td></td>
          <td></td>
          <td></td>
      </tr>
      <tr>
          <td>Other</td>
          <td></td>
          <td></td>
          <td></td>
      </tr>
  </tbody>
</table>


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">è½¬å½•ç»„</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/045%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90r%E5%8C%85-deseq2&#43;edger&#43;limma/">
    <span class="title">Â« Prev Page</span>
    <br>
    <span>å·®å¼‚åˆ†æRåŒ…-DESeq2&#43;edgeR&#43;limma</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/047%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5tpmfpkm%E6%A0%87%E5%87%86%E5%8C%96/">
    <span class="title">Next Page Â»</span>
    <br>
    <span>è¡¨è¾¾çŸ©é˜µTPM,FPKMæ ‡å‡†åŒ–</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>æ‚¨æ˜¯æœ¬ç«™ç¬¬ <span id="busuanzi_value_site_uv"></span> ä½è®¿é—®è€…ï¼Œæ€»æµè§ˆé‡ä¸º <span id="busuanzi_value_site_pv"></span> æ¬¡
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
