<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>转录组 | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="">
<meta name="description" content="李申锁的博客，记录生信学习过程。R/Python/Shell">
<meta name="author" content="">
<link rel="canonical" href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" type="application/rss+xml" href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/index.xml">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="转录组" />
<meta property="og:description" content="李申锁的博客，记录生信学习过程。R/Python/Shell" />
<meta property="og:type" content="website" />
<meta property="og:url" content="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="转录组"/>
<meta name="twitter:description" content="李申锁的博客，记录生信学习过程。R/Python/Shell"/>

<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="list" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main"> 
<header class="page-header"><div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/tags/">Tags</a></div>
  <h1>转录组</h1>
</header>





<article class="post-entry tag-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>xCell与CIBERSORT等免疫浸润分析
			</h2>
		</header>
		<section class="entry-content">
			<p>xCell xCell包 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 devtools::install_github(&#39;dviraran/xCell&#39;) library(xCell) data(&#34;xCell.data&#34;) ##查看支持的64种细胞类型，同下图 colnames(xCell.data$spill$K) ##预测函数的关键参数解释 ?xCellAnalysis() # expr = 交代表达矩阵； ##如果是array，不需要额外标准化；如果是RNAseq，需要TPM/FPKM/TPM。 ##对于基因ID格式需要是symbol格式。 # rnaseq = TRUE 数据是否为RNAseq数据，如果是芯片数据设置为FALSE # cell.types.use = NULL 提供一个字符串，说明想要预测64种细胞中的哪些细胞类型 # parallel.sz = 4 调用的线程数，默认为4 NOTE：
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-25 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-25&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-02-18&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2554&amp;nbsp;|&amp;nbsp;6 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to xCell与CIBERSORT等免疫浸润分析" href="https://lishensuo.github.io/en/posts/bioinfo/072xcell%E4%B8%8Ecibersort%E7%AD%89%E5%85%8D%E7%96%AB%E6%B5%B8%E6%B6%A6%E5%88%86%E6%9E%90/"></a>
</article>





<article class="post-entry tag-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>ClusterGVis包绘制基因表达矩阵热图
			</h2>
		</header>
		<section class="entry-content">
			<p> ClusterGVis包是中国药科大学Jun Zhang博士开发的系列可视化工具包之一，可以基因表达矩阵进行高级的热图可视化分析。如下根据其github以及微信教程简单整理一下自己感兴趣的用法。
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-03-19 00:00:00 &#43;0000 UTC&#39;&gt;2023-03-19&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-03-19&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1261&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to ClusterGVis包绘制基因表达矩阵热图" href="https://lishensuo.github.io/en/posts/bioinfo/073clustergvis%E5%8C%85%E7%BB%98%E5%88%B6%E5%9F%BA%E5%9B%A0%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5%E7%83%AD%E5%9B%BE/"></a>
</article>





<article class="post-entry tag-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>MuSiC包根据scRNAseq预测Bulk细胞组成
			</h2>
		</header>
		<section class="entry-content">
			<p> MuSiC(MUlti-Subject SIngle Cell deconvolution)是来自宾夕法尼亚大学Biostatistics, Epidemiology and Informatics系的Mingyao Li课题组于2019年发表于Nature Communication的一个工具R包，可根据单细胞转录组信息推测Bulk RNA-seq细胞组成。而后，该团队又于2022年在Briefing in bioinformatics发表了扩展版本MuSiC2，可以考虑更复杂的场景。
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-03-26 00:00:00 &#43;0000 UTC&#39;&gt;2023-03-26&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-03-26&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1799&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to MuSiC包根据scRNAseq预测Bulk细胞组成" href="https://lishensuo.github.io/en/posts/bioinfo/074music%E5%8C%85%E6%A0%B9%E6%8D%AEscrnaseq%E9%A2%84%E6%B5%8Bbulk%E7%BB%86%E8%83%9E%E7%BB%84%E6%88%90/"></a>
</article>





<article class="post-entry tag-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>从RNAseq的fastq.gz提取表达矩阵
			</h2>
		</header>
		<section class="entry-content">
			<p> RNA-seq数据比对流程主要分为三步（1）整理数据；（2）质控；（3）比对。其中每一步都涉及到若干软件的用法，如下简单整理基本的分析流程。
示例数据：GSE158623中6个样本的RNA-seq测序结果(human)，对应SRR12720999~SRR12721004
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-29 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-29&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-29&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;2347&amp;nbsp;|&amp;nbsp;5 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to 从RNAseq的fastq.gz提取表达矩阵" href="https://lishensuo.github.io/en/posts/bioinfo/060%E4%BB%8Ernaseq%E7%9A%84fastq.gz%E6%8F%90%E5%8F%96%E8%A1%A8%E8%BE%BE%E7%9F%A9%E9%98%B5/"></a>
</article>





<article class="post-entry tag-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>clusterProfiler包富集分析与可视化
			</h2>
		</header>
		<section class="entry-content">
			<p>参考
https://gseapy.readthedocs.io/en/latest/index.html 示例数据来自 https://github.com/zqfang/GSEApy/tree/master/tests/data 1 2 3 4 5 6 7 8 import os import pandas as pd import scanpy as sc import gseapy as gp from gseapy import barplot, dotplot import matplotlib.pyplot as plt 0. 基因集获取 Enrichr API supported organisms: Human, Mouse, Yeast, Fly, Fish, Worm 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 names = gp.get_library_name(organism = &#34;Human&#34;) len(names) # 217 [name for name in names if &#34;GO&#34; in name] # [&#39;GO_Biological_Process_2021&#39;, # &#39;GO_Biological_Process_2023&#39;, # &#39;GO_Biological_Process_2025&#39;, # &#39;GO_Cellular_Component_2021&#39;, # &#39;GO_Cellular_Component_2023&#39;, # &#39;GO_Cellular_Component_2025&#39;, # &#39;GO_Molecular_Function_2021&#39;, # &#39;GO_Molecular_Function_2023&#39;, # &#39;GO_Molecular_Function_2025&#39;, # &#39;SynGO_2022&#39;, # &#39;SynGO_2024&#39;] [name for name in names if &#34;KEGG&#34; in name] # [&#39;KEGG_2013&#39;, # &#39;KEGG_2015&#39;, # &#39;KEGG_2016&#39;, # &#39;KEGG_2019_Human&#39;, # &#39;KEGG_2019_Mouse&#39;, # &#39;KEGG_2021_Human&#39;] [name for name in names if &#34;MSigDB&#34; in name] # [&#39;MSigDB_Computational&#39;, &#39;MSigDB_Hallmark_2020&#39;, &#39;MSigDB_Oncogenic_Signatures&#39;] kegg = gp.get_library(name=&#39;KEGG_2021_Human&#39;, organism=&#39;Human&#39;) type(kegg) # dict kegg[&#39;AMPK signaling pathway&#39;][:4] # [&#39;RAB2A&#39;, &#39;PPP2R1A&#39;, &#39;TSC2&#39;, &#39;TSC1&#39;] 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 # 还支持对MSigDB基因集数据库的获取 from gseapy import Msigdb msig = Msigdb() # all available versions msig.list_dbver() # all types of gene sets given a dbver msig.list_category(dbver=&#34;2025.1.Hs&#34;) # all gene sets of one type gmt = msig.get_gmt(category=&#39;h.all&#39;, dbver=&#34;2025.1.Hs&#34;) type(gmt) # dict gmt.keys() # hallmark pathway names gmt[&#34;HALLMARK_ADIPOGENESIS&#34;] # gene list 1. ORA富集 1.1 分析 Online analysis 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 ## 1) 两个最重要的参数 （后同） # gene_list gene_list=&#34;tests/data/gene_list.txt&#34; # file path, one gene per row gene_list= [&#34;RAB2A&#34;,&#34;PRKAB2&#34;,&#34;PRKAA1&#34;] # list of genes # gene_sets (from enrichr API) gene_sets=&#39;KEGG_2016&#39; gene_sets=[&#39;KEGG_2021_Human&#39;,&#39;GO_Biological_Process_2025&#39;] gene_list = pd.read_csv(&#34;tests/data/gene_list.txt&#34;, header=None)[0].tolist() gene_list[:4] # [&#39;IGKV4-1&#39;, &#39;CD55&#39;, &#39;IGKC&#39;, &#39;PPFIBP1&#39;] enr = gp.enrichr(gene_list=gene_list, # or &#34;./tests/data/gene_list.txt&#34;, gene_sets=[&#39;GO_Biological_Process_2025&#39;], organism=&#39;human&#39;, outdir=None, ) enr.results.head(5) # Index([&#39;Gene_set&#39;, &#39;Term&#39;, &#39;Overlap&#39;, &#39;P-value&#39;, &#39;Adjusted P-value&#39;, # &#39;Old P-value&#39;, &#39;Old Adjusted P-value&#39;, &#39;Odds Ratio&#39;, &#39;Combined Score&#39;, # &#39;Genes&#39;], # dtype=&#39;object&#39;) ## 2) 设置结果导出路径 enr = gp.enrichr(gene_list=gene_list, gene_sets=[&#39;GO_Biological_Process_2025&#39;], organism=&#39;human&#39;, outdir=&#34;enrichr_kegg&#34;, ) os.listdir(&#34;enrichr_kegg&#34;) # [&#39;GO_Biological_Process_2025.human.enrichr.reports.txt&#39;, # &#39;GO_Biological_Process_2025.human.enrichr.reports.pdf&#39;] ## 3) 自定义Background genes (P值会存在一定差异) enr = gp.enrichr(gene_list=gene_list, gene_sets=[&#39;GO_Biological_Process_2025&#39;], organism=&#39;human&#39;, background=&#34;tests/data/background.txt&#34;, # 自定义背景基因 / or list object outdir=None, ) Offline analysis (本地提供通路基因集数据) 1 2 3 4 5 6 7 8 9 10 11 from gseapy.parser import read_gmt gmt_dict = read_gmt(&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;) gmt_dict.keys() enr2 = gp.enrich(gene_list=gene_list, # or gene_list=glist gene_sets=[&#34;tests/data/c2.cp.kegg.v7.5.1.symbols.gmt&#34;], background=None, outdir=None, verbose=True) enr2.results.head(5) 1.2 可视化 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 # ofname 设置图片的导出路径，若为None则不保存。也支持后续使用plt语法进一步调整图片 ax = dotplot(enr.results, column=&#34;Adjusted P-value&#34;, x=&#39;Gene_set&#39;, # set x axis, so you could do a multi-sample/library comparsion size=5, cutoff=0.05, top_term=10, figsize=(3, 5), title = &#34;GO_BP&#34;, xticklabels_rot=45, # rotate xtick labels show_ring=True, # set to False to revmove outer ring marker=&#39;o&#39;, ofname=&#39;./plot/p1.png&#39; ) ax = barplot(enr.results, column=&#34;Adjusted P-value&#34;, group=&#39;Gene_set&#39;, # set group, so you could do a multi-sample/library comparsion size=10, top_term=10, figsize=(3,5), #color=[&#39;darkred&#39;, &#39;darkblue&#39;] # set colors for group color = {&#39;GO_Biological_Process_2025&#39;: &#39;salmon&#39;}, ofname=&#39;./plot/p2.png&#39; ) 1 2 3 ax = dotplot(enr.res2d, title=&#39;Enrichr&#39;,cmap=&#39;viridis_r&#39;, size=10, figsize=(3,5), ofname=&#39;./plot/p3.png&#39;) ax = barplot(enr.res2d, title=&#39;Enrichr&#39;,cmap=&#39;viridis_r&#39;, size=4, figsize=(4, 5), color=&#39;darkred&#39;, ofname=&#39;./plot/p4.png&#39;) # 差别不大 2. Prerank富集 基于差异分析结果的GSEA
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2025-07-22 00:00:00 &#43;0000 UTC&#39;&gt;2025-07-22&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2025-07-22&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1367&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to clusterProfiler包富集分析与可视化" href="https://lishensuo.github.io/en/posts/bioinfo/061gseapy%E5%AF%8C%E9%9B%86%E5%88%86%E6%9E%90%E4%B8%8E%E5%8F%AF%E8%A7%86%E5%8C%96/"></a>
</article>





<article class="post-entry tag-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>肿瘤亚型分群工具ConsensusClusterPlus
			</h2>
		</header>
		<section class="entry-content">
			<p> ConsensusClusterPlus包是肿瘤分型研究的常用工具，其于2010年发表于Bioinformatics。
Paper：https://academic.oup.com/bioinformatics/article/26/12/1572/281699 Tutorial：https://bioconductor.org/packages/release/bioc/vignettes/ConsensusClusterPlus/inst/doc/ConsensusClusterPlus.pdf 1 2 # BiocManager::install(&#34;ConsensusClusterPlus&#34;) library(ConsensusClusterPlus) 1、示例表达矩阵 行名为基因，列名为样本的基因表达矩阵；根据需要，进行标准化处理 基因集的选择是关键的一步，可根据统计学（高变基因）或者生物学（特定功能相关基因）进行选择 如下是参考教程的数据 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 library(ALL) data(ALL) d=exprs(ALL) dim(d) # [1] 12625 128 mads=apply(d,1,mad) d=d[rev(order(mads))[1:5000],] # 筛选高方差基因 dim(d) # [1] 5000 128 ## 标准化处理：每行（基因）减去每行的中位数 d = sweep(d,1, apply(d,1,median,na.rm=T)) d[1:4,1:4] # 01005 01010 03002 04006 # 36638_at 1.556121 0.9521271 -0.05018082 4.780378 # 39318_at 1.191353 2.5013225 -2.38793537 -1.199521 # 38514_at 1.020716 3.2785671 1.55949145 -3.345919 # 266_s_at 1.829260 0.3624327 1.54913247 -1.286294 2、样本亚型鉴定 实际就是一个函数即可：ConsensusClusterPlus() 其中涉及到较多参数的选择，具体如下 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 ## 默认参数 ConsensusClusterPlus(d, maxK=3, reps=10, pItem=0.8, pFeature=1, title=&#34;untitled_consensus_cluster&#34;, clusterAlg=&#34;hc&#34;, distance=&#34;pearson&#34;, plot=NULL, writeTable=FALSE, seed=42) ## 参数含义 # maxK：考虑的最大聚类数，建议取10~20 # reps：抽样次数，建议1000 # pItem与pFeature：分别表示对样本与基因的抽样比例 # title：图片或文件的保存路径 # clusterAlg：聚类方式 c(&#34;hc&#34;,&#34;pam&#34;,&#34;km&#34;) # distance: 距离计算方式 c(&#34;pearson&#34;,&#34;spearman&#34;,&#34;euclidean&#34;,&#34;binay&#34;,&#34;maximum&#34;,&#34;canberra&#34;,&#34;minkowski&#34;) # plot : 是否绘图，以及图形类型 c(NULL, &#34;pdf&#34;, &#34;png&#34;, &#34;pngBMP&#34;) # writeTable: 是否保存文件 # seed：随机种子 results = ConsensusClusterPlus(d, maxK=10, reps=1000, pItem=0.8, pFeature=1, title=&#34;./tmp/&#34;, clusterAlg=&#34;hc&#34;, distance=&#34;pearson&#34;, plot=&#34;pdf&#34;, writeTable=FALSE, seed=42) R对象结果：list格式 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 ## 查看聚类数为2的结果信息 names(results[[3]]) # [1] &#34;consensusMatrix&#34; &#34;consensusTree&#34; &#34;consensusClass&#34; &#34;ml&#34; &#34;clrs&#34; table(results[[3]]$consensusClass) # 1 2 3 # 69 28 31 dim(results[[3]]$consensusMatrix) # [1] 128 128 results[[3]]$consensusMatrix[1:4,1:4] # [,1] [,2] [,3] [,4] # [1,] 1.0000000 0.3408360 0.7996870 0.3515249 # [2,] 0.3408360 1.0000000 0.1224806 1.0000000 # [3,] 0.7996870 0.1224806 1.0000000 0.1190108 # [4,] 0.3515249 1.0000000 0.1190108 1.0000000 ## consensus values 0 (never clustered together) ## consensus values 1 (always clustered together) 图形输出结果，主要两类图 （1）特定聚类结果的热图
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-04-20 00:00:00 &#43;0000 UTC&#39;&gt;2023-04-20&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-04-20&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;905&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to 肿瘤亚型分群工具ConsensusClusterPlus" href="https://lishensuo.github.io/en/posts/bioinfo/077%E8%82%BF%E7%98%A4%E4%BA%9A%E5%9E%8B%E5%88%86%E7%BE%A4%E5%B7%A5%E5%85%B7consensusclusterplus/"></a>
</article>





<article class="post-entry tag-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>使用IsoformSwitchAnalyzeR包进行可变剪切分析
			</h2>
		</header>
		<section class="entry-content">
			<p> https://bioconductor.org/packages/release/bioc/html/IsoformSwitchAnalyzeR.html 1、背景知识 The usage of Alternative Transcription Start sites (aTSS可选择转录起始位点), Alternative Splicing (AS可选择剪切位点) and alternative Transcription Termination Sites (aTTS可选择终止位点) are collectively collectively results in the production of different isoforms. Alternative isoforms are widely used as recently demonstrated by The ENCODE Consortium, which found that on average, 6.3 different transcripts are generated per gene; a number which may vary considerably per gene.
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-16 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-16&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-04-16&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;3284&amp;nbsp;|&amp;nbsp;7 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to 使用IsoformSwitchAnalyzeR包进行可变剪切分析" href="https://lishensuo.github.io/en/posts/bioinfo/070%E4%BD%BF%E7%94%A8isoformswitchanalyzer%E5%8C%85%E8%BF%9B%E8%A1%8C%E5%8F%AF%E5%8F%98%E5%89%AA%E5%88%87%E5%88%86%E6%9E%90/"></a>
</article>





<article class="post-entry tag-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>诺模图或列线图绘制分析
			</h2>
		</header>
		<section class="entry-content">
			<p> 肿瘤预后类文章的常规步骤之一是绘制诺模图，并进行相关分析。以下总结了相关基础绘制工具。
示例数据集 1 2 3 4 5 6 library(survival) head(lung) # inst time status age sex ph.ecog ph.karno pat.karno meal.cal wt.loss # 1 3 306 2 74 1 1 90 100 1175 NA # 2 3 455 2 68 1 0 90 90 1225 15 # 3 3 1010 1 56 1 0 90 90 NA 15 1、rms包 参考用法：https://atm.amegroups.com/article/view/14736/15089 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 library(rms) # var.labels = c(age=&#34;Age in Years&#34;, # lac=&#34;lactate&#34;, # sex=&#34;Sex of the participant&#34;, # shock=&#34;shock&#34;, # y=&#34;outcome&#34;, # Y=&#34;ordinal&#34;) # label(data) = lapply(names(var.labels), # function(x) label(data[,x]) = var.labels[x]) mod.cox &lt;- cph(Surv(time,status) ~ ph.ecog&#43;sex&#43;age,lung, surv=TRUE) ddist &lt;- datadist(lung) options(datadist=&#39;ddist&#39;) surv.cox &lt;- Survival(mod.cox) # 定义函数 med &lt;- Quantile(mod.cox) surv &lt;- Survival(mod.cox) nom.cox &lt;- nomogram(mod.cox, # 根据total point进行特定函数计算 fun=list(function(x) surv.cox(365, x), # 一年生存率 function(x) med(lp=x, q=0.5)), # 中位生存时间 funlabel=c(&#34;200-Day Sur. Prob.&#34;, &#34;Median Survival Time&#34;), lp=F, # 不展示Linear Predictor conf.int=c(0.1,0.5) # 两个置信区间 ) plot(nom.cox, col.conf=c(&#39;red&#39;,&#39;green&#39;), # 置信区间的颜色 col.grid = c(&#34;grey30&#34;,&#34;grey&#34;) # 网格的颜色 ) # f = cph(Surv(time, status) ~ age &#43; sex &#43; ph.karno, data = lung, # x = T, y = T, sur = T) # pred_score = apply(lung, 1, function(x){ # pred=Predict(f, age=x[&#34;age&#34;], sex=x[&#34;sex&#34;], ph.karno=x[&#34;ph.karno&#34;]) # return(pred$yhat) # }) %&gt;% unlist() # summary(pred_score) 校准曲线
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-04-28 00:00:00 &#43;0000 UTC&#39;&gt;2023-04-28&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-04-28&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;585&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to 诺模图或列线图绘制分析" href="https://lishensuo.github.io/en/posts/bioinfo/078%E8%AF%BA%E6%A8%A1%E5%9B%BE%E6%88%96%E5%88%97%E7%BA%BF%E5%9B%BE%E7%BB%98%E5%88%B6%E5%88%86%E6%9E%90/"></a>
</article>





<article class="post-entry tag-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>EnhancedVolcano包绘制火山图
			</h2>
		</header>
		<section class="entry-content">
			<p> EnhancedVolcano包可根据差异分析结果，基于ggplot2绘图结构，方便地绘制美观的火山图，下面根据自己的理解小结下基本用法。
官方全面的教程：https://github.com/kevinblighe/EnhancedVolcano 示例差异基因数据 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 library(airway) library(magrittr) data(&#39;airway&#39;) airway$dex %&lt;&gt;% relevel(&#39;untrt&#39;) ens &lt;- rownames(airway) library(org.Hs.eg.db) symbols &lt;- mapIds(org.Hs.eg.db, keys = ens, column = c(&#39;SYMBOL&#39;), keytype = &#39;ENSEMBL&#39;) symbols &lt;- symbols[!is.na(symbols)] symbols &lt;- symbols[match(rownames(airway), names(symbols))] rownames(airway) &lt;- symbols keep &lt;- !is.na(rownames(airway)) airway &lt;- airway[keep,] library(&#39;DESeq2&#39;) dds &lt;- DESeqDataSet(airway, design = ~ cell &#43; dex) dds &lt;- DESeq(dds, betaPrior=FALSE) res &lt;- results(dds, contrast = c(&#39;dex&#39;,&#39;trt&#39;,&#39;untrt&#39;)) res &lt;- lfcShrink(dds, contrast = c(&#39;dex&#39;,&#39;trt&#39;,&#39;untrt&#39;), res=res, type = &#39;normal&#39;) res &lt;- as.data.frame(res) head(res) # baseMean log2FoldChange lfcSE stat pvalue padj # TSPAN6 710.0931707 -0.37807189 0.09851236 -3.8404448 0.0001228116 0.0009522932 # TNMD 0.0000000 NA NA NA NA NA # DPM1 521.2572396 0.19826365 0.10931684 1.8155169 0.0694445184 0.1910397405 # SCYL3 237.6068046 0.03234467 0.13821470 0.2371917 0.8125081096 0.9118161375 # C1orf112 58.0358739 -0.08835419 0.25056704 -0.3194810 0.7493618190 0.8773885438 # FGR 0.3194343 -0.08459224 0.15186225 -0.3948862 0.6929268648 NA 如上，只要包含包含基因名、差异倍数、P值三部分信息的差异结果就可以用于绘制火山图。
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2022-04-16 00:00:00 &#43;0000 UTC&#39;&gt;2022-04-16&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2022-04-16&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;980&amp;nbsp;|&amp;nbsp;2 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to EnhancedVolcano包绘制火山图" href="https://lishensuo.github.io/en/posts/bioinfo/080enhancedvolcano%E5%8C%85%E7%BB%98%E5%88%B6%E7%81%AB%E5%B1%B1%E5%9B%BE/"></a>
</article>





<article class="post-entry tag-entry">
	<div class="post-info">  
		<header class="entry-header">
			<h2>IOBR包肿瘤signature打分与免疫浸润分析
			</h2>
		</header>
		<section class="entry-content">
			<p> IOBR包集signature打分与免疫浸润分析为一体的肿瘤数据分析工具，由南方医科大学南方医院廖旺军教授，曾东强博士等人于2021年7月发表于Frontiers in Immunology，引用数以超过100余次。现根据其github教程学习其部分功能用法。
...</p>
		</section>
		<footer class="entry-footer">













Create:&amp;nbsp;&lt;span title=&#39;2023-05-15 00:00:00 &#43;0000 UTC&#39;&gt;2023-05-15&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-05-15&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1667&amp;nbsp;|&amp;nbsp;4 min&amp;nbsp;|&amp;nbsp;Lishensuo</footer>
	</div>   
  <a class="entry-link" aria-label="post link to IOBR包肿瘤signature打分与免疫浸润分析" href="https://lishensuo.github.io/en/posts/bioinfo/081iobr%E5%8C%85%E8%82%BF%E7%98%A4signature%E6%89%93%E5%88%86%E4%B8%8E%E5%85%8D%E7%96%AB%E6%B5%B8%E6%B6%A6%E5%88%86%E6%9E%90/"></a>
</article>
<footer class="page-footer">
  <nav class="pagination">
    <a class="prev" href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/">« Prev Page</a>
    <a class="next" href="https://lishensuo.github.io/en/tags/%E8%BD%AC%E5%BD%95%E7%BB%84/page/3/">Next Page »</a>
  </nav>
</footer>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
