<!DOCTYPE html>
<html lang="en" dir="auto">

<head><meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="robots" content="index, follow">

<link rel="icon" href="/favicon.ico" type="image/x-icon"> 
<title>单细胞分析工具--scanpy基础用法(anndata,loom格式) | Li&#39;s Bioinfo-Blog</title>
<meta name="keywords" content="scRNA-seq">
<meta name="description" content="
scanpy是单细胞数据分析的python工具包，将数据以Anndata对象的格式进行储存。如下将学习Anndata对象操作以及scanpy分析的基础用法。">
<meta name="author" content="Lishensuo">
<link rel="canonical" href="https://lishensuo.github.io/en/posts/bioinfo/023%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scanpy%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/">
<link crossorigin="anonymous" href="/assets/css/stylesheet.min.9e4de5e3ba61ea358168341aa7cdf70abfaafb7c697dfe8624af3ddff9a35c2f.css" integrity="sha256-nk3l47ph6jWBaDQap833Cr&#43;q&#43;3xpff6GJK893/mjXC8=" rel="preload stylesheet" as="style">
<script defer crossorigin="anonymous" src="/assets/js/highlight.min.555af97124d54bb1457985dd081b8f5616a48103aafeb30ac89fde835d65aa6c.js" integrity="sha256-VVr5cSTVS7FFeYXdCBuPVhakgQOq/rMKyJ/eg11lqmw="
    onload="hljs.initHighlightingOnLoad();"></script>
<link rel="icon" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="16x16" href="https://lishensuo.github.io/img/Q.gif">
<link rel="icon" type="image/png" sizes="32x32" href="https://lishensuo.github.io/img/Q.gif">
<link rel="apple-touch-icon" href="https://lishensuo.github.io/Q.gif">
<link rel="mask-icon" href="https://lishensuo.github.io/Q.gif">
<meta name="theme-color" content="#2e2e33">
<meta name="msapplication-TileColor" content="#2e2e33">
<link rel="alternate" hreflang="en" href="https://lishensuo.github.io/en/posts/bioinfo/023%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scanpy%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/">
<noscript>
    <style>
        #theme-toggle,
        .top-link {
            display: none;
        }

    </style>
    <style>
        @media (prefers-color-scheme: dark) {
            :root {
                --theme: rgb(29, 30, 32);
                --entry: rgb(46, 46, 51);
                --primary: rgb(218, 218, 219);
                --secondary: rgb(155, 156, 157);
                --tertiary: rgb(65, 66, 68);
                --content: rgb(196, 196, 197);
                --hljs-bg: rgb(46, 46, 51);
                --code-bg: rgb(55, 56, 62);
                --border: rgb(51, 51, 51);
            }

            .list {
                background: var(--theme);
            }

            .list:not(.dark)::-webkit-scrollbar-track {
                background: 0 0;
            }

            .list:not(.dark)::-webkit-scrollbar-thumb {
                border-color: var(--theme);
            }
        }

    </style>
</noscript><meta property="og:title" content="单细胞分析工具--scanpy基础用法(anndata,loom格式)" />
<meta property="og:description" content="
scanpy是单细胞数据分析的python工具包，将数据以Anndata对象的格式进行储存。如下将学习Anndata对象操作以及scanpy分析的基础用法。" />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://lishensuo.github.io/en/posts/bioinfo/023%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scanpy%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-01-23T00:00:00&#43;00:00" />
<meta property="article:modified_time" content="2023-01-23T00:00:00&#43;00:00" />

<meta name="twitter:card" content="summary"/>
<meta name="twitter:title" content="单细胞分析工具--scanpy基础用法(anndata,loom格式)"/>
<meta name="twitter:description" content="
scanpy是单细胞数据分析的python工具包，将数据以Anndata对象的格式进行储存。如下将学习Anndata对象操作以及scanpy分析的基础用法。"/>


<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position":  1 ,
      "name": "分类",
      "item": "https://lishensuo.github.io/en/posts/"
    }, 
    {
      "@type": "ListItem",
      "position":  2 ,
      "name": "📖 生信数据分析--分析流程，工具包等",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/"
    }, 
    {
      "@type": "ListItem",
      "position":  3 ,
      "name": "单细胞分析工具--scanpy基础用法(anndata,loom格式)",
      "item": "https://lishensuo.github.io/en/posts/bioinfo/023%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scanpy%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/"
    }
  ]
}
</script>
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "单细胞分析工具--scanpy基础用法(anndata,loom格式)",
  "name": "单细胞分析工具--scanpy基础用法(anndata,loom格式)",
  "description": " scanpy是单细胞数据分析的python工具包，将数据以Anndata对象的格式进行储存。如下将学习Anndata对象操作以及scanpy分析的基础用法。\n",
  "keywords": [
    "scRNA-seq"
  ],
  "articleBody": " scanpy是单细胞数据分析的python工具包，将数据以Anndata对象的格式进行储存。如下将学习Anndata对象操作以及scanpy分析的基础用法。\n1 2 3 import numpy as np import pandas as pd import anndata as ad 1、数据导入 （1）10X的三件套数据 1 2 3 4 5 6 7 8 9 10 # !wget http://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz # tar -xzf pbmc3k_filtered_gene_bc_matrices.tar.gz adata = sc.read_10x_mtx( 'filtered_gene_bc_matrices/hg19/', # the directory with the `.mtx` file var_names='gene_symbols') # use gene symbols for the variable names (variables-axis index) adata # AnnData object with n_obs × n_vars = 2700 × 32738 # var: 'gene_ids' （2）Seurat对象转换\nhttps://github.com/LuckyMD/Code_snippets/blob/master/Seurat_to_anndata.ipynb\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 ## 同样是上述示例数据，先用R包Seurat处理，储存结果 library(Seurat) pbmc.data = Read10X(data.dir = \"filtered_gene_bc_matrices/hg19/\") pbmc = CreateSeuratObject(counts = pbmc.data, project = \"pbmc3k\") pbmc = NormalizeData(pbmc) %\u003e% FindVariableFeatures() pbmc = pbmc %\u003e% ScaleData() %\u003e% RunPCA() %\u003e% RunUMAP(dims = 1:30) %\u003e% FindNeighbors(dims = 1:30) %\u003e% FindClusters(resolution = 0.1) head(pbmc@meta.data) saveRDS(pbmc, file = \"pbmc.rds\") 1 2 3 4 5 6 7 8 9 ## 准备Python环境 import scanpy as sc import numpy as np import anndata2ri # Activate the anndata2ri conversion between SingleCellExperiment and AnnData anndata2ri.activate() #Loading the rpy2 extension enables cell magic to be used #This runs R code in jupyter notebook cells %load_ext rpy2.ipython 1 2 3 4 5 ## 读入对象 %%R library(Seurat) pbmc = readRDS(\"pbmc.rds\") pbmc 1 2 3 4 5 ## 转换对象 %%R -o pbmc_sce #convert the Seurat object to a SingleCellExperiment object pbmc_sce \u003c- as.SingleCellExperiment(pbmc) pbmc_sce 1 2 3 4 5 6 ## 查看转换后的anndata对象 pbmc_sce # AnnData object with n_obs × n_vars = 2700 × 32738 # obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'RNA_snn_res.0.1', 'seurat_clusters', 'ident' # obsm: 'X_pca', 'X_umap' # layers: 'logcounts' 2、Anndata结构 https://anndata-tutorials.readthedocs.io/en/latest/getting-started.html\n以上述1.2的pbmc_sce对象为例学习anndata结构组成\n1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 adata = pbmc_sce # (1) 矩阵 # ！注意：行是细胞，列是基因。与Seurat读取结果相反！ adata.obs_names #行名(细胞名) adata.var_names #列名(基因名) ## 原始信息以稀疏矩阵形式储存 adata.X print(adata.X) adata.X.todense() adata.layers[\"log_transformed\"] = np.log1p(adata.X) # adata.layers[\"counts\"] = adata.X.copy() adata.to_df(layer=\"log_transformed\") adata.layers[\"X\"] = adata.X adata.to_df(layer=\"X\") # (2) metadata，按pd.Dataframe格式储存 adata.obs # 细胞的metadata(常用) adata.var # 基因的metadata # (3) 降维数据：多是数组格式，按字典逻辑储存 adata.obsm adata.obsm.keys adata.obsm[\"X_umap\"] # (4) subset子集 adata[[\"Cell_1\", \"Cell_10\"], [\"Gene_5\", \"Gene_1900\"]] adata[adata.obs.cell_type == \"B\"] 3、scanpy基础分析 https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html\n以上述1.1的原始读取结果adata对象为例\n（1）质控过滤 1 2 3 4 5 6 7 8 9 10 11 sc.pp.filter_cells(adata, min_genes=200) sc.pp.filter_genes(adata, min_cells=3) adata.var['mt'] = adata.var_names.str.startswith('MT-') sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True) # sc.pl.violin(adata, # ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'], # jitter=0.4, multi_panel=True) adata = adata[adata.obs.n_genes_by_counts \u003c 2500, :] adata = adata[adata.obs.pct_counts_mt \u003c 5, :] （2）标准化 1 2 3 4 sc.pp.normalize_total(adata, target_sum=1e4) sc.pp.log1p(adata) sc.pp.scale(adata, max_value=10) sc.pp.regress_out(adata, ['total_counts', 'pct_counts_mt']) （3）降维分群 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 ## 高变基因 sc.pp.highly_variable_genes(adata) adata.var ## PCA降维 sc.tl.pca(adata) adata.obsm[\"X_pca\"].shape sc.pl.pca(adata) # 默认按X_pca的降维结果可视化 sc.pl.pca(adata, color='CST3') ## UMAP降维 sc.pp.neighbors(adata, n_neighbors=10, n_pcs=40) sc.tl.umap(adata) sc.pl.umap(adata) ## 分群 sc.tl.leiden(adata, resolution=0.5,key_added=\"cluster\") adata.obs sc.pl.umap(adata, color=['cluster']) （4）保存 1 adata.write(\"your_data.h5ad\") 2024-12-15\nLoompy 是一个用于管理、存储和处理大规模单细胞基因表达数据的 Python 库，可以与Scanpy，Seurat等兼容。其中Loom 格式专门用于高效地存储稀疏矩阵及其相关的元数据。\nhttps://linnarssonlab.org/loompy/ https://github.com/linnarsson-lab/loompy 下面学习其基本用法—\n1 2 3 import scanpy as sc import loompy import numpy as np anndata保存为loom文件 1 2 3 4 5 6 7 8 9 adata = sc.read_h5ad(\"demo_adata.h5ad\") # 添加一个Layer adata.layers[\"count\"] = adata.X.copy() # 选择需要保存的row/col metadata adata.var = adata.var[[\"feature_id\"]] adata.obs = adata.obs[[\"raw_sum\"]] # 最后保存： adata.write_loom(\"./demo_adata.loom\") 与anndata的两点区别：\nloom储存的矩阵，row代表基因，column代表细胞。这anndata正好相反 loom储存更加高效，占用硬盘空间相比h5df格式，可显著节约内存。 scanpy读取loom文件 直接读取为anndata对象\n1 2 adata = sc.read_loom(\"./demo_adata.loom\") adata loompy读取loom文件 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 # (1) 两种方式读取，后者适用于在交互式环境操作 with loompy.connect(\"filename.loom\") as ds: # do something with the connection object ds ds = loompy.connect(\"filename.loom\") # do something with the connection object ds ds.close() # (2) Layers ds.layers.keys() # ['', 'count'] ds[:4,:4] ds[\"count\"][:4,:4] # (3) row/col attribute ds.ra.keys() # ['ensembl_id', 'var_names'] ds.ca.keys() # ['n_counts', 'obs_names'] ds[:, ds.ca.n_counts \u003c 3000].shape # (4) map func ds.map([np.mean, np.std], axis=1) # (5) scan: 对每个batch细胞分别处理 for (ix, selection, view) in ds.scan(axis=1, batch_size=1024): print(view.shape) break ds.close() 在使用过程中，遇到numpy\u003e2.0的版本冲突问题。根据提示，修改源码即可。\n",
  "wordCount" : "1357",
  "inLanguage": "en",
  "datePublished": "2023-01-23T00:00:00Z",
  "dateModified": "2023-01-23T00:00:00Z",
  "author":[{
    "@type": "Person",
    "name": "Lishensuo"
  }],
  "mainEntityOfPage": {
    "@type": "WebPage",
    "@id": "https://lishensuo.github.io/en/posts/bioinfo/023%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--scanpy%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95/"
  },
  "publisher": {
    "@type": "Organization",
    "name": "Li's Bioinfo-Blog",
    "logo": {
      "@type": "ImageObject",
      "url": "https://lishensuo.github.io/img/Q.gif"
    }
  }
}
</script>
<script async src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script>
</head>

<body class="" id="top">
<script>
    if (localStorage.getItem("pref-theme") === "dark") {
        document.body.classList.add('dark');
    } else if (localStorage.getItem("pref-theme") === "light") {
        document.body.classList.remove('dark')
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
        document.body.classList.add('dark');
    }

</script>

<header class="header">
    <nav class="nav">
        <div class="logo">
            <a href="https://lishensuo.github.io/en/" accesskey="h" title="Li&#39;s Bioinfo-Blog (Alt + H)">Li&#39;s Bioinfo-Blog</a>
            <span class="logo-switches">
                <button id="theme-toggle" accesskey="t" title="(Alt + T)">
                    <svg id="moon" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <path d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"></path>
                    </svg>
                    <svg id="sun" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"
                        fill="none" stroke="currentColor" stroke-width="2" stroke-linecap="round"
                        stroke-linejoin="round">
                        <circle cx="12" cy="12" r="5"></circle>
                        <line x1="12" y1="1" x2="12" y2="3"></line>
                        <line x1="12" y1="21" x2="12" y2="23"></line>
                        <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
                        <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
                        <line x1="1" y1="12" x2="3" y2="12"></line>
                        <line x1="21" y1="12" x2="23" y2="12"></line>
                        <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
                        <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
                    </svg>
                </button>
                <ul class="lang-switch"><li>|</li>
                </ul>
            </span>
        </div>
        <ul id="menu">
            <li>
                <a href="https://lishensuo.github.io/en/" title="主页">
                    <span>主页</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/posts" title="分类">
                    <span>分类</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/tags" title="标签">
                    <span>标签</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/archives/" title="归档">
                    <span>归档</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/about" title="关于">
                    <span>关于</span>
                </a>
            </li>
            <li>
                <a href="https://lishensuo.github.io/en/search" title="搜索 (Alt &#43; /)" accesskey=/>
                    <span>搜索</span>
                </a>
            </li>
        </ul>
    </nav>
</header>
<main class="main">

<article class="post-single">
  <header class="post-header">
    <div class="breadcrumbs"><a href="https://lishensuo.github.io/en/">Home</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/">分类</a>&nbsp;»&nbsp;<a href="https://lishensuo.github.io/en/posts/bioinfo/">📖 生信数据分析--分析流程，工具包等</a></div>
    <h1 class="post-title">
      单细胞分析工具--scanpy基础用法(anndata,loom格式)
    </h1>
    <div class="post-meta">













Create:&amp;nbsp;&lt;span title=&#39;2023-01-23 00:00:00 &#43;0000 UTC&#39;&gt;2023-01-23&lt;/span&gt;&amp;nbsp;|&amp;nbsp;Update:&amp;nbsp;2023-01-23&amp;nbsp;|&amp;nbsp;Words:&amp;nbsp;1357&amp;nbsp;|&amp;nbsp;3 min&amp;nbsp;|&amp;nbsp;Lishensuo

|  Viewers: <span id="busuanzi_value_page_pv"></span> 
	  
    </div>
  </header> <aside id="toc-container" class="toc-container wide">
    <div class="toc">
        <details  open>
            <summary accesskey="c" title="(Alt + C)">
                <span class="details">Table of Contents</span>
            </summary>

            <div class="inner"><ul>
                    <li>
                        <a href="#1%e6%95%b0%e6%8d%ae%e5%af%bc%e5%85%a5" aria-label="1、数据导入">1、数据导入</a></li>
                    <li>
                        <a href="#2anndata%e7%bb%93%e6%9e%84" aria-label="2、Anndata结构">2、Anndata结构</a></li>
                    <li>
                        <a href="#3scanpy%e5%9f%ba%e7%a1%80%e5%88%86%e6%9e%90" aria-label="3、scanpy基础分析">3、scanpy基础分析</a>
                    </li>
                </ul>
            </div>
        </details>
    </div>
</aside>
<script>
    let activeElement;
    let elements;
    window.addEventListener('DOMContentLoaded', function (event) {
        checkTocPosition();

        elements = document.querySelectorAll('h1[id],h2[id],h3[id],h4[id],h5[id],h6[id]');
         
         activeElement = elements[0];
         const id = encodeURI(activeElement.getAttribute('id')).toLowerCase();
         document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
     }, false);

    window.addEventListener('resize', function(event) {
        checkTocPosition();
    }, false);

    window.addEventListener('scroll', () => {
        
        activeElement = Array.from(elements).find((element) => {
            if ((getOffsetTop(element) - window.pageYOffset) > 0 && 
                (getOffsetTop(element) - window.pageYOffset) < window.innerHeight/2) {
                return element;
            }
        }) || activeElement

        elements.forEach(element => {
             const id = encodeURI(element.getAttribute('id')).toLowerCase();
             if (element === activeElement){
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.add('active');
             } else {
                 document.querySelector(`.inner ul li a[href="#${id}"]`).classList.remove('active');
             }
         })
     }, false);

    const main = parseInt(getComputedStyle(document.body).getPropertyValue('--article-width'), 10);
    const toc = parseInt(getComputedStyle(document.body).getPropertyValue('--toc-width'), 10);
    const gap = parseInt(getComputedStyle(document.body).getPropertyValue('--gap'), 10);

    function checkTocPosition() {
        const width = document.body.scrollWidth;

        if (width - main - (toc * 2) - (gap * 4) > 0) {
            document.getElementById("toc-container").classList.add("wide");
        } else {
            document.getElementById("toc-container").classList.remove("wide");
        }
    }

    function getOffsetTop(element) {
        if (!element.getClientRects().length) {
            return 0;
        }
        let rect = element.getBoundingClientRect();
        let win = element.ownerDocument.defaultView;
        return rect.top + win.pageYOffset;   
    }
</script>


  <div class="post-content"><blockquote>
<p>scanpy是单细胞数据分析的python工具包，将数据以Anndata对象的格式进行储存。如下将学习Anndata对象操作以及scanpy分析的基础用法。</p></blockquote>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> numpy <span style="color:#fff;font-weight:bold">as</span> np
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> pandas <span style="color:#fff;font-weight:bold">as</span> pd
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> anndata <span style="color:#fff;font-weight:bold">as</span> ad
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="1数据导入">1、数据导入<a hidden class="anchor" aria-hidden="true" href="#1数据导入">#</a></h2>
<ul>
<li>（1）10X的三件套数据</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f"># !wget http://cf.10xgenomics.com/samples/cell-exp/1.1.0/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># tar -xzf pbmc3k_filtered_gene_bc_matrices.tar.gz</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>adata = sc.read_10x_mtx(
</span></span><span style="display:flex;"><span>    <span style="color:#0ff;font-weight:bold">&#39;filtered_gene_bc_matrices/hg19/&#39;</span>,  <span style="color:#007f7f"># the directory with the `.mtx` file</span>
</span></span><span style="display:flex;"><span>    var_names=<span style="color:#0ff;font-weight:bold">&#39;gene_symbols&#39;</span>)           <span style="color:#007f7f"># use gene symbols for the variable names (variables-axis index)</span>
</span></span><span style="display:flex;"><span>adata
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AnnData object with n_obs × n_vars = 2700 × 32738</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     var: &#39;gene_ids&#39;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>
<p>（2）Seurat对象转换</p>
<p><a href="https://github.com/LuckyMD/Code_snippets/blob/master/Seurat_to_anndata.ipynb">https://github.com/LuckyMD/Code_snippets/blob/master/Seurat_to_anndata.ipynb</a></p>
</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-R" data-lang="R"><span style="display:flex;"><span><span style="color:#007f7f">## 同样是上述示例数据，先用R包Seurat处理，储存结果</span>
</span></span><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>pbmc.data = Read10X(data.dir = <span style="color:#0ff;font-weight:bold">&#34;filtered_gene_bc_matrices/hg19/&#34;</span>)
</span></span><span style="display:flex;"><span>pbmc = CreateSeuratObject(counts = pbmc.data, project = <span style="color:#0ff;font-weight:bold">&#34;pbmc3k&#34;</span>)
</span></span><span style="display:flex;"><span>pbmc = NormalizeData(pbmc) %&gt;% 
</span></span><span style="display:flex;"><span>  FindVariableFeatures()
</span></span><span style="display:flex;"><span>pbmc = pbmc %&gt;% 
</span></span><span style="display:flex;"><span>  ScaleData() %&gt;% 
</span></span><span style="display:flex;"><span>  RunPCA() %&gt;% 
</span></span><span style="display:flex;"><span>  RunUMAP(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;% 
</span></span><span style="display:flex;"><span>  FindNeighbors(dims = <span style="color:#ff0;font-weight:bold">1</span>:<span style="color:#ff0;font-weight:bold">30</span>) %&gt;%
</span></span><span style="display:flex;"><span>  FindClusters(resolution = <span style="color:#ff0;font-weight:bold">0.1</span>)
</span></span><span style="display:flex;"><span>head(pbmc@meta.data)
</span></span><span style="display:flex;"><span>saveRDS(pbmc, file = <span style="color:#0ff;font-weight:bold">&#34;pbmc.rds&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f">## 准备Python环境</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> scanpy <span style="color:#fff;font-weight:bold">as</span> sc
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> numpy <span style="color:#fff;font-weight:bold">as</span> np
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> anndata2ri
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># Activate the anndata2ri conversion between SingleCellExperiment and AnnData</span>
</span></span><span style="display:flex;"><span>anndata2ri.activate()
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#Loading the rpy2 extension enables cell magic to be used</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#This runs R code in jupyter notebook cells</span>
</span></span><span style="display:flex;"><span>%load_ext rpy2.ipython
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f">## 读入对象</span>
</span></span><span style="display:flex;"><span>%%R
</span></span><span style="display:flex;"><span>library(Seurat)
</span></span><span style="display:flex;"><span>pbmc = readRDS(<span style="color:#0ff;font-weight:bold">&#34;pbmc.rds&#34;</span>)
</span></span><span style="display:flex;"><span>pbmc
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f">## 转换对象</span>
</span></span><span style="display:flex;"><span>%%R -o pbmc_sce
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#convert the Seurat object to a SingleCellExperiment object</span>
</span></span><span style="display:flex;"><span>pbmc_sce &lt;- <span style="color:#fff;font-weight:bold">as</span>.SingleCellExperiment(pbmc)
</span></span><span style="display:flex;"><span>pbmc_sce
</span></span></code></pre></td></tr></table>
</div>
</div><div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f">## 查看转换后的anndata对象</span>
</span></span><span style="display:flex;"><span>pbmc_sce
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># AnnData object with n_obs × n_vars = 2700 × 32738</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obs: &#39;orig.ident&#39;, &#39;nCount_RNA&#39;, &#39;nFeature_RNA&#39;, &#39;RNA_snn_res.0.1&#39;, &#39;seurat_clusters&#39;, &#39;ident&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     obsm: &#39;X_pca&#39;, &#39;X_umap&#39;</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#     layers: &#39;logcounts&#39;</span>
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="2anndata结构">2、Anndata结构<a hidden class="anchor" aria-hidden="true" href="#2anndata结构">#</a></h2>
<p><a href="https://anndata-tutorials.readthedocs.io/en/latest/getting-started.html">https://anndata-tutorials.readthedocs.io/en/latest/getting-started.html</a></p>
<p>以上述1.2的<code>pbmc_sce</code>对象为例学习anndata结构组成</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>adata = pbmc_sce
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (1) 矩阵</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># ！注意：行是细胞，列是基因。与Seurat读取结果相反！</span>
</span></span><span style="display:flex;"><span>adata.obs_names <span style="color:#007f7f">#行名(细胞名)</span>
</span></span><span style="display:flex;"><span>adata.var_names <span style="color:#007f7f">#列名(基因名)</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 原始信息以稀疏矩阵形式储存</span>
</span></span><span style="display:flex;"><span>adata.X
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">print</span>(adata.X)
</span></span><span style="display:flex;"><span>adata.X.todense()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>adata.layers[<span style="color:#0ff;font-weight:bold">&#34;log_transformed&#34;</span>] = np.log1p(adata.X)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># adata.layers[&#34;counts&#34;] = adata.X.copy()</span>
</span></span><span style="display:flex;"><span>adata.to_df(layer=<span style="color:#0ff;font-weight:bold">&#34;log_transformed&#34;</span>)
</span></span><span style="display:flex;"><span>adata.layers[<span style="color:#0ff;font-weight:bold">&#34;X&#34;</span>] = adata.X
</span></span><span style="display:flex;"><span>adata.to_df(layer=<span style="color:#0ff;font-weight:bold">&#34;X&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2) metadata，按pd.Dataframe格式储存</span>
</span></span><span style="display:flex;"><span>adata.obs  <span style="color:#007f7f"># 细胞的metadata(常用)</span>
</span></span><span style="display:flex;"><span>adata.var  <span style="color:#007f7f"># 基因的metadata</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (3) 降维数据：多是数组格式，按字典逻辑储存</span>
</span></span><span style="display:flex;"><span>adata.obsm
</span></span><span style="display:flex;"><span>adata.obsm.keys
</span></span><span style="display:flex;"><span>adata.obsm[<span style="color:#0ff;font-weight:bold">&#34;X_umap&#34;</span>]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (4) subset子集</span>
</span></span><span style="display:flex;"><span>adata[[<span style="color:#0ff;font-weight:bold">&#34;Cell_1&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Cell_10&#34;</span>], [<span style="color:#0ff;font-weight:bold">&#34;Gene_5&#34;</span>, <span style="color:#0ff;font-weight:bold">&#34;Gene_1900&#34;</span>]]
</span></span><span style="display:flex;"><span>adata[adata.obs.cell_type == <span style="color:#0ff;font-weight:bold">&#34;B&#34;</span>]
</span></span></code></pre></td></tr></table>
</div>
</div><h2 id="3scanpy基础分析">3、scanpy基础分析<a hidden class="anchor" aria-hidden="true" href="#3scanpy基础分析">#</a></h2>
<p><a href="https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html">https://scanpy-tutorials.readthedocs.io/en/latest/pbmc3k.html</a></p>
<p>以上述1.1的原始读取结果<code>adata</code>对象为例</p>
<ul>
<li>（1）质控过滤</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>sc.pp.filter_cells(adata, min_genes=<span style="color:#ff0;font-weight:bold">200</span>)
</span></span><span style="display:flex;"><span>sc.pp.filter_genes(adata, min_cells=<span style="color:#ff0;font-weight:bold">3</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>adata.var[<span style="color:#0ff;font-weight:bold">&#39;mt&#39;</span>] = adata.var_names.str.startswith(<span style="color:#0ff;font-weight:bold">&#39;MT-&#39;</span>) 
</span></span><span style="display:flex;"><span>sc.pp.calculate_qc_metrics(adata, qc_vars=[<span style="color:#0ff;font-weight:bold">&#39;mt&#39;</span>], 
</span></span><span style="display:flex;"><span>                           percent_top=<span style="color:#fff;font-weight:bold">None</span>, log1p=<span style="color:#fff;font-weight:bold">False</span>, inplace=<span style="color:#fff;font-weight:bold">True</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># sc.pl.violin(adata, </span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 			 [&#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;pct_counts_mt&#39;],</span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">#              jitter=0.4, multi_panel=True)</span>
</span></span><span style="display:flex;"><span>adata = adata[adata.obs.n_genes_by_counts &lt; <span style="color:#ff0;font-weight:bold">2500</span>, :]
</span></span><span style="display:flex;"><span>adata = adata[adata.obs.pct_counts_mt &lt; <span style="color:#ff0;font-weight:bold">5</span>, :]
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（2）标准化</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>sc.pp.normalize_total(adata, target_sum=<span style="color:#ff0;font-weight:bold">1e4</span>)
</span></span><span style="display:flex;"><span>sc.pp.log1p(adata)
</span></span><span style="display:flex;"><span>sc.pp.scale(adata, max_value=<span style="color:#ff0;font-weight:bold">10</span>)
</span></span><span style="display:flex;"><span>sc.pp.regress_out(adata, [<span style="color:#0ff;font-weight:bold">&#39;total_counts&#39;</span>, <span style="color:#0ff;font-weight:bold">&#39;pct_counts_mt&#39;</span>])
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（3）降维分群</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f">## 高变基因</span>
</span></span><span style="display:flex;"><span>sc.pp.highly_variable_genes(adata)
</span></span><span style="display:flex;"><span>adata.var
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## PCA降维</span>
</span></span><span style="display:flex;"><span>sc.tl.pca(adata)
</span></span><span style="display:flex;"><span>adata.obsm[<span style="color:#0ff;font-weight:bold">&#34;X_pca&#34;</span>].shape 
</span></span><span style="display:flex;"><span>sc.pl.pca(adata) <span style="color:#007f7f"># 默认按X_pca的降维结果可视化</span>
</span></span><span style="display:flex;"><span>sc.pl.pca(adata, color=<span style="color:#0ff;font-weight:bold">&#39;CST3&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## UMAP降维</span>
</span></span><span style="display:flex;"><span>sc.pp.neighbors(adata, n_neighbors=<span style="color:#ff0;font-weight:bold">10</span>, n_pcs=<span style="color:#ff0;font-weight:bold">40</span>)
</span></span><span style="display:flex;"><span>sc.tl.umap(adata)
</span></span><span style="display:flex;"><span>sc.pl.umap(adata)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f">## 分群</span>
</span></span><span style="display:flex;"><span>sc.tl.leiden(adata, resolution=<span style="color:#ff0;font-weight:bold">0.5</span>,key_added=<span style="color:#0ff;font-weight:bold">&#34;cluster&#34;</span>)
</span></span><span style="display:flex;"><span>adata.obs
</span></span><span style="display:flex;"><span>sc.pl.umap(adata, color=[<span style="color:#0ff;font-weight:bold">&#39;cluster&#39;</span>])
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li>（4）保存</li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>adata.write(<span style="color:#0ff;font-weight:bold">&#34;your_data.h5ad&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><hr>
<p>2024-12-15</p>
<p>Loompy 是一个用于管理、存储和处理大规模单细胞基因表达数据的 Python 库，可以与Scanpy，Seurat等兼容。其中Loom 格式专门用于高效地存储稀疏矩阵及其相关的元数据。</p>
<ul>
<li><a href="https://linnarssonlab.org/loompy/">https://linnarssonlab.org/loompy/</a></li>
<li><a href="https://github.com/linnarsson-lab/loompy">https://github.com/linnarsson-lab/loompy</a></li>
</ul>
<p><img loading="lazy" src="https://raw.githubusercontent.com/lishensuo/images2/main/img01/image-20241215101650062.png" alt="image-20241215101650062"  />
</p>
<p>下面学习其基本用法&mdash;</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> scanpy <span style="color:#fff;font-weight:bold">as</span> sc
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> loompy
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">import</span> numpy <span style="color:#fff;font-weight:bold">as</span> np
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li><strong>anndata保存为loom文件</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">9
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>adata = sc.read_h5ad(<span style="color:#0ff;font-weight:bold">&#34;demo_adata.h5ad&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 添加一个Layer</span>
</span></span><span style="display:flex;"><span>adata.layers[<span style="color:#0ff;font-weight:bold">&#34;count&#34;</span>] = adata.X.copy()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 选择需要保存的row/col metadata</span>
</span></span><span style="display:flex;"><span>adata.var = adata.var[[<span style="color:#0ff;font-weight:bold">&#34;feature_id&#34;</span>]]
</span></span><span style="display:flex;"><span>adata.obs = adata.obs[[<span style="color:#0ff;font-weight:bold">&#34;raw_sum&#34;</span>]]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># 最后保存：</span>
</span></span><span style="display:flex;"><span>adata.write_loom(<span style="color:#0ff;font-weight:bold">&#34;./demo_adata.loom&#34;</span>)
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>与anndata的两点区别：</p>
<ol>
<li>loom储存的矩阵，row代表基因，column代表细胞。这anndata正好相反</li>
<li>loom储存更加高效，占用硬盘空间相比h5df格式，可显著节约内存。</li>
</ol></blockquote>
<ul>
<li><strong>scanpy读取loom文件</strong></li>
</ul>
<p>直接读取为anndata对象</p>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">2
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span>adata = sc.read_loom(<span style="color:#0ff;font-weight:bold">&#34;./demo_adata.loom&#34;</span>)
</span></span><span style="display:flex;"><span>adata
</span></span></code></pre></td></tr></table>
</div>
</div><ul>
<li><strong>loompy读取loom文件</strong></li>
</ul>
<div class="highlight"><div style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;">
<table style="border-spacing:0;padding:0;margin:0;border:0;"><tr><td style="vertical-align:top;padding:0;margin:0;border:0;">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 1
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 2
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 3
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 4
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 5
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 6
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 7
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 8
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272"> 9
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">10
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">11
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">12
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">13
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">14
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">15
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">16
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">17
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">18
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">19
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">20
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">21
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">22
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">23
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">24
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">25
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">26
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">27
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">28
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">29
</span><span style="white-space:pre;-webkit-user-select:none;user-select:none;margin-right:0.4em;padding:0 0.4em 0 0.4em;color:#727272">30
</span></code></pre></td>
<td style="vertical-align:top;padding:0;margin:0;border:0;;width:100%">
<pre tabindex="0" style="color:#e5e5e5;background-color:#000;-moz-tab-size:4;-o-tab-size:4;tab-size:4;"><code class="language-python" data-lang="python"><span style="display:flex;"><span><span style="color:#007f7f"># (1) 两种方式读取，后者适用于在交互式环境操作</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">with</span> loompy.connect(<span style="color:#0ff;font-weight:bold">&#34;filename.loom&#34;</span>) <span style="color:#fff;font-weight:bold">as</span> ds:
</span></span><span style="display:flex;"><span>    <span style="color:#007f7f"># do something with the connection object ds</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>ds = loompy.connect(<span style="color:#0ff;font-weight:bold">&#34;filename.loom&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># do something with the connection object ds</span>
</span></span><span style="display:flex;"><span>ds.close()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (2) Layers</span>
</span></span><span style="display:flex;"><span>ds.layers.keys()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;&#39;, &#39;count&#39;]</span>
</span></span><span style="display:flex;"><span>ds[:<span style="color:#ff0;font-weight:bold">4</span>,:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span>ds[<span style="color:#0ff;font-weight:bold">&#34;count&#34;</span>][:<span style="color:#ff0;font-weight:bold">4</span>,:<span style="color:#ff0;font-weight:bold">4</span>]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (3) row/col attribute</span>
</span></span><span style="display:flex;"><span>ds.ra.keys()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;ensembl_id&#39;, &#39;var_names&#39;]</span>
</span></span><span style="display:flex;"><span>ds.ca.keys()
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># [&#39;n_counts&#39;, &#39;obs_names&#39;]</span>
</span></span><span style="display:flex;"><span>ds[:, ds.ca.n_counts &lt; <span style="color:#ff0;font-weight:bold">3000</span>].shape
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (4) map func</span>
</span></span><span style="display:flex;"><span>ds.map([np.mean, np.std], axis=<span style="color:#ff0;font-weight:bold">1</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#007f7f"># (5) scan: 对每个batch细胞分别处理</span>
</span></span><span style="display:flex;"><span><span style="color:#fff;font-weight:bold">for</span> (ix, selection, view) in ds.scan(axis=<span style="color:#ff0;font-weight:bold">1</span>, batch_size=<span style="color:#ff0;font-weight:bold">1024</span>):
</span></span><span style="display:flex;"><span>    <span style="color:#fff;font-weight:bold">print</span>(view.shape)
</span></span><span style="display:flex;"><span>    <span style="color:#fff;font-weight:bold">break</span>
</span></span><span style="display:flex;"><span>    
</span></span><span style="display:flex;"><span>ds.close()
</span></span></code></pre></td></tr></table>
</div>
</div><blockquote>
<p>在使用过程中，遇到numpy&gt;2.0的版本冲突问题。根据提示，修改源码即可。</p></blockquote>


  </div>

  <footer class="post-footer">
    <ul class="post-tags">
      <li><a href="https://lishensuo.github.io/en/tags/scrna-seq/">ScRNA-Seq</a></li>
    </ul>
<nav class="paginav">
  <a class="prev" href="https://lishensuo.github.io/en/posts/bioinfo/018%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--infercnv%E6%8B%B7%E8%B4%9D%E6%95%B0%E5%8F%98%E5%BC%82%E9%89%B4%E5%AE%9A/">
    <span class="title">« Prev Page</span>
    <br>
    <span>单细胞分析工具--infercnv拷贝数变异鉴定</span>
  </a>
  <a class="next" href="https://lishensuo.github.io/en/posts/bioinfo/019%E5%8D%95%E7%BB%86%E8%83%9E%E5%88%86%E6%9E%90%E5%B7%A5%E5%85%B7--singler%E7%BB%86%E8%83%9E%E7%B1%BB%E5%9E%8B%E6%B3%A8%E9%87%8A/">
    <span class="title">Next Page »</span>
    <br>
    <span>单细胞分析工具--SingleR细胞类型注释</span>
  </a>
</nav>

  </footer>
</article>
    </main>
    
<footer class="footer">
    <span>&copy; 2025 <a href="https://lishensuo.github.io/en/">Li&#39;s Bioinfo-Blog</a></span>
    <span>
        Powered by
        <a href="https://gohugo.io/" rel="noopener noreferrer" target="_blank">Hugo</a> &
        <a href="https://git.io/hugopapermod" rel="noopener" target="_blank">PaperMod</a>
		<br/>您是本站第 <span id="busuanzi_value_site_uv"></span> 位访问者，总浏览量为 <span id="busuanzi_value_site_pv"></span> 次
    </span>
</footer>
<a href="#top" aria-label="go to top" title="Go to Top (Alt + G)" class="top-link" id="top-link" accesskey="g">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 6" fill="currentColor">
        <path d="M12 6H0l6-6z" />
    </svg>
</a>

<script>
    let menu = document.getElementById('menu')
    if (menu) {
        menu.scrollLeft = localStorage.getItem("menu-scroll-position");
        menu.onscroll = function () {
            localStorage.setItem("menu-scroll-position", menu.scrollLeft);
        }
    }

    document.querySelectorAll('a[href^="#"]').forEach(anchor => {
        anchor.addEventListener("click", function (e) {
            e.preventDefault();
            var id = this.getAttribute("href").substr(1);
            if (!window.matchMedia('(prefers-reduced-motion: reduce)').matches) {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView({
                    behavior: "smooth"
                });
            } else {
                document.querySelector(`[id='${decodeURIComponent(id)}']`).scrollIntoView();
            }
            if (id === "top") {
                history.replaceState(null, null, " ");
            } else {
                history.pushState(null, null, `#${id}`);
            }
        });
    });

</script>
<script>
    var mybutton = document.getElementById("top-link");
    window.onscroll = function () {
        if (document.body.scrollTop > 800 || document.documentElement.scrollTop > 800) {
            mybutton.style.visibility = "visible";
            mybutton.style.opacity = "1";
        } else {
            mybutton.style.visibility = "hidden";
            mybutton.style.opacity = "0";
        }
    };

</script>
<script>
    document.getElementById("theme-toggle").addEventListener("click", () => {
        if (document.body.className.includes("dark")) {
            document.body.classList.remove('dark');
            localStorage.setItem("pref-theme", 'light');
        } else {
            document.body.classList.add('dark');
            localStorage.setItem("pref-theme", 'dark');
        }
    })

</script>
<script>
    document.querySelectorAll('pre > code').forEach((codeblock) => {
        const container = codeblock.parentNode.parentNode;

        const copybutton = document.createElement('button');
        copybutton.classList.add('copy-code');
        copybutton.innerText = 'copy';

        function copyingDone() {
            copybutton.innerText = 'copied!';
            setTimeout(() => {
                copybutton.innerText = 'copy';
            }, 2000);
        }

        copybutton.addEventListener('click', (cb) => {
            if ('clipboard' in navigator) {
                navigator.clipboard.writeText(codeblock.textContent);
                copyingDone();
                return;
            }

            const range = document.createRange();
            range.selectNodeContents(codeblock);
            const selection = window.getSelection();
            selection.removeAllRanges();
            selection.addRange(range);
            try {
                document.execCommand('copy');
                copyingDone();
            } catch (e) { };
            selection.removeRange(range);
        });

        if (container.classList.contains("highlight")) {
            container.appendChild(copybutton);
        } else if (container.parentNode.firstChild == container) {
            
        } else if (codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.nodeName == "TABLE") {
            
            codeblock.parentNode.parentNode.parentNode.parentNode.parentNode.appendChild(copybutton);
        } else {
            
            codeblock.parentNode.appendChild(copybutton);
        }
    });
</script>

<script type="text/javascript"
async
src="https://cdn.bootcss.com/mathjax/2.7.3/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
MathJax.Hub.Config({
tex2jax: {
inlineMath: [['$','$'], ['\\(','\\)']],
displayMath: [['$$','$$'], ['\[\[','\]\]']],
processEscapes: true,
processEnvironments: true,
skipTags: ['script', 'noscript', 'style', 'textarea', 'pre'],
TeX: { equationNumbers: { autoNumber: "AMS" },
extensions: ["AMSmath.js", "AMSsymbols.js"] }
}
});

MathJax.Hub.Queue(function() {



var all = MathJax.Hub.getAllJax(), i;
for(i = 0; i < all.length; i += 1) {
all[i].SourceElement().parentNode.className += ' has-jax';
}
});
</script>

<style>
code.has-jax {
font: inherit;
font-size: 100%;
background: inherit;
border: inherit;
color: #515151;
}
</style></body>
</html>
